{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((130, 11),\n",
       " Index(['type', 'Histone variant', 'Clustered (canonical) isoform',\n",
       "        'hgnc_gene_name', 'ncbi_gene_id', 'Ensembl gene ID',\n",
       "        'Transcript stable ID', 'RefSeq mRNA ID', 'RefSeq peptide ID',\n",
       "        'Protein length', 'references'],\n",
       "       dtype='object'))"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"https://raw.githubusercontent.com/intbio/hist_gene_prot/master/docs/human_histones.csv?token=GHSAT0AAAAAAB32GIFDIRWV3UE7YRNO4SBCY55W5JQ\",\n",
    "                 engine='python').fillna('')\n",
    "df.columns = ['type', 'Histone variant', 'Clustered (canonical) isoform',\n",
    "              'hgnc_gene_name', 'ncbi_gene_id', 'Ensembl gene ID',\n",
    "              'Transcript stable ID', 'RefSeq mRNA ID', 'RefSeq peptide ID',\n",
    "              'Protein length', 'references']\n",
    "df.shape, df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>Histone variant</th>\n",
       "      <th>Clustered (canonical) isoform</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>Ensembl gene ID</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>RefSeq mRNA ID</th>\n",
       "      <th>RefSeq peptide ID</th>\n",
       "      <th>Protein length</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NP_005309</th>\n",
       "      <td>NP_005309</td>\n",
       "      <td>H1</td>\n",
       "      <td></td>\n",
       "      <td>H1.0</td>\n",
       "      <td>H1.0</td>\n",
       "      <td></td>\n",
       "      <td>3005</td>\n",
       "      <td>H1-0</td>\n",
       "      <td>ENSG00000189060</td>\n",
       "      <td>ENST00000340857</td>\n",
       "      <td>NM_005318</td>\n",
       "      <td>NP_005309</td>\n",
       "      <td>194.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_005316</th>\n",
       "      <td>NP_005316</td>\n",
       "      <td>H1</td>\n",
       "      <td></td>\n",
       "      <td>H1.1</td>\n",
       "      <td>H1.1</td>\n",
       "      <td></td>\n",
       "      <td>3024</td>\n",
       "      <td>H1-1</td>\n",
       "      <td>ENSG00000124610</td>\n",
       "      <td>ENST00000244573</td>\n",
       "      <td>NM_005325</td>\n",
       "      <td>NP_005316</td>\n",
       "      <td>215.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_005310</th>\n",
       "      <td>NP_005310</td>\n",
       "      <td>H1</td>\n",
       "      <td></td>\n",
       "      <td>H1.2</td>\n",
       "      <td>H1.2</td>\n",
       "      <td></td>\n",
       "      <td>3006</td>\n",
       "      <td>H1-2</td>\n",
       "      <td>ENSG00000187837</td>\n",
       "      <td>ENST00000343677</td>\n",
       "      <td>NM_005319</td>\n",
       "      <td>NP_005310</td>\n",
       "      <td>213.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_005311</th>\n",
       "      <td>NP_005311</td>\n",
       "      <td>H1</td>\n",
       "      <td></td>\n",
       "      <td>H1.3</td>\n",
       "      <td>H1.3</td>\n",
       "      <td></td>\n",
       "      <td>3007</td>\n",
       "      <td>H1-3</td>\n",
       "      <td>ENSG00000124575</td>\n",
       "      <td>ENST00000244534</td>\n",
       "      <td>NM_005320</td>\n",
       "      <td>NP_005311</td>\n",
       "      <td>221.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_005312</th>\n",
       "      <td>NP_005312</td>\n",
       "      <td>H1</td>\n",
       "      <td></td>\n",
       "      <td>H1.4</td>\n",
       "      <td>H1.4</td>\n",
       "      <td></td>\n",
       "      <td>3008</td>\n",
       "      <td>H1-4</td>\n",
       "      <td>ENSG00000168298</td>\n",
       "      <td>ENST00000304218</td>\n",
       "      <td>NM_005321</td>\n",
       "      <td>NP_005312</td>\n",
       "      <td>219.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession type variant_group variant Histone variant  \\\n",
       "accession                                                         \n",
       "NP_005309  NP_005309   H1                  H1.0            H1.0   \n",
       "NP_005316  NP_005316   H1                  H1.1            H1.1   \n",
       "NP_005310  NP_005310   H1                  H1.2            H1.2   \n",
       "NP_005311  NP_005311   H1                  H1.3            H1.3   \n",
       "NP_005312  NP_005312   H1                  H1.4            H1.4   \n",
       "\n",
       "          Clustered (canonical) isoform  ncbi_gene_id hgnc_gene_name  \\\n",
       "accession                                                              \n",
       "NP_005309                                        3005           H1-0   \n",
       "NP_005316                                        3024           H1-1   \n",
       "NP_005310                                        3006           H1-2   \n",
       "NP_005311                                        3007           H1-3   \n",
       "NP_005312                                        3008           H1-4   \n",
       "\n",
       "           Ensembl gene ID Transcript stable ID RefSeq mRNA ID  \\\n",
       "accession                                                        \n",
       "NP_005309  ENSG00000189060      ENST00000340857      NM_005318   \n",
       "NP_005316  ENSG00000124610      ENST00000244573      NM_005325   \n",
       "NP_005310  ENSG00000187837      ENST00000343677      NM_005319   \n",
       "NP_005311  ENSG00000124575      ENST00000244534      NM_005320   \n",
       "NP_005312  ENSG00000168298      ENST00000304218      NM_005321   \n",
       "\n",
       "          RefSeq peptide ID Protein length taxonomy_id      organism  \\\n",
       "accession                                                              \n",
       "NP_005309         NP_005309          194.0        9606  Homo sapiens   \n",
       "NP_005316         NP_005316          215.0        9606  Homo sapiens   \n",
       "NP_005310         NP_005310          213.0        9606  Homo sapiens   \n",
       "NP_005311         NP_005311          221.0        9606  Homo sapiens   \n",
       "NP_005312         NP_005312          219.0        9606  Homo sapiens   \n",
       "\n",
       "             phylum     class taxonomy_group info references  \n",
       "accession                                                     \n",
       "NP_005309  Chordata  Mammalia       Mammalia        26689747  \n",
       "NP_005316  Chordata  Mammalia       Mammalia        26689747  \n",
       "NP_005310  Chordata  Mammalia       Mammalia        26689747  \n",
       "NP_005311  Chordata  Mammalia       Mammalia        26689747  \n",
       "NP_005312  Chordata  Mammalia       Mammalia        26689747  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['accession'] = df[\"RefSeq peptide ID\"]\n",
    "df['variant_group'] = ['']*df.shape[0]\n",
    "df['variant'] = df[\"Histone variant\"]\n",
    "df['taxonomy_id'] = ['9606']*df.shape[0]\n",
    "df['organism'] = ['Homo sapiens']*df.shape[0]\n",
    "df['phylum'] = ['Chordata']*df.shape[0]\n",
    "df['class'] = ['Mammalia']*df.shape[0]\n",
    "df['taxonomy_group'] = ['Mammalia']*df.shape[0]\n",
    "df['info'] = ['']*df.shape[0]\n",
    "df = df[['accession', 'type', 'variant_group', 'variant',\n",
    "         'Histone variant', 'Clustered (canonical) isoform',\n",
    "         'ncbi_gene_id', 'hgnc_gene_name',\n",
    "         'Ensembl gene ID', 'Transcript stable ID', 'RefSeq mRNA ID', 'RefSeq peptide ID', 'Protein length',\n",
    "         'taxonomy_id', 'organism', 'phylum', 'class', 'taxonomy_group', 'info', 'references']]\n",
    "df.index = df['accession']\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>Histone variant</th>\n",
       "      <th>Clustered (canonical) isoform</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>Ensembl gene ID</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>RefSeq mRNA ID</th>\n",
       "      <th>RefSeq peptide ID</th>\n",
       "      <th>Protein length</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NP_005309</th>\n",
       "      <td>NP_005309</td>\n",
       "      <td>H1</td>\n",
       "      <td></td>\n",
       "      <td>H1.0</td>\n",
       "      <td>H1.0</td>\n",
       "      <td></td>\n",
       "      <td>3005</td>\n",
       "      <td>H1-0</td>\n",
       "      <td>ENSG00000189060</td>\n",
       "      <td>ENST00000340857</td>\n",
       "      <td>NM_005318</td>\n",
       "      <td>NP_005309</td>\n",
       "      <td>194</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_005316</th>\n",
       "      <td>NP_005316</td>\n",
       "      <td>H1</td>\n",
       "      <td></td>\n",
       "      <td>H1.1</td>\n",
       "      <td>H1.1</td>\n",
       "      <td></td>\n",
       "      <td>3024</td>\n",
       "      <td>H1-1</td>\n",
       "      <td>ENSG00000124610</td>\n",
       "      <td>ENST00000244573</td>\n",
       "      <td>NM_005325</td>\n",
       "      <td>NP_005316</td>\n",
       "      <td>215</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_005310</th>\n",
       "      <td>NP_005310</td>\n",
       "      <td>H1</td>\n",
       "      <td></td>\n",
       "      <td>H1.2</td>\n",
       "      <td>H1.2</td>\n",
       "      <td></td>\n",
       "      <td>3006</td>\n",
       "      <td>H1-2</td>\n",
       "      <td>ENSG00000187837</td>\n",
       "      <td>ENST00000343677</td>\n",
       "      <td>NM_005319</td>\n",
       "      <td>NP_005310</td>\n",
       "      <td>213</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_005311</th>\n",
       "      <td>NP_005311</td>\n",
       "      <td>H1</td>\n",
       "      <td></td>\n",
       "      <td>H1.3</td>\n",
       "      <td>H1.3</td>\n",
       "      <td></td>\n",
       "      <td>3007</td>\n",
       "      <td>H1-3</td>\n",
       "      <td>ENSG00000124575</td>\n",
       "      <td>ENST00000244534</td>\n",
       "      <td>NM_005320</td>\n",
       "      <td>NP_005311</td>\n",
       "      <td>221</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_005312</th>\n",
       "      <td>NP_005312</td>\n",
       "      <td>H1</td>\n",
       "      <td></td>\n",
       "      <td>H1.4</td>\n",
       "      <td>H1.4</td>\n",
       "      <td></td>\n",
       "      <td>3008</td>\n",
       "      <td>H1-4</td>\n",
       "      <td>ENSG00000168298</td>\n",
       "      <td>ENST00000304218</td>\n",
       "      <td>NM_005321</td>\n",
       "      <td>NP_005312</td>\n",
       "      <td>219</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession type variant_group variant Histone variant  \\\n",
       "accession                                                         \n",
       "NP_005309  NP_005309   H1                  H1.0            H1.0   \n",
       "NP_005316  NP_005316   H1                  H1.1            H1.1   \n",
       "NP_005310  NP_005310   H1                  H1.2            H1.2   \n",
       "NP_005311  NP_005311   H1                  H1.3            H1.3   \n",
       "NP_005312  NP_005312   H1                  H1.4            H1.4   \n",
       "\n",
       "          Clustered (canonical) isoform  ncbi_gene_id hgnc_gene_name  \\\n",
       "accession                                                              \n",
       "NP_005309                                        3005           H1-0   \n",
       "NP_005316                                        3024           H1-1   \n",
       "NP_005310                                        3006           H1-2   \n",
       "NP_005311                                        3007           H1-3   \n",
       "NP_005312                                        3008           H1-4   \n",
       "\n",
       "           Ensembl gene ID Transcript stable ID RefSeq mRNA ID  \\\n",
       "accession                                                        \n",
       "NP_005309  ENSG00000189060      ENST00000340857      NM_005318   \n",
       "NP_005316  ENSG00000124610      ENST00000244573      NM_005325   \n",
       "NP_005310  ENSG00000187837      ENST00000343677      NM_005319   \n",
       "NP_005311  ENSG00000124575      ENST00000244534      NM_005320   \n",
       "NP_005312  ENSG00000168298      ENST00000304218      NM_005321   \n",
       "\n",
       "          RefSeq peptide ID Protein length taxonomy_id      organism  \\\n",
       "accession                                                              \n",
       "NP_005309         NP_005309            194        9606  Homo sapiens   \n",
       "NP_005316         NP_005316            215        9606  Homo sapiens   \n",
       "NP_005310         NP_005310            213        9606  Homo sapiens   \n",
       "NP_005311         NP_005311            221        9606  Homo sapiens   \n",
       "NP_005312         NP_005312            219        9606  Homo sapiens   \n",
       "\n",
       "             phylum     class taxonomy_group info references  \n",
       "accession                                                     \n",
       "NP_005309  Chordata  Mammalia       Mammalia        26689747  \n",
       "NP_005316  Chordata  Mammalia       Mammalia        26689747  \n",
       "NP_005310  Chordata  Mammalia       Mammalia        26689747  \n",
       "NP_005311  Chordata  Mammalia       Mammalia        26689747  \n",
       "NP_005312  Chordata  Mammalia       Mammalia        26689747  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Protein length'] = [int(l) if l != '' else '' for l in df['Protein length']]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_human = pd.read_csv(\"human_histones.csv\",engine='python').fillna('')\n",
    "# df_human.index = [acc.split('.')[0] for acc in df_human['accession']]\n",
    "# for accession, row in df_human.iterrows():\n",
    "#     df.at[accession, 'variant_group'] = row['variant_group']\n",
    "#     df.at[accession, 'variant'] = row['variant']\n",
    "#     df.at[accession, 'info'] = row['info']\n",
    "# df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(130, 20)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('human_histones.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-histdb_env]",
   "language": "python",
   "name": "conda-env-.conda-histdb_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
