{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "bde51c46-a7d0-4254-8492-29d84551ca32",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import io\n",
    "import json\n",
    "import os\n",
    "import shutil\n",
    "import subprocess\n",
    "import sys\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import Entrez, SearchIO, SeqIO\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio.Align.AlignInfo import SummaryInfo\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from ete3 import Tree\n",
    "from mysql.connector import connection\n",
    "from sshtunnel import SSHTunnelForwarder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "95503090-7bc9-4909-b81b-54d5db2b6edf",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open(\"db_curated_server_info.txt\", \"r\") as file:\n",
    "    lines = file.readlines()\n",
    "\n",
    "config = {}\n",
    "\n",
    "for line in lines:\n",
    "    line = line.strip()\n",
    "    if line and not line.startswith(\"#\"):\n",
    "        key, value = line.split(\"=\", 1)\n",
    "        config[key] = value.strip()\n",
    "\n",
    "server_name = config.get(\"server_name\")\n",
    "srever_port = int(config.get(\"srever_port\"))\n",
    "ssh_password = config.get(\"ssh_password\")\n",
    "ssh_username = config.get(\"ssh_username\")\n",
    "db_adress = config.get(\"db_adress\")\n",
    "db_port = int(config.get(\"db_port\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8deea4bb-e5aa-4363-92d7-616ccf86eb4a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44071\n"
     ]
    }
   ],
   "source": [
    "tunnel = SSHTunnelForwarder(\n",
    "    (server_name, srever_port),\n",
    "    ssh_password=ssh_password,\n",
    "    ssh_username=ssh_username,\n",
    "    remote_bind_address=(db_adress, db_port),\n",
    ")\n",
    "tunnel.start()\n",
    "print(tunnel.local_bind_port)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1f6e1cba-8d17-4bd2-8d1a-3a12e173dac6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "conn = connection.MySQLConnection(\n",
    "    user=\"db_user\",\n",
    "    password=\"db_password\",\n",
    "    host=\"localhost\",\n",
    "    port=tunnel.local_bind_port,\n",
    "    database=\"db_name\",\n",
    ")\n",
    "cursor = conn.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c8290ee5-a414-4e14-b961-efe14968776e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('alternative_name',),\n",
       " ('histone',),\n",
       " ('histone_description',),\n",
       " ('histone_has_publication',),\n",
       " ('publication',),\n",
       " ('sequence',),\n",
       " ('sequence_has_publication',)]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SHOW TABLES;\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d43c2a65-2e42-4e26-9d38-ec43e30cf591",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'id, level, taxonomic_span, taxonomic_span_id, description, parent'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()\n",
    "\", \".join([i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f59aa7bd-7ef7-4f02-81e5-7086702e4583",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()\n",
    "\", \".join([i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d187cd86-8a35-4145-b65a-65b6adf959a7",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Archaeal</td>\n",
       "      <td>type</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>H2A.Z.2_(Primates)</td>\n",
       "      <td>variant</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>H2A.Z.2.s2_(Primates)</td>\n",
       "      <td>variant</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>H2A.Z.2.s1_(Primates)</td>\n",
       "      <td>variant</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>H2A.Z.1_(Primates)</td>\n",
       "      <td>variant</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>H2A.Z</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>cH3</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>H2A.Q</td>\n",
       "      <td>variant</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>cH2B_(Chlorophyta)</td>\n",
       "      <td>variant</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>H2B.S</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>204 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        id          level  count\n",
       "0                 Archaeal           type      0\n",
       "141     H2A.Z.2_(Primates)        variant      0\n",
       "139  H2A.Z.2.s2_(Primates)        variant      0\n",
       "137  H2A.Z.2.s1_(Primates)        variant      0\n",
       "135     H2A.Z.1_(Primates)        variant      0\n",
       "..                     ...            ...    ...\n",
       "132                  H2A.Z  variant_group     20\n",
       "59                     cH3  variant_group     23\n",
       "120                  H2A.Q        variant     28\n",
       "50      cH2B_(Chlorophyta)        variant     35\n",
       "149                  H2B.S  variant_group     47\n",
       "\n",
       "[204 rows x 3 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT h.id, h.level, COUNT(accession) as count FROM histone h LEFT JOIN sequence s ON h.id = s.variant \"\n",
    "    \"GROUP BY h.id \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "count_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").sort_values([\"count\"])\n",
    "count_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "45ef0dd7-7887-46ef-a7d1-bcb825abf3ff",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Archaeal</td>\n",
       "      <td>type</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>H2A.Z.2_(Primates)</td>\n",
       "      <td>variant</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>H2A.Z.2.s2_(Primates)</td>\n",
       "      <td>variant</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>H2A.Z.2.s1_(Primates)</td>\n",
       "      <td>variant</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>H2A.Z.1_(Primates)</td>\n",
       "      <td>variant</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>cH2A_(Mus_musculus)</td>\n",
       "      <td>variant</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>cH2B_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>cH2B</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>cH2A_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>cH1</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        id          level  count\n",
       "0                 Archaeal           type      0\n",
       "141     H2A.Z.2_(Primates)        variant      0\n",
       "139  H2A.Z.2.s2_(Primates)        variant      0\n",
       "137  H2A.Z.2.s1_(Primates)        variant      0\n",
       "135     H2A.Z.1_(Primates)        variant      0\n",
       "..                     ...            ...    ...\n",
       "27     cH2A_(Mus_musculus)        variant      0\n",
       "53     cH2B_(Homo_sapiens)        variant      0\n",
       "31                    cH2B  variant_group      0\n",
       "25     cH2A_(Homo_sapiens)        variant      0\n",
       "7                      cH1  variant_group      0\n",
       "\n",
       "[67 rows x 3 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_df[count_df[\"count\"] == 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4146be8b-f619-4334-b010-7a862a8d2d2c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE variant='H2A.Z.2_(Primates)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "5b731573-21ec-4f74-ac03-59354d94adf5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_tree_dict(cursor, paretnt_id):\n",
    "    cursor.execute(f\"SELECT id FROM histone WHERE parent = '{paretnt_id}'\")\n",
    "    res = cursor.fetchall()\n",
    "    if len(res) < 1:\n",
    "        return \"null\"\n",
    "    return {v: get_tree_dict(cursor, v) for v, *_ in res}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "5fa4cb6f-349a-4a5b-b5b8-4f65b6524182",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def dict2tree(tree, d):\n",
    "    \"\"\"\n",
    "    converts tree from classification.json to a ete3 object\n",
    "    d is\n",
    "    with open('classification.json') as json_file:\n",
    "        data = json.load(json_file)\n",
    "    d=data['tree']\n",
    "    \"\"\"\n",
    "    for k, v in d.items():\n",
    "        CH = tree.add_child(name=k)\n",
    "        if isinstance(v, dict):\n",
    "            dict2tree(CH, v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "8e2bf14d-7d44-42e9-afe3-330295bcae67",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def muscle_aln(sequences, options=[], debug=False, sort=True):\n",
    "    muscle = os.path.join(os.path.dirname(sys.executable), \"muscle\")\n",
    "    process = subprocess.Popen(\n",
    "        [muscle] + options,\n",
    "        stdin=subprocess.PIPE,\n",
    "        stdout=subprocess.PIPE,\n",
    "        stderr=subprocess.PIPE,\n",
    "    )\n",
    "    aln, error = process.communicate(sequences.encode(\"utf-8\"))\n",
    "    if debug:\n",
    "        print(sequences)\n",
    "        print()\n",
    "        print(\"Stderr:\")\n",
    "        print(error.decode(\"utf-8\"))\n",
    "        print(\"Stdout:\")\n",
    "        print(aln.decode(\"utf-8\"))\n",
    "    seqFile = io.StringIO()\n",
    "    seqFile.write(aln.decode(\"utf-8\"))\n",
    "    seqFile.seek(0)\n",
    "    sequences_ids = [s.split(\" \", 1)[0] for s in sequences.split(\">\")]\n",
    "    sequences = list(\n",
    "        SeqIO.parse(seqFile, \"fasta\")\n",
    "    )  # Not in same order, but does it matter?\n",
    "    if sort:\n",
    "        sequences.sort(key=lambda x: sequences_ids.index(x.id))  # Yes, it matters\n",
    "    msa = MultipleSeqAlignment(sequences)\n",
    "    return msa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "d20dab81-4d92-447c-8a3e-7e8405c92234",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def muscle_p2p_aln(msa1, msa2, options=[], debug=False):\n",
    "    \"\"\"\n",
    "    align two alignments\n",
    "    :return: MultipleSeqAlignment object\n",
    "    \"\"\"\n",
    "    os.makedirs(\"tmp/\")  # create tmp dir to save msa profiles\n",
    "    try:\n",
    "        with open(\"tmp/one.afa\", \"w\") as f:\n",
    "            f.write(format(msa1, \"fasta\"))\n",
    "        with open(\"tmp/two.afa\", \"w\") as f:\n",
    "            f.write(format(msa2, \"fasta\"))\n",
    "\n",
    "        muscle = os.path.join(os.path.dirname(sys.executable), \"muscle\")\n",
    "        process = subprocess.Popen(\n",
    "            [muscle]\n",
    "            + options\n",
    "            + [\"-profile\", \"-in1\", \"tmp/one.afa\", \"-in2\", \"tmp/two.afa\"],\n",
    "            stdin=subprocess.PIPE,\n",
    "            stdout=subprocess.PIPE,\n",
    "            stderr=subprocess.PIPE,\n",
    "        )\n",
    "\n",
    "        aln, error = process.communicate()\n",
    "        if debug:\n",
    "            print(\"Stderr:\")\n",
    "            print(error.decode(\"utf-8\"))\n",
    "            print(\"Stdout:\")\n",
    "            print(aln.decode(\"utf-8\"))\n",
    "\n",
    "        seqFile = io.StringIO()\n",
    "        seqFile.write(aln.decode(\"utf-8\"))\n",
    "        seqFile.seek(0)\n",
    "        sequences = list(\n",
    "            SeqIO.parse(seqFile, \"fasta\")\n",
    "        )  # Not in same order, but does it matter?\n",
    "        msa = MultipleSeqAlignment(sequences)\n",
    "    except:\n",
    "        shutil.rmtree(\"tmp/\")\n",
    "        raise\n",
    "    shutil.rmtree(\"tmp/\")  # rm -rf tmp dir\n",
    "    return msa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "650ed4b1-60d9-456d-977d-125949071539",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_fasta_seq(data):\n",
    "    return \"\\n\".join(\n",
    "        [\n",
    "            SeqRecord(\n",
    "                Seq(row[\"sequence\"]),\n",
    "                id=f\"{row['organism'].split()[0]}|{row['accession']}|{row['variant']}\",\n",
    "                name=row[\"accession\"],\n",
    "                description=f\"organism={row['organism']} phylum={row['phylum']} class={row['class']}\",\n",
    "            ).format(\"fasta\")\n",
    "            for i, row in data.iterrows()\n",
    "        ]\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "bf521b06-4d78-4948-90a2-02df8a2c7804",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def generate_draft_seeds(hist_tree, sequence_data, save_directory):\n",
    "    # generate draft seeds - needs debugging\n",
    "    draft_seeds_msa = {}\n",
    "\n",
    "    # create directory if not exist or rewrite the directory\n",
    "    try:\n",
    "        os.makedirs(save_directory)\n",
    "    except FileExistsError:\n",
    "        # directory already exists\n",
    "        shutil.rmtree(save_directory)\n",
    "        os.makedirs(save_directory)\n",
    "\n",
    "    # hist_tree traversal\n",
    "    for node in hist_tree.traverse(\"postorder\"):\n",
    "        print(\"Processing \", node.name)\n",
    "        if node.is_root(): continue\n",
    "        if node.is_leaf():  # we get sequences for that variant and align them.\n",
    "            draft_seeds_msa[node.name] = muscle_aln(\n",
    "                get_fasta_seq(\n",
    "                    sequence_data.query(\n",
    "                        f'type==\"{node.name}\" | variant==\"{node.name}\"',\n",
    "                        engine=\"python\",\n",
    "                    )\n",
    "                )\n",
    "            )\n",
    "            print(node.name, \"Alignment length:\", len(draft_seeds_msa[node.name]))\n",
    "            with open(f\"{save_directory}/{node.name}.fasta\", \"w\") as f:\n",
    "                f.write(format(draft_seeds_msa[node.name], \"fasta\"))\n",
    "        elif not node.is_root():  # we will do profile to profile alignment\n",
    "            # we should first check if there are seqs with this subvariant as the most specific one\n",
    "            print(f\"\\t Node is internal, progressive alignment:\")\n",
    "            msa = muscle_aln(\n",
    "                get_fasta_seq(\n",
    "                    sequence_data.query(\n",
    "                        f'variant==\"{node.name}\"'\n",
    "                    )\n",
    "                )\n",
    "            )\n",
    "            draft_seeds_msa[node.name + \"_only\"] = msa\n",
    "            print(f\"\\t\\t For {node.name} aligned {len(msa)} sequences\")\n",
    "            # progressively align\n",
    "            for ch in node.get_children():\n",
    "                if len(msa) == 0:\n",
    "                    msa = draft_seeds_msa[ch.name]\n",
    "                    print(\n",
    "                        f\"\\t\\t Adding child {node.name} aligned {len(draft_seeds_msa[ch.name])} sequences\"\n",
    "                    )\n",
    "                    continue\n",
    "                elif len(draft_seeds_msa[ch.name]) != 0:\n",
    "                    msa = muscle_p2p_aln(msa, draft_seeds_msa[ch.name])\n",
    "                    print(\n",
    "                        f\"\\t\\t Adding child {node.name} aligned {len(draft_seeds_msa[ch.name])} sequences\"\n",
    "                    )\n",
    "                else:\n",
    "                    continue\n",
    "            draft_seeds_msa[node.name] = msa\n",
    "            print(node.name, \"Alignment length:\", len(draft_seeds_msa[node.name]))\n",
    "            with open(f\"{save_directory}/{node.name}.fasta\", \"w\") as f:\n",
    "                f.write(format(draft_seeds_msa[node.name], \"fasta\"))\n",
    "            with open(f\"{save_directory}/{node.name}_only.fasta\", \"w\") as f:\n",
    "                f.write(format(draft_seeds_msa[node.name + \"_only\"], \"fasta\"))\n",
    "        #         print(f\"\\t\\t Final for {node.name} aligned {len(draft_seeds_msa[node.name])} sequences\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "9699c9c2-d4d2-4132-8b45-88a8557712ed",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Archaeal': 'null',\n",
       " 'H1': {'cH1': 'null',\n",
       "  'generic_H1': 'null',\n",
       "  'H1.0': {'H1.0_(Homo_sapiens)': 'null'},\n",
       "  'H1.1': {'H1.1_(Homo_sapiens)': 'null'},\n",
       "  'H1.10': {'H1.10_(Homo_sapiens)': 'null'},\n",
       "  'H1.2': {'H1.2_(Homo_sapiens)': 'null'},\n",
       "  'H1.3': {'H1.3_(Homo_sapiens)': 'null'},\n",
       "  'H1.4': {'H1.4_(Homo_sapiens)': 'null'},\n",
       "  'H1.5': {'H1.5_(Homo_sapiens)': 'null'},\n",
       "  'OO_H1.8': {'H1.8_(Homo_sapiens)': 'null'},\n",
       "  'scH1': 'null',\n",
       "  'TS_H1.6': {'H1.6_(Homo_sapiens)': 'null'},\n",
       "  'TS_H1.7': {'H1.7_(Homo_sapiens)': 'null'},\n",
       "  'TS_H1.9': 'null'},\n",
       " 'H2A': {'cH2A': {'cH2A_(Animals)': {'cH2A_(Vertebrata)': {'cH2A_(Mammalia)': {'cH2A_(Homo_sapiens)': {'cH2A.10_(Homo_sapiens)': 'null',\n",
       "       'cH2A.11_(Homo_sapiens)': 'null',\n",
       "       'cH2A.1_(Homo_sapiens)': 'null',\n",
       "       'cH2A.2_(Homo_sapiens)': 'null',\n",
       "       'cH2A.3_(Homo_sapiens)': 'null',\n",
       "       'cH2A.4_(Homo_sapiens)': 'null',\n",
       "       'cH2A.5_(Homo_sapiens)': 'null',\n",
       "       'cH2A.6_(Homo_sapiens)': 'null',\n",
       "       'cH2A.7_(Homo_sapiens)': 'null',\n",
       "       'cH2A.8_(Homo_sapiens)': 'null',\n",
       "       'cH2A.9_(Homo_sapiens)': 'null'},\n",
       "      'cH2A_(Mus_musculus)': {'cH2A.1_(Mus_musculus)': 'null'}}}},\n",
       "   'cH2A_(Fungi)': 'null',\n",
       "   'cH2A_(Plants)': {'cH2A_(Chlorophyta)': 'null',\n",
       "    'cH2A_(Embryophyta)': 'null'},\n",
       "   'cH2A_(Protists)': 'null'},\n",
       "  'gH2A': 'null',\n",
       "  'H2A.J': {'H2A.J_(Homo_sapiens)': 'null'},\n",
       "  'H2A.M': 'null',\n",
       "  'H2A.R': 'null',\n",
       "  'H2A.W': 'null',\n",
       "  'H2A.X': {'H2A.X_(Animals)': {'H2A.X_(Vertebrata)': {'H2A.X_(Mammalia)': {'H2A.X_(Homo_sapiens)': 'null',\n",
       "      'H2A.X_(Mus_musculus)': 'null'}}},\n",
       "   'H2A.X_(Fungi)': 'null',\n",
       "   'H2A.X_(Plants)': 'null',\n",
       "   'H2A.X_(Protists)': 'null'},\n",
       "  'H2A.Z': {'H2A.Z.1_(Chordata)': {'H2A.Z.1_(Primates)': {'H2A.Z.1_(Homo_sapiens)': 'null'}},\n",
       "   'H2A.Z.2_(Chordata)': {'H2A.Z.2_(Primates)': {'H2A.Z.2.s1_(Primates)': {'H2A.Z.2.s1_(Homo_sapiens)': 'null'},\n",
       "     'H2A.Z.2.s2_(Primates)': {'H2A.Z.2.s2_(Homo_sapiens)': 'null'}}}},\n",
       "  'macroH2A': {'macroH2A.1_(Mammalia)': {'macroH2A.1.s1_(Mammalia)': {'macroH2A.1.s1_(Homo_sapiens)': 'null'},\n",
       "    'macroH2A.1.s2_(Mammalia)': {'macroH2A.1.s2_(Homo_sapiens)': 'null'}},\n",
       "   'macroH2A.2_(Mammalia)': {'macroH2A.2_(Homo_sapiens)': 'null'}},\n",
       "  'short_H2A': {'H2A.B': {'H2A.B_(Homo_sapiens)': {'H2A.B.1_(Homo_sapiens)': 'null',\n",
       "     'H2A.B.2_(Homo_sapiens)': 'null'},\n",
       "    'H2A.B_(Mus_musculus)': {'H2A.B.1_(Mus_musculus)': 'null',\n",
       "     'H2A.B.2_(Mus_musculus)': 'null',\n",
       "     'H2A.B.3_(Mus_musculus)': 'null'}},\n",
       "   'H2A.L': {'H2A.L_(Homo_sapiens)': {'H2A.L.1_(Homo_sapiens)': 'null',\n",
       "     'H2A.L.3_(Homo_sapiens)': 'null'},\n",
       "    'H2A.L_(Mus_musculus)': {'H2A.L.1_(Mus_musculus)': 'null',\n",
       "     'H2A.L.2_(Mus_musculus)': 'null',\n",
       "     'H2A.L.3_(Mus_musculus)': 'null'}},\n",
       "   'H2A.P': {'H2A.P_(Homo_sapiens)': 'null', 'H2A.P_(Mus_musculus)': 'null'},\n",
       "   'H2A.Q': 'null'}},\n",
       " 'H2B': {'cH2B': {'cH2B_(Animals)': {'cH2B_(Vertebrata)': {'cH2B_(Mammalia)': {'cH2B_(Homo_sapiens)': {'cH2B.10_(Homo_sapiens)': 'null',\n",
       "       'cH2B.11_(Homo_sapiens)': 'null',\n",
       "       'cH2B.12_(Homo_sapiens)': 'null',\n",
       "       'cH2B.13_(Homo_sapiens)': 'null',\n",
       "       'cH2B.14_(Homo_sapiens)': 'null',\n",
       "       'cH2B.15_(Homo_sapiens)': 'null',\n",
       "       'cH2B.1_(Homo_sapiens)': 'null',\n",
       "       'cH2B.2_(Homo_sapiens)': 'null',\n",
       "       'cH2B.3_(Homo_sapiens)': 'null',\n",
       "       'cH2B.4_(Homo_sapiens)': 'null',\n",
       "       'cH2B.5_(Homo_sapiens)': 'null',\n",
       "       'cH2B.6_(Homo_sapiens)': 'null',\n",
       "       'cH2B.7_(Homo_sapiens)': 'null',\n",
       "       'cH2B.8_(Homo_sapiens)': 'null',\n",
       "       'cH2B.9_(Homo_sapiens)': 'null'},\n",
       "      'cH2B_(Mus_musculus)': {'cH2B.1_(Mus_musculus)': 'null',\n",
       "       'cH2B.E_(Mus_musculus)': 'null'}}}},\n",
       "   'cH2B_(Fungi)': 'null',\n",
       "   'cH2B_(Plants)': {'cH2B_(Chlorophyta)': 'null',\n",
       "    'cH2B_(Embryophyta)': 'null'},\n",
       "   'cH2B_(Protists)': 'null'},\n",
       "  'CS_H2B_(Echinoidea)': 'null',\n",
       "  'early_H2B_(Echinoidea)': 'null',\n",
       "  'gH2B': 'null',\n",
       "  'H2B.K': {'H2B.K_(Homo_sapiens)': 'null'},\n",
       "  'H2B.L': 'null',\n",
       "  'H2B.N': {'H2B.N_(Homo_sapiens)': 'null'},\n",
       "  'H2B.O': 'null',\n",
       "  'H2B.S': 'null',\n",
       "  'H2B.V': 'null',\n",
       "  'H2B.W': {'H2B.W_(Homo_sapiens)': {'H2B.W.1_(Homo_sapiens)': 'null',\n",
       "    'H2B.W.2_(Homo_sapiens)': 'null'}},\n",
       "  'H2B.Z': 'null',\n",
       "  'late_H2B_(Echinoidea)': 'null',\n",
       "  'sperm_H2B_(Echinoidea)': 'null'},\n",
       " 'H3': {'cenH3_(Eukarya)': {'cenH3_(Animals)': {'cenH3_(Mammalia)': {'cenH3_(Homo_sapiens)': 'null'}},\n",
       "   'cenH3_(Fungi)': 'null',\n",
       "   'cenH3_(Plants)': 'null'},\n",
       "  'cH3': {'cH3_(Animals)': {'cH3_(Vertebrata)': {'cH3_(Mammalia)': {'cH3.1_(Mammalia)': {'cH3.1_(Homo_sapiens)': 'null'},\n",
       "      'cH3.2_(Mammalia)': {'cH3.2_(Homo_sapiens)': 'null'}}}},\n",
       "   'cH3_(Fungi)': 'null',\n",
       "   'cH3_(Plants)': {'cH3_(Chlorophyta)': 'null', 'cH3_(Embryophyta)': 'null'},\n",
       "   'cH3_(Protists)': 'null'},\n",
       "  'H3.1-like_(Plants)': 'null',\n",
       "  'H3.3': {'H3.3_(Animals)': {'H3.3_(Homo_sapiens)': 'null'},\n",
       "   'H3.3_(Fungi)': {'H3.3_(Ascomycota)': 'null'},\n",
       "   'H3.3_(Plants)': 'null',\n",
       "   'H3.3_(Protists)': 'null'},\n",
       "  'H3.3-like?': {'H3.3-like_(Animals)': {'H3.X_(Primates?)': {'H3.X_(Homo_sapiens)': 'null'},\n",
       "    'H3.Y_(Primates?)': {'H3.Y_(Homo_sapiens)': {'H3.Y.1_(Homo_sapiens)': 'null',\n",
       "      'H3.Y.2_(Homo_sapiens)': 'null'}}},\n",
       "   'H3.3-like_(Plants)': {'H3_(Lilly???)': 'null', 'TS H3.10': 'null'}},\n",
       "  'H3.4_(Mammalia)': {'H3.4_(Homo_sapiens)': 'null'},\n",
       "  'H3.5_(Hominidae)': {'H3.5_(Homo_sapiens)': 'null'},\n",
       "  'H3.6_(Mammals?)?': 'null',\n",
       "  'H3.7_(Mammals?)?': 'null',\n",
       "  'H3.8_(Mammals?)?': 'null',\n",
       "  'H3.B_(Giardia?)': 'null',\n",
       "  'H3.P_(Moneuplotes?)': 'null',\n",
       "  'H3.V_(Trypanosomatidae)': 'null'},\n",
       " 'H4': {'cH4': {'cH4_(Homo_sapiens)': 'null'}},\n",
       " 'Viral': 'null'}"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT id FROM histone WHERE parent IS NULL\"\n",
    "cursor.execute(query)\n",
    "types = cursor.fetchall()\n",
    "variants_tree = {}\n",
    "for t, *_ in types:\n",
    "    variants_tree[t] = get_tree_dict(cursor, t)\n",
    "variants_tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "36cb0bf3-69bf-44f5-8ad4-0c15a62023ca",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "   /-Archaeal\n",
      "  |\n",
      "  |   /-cH1\n",
      "  |  |\n",
      "  |  |--generic_H1\n",
      "  |  |\n",
      "  |  |-H1.0-H1.0_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-H1.1-H1.1_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-H1.10-H1.10_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-H1.2-H1.2_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-H1.3-H1.3_(Homo_sapiens)\n",
      "  |-H1\n",
      "  |  |-H1.4-H1.4_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-H1.5-H1.5_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-OO_H1.8-H1.8_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |--scH1\n",
      "  |  |\n",
      "  |  |-TS_H1.6-H1.6_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-TS_H1.7-H1.7_(Homo_sapiens)\n",
      "  |  |\n",
      "  |   \\-TS_H1.9\n",
      "  |\n",
      "  |                                                                        /-cH2A.10_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2A.11_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2A.1_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2A.2_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2A.3_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                     /cH2A_(Homo_sapiens)-cH2A.4_(Homo_sapiens)\n",
      "  |                                                    |                  |\n",
      "  |                                                    |                  |--cH2A.5_(Homo_sapiens)\n",
      "  |                                                    |                  |\n",
      "  |                                                    |                  |--cH2A.6_(Homo_sapiens)\n",
      "  |                                                    |                  |\n",
      "  |       /cH2A_(Animals)cH2A_(Vertebrata)cH2A_(Mammalia)                 |--cH2A.7_(Homo_sapiens)\n",
      "  |      |                                             |                  |\n",
      "  |      |                                             |                  |--cH2A.8_(Homo_sapiens)\n",
      "  |      |                                             |                  |\n",
      "  |      |                                             |                   \\-cH2A.9_(Homo_sapiens)\n",
      "  |      |                                             |\n",
      "  |      |                                              \\cH2A_(Mus_musculus)-cH2A.1_(Mus_musculus)\n",
      "  |   /cH2A\n",
      "  |  |   |--cH2A_(Fungi)\n",
      "  |  |   |\n",
      "  |  |   |             /-cH2A_(Chlorophyta)\n",
      "  |  |   |-cH2A_(Plants)\n",
      "  |  |   |             \\-cH2A_(Embryophyta)\n",
      "  |  |   |\n",
      "  |  |    \\-cH2A_(Protists)\n",
      "  |  |\n",
      "  |  |--gH2A\n",
      "  |  |\n",
      "  |  |-H2A.J-H2A.J_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |--H2A.M\n",
      "  |  |\n",
      "  |  |--H2A.R\n",
      "  |  |\n",
      "  |  |--H2A.W\n",
      "  |  |\n",
      "  |  |                                                      /-H2A.X_(Homo_sapiens)\n",
      "  |  |     /H2A.X_(Animals)H2A.X_(Vertebrata)H2A.X_(Mammalia)\n",
      "  |  |    |                                                 \\-H2A.X_(Mus_musculus)\n",
      "  |  |    |\n",
      "  |  |-H2A.X-H2A.X_(Fungi)\n",
      "  |  |    |\n",
      "  |  |    |--H2A.X_(Plants)\n",
      "  |  |    |\n",
      "  |-H2A    \\-H2A.X_(Protists)\n",
      "  |  |\n",
      "  |  |     /H2A.Z.1_(Chordata)H2A.Z.1_(Primates)-H2A.Z.1_(Homo_sapiens)\n",
      "  |  |-H2A.Z\n",
      "  |  |    |                                    /H2A.Z.2.s1_(Primates)-H2A.Z.2.s1_(Homo_sapiens)\n",
      "  |  |     \\H2A.Z.2_(Chordata)H2A.Z.2_(Primates)\n",
      "  |  |                                         \\H2A.Z.2.s2_(Primates)-H2A.Z.2.s2_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |                             /macroH2A.1.s1_(Mammalia)-macroH2A.1.s1_(Homo_sapiens)\n",
      "  |  |        /macroH2A.1_(Mammalia)\n",
      "  |  |-macroH2A                    \\macroH2A.1.s2_(Mammalia)-macroH2A.1.s2_(Homo_sapiens)\n",
      "  |  |       |\n",
      "  |  |        \\macroH2A.2_(Mammalia)-macroH2A.2_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |                                  /-H2A.B.1_(Homo_sapiens)\n",
      "  |  |              /H2A.B_(Homo_sapiens)\n",
      "  |  |             |                    \\-H2A.B.2_(Homo_sapiens)\n",
      "  |  |         /H2A.B\n",
      "  |  |        |    |                    /-H2A.B.1_(Mus_musculus)\n",
      "  |  |        |    |                   |\n",
      "  |  |        |     \\H2A.B_(Mus_musculus)-H2A.B.2_(Mus_musculus)\n",
      "  |  |        |                        |\n",
      "  |  |        |                         \\-H2A.B.3_(Mus_musculus)\n",
      "  |  |        |\n",
      "  |  |        |                         /-H2A.L.1_(Homo_sapiens)\n",
      "  |  |        |     /H2A.L_(Homo_sapiens)\n",
      "  |  |        |    |                    \\-H2A.L.3_(Homo_sapiens)\n",
      "  |   \\short_H2AH2A.L\n",
      "  |           |    |                    /-H2A.L.1_(Mus_musculus)\n",
      "  |           |    |                   |\n",
      "  |           |     \\H2A.L_(Mus_musculus)-H2A.L.2_(Mus_musculus)\n",
      "  |           |                        |\n",
      "  |           |                         \\-H2A.L.3_(Mus_musculus)\n",
      "  |           |\n",
      "  |           |     /-H2A.P_(Homo_sapiens)\n",
      "  |           |-H2A.P\n",
      "  |           |     \\-H2A.P_(Mus_musculus)\n",
      "  |           |\n",
      "  |            \\-H2A.Q\n",
      "  |\n",
      "  |                                                                        /-cH2B.10_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2B.11_(Homo_sapiens)\n",
      "--|                                                                       |\n",
      "  |                                                                       |--cH2B.12_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2B.13_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2B.14_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2B.15_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2B.1_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                     /cH2B_(Homo_sapiens)-cH2B.2_(Homo_sapiens)\n",
      "  |                                                    |                  |\n",
      "  |                                                    |                  |--cH2B.3_(Homo_sapiens)\n",
      "  |                                                    |                  |\n",
      "  |                                                    |                  |--cH2B.4_(Homo_sapiens)\n",
      "  |                                                    |                  |\n",
      "  |                                                    |                  |--cH2B.5_(Homo_sapiens)\n",
      "  |                                                    |                  |\n",
      "  |       /cH2B_(Animals)cH2B_(Vertebrata)cH2B_(Mammalia)                 |--cH2B.6_(Homo_sapiens)\n",
      "  |      |                                             |                  |\n",
      "  |      |                                             |                  |--cH2B.7_(Homo_sapiens)\n",
      "  |      |                                             |                  |\n",
      "  |      |                                             |                  |--cH2B.8_(Homo_sapiens)\n",
      "  |      |                                             |                  |\n",
      "  |      |                                             |                   \\-cH2B.9_(Homo_sapiens)\n",
      "  |      |                                             |\n",
      "  |      |                                             |                   /-cH2B.1_(Mus_musculus)\n",
      "  |   /cH2B                                             \\cH2B_(Mus_musculus)\n",
      "  |  |   |                                                                 \\-cH2B.E_(Mus_musculus)\n",
      "  |  |   |\n",
      "  |  |   |--cH2B_(Fungi)\n",
      "  |  |   |\n",
      "  |  |   |             /-cH2B_(Chlorophyta)\n",
      "  |  |   |-cH2B_(Plants)\n",
      "  |  |   |             \\-cH2B_(Embryophyta)\n",
      "  |  |   |\n",
      "  |  |    \\-cH2B_(Protists)\n",
      "  |  |\n",
      "  |  |--CS_H2B_(Echinoidea)\n",
      "  |  |\n",
      "  |  |--early_H2B_(Echinoidea)\n",
      "  |  |\n",
      "  |  |--gH2B\n",
      "  |  |\n",
      "  |  |-H2B.K-H2B.K_(Homo_sapiens)\n",
      "  |-H2B\n",
      "  |  |--H2B.L\n",
      "  |  |\n",
      "  |  |-H2B.N-H2B.N_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |--H2B.O\n",
      "  |  |\n",
      "  |  |--H2B.S\n",
      "  |  |\n",
      "  |  |--H2B.V\n",
      "  |  |\n",
      "  |  |                         /-H2B.W.1_(Homo_sapiens)\n",
      "  |  |-H2B.WH2B.W_(Homo_sapiens)\n",
      "  |  |                         \\-H2B.W.2_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |--H2B.Z\n",
      "  |  |\n",
      "  |  |--late_H2B_(Echinoidea)\n",
      "  |  |\n",
      "  |   \\-sperm_H2B_(Echinoidea)\n",
      "  |\n",
      "  |                  /cenH3_(Animals)cenH3_(Mammalia)-cenH3_(Homo_sapiens)\n",
      "  |                 |\n",
      "  |   /cenH3_(Eukarya)-cenH3_(Fungi)\n",
      "  |  |              |\n",
      "  |  |               \\-cenH3_(Plants)\n",
      "  |  |\n",
      "  |  |                                              /cH3.1_(Mammalia)-cH3.1_(Homo_sapiens)\n",
      "  |  |   /cH3_(Animals)cH3_(Vertebrata)cH3_(Mammalia)\n",
      "  |  |  |                                           \\cH3.2_(Mammalia)-cH3.2_(Homo_sapiens)\n",
      "  |  |  |\n",
      "  |  |  |--cH3_(Fungi)\n",
      "  |  |-cH3\n",
      "  |  |  |            /-cH3_(Chlorophyta)\n",
      "  |  |  |-cH3_(Plants)\n",
      "  |  |  |            \\-cH3_(Embryophyta)\n",
      "  |  |  |\n",
      "  |  |   \\-cH3_(Protists)\n",
      "  |  |\n",
      "  |  |--H3.1-like_(Plants)\n",
      "  |  |\n",
      "  |  |    /H3.3_(Animals)-H3.3_(Homo_sapiens)\n",
      "  |  |   |\n",
      "  |  |   |-H3.3_(Fungi)-H3.3_(Ascomycota)\n",
      "  |  |-H3.3\n",
      "  |  |   |--H3.3_(Plants)\n",
      "  |  |   |\n",
      "  |  |    \\-H3.3_(Protists)\n",
      "  |-H3\n",
      "  |  |                             /H3.X_(Primates?)-H3.X_(Homo_sapiens)\n",
      "  |  |          /H3.3-like_(Animals)\n",
      "  |  |         |                  |                                   /-H3.Y.1_(Homo_sapiens)\n",
      "  |  |         |                   \\H3.Y_(Primates?)H3.Y_(Homo_sapiens)\n",
      "  |  |-H3.3-like?                                                     \\-H3.Y.2_(Homo_sapiens)\n",
      "  |  |         |\n",
      "  |  |         |                  /-H3_(Lilly???)\n",
      "  |  |          \\H3.3-like_(Plants)\n",
      "  |  |                            \\-TS H3.10\n",
      "  |  |\n",
      "  |  |-H3.4_(Mammalia)-H3.4_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-H3.5_(Hominidae)-H3.5_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |--H3.6_(Mammals?)?\n",
      "  |  |\n",
      "  |  |--H3.7_(Mammals?)?\n",
      "  |  |\n",
      "  |  |--H3.8_(Mammals?)?\n",
      "  |  |\n",
      "  |  |--H3.B_(Giardia?)\n",
      "  |  |\n",
      "  |  |--H3.P_(Moneuplotes?)\n",
      "  |  |\n",
      "  |   \\-H3.V_(Trypanosomatidae)\n",
      "  |\n",
      "  |-H4/cH4-cH4_(Homo_sapiens)\n",
      "  |\n",
      "   \\-Viral\n"
     ]
    }
   ],
   "source": [
    "hist_tree = Tree()\n",
    "dict2tree(hist_tree, variants_tree)\n",
    "print(hist_tree.get_ascii(show_internal=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "2e65711f-6265-4f31-a614-1d9b797001cd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sequence_df = pd.read_csv(\n",
    "    \"/home/l_singh/_scratch/hdb/project_dir/histonedb/CURATED_SET/histones.csv\"\n",
    ").fillna(\"\")\n",
    "sequence_df.index = list(sequence_df[\"accession\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "c572e59d-256a-4d17-94c3-a0b51c2c44f5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, type, variant_group, variant, doublet, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, references, sequence]\n",
       "Index: []"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df[sequence_df[\"variant\"].str.contains(\"H3_(Lilly\\?\\?\\?)\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "d0766ba9-4966-40de-839f-b03196df7610",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "ename": "error",
     "evalue": "multiple repeat at position 11",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31merror\u001b[0m                                     Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[75], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43msequence_df\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mquery\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m      2\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mvariant.str.contains(\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mH3_(Lilly???)\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m)\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      3\u001b[0m \u001b[43m    \u001b[49m\u001b[43mengine\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpython\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      4\u001b[0m \u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.conda/envs/histdb_env/lib/python3.8/site-packages/pandas/core/frame.py:4440\u001b[0m, in \u001b[0;36mDataFrame.query\u001b[0;34m(self, expr, inplace, **kwargs)\u001b[0m\n\u001b[1;32m   4438\u001b[0m kwargs[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlevel\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwargs\u001b[38;5;241m.\u001b[39mpop(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlevel\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;241m0\u001b[39m) \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m   4439\u001b[0m kwargs[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtarget\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 4440\u001b[0m res \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43meval\u001b[49m\u001b[43m(\u001b[49m\u001b[43mexpr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4442\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   4443\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mloc[res]\n",
      "File \u001b[0;32m~/.conda/envs/histdb_env/lib/python3.8/site-packages/pandas/core/frame.py:4566\u001b[0m, in \u001b[0;36mDataFrame.eval\u001b[0;34m(self, expr, inplace, **kwargs)\u001b[0m\n\u001b[1;32m   4563\u001b[0m     kwargs[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtarget\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\n\u001b[1;32m   4564\u001b[0m kwargs[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mresolvers\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mtuple\u001b[39m(kwargs\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mresolvers\u001b[39m\u001b[38;5;124m\"\u001b[39m, ())) \u001b[38;5;241m+\u001b[39m resolvers\n\u001b[0;32m-> 4566\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_eval\u001b[49m\u001b[43m(\u001b[49m\u001b[43mexpr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minplace\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minplace\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.conda/envs/histdb_env/lib/python3.8/site-packages/pandas/core/computation/eval.py:336\u001b[0m, in \u001b[0;36meval\u001b[0;34m(expr, parser, engine, local_dict, global_dict, resolvers, level, target, inplace)\u001b[0m\n\u001b[1;32m    327\u001b[0m \u001b[38;5;66;03m# get our (possibly passed-in) scope\u001b[39;00m\n\u001b[1;32m    328\u001b[0m env \u001b[38;5;241m=\u001b[39m ensure_scope(\n\u001b[1;32m    329\u001b[0m     level \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m,\n\u001b[1;32m    330\u001b[0m     global_dict\u001b[38;5;241m=\u001b[39mglobal_dict,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    333\u001b[0m     target\u001b[38;5;241m=\u001b[39mtarget,\n\u001b[1;32m    334\u001b[0m )\n\u001b[0;32m--> 336\u001b[0m parsed_expr \u001b[38;5;241m=\u001b[39m \u001b[43mExpr\u001b[49m\u001b[43m(\u001b[49m\u001b[43mexpr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mengine\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparser\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparser\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43menv\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43menv\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    338\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m engine \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnumexpr\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m (\n\u001b[1;32m    339\u001b[0m     is_extension_array_dtype(parsed_expr\u001b[38;5;241m.\u001b[39mterms\u001b[38;5;241m.\u001b[39mreturn_type)\n\u001b[1;32m    340\u001b[0m     \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(parsed_expr\u001b[38;5;241m.\u001b[39mterms, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124moperand_types\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m) \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    344\u001b[0m     )\n\u001b[1;32m    345\u001b[0m ):\n\u001b[1;32m    346\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m    347\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mEngine has switched to \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpython\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m because numexpr does not support \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    348\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mextension array dtypes. Please set your engine to python manually.\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    349\u001b[0m         \u001b[38;5;167;01mRuntimeWarning\u001b[39;00m,\n\u001b[1;32m    350\u001b[0m         stacklevel\u001b[38;5;241m=\u001b[39mfind_stack_level(),\n\u001b[1;32m    351\u001b[0m     )\n",
      "File \u001b[0;32m~/.conda/envs/histdb_env/lib/python3.8/site-packages/pandas/core/computation/expr.py:809\u001b[0m, in \u001b[0;36mExpr.__init__\u001b[0;34m(self, expr, engine, parser, env, level)\u001b[0m\n\u001b[1;32m    807\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mparser \u001b[38;5;241m=\u001b[39m parser\n\u001b[1;32m    808\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_visitor \u001b[38;5;241m=\u001b[39m PARSERS[parser](\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39menv, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mengine, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mparser)\n\u001b[0;32m--> 809\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mterms \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mparse\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.conda/envs/histdb_env/lib/python3.8/site-packages/pandas/core/computation/expr.py:828\u001b[0m, in \u001b[0;36mExpr.parse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    824\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mparse\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m    825\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    826\u001b[0m \u001b[38;5;124;03m    Parse an expression.\u001b[39;00m\n\u001b[1;32m    827\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 828\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_visitor\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvisit\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexpr\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.conda/envs/histdb_env/lib/python3.8/site-packages/pandas/core/computation/expr.py:415\u001b[0m, in \u001b[0;36mBaseExprVisitor.visit\u001b[0;34m(self, node, **kwargs)\u001b[0m\n\u001b[1;32m    413\u001b[0m method \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvisit_\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(node)\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    414\u001b[0m visitor \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(\u001b[38;5;28mself\u001b[39m, method)\n\u001b[0;32m--> 415\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mvisitor\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnode\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.conda/envs/histdb_env/lib/python3.8/site-packages/pandas/core/computation/expr.py:421\u001b[0m, in \u001b[0;36mBaseExprVisitor.visit_Module\u001b[0;34m(self, node, **kwargs)\u001b[0m\n\u001b[1;32m    419\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mSyntaxError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124monly a single expression is allowed\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    420\u001b[0m expr \u001b[38;5;241m=\u001b[39m node\u001b[38;5;241m.\u001b[39mbody[\u001b[38;5;241m0\u001b[39m]\n\u001b[0;32m--> 421\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvisit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mexpr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.conda/envs/histdb_env/lib/python3.8/site-packages/pandas/core/computation/expr.py:415\u001b[0m, in \u001b[0;36mBaseExprVisitor.visit\u001b[0;34m(self, node, **kwargs)\u001b[0m\n\u001b[1;32m    413\u001b[0m method \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvisit_\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(node)\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    414\u001b[0m visitor \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(\u001b[38;5;28mself\u001b[39m, method)\n\u001b[0;32m--> 415\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mvisitor\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnode\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.conda/envs/histdb_env/lib/python3.8/site-packages/pandas/core/computation/expr.py:424\u001b[0m, in \u001b[0;36mBaseExprVisitor.visit_Expr\u001b[0;34m(self, node, **kwargs)\u001b[0m\n\u001b[1;32m    423\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mvisit_Expr\u001b[39m(\u001b[38;5;28mself\u001b[39m, node, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[0;32m--> 424\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvisit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnode\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalue\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.conda/envs/histdb_env/lib/python3.8/site-packages/pandas/core/computation/expr.py:415\u001b[0m, in \u001b[0;36mBaseExprVisitor.visit\u001b[0;34m(self, node, **kwargs)\u001b[0m\n\u001b[1;32m    413\u001b[0m method \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvisit_\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(node)\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    414\u001b[0m visitor \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(\u001b[38;5;28mself\u001b[39m, method)\n\u001b[0;32m--> 415\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mvisitor\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnode\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.conda/envs/histdb_env/lib/python3.8/site-packages/pandas/core/computation/expr.py:705\u001b[0m, in \u001b[0;36mBaseExprVisitor.visit_Call\u001b[0;34m(self, node, side, **kwargs)\u001b[0m\n\u001b[1;32m    702\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m key\u001b[38;5;241m.\u001b[39marg:\n\u001b[1;32m    703\u001b[0m         kwargs[key\u001b[38;5;241m.\u001b[39marg] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvisit(key\u001b[38;5;241m.\u001b[39mvalue)(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39menv)\n\u001b[0;32m--> 705\u001b[0m name \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39menv\u001b[38;5;241m.\u001b[39madd_tmp(\u001b[43mres\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mnew_args\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[1;32m    706\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mterm_type(name\u001b[38;5;241m=\u001b[39mname, env\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39menv)\n",
      "File \u001b[0;32m~/.conda/envs/histdb_env/lib/python3.8/site-packages/pandas/core/strings/accessor.py:129\u001b[0m, in \u001b[0;36mforbid_nonstring_types.<locals>._forbid_nonstring_types.<locals>.wrapper\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    124\u001b[0m     msg \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m    125\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot use .str.\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfunc_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m with values of \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    126\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124minferred dtype \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_inferred_dtype\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    127\u001b[0m     )\n\u001b[1;32m    128\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(msg)\n\u001b[0;32m--> 129\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.conda/envs/histdb_env/lib/python3.8/site-packages/pandas/core/strings/accessor.py:1281\u001b[0m, in \u001b[0;36mStringMethods.contains\u001b[0;34m(self, pat, case, flags, na, regex)\u001b[0m\n\u001b[1;32m   1154\u001b[0m \u001b[38;5;129m@forbid_nonstring_types\u001b[39m([\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbytes\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n\u001b[1;32m   1155\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcontains\u001b[39m(\n\u001b[1;32m   1156\u001b[0m     \u001b[38;5;28mself\u001b[39m, pat, case: \u001b[38;5;28mbool\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m, flags: \u001b[38;5;28mint\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m, na\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, regex: \u001b[38;5;28mbool\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m   1157\u001b[0m ):\n\u001b[1;32m   1158\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   1159\u001b[0m \u001b[38;5;124;03m    Test if pattern or regex is contained within a string of a Series or Index.\u001b[39;00m\n\u001b[1;32m   1160\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1279\u001b[0m \u001b[38;5;124;03m    dtype: bool\u001b[39;00m\n\u001b[1;32m   1280\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 1281\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m regex \u001b[38;5;129;01mand\u001b[39;00m \u001b[43mre\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcompile\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpat\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mgroups:\n\u001b[1;32m   1282\u001b[0m         warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m   1283\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mThis pattern is interpreted as a regular expression, and has \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1284\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmatch groups. To actually get the groups, use str.extract.\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m   1285\u001b[0m             \u001b[38;5;167;01mUserWarning\u001b[39;00m,\n\u001b[1;32m   1286\u001b[0m             stacklevel\u001b[38;5;241m=\u001b[39mfind_stack_level(),\n\u001b[1;32m   1287\u001b[0m         )\n\u001b[1;32m   1289\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_data\u001b[38;5;241m.\u001b[39marray\u001b[38;5;241m.\u001b[39m_str_contains(pat, case, flags, na, regex)\n",
      "File \u001b[0;32m~/.conda/envs/histdb_env/lib/python3.8/re.py:250\u001b[0m, in \u001b[0;36mcompile\u001b[0;34m(pattern, flags)\u001b[0m\n\u001b[1;32m    248\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcompile\u001b[39m(pattern, flags\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m):\n\u001b[1;32m    249\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCompile a regular expression pattern, returning a Pattern object.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m--> 250\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_compile\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpattern\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mflags\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.conda/envs/histdb_env/lib/python3.8/re.py:302\u001b[0m, in \u001b[0;36m_compile\u001b[0;34m(pattern, flags)\u001b[0m\n\u001b[1;32m    300\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m sre_compile\u001b[38;5;241m.\u001b[39misstring(pattern):\n\u001b[1;32m    301\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfirst argument must be string or compiled pattern\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m--> 302\u001b[0m p \u001b[38;5;241m=\u001b[39m \u001b[43msre_compile\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcompile\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpattern\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mflags\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    303\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (flags \u001b[38;5;241m&\u001b[39m DEBUG):\n\u001b[1;32m    304\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(_cache) \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m _MAXCACHE:\n\u001b[1;32m    305\u001b[0m         \u001b[38;5;66;03m# Drop the oldest item\u001b[39;00m\n",
      "File \u001b[0;32m~/.conda/envs/histdb_env/lib/python3.8/sre_compile.py:764\u001b[0m, in \u001b[0;36mcompile\u001b[0;34m(p, flags)\u001b[0m\n\u001b[1;32m    762\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m isstring(p):\n\u001b[1;32m    763\u001b[0m     pattern \u001b[38;5;241m=\u001b[39m p\n\u001b[0;32m--> 764\u001b[0m     p \u001b[38;5;241m=\u001b[39m \u001b[43msre_parse\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mparse\u001b[49m\u001b[43m(\u001b[49m\u001b[43mp\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mflags\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    765\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    766\u001b[0m     pattern \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/.conda/envs/histdb_env/lib/python3.8/sre_parse.py:948\u001b[0m, in \u001b[0;36mparse\u001b[0;34m(str, flags, state)\u001b[0m\n\u001b[1;32m    945\u001b[0m state\u001b[38;5;241m.\u001b[39mstr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m\n\u001b[1;32m    947\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 948\u001b[0m     p \u001b[38;5;241m=\u001b[39m \u001b[43m_parse_sub\u001b[49m\u001b[43m(\u001b[49m\u001b[43msource\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstate\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mflags\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m&\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mSRE_FLAG_VERBOSE\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    949\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m Verbose:\n\u001b[1;32m    950\u001b[0m     \u001b[38;5;66;03m# the VERBOSE flag was switched on inside the pattern.  to be\u001b[39;00m\n\u001b[1;32m    951\u001b[0m     \u001b[38;5;66;03m# on the safe side, we'll parse the whole thing again...\u001b[39;00m\n\u001b[1;32m    952\u001b[0m     state \u001b[38;5;241m=\u001b[39m State()\n",
      "File \u001b[0;32m~/.conda/envs/histdb_env/lib/python3.8/sre_parse.py:443\u001b[0m, in \u001b[0;36m_parse_sub\u001b[0;34m(source, state, verbose, nested)\u001b[0m\n\u001b[1;32m    441\u001b[0m start \u001b[38;5;241m=\u001b[39m source\u001b[38;5;241m.\u001b[39mtell()\n\u001b[1;32m    442\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[0;32m--> 443\u001b[0m     itemsappend(\u001b[43m_parse\u001b[49m\u001b[43m(\u001b[49m\u001b[43msource\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstate\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mverbose\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnested\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    444\u001b[0m \u001b[43m                       \u001b[49m\u001b[38;5;129;43;01mnot\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mnested\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mand\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mnot\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mitems\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[1;32m    445\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m sourcematch(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m|\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m    446\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[0;32m~/.conda/envs/histdb_env/lib/python3.8/sre_parse.py:834\u001b[0m, in \u001b[0;36m_parse\u001b[0;34m(source, state, verbose, nested, first)\u001b[0m\n\u001b[1;32m    831\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m source\u001b[38;5;241m.\u001b[39merror(err\u001b[38;5;241m.\u001b[39mmsg, \u001b[38;5;28mlen\u001b[39m(name) \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    832\u001b[0m sub_verbose \u001b[38;5;241m=\u001b[39m ((verbose \u001b[38;5;129;01mor\u001b[39;00m (add_flags \u001b[38;5;241m&\u001b[39m SRE_FLAG_VERBOSE)) \u001b[38;5;129;01mand\u001b[39;00m\n\u001b[1;32m    833\u001b[0m                \u001b[38;5;129;01mnot\u001b[39;00m (del_flags \u001b[38;5;241m&\u001b[39m SRE_FLAG_VERBOSE))\n\u001b[0;32m--> 834\u001b[0m p \u001b[38;5;241m=\u001b[39m \u001b[43m_parse_sub\u001b[49m\u001b[43m(\u001b[49m\u001b[43msource\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstate\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msub_verbose\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnested\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    835\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m source\u001b[38;5;241m.\u001b[39mmatch(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m)\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m    836\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m source\u001b[38;5;241m.\u001b[39merror(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmissing ), unterminated subpattern\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    837\u001b[0m                        source\u001b[38;5;241m.\u001b[39mtell() \u001b[38;5;241m-\u001b[39m start)\n",
      "File \u001b[0;32m~/.conda/envs/histdb_env/lib/python3.8/sre_parse.py:443\u001b[0m, in \u001b[0;36m_parse_sub\u001b[0;34m(source, state, verbose, nested)\u001b[0m\n\u001b[1;32m    441\u001b[0m start \u001b[38;5;241m=\u001b[39m source\u001b[38;5;241m.\u001b[39mtell()\n\u001b[1;32m    442\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[0;32m--> 443\u001b[0m     itemsappend(\u001b[43m_parse\u001b[49m\u001b[43m(\u001b[49m\u001b[43msource\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstate\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mverbose\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnested\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    444\u001b[0m \u001b[43m                       \u001b[49m\u001b[38;5;129;43;01mnot\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mnested\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mand\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mnot\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mitems\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[1;32m    445\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m sourcematch(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m|\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m    446\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[0;32m~/.conda/envs/histdb_env/lib/python3.8/sre_parse.py:671\u001b[0m, in \u001b[0;36m_parse\u001b[0;34m(source, state, verbose, nested, first)\u001b[0m\n\u001b[1;32m    668\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m source\u001b[38;5;241m.\u001b[39merror(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnothing to repeat\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    669\u001b[0m                        source\u001b[38;5;241m.\u001b[39mtell() \u001b[38;5;241m-\u001b[39m here \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mlen\u001b[39m(this))\n\u001b[1;32m    670\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m item[\u001b[38;5;241m0\u001b[39m][\u001b[38;5;241m0\u001b[39m] \u001b[38;5;129;01min\u001b[39;00m _REPEATCODES:\n\u001b[0;32m--> 671\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m source\u001b[38;5;241m.\u001b[39merror(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmultiple repeat\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    672\u001b[0m                        source\u001b[38;5;241m.\u001b[39mtell() \u001b[38;5;241m-\u001b[39m here \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mlen\u001b[39m(this))\n\u001b[1;32m    673\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m item[\u001b[38;5;241m0\u001b[39m][\u001b[38;5;241m0\u001b[39m] \u001b[38;5;129;01mis\u001b[39;00m SUBPATTERN:\n\u001b[1;32m    674\u001b[0m     group, add_flags, del_flags, p \u001b[38;5;241m=\u001b[39m item[\u001b[38;5;241m0\u001b[39m][\u001b[38;5;241m1\u001b[39m]\n",
      "\u001b[0;31merror\u001b[0m: multiple repeat at position 11"
     ]
    }
   ],
   "source": [
    "sequence_df.query(\n",
    "    f'type==\"H3_(Lilly???)\" | variant.str.contains(\"H3_(Lilly???)\")',\n",
    "    engine=\"python\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "a01e42cb-d68b-4941-a564-2d1a4a02da45",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NP_734466.1</th>\n",
       "      <td>NP_734466.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A.1_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>221613.0</td>\n",
       "      <td>H2AC1</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>2011515 24506885 7068607</td>\n",
       "      <td>MSGRGKQGGKARAKSKSRSSRAGLQFPVGRIHRLLRKGNYAERIGA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               accession type variant_group                variant doublet gi  \\\n",
       "NP_734466.1  NP_734466.1  H2A          cH2A  cH2A.1_(Homo_sapiens)              \n",
       "\n",
       "            ncbi_gene_id hgnc_gene_name taxonomy_id      organism    phylum  \\\n",
       "NP_734466.1     221613.0          H2AC1      9606.0  Homo sapiens  Chordata   \n",
       "\n",
       "                class taxonomy_group info                references  \\\n",
       "NP_734466.1  Mammalia       Mammalia       2011515 24506885 7068607   \n",
       "\n",
       "                                                      sequence  \n",
       "NP_734466.1  MSGRGKQGGKARAKSKSRSSRAGLQFPVGRIHRLLRKGNYAERIGA...  "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df[sequence_df[\"variant\"].str.contains(\"cH2A.1_\\(Homo_sapiens\\)\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "d2db8101-4205-4af6-9ce0-53245d6a289d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, type, variant_group, variant, doublet, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, references, sequence]\n",
       "Index: []"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df.query(\n",
    "    f'type==\"cH2A.1_(Homo_sapiens)\" | variant.str.contains(\"cH2A.1_(Homo_sapiens)\")',\n",
    "    engine=\"python\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "b09590fc-4dd0-44d2-9c99-e87ad5ff1660",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing  Archaeal\n",
      "Archaeal Alignment length: 0\n",
      "Processing  cH1\n",
      "cH1 Alignment length: 0\n",
      "Processing  generic_H1\n",
      "generic_H1 Alignment length: 13\n",
      "Processing  H1.0_(Homo_sapiens)\n",
      "H1.0_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H1.0\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H1.0 aligned 14 sequences\n",
      "\t\t Adding child H1.0 aligned 1 sequences\n",
      "H1.0 Alignment length: 15\n",
      "Processing  H1.1_(Homo_sapiens)\n",
      "H1.1_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H1.1\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H1.1 aligned 0 sequences\n",
      "\t\t Adding child H1.1 aligned 1 sequences\n",
      "H1.1 Alignment length: 1\n",
      "Processing  H1.10_(Homo_sapiens)\n",
      "H1.10_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H1.10\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H1.10 aligned 5 sequences\n",
      "\t\t Adding child H1.10 aligned 1 sequences\n",
      "H1.10 Alignment length: 6\n",
      "Processing  H1.2_(Homo_sapiens)\n",
      "H1.2_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H1.2\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H1.2 aligned 0 sequences\n",
      "\t\t Adding child H1.2 aligned 1 sequences\n",
      "H1.2 Alignment length: 1\n",
      "Processing  H1.3_(Homo_sapiens)\n",
      "H1.3_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H1.3\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H1.3 aligned 0 sequences\n",
      "\t\t Adding child H1.3 aligned 1 sequences\n",
      "H1.3 Alignment length: 1\n",
      "Processing  H1.4_(Homo_sapiens)\n",
      "H1.4_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H1.4\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H1.4 aligned 0 sequences\n",
      "\t\t Adding child H1.4 aligned 1 sequences\n",
      "H1.4 Alignment length: 1\n",
      "Processing  H1.5_(Homo_sapiens)\n",
      "H1.5_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H1.5\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H1.5 aligned 0 sequences\n",
      "\t\t Adding child H1.5 aligned 1 sequences\n",
      "H1.5 Alignment length: 1\n",
      "Processing  H1.8_(Homo_sapiens)\n",
      "H1.8_(Homo_sapiens) Alignment length: 2\n",
      "Processing  OO_H1.8\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For OO_H1.8 aligned 1 sequences\n",
      "\t\t Adding child OO_H1.8 aligned 2 sequences\n",
      "OO_H1.8 Alignment length: 3\n",
      "Processing  scH1\n",
      "scH1 Alignment length: 2\n",
      "Processing  H1.6_(Homo_sapiens)\n",
      "H1.6_(Homo_sapiens) Alignment length: 1\n",
      "Processing  TS_H1.6\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For TS_H1.6 aligned 7 sequences\n",
      "\t\t Adding child TS_H1.6 aligned 1 sequences\n",
      "TS_H1.6 Alignment length: 8\n",
      "Processing  H1.7_(Homo_sapiens)\n",
      "H1.7_(Homo_sapiens) Alignment length: 1\n",
      "Processing  TS_H1.7\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For TS_H1.7 aligned 1 sequences\n",
      "\t\t Adding child TS_H1.7 aligned 1 sequences\n",
      "TS_H1.7 Alignment length: 2\n",
      "Processing  TS_H1.9\n",
      "TS_H1.9 Alignment length: 3\n",
      "Processing  H1\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H1 aligned 0 sequences\n",
      "\t\t Adding child H1 aligned 0 sequences\n",
      "\t\t Adding child H1 aligned 13 sequences\n",
      "\t\t Adding child H1 aligned 15 sequences\n",
      "\t\t Adding child H1 aligned 1 sequences\n",
      "\t\t Adding child H1 aligned 6 sequences\n",
      "\t\t Adding child H1 aligned 1 sequences\n",
      "\t\t Adding child H1 aligned 1 sequences\n",
      "\t\t Adding child H1 aligned 1 sequences\n",
      "\t\t Adding child H1 aligned 1 sequences\n",
      "\t\t Adding child H1 aligned 3 sequences\n",
      "\t\t Adding child H1 aligned 2 sequences\n",
      "\t\t Adding child H1 aligned 8 sequences\n",
      "\t\t Adding child H1 aligned 2 sequences\n",
      "\t\t Adding child H1 aligned 3 sequences\n",
      "H1 Alignment length: 57\n",
      "Processing  cH2A.10_(Homo_sapiens)\n",
      "cH2A.10_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2A.11_(Homo_sapiens)\n",
      "cH2A.11_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2A.1_(Homo_sapiens)\n",
      "cH2A.1_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2A.2_(Homo_sapiens)\n",
      "cH2A.2_(Homo_sapiens) Alignment length: 2\n",
      "Processing  cH2A.3_(Homo_sapiens)\n",
      "cH2A.3_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2A.4_(Homo_sapiens)\n",
      "cH2A.4_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2A.5_(Homo_sapiens)\n",
      "cH2A.5_(Homo_sapiens) Alignment length: 5\n",
      "Processing  cH2A.6_(Homo_sapiens)\n",
      "cH2A.6_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2A.7_(Homo_sapiens)\n",
      "cH2A.7_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2A.8_(Homo_sapiens)\n",
      "cH2A.8_(Homo_sapiens) Alignment length: 2\n",
      "Processing  cH2A.9_(Homo_sapiens)\n",
      "cH2A.9_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2A_(Homo_sapiens)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2A_(Homo_sapiens) aligned 0 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 2 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 5 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 2 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 1 sequences\n",
      "cH2A_(Homo_sapiens) Alignment length: 17\n",
      "Processing  cH2A.1_(Mus_musculus)\n",
      "cH2A.1_(Mus_musculus) Alignment length: 1\n",
      "Processing  cH2A_(Mus_musculus)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2A_(Mus_musculus) aligned 0 sequences\n",
      "\t\t Adding child cH2A_(Mus_musculus) aligned 1 sequences\n",
      "cH2A_(Mus_musculus) Alignment length: 1\n",
      "Processing  cH2A_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2A_(Mammalia) aligned 16 sequences\n",
      "\t\t Adding child cH2A_(Mammalia) aligned 17 sequences\n",
      "\t\t Adding child cH2A_(Mammalia) aligned 1 sequences\n",
      "cH2A_(Mammalia) Alignment length: 34\n",
      "Processing  cH2A_(Vertebrata)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2A_(Vertebrata) aligned 3 sequences\n",
      "\t\t Adding child cH2A_(Vertebrata) aligned 34 sequences\n",
      "cH2A_(Vertebrata) Alignment length: 37\n",
      "Processing  cH2A_(Animals)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2A_(Animals) aligned 5 sequences\n",
      "\t\t Adding child cH2A_(Animals) aligned 37 sequences\n",
      "cH2A_(Animals) Alignment length: 42\n",
      "Processing  cH2A_(Fungi)\n",
      "cH2A_(Fungi) Alignment length: 2\n",
      "Processing  cH2A_(Chlorophyta)\n",
      "cH2A_(Chlorophyta) Alignment length: 1\n",
      "Processing  cH2A_(Embryophyta)\n",
      "cH2A_(Embryophyta) Alignment length: 8\n",
      "Processing  cH2A_(Plants)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2A_(Plants) aligned 0 sequences\n",
      "\t\t Adding child cH2A_(Plants) aligned 1 sequences\n",
      "\t\t Adding child cH2A_(Plants) aligned 8 sequences\n",
      "cH2A_(Plants) Alignment length: 9\n",
      "Processing  cH2A_(Protists)\n",
      "cH2A_(Protists) Alignment length: 4\n",
      "Processing  cH2A\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2A aligned 0 sequences\n",
      "\t\t Adding child cH2A aligned 42 sequences\n",
      "\t\t Adding child cH2A aligned 2 sequences\n",
      "\t\t Adding child cH2A aligned 9 sequences\n",
      "\t\t Adding child cH2A aligned 4 sequences\n",
      "cH2A Alignment length: 57\n",
      "Processing  gH2A\n",
      "gH2A Alignment length: 1\n",
      "Processing  H2A.J_(Homo_sapiens)\n",
      "H2A.J_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2A.J\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.J aligned 0 sequences\n",
      "\t\t Adding child H2A.J aligned 1 sequences\n",
      "H2A.J Alignment length: 1\n",
      "Processing  H2A.M\n",
      "H2A.M Alignment length: 7\n",
      "Processing  H2A.R\n",
      "H2A.R Alignment length: 5\n",
      "Processing  H2A.W\n",
      "H2A.W Alignment length: 9\n",
      "Processing  H2A.X_(Homo_sapiens)\n",
      "H2A.X_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2A.X_(Mus_musculus)\n",
      "H2A.X_(Mus_musculus) Alignment length: 1\n",
      "Processing  H2A.X_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.X_(Mammalia) aligned 3 sequences\n",
      "\t\t Adding child H2A.X_(Mammalia) aligned 1 sequences\n",
      "\t\t Adding child H2A.X_(Mammalia) aligned 1 sequences\n",
      "H2A.X_(Mammalia) Alignment length: 5\n",
      "Processing  H2A.X_(Vertebrata)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.X_(Vertebrata) aligned 1 sequences\n",
      "\t\t Adding child H2A.X_(Vertebrata) aligned 5 sequences\n",
      "H2A.X_(Vertebrata) Alignment length: 6\n",
      "Processing  H2A.X_(Animals)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.X_(Animals) aligned 2 sequences\n",
      "\t\t Adding child H2A.X_(Animals) aligned 6 sequences\n",
      "H2A.X_(Animals) Alignment length: 8\n",
      "Processing  H2A.X_(Fungi)\n",
      "H2A.X_(Fungi) Alignment length: 4\n",
      "Processing  H2A.X_(Plants)\n",
      "H2A.X_(Plants) Alignment length: 5\n",
      "Processing  H2A.X_(Protists)\n",
      "H2A.X_(Protists) Alignment length: 6\n",
      "Processing  H2A.X\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.X aligned 0 sequences\n",
      "\t\t Adding child H2A.X aligned 8 sequences\n",
      "\t\t Adding child H2A.X aligned 4 sequences\n",
      "\t\t Adding child H2A.X aligned 5 sequences\n",
      "\t\t Adding child H2A.X aligned 6 sequences\n",
      "H2A.X Alignment length: 23\n",
      "Processing  H2A.Z.1_(Homo_sapiens)\n",
      "H2A.Z.1_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2A.Z.1_(Primates)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.Z.1_(Primates) aligned 0 sequences\n",
      "\t\t Adding child H2A.Z.1_(Primates) aligned 1 sequences\n",
      "H2A.Z.1_(Primates) Alignment length: 1\n",
      "Processing  H2A.Z.1_(Chordata)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.Z.1_(Chordata) aligned 2 sequences\n",
      "\t\t Adding child H2A.Z.1_(Chordata) aligned 1 sequences\n",
      "H2A.Z.1_(Chordata) Alignment length: 3\n",
      "Processing  H2A.Z.2.s1_(Homo_sapiens)\n",
      "H2A.Z.2.s1_(Homo_sapiens) Alignment length: 3\n",
      "Processing  H2A.Z.2.s1_(Primates)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.Z.2.s1_(Primates) aligned 0 sequences\n",
      "\t\t Adding child H2A.Z.2.s1_(Primates) aligned 3 sequences\n",
      "H2A.Z.2.s1_(Primates) Alignment length: 3\n",
      "Processing  H2A.Z.2.s2_(Homo_sapiens)\n",
      "H2A.Z.2.s2_(Homo_sapiens) Alignment length: 2\n",
      "Processing  H2A.Z.2.s2_(Primates)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.Z.2.s2_(Primates) aligned 0 sequences\n",
      "\t\t Adding child H2A.Z.2.s2_(Primates) aligned 2 sequences\n",
      "H2A.Z.2.s2_(Primates) Alignment length: 2\n",
      "Processing  H2A.Z.2_(Primates)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.Z.2_(Primates) aligned 0 sequences\n",
      "\t\t Adding child H2A.Z.2_(Primates) aligned 3 sequences\n",
      "\t\t Adding child H2A.Z.2_(Primates) aligned 2 sequences\n",
      "H2A.Z.2_(Primates) Alignment length: 5\n",
      "Processing  H2A.Z.2_(Chordata)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.Z.2_(Chordata) aligned 2 sequences\n",
      "\t\t Adding child H2A.Z.2_(Chordata) aligned 5 sequences\n",
      "H2A.Z.2_(Chordata) Alignment length: 7\n",
      "Processing  H2A.Z\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.Z aligned 20 sequences\n",
      "\t\t Adding child H2A.Z aligned 3 sequences\n",
      "\t\t Adding child H2A.Z aligned 7 sequences\n",
      "H2A.Z Alignment length: 30\n",
      "Processing  macroH2A.1.s1_(Homo_sapiens)\n",
      "macroH2A.1.s1_(Homo_sapiens) Alignment length: 1\n",
      "Processing  macroH2A.1.s1_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For macroH2A.1.s1_(Mammalia) aligned 0 sequences\n",
      "\t\t Adding child macroH2A.1.s1_(Mammalia) aligned 1 sequences\n",
      "macroH2A.1.s1_(Mammalia) Alignment length: 1\n",
      "Processing  macroH2A.1.s2_(Homo_sapiens)\n",
      "macroH2A.1.s2_(Homo_sapiens) Alignment length: 2\n",
      "Processing  macroH2A.1.s2_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For macroH2A.1.s2_(Mammalia) aligned 0 sequences\n",
      "\t\t Adding child macroH2A.1.s2_(Mammalia) aligned 2 sequences\n",
      "macroH2A.1.s2_(Mammalia) Alignment length: 2\n",
      "Processing  macroH2A.1_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For macroH2A.1_(Mammalia) aligned 0 sequences\n",
      "\t\t Adding child macroH2A.1_(Mammalia) aligned 1 sequences\n",
      "\t\t Adding child macroH2A.1_(Mammalia) aligned 2 sequences\n",
      "macroH2A.1_(Mammalia) Alignment length: 3\n",
      "Processing  macroH2A.2_(Homo_sapiens)\n",
      "macroH2A.2_(Homo_sapiens) Alignment length: 1\n",
      "Processing  macroH2A.2_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For macroH2A.2_(Mammalia) aligned 0 sequences\n",
      "\t\t Adding child macroH2A.2_(Mammalia) aligned 1 sequences\n",
      "macroH2A.2_(Mammalia) Alignment length: 1\n",
      "Processing  macroH2A\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For macroH2A aligned 7 sequences\n",
      "\t\t Adding child macroH2A aligned 3 sequences\n",
      "\t\t Adding child macroH2A aligned 1 sequences\n",
      "macroH2A Alignment length: 11\n",
      "Processing  H2A.B.1_(Homo_sapiens)\n",
      "H2A.B.1_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2A.B.2_(Homo_sapiens)\n",
      "H2A.B.2_(Homo_sapiens) Alignment length: 2\n",
      "Processing  H2A.B_(Homo_sapiens)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.B_(Homo_sapiens) aligned 0 sequences\n",
      "\t\t Adding child H2A.B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child H2A.B_(Homo_sapiens) aligned 2 sequences\n",
      "H2A.B_(Homo_sapiens) Alignment length: 3\n",
      "Processing  H2A.B.1_(Mus_musculus)\n",
      "H2A.B.1_(Mus_musculus) Alignment length: 1\n",
      "Processing  H2A.B.2_(Mus_musculus)\n",
      "H2A.B.2_(Mus_musculus) Alignment length: 1\n",
      "Processing  H2A.B.3_(Mus_musculus)\n",
      "H2A.B.3_(Mus_musculus) Alignment length: 1\n",
      "Processing  H2A.B_(Mus_musculus)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.B_(Mus_musculus) aligned 0 sequences\n",
      "\t\t Adding child H2A.B_(Mus_musculus) aligned 1 sequences\n",
      "\t\t Adding child H2A.B_(Mus_musculus) aligned 1 sequences\n",
      "\t\t Adding child H2A.B_(Mus_musculus) aligned 1 sequences\n",
      "H2A.B_(Mus_musculus) Alignment length: 3\n",
      "Processing  H2A.B\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.B aligned 11 sequences\n",
      "\t\t Adding child H2A.B aligned 3 sequences\n",
      "\t\t Adding child H2A.B aligned 3 sequences\n",
      "H2A.B Alignment length: 17\n",
      "Processing  H2A.L.1_(Homo_sapiens)\n",
      "H2A.L.1_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2A.L.3_(Homo_sapiens)\n",
      "H2A.L.3_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2A.L_(Homo_sapiens)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.L_(Homo_sapiens) aligned 0 sequences\n",
      "\t\t Adding child H2A.L_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child H2A.L_(Homo_sapiens) aligned 1 sequences\n",
      "H2A.L_(Homo_sapiens) Alignment length: 2\n",
      "Processing  H2A.L.1_(Mus_musculus)\n",
      "H2A.L.1_(Mus_musculus) Alignment length: 1\n",
      "Processing  H2A.L.2_(Mus_musculus)\n",
      "H2A.L.2_(Mus_musculus) Alignment length: 1\n",
      "Processing  H2A.L.3_(Mus_musculus)\n",
      "H2A.L.3_(Mus_musculus) Alignment length: 1\n",
      "Processing  H2A.L_(Mus_musculus)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.L_(Mus_musculus) aligned 0 sequences\n",
      "\t\t Adding child H2A.L_(Mus_musculus) aligned 1 sequences\n",
      "\t\t Adding child H2A.L_(Mus_musculus) aligned 1 sequences\n",
      "\t\t Adding child H2A.L_(Mus_musculus) aligned 1 sequences\n",
      "H2A.L_(Mus_musculus) Alignment length: 3\n",
      "Processing  H2A.L\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.L aligned 15 sequences\n",
      "\t\t Adding child H2A.L aligned 2 sequences\n",
      "\t\t Adding child H2A.L aligned 3 sequences\n",
      "H2A.L Alignment length: 20\n",
      "Processing  H2A.P_(Homo_sapiens)\n",
      "H2A.P_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2A.P_(Mus_musculus)\n",
      "H2A.P_(Mus_musculus) Alignment length: 1\n",
      "Processing  H2A.P\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.P aligned 8 sequences\n",
      "\t\t Adding child H2A.P aligned 1 sequences\n",
      "\t\t Adding child H2A.P aligned 1 sequences\n",
      "H2A.P Alignment length: 10\n",
      "Processing  H2A.Q\n",
      "H2A.Q Alignment length: 28\n",
      "Processing  short_H2A\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For short_H2A aligned 0 sequences\n",
      "\t\t Adding child short_H2A aligned 17 sequences\n",
      "\t\t Adding child short_H2A aligned 20 sequences\n",
      "\t\t Adding child short_H2A aligned 10 sequences\n",
      "\t\t Adding child short_H2A aligned 28 sequences\n",
      "short_H2A Alignment length: 75\n",
      "Processing  H2A\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A aligned 0 sequences\n",
      "\t\t Adding child H2A aligned 57 sequences\n",
      "\t\t Adding child H2A aligned 1 sequences\n",
      "\t\t Adding child H2A aligned 1 sequences\n",
      "\t\t Adding child H2A aligned 7 sequences\n",
      "\t\t Adding child H2A aligned 5 sequences\n",
      "\t\t Adding child H2A aligned 9 sequences\n",
      "\t\t Adding child H2A aligned 23 sequences\n",
      "\t\t Adding child H2A aligned 30 sequences\n",
      "\t\t Adding child H2A aligned 11 sequences\n",
      "\t\t Adding child H2A aligned 75 sequences\n",
      "H2A Alignment length: 219\n",
      "Processing  cH2B.10_(Homo_sapiens)\n",
      "cH2B.10_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.11_(Homo_sapiens)\n",
      "cH2B.11_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.12_(Homo_sapiens)\n",
      "cH2B.12_(Homo_sapiens) Alignment length: 2\n",
      "Processing  cH2B.13_(Homo_sapiens)\n",
      "cH2B.13_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.14_(Homo_sapiens)\n",
      "cH2B.14_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.15_(Homo_sapiens)\n",
      "cH2B.15_(Homo_sapiens) Alignment length: 0\n",
      "Processing  cH2B.1_(Homo_sapiens)\n",
      "cH2B.1_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.2_(Homo_sapiens)\n",
      "cH2B.2_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.3_(Homo_sapiens)\n",
      "cH2B.3_(Homo_sapiens) Alignment length: 5\n",
      "Processing  cH2B.4_(Homo_sapiens)\n",
      "cH2B.4_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.5_(Homo_sapiens)\n",
      "cH2B.5_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.6_(Homo_sapiens)\n",
      "cH2B.6_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.7_(Homo_sapiens)\n",
      "cH2B.7_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.8_(Homo_sapiens)\n",
      "cH2B.8_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.9_(Homo_sapiens)\n",
      "cH2B.9_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B_(Homo_sapiens)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2B_(Homo_sapiens) aligned 0 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 2 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 5 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "cH2B_(Homo_sapiens) Alignment length: 19\n",
      "Processing  cH2B.1_(Mus_musculus)\n",
      "cH2B.1_(Mus_musculus) Alignment length: 2\n",
      "Processing  cH2B.E_(Mus_musculus)\n",
      "cH2B.E_(Mus_musculus) Alignment length: 1\n",
      "Processing  cH2B_(Mus_musculus)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2B_(Mus_musculus) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Mus_musculus) aligned 2 sequences\n",
      "\t\t Adding child cH2B_(Mus_musculus) aligned 1 sequences\n",
      "cH2B_(Mus_musculus) Alignment length: 4\n",
      "Processing  cH2B_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2B_(Mammalia) aligned 16 sequences\n",
      "\t\t Adding child cH2B_(Mammalia) aligned 19 sequences\n",
      "\t\t Adding child cH2B_(Mammalia) aligned 4 sequences\n",
      "cH2B_(Mammalia) Alignment length: 39\n",
      "Processing  cH2B_(Vertebrata)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2B_(Vertebrata) aligned 2 sequences\n",
      "\t\t Adding child cH2B_(Vertebrata) aligned 39 sequences\n",
      "cH2B_(Vertebrata) Alignment length: 41\n",
      "Processing  cH2B_(Animals)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2B_(Animals) aligned 4 sequences\n",
      "\t\t Adding child cH2B_(Animals) aligned 41 sequences\n",
      "cH2B_(Animals) Alignment length: 45\n",
      "Processing  cH2B_(Fungi)\n",
      "cH2B_(Fungi) Alignment length: 2\n",
      "Processing  cH2B_(Chlorophyta)\n",
      "cH2B_(Chlorophyta) Alignment length: 35\n",
      "Processing  cH2B_(Embryophyta)\n",
      "cH2B_(Embryophyta) Alignment length: 4\n",
      "Processing  cH2B_(Plants)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2B_(Plants) aligned 0 sequences\n",
      "\t\t Adding child cH2B_(Plants) aligned 35 sequences\n",
      "\t\t Adding child cH2B_(Plants) aligned 4 sequences\n",
      "cH2B_(Plants) Alignment length: 39\n",
      "Processing  cH2B_(Protists)\n",
      "cH2B_(Protists) Alignment length: 1\n",
      "Processing  cH2B\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2B aligned 0 sequences\n",
      "\t\t Adding child cH2B aligned 45 sequences\n",
      "\t\t Adding child cH2B aligned 2 sequences\n",
      "\t\t Adding child cH2B aligned 39 sequences\n",
      "\t\t Adding child cH2B aligned 1 sequences\n",
      "cH2B Alignment length: 87\n",
      "Processing  CS_H2B_(Echinoidea)\n",
      "CS_H2B_(Echinoidea) Alignment length: 1\n",
      "Processing  early_H2B_(Echinoidea)\n",
      "early_H2B_(Echinoidea) Alignment length: 1\n",
      "Processing  gH2B\n",
      "gH2B Alignment length: 8\n",
      "Processing  H2B.K_(Homo_sapiens)\n",
      "H2B.K_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2B.K\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2B.K aligned 19 sequences\n",
      "\t\t Adding child H2B.K aligned 1 sequences\n",
      "H2B.K Alignment length: 20\n",
      "Processing  H2B.L\n",
      "H2B.L Alignment length: 11\n",
      "Processing  H2B.N_(Homo_sapiens)\n",
      "H2B.N_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2B.N\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2B.N aligned 13 sequences\n",
      "\t\t Adding child H2B.N aligned 1 sequences\n",
      "H2B.N Alignment length: 14\n",
      "Processing  H2B.O\n",
      "H2B.O Alignment length: 3\n",
      "Processing  H2B.S\n",
      "H2B.S Alignment length: 47\n",
      "Processing  H2B.V\n",
      "H2B.V Alignment length: 1\n",
      "Processing  H2B.W.1_(Homo_sapiens)\n",
      "H2B.W.1_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2B.W.2_(Homo_sapiens)\n",
      "H2B.W.2_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2B.W_(Homo_sapiens)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2B.W_(Homo_sapiens) aligned 0 sequences\n",
      "\t\t Adding child H2B.W_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child H2B.W_(Homo_sapiens) aligned 1 sequences\n",
      "H2B.W_(Homo_sapiens) Alignment length: 2\n",
      "Processing  H2B.W\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2B.W aligned 5 sequences\n",
      "\t\t Adding child H2B.W aligned 2 sequences\n",
      "H2B.W Alignment length: 7\n",
      "Processing  H2B.Z\n",
      "H2B.Z Alignment length: 9\n",
      "Processing  late_H2B_(Echinoidea)\n",
      "late_H2B_(Echinoidea) Alignment length: 2\n",
      "Processing  sperm_H2B_(Echinoidea)\n",
      "sperm_H2B_(Echinoidea) Alignment length: 7\n",
      "Processing  H2B\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2B aligned 0 sequences\n",
      "\t\t Adding child H2B aligned 87 sequences\n",
      "\t\t Adding child H2B aligned 1 sequences\n",
      "\t\t Adding child H2B aligned 1 sequences\n",
      "\t\t Adding child H2B aligned 8 sequences\n",
      "\t\t Adding child H2B aligned 20 sequences\n",
      "\t\t Adding child H2B aligned 11 sequences\n",
      "\t\t Adding child H2B aligned 14 sequences\n",
      "\t\t Adding child H2B aligned 3 sequences\n",
      "\t\t Adding child H2B aligned 47 sequences\n",
      "\t\t Adding child H2B aligned 1 sequences\n",
      "\t\t Adding child H2B aligned 7 sequences\n",
      "\t\t Adding child H2B aligned 9 sequences\n",
      "\t\t Adding child H2B aligned 2 sequences\n",
      "\t\t Adding child H2B aligned 7 sequences\n",
      "H2B Alignment length: 218\n",
      "Processing  cenH3_(Homo_sapiens)\n",
      "cenH3_(Homo_sapiens) Alignment length: 2\n",
      "Processing  cenH3_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cenH3_(Mammalia) aligned 1 sequences\n",
      "\t\t Adding child cenH3_(Mammalia) aligned 2 sequences\n",
      "cenH3_(Mammalia) Alignment length: 3\n",
      "Processing  cenH3_(Animals)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cenH3_(Animals) aligned 2 sequences\n",
      "\t\t Adding child cenH3_(Animals) aligned 3 sequences\n",
      "cenH3_(Animals) Alignment length: 5\n",
      "Processing  cenH3_(Fungi)\n",
      "cenH3_(Fungi) Alignment length: 2\n",
      "Processing  cenH3_(Plants)\n",
      "cenH3_(Plants) Alignment length: 3\n",
      "Processing  cenH3_(Eukarya)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cenH3_(Eukarya) aligned 5 sequences\n",
      "\t\t Adding child cenH3_(Eukarya) aligned 5 sequences\n",
      "\t\t Adding child cenH3_(Eukarya) aligned 2 sequences\n",
      "\t\t Adding child cenH3_(Eukarya) aligned 3 sequences\n",
      "cenH3_(Eukarya) Alignment length: 15\n",
      "Processing  cH3.1_(Homo_sapiens)\n",
      "cH3.1_(Homo_sapiens) Alignment length: 10\n",
      "Processing  cH3.1_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH3.1_(Mammalia) aligned 0 sequences\n",
      "\t\t Adding child cH3.1_(Mammalia) aligned 10 sequences\n",
      "cH3.1_(Mammalia) Alignment length: 10\n",
      "Processing  cH3.2_(Homo_sapiens)\n",
      "cH3.2_(Homo_sapiens) Alignment length: 3\n",
      "Processing  cH3.2_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH3.2_(Mammalia) aligned 0 sequences\n",
      "\t\t Adding child cH3.2_(Mammalia) aligned 3 sequences\n",
      "cH3.2_(Mammalia) Alignment length: 3\n",
      "Processing  cH3_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH3_(Mammalia) aligned 0 sequences\n",
      "\t\t Adding child cH3_(Mammalia) aligned 10 sequences\n",
      "\t\t Adding child cH3_(Mammalia) aligned 3 sequences\n",
      "cH3_(Mammalia) Alignment length: 13\n",
      "Processing  cH3_(Vertebrata)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH3_(Vertebrata) aligned 0 sequences\n",
      "\t\t Adding child cH3_(Vertebrata) aligned 13 sequences\n",
      "cH3_(Vertebrata) Alignment length: 13\n",
      "Processing  cH3_(Animals)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH3_(Animals) aligned 0 sequences\n",
      "\t\t Adding child cH3_(Animals) aligned 13 sequences\n",
      "cH3_(Animals) Alignment length: 13\n",
      "Processing  cH3_(Fungi)\n",
      "cH3_(Fungi) Alignment length: 0\n",
      "Processing  cH3_(Chlorophyta)\n",
      "cH3_(Chlorophyta) Alignment length: 0\n",
      "Processing  cH3_(Embryophyta)\n",
      "cH3_(Embryophyta) Alignment length: 0\n",
      "Processing  cH3_(Plants)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH3_(Plants) aligned 0 sequences\n",
      "\t\t Adding child cH3_(Plants) aligned 0 sequences\n",
      "\t\t Adding child cH3_(Plants) aligned 0 sequences\n",
      "cH3_(Plants) Alignment length: 0\n",
      "Processing  cH3_(Protists)\n",
      "cH3_(Protists) Alignment length: 0\n",
      "Processing  cH3\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH3 aligned 23 sequences\n",
      "\t\t Adding child cH3 aligned 13 sequences\n",
      "cH3 Alignment length: 36\n",
      "Processing  H3.1-like_(Plants)\n",
      "H3.1-like_(Plants) Alignment length: 0\n",
      "Processing  H3.3_(Homo_sapiens)\n",
      "H3.3_(Homo_sapiens) Alignment length: 2\n",
      "Processing  H3.3_(Animals)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.3_(Animals) aligned 0 sequences\n",
      "\t\t Adding child H3.3_(Animals) aligned 2 sequences\n",
      "H3.3_(Animals) Alignment length: 2\n",
      "Processing  H3.3_(Ascomycota)\n",
      "H3.3_(Ascomycota) Alignment length: 0\n",
      "Processing  H3.3_(Fungi)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.3_(Fungi) aligned 0 sequences\n",
      "\t\t Adding child H3.3_(Fungi) aligned 0 sequences\n",
      "H3.3_(Fungi) Alignment length: 0\n",
      "Processing  H3.3_(Plants)\n",
      "H3.3_(Plants) Alignment length: 0\n",
      "Processing  H3.3_(Protists)\n",
      "H3.3_(Protists) Alignment length: 0\n",
      "Processing  H3.3\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.3 aligned 18 sequences\n",
      "\t\t Adding child H3.3 aligned 2 sequences\n",
      "H3.3 Alignment length: 20\n",
      "Processing  H3.X_(Homo_sapiens)\n",
      "H3.X_(Homo_sapiens) Alignment length: 0\n",
      "Processing  H3.X_(Primates?)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.X_(Primates?) aligned 0 sequences\n",
      "\t\t Adding child H3.X_(Primates?) aligned 0 sequences\n",
      "H3.X_(Primates?) Alignment length: 0\n",
      "Processing  H3.Y.1_(Homo_sapiens)\n",
      "H3.Y.1_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H3.Y.2_(Homo_sapiens)\n",
      "H3.Y.2_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H3.Y_(Homo_sapiens)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.Y_(Homo_sapiens) aligned 0 sequences\n",
      "\t\t Adding child H3.Y_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child H3.Y_(Homo_sapiens) aligned 1 sequences\n",
      "H3.Y_(Homo_sapiens) Alignment length: 2\n",
      "Processing  H3.Y_(Primates?)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.Y_(Primates?) aligned 4 sequences\n",
      "\t\t Adding child H3.Y_(Primates?) aligned 2 sequences\n",
      "H3.Y_(Primates?) Alignment length: 6\n",
      "Processing  H3.3-like_(Animals)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.3-like_(Animals) aligned 0 sequences\n",
      "\t\t Adding child H3.3-like_(Animals) aligned 0 sequences\n",
      "\t\t Adding child H3.3-like_(Animals) aligned 6 sequences\n",
      "H3.3-like_(Animals) Alignment length: 6\n",
      "Processing  H3_(Lilly???)\n",
      "H3_(Lilly???) Alignment length: 0\n",
      "Processing  TS H3.10\n",
      "TS H3.10 Alignment length: 0\n",
      "Processing  H3.3-like_(Plants)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.3-like_(Plants) aligned 0 sequences\n",
      "\t\t Adding child H3.3-like_(Plants) aligned 0 sequences\n",
      "\t\t Adding child H3.3-like_(Plants) aligned 0 sequences\n",
      "H3.3-like_(Plants) Alignment length: 0\n",
      "Processing  H3.3-like?\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.3-like? aligned 0 sequences\n",
      "\t\t Adding child H3.3-like? aligned 6 sequences\n",
      "H3.3-like? Alignment length: 6\n",
      "Processing  H3.4_(Homo_sapiens)\n",
      "H3.4_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H3.4_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.4_(Mammalia) aligned 1 sequences\n",
      "\t\t Adding child H3.4_(Mammalia) aligned 1 sequences\n",
      "H3.4_(Mammalia) Alignment length: 2\n",
      "Processing  H3.5_(Homo_sapiens)\n",
      "H3.5_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H3.5_(Hominidae)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.5_(Hominidae) aligned 1 sequences\n",
      "\t\t Adding child H3.5_(Hominidae) aligned 1 sequences\n",
      "H3.5_(Hominidae) Alignment length: 2\n",
      "Processing  H3.6_(Mammals?)?\n",
      "H3.6_(Mammals?)? Alignment length: 0\n",
      "Processing  H3.7_(Mammals?)?\n",
      "H3.7_(Mammals?)? Alignment length: 0\n",
      "Processing  H3.8_(Mammals?)?\n",
      "H3.8_(Mammals?)? Alignment length: 0\n",
      "Processing  H3.B_(Giardia?)\n",
      "H3.B_(Giardia?) Alignment length: 0\n",
      "Processing  H3.P_(Moneuplotes?)\n",
      "H3.P_(Moneuplotes?) Alignment length: 0\n",
      "Processing  H3.V_(Trypanosomatidae)\n",
      "H3.V_(Trypanosomatidae) Alignment length: 0\n",
      "Processing  H3\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3 aligned 0 sequences\n",
      "\t\t Adding child H3 aligned 15 sequences\n",
      "\t\t Adding child H3 aligned 36 sequences\n",
      "\t\t Adding child H3 aligned 20 sequences\n",
      "\t\t Adding child H3 aligned 6 sequences\n",
      "\t\t Adding child H3 aligned 2 sequences\n",
      "\t\t Adding child H3 aligned 2 sequences\n",
      "H3 Alignment length: 81\n",
      "Processing  cH4_(Homo_sapiens)\n",
      "cH4_(Homo_sapiens) Alignment length: 14\n",
      "Processing  cH4\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH4 aligned 12 sequences\n",
      "\t\t Adding child cH4 aligned 14 sequences\n",
      "cH4 Alignment length: 26\n",
      "Processing  H4\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H4 aligned 0 sequences\n",
      "\t\t Adding child H4 aligned 26 sequences\n",
      "H4 Alignment length: 26\n",
      "Processing  Viral\n",
      "Viral Alignment length: 0\n",
      "Processing  \n"
     ]
    }
   ],
   "source": [
    "generate_draft_seeds(\n",
    "    hist_tree,\n",
    "    sequence_df,\n",
    "    \"/home/l_singh/_scratch/hdb/project_dir/histonedb/CURATED_SET/draft_seeds\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2569fada-becb-4419-8926-e0c131b4b7bf",
   "metadata": {},
   "source": [
    "# Close connection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b7719968-002f-4229-846d-e4a8f180ec6b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cursor.close()\n",
    "conn.close()\n",
    "tunnel.stop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "763540a8-72aa-4529-88dc-95ccee1dfdff",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".conda-histdb_env",
   "language": "python",
   "name": "conda-env-.conda-histdb_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
