{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bde51c46-a7d0-4254-8492-29d84551ca32",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/miniconda3/envs/moldyn/lib/python3.7/site-packages/Bio/Align/substitution_matrices/__init__.py:21: BiopythonExperimentalWarning: Bio.Align.substitution_matrices is an experimental module which may still undergo significant changes. In particular, the location of this module may change, and the Array class defined in this module may be moved to other existing or new modules in Biopython.\n",
      "  BiopythonExperimentalWarning)\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c5ed42165c15425ea4f7bc8fb7a38ad7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import glob\n",
    "import io\n",
    "import json\n",
    "import os\n",
    "import shutil\n",
    "import subprocess\n",
    "import sys\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import AlignIO, Entrez, SearchIO, SeqIO\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio.Align.AlignInfo import SummaryInfo\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from ete3 import Tree\n",
    "from pynucl.hist_features import hist_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5fa4cb6f-349a-4a5b-b5b8-4f65b6524182",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def dict2tree(tree, d):\n",
    "    \"\"\"\n",
    "    converts tree from classification.json to a ete3 object\n",
    "    d is\n",
    "    with open('classification.json') as json_file:\n",
    "        data = json.load(json_file)\n",
    "    d=data['tree']\n",
    "    \"\"\"\n",
    "    for k, v in d.items():\n",
    "        CH = tree.add_child(name=k)\n",
    "        if isinstance(v, dict):\n",
    "            dict2tree(CH, v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8e2bf14d-7d44-42e9-afe3-330295bcae67",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def muscle_aln(sequences, options=[], debug=False, sort=True):\n",
    "    muscle = os.path.join(os.path.dirname(sys.executable), \"muscle\")\n",
    "    process = subprocess.Popen(\n",
    "        [muscle] + options,\n",
    "        stdin=subprocess.PIPE,\n",
    "        stdout=subprocess.PIPE,\n",
    "        stderr=subprocess.PIPE,\n",
    "    )\n",
    "    aln, error = process.communicate(sequences.encode(\"utf-8\"))\n",
    "    if debug:\n",
    "        print(sequences)\n",
    "        print()\n",
    "        print(\"Stderr:\")\n",
    "        print(error.decode(\"utf-8\"))\n",
    "        print(\"Stdout:\")\n",
    "        print(aln.decode(\"utf-8\"))\n",
    "    seqFile = io.StringIO()\n",
    "    seqFile.write(aln.decode(\"utf-8\"))\n",
    "    seqFile.seek(0)\n",
    "    sequences_ids = [s.split(\" \", 1)[0] for s in sequences.split(\">\")]\n",
    "    sequences = list(\n",
    "        SeqIO.parse(seqFile, \"fasta\")\n",
    "    )  # Not in same order, but does it matter?\n",
    "    if sort:\n",
    "        sequences.sort(key=lambda x: sequences_ids.index(x.id))  # Yes, it matters\n",
    "    msa = MultipleSeqAlignment(sequences)\n",
    "    return msa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d20dab81-4d92-447c-8a3e-7e8405c92234",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def muscle_p2p_aln(msa1, msa2, options=[], debug=False):\n",
    "    \"\"\"\n",
    "    align two alignments\n",
    "    :return: MultipleSeqAlignment object\n",
    "    \"\"\"\n",
    "    os.makedirs(\"tmp/\")  # create tmp dir to save msa profiles\n",
    "    try:\n",
    "        with open(\"tmp/one.afa\", \"w\") as f:\n",
    "            f.write(format(msa1, \"fasta\"))\n",
    "        with open(\"tmp/two.afa\", \"w\") as f:\n",
    "            f.write(format(msa2, \"fasta\"))\n",
    "\n",
    "        muscle = os.path.join(os.path.dirname(sys.executable), \"muscle\")\n",
    "        process = subprocess.Popen(\n",
    "            [muscle]\n",
    "            + options\n",
    "            + [\"-profile\", \"-in1\", \"tmp/one.afa\", \"-in2\", \"tmp/two.afa\"],\n",
    "            stdin=subprocess.PIPE,\n",
    "            stdout=subprocess.PIPE,\n",
    "            stderr=subprocess.PIPE,\n",
    "        )\n",
    "\n",
    "        aln, error = process.communicate()\n",
    "        if debug:\n",
    "            print(\"Stderr:\")\n",
    "            print(error.decode(\"utf-8\"))\n",
    "            print(\"Stdout:\")\n",
    "            print(aln.decode(\"utf-8\"))\n",
    "\n",
    "        seqFile = io.StringIO()\n",
    "        seqFile.write(aln.decode(\"utf-8\"))\n",
    "        seqFile.seek(0)\n",
    "        sequences = list(\n",
    "            SeqIO.parse(seqFile, \"fasta\")\n",
    "        )  # Not in same order, but does it matter?\n",
    "        msa = MultipleSeqAlignment(sequences)\n",
    "    except:\n",
    "        shutil.rmtree(\"tmp/\")\n",
    "        raise\n",
    "    shutil.rmtree(\"tmp/\")  # rm -rf tmp dir\n",
    "    return msa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "650ed4b1-60d9-456d-977d-125949071539",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_fasta_seq(data):\n",
    "    return \"\\n\".join(\n",
    "        [\n",
    "            SeqRecord(\n",
    "                Seq(row[\"sequence\"]),\n",
    "                # id=f\"{row['organism'].split()[0]}|{row['accession']}|{row['variant']}\",\n",
    "                id=f\"{row['variant'].replace('(', '').replace(')', '')}_{row['organism'].replace(' ', '_')}_{row['accession']}\",\n",
    "                name=row[\"accession\"],\n",
    "                description=f\"organism={row['organism']} phylum={row['phylum']} class={row['class']}\",\n",
    "            ).format(\"fasta\")\n",
    "            for i, row in data.iterrows()\n",
    "        ]\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "129b46be-6737-40c9-8978-e4c1a63c2b5b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def delete_gff_files(directory):\n",
    "    # Находим все файлы с расширением .gff в указанной папке\n",
    "    gff_files = glob.glob(os.path.join(directory, \"*.gff\"))\n",
    "\n",
    "    # Удаляем каждый файл\n",
    "    for file_path in gff_files:\n",
    "        try:\n",
    "            os.remove(file_path)\n",
    "        except Exception as e:\n",
    "            print(f\"Не удалось удалить файл {file_path}: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "1e32607e-4714-4f1f-8a21-12abb93f2485",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def delete_tree_files(directory):\n",
    "    # Находим все файлы с расширением .gff в указанной папке\n",
    "    gff_files = glob.glob(os.path.join(directory, \"*.tree\"))\n",
    "\n",
    "    # Удаляем каждый файл\n",
    "    for file_path in gff_files:\n",
    "        try:\n",
    "            os.remove(file_path)\n",
    "        except Exception as e:\n",
    "            print(f\"Не удалось удалить файл {file_path}: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "260ec5fb-85f2-452d-b134-c63a7f75bf07",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "class _IdHandler:\n",
    "    \"\"\"Generate IDs for GFF3 Parent/Child relationships where they don't exist.\"\"\"\n",
    "\n",
    "    def __init__(self):\n",
    "        self._prefix = \"biopygen\"\n",
    "        self._counter = 1\n",
    "        self._seen_ids = []\n",
    "\n",
    "    def _generate_id(self, quals):\n",
    "        \"\"\"Generate a unique ID not present in our existing IDs.\"\"\"\n",
    "        gen_id = self._get_standard_id(quals)\n",
    "        if gen_id is None:\n",
    "            while 1:\n",
    "                gen_id = \"%s%s\" % (self._prefix, self._counter)\n",
    "                if gen_id not in self._seen_ids:\n",
    "                    break\n",
    "                self._counter += 1\n",
    "        return gen_id\n",
    "\n",
    "    def _get_standard_id(self, quals):\n",
    "        \"\"\"Retrieve standardized IDs from other sources like NCBI GenBank.\n",
    "\n",
    "        This tries to find IDs from known key/values when stored differently\n",
    "        than GFF3 specifications.\n",
    "        \"\"\"\n",
    "        possible_keys = [\"transcript_id\", \"protein_id\"]\n",
    "        for test_key in possible_keys:\n",
    "            if test_key in quals:\n",
    "                cur_id = quals[test_key]\n",
    "                if isinstance(cur_id, tuple) or isinstance(cur_id, list):\n",
    "                    return cur_id[0]\n",
    "                else:\n",
    "                    return cur_id\n",
    "        return None\n",
    "\n",
    "    def update_quals(self, quals, has_children):\n",
    "        \"\"\"Update a set of qualifiers, adding an ID if necessary.\"\"\"\n",
    "        cur_id = quals.get(\"ID\", None)\n",
    "        # if we have an ID, record it\n",
    "        if cur_id:\n",
    "            if not isinstance(cur_id, list) and not isinstance(cur_id, tuple):\n",
    "                cur_id = [cur_id]\n",
    "            for add_id in cur_id:\n",
    "                self._seen_ids.append(add_id)\n",
    "        # if we need one and don't have it, create a new one\n",
    "        elif has_children:\n",
    "            new_id = self._generate_id(quals)\n",
    "            self._seen_ids.append(new_id)\n",
    "            quals[\"ID\"] = [new_id]\n",
    "        return quals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3712dca3-960d-43ef-ae40-8cb93f91846d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "class GFF3Writer:\n",
    "    \"\"\"Write GFF3 files starting with standard Biopython objects.\"\"\"\n",
    "\n",
    "    def __init__(self):\n",
    "        pass\n",
    "\n",
    "    def write(self, recs, out_handle, include_fasta=False, filterids=[]):\n",
    "        \"\"\"Write the provided records to the given handle in GFF3 format.\"\"\"\n",
    "        self.filterids = filterids\n",
    "        id_handler = _IdHandler()\n",
    "        self._write_header(out_handle)\n",
    "        fasta_recs = []\n",
    "        try:\n",
    "            recs = iter(recs)\n",
    "        except TypeError:\n",
    "            recs = [recs]\n",
    "        for rec in recs:\n",
    "            self._write_rec(rec, out_handle)\n",
    "            self._write_annotations(rec.annotations, rec.id, len(rec.seq), out_handle)\n",
    "            for sf in rec.features:\n",
    "                if sf.id in self.filterids:\n",
    "                    continue\n",
    "                sf = self._clean_feature(sf)\n",
    "                id_handler = self._write_feature(sf, rec.id, out_handle, id_handler)\n",
    "            if include_fasta and len(rec.seq) > 0:\n",
    "                fasta_recs.append(rec)\n",
    "        if len(fasta_recs) > 0:\n",
    "            self._write_fasta(fasta_recs, out_handle)\n",
    "\n",
    "    def _clean_feature(self, feature):\n",
    "        quals = {}\n",
    "        for key, val in feature.qualifiers.items():\n",
    "            if not isinstance(val, (list, tuple)):\n",
    "                val = [val]\n",
    "            val = [str(x) for x in val]\n",
    "            quals[key] = val\n",
    "        feature.qualifiers = quals\n",
    "        # Support for Biopython 1.68 and above, which removed sub_features\n",
    "        if not hasattr(feature, \"sub_features\"):\n",
    "            feature.sub_features = []\n",
    "        clean_sub = [self._clean_feature(f) for f in feature.sub_features]\n",
    "        feature.sub_features = clean_sub\n",
    "        return feature\n",
    "\n",
    "    def _write_rec(self, rec, out_handle):\n",
    "        # if we have a SeqRecord, write out optional directive\n",
    "        if len(rec.seq) > 0:\n",
    "            out_handle.write(\"##sequence-region %s 1 %s\\n\" % (rec.id, len(rec.seq)))\n",
    "\n",
    "    def _get_phase(self, feature):\n",
    "        if \"phase\" in feature.qualifiers:\n",
    "            phase = feature.qualifiers[\"phase\"][0]\n",
    "        elif feature.type == \"CDS\":\n",
    "            phase = int(feature.qualifiers.get(\"codon_start\", [1])[0]) - 1\n",
    "        else:\n",
    "            phase = \".\"\n",
    "        return str(phase)\n",
    "\n",
    "    def _write_feature(self, feature, rec_id, out_handle, id_handler, parent_id=None):\n",
    "        \"\"\"Write a feature with location information.\"\"\"\n",
    "        if feature.strand == 1:\n",
    "            strand = \"+\"\n",
    "        elif feature.strand == -1:\n",
    "            strand = \"-\"\n",
    "        else:\n",
    "            strand = \".\"\n",
    "        # remove any standard features from the qualifiers\n",
    "        quals = feature.qualifiers.copy()\n",
    "        for std_qual in [\"source\", \"score\", \"phase\"]:\n",
    "            if std_qual in quals and len(quals[std_qual]) == 1:\n",
    "                del quals[std_qual]\n",
    "        # add a link to a parent identifier if it exists\n",
    "        if parent_id:\n",
    "            if not \"Parent\" in quals:\n",
    "                quals[\"Parent\"] = []\n",
    "            quals[\"Parent\"].append(parent_id)\n",
    "        quals = id_handler.update_quals(quals, len(feature.sub_features) > 0)\n",
    "        if feature.type:\n",
    "            ftype = feature.type\n",
    "        else:\n",
    "            ftype = \"sequence_feature\"\n",
    "        parts = [\n",
    "            str(rec_id),\n",
    "            feature.qualifiers.get(\"source\", [\"feature\"])[0],\n",
    "            ftype,\n",
    "            str(feature.location.nofuzzy_start + 1),  # 1-based indexing\n",
    "            str(feature.location.nofuzzy_end),\n",
    "            feature.qualifiers.get(\"score\", [\".\"])[0],\n",
    "            strand,\n",
    "            self._get_phase(feature),\n",
    "            self._format_keyvals(quals),\n",
    "        ]\n",
    "        out_handle.write(\"\\t\".join(parts) + \"\\n\")\n",
    "        for sub_feature in feature.sub_features:\n",
    "            id_handler = self._write_feature(\n",
    "                sub_feature, rec_id, out_handle, id_handler, quals[\"ID\"][0]\n",
    "            )\n",
    "        return id_handler\n",
    "\n",
    "    def _format_keyvals(self, keyvals):\n",
    "        format_kvs = []\n",
    "        for key in sorted(keyvals.keys()):\n",
    "            values = keyvals[key]\n",
    "            key = key.strip()\n",
    "            format_vals = []\n",
    "            if not isinstance(values, list) or isinstance(values, tuple):\n",
    "                values = [values]\n",
    "            for val in values:\n",
    "                # val = urllib.parse.quote(str(val).strip(), safe=\":/ \")\n",
    "                if (key and val) and val not in format_vals:\n",
    "                    format_vals.append(val)\n",
    "            format_kvs.append(\"%s=%s\" % (key, \",\".join(format_vals)))\n",
    "        return \";\".join(format_kvs)\n",
    "\n",
    "    def _write_annotations(self, anns, rec_id, size, out_handle):\n",
    "        \"\"\"Add annotations which refer to an entire sequence.\"\"\"\n",
    "        format_anns = self._format_keyvals(anns)\n",
    "        if format_anns:\n",
    "            parts = [\n",
    "                rec_id,\n",
    "                \"annotation\",\n",
    "                \"remark\",\n",
    "                \"1\",\n",
    "                str(size if size > 1 else 1),\n",
    "                \".\",\n",
    "                \".\",\n",
    "                \".\",\n",
    "                format_anns,\n",
    "            ]\n",
    "            out_handle.write(\"\\t\".join(parts) + \"\\n\")\n",
    "\n",
    "    def _write_header(self, out_handle):\n",
    "        \"\"\"Write out standard header directives.\"\"\"\n",
    "        out_handle.write(\"##gff-version 3\\n\")\n",
    "\n",
    "    def _write_fasta(self, recs, out_handle):\n",
    "        \"\"\"Write sequence records using the ##FASTA directive.\"\"\"\n",
    "        out_handle.write(\"##FASTA\\n\")\n",
    "        SeqIO.write(recs, out_handle, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "61926d52-1b5d-4c61-b141-cb47eb89b19a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def GFFwrite(recs, out_handle, include_fasta=False, filterids=[]):\n",
    "    \"\"\"High level interface to write GFF3 files from SeqRecords and SeqFeatures.\n",
    "\n",
    "    If include_fasta is True, the GFF3 file will include sequence information\n",
    "    using the ##FASTA directive.\n",
    "    \"\"\"\n",
    "    writer = GFF3Writer()\n",
    "    return writer.write(recs, out_handle, include_fasta, filterids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "bf521b06-4d78-4948-90a2-02df8a2c7804",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def generate_draft_seeds(hist_tree, sequence_data, save_directory):\n",
    "    # generate draft seeds - needs debugging\n",
    "    draft_seeds_msa = {}\n",
    "\n",
    "    # create directory if not exist or rewrite the directory\n",
    "    try:\n",
    "        os.makedirs(save_directory)\n",
    "    except FileExistsError:\n",
    "        # directory already exists\n",
    "        shutil.rmtree(save_directory)\n",
    "        os.makedirs(save_directory)\n",
    "\n",
    "    # hist_tree traversal\n",
    "    for node in hist_tree.traverse(\"postorder\"):\n",
    "        print(\"Processing \", node.name)\n",
    "        if node.is_root():\n",
    "            continue\n",
    "        if node.is_leaf():  # we get sequences for that variant and align them.\n",
    "            draft_seeds_msa[node.name] = muscle_aln(\n",
    "                get_fasta_seq(\n",
    "                    sequence_data.query(\n",
    "                        f'type==\"{node.name}\" | variant==\"{node.name}\"',\n",
    "                        engine=\"python\",\n",
    "                    )\n",
    "                )\n",
    "            )\n",
    "            print(node.name, \"Alignment length:\", len(draft_seeds_msa[node.name]))\n",
    "            with open(\n",
    "                f\"{save_directory}/{node.name.replace('(', '').replace(')', '')}.fasta\",\n",
    "                \"w\",\n",
    "            ) as f:\n",
    "                f.write(format(draft_seeds_msa[node.name], \"fasta\"))\n",
    "        elif not node.is_root():  # we will do profile to profile alignment\n",
    "            # we should first check if there are seqs with this subvariant as the most specific one\n",
    "            print(f\"\\t Node is internal, progressive alignment:\")\n",
    "            msa = muscle_aln(\n",
    "                get_fasta_seq(sequence_data.query(f'variant==\"{node.name}\"'))\n",
    "            )\n",
    "            draft_seeds_msa[node.name + \"_only\"] = msa\n",
    "            print(f\"\\t\\t For {node.name} aligned {len(msa)} sequences\")\n",
    "            # progressively align\n",
    "            for ch in node.get_children():\n",
    "                if len(msa) == 0:\n",
    "                    msa = draft_seeds_msa[ch.name]\n",
    "                    print(\n",
    "                        f\"\\t\\t Adding child {node.name} aligned {len(draft_seeds_msa[ch.name])} sequences\"\n",
    "                    )\n",
    "                    continue\n",
    "                elif len(draft_seeds_msa[ch.name]) != 0:\n",
    "                    msa = muscle_p2p_aln(msa, draft_seeds_msa[ch.name])\n",
    "                    print(\n",
    "                        f\"\\t\\t Adding child {node.name} aligned {len(draft_seeds_msa[ch.name])} sequences\"\n",
    "                    )\n",
    "                else:\n",
    "                    continue\n",
    "            draft_seeds_msa[node.name] = msa\n",
    "            print(node.name, \"Alignment length:\", len(draft_seeds_msa[node.name]))\n",
    "            with open(\n",
    "                f\"{save_directory}/{node.name.replace('(', '').replace(')', '')}.fasta\",\n",
    "                \"w\",\n",
    "            ) as f:\n",
    "                f.write(format(draft_seeds_msa[node.name], \"fasta\"))\n",
    "            with open(\n",
    "                f\"{save_directory}/{node.name.replace('(', '').replace(')', '')}_only.fasta\",\n",
    "                \"w\",\n",
    "            ) as f:\n",
    "                f.write(format(draft_seeds_msa[node.name + \"_only\"], \"fasta\"))\n",
    "        #         print(f\"\\t\\t Final for {node.name} aligned {len(draft_seeds_msa[node.name])} sequences\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4bb6ffa3-5d78-49aa-bb8c-2c67c80cfb4b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Load curated seeds by traversing the tree\n",
    "def generate_gff(curated_seeds_msa_form_files, save_directory):\n",
    "    # firstly, delete all gff files\n",
    "    # delete_gff_files(save_directory)\n",
    "    for k, v in curated_seeds_msa_form_files.items():\n",
    "        if not isinstance(v, list):\n",
    "            cons = SummaryInfo(v).dumb_consensus(threshold=0.1, ambiguous=\"X\")\n",
    "            f = hist_features(cons)\n",
    "            r = SeqRecord(seq=cons, id=\"Consensus\", name=\"Consensus\", features=f)\n",
    "            with open(\n",
    "                f\"{save_directory}/{k.replace('(', '').replace(')', '')}.gff3\", \"w\"\n",
    "            ) as j:\n",
    "                GFFwrite([r], j, filterids=[\"inner_core\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "27771ede-f821-4aa1-b63d-1e6104f2839a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def generate_core_seeds(curated_seeds_msa_form_files, save_directory):\n",
    "    # create directory if not exist or rewrite the directory\n",
    "    try:\n",
    "        os.makedirs(save_directory)\n",
    "    except FileExistsError:\n",
    "        # directory already exists\n",
    "        shutil.rmtree(save_directory)\n",
    "        os.makedirs(save_directory)\n",
    "\n",
    "    for k, msa in curated_seeds_msa_form_files.items():\n",
    "        if not isinstance(msa, list):\n",
    "            core = hist_features(\n",
    "                SummaryInfo(msa).dumb_consensus(threshold=0.1, ambiguous=\"X\")\n",
    "            )[1]\n",
    "            msa_core = msa[:, core.location.start : core.location.end]\n",
    "            with open(\n",
    "                f\"{save_directory}/{k.replace('(', '').replace(')', '')}.fasta\", \"w\"\n",
    "            ) as f:\n",
    "                f.write(msa_core.format(\"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "65c5f837-d62f-43b5-8305-53b946a43301",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "DIR = \"/home/l_singh/_scratch/hdb/project_dir/histonedb/CURATED_SET\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "36cb0bf3-69bf-44f5-8ad4-0c15a62023ca",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "   /-Archaeal\n",
      "  |\n",
      "  |   /-cH1\n",
      "  |  |\n",
      "  |  |--generic_H1\n",
      "  |  |\n",
      "  |  |-H1.0-H1.0_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-H1.1-H1.1_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-H1.10-H1.10_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-H1.2-H1.2_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-H1.3-H1.3_(Homo_sapiens)\n",
      "  |-H1\n",
      "  |  |-H1.4-H1.4_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-H1.5-H1.5_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-OO_H1.8-H1.8_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |--scH1\n",
      "  |  |\n",
      "  |  |-TS_H1.6-H1.6_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-TS_H1.7-H1.7_(Homo_sapiens)\n",
      "  |  |\n",
      "  |   \\-TS_H1.9\n",
      "  |\n",
      "  |                                                                        /-cH2A.10_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2A.11_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2A.1_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2A.2_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2A.3_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                     /cH2A_(Homo_sapiens)-cH2A.4_(Homo_sapiens)\n",
      "  |                                                    |                  |\n",
      "  |                                                    |                  |--cH2A.5_(Homo_sapiens)\n",
      "  |                                                    |                  |\n",
      "  |                                                    |                  |--cH2A.6_(Homo_sapiens)\n",
      "  |                                                    |                  |\n",
      "  |       /cH2A_(Animals)cH2A_(Vertebrata)cH2A_(Mammalia)                 |--cH2A.7_(Homo_sapiens)\n",
      "  |      |                                             |                  |\n",
      "  |      |                                             |                  |--cH2A.8_(Homo_sapiens)\n",
      "  |      |                                             |                  |\n",
      "  |      |                                             |                   \\-cH2A.9_(Homo_sapiens)\n",
      "  |      |                                             |\n",
      "  |      |                                              \\cH2A_(Mus_musculus)-cH2A.1_(Mus_musculus)\n",
      "  |   /cH2A\n",
      "  |  |   |--cH2A_(Fungi)\n",
      "  |  |   |\n",
      "  |  |   |             /-cH2A_(Chlorophyta)\n",
      "  |  |   |-cH2A_(Plants)\n",
      "  |  |   |             \\-cH2A_(Embryophyta)\n",
      "  |  |   |\n",
      "  |  |    \\-cH2A_(Protists)\n",
      "  |  |\n",
      "  |  |--gH2A\n",
      "  |  |\n",
      "  |  |-H2A.J-H2A.J_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |--H2A.M\n",
      "  |  |\n",
      "  |  |--H2A.R\n",
      "  |  |\n",
      "  |  |--H2A.W\n",
      "  |  |\n",
      "  |  |                                                      /-H2A.X_(Homo_sapiens)\n",
      "  |  |     /H2A.X_(Animals)H2A.X_(Vertebrata)H2A.X_(Mammalia)\n",
      "  |  |    |                                                 \\-H2A.X_(Mus_musculus)\n",
      "  |  |    |\n",
      "  |  |-H2A.X-H2A.X_(Fungi)\n",
      "  |  |    |\n",
      "  |  |    |--H2A.X_(Plants)\n",
      "  |  |    |\n",
      "  |-H2A    \\-H2A.X_(Protists)\n",
      "  |  |\n",
      "  |  |     /H2A.Z.1_(Chordata)H2A.Z.1_(Primates)-H2A.Z.1_(Homo_sapiens)\n",
      "  |  |-H2A.Z\n",
      "  |  |    |                                    /H2A.Z.2.s1_(Primates)-H2A.Z.2.s1_(Homo_sapiens)\n",
      "  |  |     \\H2A.Z.2_(Chordata)H2A.Z.2_(Primates)\n",
      "  |  |                                         \\H2A.Z.2.s2_(Primates)-H2A.Z.2.s2_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |                             /macroH2A.1.s1_(Mammalia)-macroH2A.1.s1_(Homo_sapiens)\n",
      "  |  |        /macroH2A.1_(Mammalia)\n",
      "  |  |-macroH2A                    \\macroH2A.1.s2_(Mammalia)-macroH2A.1.s2_(Homo_sapiens)\n",
      "  |  |       |\n",
      "  |  |        \\macroH2A.2_(Mammalia)-macroH2A.2_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |                                  /-H2A.B.1_(Homo_sapiens)\n",
      "  |  |              /H2A.B_(Homo_sapiens)\n",
      "  |  |             |                    \\-H2A.B.2_(Homo_sapiens)\n",
      "  |  |         /H2A.B\n",
      "  |  |        |    |                    /-H2A.B.1_(Mus_musculus)\n",
      "  |  |        |    |                   |\n",
      "  |  |        |     \\H2A.B_(Mus_musculus)-H2A.B.2_(Mus_musculus)\n",
      "  |  |        |                        |\n",
      "  |  |        |                         \\-H2A.B.3_(Mus_musculus)\n",
      "  |  |        |\n",
      "  |  |        |                         /-H2A.L.1_(Homo_sapiens)\n",
      "  |  |        |     /H2A.L_(Homo_sapiens)\n",
      "  |  |        |    |                    \\-H2A.L.3_(Homo_sapiens)\n",
      "  |   \\short_H2AH2A.L\n",
      "  |           |    |                    /-H2A.L.1_(Mus_musculus)\n",
      "  |           |    |                   |\n",
      "  |           |     \\H2A.L_(Mus_musculus)-H2A.L.2_(Mus_musculus)\n",
      "  |           |                        |\n",
      "  |           |                         \\-H2A.L.3_(Mus_musculus)\n",
      "  |           |\n",
      "  |           |     /-H2A.P_(Homo_sapiens)\n",
      "  |           |-H2A.P\n",
      "  |           |     \\-H2A.P_(Mus_musculus)\n",
      "  |           |\n",
      "  |            \\-H2A.Q\n",
      "  |\n",
      "  |                                                                        /-cH2B.10_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2B.11_(Homo_sapiens)\n",
      "--|                                                                       |\n",
      "  |                                                                       |--cH2B.12_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2B.13_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2B.14_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2B.15_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2B.1_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                     /cH2B_(Homo_sapiens)-cH2B.2_(Homo_sapiens)\n",
      "  |                                                    |                  |\n",
      "  |                                                    |                  |--cH2B.3_(Homo_sapiens)\n",
      "  |                                                    |                  |\n",
      "  |                                                    |                  |--cH2B.4_(Homo_sapiens)\n",
      "  |                                                    |                  |\n",
      "  |                                                    |                  |--cH2B.5_(Homo_sapiens)\n",
      "  |                                                    |                  |\n",
      "  |       /cH2B_(Animals)cH2B_(Vertebrata)cH2B_(Mammalia)                 |--cH2B.6_(Homo_sapiens)\n",
      "  |      |                                             |                  |\n",
      "  |      |                                             |                  |--cH2B.7_(Homo_sapiens)\n",
      "  |      |                                             |                  |\n",
      "  |      |                                             |                  |--cH2B.8_(Homo_sapiens)\n",
      "  |      |                                             |                  |\n",
      "  |      |                                             |                   \\-cH2B.9_(Homo_sapiens)\n",
      "  |      |                                             |\n",
      "  |      |                                             |                   /-cH2B.1_(Mus_musculus)\n",
      "  |   /cH2B                                             \\cH2B_(Mus_musculus)\n",
      "  |  |   |                                                                 \\-cH2B.E_(Mus_musculus)\n",
      "  |  |   |\n",
      "  |  |   |--cH2B_(Fungi)\n",
      "  |  |   |\n",
      "  |  |   |             /-cH2B_(Chlorophyta)\n",
      "  |  |   |-cH2B_(Plants)\n",
      "  |  |   |             \\-cH2B_(Embryophyta)\n",
      "  |  |   |\n",
      "  |  |    \\-cH2B_(Protists)\n",
      "  |  |\n",
      "  |  |--CS_H2B_(Echinoidea)\n",
      "  |  |\n",
      "  |  |--early_H2B_(Echinoidea)\n",
      "  |  |\n",
      "  |  |--gH2B\n",
      "  |  |\n",
      "  |  |-H2B.K-H2B.K_(Homo_sapiens)\n",
      "  |-H2B\n",
      "  |  |--H2B.L\n",
      "  |  |\n",
      "  |  |-H2B.N-H2B.N_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |--H2B.O\n",
      "  |  |\n",
      "  |  |--H2B.S\n",
      "  |  |\n",
      "  |  |--H2B.V\n",
      "  |  |\n",
      "  |  |                         /-H2B.W.1_(Homo_sapiens)\n",
      "  |  |-H2B.WH2B.W_(Homo_sapiens)\n",
      "  |  |                         \\-H2B.W.2_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |--H2B.Z\n",
      "  |  |\n",
      "  |  |--late_H2B_(Echinoidea)\n",
      "  |  |\n",
      "  |   \\-sperm_H2B_(Echinoidea)\n",
      "  |\n",
      "  |                  /cenH3_(Animals)cenH3_(Mammalia)-cenH3_(Homo_sapiens)\n",
      "  |                 |\n",
      "  |   /cenH3_(Eukarya)-cenH3_(Fungi)\n",
      "  |  |              |\n",
      "  |  |               \\-cenH3_(Plants)\n",
      "  |  |\n",
      "  |  |                                              /cH3.1_(Mammalia)-cH3.1_(Homo_sapiens)\n",
      "  |  |   /cH3_(Animals)cH3_(Vertebrata)cH3_(Mammalia)\n",
      "  |  |  |                                           \\cH3.2_(Mammalia)-cH3.2_(Homo_sapiens)\n",
      "  |  |  |\n",
      "  |  |  |--cH3_(Fungi)\n",
      "  |  |-cH3\n",
      "  |  |  |            /-cH3_(Chlorophyta)\n",
      "  |  |  |-cH3_(Plants)\n",
      "  |  |  |            \\-cH3_(Embryophyta)\n",
      "  |  |  |\n",
      "  |  |   \\-cH3_(Protists)\n",
      "  |  |\n",
      "  |  |--H3.1-like_(Plants)\n",
      "  |  |\n",
      "  |  |    /H3.3_(Animals)-H3.3_(Homo_sapiens)\n",
      "  |  |   |\n",
      "  |  |   |-H3.3_(Fungi)-H3.3_(Ascomycota)\n",
      "  |  |-H3.3\n",
      "  |  |   |--H3.3_(Plants)\n",
      "  |  |   |\n",
      "  |  |    \\-H3.3_(Protists)\n",
      "  |-H3\n",
      "  |  |                             /H3.X_(Primates?)-H3.X_(Homo_sapiens)\n",
      "  |  |          /H3.3-like_(Animals)\n",
      "  |  |         |                  |                                   /-H3.Y.1_(Homo_sapiens)\n",
      "  |  |         |                   \\H3.Y_(Primates?)H3.Y_(Homo_sapiens)\n",
      "  |  |-H3.3-like?                                                     \\-H3.Y.2_(Homo_sapiens)\n",
      "  |  |         |\n",
      "  |  |         |                  /-H3_(Lilly???)\n",
      "  |  |          \\H3.3-like_(Plants)\n",
      "  |  |                            \\-TS H3.10\n",
      "  |  |\n",
      "  |  |-H3.4_(Mammalia)-H3.4_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-H3.5_(Hominidae)-H3.5_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |--H3.6_(Mammals?)?\n",
      "  |  |\n",
      "  |  |--H3.7_(Mammals?)?\n",
      "  |  |\n",
      "  |  |--H3.8_(Mammals?)?\n",
      "  |  |\n",
      "  |  |--H3.B_(Giardia?)\n",
      "  |  |\n",
      "  |  |--H3.P_(Moneuplotes?)\n",
      "  |  |\n",
      "  |   \\-H3.V_(Trypanosomatidae)\n",
      "  |\n",
      "  |-H4/cH4-cH4_(Homo_sapiens)\n",
      "  |\n",
      "   \\-Viral\n"
     ]
    }
   ],
   "source": [
    "with open(f\"{DIR}/classification.json\") as json_file:\n",
    "    data = json.load(json_file)\n",
    "hist_tree = Tree()\n",
    "dict2tree(hist_tree, data[\"tree\"])\n",
    "print(hist_tree.get_ascii(show_internal=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "2e65711f-6265-4f31-a614-1d9b797001cd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sequence_df = pd.read_csv(f\"{DIR}/histones.csv\").fillna(\"\")\n",
    "sequence_df.index = list(sequence_df[\"accession\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "b09590fc-4dd0-44d2-9c99-e87ad5ff1660",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing  Archaeal\n",
      "Archaeal Alignment length: 0\n",
      "Processing  cH1\n",
      "cH1 Alignment length: 0\n",
      "Processing  generic_H1\n",
      "generic_H1 Alignment length: 13\n",
      "Processing  H1.0_(Homo_sapiens)\n",
      "H1.0_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H1.0\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H1.0 aligned 14 sequences\n",
      "\t\t Adding child H1.0 aligned 1 sequences\n",
      "H1.0 Alignment length: 15\n",
      "Processing  H1.1_(Homo_sapiens)\n",
      "H1.1_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H1.1\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H1.1 aligned 0 sequences\n",
      "\t\t Adding child H1.1 aligned 1 sequences\n",
      "H1.1 Alignment length: 1\n",
      "Processing  H1.10_(Homo_sapiens)\n",
      "H1.10_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H1.10\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H1.10 aligned 5 sequences\n",
      "\t\t Adding child H1.10 aligned 1 sequences\n",
      "H1.10 Alignment length: 6\n",
      "Processing  H1.2_(Homo_sapiens)\n",
      "H1.2_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H1.2\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H1.2 aligned 0 sequences\n",
      "\t\t Adding child H1.2 aligned 1 sequences\n",
      "H1.2 Alignment length: 1\n",
      "Processing  H1.3_(Homo_sapiens)\n",
      "H1.3_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H1.3\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H1.3 aligned 0 sequences\n",
      "\t\t Adding child H1.3 aligned 1 sequences\n",
      "H1.3 Alignment length: 1\n",
      "Processing  H1.4_(Homo_sapiens)\n",
      "H1.4_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H1.4\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H1.4 aligned 0 sequences\n",
      "\t\t Adding child H1.4 aligned 1 sequences\n",
      "H1.4 Alignment length: 1\n",
      "Processing  H1.5_(Homo_sapiens)\n",
      "H1.5_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H1.5\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H1.5 aligned 0 sequences\n",
      "\t\t Adding child H1.5 aligned 1 sequences\n",
      "H1.5 Alignment length: 1\n",
      "Processing  H1.8_(Homo_sapiens)\n",
      "H1.8_(Homo_sapiens) Alignment length: 2\n",
      "Processing  OO_H1.8\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For OO_H1.8 aligned 1 sequences\n",
      "\t\t Adding child OO_H1.8 aligned 2 sequences\n",
      "OO_H1.8 Alignment length: 3\n",
      "Processing  scH1\n",
      "scH1 Alignment length: 2\n",
      "Processing  H1.6_(Homo_sapiens)\n",
      "H1.6_(Homo_sapiens) Alignment length: 1\n",
      "Processing  TS_H1.6\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For TS_H1.6 aligned 7 sequences\n",
      "\t\t Adding child TS_H1.6 aligned 1 sequences\n",
      "TS_H1.6 Alignment length: 8\n",
      "Processing  H1.7_(Homo_sapiens)\n",
      "H1.7_(Homo_sapiens) Alignment length: 1\n",
      "Processing  TS_H1.7\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For TS_H1.7 aligned 1 sequences\n",
      "\t\t Adding child TS_H1.7 aligned 1 sequences\n",
      "TS_H1.7 Alignment length: 2\n",
      "Processing  TS_H1.9\n",
      "TS_H1.9 Alignment length: 3\n",
      "Processing  H1\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H1 aligned 0 sequences\n",
      "\t\t Adding child H1 aligned 0 sequences\n",
      "\t\t Adding child H1 aligned 13 sequences\n",
      "\t\t Adding child H1 aligned 15 sequences\n",
      "\t\t Adding child H1 aligned 1 sequences\n",
      "\t\t Adding child H1 aligned 6 sequences\n",
      "\t\t Adding child H1 aligned 1 sequences\n",
      "\t\t Adding child H1 aligned 1 sequences\n",
      "\t\t Adding child H1 aligned 1 sequences\n",
      "\t\t Adding child H1 aligned 1 sequences\n",
      "\t\t Adding child H1 aligned 3 sequences\n",
      "\t\t Adding child H1 aligned 2 sequences\n",
      "\t\t Adding child H1 aligned 8 sequences\n",
      "\t\t Adding child H1 aligned 2 sequences\n",
      "\t\t Adding child H1 aligned 3 sequences\n",
      "H1 Alignment length: 57\n",
      "Processing  cH2A.10_(Homo_sapiens)\n",
      "cH2A.10_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2A.11_(Homo_sapiens)\n",
      "cH2A.11_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2A.1_(Homo_sapiens)\n",
      "cH2A.1_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2A.2_(Homo_sapiens)\n",
      "cH2A.2_(Homo_sapiens) Alignment length: 2\n",
      "Processing  cH2A.3_(Homo_sapiens)\n",
      "cH2A.3_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2A.4_(Homo_sapiens)\n",
      "cH2A.4_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2A.5_(Homo_sapiens)\n",
      "cH2A.5_(Homo_sapiens) Alignment length: 5\n",
      "Processing  cH2A.6_(Homo_sapiens)\n",
      "cH2A.6_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2A.7_(Homo_sapiens)\n",
      "cH2A.7_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2A.8_(Homo_sapiens)\n",
      "cH2A.8_(Homo_sapiens) Alignment length: 2\n",
      "Processing  cH2A.9_(Homo_sapiens)\n",
      "cH2A.9_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2A_(Homo_sapiens)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2A_(Homo_sapiens) aligned 0 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 2 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 5 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 2 sequences\n",
      "\t\t Adding child cH2A_(Homo_sapiens) aligned 1 sequences\n",
      "cH2A_(Homo_sapiens) Alignment length: 17\n",
      "Processing  cH2A.1_(Mus_musculus)\n",
      "cH2A.1_(Mus_musculus) Alignment length: 1\n",
      "Processing  cH2A_(Mus_musculus)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2A_(Mus_musculus) aligned 0 sequences\n",
      "\t\t Adding child cH2A_(Mus_musculus) aligned 1 sequences\n",
      "cH2A_(Mus_musculus) Alignment length: 1\n",
      "Processing  cH2A_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2A_(Mammalia) aligned 16 sequences\n",
      "\t\t Adding child cH2A_(Mammalia) aligned 17 sequences\n",
      "\t\t Adding child cH2A_(Mammalia) aligned 1 sequences\n",
      "cH2A_(Mammalia) Alignment length: 34\n",
      "Processing  cH2A_(Vertebrata)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2A_(Vertebrata) aligned 3 sequences\n",
      "\t\t Adding child cH2A_(Vertebrata) aligned 34 sequences\n",
      "cH2A_(Vertebrata) Alignment length: 37\n",
      "Processing  cH2A_(Animals)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2A_(Animals) aligned 5 sequences\n",
      "\t\t Adding child cH2A_(Animals) aligned 37 sequences\n",
      "cH2A_(Animals) Alignment length: 42\n",
      "Processing  cH2A_(Fungi)\n",
      "cH2A_(Fungi) Alignment length: 2\n",
      "Processing  cH2A_(Chlorophyta)\n",
      "cH2A_(Chlorophyta) Alignment length: 1\n",
      "Processing  cH2A_(Embryophyta)\n",
      "cH2A_(Embryophyta) Alignment length: 8\n",
      "Processing  cH2A_(Plants)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2A_(Plants) aligned 0 sequences\n",
      "\t\t Adding child cH2A_(Plants) aligned 1 sequences\n",
      "\t\t Adding child cH2A_(Plants) aligned 8 sequences\n",
      "cH2A_(Plants) Alignment length: 9\n",
      "Processing  cH2A_(Protists)\n",
      "cH2A_(Protists) Alignment length: 4\n",
      "Processing  cH2A\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2A aligned 0 sequences\n",
      "\t\t Adding child cH2A aligned 42 sequences\n",
      "\t\t Adding child cH2A aligned 2 sequences\n",
      "\t\t Adding child cH2A aligned 9 sequences\n",
      "\t\t Adding child cH2A aligned 4 sequences\n",
      "cH2A Alignment length: 57\n",
      "Processing  gH2A\n",
      "gH2A Alignment length: 1\n",
      "Processing  H2A.J_(Homo_sapiens)\n",
      "H2A.J_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2A.J\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.J aligned 0 sequences\n",
      "\t\t Adding child H2A.J aligned 1 sequences\n",
      "H2A.J Alignment length: 1\n",
      "Processing  H2A.M\n",
      "H2A.M Alignment length: 7\n",
      "Processing  H2A.R\n",
      "H2A.R Alignment length: 5\n",
      "Processing  H2A.W\n",
      "H2A.W Alignment length: 9\n",
      "Processing  H2A.X_(Homo_sapiens)\n",
      "H2A.X_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2A.X_(Mus_musculus)\n",
      "H2A.X_(Mus_musculus) Alignment length: 1\n",
      "Processing  H2A.X_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.X_(Mammalia) aligned 3 sequences\n",
      "\t\t Adding child H2A.X_(Mammalia) aligned 1 sequences\n",
      "\t\t Adding child H2A.X_(Mammalia) aligned 1 sequences\n",
      "H2A.X_(Mammalia) Alignment length: 5\n",
      "Processing  H2A.X_(Vertebrata)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.X_(Vertebrata) aligned 1 sequences\n",
      "\t\t Adding child H2A.X_(Vertebrata) aligned 5 sequences\n",
      "H2A.X_(Vertebrata) Alignment length: 6\n",
      "Processing  H2A.X_(Animals)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.X_(Animals) aligned 2 sequences\n",
      "\t\t Adding child H2A.X_(Animals) aligned 6 sequences\n",
      "H2A.X_(Animals) Alignment length: 8\n",
      "Processing  H2A.X_(Fungi)\n",
      "H2A.X_(Fungi) Alignment length: 4\n",
      "Processing  H2A.X_(Plants)\n",
      "H2A.X_(Plants) Alignment length: 5\n",
      "Processing  H2A.X_(Protists)\n",
      "H2A.X_(Protists) Alignment length: 6\n",
      "Processing  H2A.X\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.X aligned 0 sequences\n",
      "\t\t Adding child H2A.X aligned 8 sequences\n",
      "\t\t Adding child H2A.X aligned 4 sequences\n",
      "\t\t Adding child H2A.X aligned 5 sequences\n",
      "\t\t Adding child H2A.X aligned 6 sequences\n",
      "H2A.X Alignment length: 23\n",
      "Processing  H2A.Z.1_(Homo_sapiens)\n",
      "H2A.Z.1_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2A.Z.1_(Primates)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.Z.1_(Primates) aligned 0 sequences\n",
      "\t\t Adding child H2A.Z.1_(Primates) aligned 1 sequences\n",
      "H2A.Z.1_(Primates) Alignment length: 1\n",
      "Processing  H2A.Z.1_(Chordata)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.Z.1_(Chordata) aligned 2 sequences\n",
      "\t\t Adding child H2A.Z.1_(Chordata) aligned 1 sequences\n",
      "H2A.Z.1_(Chordata) Alignment length: 3\n",
      "Processing  H2A.Z.2.s1_(Homo_sapiens)\n",
      "H2A.Z.2.s1_(Homo_sapiens) Alignment length: 3\n",
      "Processing  H2A.Z.2.s1_(Primates)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.Z.2.s1_(Primates) aligned 0 sequences\n",
      "\t\t Adding child H2A.Z.2.s1_(Primates) aligned 3 sequences\n",
      "H2A.Z.2.s1_(Primates) Alignment length: 3\n",
      "Processing  H2A.Z.2.s2_(Homo_sapiens)\n",
      "H2A.Z.2.s2_(Homo_sapiens) Alignment length: 2\n",
      "Processing  H2A.Z.2.s2_(Primates)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.Z.2.s2_(Primates) aligned 0 sequences\n",
      "\t\t Adding child H2A.Z.2.s2_(Primates) aligned 2 sequences\n",
      "H2A.Z.2.s2_(Primates) Alignment length: 2\n",
      "Processing  H2A.Z.2_(Primates)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.Z.2_(Primates) aligned 0 sequences\n",
      "\t\t Adding child H2A.Z.2_(Primates) aligned 3 sequences\n",
      "\t\t Adding child H2A.Z.2_(Primates) aligned 2 sequences\n",
      "H2A.Z.2_(Primates) Alignment length: 5\n",
      "Processing  H2A.Z.2_(Chordata)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.Z.2_(Chordata) aligned 2 sequences\n",
      "\t\t Adding child H2A.Z.2_(Chordata) aligned 5 sequences\n",
      "H2A.Z.2_(Chordata) Alignment length: 7\n",
      "Processing  H2A.Z\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.Z aligned 20 sequences\n",
      "\t\t Adding child H2A.Z aligned 3 sequences\n",
      "\t\t Adding child H2A.Z aligned 7 sequences\n",
      "H2A.Z Alignment length: 30\n",
      "Processing  macroH2A.1.s1_(Homo_sapiens)\n",
      "macroH2A.1.s1_(Homo_sapiens) Alignment length: 1\n",
      "Processing  macroH2A.1.s1_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For macroH2A.1.s1_(Mammalia) aligned 0 sequences\n",
      "\t\t Adding child macroH2A.1.s1_(Mammalia) aligned 1 sequences\n",
      "macroH2A.1.s1_(Mammalia) Alignment length: 1\n",
      "Processing  macroH2A.1.s2_(Homo_sapiens)\n",
      "macroH2A.1.s2_(Homo_sapiens) Alignment length: 2\n",
      "Processing  macroH2A.1.s2_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For macroH2A.1.s2_(Mammalia) aligned 0 sequences\n",
      "\t\t Adding child macroH2A.1.s2_(Mammalia) aligned 2 sequences\n",
      "macroH2A.1.s2_(Mammalia) Alignment length: 2\n",
      "Processing  macroH2A.1_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For macroH2A.1_(Mammalia) aligned 0 sequences\n",
      "\t\t Adding child macroH2A.1_(Mammalia) aligned 1 sequences\n",
      "\t\t Adding child macroH2A.1_(Mammalia) aligned 2 sequences\n",
      "macroH2A.1_(Mammalia) Alignment length: 3\n",
      "Processing  macroH2A.2_(Homo_sapiens)\n",
      "macroH2A.2_(Homo_sapiens) Alignment length: 1\n",
      "Processing  macroH2A.2_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For macroH2A.2_(Mammalia) aligned 0 sequences\n",
      "\t\t Adding child macroH2A.2_(Mammalia) aligned 1 sequences\n",
      "macroH2A.2_(Mammalia) Alignment length: 1\n",
      "Processing  macroH2A\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For macroH2A aligned 7 sequences\n",
      "\t\t Adding child macroH2A aligned 3 sequences\n",
      "\t\t Adding child macroH2A aligned 1 sequences\n",
      "macroH2A Alignment length: 11\n",
      "Processing  H2A.B.1_(Homo_sapiens)\n",
      "H2A.B.1_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2A.B.2_(Homo_sapiens)\n",
      "H2A.B.2_(Homo_sapiens) Alignment length: 2\n",
      "Processing  H2A.B_(Homo_sapiens)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.B_(Homo_sapiens) aligned 0 sequences\n",
      "\t\t Adding child H2A.B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child H2A.B_(Homo_sapiens) aligned 2 sequences\n",
      "H2A.B_(Homo_sapiens) Alignment length: 3\n",
      "Processing  H2A.B.1_(Mus_musculus)\n",
      "H2A.B.1_(Mus_musculus) Alignment length: 1\n",
      "Processing  H2A.B.2_(Mus_musculus)\n",
      "H2A.B.2_(Mus_musculus) Alignment length: 1\n",
      "Processing  H2A.B.3_(Mus_musculus)\n",
      "H2A.B.3_(Mus_musculus) Alignment length: 1\n",
      "Processing  H2A.B_(Mus_musculus)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.B_(Mus_musculus) aligned 0 sequences\n",
      "\t\t Adding child H2A.B_(Mus_musculus) aligned 1 sequences\n",
      "\t\t Adding child H2A.B_(Mus_musculus) aligned 1 sequences\n",
      "\t\t Adding child H2A.B_(Mus_musculus) aligned 1 sequences\n",
      "H2A.B_(Mus_musculus) Alignment length: 3\n",
      "Processing  H2A.B\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.B aligned 11 sequences\n",
      "\t\t Adding child H2A.B aligned 3 sequences\n",
      "\t\t Adding child H2A.B aligned 3 sequences\n",
      "H2A.B Alignment length: 17\n",
      "Processing  H2A.L.1_(Homo_sapiens)\n",
      "H2A.L.1_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2A.L.3_(Homo_sapiens)\n",
      "H2A.L.3_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2A.L_(Homo_sapiens)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.L_(Homo_sapiens) aligned 0 sequences\n",
      "\t\t Adding child H2A.L_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child H2A.L_(Homo_sapiens) aligned 1 sequences\n",
      "H2A.L_(Homo_sapiens) Alignment length: 2\n",
      "Processing  H2A.L.1_(Mus_musculus)\n",
      "H2A.L.1_(Mus_musculus) Alignment length: 1\n",
      "Processing  H2A.L.2_(Mus_musculus)\n",
      "H2A.L.2_(Mus_musculus) Alignment length: 1\n",
      "Processing  H2A.L.3_(Mus_musculus)\n",
      "H2A.L.3_(Mus_musculus) Alignment length: 1\n",
      "Processing  H2A.L_(Mus_musculus)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.L_(Mus_musculus) aligned 0 sequences\n",
      "\t\t Adding child H2A.L_(Mus_musculus) aligned 1 sequences\n",
      "\t\t Adding child H2A.L_(Mus_musculus) aligned 1 sequences\n",
      "\t\t Adding child H2A.L_(Mus_musculus) aligned 1 sequences\n",
      "H2A.L_(Mus_musculus) Alignment length: 3\n",
      "Processing  H2A.L\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.L aligned 15 sequences\n",
      "\t\t Adding child H2A.L aligned 2 sequences\n",
      "\t\t Adding child H2A.L aligned 3 sequences\n",
      "H2A.L Alignment length: 20\n",
      "Processing  H2A.P_(Homo_sapiens)\n",
      "H2A.P_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2A.P_(Mus_musculus)\n",
      "H2A.P_(Mus_musculus) Alignment length: 1\n",
      "Processing  H2A.P\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A.P aligned 8 sequences\n",
      "\t\t Adding child H2A.P aligned 1 sequences\n",
      "\t\t Adding child H2A.P aligned 1 sequences\n",
      "H2A.P Alignment length: 10\n",
      "Processing  H2A.Q\n",
      "H2A.Q Alignment length: 28\n",
      "Processing  short_H2A\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For short_H2A aligned 0 sequences\n",
      "\t\t Adding child short_H2A aligned 17 sequences\n",
      "\t\t Adding child short_H2A aligned 20 sequences\n",
      "\t\t Adding child short_H2A aligned 10 sequences\n",
      "\t\t Adding child short_H2A aligned 28 sequences\n",
      "short_H2A Alignment length: 75\n",
      "Processing  H2A\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2A aligned 0 sequences\n",
      "\t\t Adding child H2A aligned 57 sequences\n",
      "\t\t Adding child H2A aligned 1 sequences\n",
      "\t\t Adding child H2A aligned 1 sequences\n",
      "\t\t Adding child H2A aligned 7 sequences\n",
      "\t\t Adding child H2A aligned 5 sequences\n",
      "\t\t Adding child H2A aligned 9 sequences\n",
      "\t\t Adding child H2A aligned 23 sequences\n",
      "\t\t Adding child H2A aligned 30 sequences\n",
      "\t\t Adding child H2A aligned 11 sequences\n",
      "\t\t Adding child H2A aligned 75 sequences\n",
      "H2A Alignment length: 219\n",
      "Processing  cH2B.10_(Homo_sapiens)\n",
      "cH2B.10_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.11_(Homo_sapiens)\n",
      "cH2B.11_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.12_(Homo_sapiens)\n",
      "cH2B.12_(Homo_sapiens) Alignment length: 2\n",
      "Processing  cH2B.13_(Homo_sapiens)\n",
      "cH2B.13_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.14_(Homo_sapiens)\n",
      "cH2B.14_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.15_(Homo_sapiens)\n",
      "cH2B.15_(Homo_sapiens) Alignment length: 0\n",
      "Processing  cH2B.1_(Homo_sapiens)\n",
      "cH2B.1_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.2_(Homo_sapiens)\n",
      "cH2B.2_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.3_(Homo_sapiens)\n",
      "cH2B.3_(Homo_sapiens) Alignment length: 5\n",
      "Processing  cH2B.4_(Homo_sapiens)\n",
      "cH2B.4_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.5_(Homo_sapiens)\n",
      "cH2B.5_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.6_(Homo_sapiens)\n",
      "cH2B.6_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.7_(Homo_sapiens)\n",
      "cH2B.7_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.8_(Homo_sapiens)\n",
      "cH2B.8_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B.9_(Homo_sapiens)\n",
      "cH2B.9_(Homo_sapiens) Alignment length: 1\n",
      "Processing  cH2B_(Homo_sapiens)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2B_(Homo_sapiens) aligned 0 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 2 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 5 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Homo_sapiens) aligned 1 sequences\n",
      "cH2B_(Homo_sapiens) Alignment length: 19\n",
      "Processing  cH2B.1_(Mus_musculus)\n",
      "cH2B.1_(Mus_musculus) Alignment length: 2\n",
      "Processing  cH2B.E_(Mus_musculus)\n",
      "cH2B.E_(Mus_musculus) Alignment length: 1\n",
      "Processing  cH2B_(Mus_musculus)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2B_(Mus_musculus) aligned 1 sequences\n",
      "\t\t Adding child cH2B_(Mus_musculus) aligned 2 sequences\n",
      "\t\t Adding child cH2B_(Mus_musculus) aligned 1 sequences\n",
      "cH2B_(Mus_musculus) Alignment length: 4\n",
      "Processing  cH2B_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2B_(Mammalia) aligned 16 sequences\n",
      "\t\t Adding child cH2B_(Mammalia) aligned 19 sequences\n",
      "\t\t Adding child cH2B_(Mammalia) aligned 4 sequences\n",
      "cH2B_(Mammalia) Alignment length: 39\n",
      "Processing  cH2B_(Vertebrata)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2B_(Vertebrata) aligned 2 sequences\n",
      "\t\t Adding child cH2B_(Vertebrata) aligned 39 sequences\n",
      "cH2B_(Vertebrata) Alignment length: 41\n",
      "Processing  cH2B_(Animals)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2B_(Animals) aligned 4 sequences\n",
      "\t\t Adding child cH2B_(Animals) aligned 41 sequences\n",
      "cH2B_(Animals) Alignment length: 45\n",
      "Processing  cH2B_(Fungi)\n",
      "cH2B_(Fungi) Alignment length: 2\n",
      "Processing  cH2B_(Chlorophyta)\n",
      "cH2B_(Chlorophyta) Alignment length: 35\n",
      "Processing  cH2B_(Embryophyta)\n",
      "cH2B_(Embryophyta) Alignment length: 4\n",
      "Processing  cH2B_(Plants)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2B_(Plants) aligned 0 sequences\n",
      "\t\t Adding child cH2B_(Plants) aligned 35 sequences\n",
      "\t\t Adding child cH2B_(Plants) aligned 4 sequences\n",
      "cH2B_(Plants) Alignment length: 39\n",
      "Processing  cH2B_(Protists)\n",
      "cH2B_(Protists) Alignment length: 1\n",
      "Processing  cH2B\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH2B aligned 0 sequences\n",
      "\t\t Adding child cH2B aligned 45 sequences\n",
      "\t\t Adding child cH2B aligned 2 sequences\n",
      "\t\t Adding child cH2B aligned 39 sequences\n",
      "\t\t Adding child cH2B aligned 1 sequences\n",
      "cH2B Alignment length: 87\n",
      "Processing  CS_H2B_(Echinoidea)\n",
      "CS_H2B_(Echinoidea) Alignment length: 1\n",
      "Processing  early_H2B_(Echinoidea)\n",
      "early_H2B_(Echinoidea) Alignment length: 1\n",
      "Processing  gH2B\n",
      "gH2B Alignment length: 8\n",
      "Processing  H2B.K_(Homo_sapiens)\n",
      "H2B.K_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2B.K\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2B.K aligned 19 sequences\n",
      "\t\t Adding child H2B.K aligned 1 sequences\n",
      "H2B.K Alignment length: 20\n",
      "Processing  H2B.L\n",
      "H2B.L Alignment length: 11\n",
      "Processing  H2B.N_(Homo_sapiens)\n",
      "H2B.N_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2B.N\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2B.N aligned 13 sequences\n",
      "\t\t Adding child H2B.N aligned 1 sequences\n",
      "H2B.N Alignment length: 14\n",
      "Processing  H2B.O\n",
      "H2B.O Alignment length: 3\n",
      "Processing  H2B.S\n",
      "H2B.S Alignment length: 47\n",
      "Processing  H2B.V\n",
      "H2B.V Alignment length: 1\n",
      "Processing  H2B.W.1_(Homo_sapiens)\n",
      "H2B.W.1_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2B.W.2_(Homo_sapiens)\n",
      "H2B.W.2_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H2B.W_(Homo_sapiens)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2B.W_(Homo_sapiens) aligned 0 sequences\n",
      "\t\t Adding child H2B.W_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child H2B.W_(Homo_sapiens) aligned 1 sequences\n",
      "H2B.W_(Homo_sapiens) Alignment length: 2\n",
      "Processing  H2B.W\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2B.W aligned 5 sequences\n",
      "\t\t Adding child H2B.W aligned 2 sequences\n",
      "H2B.W Alignment length: 7\n",
      "Processing  H2B.Z\n",
      "H2B.Z Alignment length: 9\n",
      "Processing  late_H2B_(Echinoidea)\n",
      "late_H2B_(Echinoidea) Alignment length: 2\n",
      "Processing  sperm_H2B_(Echinoidea)\n",
      "sperm_H2B_(Echinoidea) Alignment length: 7\n",
      "Processing  H2B\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H2B aligned 0 sequences\n",
      "\t\t Adding child H2B aligned 87 sequences\n",
      "\t\t Adding child H2B aligned 1 sequences\n",
      "\t\t Adding child H2B aligned 1 sequences\n",
      "\t\t Adding child H2B aligned 8 sequences\n",
      "\t\t Adding child H2B aligned 20 sequences\n",
      "\t\t Adding child H2B aligned 11 sequences\n",
      "\t\t Adding child H2B aligned 14 sequences\n",
      "\t\t Adding child H2B aligned 3 sequences\n",
      "\t\t Adding child H2B aligned 47 sequences\n",
      "\t\t Adding child H2B aligned 1 sequences\n",
      "\t\t Adding child H2B aligned 7 sequences\n",
      "\t\t Adding child H2B aligned 9 sequences\n",
      "\t\t Adding child H2B aligned 2 sequences\n",
      "\t\t Adding child H2B aligned 7 sequences\n",
      "H2B Alignment length: 218\n",
      "Processing  cenH3_(Homo_sapiens)\n",
      "cenH3_(Homo_sapiens) Alignment length: 2\n",
      "Processing  cenH3_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cenH3_(Mammalia) aligned 1 sequences\n",
      "\t\t Adding child cenH3_(Mammalia) aligned 2 sequences\n",
      "cenH3_(Mammalia) Alignment length: 3\n",
      "Processing  cenH3_(Animals)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cenH3_(Animals) aligned 2 sequences\n",
      "\t\t Adding child cenH3_(Animals) aligned 3 sequences\n",
      "cenH3_(Animals) Alignment length: 5\n",
      "Processing  cenH3_(Fungi)\n",
      "cenH3_(Fungi) Alignment length: 2\n",
      "Processing  cenH3_(Plants)\n",
      "cenH3_(Plants) Alignment length: 3\n",
      "Processing  cenH3_(Eukarya)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cenH3_(Eukarya) aligned 5 sequences\n",
      "\t\t Adding child cenH3_(Eukarya) aligned 5 sequences\n",
      "\t\t Adding child cenH3_(Eukarya) aligned 2 sequences\n",
      "\t\t Adding child cenH3_(Eukarya) aligned 3 sequences\n",
      "cenH3_(Eukarya) Alignment length: 15\n",
      "Processing  cH3.1_(Homo_sapiens)\n",
      "cH3.1_(Homo_sapiens) Alignment length: 10\n",
      "Processing  cH3.1_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH3.1_(Mammalia) aligned 0 sequences\n",
      "\t\t Adding child cH3.1_(Mammalia) aligned 10 sequences\n",
      "cH3.1_(Mammalia) Alignment length: 10\n",
      "Processing  cH3.2_(Homo_sapiens)\n",
      "cH3.2_(Homo_sapiens) Alignment length: 3\n",
      "Processing  cH3.2_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH3.2_(Mammalia) aligned 0 sequences\n",
      "\t\t Adding child cH3.2_(Mammalia) aligned 3 sequences\n",
      "cH3.2_(Mammalia) Alignment length: 3\n",
      "Processing  cH3_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH3_(Mammalia) aligned 0 sequences\n",
      "\t\t Adding child cH3_(Mammalia) aligned 10 sequences\n",
      "\t\t Adding child cH3_(Mammalia) aligned 3 sequences\n",
      "cH3_(Mammalia) Alignment length: 13\n",
      "Processing  cH3_(Vertebrata)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH3_(Vertebrata) aligned 0 sequences\n",
      "\t\t Adding child cH3_(Vertebrata) aligned 13 sequences\n",
      "cH3_(Vertebrata) Alignment length: 13\n",
      "Processing  cH3_(Animals)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH3_(Animals) aligned 0 sequences\n",
      "\t\t Adding child cH3_(Animals) aligned 13 sequences\n",
      "cH3_(Animals) Alignment length: 13\n",
      "Processing  cH3_(Fungi)\n",
      "cH3_(Fungi) Alignment length: 0\n",
      "Processing  cH3_(Chlorophyta)\n",
      "cH3_(Chlorophyta) Alignment length: 0\n",
      "Processing  cH3_(Embryophyta)\n",
      "cH3_(Embryophyta) Alignment length: 0\n",
      "Processing  cH3_(Plants)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH3_(Plants) aligned 0 sequences\n",
      "\t\t Adding child cH3_(Plants) aligned 0 sequences\n",
      "\t\t Adding child cH3_(Plants) aligned 0 sequences\n",
      "cH3_(Plants) Alignment length: 0\n",
      "Processing  cH3_(Protists)\n",
      "cH3_(Protists) Alignment length: 0\n",
      "Processing  cH3\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH3 aligned 23 sequences\n",
      "\t\t Adding child cH3 aligned 13 sequences\n",
      "cH3 Alignment length: 36\n",
      "Processing  H3.1-like_(Plants)\n",
      "H3.1-like_(Plants) Alignment length: 0\n",
      "Processing  H3.3_(Homo_sapiens)\n",
      "H3.3_(Homo_sapiens) Alignment length: 2\n",
      "Processing  H3.3_(Animals)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.3_(Animals) aligned 0 sequences\n",
      "\t\t Adding child H3.3_(Animals) aligned 2 sequences\n",
      "H3.3_(Animals) Alignment length: 2\n",
      "Processing  H3.3_(Ascomycota)\n",
      "H3.3_(Ascomycota) Alignment length: 0\n",
      "Processing  H3.3_(Fungi)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.3_(Fungi) aligned 0 sequences\n",
      "\t\t Adding child H3.3_(Fungi) aligned 0 sequences\n",
      "H3.3_(Fungi) Alignment length: 0\n",
      "Processing  H3.3_(Plants)\n",
      "H3.3_(Plants) Alignment length: 0\n",
      "Processing  H3.3_(Protists)\n",
      "H3.3_(Protists) Alignment length: 0\n",
      "Processing  H3.3\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.3 aligned 18 sequences\n",
      "\t\t Adding child H3.3 aligned 2 sequences\n",
      "H3.3 Alignment length: 20\n",
      "Processing  H3.X_(Homo_sapiens)\n",
      "H3.X_(Homo_sapiens) Alignment length: 0\n",
      "Processing  H3.X_(Primates?)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.X_(Primates?) aligned 0 sequences\n",
      "\t\t Adding child H3.X_(Primates?) aligned 0 sequences\n",
      "H3.X_(Primates?) Alignment length: 0\n",
      "Processing  H3.Y.1_(Homo_sapiens)\n",
      "H3.Y.1_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H3.Y.2_(Homo_sapiens)\n",
      "H3.Y.2_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H3.Y_(Homo_sapiens)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.Y_(Homo_sapiens) aligned 0 sequences\n",
      "\t\t Adding child H3.Y_(Homo_sapiens) aligned 1 sequences\n",
      "\t\t Adding child H3.Y_(Homo_sapiens) aligned 1 sequences\n",
      "H3.Y_(Homo_sapiens) Alignment length: 2\n",
      "Processing  H3.Y_(Primates?)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.Y_(Primates?) aligned 4 sequences\n",
      "\t\t Adding child H3.Y_(Primates?) aligned 2 sequences\n",
      "H3.Y_(Primates?) Alignment length: 6\n",
      "Processing  H3.3-like_(Animals)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.3-like_(Animals) aligned 0 sequences\n",
      "\t\t Adding child H3.3-like_(Animals) aligned 0 sequences\n",
      "\t\t Adding child H3.3-like_(Animals) aligned 6 sequences\n",
      "H3.3-like_(Animals) Alignment length: 6\n",
      "Processing  H3_(Lilly???)\n",
      "H3_(Lilly???) Alignment length: 0\n",
      "Processing  TS H3.10\n",
      "TS H3.10 Alignment length: 0\n",
      "Processing  H3.3-like_(Plants)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.3-like_(Plants) aligned 0 sequences\n",
      "\t\t Adding child H3.3-like_(Plants) aligned 0 sequences\n",
      "\t\t Adding child H3.3-like_(Plants) aligned 0 sequences\n",
      "H3.3-like_(Plants) Alignment length: 0\n",
      "Processing  H3.3-like?\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.3-like? aligned 0 sequences\n",
      "\t\t Adding child H3.3-like? aligned 6 sequences\n",
      "H3.3-like? Alignment length: 6\n",
      "Processing  H3.4_(Homo_sapiens)\n",
      "H3.4_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H3.4_(Mammalia)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.4_(Mammalia) aligned 1 sequences\n",
      "\t\t Adding child H3.4_(Mammalia) aligned 1 sequences\n",
      "H3.4_(Mammalia) Alignment length: 2\n",
      "Processing  H3.5_(Homo_sapiens)\n",
      "H3.5_(Homo_sapiens) Alignment length: 1\n",
      "Processing  H3.5_(Hominidae)\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3.5_(Hominidae) aligned 1 sequences\n",
      "\t\t Adding child H3.5_(Hominidae) aligned 1 sequences\n",
      "H3.5_(Hominidae) Alignment length: 2\n",
      "Processing  H3.6_(Mammals?)?\n",
      "H3.6_(Mammals?)? Alignment length: 0\n",
      "Processing  H3.7_(Mammals?)?\n",
      "H3.7_(Mammals?)? Alignment length: 0\n",
      "Processing  H3.8_(Mammals?)?\n",
      "H3.8_(Mammals?)? Alignment length: 0\n",
      "Processing  H3.B_(Giardia?)\n",
      "H3.B_(Giardia?) Alignment length: 0\n",
      "Processing  H3.P_(Moneuplotes?)\n",
      "H3.P_(Moneuplotes?) Alignment length: 0\n",
      "Processing  H3.V_(Trypanosomatidae)\n",
      "H3.V_(Trypanosomatidae) Alignment length: 0\n",
      "Processing  H3\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H3 aligned 0 sequences\n",
      "\t\t Adding child H3 aligned 15 sequences\n",
      "\t\t Adding child H3 aligned 36 sequences\n",
      "\t\t Adding child H3 aligned 20 sequences\n",
      "\t\t Adding child H3 aligned 6 sequences\n",
      "\t\t Adding child H3 aligned 2 sequences\n",
      "\t\t Adding child H3 aligned 2 sequences\n",
      "H3 Alignment length: 81\n",
      "Processing  cH4_(Homo_sapiens)\n",
      "cH4_(Homo_sapiens) Alignment length: 14\n",
      "Processing  cH4\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For cH4 aligned 12 sequences\n",
      "\t\t Adding child cH4 aligned 14 sequences\n",
      "cH4 Alignment length: 26\n",
      "Processing  H4\n",
      "\t Node is internal, progressive alignment:\n",
      "\t\t For H4 aligned 0 sequences\n",
      "\t\t Adding child H4 aligned 26 sequences\n",
      "H4 Alignment length: 26\n",
      "Processing  Viral\n",
      "Viral Alignment length: 0\n",
      "Processing  \n"
     ]
    }
   ],
   "source": [
    "generate_draft_seeds(\n",
    "    hist_tree,\n",
    "    sequence_df,\n",
    "    f\"{DIR}/draft_seeds\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "889b517f-cf89-460f-8ccc-da0c2ec1e001",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading  Archaeal\n",
      "Loading  cH1\n",
      "Loading  generic_H1\n",
      "Loading  H1.0_(Homo_sapiens)\n",
      "Loading  H1.0\n",
      "Loading  H1.1_(Homo_sapiens)\n",
      "Loading  H1.1\n",
      "Loading  H1.10_(Homo_sapiens)\n",
      "Loading  H1.10\n",
      "Loading  H1.2_(Homo_sapiens)\n",
      "Loading  H1.2\n",
      "Loading  H1.3_(Homo_sapiens)\n",
      "Loading  H1.3\n",
      "Loading  H1.4_(Homo_sapiens)\n",
      "Loading  H1.4\n",
      "Loading  H1.5_(Homo_sapiens)\n",
      "Loading  H1.5\n",
      "Loading  H1.8_(Homo_sapiens)\n",
      "Loading  OO_H1.8\n",
      "Loading  scH1\n",
      "Loading  H1.6_(Homo_sapiens)\n",
      "Loading  TS_H1.6\n",
      "Loading  H1.7_(Homo_sapiens)\n",
      "Loading  TS_H1.7\n",
      "Loading  TS_H1.9\n",
      "Loading  H1\n",
      "Loading  cH2A.10_(Homo_sapiens)\n",
      "Loading  cH2A.11_(Homo_sapiens)\n",
      "Loading  cH2A.1_(Homo_sapiens)\n",
      "Loading  cH2A.2_(Homo_sapiens)\n",
      "Loading  cH2A.3_(Homo_sapiens)\n",
      "Loading  cH2A.4_(Homo_sapiens)\n",
      "Loading  cH2A.5_(Homo_sapiens)\n",
      "Loading  cH2A.6_(Homo_sapiens)\n",
      "Loading  cH2A.7_(Homo_sapiens)\n",
      "Loading  cH2A.8_(Homo_sapiens)\n",
      "Loading  cH2A.9_(Homo_sapiens)\n",
      "Loading  cH2A_(Homo_sapiens)\n",
      "Loading  cH2A.1_(Mus_musculus)\n",
      "Loading  cH2A_(Mus_musculus)\n",
      "Loading  cH2A_(Mammalia)\n",
      "Loading  cH2A_(Vertebrata)\n",
      "Loading  cH2A_(Animals)\n",
      "Loading  cH2A_(Fungi)\n",
      "Loading  cH2A_(Chlorophyta)\n",
      "Loading  cH2A_(Embryophyta)\n",
      "Loading  cH2A_(Plants)\n",
      "Loading  cH2A_(Protists)\n",
      "Loading  cH2A\n",
      "Loading  gH2A\n",
      "Loading  H2A.J_(Homo_sapiens)\n",
      "Loading  H2A.J\n",
      "Loading  H2A.M\n",
      "Loading  H2A.R\n",
      "Loading  H2A.W\n",
      "Loading  H2A.X_(Homo_sapiens)\n",
      "Loading  H2A.X_(Mus_musculus)\n",
      "Loading  H2A.X_(Mammalia)\n",
      "Loading  H2A.X_(Vertebrata)\n",
      "Loading  H2A.X_(Animals)\n",
      "Loading  H2A.X_(Fungi)\n",
      "Loading  H2A.X_(Plants)\n",
      "Loading  H2A.X_(Protists)\n",
      "Loading  H2A.X\n",
      "Loading  H2A.Z.1_(Homo_sapiens)\n",
      "Loading  H2A.Z.1_(Primates)\n",
      "Loading  H2A.Z.1_(Chordata)\n",
      "Loading  H2A.Z.2.s1_(Homo_sapiens)\n",
      "Loading  H2A.Z.2.s1_(Primates)\n",
      "Loading  H2A.Z.2.s2_(Homo_sapiens)\n",
      "Loading  H2A.Z.2.s2_(Primates)\n",
      "Loading  H2A.Z.2_(Primates)\n",
      "Loading  H2A.Z.2_(Chordata)\n",
      "Loading  H2A.Z\n",
      "Loading  macroH2A.1.s1_(Homo_sapiens)\n",
      "Loading  macroH2A.1.s1_(Mammalia)\n",
      "Loading  macroH2A.1.s2_(Homo_sapiens)\n",
      "Loading  macroH2A.1.s2_(Mammalia)\n",
      "Loading  macroH2A.1_(Mammalia)\n",
      "Loading  macroH2A.2_(Homo_sapiens)\n",
      "Loading  macroH2A.2_(Mammalia)\n",
      "Loading  macroH2A\n",
      "Loading  H2A.B.1_(Homo_sapiens)\n",
      "Loading  H2A.B.2_(Homo_sapiens)\n",
      "Loading  H2A.B_(Homo_sapiens)\n",
      "Loading  H2A.B.1_(Mus_musculus)\n",
      "Loading  H2A.B.2_(Mus_musculus)\n",
      "Loading  H2A.B.3_(Mus_musculus)\n",
      "Loading  H2A.B_(Mus_musculus)\n",
      "Loading  H2A.B\n",
      "Loading  H2A.L.1_(Homo_sapiens)\n",
      "Loading  H2A.L.3_(Homo_sapiens)\n",
      "Loading  H2A.L_(Homo_sapiens)\n",
      "Loading  H2A.L.1_(Mus_musculus)\n",
      "Loading  H2A.L.2_(Mus_musculus)\n",
      "Loading  H2A.L.3_(Mus_musculus)\n",
      "Loading  H2A.L_(Mus_musculus)\n",
      "Loading  H2A.L\n",
      "Loading  H2A.P_(Homo_sapiens)\n",
      "Loading  H2A.P_(Mus_musculus)\n",
      "Loading  H2A.P\n",
      "Loading  H2A.Q\n",
      "Loading  short_H2A\n",
      "Loading  H2A\n",
      "Loading  cH2B.10_(Homo_sapiens)\n",
      "Loading  cH2B.11_(Homo_sapiens)\n",
      "Loading  cH2B.12_(Homo_sapiens)\n",
      "Loading  cH2B.13_(Homo_sapiens)\n",
      "Loading  cH2B.14_(Homo_sapiens)\n",
      "Loading  cH2B.15_(Homo_sapiens)\n",
      "Loading  cH2B.1_(Homo_sapiens)\n",
      "Loading  cH2B.2_(Homo_sapiens)\n",
      "Loading  cH2B.3_(Homo_sapiens)\n",
      "Loading  cH2B.4_(Homo_sapiens)\n",
      "Loading  cH2B.5_(Homo_sapiens)\n",
      "Loading  cH2B.6_(Homo_sapiens)\n",
      "Loading  cH2B.7_(Homo_sapiens)\n",
      "Loading  cH2B.8_(Homo_sapiens)\n",
      "Loading  cH2B.9_(Homo_sapiens)\n",
      "Loading  cH2B_(Homo_sapiens)\n",
      "Loading  cH2B.1_(Mus_musculus)\n",
      "Loading  cH2B.E_(Mus_musculus)\n",
      "Loading  cH2B_(Mus_musculus)\n",
      "Loading  cH2B_(Mammalia)\n",
      "Loading  cH2B_(Vertebrata)\n",
      "Loading  cH2B_(Animals)\n",
      "Loading  cH2B_(Fungi)\n",
      "Loading  cH2B_(Chlorophyta)\n",
      "Loading  cH2B_(Embryophyta)\n",
      "Loading  cH2B_(Plants)\n",
      "Loading  cH2B_(Protists)\n",
      "Loading  cH2B\n",
      "Loading  CS_H2B_(Echinoidea)\n",
      "Loading  early_H2B_(Echinoidea)\n",
      "Loading  gH2B\n",
      "Loading  H2B.K_(Homo_sapiens)\n",
      "Loading  H2B.K\n",
      "Loading  H2B.L\n",
      "Loading  H2B.N_(Homo_sapiens)\n",
      "Loading  H2B.N\n",
      "Loading  H2B.O\n",
      "Loading  H2B.S\n",
      "Loading  H2B.V\n",
      "Loading  H2B.W.1_(Homo_sapiens)\n",
      "Loading  H2B.W.2_(Homo_sapiens)\n",
      "Loading  H2B.W_(Homo_sapiens)\n",
      "Loading  H2B.W\n",
      "Loading  H2B.Z\n",
      "Loading  late_H2B_(Echinoidea)\n",
      "Loading  sperm_H2B_(Echinoidea)\n",
      "Loading  H2B\n",
      "Loading  cenH3_(Homo_sapiens)\n",
      "Loading  cenH3_(Mammalia)\n",
      "Loading  cenH3_(Animals)\n",
      "Loading  cenH3_(Fungi)\n",
      "Loading  cenH3_(Plants)\n",
      "Loading  cenH3_(Eukarya)\n",
      "Loading  cH3.1_(Homo_sapiens)\n",
      "Loading  cH3.1_(Mammalia)\n",
      "Loading  cH3.2_(Homo_sapiens)\n",
      "Loading  cH3.2_(Mammalia)\n",
      "Loading  cH3_(Mammalia)\n",
      "Loading  cH3_(Vertebrata)\n",
      "Loading  cH3_(Animals)\n",
      "Loading  cH3_(Fungi)\n",
      "Loading  cH3_(Chlorophyta)\n",
      "Loading  cH3_(Embryophyta)\n",
      "Loading  cH3_(Plants)\n",
      "Loading  cH3_(Protists)\n",
      "Loading  cH3\n",
      "Loading  H3.1-like_(Plants)\n",
      "Loading  H3.3_(Homo_sapiens)\n",
      "Loading  H3.3_(Animals)\n",
      "Loading  H3.3_(Ascomycota)\n",
      "Loading  H3.3_(Fungi)\n",
      "Loading  H3.3_(Plants)\n",
      "Loading  H3.3_(Protists)\n",
      "Loading  H3.3\n",
      "Loading  H3.X_(Homo_sapiens)\n",
      "Loading  H3.X_(Primates?)\n",
      "Loading  H3.Y.1_(Homo_sapiens)\n",
      "Loading  H3.Y.2_(Homo_sapiens)\n",
      "Loading  H3.Y_(Homo_sapiens)\n",
      "Loading  H3.Y_(Primates?)\n",
      "Loading  H3.3-like_(Animals)\n",
      "Loading  H3_(Lilly???)\n",
      "Loading  TS H3.10\n",
      "Loading  H3.3-like_(Plants)\n",
      "Loading  H3.3-like?\n",
      "Loading  H3.4_(Homo_sapiens)\n",
      "Loading  H3.4_(Mammalia)\n",
      "Loading  H3.5_(Homo_sapiens)\n",
      "Loading  H3.5_(Hominidae)\n",
      "Loading  H3.6_(Mammals?)?\n",
      "Loading  H3.7_(Mammals?)?\n",
      "Loading  H3.8_(Mammals?)?\n",
      "Loading  H3.B_(Giardia?)\n",
      "Loading  H3.P_(Moneuplotes?)\n",
      "Loading  H3.V_(Trypanosomatidae)\n",
      "Loading  H3\n",
      "Loading  cH4_(Homo_sapiens)\n",
      "Loading  cH4\n",
      "Loading  H4\n",
      "Loading  Viral\n"
     ]
    }
   ],
   "source": [
    "curated_seeds_msa_form_files = {}\n",
    "for node in hist_tree.traverse(\"postorder\"):\n",
    "    if os.path.isfile(\n",
    "        f\"{DIR}/draft_seeds/{node.name.replace('(', '').replace(')', '')}.fasta\"\n",
    "    ):\n",
    "        print(\"Loading \", node.name)\n",
    "        curated_seeds_msa_form_files[node.name] = (\n",
    "            AlignIO.read(\n",
    "                f\"{DIR}/draft_seeds/{node.name.replace('(', '').replace(')', '')}.fasta\",\n",
    "                \"fasta\",\n",
    "            )\n",
    "            if os.path.getsize(\n",
    "                f\"{DIR}/draft_seeds/{node.name.replace('(', '').replace(')', '')}.fasta\"\n",
    "            )\n",
    "            > 0\n",
    "            else []\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "57cbb132-2ce4-46e2-b213-f045d5e01d90",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "generate_gff(curated_seeds_msa_form_files, f\"{DIR}/draft_seeds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "308d3930-5c07-4367-9a57-358fb30b713c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "generate_core_seeds(curated_seeds_msa_form_files, f\"{DIR}/draft_seeds_core\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "667e6544-087f-47ed-9251-0f5f559e3a06",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def run_bash_command(command_string):\n",
    "    try:\n",
    "        # Запускаем команду bash из строки\n",
    "        result = subprocess.run(\n",
    "            [\"bash\", \"-c\", command_string], check=True, text=True, capture_output=True\n",
    "        )\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(\"Ошибка выполнения скрипта:\", e.stderr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "1fc2d159-8d05-424b-8bcd-6dd6140bbb97",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "shutil.rmtree(\"tmp/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "08671eff-0312-4d18-86f9-495df45db45f",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building  generic_H1\n",
      "Building  H1.0_(Homo_sapiens)\n",
      "Building  H1.0\n",
      "Building  H1.1_(Homo_sapiens)\n",
      "Building  H1.1\n",
      "Building  H1.10_(Homo_sapiens)\n",
      "Building  H1.10\n",
      "Building  H1.2_(Homo_sapiens)\n",
      "Building  H1.2\n",
      "Building  H1.3_(Homo_sapiens)\n",
      "Building  H1.3\n",
      "Building  H1.4_(Homo_sapiens)\n",
      "Building  H1.4\n",
      "Building  H1.5_(Homo_sapiens)\n",
      "Building  H1.5\n",
      "Building  H1.8_(Homo_sapiens)\n",
      "Building  OO_H1.8\n",
      "Building  scH1\n",
      "Building  H1.6_(Homo_sapiens)\n",
      "Building  TS_H1.6\n",
      "Building  H1.7_(Homo_sapiens)\n",
      "Building  TS_H1.7\n",
      "Building  TS_H1.9\n",
      "Building  H1\n",
      "Building  cH2A.10_(Homo_sapiens)\n",
      "Building  cH2A.11_(Homo_sapiens)\n",
      "Building  cH2A.1_(Homo_sapiens)\n",
      "Building  cH2A.2_(Homo_sapiens)\n",
      "Building  cH2A.3_(Homo_sapiens)\n",
      "Building  cH2A.4_(Homo_sapiens)\n",
      "Building  cH2A.5_(Homo_sapiens)\n",
      "Building  cH2A.6_(Homo_sapiens)\n",
      "Building  cH2A.7_(Homo_sapiens)\n",
      "Building  cH2A.8_(Homo_sapiens)\n",
      "Building  cH2A.9_(Homo_sapiens)\n",
      "Building  cH2A_(Homo_sapiens)\n",
      "Building  cH2A.1_(Mus_musculus)\n",
      "Building  cH2A_(Mus_musculus)\n",
      "Building  cH2A_(Mammalia)\n",
      "Building  cH2A_(Vertebrata)\n",
      "Building  cH2A_(Animals)\n",
      "Building  cH2A_(Fungi)\n",
      "Building  cH2A_(Chlorophyta)\n",
      "Building  cH2A_(Embryophyta)\n",
      "Building  cH2A_(Plants)\n",
      "Building  cH2A_(Protists)\n",
      "Building  cH2A\n",
      "Building  gH2A\n",
      "Building  H2A.J_(Homo_sapiens)\n",
      "Building  H2A.J\n",
      "Building  H2A.M\n",
      "Building  H2A.R\n",
      "Building  H2A.W\n",
      "Building  H2A.X_(Homo_sapiens)\n",
      "Building  H2A.X_(Mus_musculus)\n",
      "Building  H2A.X_(Mammalia)\n",
      "Building  H2A.X_(Vertebrata)\n",
      "Building  H2A.X_(Animals)\n",
      "Building  H2A.X_(Fungi)\n",
      "Building  H2A.X_(Plants)\n",
      "Building  H2A.X_(Protists)\n",
      "Building  H2A.X\n",
      "Building  H2A.Z.1_(Homo_sapiens)\n",
      "Building  H2A.Z.1_(Primates)\n",
      "Building  H2A.Z.1_(Chordata)\n",
      "Building  H2A.Z.2.s1_(Homo_sapiens)\n",
      "Building  H2A.Z.2.s1_(Primates)\n",
      "Building  H2A.Z.2.s2_(Homo_sapiens)\n",
      "Building  H2A.Z.2.s2_(Primates)\n",
      "Building  H2A.Z.2_(Primates)\n",
      "Building  H2A.Z.2_(Chordata)\n",
      "Building  H2A.Z\n",
      "Building  macroH2A.1.s1_(Homo_sapiens)\n",
      "Building  macroH2A.1.s1_(Mammalia)\n",
      "Building  macroH2A.1.s2_(Homo_sapiens)\n",
      "Building  macroH2A.1.s2_(Mammalia)\n",
      "Building  macroH2A.1_(Mammalia)\n",
      "Building  macroH2A.2_(Homo_sapiens)\n",
      "Building  macroH2A.2_(Mammalia)\n",
      "Building  macroH2A\n",
      "Building  H2A.B.1_(Homo_sapiens)\n",
      "Building  H2A.B.2_(Homo_sapiens)\n",
      "Building  H2A.B_(Homo_sapiens)\n",
      "Building  H2A.B.1_(Mus_musculus)\n",
      "Building  H2A.B.2_(Mus_musculus)\n",
      "Building  H2A.B.3_(Mus_musculus)\n",
      "Building  H2A.B_(Mus_musculus)\n",
      "Building  H2A.B\n",
      "Building  H2A.L.1_(Homo_sapiens)\n",
      "Building  H2A.L.3_(Homo_sapiens)\n",
      "Building  H2A.L_(Homo_sapiens)\n",
      "Building  H2A.L.1_(Mus_musculus)\n",
      "Building  H2A.L.2_(Mus_musculus)\n",
      "Building  H2A.L.3_(Mus_musculus)\n",
      "Building  H2A.L_(Mus_musculus)\n",
      "Building  H2A.L\n",
      "Building  H2A.P_(Homo_sapiens)\n",
      "Building  H2A.P_(Mus_musculus)\n",
      "Building  H2A.P\n",
      "Building  H2A.Q\n",
      "Building  short_H2A\n",
      "Building  H2A\n",
      "Building  cH2B.10_(Homo_sapiens)\n",
      "Building  cH2B.11_(Homo_sapiens)\n",
      "Building  cH2B.12_(Homo_sapiens)\n",
      "Building  cH2B.13_(Homo_sapiens)\n",
      "Building  cH2B.14_(Homo_sapiens)\n",
      "Building  cH2B.1_(Homo_sapiens)\n",
      "Building  cH2B.2_(Homo_sapiens)\n",
      "Building  cH2B.3_(Homo_sapiens)\n",
      "Building  cH2B.4_(Homo_sapiens)\n",
      "Building  cH2B.5_(Homo_sapiens)\n",
      "Building  cH2B.6_(Homo_sapiens)\n",
      "Building  cH2B.7_(Homo_sapiens)\n",
      "Building  cH2B.8_(Homo_sapiens)\n",
      "Building  cH2B.9_(Homo_sapiens)\n",
      "Building  cH2B_(Homo_sapiens)\n",
      "Building  cH2B.1_(Mus_musculus)\n",
      "Building  cH2B.E_(Mus_musculus)\n",
      "Building  cH2B_(Mus_musculus)\n",
      "Building  cH2B_(Mammalia)\n",
      "Building  cH2B_(Vertebrata)\n",
      "Building  cH2B_(Animals)\n",
      "Building  cH2B_(Fungi)\n",
      "Building  cH2B_(Chlorophyta)\n",
      "Building  cH2B_(Embryophyta)\n",
      "Building  cH2B_(Plants)\n",
      "Building  cH2B_(Protists)\n",
      "Building  cH2B\n",
      "Building  CS_H2B_(Echinoidea)\n",
      "Building  early_H2B_(Echinoidea)\n",
      "Building  gH2B\n",
      "Building  H2B.K_(Homo_sapiens)\n",
      "Building  H2B.K\n",
      "Building  H2B.L\n",
      "Building  H2B.N_(Homo_sapiens)\n",
      "Building  H2B.N\n",
      "Building  H2B.O\n",
      "Building  H2B.S\n",
      "Building  H2B.V\n",
      "Building  H2B.W.1_(Homo_sapiens)\n",
      "Building  H2B.W.2_(Homo_sapiens)\n",
      "Building  H2B.W_(Homo_sapiens)\n",
      "Building  H2B.W\n",
      "Building  H2B.Z\n",
      "Building  late_H2B_(Echinoidea)\n",
      "Building  sperm_H2B_(Echinoidea)\n",
      "Building  H2B\n",
      "Building  cenH3_(Homo_sapiens)\n",
      "Building  cenH3_(Mammalia)\n",
      "Building  cenH3_(Animals)\n",
      "Building  cenH3_(Fungi)\n",
      "Building  cenH3_(Plants)\n",
      "Building  cenH3_(Eukarya)\n",
      "Building  cH3.1_(Homo_sapiens)\n",
      "Building  cH3.1_(Mammalia)\n",
      "Building  cH3.2_(Homo_sapiens)\n",
      "Building  cH3.2_(Mammalia)\n",
      "Building  cH3_(Mammalia)\n",
      "Building  cH3_(Vertebrata)\n",
      "Building  cH3_(Animals)\n",
      "Building  cH3\n",
      "Building  H3.3_(Homo_sapiens)\n",
      "Building  H3.3_(Animals)\n",
      "Building  H3.3\n",
      "Building  H3.Y.1_(Homo_sapiens)\n",
      "Building  H3.Y.2_(Homo_sapiens)\n",
      "Building  H3.Y_(Homo_sapiens)\n",
      "Building  H3.Y_(Primates?)\n",
      "Building  H3.3-like_(Animals)\n",
      "Building  H3.3-like?\n",
      "Building  H3.4_(Homo_sapiens)\n",
      "Building  H3.4_(Mammalia)\n",
      "Building  H3.5_(Homo_sapiens)\n",
      "Building  H3.5_(Hominidae)\n",
      "Building  H3\n",
      "Building  cH4_(Homo_sapiens)\n",
      "Building  cH4\n",
      "Building  H4\n"
     ]
    }
   ],
   "source": [
    "for node in hist_tree.traverse(\"postorder\"):\n",
    "    if os.path.isfile(\n",
    "        f\"{DIR}/draft_seeds/{node.name.replace('(', '').replace(')', '')}.fasta\"\n",
    "    ) and os.path.isfile(\n",
    "        f\"{DIR}/draft_seeds_core/{node.name.replace('(', '').replace(')', '')}.fasta\"\n",
    "    ):\n",
    "        print(\"Building \", node.name)\n",
    "        name = node.name.replace(\"(\", \"\").replace(\")\", \"\")\n",
    "        os.makedirs(\"tmp/\")\n",
    "        run_bash_command(\n",
    "            f\"FastTree -nosupport < {DIR}/draft_seeds_core/{name}.fasta > tmp/tree 2> /dev/null\"\n",
    "        )\n",
    "        run_bash_command(\n",
    "            f\"sed '$ s/.$/Consensus;/' tmp/tree > {DIR}/draft_seeds/{name}.tree\"\n",
    "        )\n",
    "        shutil.rmtree(\"tmp/\")  # rm -rf tmp dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00aaf236-ba98-4f1e-ac0e-a752302f0bf8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "763540a8-72aa-4529-88dc-95ccee1dfdff",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "moldyn",
   "language": "python",
   "name": "conda-env-moldyn-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
