{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7875be0a-21e6-4975-ae59-62c46e4ac202",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from mysql.connector import connection\n",
    "from sshtunnel import SSHTunnelForwarder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c1a5725b-b7e9-4301-a6d7-03fd502d0f25",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open('db_curated_server_info.txt', 'r') as file:\n",
    "    lines = file.readlines()\n",
    "\n",
    "config = {}\n",
    "\n",
    "for line in lines:\n",
    "    line = line.strip()  \n",
    "    if line and not line.startswith('#'):  \n",
    "        key, value = line.split('=', 1)  \n",
    "        config[key] = value.strip()  \n",
    "\n",
    "server_name = config.get('server_name')\n",
    "srever_port = int(config.get('srever_port'))\n",
    "ssh_password = config.get('ssh_password')\n",
    "ssh_username = config.get('ssh_username')\n",
    "db_adress = config.get('db_adress')\n",
    "db_port = int(config.get('db_port'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d6e32db5-5fb0-4fd1-8194-e1cecf6e3621",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32771\n"
     ]
    }
   ],
   "source": [
    "tunnel = SSHTunnelForwarder(\n",
    "    (server_name, srever_port),\n",
    "    ssh_password=ssh_password,\n",
    "    ssh_username=ssh_username,\n",
    "    remote_bind_address=(db_adress, db_port),\n",
    ")\n",
    "tunnel.start()\n",
    "print(tunnel.local_bind_port)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a0167244-a996-453a-8a6d-8986414aca57",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "conn = connection.MySQLConnection(\n",
    "    user=\"db_user\",\n",
    "    password=\"db_password\",\n",
    "    host=\"localhost\",\n",
    "    port=tunnel.local_bind_port,\n",
    "    database=\"db_name\",\n",
    ")\n",
    "cursor = conn.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8dfef833-0997-4b65-8367-8b1f8d17d064",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('alternative_name',),\n",
       " ('histone',),\n",
       " ('histone_description',),\n",
       " ('histone_has_publication',),\n",
       " ('publication',),\n",
       " ('sequence',),\n",
       " ('sequence_has_publication',)]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SHOW TABLES;\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9c7bf588-0553-42fe-89da-65c8a1ab651f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'id, level, taxonomic_span, taxonomic_span_id, description, parent'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()\n",
    "\", \".join([i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "780d536d-7bc7-413c-ad09-55656fa72a97",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'id, name, taxonomy, gene, splice, histone'"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()\n",
    "\", \".join([i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "ecd452b4-eaa2-4b98-8a20-a307202ee9d6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'id, title, doi, author, year'"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()\n",
    "\", \".join([i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "5f230ad0-5e98-4bf7-8937-a621c2771e8d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'histone_id, publication_id'"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone_has_publication\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()\n",
    "\", \".join([i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "1381f91b-1543-43c4-8545-5e30fcd50074",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "add_histone_description = (\n",
    "    \"INSERT INTO histone_description \"\n",
    "    \"(summary, taxonomy, genes, evolution, expression, knock_out, function, sequence, localization, deposition, structure, interactions, disease, caveats) \"\n",
    "    \"VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)\"\n",
    ")\n",
    "add_histone = (\n",
    "    \"INSERT INTO histone \"\n",
    "    \"(id, level, taxonomic_span, taxonomic_span_id, description, parent) \"\n",
    "    \"VALUES (%(id)s, %(level)s, %(taxonomic_span)s, %(taxonomic_span_id)s, %(description)s, %(parent)s)\"\n",
    ")\n",
    "add_alternate_names = (\n",
    "    \"INSERT INTO alternative_name \"\n",
    "    \"(name, taxonomy, gene, splice, histone) \"\n",
    "    \"VALUES (%(name)s, %(taxonomy)s, %(gene)s, %(splice)s, %(histone)s)\"\n",
    ")\n",
    "add_publication = (\n",
    "    \"INSERT INTO publication \"\n",
    "    \"(id, title, doi, author, year) \"\n",
    "    \"VALUES (%(id)s, %(title)s, %(doi)s, %(author)s, %(year)s)\"\n",
    ")\n",
    "add_histone_has_publication = (\n",
    "    \"INSERT INTO histone_has_publication \"\n",
    "    \"(histone_id, publication_id) \"\n",
    "    \"VALUES (%s, %s)\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0a4869c-eba8-4231-bcf5-e2918e06f159",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Add histone variant cH2B.15_(Homo_sapiens)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e77b8375-457d-4f55-b716-737f6641867a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, level, taxonomic_span, taxonomic_span_id, description, parent]\n",
       "Index: []"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone\"\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df[histone_df[\"id\"] == \"cH2B.15_(Homo_sapiens)\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "81b300a9-e18b-471b-8a1e-d08b32cf2dac",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "histone_description_summary = \"cH2B.15_(Homo_sapiens)  is an isoform (variant) of clustered (canonical) H2B histones in human endoded by H2BC12L gene. H2BC12L is represented by a human-specific duplication of the H2BC12 gene from the chromosome 6 onto chromosome 21, the gene appears to be expressed, its protein sequence is expected to have two nonsynonymous substitutions with respect to H2BC12 gene.\"\n",
    "data_histone_description = (histone_description_summary,) + (\"null\",) * 13\n",
    "cursor.execute(add_histone_description, data_histone_description)\n",
    "\n",
    "histone_description_id = cursor.lastrowid\n",
    "data_histone = {\n",
    "    \"id\": \"cH2B.15_(Homo_sapiens)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Homo sapiens\",\n",
    "    \"taxonomic_span_id\": \"9606\",\n",
    "    \"description\": histone_description_id,\n",
    "    \"parent\": \"cH2B_(Homo_sapiens)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdf8f8df-2bca-4847-a46d-a80fce2088cf",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Checking added data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a6aa4fdd-b504-4885-ad43-90281ddb7941",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>evolution</th>\n",
       "      <th>expression</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>cH2B.15_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>156.0</td>\n",
       "      <td>cH2B_(Homo_sapiens)</td>\n",
       "      <td>cH2B.15_(Homo_sapiens)  is an isoform (variant...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        id    level taxonomic_span taxonomic_span_id  \\\n",
       "32  cH2B.15_(Homo_sapiens)  variant   Homo sapiens              9606   \n",
       "\n",
       "    description               parent  \\\n",
       "32        156.0  cH2B_(Homo_sapiens)   \n",
       "\n",
       "                                              summary taxonomy genes  \\\n",
       "32  cH2B.15_(Homo_sapiens)  is an isoform (variant...     null  null   \n",
       "\n",
       "   evolution expression knock_out function sequence localization deposition  \\\n",
       "32      null       null      null     null     null         null       null   \n",
       "\n",
       "   structure interactions disease caveats  \n",
       "32      null         null    null    null  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT h.*, d.summary, d.taxonomy, d.genes, d.evolution, d.expression, d.knock_out, d.function, d.sequence, d.localization, d.deposition, d.structure, d.interactions, d.disease, d.caveats \"\n",
    "    \"FROM histone h LEFT JOIN histone_description d ON h.description = d.id\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df[histone_df[\"id\"] == \"cH2B.15_(Homo_sapiens)\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "beb25a14-ceaf-4425-bc31-e5517477ead6",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Add histone variant cenH3_(Plants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d58d565f-f7ab-4bd6-bc7e-b959caf9f866",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, level, taxonomic_span, taxonomic_span_id, description, parent]\n",
       "Index: []"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone\"\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df[histone_df[\"id\"] == \"cenH3_(Plants)\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "ac50ec21-04d3-4924-a2cb-5d0bb570c306",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                id          level taxonomic_span taxonomic_span_id  \\\n",
       "1  cenH3_(Eukarya)  variant_group           None              None   \n",
       "\n",
       "   description parent  \n",
       "1          NaN     H3  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "histone_df[histone_df[\"id\"] == \"cenH3_(Eukarya)\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ae796ee1-391b-4b15-aa8b-fcf512cbe1b2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>evolution</th>\n",
       "      <th>expression</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H3</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                id          level taxonomic_span taxonomic_span_id  \\\n",
       "1  cenH3_(Eukarya)  variant_group           None              None   \n",
       "\n",
       "   description parent summary taxonomy genes evolution expression knock_out  \\\n",
       "1          NaN     H3    None     None  None      None       None      None   \n",
       "\n",
       "  function sequence localization deposition structure interactions disease  \\\n",
       "1     None     None         None       None      None         None    None   \n",
       "\n",
       "  caveats  \n",
       "1    None  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT h.*, d.summary, d.taxonomy, d.genes, d.evolution, d.expression, d.knock_out, d.function, d.sequence, d.localization, d.deposition, d.structure, d.interactions, d.disease, d.caveats \"\n",
    "    \"FROM histone h LEFT JOIN histone_description d ON h.description = d.id\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df[histone_df[\"id\"] == \"cenH3_(Eukarya)\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "989e9d2a-f0b5-43ee-a9ef-e74213b4b67e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "histone_description_summary = \"cenH3 is a centromere-specific histone variant, which replaces canonical H3 in centromeric nucleosomes. It is required for kinetochore formation, mitotic progression and chromosome segregation. cenH3 has an extended L1-loop and its N-terminal tail is very different from other H3 variants. cenH3s have an extended L1-loop and usually replace Phe84 in canonical H3 with Trp, and Thr 107 with Ala, Cys, or Ser. cenH3s ususally lack a conserved glutamine in the alpha1 helix of the histone fold. cenH3s typically have only about 50-60% amino acid identity to canonical H3 in the histone fold domain and no conservation of the N-terminus.\"\n",
    "data_histone_description = (histone_description_summary,) + (\"null\",) * 13\n",
    "cursor.execute(add_histone_description, data_histone_description)\n",
    "\n",
    "histone_description_id = cursor.lastrowid\n",
    "data_histone = {\n",
    "    \"id\": \"cenH3_(Plants)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Eukaryotes\",\n",
    "    \"taxonomic_span_id\": \"2759\",\n",
    "    \"description\": histone_description_id,\n",
    "    \"parent\": \"cenH3_(Eukarya)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8164e04-c093-45b8-a523-0c9cf04fbcbe",
   "metadata": {},
   "source": [
    "### Add alternate names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "22dc1ee4-ebd1-4567-bc82-5113e73870e4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_alternative_names = [\n",
    "    {\n",
    "        \"name\": \"CENP-A\",\n",
    "        \"taxonomy\": \"null\",\n",
    "        \"gene\": \"null\",\n",
    "        \"splice\": \"null\",\n",
    "        \"histone\": \"cenH3_(Plants)\",\n",
    "    },\n",
    "    {\n",
    "        \"name\": \"Cse4\",\n",
    "        \"taxonomy\": \"null\",\n",
    "        \"gene\": \"null\",\n",
    "        \"splice\": \"null\",\n",
    "        \"histone\": \"cenH3_(Plants)\",\n",
    "    },\n",
    "    {\n",
    "        \"name\": \"HCP-3\",\n",
    "        \"taxonomy\": \"null\",\n",
    "        \"gene\": \"null\",\n",
    "        \"splice\": \"null\",\n",
    "        \"histone\": \"cenH3_(Plants)\",\n",
    "    },\n",
    "    {\n",
    "        \"name\": \"CNP1\",\n",
    "        \"taxonomy\": \"null\",\n",
    "        \"gene\": \"null\",\n",
    "        \"splice\": \"null\",\n",
    "        \"histone\": \"cenH3_(Plants)\",\n",
    "    },\n",
    "    {\n",
    "        \"name\": \"HTR12\",\n",
    "        \"taxonomy\": \"null\",\n",
    "        \"gene\": \"null\",\n",
    "        \"splice\": \"null\",\n",
    "        \"histone\": \"cenH3_(Plants)\",\n",
    "    },\n",
    "    {\n",
    "        \"name\": \"CNA1\",\n",
    "        \"taxonomy\": \"null\",\n",
    "        \"gene\": \"null\",\n",
    "        \"splice\": \"null\",\n",
    "        \"histone\": \"cenH3_(Plants)\",\n",
    "    },\n",
    "    {\n",
    "        \"name\": \"cid\",\n",
    "        \"taxonomy\": \"null\",\n",
    "        \"gene\": \"null\",\n",
    "        \"splice\": \"null\",\n",
    "        \"histone\": \"cenH3_(Plants)\",\n",
    "    },\n",
    "]\n",
    "for data_an in data_alternative_names:\n",
    "    cursor.execute(add_alternate_names, data_an)\n",
    "\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "44519d2f-6e24-40db-935a-c633866ba898",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>82</td>\n",
       "      <td>CENP-A</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>83</td>\n",
       "      <td>Cse4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>84</td>\n",
       "      <td>HCP-3</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>85</td>\n",
       "      <td>CNP1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>86</td>\n",
       "      <td>HTR12</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>87</td>\n",
       "      <td>CNA1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>88</td>\n",
       "      <td>cid</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    id    name taxonomy  gene splice         histone\n",
       "81  82  CENP-A     None  None   None  cenH3_(Plants)\n",
       "82  83    Cse4     None  None   None  cenH3_(Plants)\n",
       "83  84   HCP-3     None  None   None  cenH3_(Plants)\n",
       "84  85    CNP1     None  None   None  cenH3_(Plants)\n",
       "85  86   HTR12     None  None   None  cenH3_(Plants)\n",
       "86  87    CNA1     None  None   None  cenH3_(Plants)\n",
       "87  88     cid     None  None   None  cenH3_(Plants)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name\"\n",
    "cursor.execute(query)\n",
    "alternative_name_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ")\n",
    "alternative_name_df[alternative_name_df[\"histone\"] == \"cenH3_(Plants)\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c02f9327-67cf-4264-b900-c606a0a3d840",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for i in alternative_name_df[alternative_name_df[\"histone\"] == \"cenH3_(Plants)\"][\"id\"]:\n",
    "    query = f\"UPDATE alternative_name SET taxonomy=null WHERE id={i}\"\n",
    "    cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "368ee4ce-c44f-45d7-aee0-57bfdf642301",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for i in alternative_name_df[alternative_name_df[\"histone\"] == \"cenH3_(Plants)\"][\"id\"]:\n",
    "    query = f\"UPDATE alternative_name SET gene=null WHERE id={i}\"\n",
    "    cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "75ae601d-3e14-4cdf-a966-5df57ea6c230",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for i in alternative_name_df[alternative_name_df[\"histone\"] == \"cenH3_(Plants)\"][\"id\"]:\n",
    "    query = f\"UPDATE alternative_name SET splice=null WHERE id={i}\"\n",
    "    cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2351bcbc-e5d3-4763-aca6-bf66a09bd920",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Add publications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "b401d983-c578-499a-b73e-f5c442990a46",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "publication_list = [\n",
    "    \"22650316\",\n",
    "    \"19766562\",\n",
    "    \"25956076\",\n",
    "    \"23324462\",\n",
    "    \"21743476\",\n",
    "    \"14583738\",\n",
    "]\n",
    "query = \"SELECT id FROM publication\"\n",
    "cursor.execute(query)\n",
    "for pub in set(publication_list) - set([i[0] for i in cursor]):\n",
    "    data_publication = {\n",
    "        \"id\": pub,\n",
    "        \"title\": \"null\",\n",
    "        \"doi\": \"null\",\n",
    "        \"author\": \"null\",\n",
    "        \"year\": \"null\",\n",
    "    }\n",
    "    cursor.execute(add_publication, data_publication)\n",
    "    cursor.execute(add_histone_has_publication, (\"cenH3_(Plants)\", pub))\n",
    "\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "0e4a0c38-66eb-43d3-868e-607d26dd39a7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "query = \"SELECT publication_id FROM histone h LEFT JOIN histone_has_publication p ON h.id = p.histone_id\"\n",
    "cursor.execute(query)\n",
    "for pub in set(publication_list) - set([i[0] for i in cursor]):\n",
    "    cursor.execute(add_histone_has_publication, (\"cenH3_(Plants)\", pub))\n",
    "\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "fe373b50-bfb6-46e3-9397-b0bd8b4829b6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cursor.execute(add_histone_has_publication, (\"cenH3_(Plants)\", \"22650316\"))\n",
    "\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2387bcc9-c843-493c-bad9-04768a8bf8be",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Checking added data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "48a23a08-14f2-42a2-b9f4-2c775db1ed71",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>evolution</th>\n",
       "      <th>expression</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>157.0</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3 is a centromere-specific histone variant...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "2  cenH3_(Plants)  variant     Eukaryotes              2759        157.0   \n",
       "\n",
       "            parent                                            summary  \\\n",
       "2  cenH3_(Eukarya)  cenH3 is a centromere-specific histone variant...   \n",
       "\n",
       "  taxonomy genes evolution expression knock_out function sequence  \\\n",
       "2     null  null      null       null      null     null     null   \n",
       "\n",
       "  localization deposition structure interactions disease caveats  \n",
       "2         null       null      null         null    null    null  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT h.*, d.summary, d.taxonomy, d.genes, d.evolution, d.expression, d.knock_out, d.function, d.sequence, d.localization, d.deposition, d.structure, d.interactions, d.disease, d.caveats \"\n",
    "    \"FROM histone h LEFT JOIN histone_description d ON h.description = d.id\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df[histone_df[\"id\"] == \"cenH3_(Plants)\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c192f9e9-feef-4004-a4b2-ff4844f13f1e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>82</td>\n",
       "      <td>CENP-A</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>83</td>\n",
       "      <td>Cse4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>84</td>\n",
       "      <td>HCP-3</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>85</td>\n",
       "      <td>CNP1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>86</td>\n",
       "      <td>HTR12</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>87</td>\n",
       "      <td>CNA1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>88</td>\n",
       "      <td>cid</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    id    name taxonomy  gene splice         histone\n",
       "81  82  CENP-A     None  None   None  cenH3_(Plants)\n",
       "82  83    Cse4     None  None   None  cenH3_(Plants)\n",
       "83  84   HCP-3     None  None   None  cenH3_(Plants)\n",
       "84  85    CNP1     None  None   None  cenH3_(Plants)\n",
       "85  86   HTR12     None  None   None  cenH3_(Plants)\n",
       "86  87    CNA1     None  None   None  cenH3_(Plants)\n",
       "87  88     cid     None  None   None  cenH3_(Plants)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name\"\n",
    "cursor.execute(query)\n",
    "alternative_name_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ")\n",
    "alternative_name_df[alternative_name_df[\"histone\"] == \"cenH3_(Plants)\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "42b4a21d-8792-4b80-8022-c43f360b7198",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>histone_id</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>157.0</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>14583738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>157.0</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>19766562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>157.0</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>21743476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>157.0</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>157.0</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>23324462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>157.0</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>25956076</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "2  cenH3_(Plants)  variant     Eukaryotes              2759        157.0   \n",
       "3  cenH3_(Plants)  variant     Eukaryotes              2759        157.0   \n",
       "4  cenH3_(Plants)  variant     Eukaryotes              2759        157.0   \n",
       "5  cenH3_(Plants)  variant     Eukaryotes              2759        157.0   \n",
       "6  cenH3_(Plants)  variant     Eukaryotes              2759        157.0   \n",
       "7  cenH3_(Plants)  variant     Eukaryotes              2759        157.0   \n",
       "\n",
       "            parent      histone_id publication_id  \n",
       "2  cenH3_(Eukarya)  cenH3_(Plants)       14583738  \n",
       "3  cenH3_(Eukarya)  cenH3_(Plants)       19766562  \n",
       "4  cenH3_(Eukarya)  cenH3_(Plants)       21743476  \n",
       "5  cenH3_(Eukarya)  cenH3_(Plants)       22650316  \n",
       "6  cenH3_(Eukarya)  cenH3_(Plants)       23324462  \n",
       "7  cenH3_(Eukarya)  cenH3_(Plants)       25956076  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * \"\n",
    "    \"FROM histone h LEFT JOIN histone_has_publication p ON h.id = p.histone_id\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df[histone_df[\"id\"] == \"cenH3_(Plants)\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27f1efde-0cbd-4fd2-9cc7-80adc3fbcfec",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Add histone variant TS_H3.4\n",
    "\n",
    "Эта запись дублиует запись `H3.4_(Mammalia)`. Поэтому просто совместим информацию из обоих записей. Для эотого нужно добавить в запись `H3.4_(Mammalia)` еще одно альтенативное имя (\"H3.1t\"), а также добавить публикации: [\"22650316\", \"8986613\"]."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "f67acace-042c-4f92-89ae-b24ae2ecd008",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, level, taxonomic_span, taxonomic_span_id, description, parent]\n",
       "Index: []"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone\"\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df[histone_df[\"id\"] == \"TS_H3.4\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "647213c2-7f5b-46b0-b4b0-f9275e131964",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>32.0</td>\n",
       "      <td>H3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  id          level taxonomic_span taxonomic_span_id  \\\n",
       "153  H3.4_(Mammalia)  variant_group       Mammalia             40674   \n",
       "\n",
       "     description parent  \n",
       "153         32.0     H3  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "histone_df[histone_df[\"id\"] == \"H3.4_(Mammalia)\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "8d08c194-568f-4286-b28e-6bafafe0acbd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>33</td>\n",
       "      <td>H3T</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    id name taxonomy  gene splice          histone\n",
       "32  33  H3T     None  None   None  H3.4_(Mammalia)"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name\"\n",
    "cursor.execute(query)\n",
    "alternative_name_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ")\n",
    "alternative_name_df[alternative_name_df[\"histone\"] == \"H3.4_(Mammalia)\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "3b2de792-fe2f-47c8-8486-bb43054b5742",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>histone_id</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>336</th>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>32.0</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>kycia_tudor_2014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>337</th>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>32.0</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>dong_structural_2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338</th>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>32.0</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>kycia_tudor_2014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>32.0</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>tachiwana_nucleosome_2008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>32.0</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>tachiwana_structural_2010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>32.0</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>ueda_testis-specific_2017</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  id          level taxonomic_span taxonomic_span_id  \\\n",
       "336  H3.4_(Mammalia)  variant_group       Mammalia             40674   \n",
       "337  H3.4_(Mammalia)  variant_group       Mammalia             40674   \n",
       "338  H3.4_(Mammalia)  variant_group       Mammalia             40674   \n",
       "339  H3.4_(Mammalia)  variant_group       Mammalia             40674   \n",
       "340  H3.4_(Mammalia)  variant_group       Mammalia             40674   \n",
       "341  H3.4_(Mammalia)  variant_group       Mammalia             40674   \n",
       "\n",
       "     description parent       histone_id             publication_id  \n",
       "336         32.0     H3  H3.4_(Mammalia)           kycia_tudor_2014  \n",
       "337         32.0     H3  H3.4_(Mammalia)       dong_structural_2020  \n",
       "338         32.0     H3  H3.4_(Mammalia)           kycia_tudor_2014  \n",
       "339         32.0     H3  H3.4_(Mammalia)  tachiwana_nucleosome_2008  \n",
       "340         32.0     H3  H3.4_(Mammalia)  tachiwana_structural_2010  \n",
       "341         32.0     H3  H3.4_(Mammalia)  ueda_testis-specific_2017  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * \"\n",
    "    \"FROM histone h LEFT JOIN histone_has_publication p ON h.id = p.histone_id\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df[histone_df[\"id\"] == \"H3.4_(Mammalia)\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44bf65a0-a6d2-48e5-8952-0257a7faa2b4",
   "metadata": {},
   "source": [
    "### Add alternate names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "03ad64ce-c686-4ce5-9236-71509cc9c7af",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_alternative_name = {\n",
    "    \"name\": \"H3.1t\",\n",
    "    \"taxonomy\": \"null\",\n",
    "    \"gene\": \"null\",\n",
    "    \"splice\": \"null\",\n",
    "    \"histone\": \"H3.4_(Mammalia)\",\n",
    "}\n",
    "cursor.execute(add_alternate_names, data_alternative_name)\n",
    "\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e447607-94d9-4511-a49f-5d4323df7f8a",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Add publications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "a83419ae-2fea-48fd-8263-960fe1478cc6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "publication_list = [\"22650316\", \"8986613\"]\n",
    "query = \"SELECT id FROM publication\"\n",
    "cursor.execute(query)\n",
    "for pub in set(publication_list) - set([i[0] for i in cursor]):\n",
    "    data_publication = {\n",
    "        \"id\": pub,\n",
    "        \"title\": \"null\",\n",
    "        \"doi\": \"null\",\n",
    "        \"author\": \"null\",\n",
    "        \"year\": \"null\",\n",
    "    }\n",
    "    cursor.execute(add_publication, data_publication)\n",
    "    cursor.execute(add_histone_has_publication, (\"H3.4_(Mammalia)\", pub))\n",
    "\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "0ac57fad-e26b-4414-aee0-6dacd5737edb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "query = \"SELECT publication_id FROM histone h LEFT JOIN histone_has_publication p ON h.id = p.histone_id\"\n",
    "cursor.execute(query)\n",
    "for pub in set(publication_list) - set([i[0] for i in cursor]):\n",
    "    cursor.execute(add_histone_has_publication, (\"H3.4_(Mammalia)\", pub))\n",
    "\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "7f7c28ed-4261-4bee-ada0-e9143ba449b8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cursor.execute(add_histone_has_publication, (\"H3.4_(Mammalia)\", \"22650316\"))\n",
    "\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aad7e51b-cdca-4f08-9792-0398f511dcd5",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Checking added data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "0229bc31-323f-451c-8a40-d9887c8fab7b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>33</td>\n",
       "      <td>H3T</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>90</td>\n",
       "      <td>H3.1t</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    id   name taxonomy  gene splice          histone\n",
       "32  33    H3T     None  None   None  H3.4_(Mammalia)\n",
       "88  90  H3.1t     null  null   null  H3.4_(Mammalia)"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name\"\n",
    "cursor.execute(query)\n",
    "alternative_name_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ")\n",
    "alternative_name_df[alternative_name_df[\"histone\"] == \"H3.4_(Mammalia)\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27b45b89-deb5-40af-9dcc-f649265c34a8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>histone_id</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>336</th>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>32.0</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>kycia_tudor_2014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>337</th>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>32.0</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338</th>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>32.0</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>8986613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>32.0</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>dong_structural_2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>32.0</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>kycia_tudor_2014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>32.0</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>tachiwana_nucleosome_2008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>32.0</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>tachiwana_structural_2010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>343</th>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>32.0</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>ueda_testis-specific_2017</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  id          level taxonomic_span taxonomic_span_id  \\\n",
       "336  H3.4_(Mammalia)  variant_group       Mammalia             40674   \n",
       "337  H3.4_(Mammalia)  variant_group       Mammalia             40674   \n",
       "338  H3.4_(Mammalia)  variant_group       Mammalia             40674   \n",
       "339  H3.4_(Mammalia)  variant_group       Mammalia             40674   \n",
       "340  H3.4_(Mammalia)  variant_group       Mammalia             40674   \n",
       "341  H3.4_(Mammalia)  variant_group       Mammalia             40674   \n",
       "342  H3.4_(Mammalia)  variant_group       Mammalia             40674   \n",
       "343  H3.4_(Mammalia)  variant_group       Mammalia             40674   \n",
       "\n",
       "     description parent       histone_id             publication_id  \n",
       "336         32.0     H3  H3.4_(Mammalia)           kycia_tudor_2014  \n",
       "337         32.0     H3  H3.4_(Mammalia)                   22650316  \n",
       "338         32.0     H3  H3.4_(Mammalia)                    8986613  \n",
       "339         32.0     H3  H3.4_(Mammalia)       dong_structural_2020  \n",
       "340         32.0     H3  H3.4_(Mammalia)           kycia_tudor_2014  \n",
       "341         32.0     H3  H3.4_(Mammalia)  tachiwana_nucleosome_2008  \n",
       "342         32.0     H3  H3.4_(Mammalia)  tachiwana_structural_2010  \n",
       "343         32.0     H3  H3.4_(Mammalia)  ueda_testis-specific_2017  "
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * \"\n",
    "    \"FROM histone h LEFT JOIN histone_has_publication p ON h.id = p.histone_id\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df[histone_df[\"id\"] == \"H3.4_(Mammalia)\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0915ed7d-d7d7-470e-b59a-442167b402e7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Edit histone variant H2A.W"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "69e64035-e3c7-4523-9cdb-f20d38c69428",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>evolution</th>\n",
       "      <th>expression</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>H2A.W</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>3398</td>\n",
       "      <td>13.0</td>\n",
       "      <td>H2A</td>\n",
       "      <td>H2A.W is a plant specific variant found in ang...</td>\n",
       "      <td>H2A.W variant is found exclusively in angiospe...</td>\n",
       "      <td>Arabidopsis has three H2A.W genes (HTA6, HTA7,...</td>\n",
       "      <td>H2A.W is a plant-lineage-specific variant that...</td>\n",
       "      <td>In Arabidopsis HTA6 and HTA7 were found to hav...</td>\n",
       "      <td>None</td>\n",
       "      <td>H2A.W participates in constitutive heterochrom...</td>\n",
       "      <td>A characteristic feature of H2A.W sequences is...</td>\n",
       "      <td>Genome-wide analysis showed that H2A.W variant...</td>\n",
       "      <td>Specific deposition mechanism are not known. C...</td>\n",
       "      <td>The extended C-terminal tail of H2A.W interact...</td>\n",
       "      <td>It was hypothesized that C-terminal tail of H2...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        id          level taxonomic_span taxonomic_span_id  description  \\\n",
       "107  H2A.W  variant_group  Magnoliopsida              3398         13.0   \n",
       "\n",
       "    parent                                            summary  \\\n",
       "107    H2A  H2A.W is a plant specific variant found in ang...   \n",
       "\n",
       "                                              taxonomy  \\\n",
       "107  H2A.W variant is found exclusively in angiospe...   \n",
       "\n",
       "                                                 genes  \\\n",
       "107  Arabidopsis has three H2A.W genes (HTA6, HTA7,...   \n",
       "\n",
       "                                             evolution  \\\n",
       "107  H2A.W is a plant-lineage-specific variant that...   \n",
       "\n",
       "                                            expression knock_out  \\\n",
       "107  In Arabidopsis HTA6 and HTA7 were found to hav...      None   \n",
       "\n",
       "                                              function  \\\n",
       "107  H2A.W participates in constitutive heterochrom...   \n",
       "\n",
       "                                              sequence  \\\n",
       "107  A characteristic feature of H2A.W sequences is...   \n",
       "\n",
       "                                          localization  \\\n",
       "107  Genome-wide analysis showed that H2A.W variant...   \n",
       "\n",
       "                                            deposition  \\\n",
       "107  Specific deposition mechanism are not known. C...   \n",
       "\n",
       "                                             structure  \\\n",
       "107  The extended C-terminal tail of H2A.W interact...   \n",
       "\n",
       "                                          interactions disease caveats  \n",
       "107  It was hypothesized that C-terminal tail of H2...    None    None  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT h.*, d.summary, d.taxonomy, d.genes, d.evolution, d.expression, d.knock_out, d.function, d.sequence, d.localization, d.deposition, d.structure, d.interactions, d.disease, d.caveats \"\n",
    "    \"FROM histone h LEFT JOIN histone_description d ON h.description = d.id\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df[histone_df[\"id\"] == \"H2A.W\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "7e9cdbe3-92bd-4634-9f85-6ed5e0848858",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "knock_out_text = \"In Arabidopsis single mutants of H2A.W genes do not dispay any phenotype; double mutants, HtA6 HTA7 and HTA6 HTA12, and triple mutants result in growth defects that were even more severe in the triple mutant. This suggests that the three H2A.W paralogs in Arabidopsis are functionally redundant [alvarez-venegas_canonical_2019,yelagandula_histone_2014]. HTA7 knock-out in Arabidopsis results in sensitivity to genotoxic agents [lorkovic_compartmentalization_2017].\"\n",
    "query = f\"UPDATE histone_description SET knock_out='{knock_out_text}' WHERE id=13\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ecb3f81-a968-4f1d-9b08-8ae1abcc7b8c",
   "metadata": {},
   "source": [
    "### Checking added data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "5f409ed7-f28b-4e12-b6fe-86e4a984c571",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>evolution</th>\n",
       "      <th>expression</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>H2A.W</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>3398</td>\n",
       "      <td>13.0</td>\n",
       "      <td>H2A</td>\n",
       "      <td>H2A.W is a plant specific variant found in ang...</td>\n",
       "      <td>H2A.W variant is found exclusively in angiospe...</td>\n",
       "      <td>Arabidopsis has three H2A.W genes (HTA6, HTA7,...</td>\n",
       "      <td>H2A.W is a plant-lineage-specific variant that...</td>\n",
       "      <td>In Arabidopsis HTA6 and HTA7 were found to hav...</td>\n",
       "      <td>In Arabidopsis single mutants of H2A.W genes d...</td>\n",
       "      <td>H2A.W participates in constitutive heterochrom...</td>\n",
       "      <td>A characteristic feature of H2A.W sequences is...</td>\n",
       "      <td>Genome-wide analysis showed that H2A.W variant...</td>\n",
       "      <td>Specific deposition mechanism are not known. C...</td>\n",
       "      <td>The extended C-terminal tail of H2A.W interact...</td>\n",
       "      <td>It was hypothesized that C-terminal tail of H2...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        id          level taxonomic_span taxonomic_span_id  description  \\\n",
       "107  H2A.W  variant_group  Magnoliopsida              3398         13.0   \n",
       "\n",
       "    parent                                            summary  \\\n",
       "107    H2A  H2A.W is a plant specific variant found in ang...   \n",
       "\n",
       "                                              taxonomy  \\\n",
       "107  H2A.W variant is found exclusively in angiospe...   \n",
       "\n",
       "                                                 genes  \\\n",
       "107  Arabidopsis has three H2A.W genes (HTA6, HTA7,...   \n",
       "\n",
       "                                             evolution  \\\n",
       "107  H2A.W is a plant-lineage-specific variant that...   \n",
       "\n",
       "                                            expression  \\\n",
       "107  In Arabidopsis HTA6 and HTA7 were found to hav...   \n",
       "\n",
       "                                             knock_out  \\\n",
       "107  In Arabidopsis single mutants of H2A.W genes d...   \n",
       "\n",
       "                                              function  \\\n",
       "107  H2A.W participates in constitutive heterochrom...   \n",
       "\n",
       "                                              sequence  \\\n",
       "107  A characteristic feature of H2A.W sequences is...   \n",
       "\n",
       "                                          localization  \\\n",
       "107  Genome-wide analysis showed that H2A.W variant...   \n",
       "\n",
       "                                            deposition  \\\n",
       "107  Specific deposition mechanism are not known. C...   \n",
       "\n",
       "                                             structure  \\\n",
       "107  The extended C-terminal tail of H2A.W interact...   \n",
       "\n",
       "                                          interactions disease caveats  \n",
       "107  It was hypothesized that C-terminal tail of H2...    None    None  "
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT h.*, d.summary, d.taxonomy, d.genes, d.evolution, d.expression, d.knock_out, d.function, d.sequence, d.localization, d.deposition, d.structure, d.interactions, d.disease, d.caveats \"\n",
    "    \"FROM histone h LEFT JOIN histone_description d ON h.description = d.id\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df[histone_df[\"id\"] == \"H2A.W\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b10aa262-bc85-4808-b730-4a52d396c058",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Edit histone variant H3.3_(Animals)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "7dc8b93c-f20f-4fb2-a59d-f2f3c07949d4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>evolution</th>\n",
       "      <th>expression</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>H3.3_(Animals)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>33208</td>\n",
       "      <td>70.0</td>\n",
       "      <td>H3.3</td>\n",
       "      <td>H3.3_(Animals)  is a class of replication-inde...</td>\n",
       "      <td>Animals (Metazoa)</td>\n",
       "      <td>In human H3.3 is encoded by H3-3A and H3-3B ge...</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3 mRNAs are polyadenilated. Expressed indep...</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3 has been implicated in a variety of biolo...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>In humans and other animals specialized chaper...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Mutations in histone genes can affect sites of...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "147  H3.3_(Animals)  variant        Metazoa             33208         70.0   \n",
       "\n",
       "    parent                                            summary  \\\n",
       "147   H3.3  H3.3_(Animals)  is a class of replication-inde...   \n",
       "\n",
       "              taxonomy                                              genes  \\\n",
       "147  Animals (Metazoa)  In human H3.3 is encoded by H3-3A and H3-3B ge...   \n",
       "\n",
       "    evolution                                         expression knock_out  \\\n",
       "147      None  H3.3 mRNAs are polyadenilated. Expressed indep...      None   \n",
       "\n",
       "                                              function sequence localization  \\\n",
       "147  H3.3 has been implicated in a variety of biolo...     None         None   \n",
       "\n",
       "                                            deposition structure interactions  \\\n",
       "147  In humans and other animals specialized chaper...      None         None   \n",
       "\n",
       "                                               disease caveats  \n",
       "147  Mutations in histone genes can affect sites of...    None  "
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT h.*, d.summary, d.taxonomy, d.genes, d.evolution, d.expression, d.knock_out, d.function, d.sequence, d.localization, d.deposition, d.structure, d.interactions, d.disease, d.caveats \"\n",
    "    \"FROM histone h LEFT JOIN histone_description d ON h.description = d.id\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df[histone_df[\"id\"] == \"H3.3_(Animals)\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "fe976192-c4cc-4baa-8d69-551c493d411c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "knock_out_text = \"In mice complete depletion of H3.3 leads to developmental retardation and early embryonic lethality. At the cellular level, H3.3 loss triggers cell cycle suppression and cell death. Surprisingly, H3.3 depletion does not dramatically disrupt gene regulation in the developing embryo. Instead, H3.3 depletion causes dysfunction of heterochromatin structures at telomeres, centromeres, and pericentromeric regions of chromosomes, leading to mitotic defects [jang_histone_2015]. In C. elegans Despite these specific expression patterns, we find that neither loss of individual H3.3 homologs nor the knockout of all five H3.3-coding genes causes sterility or lethality. However, we demonstrate an essential role for the conserved histone chaperone HIRA in the nucleosomal loading of all H3.3 variants. This requirement can be bypassed by mutation of the H3.3-specific residues to those found in H3. While even removal of all H3.3 homologs does not result in lethality, it leads to reduced fertility and viability in response to high-temperature stress [delaney_differential_2018]. Flies that lack both H3.3 genes have reduced viability and individuals that survive to adulthood are completely sterile in both sexes [sakai_transcriptional_2009]. Targeted disruption of one gene (H3f3b) results in a number of phenotypic abnormalities, including a reduction in H3.3 histone levels, leading to male infertility, as well as abnormal sperm and testes morphology. Additionally, null germ cell populations at specific stages in spermatogenesis, in particular spermatocytes and spermatogonia, exhibited increased rates of apoptosis. Disruption of H3f3b also altered histone post-translational modifications and gene expression in the testes, with the most prominent changes occurring at genes involved in spermatogenesis. Finally, H3f3b null testes also exhibited abnormal germ cell chromatin reorganization and reduced protamine incorporation [yuen_histone_2014].\"\n",
    "query = f\"UPDATE histone_description SET knock_out='{knock_out_text}' WHERE id=70\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cccaf34-c3f3-45f6-8dbf-74e5393255ef",
   "metadata": {},
   "source": [
    "### Checking added data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "ed9ce61f-e4c7-4bff-b01d-1aa4996f4881",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>evolution</th>\n",
       "      <th>expression</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>H3.3_(Animals)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>33208</td>\n",
       "      <td>70.0</td>\n",
       "      <td>H3.3</td>\n",
       "      <td>H3.3_(Animals)  is a class of replication-inde...</td>\n",
       "      <td>Animals (Metazoa)</td>\n",
       "      <td>In human H3.3 is encoded by H3-3A and H3-3B ge...</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3 mRNAs are polyadenilated. Expressed indep...</td>\n",
       "      <td>In mice complete depletion of H3.3 leads to de...</td>\n",
       "      <td>H3.3 has been implicated in a variety of biolo...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>In humans and other animals specialized chaper...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Mutations in histone genes can affect sites of...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "147  H3.3_(Animals)  variant        Metazoa             33208         70.0   \n",
       "\n",
       "    parent                                            summary  \\\n",
       "147   H3.3  H3.3_(Animals)  is a class of replication-inde...   \n",
       "\n",
       "              taxonomy                                              genes  \\\n",
       "147  Animals (Metazoa)  In human H3.3 is encoded by H3-3A and H3-3B ge...   \n",
       "\n",
       "    evolution                                         expression  \\\n",
       "147      None  H3.3 mRNAs are polyadenilated. Expressed indep...   \n",
       "\n",
       "                                             knock_out  \\\n",
       "147  In mice complete depletion of H3.3 leads to de...   \n",
       "\n",
       "                                              function sequence localization  \\\n",
       "147  H3.3 has been implicated in a variety of biolo...     None         None   \n",
       "\n",
       "                                            deposition structure interactions  \\\n",
       "147  In humans and other animals specialized chaper...      None         None   \n",
       "\n",
       "                                               disease caveats  \n",
       "147  Mutations in histone genes can affect sites of...    None  "
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT h.*, d.summary, d.taxonomy, d.genes, d.evolution, d.expression, d.knock_out, d.function, d.sequence, d.localization, d.deposition, d.structure, d.interactions, d.disease, d.caveats \"\n",
    "    \"FROM histone h LEFT JOIN histone_description d ON h.description = d.id\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df[histone_df[\"id\"] == \"H3.3_(Animals)\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3645b603-53ef-48f7-8f28-d7ce78b21b40",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Edit histone variant H3.3_(Plants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "ba0f44dc-44f3-4941-be76-8af0f2263c09",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>evolution</th>\n",
       "      <th>expression</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>H3.3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Viridiplantae</td>\n",
       "      <td>33090</td>\n",
       "      <td>71.0</td>\n",
       "      <td>H3.3</td>\n",
       "      <td>H3.3_(Plants)  is a class of replication-indep...</td>\n",
       "      <td>Green plants (Viridiplantae)</td>\n",
       "      <td>In contrast to intronless cH3s, H3.3 in plants...</td>\n",
       "      <td>None</td>\n",
       "      <td>In Arabidopsis, H3.3 genes do not show replica...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Characteristic changes between cH3 and H3.3 in...</td>\n",
       "      <td>None</td>\n",
       "      <td>Similar to animals, plants have a specialized ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "151  H3.3_(Plants)  variant  Viridiplantae             33090         71.0   \n",
       "\n",
       "    parent                                            summary  \\\n",
       "151   H3.3  H3.3_(Plants)  is a class of replication-indep...   \n",
       "\n",
       "                         taxonomy  \\\n",
       "151  Green plants (Viridiplantae)   \n",
       "\n",
       "                                                 genes evolution  \\\n",
       "151  In contrast to intronless cH3s, H3.3 in plants...      None   \n",
       "\n",
       "                                            expression knock_out function  \\\n",
       "151  In Arabidopsis, H3.3 genes do not show replica...      None     None   \n",
       "\n",
       "                                              sequence localization  \\\n",
       "151  Characteristic changes between cH3 and H3.3 in...         None   \n",
       "\n",
       "                                            deposition structure interactions  \\\n",
       "151  Similar to animals, plants have a specialized ...      None         None   \n",
       "\n",
       "    disease caveats  \n",
       "151    None    None  "
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT h.*, d.summary, d.taxonomy, d.genes, d.evolution, d.expression, d.knock_out, d.function, d.sequence, d.localization, d.deposition, d.structure, d.interactions, d.disease, d.caveats \"\n",
    "    \"FROM histone h LEFT JOIN histone_description d ON h.description = d.id\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df[histone_df[\"id\"] == \"H3.3_(Plants)\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "513c1382-9109-4f18-ac33-26d817e070f5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "knock_out_text = \"In Arabidopsis, removal of three H3.3 genes (HTR4, HTR5,and HTR8) causes defects in male gametogenesis and results in embryonic lethality [wollmann_histone_2017].\"\n",
    "query = f\"UPDATE histone_description SET knock_out='{knock_out_text}' WHERE id=71\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "139dc919-e429-4354-8c11-fbb900b9482e",
   "metadata": {},
   "source": [
    "### Checking added data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "bfca7bfd-5108-4126-b21c-73e65102dbf5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>evolution</th>\n",
       "      <th>expression</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>H3.3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Viridiplantae</td>\n",
       "      <td>33090</td>\n",
       "      <td>71.0</td>\n",
       "      <td>H3.3</td>\n",
       "      <td>H3.3_(Plants)  is a class of replication-indep...</td>\n",
       "      <td>Green plants (Viridiplantae)</td>\n",
       "      <td>In contrast to intronless cH3s, H3.3 in plants...</td>\n",
       "      <td>None</td>\n",
       "      <td>In Arabidopsis, H3.3 genes do not show replica...</td>\n",
       "      <td>In Arabidopsis, removal of three H3.3 genes (H...</td>\n",
       "      <td>None</td>\n",
       "      <td>Characteristic changes between cH3 and H3.3 in...</td>\n",
       "      <td>None</td>\n",
       "      <td>Similar to animals, plants have a specialized ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "151  H3.3_(Plants)  variant  Viridiplantae             33090         71.0   \n",
       "\n",
       "    parent                                            summary  \\\n",
       "151   H3.3  H3.3_(Plants)  is a class of replication-indep...   \n",
       "\n",
       "                         taxonomy  \\\n",
       "151  Green plants (Viridiplantae)   \n",
       "\n",
       "                                                 genes evolution  \\\n",
       "151  In contrast to intronless cH3s, H3.3 in plants...      None   \n",
       "\n",
       "                                            expression  \\\n",
       "151  In Arabidopsis, H3.3 genes do not show replica...   \n",
       "\n",
       "                                             knock_out function  \\\n",
       "151  In Arabidopsis, removal of three H3.3 genes (H...     None   \n",
       "\n",
       "                                              sequence localization  \\\n",
       "151  Characteristic changes between cH3 and H3.3 in...         None   \n",
       "\n",
       "                                            deposition structure interactions  \\\n",
       "151  Similar to animals, plants have a specialized ...      None         None   \n",
       "\n",
       "    disease caveats  \n",
       "151    None    None  "
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT h.*, d.summary, d.taxonomy, d.genes, d.evolution, d.expression, d.knock_out, d.function, d.sequence, d.localization, d.deposition, d.structure, d.interactions, d.disease, d.caveats \"\n",
    "    \"FROM histone h LEFT JOIN histone_description d ON h.description = d.id\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df[histone_df[\"id\"] == \"H3.3_(Plants)\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47a27e81-42de-4c8e-a23a-f56915af5f0b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Edit histone variant H3.4_(Mammalia)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "703f2882-b9d1-41cc-910a-4fa020027dc0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>evolution</th>\n",
       "      <th>expression</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>32.0</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.4_(Mammalia)  is a mammal-specific H3 histo...</td>\n",
       "      <td>Mammals</td>\n",
       "      <td>In human is encoded by H3-4 gene (formerly HIS...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PDB structure of human H3.4 containing nucleos...</td>\n",
       "      <td>None</td>\n",
       "      <td>The single‑nucleotide polymorphism c190C&gt;T (Ar...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  id          level taxonomic_span taxonomic_span_id  \\\n",
       "153  H3.4_(Mammalia)  variant_group       Mammalia             40674   \n",
       "\n",
       "     description parent                                            summary  \\\n",
       "153         32.0     H3  H3.4_(Mammalia)  is a mammal-specific H3 histo...   \n",
       "\n",
       "    taxonomy                                              genes evolution  \\\n",
       "153  Mammals  In human is encoded by H3-4 gene (formerly HIS...      None   \n",
       "\n",
       "    expression knock_out function sequence localization deposition  \\\n",
       "153       None      None     None     None         None       None   \n",
       "\n",
       "                                             structure interactions  \\\n",
       "153  PDB structure of human H3.4 containing nucleos...         None   \n",
       "\n",
       "                                               disease caveats  \n",
       "153  The single‑nucleotide polymorphism c190C>T (Ar...    None  "
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT h.*, d.summary, d.taxonomy, d.genes, d.evolution, d.expression, d.knock_out, d.function, d.sequence, d.localization, d.deposition, d.structure, d.interactions, d.disease, d.caveats \"\n",
    "    \"FROM histone h LEFT JOIN histone_description d ON h.description = d.id\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df[histone_df[\"id\"] == \"H3.4_(Mammalia)\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "e5befe59-fabb-43d8-ac9c-0877ae65cf29",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "knock_out_text = \"Knockout mice for H3.4 were first generated in 2017; both male and female H3t null mice were viable and healthy, but the male mice were sterile. H3.4 deficiency leads to azoospermia because of the loss of haploid germ cells[ueda_testis-specific_2017].\"\n",
    "query = f\"UPDATE histone_description SET knock_out='{knock_out_text}' WHERE id=32\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cb0ccf4-e4da-402b-9c5f-615878513b0a",
   "metadata": {},
   "source": [
    "### Checking added data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "e77c444c-0062-4999-bc72-830772c0b906",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>evolution</th>\n",
       "      <th>expression</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>32.0</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.4_(Mammalia)  is a mammal-specific H3 histo...</td>\n",
       "      <td>Mammals</td>\n",
       "      <td>In human is encoded by H3-4 gene (formerly HIS...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Knockout mice for H3.4 were first generated in...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PDB structure of human H3.4 containing nucleos...</td>\n",
       "      <td>None</td>\n",
       "      <td>The single‑nucleotide polymorphism c190C&gt;T (Ar...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  id          level taxonomic_span taxonomic_span_id  \\\n",
       "153  H3.4_(Mammalia)  variant_group       Mammalia             40674   \n",
       "\n",
       "     description parent                                            summary  \\\n",
       "153         32.0     H3  H3.4_(Mammalia)  is a mammal-specific H3 histo...   \n",
       "\n",
       "    taxonomy                                              genes evolution  \\\n",
       "153  Mammals  In human is encoded by H3-4 gene (formerly HIS...      None   \n",
       "\n",
       "    expression                                          knock_out function  \\\n",
       "153       None  Knockout mice for H3.4 were first generated in...     None   \n",
       "\n",
       "    sequence localization deposition  \\\n",
       "153     None         None       None   \n",
       "\n",
       "                                             structure interactions  \\\n",
       "153  PDB structure of human H3.4 containing nucleos...         None   \n",
       "\n",
       "                                               disease caveats  \n",
       "153  The single‑nucleotide polymorphism c190C>T (Ar...    None  "
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT h.*, d.summary, d.taxonomy, d.genes, d.evolution, d.expression, d.knock_out, d.function, d.sequence, d.localization, d.deposition, d.structure, d.interactions, d.disease, d.caveats \"\n",
    "    \"FROM histone h LEFT JOIN histone_description d ON h.description = d.id\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df[histone_df[\"id\"] == \"H3.4_(Mammalia)\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de2712dc-cf76-4cfd-af48-b8e8afd0b497",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# For csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "52a01444-7419-4e05-986c-2588e79f8760",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "pattern = r\"__\\?\\?\\?\"  # Шаблон для поиска '__???'\n",
    "sequence_df[sequence_df[\"variant\"].str.contains(pattern, regex=True)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ec205066-c821-49ed-9349-44933e63656b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(53, 13)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df[\n",
    "    sequence_df[\"variant_under_consideration\"].str.contains(pattern, regex=True)\n",
    "].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b70bf805-beee-4ea7-99e0-65b3736fb8f1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28c756c2-da06-4f0a-9229-5714d7af25c0",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Remove histone sequences classified as clustered H2B(?), H3.6(?), H3.7(?), H3.8(?), clustered H4(?) (these histones are putative)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a9d61b44-be46-4822-a326-a7bfe6d54430",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>NP_059141.1</td>\n",
       "      <td></td>\n",
       "      <td>cH2B(?)_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession variant  variant_under_consideration\n",
       "245  NP_059141.1          cH2B(?)_(Homo_sapiens)__???"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df[sequence_df[\"variant_under_consideration\"] == \"cH2B(?)_(Homo_sapiens)__???\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "83c7d1b8-0b66-412e-817d-11fb1b9be189",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "query = \"DELETE FROM sequence WHERE variant_under_consideration='cH2B(?)_(Homo_sapiens)__???'\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "cb33db86-6b3a-4a90-85f8-f46adaa326df",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "]\n",
    "sequence_df[sequence_df[\"variant_under_consideration\"] == \"cH2B(?)_(Homo_sapiens)__???\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c5c1f2d9-b1a8-46db-b8d2-e21b662efe9e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>NP_001342338.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.7(?)_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession variant  variant_under_consideration\n",
       "167  NP_001342338.1          H3.7(?)_(Homo_sapiens)__???"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df[sequence_df[\"variant_under_consideration\"] == \"H3.7(?)_(Homo_sapiens)__???\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "90b2fd2d-3448-4def-9eec-ac898163af86",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NP_001342338.1</td>\n",
       "      <td>12408966</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sequence_accession publication_id\n",
       "9     NP_001342338.1       12408966"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence_has_publication\"\n",
    "cursor.execute(query)\n",
    "sp_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "sp_df[sp_df[\"sequence_accession\"] == \"NP_001342338.1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "bd5fa1c1-c968-4ab9-9156-c61fca5d502b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "query = (\n",
    "    \"DELETE FROM sequence_has_publication WHERE sequence_accession='NP_001342338.1'; \"\n",
    "    \"DELETE FROM sequence WHERE variant_under_consideration='H3.7(?)_(Homo_sapiens)__???'\"\n",
    ")\n",
    "cursor.execute(query, multi=True)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "92a7c83c-13d8-4f84-b1a1-ff3dddba92aa",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "]\n",
    "sequence_df[sequence_df[\"variant_under_consideration\"] == \"H3.7(?)_(Homo_sapiens)__???\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3f060ef2-d7d1-4cf6-8dea-9ad5bcb327ab",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [sequence_accession, publication_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence_has_publication\"\n",
    "cursor.execute(query)\n",
    "sp_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "sp_df[sp_df[\"sequence_accession\"] == \"NP_001342338.1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3af0f2d8-ec93-4cd3-be1c-642bf81f8e53",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>NP_003538.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4(?)_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession variant variant_under_consideration\n",
       "214  NP_003538.1          cH4(?)_(Homo_sapiens)__???"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df[sequence_df[\"variant_under_consideration\"] == \"cH4(?)_(Homo_sapiens)__???\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5bf19f9e-a6a1-4e21-acc7-bd7efd8f2b78",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>NP_003538.1</td>\n",
       "      <td>12408966</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sequence_accession publication_id\n",
       "48        NP_003538.1       12408966"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence_has_publication\"\n",
    "cursor.execute(query)\n",
    "sp_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "sp_df[sp_df[\"sequence_accession\"] == \"NP_003538.1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "28e07a1b-e611-4329-8b24-e7e6fa87e1af",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DELETE FROM sequence_has_publication WHERE sequence_accession='NP_003538.1'\n"
     ]
    }
   ],
   "source": [
    "query = \"DELETE FROM sequence_has_publication WHERE sequence_accession='NP_003538.1'\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "eb62d78f-ef73-4d63-8f9c-a785a13359b7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "query = \"DELETE FROM sequence WHERE variant_under_consideration='cH4(?)_(Homo_sapiens)__???'\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "691cd0ea-0c82-4f11-8a7a-4254d4ef92a2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "]\n",
    "sequence_df[sequence_df[\"variant_under_consideration\"] == \"cH4(?)_(Homo_sapiens)__???\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "7cc1db1d-4ff5-49ca-82ed-a48fc1b2f34c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [sequence_accession, publication_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence_has_publication\"\n",
    "cursor.execute(query)\n",
    "sp_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "sp_df[sp_df[\"sequence_accession\"] == \"NP_003538.1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1146255-030e-49fa-bf7f-1ee03e2176f0",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Update other sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "3cdf98f4-7f28-42cf-ad0d-7d9d69698709",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>NP_005309.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.0_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>NP_006017.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.10_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>NP_005316.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.1_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>NP_005310.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.2_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>NP_005311.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.3_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>NP_005312.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>NP_005313.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.5_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>NP_005314.2</td>\n",
       "      <td></td>\n",
       "      <td>H1.6_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>NP_861453.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.7_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>NP_722575.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.8_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>NP_001295191.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.8_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>NP_003484.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>NP_001013721.2</td>\n",
       "      <td></td>\n",
       "      <td>H3.5_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468</th>\n",
       "      <td>XP_003954426.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.5__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>NP_001342187.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y.1_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>NP_001358848.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y.2_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>HISTDB_H3_Y_3</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>HISTDB_H3_Y_1</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>HISTDB_H3_Y_0</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>HISTDB_H3_Y_2</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467</th>\n",
       "      <td>XP_003804825.1</td>\n",
       "      <td></td>\n",
       "      <td>TS_H3.4__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>NP_068803.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>NP_003539.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>NP_003535.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>NP_003536.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>NP_001029249.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>NP_003537.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>NP_003486.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>NP_003529.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>NP_778224.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>NP_003531.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>NP_003532.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>NP_003533.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>NP_003534.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>NP_003530.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>NP_001035891.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>NP_001800.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>416</th>\n",
       "      <td>XP_002767160.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>401</th>\n",
       "      <td>XP_002287626.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369</th>\n",
       "      <td>XP_001350068.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>XP_001011273.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AAK39657.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293</th>\n",
       "      <td>NP_563627.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>NP_523730.2</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>NP_499128.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>NP_031707.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>NP_012875.2</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>475</th>\n",
       "      <td>XP_005535607.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>NP_596473.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>503</th>\n",
       "      <td>XP_009526809.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession variant variant_under_consideration\n",
       "215     NP_005309.1            H1.0_(Homo_sapiens)__???\n",
       "223     NP_006017.1           H1.10_(Homo_sapiens)__???\n",
       "222     NP_005316.1            H1.1_(Homo_sapiens)__???\n",
       "216     NP_005310.1            H1.2_(Homo_sapiens)__???\n",
       "217     NP_005311.1            H1.3_(Homo_sapiens)__???\n",
       "218     NP_005312.1            H1.4_(Homo_sapiens)__???\n",
       "219     NP_005313.1            H1.5_(Homo_sapiens)__???\n",
       "220     NP_005314.2            H1.6_(Homo_sapiens)__???\n",
       "320     NP_861453.1            H1.7_(Homo_sapiens)__???\n",
       "306     NP_722575.1            H1.8_(Homo_sapiens)__???\n",
       "164  NP_001295191.1            H1.8_(Homo_sapiens)__???\n",
       "175     NP_003484.1            H3.4_(Homo_sapiens)__???\n",
       "99   NP_001013721.2            H3.5_(Homo_sapiens)__???\n",
       "468  XP_003954426.1                           H3.5__???\n",
       "166  NP_001342187.1          H3.Y.1_(Homo_sapiens)__???\n",
       "168  NP_001358848.1          H3.Y.2_(Homo_sapiens)__???\n",
       "94    HISTDB_H3_Y_3                           H3.Y__???\n",
       "92    HISTDB_H3_Y_1                           H3.Y__???\n",
       "91    HISTDB_H3_Y_0                           H3.Y__???\n",
       "93    HISTDB_H3_Y_2                           H3.Y__???\n",
       "467  XP_003804825.1                        TS_H3.4__???\n",
       "255     NP_068803.1             cH4_(Homo_sapiens)__???\n",
       "214     NP_003539.1             cH4_(Homo_sapiens)__???\n",
       "211     NP_003535.1             cH4_(Homo_sapiens)__???\n",
       "212     NP_003536.1             cH4_(Homo_sapiens)__???\n",
       "107  NP_001029249.1             cH4_(Homo_sapiens)__???\n",
       "213     NP_003537.1             cH4_(Homo_sapiens)__???\n",
       "176     NP_003486.1             cH4_(Homo_sapiens)__???\n",
       "205     NP_003529.1             cH4_(Homo_sapiens)__???\n",
       "311     NP_778224.1             cH4_(Homo_sapiens)__???\n",
       "207     NP_003531.1             cH4_(Homo_sapiens)__???\n",
       "208     NP_003532.1             cH4_(Homo_sapiens)__???\n",
       "209     NP_003533.1             cH4_(Homo_sapiens)__???\n",
       "210     NP_003534.1             cH4_(Homo_sapiens)__???\n",
       "206     NP_003530.1             cH4_(Homo_sapiens)__???\n",
       "112  NP_001035891.1           cenH3_(Homo_sapiens)__???\n",
       "171     NP_001800.1           cenH3_(Homo_sapiens)__???\n",
       "416  XP_002767160.1                          cenH3__???\n",
       "401  XP_002287626.1                          cenH3__???\n",
       "369  XP_001350068.1                          cenH3__???\n",
       "361  XP_001011273.1                          cenH3__???\n",
       "5        AAK39657.1                          cenH3__???\n",
       "293     NP_563627.1                          cenH3__???\n",
       "289     NP_523730.2                          cenH3__???\n",
       "284     NP_499128.1                          cenH3__???\n",
       "233     NP_031707.1                          cenH3__???\n",
       "230     NP_012875.2                          cenH3__???\n",
       "475  XP_005535607.1                          cenH3__???\n",
       "298     NP_596473.1                          cenH3__???\n",
       "503  XP_009526809.1                          cenH3__???"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df = sequence_df.sort_values(by=[\"variant_under_consideration\"])\n",
    "sequence_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "141ca51d-4d72-4241-bf17-6b5b7162e42c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"H1.0_(Homo_sapiens)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Homo sapiens\",\n",
    "    \"taxonomic_span_id\": \"9606\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"H1.0\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "query = f\"UPDATE sequence SET variant='H1.0_(Homo_sapiens)', variant_under_consideration=null WHERE accession='NP_005309.1'\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "7a921177-ff77-4d95-9699-ff695e4b2727",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_005309.1</td>\n",
       "      <td>H1.0_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>3005</td>\n",
       "      <td>H1-0</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MTENSTSAPAAKPKRAKASKKSTDHPKYSDMIVAAIQAEKNRAGSS...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession              variant gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  NP_005309.1  H1.0_(Homo_sapiens)            3005           H1-0   \n",
       "\n",
       "   taxonomy_id      organism    phylum     class taxonomy_group info  \\\n",
       "0         9606  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MTENSTSAPAAKPKRAKASKKSTDHPKYSDMIVAAIQAEKNRAGSS...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                              "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE accession='NP_005309.1'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "29794af1-adad-4710-8105-3e82955ba779",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H1.0_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td></td>\n",
       "      <td>H1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  H1.0_(Homo_sapiens)  variant   Homo sapiens              9606               \n",
       "\n",
       "  parent  \n",
       "0   H1.0  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone WHERE id='H1.0_(Homo_sapiens)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "64356b8c-0776-4b42-a059-0a6f0d0a3eee",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>NP_006017.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.10_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>NP_005316.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.1_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>NP_005310.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.2_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>NP_005311.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.3_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>NP_005312.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession variant variant_under_consideration\n",
       "223  NP_006017.1           H1.10_(Homo_sapiens)__???\n",
       "222  NP_005316.1            H1.1_(Homo_sapiens)__???\n",
       "216  NP_005310.1            H1.2_(Homo_sapiens)__???\n",
       "217  NP_005311.1            H1.3_(Homo_sapiens)__???\n",
       "218  NP_005312.1            H1.4_(Homo_sapiens)__???"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "].sort_values(by=[\"variant_under_consideration\"])\n",
    "sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "517b1f73-0d32-4c91-bdfe-9e9354dd4ea5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"H1.10_(Homo_sapiens)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Homo sapiens\",\n",
    "    \"taxonomic_span_id\": \"9606\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"H1.10\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "query = f\"UPDATE sequence SET variant='H1.10_(Homo_sapiens)', variant_under_consideration=null WHERE accession='NP_006017.1'\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "057bbfbe-5fd9-41a6-a4ab-e10a9f31d472",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_006017.1</td>\n",
       "      <td>H1.10_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>8971</td>\n",
       "      <td>H1-10</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MSVELEEALPVTTAEGMAKKVTKAGGSAALSPSKKRKNSKKKNQPG...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession               variant gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  NP_006017.1  H1.10_(Homo_sapiens)            8971          H1-10   \n",
       "\n",
       "   taxonomy_id      organism    phylum     class taxonomy_group info  \\\n",
       "0         9606  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MSVELEEALPVTTAEGMAKKVTKAGGSAALSPSKKRKNSKKKNQPG...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                              "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE accession='NP_006017.1'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "e648412e-98a9-4383-841d-6def0d0c0320",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H1.10_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td></td>\n",
       "      <td>H1.10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  H1.10_(Homo_sapiens)  variant   Homo sapiens              9606               \n",
       "\n",
       "  parent  \n",
       "0  H1.10  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone WHERE id='H1.10_(Homo_sapiens)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "132c510f-3a58-4280-a6bc-234ce20bba13",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>NP_005316.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.1_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>NP_005310.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.2_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>NP_005311.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.3_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>NP_005312.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>NP_005313.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.5_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession variant variant_under_consideration\n",
       "222  NP_005316.1            H1.1_(Homo_sapiens)__???\n",
       "216  NP_005310.1            H1.2_(Homo_sapiens)__???\n",
       "217  NP_005311.1            H1.3_(Homo_sapiens)__???\n",
       "218  NP_005312.1            H1.4_(Homo_sapiens)__???\n",
       "219  NP_005313.1            H1.5_(Homo_sapiens)__???"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "].sort_values(by=[\"variant_under_consideration\"])\n",
    "sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "f3683e3d-a1b1-4731-a576-0639ac45ef5f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"H1.1_(Homo_sapiens)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Homo sapiens\",\n",
    "    \"taxonomic_span_id\": \"9606\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"H1.1\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "query = f\"UPDATE sequence SET variant='H1.1_(Homo_sapiens)', variant_under_consideration=null WHERE accession='NP_005316.1'\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "6a95b628-e821-423a-9474-09b9796c7198",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_005316.1</td>\n",
       "      <td>H1.1_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>3024</td>\n",
       "      <td>H1-1</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MSETVPPAPAASAAPEKPLAGKKAKKPAKAAAASKKKPAGPSVSEL...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession              variant gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  NP_005316.1  H1.1_(Homo_sapiens)            3024           H1-1   \n",
       "\n",
       "   taxonomy_id      organism    phylum     class taxonomy_group info  \\\n",
       "0         9606  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MSETVPPAPAASAAPEKPLAGKKAKKPAKAAAASKKKPAGPSVSEL...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                              "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE accession='NP_005316.1'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "6f08558f-4613-4b49-bc6f-2e09f13653ea",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H1.1_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td></td>\n",
       "      <td>H1.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  H1.1_(Homo_sapiens)  variant   Homo sapiens              9606               \n",
       "\n",
       "  parent  \n",
       "0   H1.1  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone WHERE id='H1.1_(Homo_sapiens)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "95bd0897-fe56-46e9-8fb4-45a5bb31f7a4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>NP_005310.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.2_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>NP_005311.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.3_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>NP_005312.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>NP_005313.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.5_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>NP_005314.2</td>\n",
       "      <td></td>\n",
       "      <td>H1.6_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession variant variant_under_consideration\n",
       "216  NP_005310.1            H1.2_(Homo_sapiens)__???\n",
       "217  NP_005311.1            H1.3_(Homo_sapiens)__???\n",
       "218  NP_005312.1            H1.4_(Homo_sapiens)__???\n",
       "219  NP_005313.1            H1.5_(Homo_sapiens)__???\n",
       "220  NP_005314.2            H1.6_(Homo_sapiens)__???"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "].sort_values(by=[\"variant_under_consideration\"])\n",
    "sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "6b2d3114-d2b9-4b31-97cd-2f0ca9a6ee49",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"H1.2_(Homo_sapiens)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Homo sapiens\",\n",
    "    \"taxonomic_span_id\": \"9606\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"H1.2\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "query = f\"UPDATE sequence SET variant='H1.2_(Homo_sapiens)', variant_under_consideration=null WHERE accession='NP_005310.1'\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "10b92ec7-14ec-4b30-a03f-ff74907c9f56",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_005310.1</td>\n",
       "      <td>H1.2_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>3006</td>\n",
       "      <td>H1-2</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MSETAPAAPAAAPPAEKAPVKKKAAKKAGGTPRKASGPPVSELITK...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession              variant gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  NP_005310.1  H1.2_(Homo_sapiens)            3006           H1-2   \n",
       "\n",
       "   taxonomy_id      organism    phylum     class taxonomy_group info  \\\n",
       "0         9606  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MSETAPAAPAAAPPAEKAPVKKKAAKKAGGTPRKASGPPVSELITK...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                              "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE accession='NP_005310.1'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "db4a81ff-9123-4248-b72d-3cf36fa6b217",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H1.2_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td></td>\n",
       "      <td>H1.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  H1.2_(Homo_sapiens)  variant   Homo sapiens              9606               \n",
       "\n",
       "  parent  \n",
       "0   H1.2  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone WHERE id='H1.2_(Homo_sapiens)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "b0525646-bc12-4dc4-9838-bfefad6a0cc3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>NP_005311.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.3_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>NP_005312.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>NP_005313.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.5_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>NP_005314.2</td>\n",
       "      <td></td>\n",
       "      <td>H1.6_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>NP_861453.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.7_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession variant variant_under_consideration\n",
       "217  NP_005311.1            H1.3_(Homo_sapiens)__???\n",
       "218  NP_005312.1            H1.4_(Homo_sapiens)__???\n",
       "219  NP_005313.1            H1.5_(Homo_sapiens)__???\n",
       "220  NP_005314.2            H1.6_(Homo_sapiens)__???\n",
       "320  NP_861453.1            H1.7_(Homo_sapiens)__???"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "].sort_values(by=[\"variant_under_consideration\"])\n",
    "sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "e0f8ca77-1a50-4816-86b1-bfb2b78ffb4f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"H1.3_(Homo_sapiens)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Homo sapiens\",\n",
    "    \"taxonomic_span_id\": \"9606\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"H1.3\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "query = f\"UPDATE sequence SET variant='H1.3_(Homo_sapiens)', variant_under_consideration=null WHERE accession='NP_005311.1'\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "4a43bfd9-1ce7-42b3-9b91-1e550dc6ac4f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_005311.1</td>\n",
       "      <td>H1.3_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>3007</td>\n",
       "      <td>H1-3</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MSETAPLAPTIPAPAEKTPVKKKAKKAGATAGKRKASGPPVSELIT...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession              variant gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  NP_005311.1  H1.3_(Homo_sapiens)            3007           H1-3   \n",
       "\n",
       "   taxonomy_id      organism    phylum     class taxonomy_group info  \\\n",
       "0         9606  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MSETAPLAPTIPAPAEKTPVKKKAKKAGATAGKRKASGPPVSELIT...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                              "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE accession='NP_005311.1'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "2e0e8365-8bf6-47fd-a4f2-14d192f2c727",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H1.3_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td></td>\n",
       "      <td>H1.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  H1.3_(Homo_sapiens)  variant   Homo sapiens              9606               \n",
       "\n",
       "  parent  \n",
       "0   H1.3  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone WHERE id='H1.3_(Homo_sapiens)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "66f644a2-bf18-43b7-9ceb-eac0a5a3bdbd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>NP_005312.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>NP_005313.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.5_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>NP_005314.2</td>\n",
       "      <td></td>\n",
       "      <td>H1.6_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>NP_861453.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.7_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>NP_722575.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.8_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession variant variant_under_consideration\n",
       "218  NP_005312.1            H1.4_(Homo_sapiens)__???\n",
       "219  NP_005313.1            H1.5_(Homo_sapiens)__???\n",
       "220  NP_005314.2            H1.6_(Homo_sapiens)__???\n",
       "320  NP_861453.1            H1.7_(Homo_sapiens)__???\n",
       "306  NP_722575.1            H1.8_(Homo_sapiens)__???"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "].sort_values(by=[\"variant_under_consideration\"])\n",
    "sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "ab86387c-6bc4-4f0e-85a0-ca97390eed3b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"H1.4_(Homo_sapiens)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Homo sapiens\",\n",
    "    \"taxonomic_span_id\": \"9606\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"H1.4\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "query = f\"UPDATE sequence SET variant='H1.4_(Homo_sapiens)', variant_under_consideration=null WHERE accession='NP_005312.1'\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "dfe8edf1-53f1-4dfb-9ee3-272f9a42193e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_005312.1</td>\n",
       "      <td>H1.4_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>3008</td>\n",
       "      <td>H1-4</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MSETAPAAPAAPAPAEKTPVKKKARKSAGAAKRKASGPPVSELITK...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession              variant gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  NP_005312.1  H1.4_(Homo_sapiens)            3008           H1-4   \n",
       "\n",
       "   taxonomy_id      organism    phylum     class taxonomy_group info  \\\n",
       "0         9606  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MSETAPAAPAAPAPAEKTPVKKKARKSAGAAKRKASGPPVSELITK...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                              "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE accession='NP_005312.1'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "ce87c153-62d8-408c-b132-8f4f81b8b2b6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H1.4_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td></td>\n",
       "      <td>H1.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  H1.4_(Homo_sapiens)  variant   Homo sapiens              9606               \n",
       "\n",
       "  parent  \n",
       "0   H1.4  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone WHERE id='H1.4_(Homo_sapiens)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "fb2fd3cb-6222-4803-9f75-dcae315acfd0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>NP_005313.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.5_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>NP_005314.2</td>\n",
       "      <td></td>\n",
       "      <td>H1.6_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>NP_861453.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.7_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>NP_722575.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.8_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>NP_001295191.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.8_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession variant variant_under_consideration\n",
       "219     NP_005313.1            H1.5_(Homo_sapiens)__???\n",
       "220     NP_005314.2            H1.6_(Homo_sapiens)__???\n",
       "320     NP_861453.1            H1.7_(Homo_sapiens)__???\n",
       "306     NP_722575.1            H1.8_(Homo_sapiens)__???\n",
       "164  NP_001295191.1            H1.8_(Homo_sapiens)__???"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "].sort_values(by=[\"variant_under_consideration\"])\n",
    "sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "576db4f3-66b7-4d3b-9996-4016feff462c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"H1.5_(Homo_sapiens)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Homo sapiens\",\n",
    "    \"taxonomic_span_id\": \"9606\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"H1.5\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "query = f\"UPDATE sequence SET variant='H1.5_(Homo_sapiens)', variant_under_consideration=null WHERE accession='NP_005313.1'\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "743e2792-f03e-4f8e-9f52-c5721ed1731e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_005313.1</td>\n",
       "      <td>H1.5_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>3009</td>\n",
       "      <td>H1-5</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MSETAPAETATPAPVEKSPAKKKATKKAAGAGAAKRKATGPPVSEL...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession              variant gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  NP_005313.1  H1.5_(Homo_sapiens)            3009           H1-5   \n",
       "\n",
       "   taxonomy_id      organism    phylum     class taxonomy_group info  \\\n",
       "0         9606  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MSETAPAETATPAPVEKSPAKKKATKKAAGAGAAKRKATGPPVSEL...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                              "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE accession='NP_005313.1'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "de103dea-a369-483f-847e-8ccae671de61",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H1.5_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td></td>\n",
       "      <td>H1.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  H1.5_(Homo_sapiens)  variant   Homo sapiens              9606               \n",
       "\n",
       "  parent  \n",
       "0   H1.5  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone WHERE id='H1.5_(Homo_sapiens)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "cab756ac-4c46-4f46-92a6-7e7fae3843bb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>NP_005314.2</td>\n",
       "      <td></td>\n",
       "      <td>H1.6_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>NP_861453.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.7_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>NP_722575.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.8_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>NP_001295191.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.8_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>NP_003484.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession variant variant_under_consideration\n",
       "220     NP_005314.2            H1.6_(Homo_sapiens)__???\n",
       "320     NP_861453.1            H1.7_(Homo_sapiens)__???\n",
       "306     NP_722575.1            H1.8_(Homo_sapiens)__???\n",
       "164  NP_001295191.1            H1.8_(Homo_sapiens)__???\n",
       "175     NP_003484.1            H3.4_(Homo_sapiens)__???"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "].sort_values(by=[\"variant_under_consideration\"])\n",
    "sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "93980f9e-3401-4afd-84c9-1b4205c8c5c6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"H1.6_(Homo_sapiens)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Homo sapiens\",\n",
    "    \"taxonomic_span_id\": \"9606\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"TS_H1.6\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "query = f\"UPDATE sequence SET variant='H1.6_(Homo_sapiens)', variant_under_consideration=null WHERE accession='NP_005314.2'\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "2aaed9e1-ad88-4588-8fe3-df82321c44ff",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_005314.2</td>\n",
       "      <td>H1.6_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>3010</td>\n",
       "      <td>H1-6</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MSETVPAASASAGVAAMEKLPTKKRGRKPAGLISASRKVPNLSVSK...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession              variant gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  NP_005314.2  H1.6_(Homo_sapiens)            3010           H1-6   \n",
       "\n",
       "   taxonomy_id      organism    phylum     class taxonomy_group info  \\\n",
       "0         9606  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MSETVPAASASAGVAAMEKLPTKKRGRKPAGLISASRKVPNLSVSK...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                              "
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE accession='NP_005314.2'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "b1d12aef-c67a-4970-9e42-fe26fff06bb7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H1.6_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td></td>\n",
       "      <td>TS_H1.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  H1.6_(Homo_sapiens)  variant   Homo sapiens              9606               \n",
       "\n",
       "    parent  \n",
       "0  TS_H1.6  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone WHERE id='H1.6_(Homo_sapiens)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "deb7e5c2-732c-43ac-acdc-990c969e5a86",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>NP_861453.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.7_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>NP_722575.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.8_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>NP_001295191.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.8_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>NP_003484.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>NP_001013721.2</td>\n",
       "      <td></td>\n",
       "      <td>H3.5_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession variant variant_under_consideration\n",
       "320     NP_861453.1            H1.7_(Homo_sapiens)__???\n",
       "306     NP_722575.1            H1.8_(Homo_sapiens)__???\n",
       "164  NP_001295191.1            H1.8_(Homo_sapiens)__???\n",
       "175     NP_003484.1            H3.4_(Homo_sapiens)__???\n",
       "99   NP_001013721.2            H3.5_(Homo_sapiens)__???"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "].sort_values(by=[\"variant_under_consideration\"])\n",
    "sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "10464a34-fdba-4c14-8e1a-a56d2f9af1ef",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"H1.7_(Homo_sapiens)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Homo sapiens\",\n",
    "    \"taxonomic_span_id\": \"9606\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"TS_H1.7\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "query = f\"UPDATE sequence SET variant='H1.7_(Homo_sapiens)', variant_under_consideration=null WHERE accession='NP_861453.1'\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "8dbe6444-3ca8-4f1b-8479-f2a46fe71809",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_861453.1</td>\n",
       "      <td>H1.7_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>341567</td>\n",
       "      <td>H1-7</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MEQALTGEAQSRWPRRGGSGAMAEAPGPSGESRGHSATQLPAEKTV...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession              variant gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  NP_861453.1  H1.7_(Homo_sapiens)          341567           H1-7   \n",
       "\n",
       "   taxonomy_id      organism    phylum     class taxonomy_group info  \\\n",
       "0         9606  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MEQALTGEAQSRWPRRGGSGAMAEAPGPSGESRGHSATQLPAEKTV...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                              "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE accession='NP_861453.1'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "f6a6c173-839b-45d5-b865-f0296a733353",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H1.7_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td></td>\n",
       "      <td>TS_H1.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  H1.7_(Homo_sapiens)  variant   Homo sapiens              9606               \n",
       "\n",
       "    parent  \n",
       "0  TS_H1.7  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone WHERE id='H1.7_(Homo_sapiens)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "3b386f78-76fc-4f0a-a0ef-2410933ffe83",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>NP_722575.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.8_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>NP_001295191.1</td>\n",
       "      <td></td>\n",
       "      <td>H1.8_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>NP_003484.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>NP_001013721.2</td>\n",
       "      <td></td>\n",
       "      <td>H3.5_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468</th>\n",
       "      <td>XP_003954426.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.5__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession variant variant_under_consideration\n",
       "306     NP_722575.1            H1.8_(Homo_sapiens)__???\n",
       "164  NP_001295191.1            H1.8_(Homo_sapiens)__???\n",
       "175     NP_003484.1            H3.4_(Homo_sapiens)__???\n",
       "99   NP_001013721.2            H3.5_(Homo_sapiens)__???\n",
       "468  XP_003954426.1                           H3.5__???"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "].sort_values(by=[\"variant_under_consideration\"])\n",
    "sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "8f0b87e9-7a11-41cb-862f-b20925fd85c2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"H1.8_(Homo_sapiens)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Homo sapiens\",\n",
    "    \"taxonomic_span_id\": \"9606\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"OO_H1.8\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "query = f\"UPDATE sequence SET variant='H1.8_(Homo_sapiens)', variant_under_consideration=null WHERE accession='NP_722575.1'\"\n",
    "cursor.execute(query)\n",
    "query = f\"UPDATE sequence SET variant='H1.8_(Homo_sapiens)', variant_under_consideration=null WHERE accession='NP_001295191.1'\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "a9c886ab-7823-44e0-acd7-498d245e22b4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_722575.1</td>\n",
       "      <td>H1.8_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>132243</td>\n",
       "      <td>H1-8</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MAPGSVTSDISPSSTSTAGSSRSPESEKPGPSHGGVPPGGPSHSSL...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession              variant gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  NP_722575.1  H1.8_(Homo_sapiens)          132243           H1-8   \n",
       "\n",
       "   taxonomy_id      organism    phylum     class taxonomy_group info  \\\n",
       "0         9606  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MAPGSVTSDISPSSTSTAGSSRSPESEKPGPSHGGVPPGGPSHSSL...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                              "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE accession='NP_722575.1'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "4c25c05c-a803-4b0e-8bd7-915563bce08a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_001295191.1</td>\n",
       "      <td>H1.8_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>132243</td>\n",
       "      <td>H1-8</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MAPATAPRRAGEAKGKGPKKPSEAKEDPPNVGKVKKAAKRPAKVQK...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession              variant gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  NP_001295191.1  H1.8_(Homo_sapiens)          132243           H1-8   \n",
       "\n",
       "   taxonomy_id      organism    phylum     class taxonomy_group info  \\\n",
       "0         9606  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MAPATAPRRAGEAKGKGPKKPSEAKEDPPNVGKVKKAAKRPAKVQK...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                              "
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE accession='NP_001295191.1'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "f11d9e05-2b3c-480f-8256-6944c5307807",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H1.8_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td></td>\n",
       "      <td>OO_H1.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  H1.8_(Homo_sapiens)  variant   Homo sapiens              9606               \n",
       "\n",
       "    parent  \n",
       "0  OO_H1.8  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone WHERE id='H1.8_(Homo_sapiens)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "69071804-f0d4-4275-9e7a-f829a7f80561",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>NP_003484.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>NP_001013721.2</td>\n",
       "      <td></td>\n",
       "      <td>H3.5_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468</th>\n",
       "      <td>XP_003954426.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.5__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>NP_001342187.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y.1_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>NP_001358848.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y.2_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession variant variant_under_consideration\n",
       "175     NP_003484.1            H3.4_(Homo_sapiens)__???\n",
       "99   NP_001013721.2            H3.5_(Homo_sapiens)__???\n",
       "468  XP_003954426.1                           H3.5__???\n",
       "166  NP_001342187.1          H3.Y.1_(Homo_sapiens)__???\n",
       "168  NP_001358848.1          H3.Y.2_(Homo_sapiens)__???"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "].sort_values(by=[\"variant_under_consideration\"])\n",
    "sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "8df3cb26-1347-4b17-9a33-702f66faa8fd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"H3.4_(Homo_sapiens)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Homo sapiens\",\n",
    "    \"taxonomic_span_id\": \"9606\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"H3.4_(Mammalia)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "query = f\"UPDATE sequence SET variant='H3.4_(Homo_sapiens)', variant_under_consideration=null WHERE accession='NP_003484.1'\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "64a5dc59-6a05-4b80-b4ed-9d34ddc959ec",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_003484.1</td>\n",
       "      <td>H3.4_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>8290</td>\n",
       "      <td>H3-4</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYRPGT...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession              variant gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  NP_003484.1  H3.4_(Homo_sapiens)            8290           H3-4   \n",
       "\n",
       "   taxonomy_id      organism    phylum     class taxonomy_group info  \\\n",
       "0         9606  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYRPGT...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                              "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE accession='NP_003484.1'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "f48949bb-f2d7-4841-8e2f-0eff18d3f90c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H3.4_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td></td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  H3.4_(Homo_sapiens)  variant   Homo sapiens              9606               \n",
       "\n",
       "            parent  \n",
       "0  H3.4_(Mammalia)  "
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone WHERE id='H3.4_(Homo_sapiens)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "8aeb6b75-8687-4f03-bb40-47e8abfa8e6b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>NP_001013721.2</td>\n",
       "      <td></td>\n",
       "      <td>H3.5_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468</th>\n",
       "      <td>XP_003954426.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.5__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>NP_001342187.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y.1_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>NP_001358848.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y.2_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>HISTDB_H3_Y_0</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession variant variant_under_consideration\n",
       "99   NP_001013721.2            H3.5_(Homo_sapiens)__???\n",
       "468  XP_003954426.1                           H3.5__???\n",
       "166  NP_001342187.1          H3.Y.1_(Homo_sapiens)__???\n",
       "168  NP_001358848.1          H3.Y.2_(Homo_sapiens)__???\n",
       "91    HISTDB_H3_Y_0                           H3.Y__???"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "].sort_values(by=[\"variant_under_consideration\"])\n",
    "sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "7777d3d2-9cbe-4a2d-a461-098d279734ee",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"H3.5_(Homo_sapiens)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Homo sapiens\",\n",
    "    \"taxonomic_span_id\": \"9606\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"H3.5_(Primates_or_Hominids?)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "query = f\"UPDATE sequence SET variant='H3.5_(Homo_sapiens)', variant_under_consideration=null WHERE accession='NP_001013721.2'\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "38963784-3b40-4a0d-a046-256a165faf87",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_001013721.2</td>\n",
       "      <td>H3.5_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>440093</td>\n",
       "      <td>H3-5</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSTPSTCGVKPHRYRPGTV...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession              variant gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  NP_001013721.2  H3.5_(Homo_sapiens)          440093           H3-5   \n",
       "\n",
       "   taxonomy_id      organism    phylum     class taxonomy_group info  \\\n",
       "0         9606  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MARTKQTARKSTGGKAPRKQLATKAARKSTPSTCGVKPHRYRPGTV...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                              "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE accession='NP_001013721.2'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "96c834e4-1a2c-45dd-9104-e0bb6780ba18",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H3.5_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td></td>\n",
       "      <td>H3.5_(Primates_or_Hominids?)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  H3.5_(Homo_sapiens)  variant   Homo sapiens              9606               \n",
       "\n",
       "                         parent  \n",
       "0  H3.5_(Primates_or_Hominids?)  "
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone WHERE id='H3.5_(Homo_sapiens)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "6776ad7e-d81a-41bd-9f24-4ebb05ef36e4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>468</th>\n",
       "      <td>XP_003954426.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.5__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>NP_001342187.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y.1_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>NP_001358848.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y.2_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>HISTDB_H3_Y_0</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>HISTDB_H3_Y_1</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession variant variant_under_consideration\n",
       "468  XP_003954426.1                           H3.5__???\n",
       "166  NP_001342187.1          H3.Y.1_(Homo_sapiens)__???\n",
       "168  NP_001358848.1          H3.Y.2_(Homo_sapiens)__???\n",
       "91    HISTDB_H3_Y_0                           H3.Y__???\n",
       "92    HISTDB_H3_Y_1                           H3.Y__???"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "].sort_values(by=[\"variant_under_consideration\"])\n",
    "sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "1ed9d969-9bee-4d45-a06a-ee2ede69c850",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "query = f\"UPDATE sequence SET variant='H3.5_(Primates_or_Hominids?)', variant_under_consideration=null WHERE accession='XP_003954426.1'\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "8957c888-cba9-4cf4-87ac-0f1c3b9d999e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_003954426.1</td>\n",
       "      <td>H3.5_(Primates_or_Hominids?)</td>\n",
       "      <td>410046862</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9598</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSTPSTXGVKKPHRYRPGT...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession                       variant         gi ncbi_gene_id  \\\n",
       "0  XP_003954426.1  H3.5_(Primates_or_Hominids?)  410046862                \n",
       "\n",
       "  hgnc_gene_name  taxonomy_id         organism    phylum     class  \\\n",
       "0                        9598  Pan troglodytes  Chordata  Mammalia   \n",
       "\n",
       "  taxonomy_group info                                           sequence  \\\n",
       "0                      MARTKQTARKSTGGKAPRKQLATKAARKSTPSTXGVKKPHRYRPGT...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                              "
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE accession='XP_003954426.1'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "3ccabdbd-fac9-46ac-ba92-51fd80e1db0a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>NP_001342187.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y.1_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>NP_001358848.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y.2_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>HISTDB_H3_Y_0</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>HISTDB_H3_Y_1</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>HISTDB_H3_Y_2</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession variant variant_under_consideration\n",
       "166  NP_001342187.1          H3.Y.1_(Homo_sapiens)__???\n",
       "168  NP_001358848.1          H3.Y.2_(Homo_sapiens)__???\n",
       "91    HISTDB_H3_Y_0                           H3.Y__???\n",
       "92    HISTDB_H3_Y_1                           H3.Y__???\n",
       "93    HISTDB_H3_Y_2                           H3.Y__???"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "].sort_values(by=[\"variant_under_consideration\"])\n",
    "sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "b4a2d920-b08c-44bf-9b83-2317a159b586",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"H3.Y.1_(Homo_sapiens)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Homo sapiens\",\n",
    "    \"taxonomic_span_id\": \"9606\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"H3.Y_(Homo_sapiens)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "query = f\"UPDATE sequence SET variant='H3.Y.1_(Homo_sapiens)', variant_under_consideration=null WHERE accession='NP_001342187.1'\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "d23e83a6-7482-41f8-aa53-8292846d06c7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_001342187.1</td>\n",
       "      <td>H3.Y.1_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>391769</td>\n",
       "      <td>H3Y1</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGT...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession                variant gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  NP_001342187.1  H3.Y.1_(Homo_sapiens)          391769           H3Y1   \n",
       "\n",
       "   taxonomy_id      organism    phylum     class taxonomy_group info  \\\n",
       "0         9606  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGT...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                              "
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE accession='NP_001342187.1'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "f068c5f5-79b5-4889-a23c-708769c6e657",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H3.Y.1_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y_(Homo_sapiens)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      id    level taxonomic_span taxonomic_span_id  \\\n",
       "0  H3.Y.1_(Homo_sapiens)  variant   Homo sapiens              9606   \n",
       "\n",
       "  description               parent  \n",
       "0              H3.Y_(Homo_sapiens)  "
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone WHERE id='H3.Y.1_(Homo_sapiens)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "bf628a19-7be2-4e7f-9894-8b5acd766135",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>NP_001358848.1</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y.2_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>HISTDB_H3_Y_0</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>HISTDB_H3_Y_1</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>HISTDB_H3_Y_2</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>HISTDB_H3_Y_3</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession variant variant_under_consideration\n",
       "168  NP_001358848.1          H3.Y.2_(Homo_sapiens)__???\n",
       "91    HISTDB_H3_Y_0                           H3.Y__???\n",
       "92    HISTDB_H3_Y_1                           H3.Y__???\n",
       "93    HISTDB_H3_Y_2                           H3.Y__???\n",
       "94    HISTDB_H3_Y_3                           H3.Y__???"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "].sort_values(by=[\"variant_under_consideration\"])\n",
    "sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "cbbe500d-2f11-4365-8f86-2b8ed8063f2d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"H3.Y.2_(Homo_sapiens)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Homo sapiens\",\n",
    "    \"taxonomic_span_id\": \"9606\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"H3.Y_(Homo_sapiens)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "query = f\"UPDATE sequence SET variant='H3.Y.2_(Homo_sapiens)', variant_under_consideration=null WHERE accession='NP_001358848.1'\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "636fe919-0426-446a-9361-7c34931d61fa",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_001358848.1</td>\n",
       "      <td>H3.Y.2_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>340096</td>\n",
       "      <td>H3Y2</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGT...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession                variant gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  NP_001358848.1  H3.Y.2_(Homo_sapiens)          340096           H3Y2   \n",
       "\n",
       "   taxonomy_id      organism    phylum     class taxonomy_group info  \\\n",
       "0         9606  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGT...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                              "
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE accession='NP_001358848.1'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "f4a86121-a374-4048-b419-2b2061d5f1a4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H3.Y.2_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y_(Homo_sapiens)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      id    level taxonomic_span taxonomic_span_id  \\\n",
       "0  H3.Y.2_(Homo_sapiens)  variant   Homo sapiens              9606   \n",
       "\n",
       "  description               parent  \n",
       "0              H3.Y_(Homo_sapiens)  "
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone WHERE id='H3.Y.2_(Homo_sapiens)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "720b4f03-e7f0-403b-80dd-d617d34931b1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>HISTDB_H3_Y_0</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>HISTDB_H3_Y_1</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>HISTDB_H3_Y_2</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>HISTDB_H3_Y_3</td>\n",
       "      <td></td>\n",
       "      <td>H3.Y__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467</th>\n",
       "      <td>XP_003804825.1</td>\n",
       "      <td></td>\n",
       "      <td>TS_H3.4__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession variant variant_under_consideration\n",
       "91    HISTDB_H3_Y_0                           H3.Y__???\n",
       "92    HISTDB_H3_Y_1                           H3.Y__???\n",
       "93    HISTDB_H3_Y_2                           H3.Y__???\n",
       "94    HISTDB_H3_Y_3                           H3.Y__???\n",
       "467  XP_003804825.1                        TS_H3.4__???"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "].sort_values(by=[\"variant_under_consideration\"])\n",
    "sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "802c64a8-542f-492f-9ede-d258d7617e15",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for i in range(4):\n",
    "    query = f\"UPDATE sequence SET variant='H3.Y_(Primates?)', variant_under_consideration=null WHERE accession='HISTDB_H3_Y_{i}'\"\n",
    "    cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "89cb8c71-1661-4992-9c0c-f3c0684c2213",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HISTDB_H3_Y_0</td>\n",
       "      <td>H3.Y_(Primates?)</td>\n",
       "      <td>NOGI</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9544</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ARTKQTARKATNWQAPRKPLATKAAAKRAPPRGGIKKPHRYKPGTQ...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HISTDB_H3_Y_1</td>\n",
       "      <td>H3.Y_(Primates?)</td>\n",
       "      <td>NOGI</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9544</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ARTKQTARKATNWQAPRKPLATKAPGKRLPPRGGIKKPHRYRPGTQ...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HISTDB_H3_Y_2</td>\n",
       "      <td>H3.Y_(Primates?)</td>\n",
       "      <td>NOGI</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9598</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGTL...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HISTDB_H3_Y_3</td>\n",
       "      <td>H3.Y_(Primates?)</td>\n",
       "      <td>NOGI</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9598</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGTL...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession           variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  HISTDB_H3_Y_0  H3.Y_(Primates?)  NOGI                               \n",
       "1  HISTDB_H3_Y_1  H3.Y_(Primates?)  NOGI                               \n",
       "2  HISTDB_H3_Y_2  H3.Y_(Primates?)  NOGI                               \n",
       "3  HISTDB_H3_Y_3  H3.Y_(Primates?)  NOGI                               \n",
       "\n",
       "   taxonomy_id         organism    phylum     class taxonomy_group info  \\\n",
       "0         9544   Macaca mulatta  Chordata  Mammalia                       \n",
       "1         9544   Macaca mulatta  Chordata  Mammalia                       \n",
       "2         9598  Pan troglodytes  Chordata  Mammalia                       \n",
       "3         9598  Pan troglodytes  Chordata  Mammalia                       \n",
       "\n",
       "                                            sequence  \\\n",
       "0  ARTKQTARKATNWQAPRKPLATKAAAKRAPPRGGIKKPHRYKPGTQ...   \n",
       "1  ARTKQTARKATNWQAPRKPLATKAPGKRLPPRGGIKKPHRYRPGTQ...   \n",
       "2  ARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGTL...   \n",
       "3  ARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGTL...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                              \n",
       "1                              \n",
       "2                              \n",
       "3                              "
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE accession LIKE 'HISTDB_H3_Y_%'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "a88b0411-83d1-4cf5-b51c-da345a136eca",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>467</th>\n",
       "      <td>XP_003804825.1</td>\n",
       "      <td></td>\n",
       "      <td>TS_H3.4__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>NP_003539.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>NP_068803.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>NP_003537.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>NP_003536.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession variant variant_under_consideration\n",
       "467  XP_003804825.1                        TS_H3.4__???\n",
       "214     NP_003539.1             cH4_(Homo_sapiens)__???\n",
       "255     NP_068803.1             cH4_(Homo_sapiens)__???\n",
       "213     NP_003537.1             cH4_(Homo_sapiens)__???\n",
       "212     NP_003536.1             cH4_(Homo_sapiens)__???"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "].sort_values(by=[\"variant_under_consideration\"])\n",
    "sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "1a3f7b53-2139-48af-87ba-ddd741f1b7fe",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "query = f\"UPDATE sequence SET variant='H3.4_(Mammalia)', variant_under_consideration=null WHERE accession='XP_003804825.1'\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "659e5cda-a558-4ec4-ad5a-6fc6305ce877",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_003804825.1</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>397466137</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9597</td>\n",
       "      <td>Pan paniscus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLVTKVARKSAPATGGVKKPHRYRPGT...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession          variant         gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  XP_003804825.1  H3.4_(Mammalia)  397466137                               \n",
       "\n",
       "   taxonomy_id      organism    phylum     class taxonomy_group info  \\\n",
       "0         9597  Pan paniscus  Chordata  Mammalia                       \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MARTKQTARKSTGGKAPRKQLVTKVARKSAPATGGVKKPHRYRPGT...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                              "
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE accession='XP_003804825.1'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "48703182-5fe7-4965-b134-0ca6586c3512",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>NP_003539.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>NP_068803.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>NP_003537.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>NP_003536.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>NP_003535.1</td>\n",
       "      <td></td>\n",
       "      <td>cH4_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession variant variant_under_consideration\n",
       "214  NP_003539.1             cH4_(Homo_sapiens)__???\n",
       "255  NP_068803.1             cH4_(Homo_sapiens)__???\n",
       "213  NP_003537.1             cH4_(Homo_sapiens)__???\n",
       "212  NP_003536.1             cH4_(Homo_sapiens)__???\n",
       "211  NP_003535.1             cH4_(Homo_sapiens)__???"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "].sort_values(by=[\"variant_under_consideration\"])\n",
    "sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "bf5d02f8-e21b-43a2-9b46-8721ba0a0dfc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(14, 3)"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df[\n",
    "    sequence_df[\"variant_under_consideration\"] == \"cH4_(Homo_sapiens)__???\"\n",
    "].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "eebf9fce-c1d2-4ab8-8fc4-bcf52298f4b8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "histone_description_id = cursor.lastrowid\n",
    "data_histone = {\n",
    "    \"id\": \"cH4_(Homo_sapiens)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Homo sapiens\",\n",
    "    \"taxonomic_span_id\": \"9606\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cH4\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "for acc in sequence_df[\n",
    "    sequence_df[\"variant_under_consideration\"] == \"cH4_(Homo_sapiens)__???\"\n",
    "][\"accession\"]:\n",
    "    query = f\"UPDATE sequence SET variant='cH4_(Homo_sapiens)', variant_under_consideration=null WHERE accession='{acc}'\"\n",
    "    cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "4b0de031-7407-4dd8-ad87-284ba64acdcf",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>NP_001029249.1</td>\n",
       "      <td>cH4_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>554313</td>\n",
       "      <td>H4C15</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>NP_003486.1</td>\n",
       "      <td>cH4_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>8294</td>\n",
       "      <td>H4C9</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>NP_003529.1</td>\n",
       "      <td>cH4_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>8359</td>\n",
       "      <td>H4C1</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>NP_003530.1</td>\n",
       "      <td>cH4_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>8360</td>\n",
       "      <td>H4C4</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>NP_003531.1</td>\n",
       "      <td>cH4_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>8361</td>\n",
       "      <td>H4C6</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>NP_003532.1</td>\n",
       "      <td>cH4_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>8362</td>\n",
       "      <td>H4C12</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>NP_003533.1</td>\n",
       "      <td>cH4_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>8364</td>\n",
       "      <td>H4C3</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>NP_003534.1</td>\n",
       "      <td>cH4_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>8365</td>\n",
       "      <td>H4C8</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>NP_003535.1</td>\n",
       "      <td>cH4_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>8366</td>\n",
       "      <td>H4C2</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>NP_003536.1</td>\n",
       "      <td>cH4_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>8367</td>\n",
       "      <td>H4C5</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>NP_003537.1</td>\n",
       "      <td>cH4_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>8368</td>\n",
       "      <td>H4C13</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>NP_003539.1</td>\n",
       "      <td>cH4_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>8370</td>\n",
       "      <td>H4C14</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>NP_068803.1</td>\n",
       "      <td>cH4_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>8363</td>\n",
       "      <td>H4C11</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>NP_778224.1</td>\n",
       "      <td>cH4_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>121504</td>\n",
       "      <td>H4C16</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession             variant gi ncbi_gene_id hgnc_gene_name  \\\n",
       "107  NP_001029249.1  cH4_(Homo_sapiens)          554313          H4C15   \n",
       "176     NP_003486.1  cH4_(Homo_sapiens)            8294           H4C9   \n",
       "205     NP_003529.1  cH4_(Homo_sapiens)            8359           H4C1   \n",
       "206     NP_003530.1  cH4_(Homo_sapiens)            8360           H4C4   \n",
       "207     NP_003531.1  cH4_(Homo_sapiens)            8361           H4C6   \n",
       "208     NP_003532.1  cH4_(Homo_sapiens)            8362          H4C12   \n",
       "209     NP_003533.1  cH4_(Homo_sapiens)            8364           H4C3   \n",
       "210     NP_003534.1  cH4_(Homo_sapiens)            8365           H4C8   \n",
       "211     NP_003535.1  cH4_(Homo_sapiens)            8366           H4C2   \n",
       "212     NP_003536.1  cH4_(Homo_sapiens)            8367           H4C5   \n",
       "213     NP_003537.1  cH4_(Homo_sapiens)            8368          H4C13   \n",
       "214     NP_003539.1  cH4_(Homo_sapiens)            8370          H4C14   \n",
       "255     NP_068803.1  cH4_(Homo_sapiens)            8363          H4C11   \n",
       "311     NP_778224.1  cH4_(Homo_sapiens)          121504          H4C16   \n",
       "\n",
       "    taxonomy_id      organism    phylum     class taxonomy_group info  \\\n",
       "107      9606.0  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "176      9606.0  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "205      9606.0  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "206      9606.0  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "207      9606.0  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "208      9606.0  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "209      9606.0  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "210      9606.0  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "211      9606.0  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "212      9606.0  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "213      9606.0  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "214      9606.0  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "255      9606.0  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "311      9606.0  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "\n",
       "                                              sequence  \\\n",
       "107  MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...   \n",
       "176  MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...   \n",
       "205  MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...   \n",
       "206  MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...   \n",
       "207  MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...   \n",
       "208  MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...   \n",
       "209  MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...   \n",
       "210  MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...   \n",
       "211  MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...   \n",
       "212  MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...   \n",
       "213  MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...   \n",
       "214  MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...   \n",
       "255  MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...   \n",
       "311  MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...   \n",
       "\n",
       "    variant_under_consideration  \n",
       "107                              \n",
       "176                              \n",
       "205                              \n",
       "206                              \n",
       "207                              \n",
       "208                              \n",
       "209                              \n",
       "210                              \n",
       "211                              \n",
       "212                              \n",
       "213                              \n",
       "214                              \n",
       "255                              \n",
       "311                              "
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "ch4df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "ch4df[\n",
    "    ch4df[\"accession\"].isin(\n",
    "        sequence_df[\n",
    "            sequence_df[\"variant_under_consideration\"] == \"cH4_(Homo_sapiens)__???\"\n",
    "        ][\"accession\"]\n",
    "    )\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "8f34dfd7-a15d-49fb-aaff-613d6cf464b2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH4_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td></td>\n",
       "      <td>cH4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  cH4_(Homo_sapiens)  variant   Homo sapiens              9606               \n",
       "\n",
       "  parent  \n",
       "0    cH4  "
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone WHERE id='cH4_(Homo_sapiens)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "1181e2b0-1ebd-4a91-b477-7077ebfe6ed0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>NP_001035891.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>NP_001800.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3_(Homo_sapiens)__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AAK39657.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>NP_012875.2</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>NP_031707.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession variant variant_under_consideration\n",
       "112  NP_001035891.1           cenH3_(Homo_sapiens)__???\n",
       "171     NP_001800.1           cenH3_(Homo_sapiens)__???\n",
       "5        AAK39657.1                          cenH3__???\n",
       "230     NP_012875.2                          cenH3__???\n",
       "233     NP_031707.1                          cenH3__???"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "].sort_values(by=[\"variant_under_consideration\"])\n",
    "sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "3d9c11a6-a7de-402d-97ec-690abd5330a2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"cenH3_(Animals)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Homo sapiens\",\n",
    "    \"taxonomic_span_id\": \"9606\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cenH3_(Eukarya)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "data_histone = {\n",
    "    \"id\": \"cenH3_(Mammalia)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Homo sapiens\",\n",
    "    \"taxonomic_span_id\": \"9606\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cenH3_(Animals)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "data_histone = {\n",
    "    \"id\": \"cenH3_(Homo_sapiens)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Homo sapiens\",\n",
    "    \"taxonomic_span_id\": \"9606\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cenH3_(Mammalia)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "query = f\"UPDATE sequence SET variant='cenH3_(Homo_sapiens)', variant_under_consideration=null WHERE accession='NP_001035891.1'\"\n",
    "cursor.execute(query)\n",
    "query = f\"UPDATE sequence SET variant='cenH3_(Homo_sapiens)', variant_under_consideration=null WHERE accession='NP_001800.1'\"\n",
    "cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "949d39c6-dbf0-4c0c-82ea-93f861e6810b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_001035891.1</td>\n",
       "      <td>cenH3_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>1058</td>\n",
       "      <td>CENPA</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MGPRRRSRKPEAPRRRSPSPTPTPGPSRRGPSLGASSHQHSRRRQG...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession               variant gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  NP_001035891.1  cenH3_(Homo_sapiens)            1058          CENPA   \n",
       "\n",
       "   taxonomy_id      organism    phylum     class taxonomy_group info  \\\n",
       "0         9606  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MGPRRRSRKPEAPRRRSPSPTPTPGPSRRGPSLGASSHQHSRRRQG...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                              "
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE accession='NP_001035891.1'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "baea941f-3993-41a2-ad68-2984069c373b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_001800.1</td>\n",
       "      <td>cenH3_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td>1058</td>\n",
       "      <td>CENPA</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>MGPRRRSRKPEAPRRRSPSPTPTPGPSRRGPSLGASSHQHSRRRQG...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession               variant gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  NP_001800.1  cenH3_(Homo_sapiens)            1058          CENPA   \n",
       "\n",
       "   taxonomy_id      organism    phylum     class taxonomy_group info  \\\n",
       "0         9606  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MGPRRRSRKPEAPRRRSPSPTPTPGPSRRGPSLGASSHQHSRRRQG...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                              "
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE accession='NP_001800.1'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "bd90e0db-3040-4915-ad3a-29f910572e69",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td></td>\n",
       "      <td>cenH3_(Mammalia)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  cenH3_(Homo_sapiens)  variant   Homo sapiens              9606               \n",
       "\n",
       "             parent  \n",
       "0  cenH3_(Mammalia)  "
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone WHERE id='cenH3_(Homo_sapiens)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "6f1a0071-d1db-4db5-b580-eab934aa2db9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AAK39657.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>NP_012875.2</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>NP_031707.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>NP_499128.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>NP_523730.2</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293</th>\n",
       "      <td>NP_563627.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>NP_596473.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>XP_001011273.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369</th>\n",
       "      <td>XP_001350068.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>401</th>\n",
       "      <td>XP_002287626.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>416</th>\n",
       "      <td>XP_002767160.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>475</th>\n",
       "      <td>XP_005535607.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>503</th>\n",
       "      <td>XP_009526809.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession variant variant_under_consideration\n",
       "5        AAK39657.1                          cenH3__???\n",
       "230     NP_012875.2                          cenH3__???\n",
       "233     NP_031707.1                          cenH3__???\n",
       "284     NP_499128.1                          cenH3__???\n",
       "289     NP_523730.2                          cenH3__???\n",
       "293     NP_563627.1                          cenH3__???\n",
       "298     NP_596473.1                          cenH3__???\n",
       "361  XP_001011273.1                          cenH3__???\n",
       "369  XP_001350068.1                          cenH3__???\n",
       "401  XP_002287626.1                          cenH3__???\n",
       "416  XP_002767160.1                          cenH3__???\n",
       "475  XP_005535607.1                          cenH3__???\n",
       "503  XP_009526809.1                          cenH3__???"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "].sort_values(by=[\"variant_under_consideration\"])\n",
    "sequence_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "9fa479e5-424c-4440-93a4-f2cafbd7a709",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>TAX</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>NP_499128.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "      <td>Animals</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>NP_523730.2</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "      <td>Animals</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>XP_001011273.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "      <td>Eukarya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369</th>\n",
       "      <td>XP_001350068.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "      <td>Eukarya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>401</th>\n",
       "      <td>XP_002287626.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "      <td>Eukarya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>416</th>\n",
       "      <td>XP_002767160.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "      <td>Eukarya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>503</th>\n",
       "      <td>XP_009526809.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "      <td>Eukarya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>NP_012875.2</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "      <td>Fungi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>NP_596473.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "      <td>Fungi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>NP_031707.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "      <td>Mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AAK39657.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "      <td>Plants</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293</th>\n",
       "      <td>NP_563627.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "      <td>Plants</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>475</th>\n",
       "      <td>XP_005535607.1</td>\n",
       "      <td></td>\n",
       "      <td>cenH3__???</td>\n",
       "      <td>Plants</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession variant variant_under_consideration       TAX\n",
       "284     NP_499128.1                          cenH3__???   Animals\n",
       "289     NP_523730.2                          cenH3__???   Animals\n",
       "361  XP_001011273.1                          cenH3__???   Eukarya\n",
       "369  XP_001350068.1                          cenH3__???   Eukarya\n",
       "401  XP_002287626.1                          cenH3__???   Eukarya\n",
       "416  XP_002767160.1                          cenH3__???   Eukarya\n",
       "503  XP_009526809.1                          cenH3__???   Eukarya\n",
       "230     NP_012875.2                          cenH3__???     Fungi\n",
       "298     NP_596473.1                          cenH3__???     Fungi\n",
       "233     NP_031707.1                          cenH3__???  Mammalia\n",
       "5        AAK39657.1                          cenH3__???    Plants\n",
       "293     NP_563627.1                          cenH3__???    Plants\n",
       "475  XP_005535607.1                          cenH3__???    Plants"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df['TAX'] = [\n",
    "    \"Plants\",\n",
    "    \"Fungi\",\n",
    "    \"Mammalia\",\n",
    "    \"Animals\",\n",
    "    \"Animals\",\n",
    "    \"Plants\",\n",
    "    \"Fungi\",\n",
    "    \"Eukarya\",\n",
    "    \"Eukarya\",\n",
    "    \"Eukarya\",\n",
    "    \"Eukarya\",\n",
    "    \"Plants\",\n",
    "    \"Eukarya\",\n",
    "]\n",
    "sequence_df = sequence_df.sort_values(by=[\"TAX\"])\n",
    "sequence_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "3410b46c-79f7-4d46-8968-cf9b99880326",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"cenH3_(Fungi)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Homo sapiens\",\n",
    "    \"taxonomic_span_id\": \"9606\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cenH3_(Eukarya)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "for i, row in sequence_df.iterrows():\n",
    "    query = f\"UPDATE sequence SET variant='cenH3_({row['TAX']})', variant_under_consideration=null WHERE accession='{row['accession']}'\"\n",
    "    cursor.execute(query)\n",
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "52abd53c-8da1-4b20-a945-7280edf3128f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AAK39657.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>13794280</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>55529.0</td>\n",
       "      <td>Guillardia theta</td>\n",
       "      <td>None</td>\n",
       "      <td>Cryptophyceae</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MMKKQNLKRFKKSSNSLVDIRKFQKSTDLLIHRLPFARLVKEISLK...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>NP_012875.2</td>\n",
       "      <td>cenH3_(Fungi)</td>\n",
       "      <td>27808712</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>559292.0</td>\n",
       "      <td>Saccharomyces cerevisiae S288C</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Saccharomycetes</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MSSKQQWVSSAIQSDSSGRSLSNVNRLAGDQQSINDRALSLLQRTR...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>NP_031707.1</td>\n",
       "      <td>cenH3_(Mammalia)</td>\n",
       "      <td>6680920</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10090.0</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MGPRRKPQTPRRRPSSPAPGPSRQSSSVGSQTLRRRQKFMWLKEIK...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>NP_499128.1</td>\n",
       "      <td>cenH3_(Animals)</td>\n",
       "      <td>17553736</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>6239.0</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>Nematoda</td>\n",
       "      <td>Chromadorea</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MADDTPIIEEIAEQNESVTRIMQRLKHDMQRVTSVPGFNTSAAGVN...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>NP_523730.2</td>\n",
       "      <td>cenH3_(Animals)</td>\n",
       "      <td>22024004</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7227.0</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MPRHSRAKRAPRPSANNSKSPNDDDTAFRSPEPEDGTDYGLEFTTS...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293</th>\n",
       "      <td>NP_563627.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>18378832</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3702.0</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MARTKHRVTRSQPRNQTDAAGASSSQAAGPTTTPTRRGGEGGDNTQ...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>NP_596473.1</td>\n",
       "      <td>cenH3_(Fungi)</td>\n",
       "      <td>19113265</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4896.0</td>\n",
       "      <td>Schizosaccharomyces pombe</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Schizosaccharomycetes</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MAKKSLMAEPGDPIPRPRKKRYRPGTTALREIRKYQRSTDLLIQRL...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>XP_001011273.1</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>118356028</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>312017.0</td>\n",
       "      <td>Tetrahymena thermophila SB210</td>\n",
       "      <td>Ciliophora</td>\n",
       "      <td>Oligohymenophorea</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MARKAYQPKRRSNSNQNQQRSDSLKKNKQDNLRSKSAGNQQGNEKN...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369</th>\n",
       "      <td>XP_001350068.1</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>124513424</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>36329.0</td>\n",
       "      <td>Plasmodium falciparum 3D7</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Aconoidasida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MVRTKKNIPNHNPLNAFNRDKSFKTNKTLPNRTVHHGISSKTTNIN...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>401</th>\n",
       "      <td>XP_002287626.1</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>223995905</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>296543.0</td>\n",
       "      <td>Thalassiosira pseudonana CCMP1335</td>\n",
       "      <td>Bacillariophyta</td>\n",
       "      <td>Coscinodiscophyceae</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MRPGEKALREIRQYQSSTSLLLRRLPFARLVREIQYGMTRQPYRWQ...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>416</th>\n",
       "      <td>XP_002767160.1</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>294874934</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>423536.0</td>\n",
       "      <td>Perkinsus marinus ATCC 50983</td>\n",
       "      <td>Perkinsozoa</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MVGVENLGVGFDELLTRGGCGVRDDAVEIAFRGVEGLEDVLKDYMV...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>475</th>\n",
       "      <td>XP_005535607.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>544210412</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>280699.0</td>\n",
       "      <td>Cyanidioschyzon merolae strain 10D</td>\n",
       "      <td>Rhodophyta</td>\n",
       "      <td>Bangiophyceae</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MVRVFGNPVGGAHQYRQTQALNRDESLREQLQRTTVAEQDENIEPE...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>503</th>\n",
       "      <td>XP_009526809.1</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>695416328</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>67593.0</td>\n",
       "      <td>Phytophthora sojae</td>\n",
       "      <td>Oomycota</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MASPRPPALPASDASSSSVTDSGTDDSTPPNSPQPQRSPSPPFAAT...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession           variant         gi ncbi_gene_id hgnc_gene_name  \\\n",
       "5        AAK39657.1    cenH3_(Plants)   13794280                               \n",
       "230     NP_012875.2     cenH3_(Fungi)   27808712                               \n",
       "233     NP_031707.1  cenH3_(Mammalia)    6680920                               \n",
       "284     NP_499128.1   cenH3_(Animals)   17553736                               \n",
       "289     NP_523730.2   cenH3_(Animals)   22024004                               \n",
       "293     NP_563627.1    cenH3_(Plants)   18378832                               \n",
       "298     NP_596473.1     cenH3_(Fungi)   19113265                               \n",
       "361  XP_001011273.1   cenH3_(Eukarya)  118356028                               \n",
       "369  XP_001350068.1   cenH3_(Eukarya)  124513424                               \n",
       "401  XP_002287626.1   cenH3_(Eukarya)  223995905                               \n",
       "416  XP_002767160.1   cenH3_(Eukarya)  294874934                               \n",
       "475  XP_005535607.1    cenH3_(Plants)  544210412                               \n",
       "503  XP_009526809.1   cenH3_(Eukarya)  695416328                               \n",
       "\n",
       "    taxonomy_id                            organism           phylum  \\\n",
       "5       55529.0                    Guillardia theta             None   \n",
       "230    559292.0      Saccharomyces cerevisiae S288C       Ascomycota   \n",
       "233     10090.0                        Mus musculus         Chordata   \n",
       "284      6239.0              Caenorhabditis elegans         Nematoda   \n",
       "289      7227.0             Drosophila melanogaster       Arthropoda   \n",
       "293      3702.0                Arabidopsis thaliana     Streptophyta   \n",
       "298      4896.0           Schizosaccharomyces pombe       Ascomycota   \n",
       "361    312017.0       Tetrahymena thermophila SB210       Ciliophora   \n",
       "369     36329.0           Plasmodium falciparum 3D7      Apicomplexa   \n",
       "401    296543.0   Thalassiosira pseudonana CCMP1335  Bacillariophyta   \n",
       "416    423536.0        Perkinsus marinus ATCC 50983      Perkinsozoa   \n",
       "475    280699.0  Cyanidioschyzon merolae strain 10D       Rhodophyta   \n",
       "503     67593.0                  Phytophthora sojae         Oomycota   \n",
       "\n",
       "                     class taxonomy_group info  \\\n",
       "5            Cryptophyceae                       \n",
       "230        Saccharomycetes                       \n",
       "233               Mammalia                       \n",
       "284            Chromadorea                       \n",
       "289                Insecta                       \n",
       "293          Magnoliopsida                       \n",
       "298  Schizosaccharomycetes                       \n",
       "361      Oligohymenophorea                       \n",
       "369           Aconoidasida                       \n",
       "401    Coscinodiscophyceae                       \n",
       "416                   None                       \n",
       "475          Bangiophyceae                       \n",
       "503                   None                       \n",
       "\n",
       "                                              sequence  \\\n",
       "5    MMKKQNLKRFKKSSNSLVDIRKFQKSTDLLIHRLPFARLVKEISLK...   \n",
       "230  MSSKQQWVSSAIQSDSSGRSLSNVNRLAGDQQSINDRALSLLQRTR...   \n",
       "233  MGPRRKPQTPRRRPSSPAPGPSRQSSSVGSQTLRRRQKFMWLKEIK...   \n",
       "284  MADDTPIIEEIAEQNESVTRIMQRLKHDMQRVTSVPGFNTSAAGVN...   \n",
       "289  MPRHSRAKRAPRPSANNSKSPNDDDTAFRSPEPEDGTDYGLEFTTS...   \n",
       "293  MARTKHRVTRSQPRNQTDAAGASSSQAAGPTTTPTRRGGEGGDNTQ...   \n",
       "298  MAKKSLMAEPGDPIPRPRKKRYRPGTTALREIRKYQRSTDLLIQRL...   \n",
       "361  MARKAYQPKRRSNSNQNQQRSDSLKKNKQDNLRSKSAGNQQGNEKN...   \n",
       "369  MVRTKKNIPNHNPLNAFNRDKSFKTNKTLPNRTVHHGISSKTTNIN...   \n",
       "401  MRPGEKALREIRQYQSSTSLLLRRLPFARLVREIQYGMTRQPYRWQ...   \n",
       "416  MVGVENLGVGFDELLTRGGCGVRDDAVEIAFRGVEGLEDVLKDYMV...   \n",
       "475  MVRVFGNPVGGAHQYRQTQALNRDESLREQLQRTTVAEQDENIEPE...   \n",
       "503  MASPRPPALPASDASSSSVTDSGTDDSTPPNSPQPQRSPSPPFAAT...   \n",
       "\n",
       "    variant_under_consideration  \n",
       "5                                \n",
       "230                              \n",
       "233                              \n",
       "284                              \n",
       "289                              \n",
       "293                              \n",
       "298                              \n",
       "361                              \n",
       "369                              \n",
       "401                              \n",
       "416                              \n",
       "475                              \n",
       "503                              "
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "cenh3 = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "cenh3[cenh3[\"accession\"].isin(sequence_df[\"accession\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "6b93e33d-2bab-4239-9067-4ebda2e653a7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3_(Fungi)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td></td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  cenH3_(Fungi)  variant   Homo sapiens              9606               \n",
       "\n",
       "            parent  \n",
       "0  cenH3_(Eukarya)  "
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone WHERE id='cenH3_(Fungi)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description]).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "ff9271eb-a3ac-4fb9-843b-f38077c5339c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ").fillna(\"\")\n",
    "sequence_df = sequence_df[sequence_df[\"variant_under_consideration\"] != \"\"][\n",
    "    [\"accession\", \"variant\", \"variant_under_consideration\"]\n",
    "].sort_values(by=[\"variant_under_consideration\"])\n",
    "sequence_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22492aa8-95fe-4ce4-a867-f507f9620337",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Close connections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1728e2b4-d030-49bc-8b58-e9d3892cb0c9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cursor.close()\n",
    "conn.close()\n",
    "tunnel.stop()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d799238-32c2-48c0-8f6d-4dff6818d413",
   "metadata": {},
   "source": [
    "# Что осталось необработанным:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3649a3f7-17f7-4e5c-8abc-bf5ec6f5d2c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "uresolved = {\n",
    "    \"H3.3_(Eukarya)\": {\n",
    "        \"level\": \"variant_group\",\n",
    "        \"description\": \"The major replication-independent or replacement H3, important for development, transcription, and chromosome segregation. It typically differs from the canonical H3 by only a few amino acids that are necessary for replication-independent assembly. \",\n",
    "        \"taxonomic_span\": \"Eukaryotes\",\n",
    "        \"taxonomic_span_id\": \"2759\",\n",
    "        \"alternate_names\": [\n",
    "            {\"name\": \"soH3-1\", \"gene\": 1},\n",
    "            {\"name\": \"soH3-2\", \"gene\": 2},\n",
    "            {\"name\": \"hv2\", \"gene\": 1},\n",
    "        ],\n",
    "        \"publications\": [\"22650316\", \"20738881\", \"19412883\", \"24229707\", \"14583738\"],\n",
    "    },  # не разобралась\n",
    "    \"H3.5_(Mammals?)\": {\n",
    "        \"level\": \"variant_group\",\n",
    "        \"description\": \"A hominid specific variant expressed in seminiferous tubules of human testis.\",\n",
    "        \"taxonomic_span\": \"Hominids\",\n",
    "        \"taxonomic_span_id\": \"9604\",\n",
    "        \"alternate_names\": [],\n",
    "        \"publications\": [\"21274551\"],\n",
    "    },  # у меня есть какое-то описание, надо сравнить\n",
    "    \"H3.5_(Hominids?)\": {\n",
    "        \"level\": \"variant_group\",\n",
    "        \"description\": \"A hominid specific variant expressed in seminiferous tubules of human testis.\",\n",
    "        \"taxonomic_span\": \"Hominids\",\n",
    "        \"taxonomic_span_id\": \"9604\",\n",
    "        \"alternate_names\": [],\n",
    "        \"publications\": [\"21274551\"],\n",
    "    },  # у меня есть какое-то описание, надо сравнить\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".conda-histdb_env",
   "language": "python",
   "name": "conda-env-.conda-histdb_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
