{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "437488f9-4c77-42a2-94ee-4bd2712c44b7",
   "metadata": {},
   "source": [
    "# Recover curatedDB from files classification.json and histones.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a546f904-36bd-4a61-ade6-4c59709f4f20",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/scratch/l_singh/hdb/project_dir/histonedb/CURATED_SET\n"
     ]
    }
   ],
   "source": [
    "%cd '../..'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3fd391a0-b09f-4eb9-92ce-a836c782344f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from mysql.connector import connection\n",
    "from sshtunnel import SSHTunnelForwarder\n",
    "from ete3 import Tree\n",
    "from curated_set_services import dict2tree\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fdc483b0-5ae5-4321-9e0a-112b427bea00",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open(\"curated_service/curatedDB/db_curated_server_info.txt\", \"r\") as file:\n",
    "    lines = file.readlines()\n",
    "\n",
    "config = {}\n",
    "\n",
    "for line in lines:\n",
    "    line = line.strip()\n",
    "    if line and not line.startswith(\"#\"):\n",
    "        key, value = line.split(\"=\", 1)\n",
    "        config[key] = value.strip()\n",
    "\n",
    "server_name = config.get(\"server_name\")\n",
    "srever_port = int(config.get(\"srever_port\"))\n",
    "ssh_password = config.get(\"ssh_password\")\n",
    "ssh_username = config.get(\"ssh_username\")\n",
    "db_adress = config.get(\"db_adress\")\n",
    "db_port = int(config.get(\"db_port\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1198219f-0248-4738-9fd0-f78cb7591eed",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43171\n"
     ]
    }
   ],
   "source": [
    "tunnel = SSHTunnelForwarder(\n",
    "    (server_name, srever_port),\n",
    "    ssh_password=ssh_password,\n",
    "    ssh_username=ssh_username,\n",
    "    remote_bind_address=(db_adress, db_port),\n",
    ")\n",
    "tunnel.start()\n",
    "print(tunnel.local_bind_port)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fd4e7e45-f55b-4214-8931-c742765f75de",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "conn = connection.MySQLConnection(\n",
    "    user=\"db_user\",\n",
    "    password=\"db_password\",\n",
    "    host=\"localhost\",\n",
    "    port=tunnel.local_bind_port,\n",
    "    database=\"db_name\",\n",
    ")\n",
    "cursor = conn.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0d5c2490-fbe2-45e8-9be9-9d8a345a242a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('alternative_name',),\n",
       " ('histone',),\n",
       " ('histone_description',),\n",
       " ('histone_has_publication',),\n",
       " ('publication',),\n",
       " ('sequence',),\n",
       " ('sequence_has_publication',)]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SHOW TABLES;\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e2a8c746-7623-4259-b0de-288c207f7f19",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Получение списка всех таблиц в базе данных\n",
    "cursor.execute(\"SHOW TABLES;\")\n",
    "tables = cursor.fetchall()\n",
    "\n",
    "# Отключаем проверку внешних ключей временно\n",
    "cursor.execute(\"SET FOREIGN_KEY_CHECKS = 0;\")\n",
    "\n",
    "# Транзакция для очистки всех таблиц\n",
    "for (table,) in tables:\n",
    "    cursor.execute(f\"TRUNCATE TABLE {table};\")\n",
    "\n",
    "# Восстанавливаем проверку внешних ключей\n",
    "cursor.execute(\"SET FOREIGN_KEY_CHECKS = 1;\")\n",
    "\n",
    "# Фиксация изменений\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f21bf7b7-b955-4897-8b5d-7406305e299e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Все таблицы в базе данных пустые.\n"
     ]
    }
   ],
   "source": [
    "# Получение списка всех таблиц в базе данных\n",
    "cursor.execute(\"SHOW TABLES;\")\n",
    "tables = cursor.fetchall()\n",
    "\n",
    "# Функция для проверки, что таблица пуста\n",
    "def is_table_empty(table_name):\n",
    "    cursor.execute(f\"SELECT COUNT(*) FROM {table_name};\")\n",
    "    result = cursor.fetchone()\n",
    "    return result[0] == 0  # Возвращает True, если количество строк равно 0\n",
    "\n",
    "all_empty = True\n",
    "non_empty_tables = []\n",
    "\n",
    "# Проверка всех таблиц\n",
    "for (table,) in tables:\n",
    "    if not is_table_empty(table):\n",
    "        all_empty = False\n",
    "        non_empty_tables.append(table)\n",
    "\n",
    "if all_empty:\n",
    "    print(\"Все таблицы в базе данных пустые.\")\n",
    "else:\n",
    "    print(f\"Следующие таблицы не пустые: {', '.join(non_empty_tables)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6b87b99-fd5c-4627-927e-c573043b2a00",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Add info from classification.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c3877e74-8321-4dc9-82c7-5d20fcf43c18",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "   /-Archaeal\n",
      "  |\n",
      "  |   /-cH1\n",
      "  |  |\n",
      "  |  |--generic_H1\n",
      "  |  |\n",
      "  |  |-H1.0-H1.0_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-H1.1-H1.1_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-H1.10-H1.10_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-H1.2-H1.2_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-H1.3-H1.3_(Homo_sapiens)\n",
      "  |-H1\n",
      "  |  |-H1.4-H1.4_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-H1.5-H1.5_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-OO_H1.8-H1.8_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |--scH1\n",
      "  |  |\n",
      "  |  |-TS_H1.6-H1.6_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-TS_H1.7-H1.7_(Homo_sapiens)\n",
      "  |  |\n",
      "  |   \\-TS_H1.9\n",
      "  |\n",
      "  |                                                                        /-cH2A.10_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2A.11_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2A.1_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2A.2_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2A.3_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                     /cH2A_(Homo_sapiens)-cH2A.4_(Homo_sapiens)\n",
      "  |                                                    |                  |\n",
      "  |                                                    |                  |--cH2A.5_(Homo_sapiens)\n",
      "  |                                                    |                  |\n",
      "  |                                                    |                  |--cH2A.6_(Homo_sapiens)\n",
      "  |                                                    |                  |\n",
      "  |       /cH2A_(Animals)cH2A_(Vertebrata)cH2A_(Mammalia)                 |--cH2A.7_(Homo_sapiens)\n",
      "  |      |                                             |                  |\n",
      "  |      |                                             |                  |--cH2A.8_(Homo_sapiens)\n",
      "  |      |                                             |                  |\n",
      "  |      |                                             |                   \\-cH2A.9_(Homo_sapiens)\n",
      "  |      |                                             |\n",
      "  |      |                                              \\cH2A_(Mus_musculus)-cH2A.1_(Mus_musculus)\n",
      "  |   /cH2A\n",
      "  |  |   |--cH2A_(Fungi)\n",
      "  |  |   |\n",
      "  |  |   |             /-cH2A_(Chlorophyta)\n",
      "  |  |   |-cH2A_(Plants)\n",
      "  |  |   |             \\-cH2A_(Embryophyta)\n",
      "  |  |   |\n",
      "  |  |    \\-cH2A_(Protists)\n",
      "  |  |\n",
      "  |  |--gH2A\n",
      "  |  |\n",
      "  |  |-H2A.J-H2A.J_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |--H2A.M\n",
      "  |  |\n",
      "  |  |--H2A.R\n",
      "  |  |\n",
      "  |  |--H2A.W\n",
      "  |  |\n",
      "  |  |                                                      /-H2A.X_(Homo_sapiens)\n",
      "  |  |     /H2A.X_(Animals)H2A.X_(Vertebrata)H2A.X_(Mammalia)\n",
      "  |  |    |                                                 \\-H2A.X_(Mus_musculus)\n",
      "  |  |    |\n",
      "  |  |-H2A.X-H2A.X_(Fungi)\n",
      "  |  |    |\n",
      "  |  |    |--H2A.X_(Plants)\n",
      "  |  |    |\n",
      "  |-H2A    \\-H2A.X_(Protists)\n",
      "  |  |\n",
      "  |  |     /H2A.Z.1_(Chordata)H2A.Z.1_(Primates)-H2A.Z.1_(Homo_sapiens)\n",
      "  |  |-H2A.Z\n",
      "  |  |    |                                    /H2A.Z.2.s1_(Primates)-H2A.Z.2.s1_(Homo_sapiens)\n",
      "  |  |     \\H2A.Z.2_(Chordata)H2A.Z.2_(Primates)\n",
      "  |  |                                         \\H2A.Z.2.s2_(Primates)-H2A.Z.2.s2_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |                             /macroH2A.1.s1_(Mammalia)-macroH2A.1.s1_(Homo_sapiens)\n",
      "  |  |        /macroH2A.1_(Mammalia)\n",
      "  |  |-macroH2A                    \\macroH2A.1.s2_(Mammalia)-macroH2A.1.s2_(Homo_sapiens)\n",
      "  |  |       |\n",
      "  |  |        \\macroH2A.2_(Mammalia)-macroH2A.2_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |                                  /-H2A.B.1_(Homo_sapiens)\n",
      "  |  |              /H2A.B_(Homo_sapiens)\n",
      "  |  |             |                    \\-H2A.B.2_(Homo_sapiens)\n",
      "  |  |         /H2A.B\n",
      "  |  |        |    |                    /-H2A.B.1_(Mus_musculus)\n",
      "  |  |        |    |                   |\n",
      "  |  |        |     \\H2A.B_(Mus_musculus)-H2A.B.2_(Mus_musculus)\n",
      "  |  |        |                        |\n",
      "  |  |        |                         \\-H2A.B.3_(Mus_musculus)\n",
      "  |  |        |\n",
      "  |  |        |                         /-H2A.L.1_(Homo_sapiens)\n",
      "  |  |        |     /H2A.L_(Homo_sapiens)\n",
      "  |  |        |    |                    \\-H2A.L.3_(Homo_sapiens)\n",
      "  |   \\short_H2AH2A.L\n",
      "  |           |    |                    /-H2A.L.1_(Mus_musculus)\n",
      "  |           |    |                   |\n",
      "  |           |     \\H2A.L_(Mus_musculus)-H2A.L.2_(Mus_musculus)\n",
      "  |           |                        |\n",
      "  |           |                         \\-H2A.L.3_(Mus_musculus)\n",
      "  |           |\n",
      "  |           |     /-H2A.P_(Homo_sapiens)\n",
      "  |           |-H2A.P\n",
      "  |           |     \\-H2A.P_(Mus_musculus)\n",
      "  |           |\n",
      "  |            \\-H2A.Q\n",
      "  |\n",
      "  |                                                                        /-cH2B.10_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2B.11_(Homo_sapiens)\n",
      "--|                                                                       |\n",
      "  |                                                                       |--cH2B.12_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2B.13_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2B.14_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2B.15_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                                       |--cH2B.1_(Homo_sapiens)\n",
      "  |                                                                       |\n",
      "  |                                                     /cH2B_(Homo_sapiens)-cH2B.2_(Homo_sapiens)\n",
      "  |                                                    |                  |\n",
      "  |                                                    |                  |--cH2B.3_(Homo_sapiens)\n",
      "  |                                                    |                  |\n",
      "  |                                                    |                  |--cH2B.4_(Homo_sapiens)\n",
      "  |                                                    |                  |\n",
      "  |                                                    |                  |--cH2B.5_(Homo_sapiens)\n",
      "  |                                                    |                  |\n",
      "  |       /cH2B_(Animals)cH2B_(Vertebrata)cH2B_(Mammalia)                 |--cH2B.6_(Homo_sapiens)\n",
      "  |      |                                             |                  |\n",
      "  |      |                                             |                  |--cH2B.7_(Homo_sapiens)\n",
      "  |      |                                             |                  |\n",
      "  |      |                                             |                  |--cH2B.8_(Homo_sapiens)\n",
      "  |      |                                             |                  |\n",
      "  |      |                                             |                   \\-cH2B.9_(Homo_sapiens)\n",
      "  |      |                                             |\n",
      "  |      |                                             |                   /-cH2B.1_(Mus_musculus)\n",
      "  |   /cH2B                                             \\cH2B_(Mus_musculus)\n",
      "  |  |   |                                                                 \\-cH2B.E_(Mus_musculus)\n",
      "  |  |   |\n",
      "  |  |   |--cH2B_(Fungi)\n",
      "  |  |   |\n",
      "  |  |   |             /-cH2B_(Chlorophyta)\n",
      "  |  |   |-cH2B_(Plants)\n",
      "  |  |   |             \\-cH2B_(Embryophyta)\n",
      "  |  |   |\n",
      "  |  |    \\-cH2B_(Protists)\n",
      "  |  |\n",
      "  |  |--CS_H2B_(Echinoidea)\n",
      "  |  |\n",
      "  |  |--early_H2B_(Echinoidea)\n",
      "  |  |\n",
      "  |  |--gH2B\n",
      "  |  |\n",
      "  |  |-H2B.K-H2B.K_(Homo_sapiens)\n",
      "  |-H2B\n",
      "  |  |--H2B.L\n",
      "  |  |\n",
      "  |  |-H2B.N-H2B.N_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |--H2B.O\n",
      "  |  |\n",
      "  |  |--H2B.S\n",
      "  |  |\n",
      "  |  |--H2B.V\n",
      "  |  |\n",
      "  |  |                         /-H2B.W.1_(Homo_sapiens)\n",
      "  |  |-H2B.WH2B.W_(Homo_sapiens)\n",
      "  |  |                         \\-H2B.W.2_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |--H2B.Z\n",
      "  |  |\n",
      "  |  |--late_H2B_(Echinoidea)\n",
      "  |  |\n",
      "  |   \\-sperm_H2B_(Echinoidea)\n",
      "  |\n",
      "  |                  /cenH3_(Animals)cenH3_(Mammalia)-cenH3_(Homo_sapiens)\n",
      "  |                 |\n",
      "  |   /cenH3_(Eukarya)-cenH3_(Fungi)\n",
      "  |  |              |\n",
      "  |  |               \\-cenH3_(Plants)\n",
      "  |  |\n",
      "  |  |                                              /cH3.1_(Mammalia)-cH3.1_(Homo_sapiens)\n",
      "  |  |   /cH3_(Animals)cH3_(Vertebrata)cH3_(Mammalia)\n",
      "  |  |  |                                           \\cH3.2_(Mammalia)-cH3.2_(Homo_sapiens)\n",
      "  |  |  |\n",
      "  |  |  |--cH3_(Fungi)\n",
      "  |  |-cH3\n",
      "  |  |  |            /-cH3_(Chlorophyta)\n",
      "  |  |  |-cH3_(Plants)\n",
      "  |  |  |            \\-cH3_(Embryophyta)\n",
      "  |  |  |\n",
      "  |  |   \\-cH3_(Protists)\n",
      "  |  |\n",
      "  |  |--H3.1-like_(Plants)\n",
      "  |  |\n",
      "  |  |    /H3.3_(Animals)-H3.3_(Homo_sapiens)\n",
      "  |  |   |\n",
      "  |  |   |-H3.3_(Fungi)-H3.3_(Ascomycota)\n",
      "  |  |-H3.3\n",
      "  |  |   |--H3.3_(Plants)\n",
      "  |  |   |\n",
      "  |  |    \\-H3.3_(Protists)\n",
      "  |-H3\n",
      "  |  |                             /H3.X_(Primates?)-H3.X_(Homo_sapiens)\n",
      "  |  |          /H3.3-like_(Animals)\n",
      "  |  |         |                  |                                   /-H3.Y.1_(Homo_sapiens)\n",
      "  |  |         |                   \\H3.Y_(Primates?)H3.Y_(Homo_sapiens)\n",
      "  |  |-H3.3-like?                                                     \\-H3.Y.2_(Homo_sapiens)\n",
      "  |  |         |\n",
      "  |  |         |                  /-H3_(Lilly???)\n",
      "  |  |          \\H3.3-like_(Plants)\n",
      "  |  |                            \\-TS H3.10\n",
      "  |  |\n",
      "  |  |-H3.4_(Mammalia)-H3.4_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |-H3.5_(Primates_or_Hominids?)-H3.5_(Homo_sapiens)\n",
      "  |  |\n",
      "  |  |--H3.6_(Mammals?)?\n",
      "  |  |\n",
      "  |  |--H3.7_(Mammals?)?\n",
      "  |  |\n",
      "  |  |--H3.8_(Mammals?)?\n",
      "  |  |\n",
      "  |  |--H3.B_(Giardia?)\n",
      "  |  |\n",
      "  |  |--H3.P_(Moneuplotes?)\n",
      "  |  |\n",
      "  |   \\-H3.V_(Trypanosomes?)\n",
      "  |\n",
      "  |-H4/cH4-cH4_(Homo_sapiens)\n",
      "  |\n",
      "   \\-Viral\n"
     ]
    }
   ],
   "source": [
    "with open('classification.json') as json_file:\n",
    "    data = json.load(json_file)\n",
    "\n",
    "hist_tree=Tree()\n",
    "dict2tree(hist_tree,data['tree'])\n",
    "print(hist_tree.get_ascii(show_internal=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "004fc5ce-f997-416d-b6e1-83bec36acf35",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sql_query = []\n",
    "desc_id = 0\n",
    "pubs_unique = []\n",
    "alt_name_id = 0\n",
    "added_histones = []\n",
    "\n",
    "for node in hist_tree.traverse(\"levelorder\"):\n",
    "    if node.is_root():\n",
    "        continue\n",
    "    \n",
    "    added_histones.append(node.name)\n",
    "    \n",
    "    # insert desc\n",
    "    desc = None if node.name not in data[\"info\"].keys() else data[\"info\"][node.name].get(\"description\", None)\n",
    "    \n",
    "    if isinstance(desc, str):\n",
    "        desc = None if desc == \"null\" else {'summary': desc}\n",
    "    \n",
    "    if desc is not None:\n",
    "        desc_id += 1\n",
    "        # Параметризованный запрос для вставки в histone_description\n",
    "        cursor.execute('''INSERT INTO histone_description \n",
    "            (id, summary, taxonomy, genes, evolution, expression, knock_out, function, sequence, localization, deposition, structure, interactions, disease, caveats)\n",
    "            VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s);''', (\n",
    "                desc_id,\n",
    "                desc.get(\"summary\", None),\n",
    "                desc.get(\"taxonomy\", None),\n",
    "                desc.get(\"genes\", None),\n",
    "                desc.get(\"evolution\", None),\n",
    "                desc.get(\"expression\", None),\n",
    "                desc.get(\"knock-out\", None),\n",
    "                desc.get(\"function\", None),\n",
    "                desc.get(\"sequence\", None),\n",
    "                desc.get(\"localization\", None),\n",
    "                desc.get(\"deposition\", None),\n",
    "                desc.get(\"structure\", None),\n",
    "                desc.get(\"interactions\", None),\n",
    "                desc.get(\"disease\", None),\n",
    "                desc.get(\"caveats\", None)\n",
    "            ))\n",
    "\n",
    "    # insert histone\n",
    "    cursor.execute('''INSERT INTO histone \n",
    "        (id, level, taxonomic_span, taxonomic_span_id, description, parent) \n",
    "        VALUES (%s, %s, %s, %s, %s, %s);''', (\n",
    "            node.name,\n",
    "            \"type\" if node.up.is_root() else \"variant_group\" if node.up.up.is_root() else \"variant\",\n",
    "            data[\"info\"].get(node.name, {}).get(\"taxonomic_span\", None),\n",
    "            data[\"info\"].get(node.name, {}).get(\"taxonomic_span_id\", None),\n",
    "            desc_id if desc is not None else None,\n",
    "            None if node.up.is_root() else node.up.name\n",
    "        ))\n",
    "\n",
    "    # insert publication\n",
    "    pubs = data[\"info\"].get(node.name, {}).get(\"publications\", None)\n",
    "    \n",
    "    if pubs is not None:\n",
    "        if not isinstance(pubs, list):\n",
    "            print(f\"PUBS IS NOT A LIST:\\n{pubs}\")\n",
    "        else:\n",
    "            for p in pubs:\n",
    "                if p not in pubs_unique:\n",
    "                    pubs_unique.append(p)\n",
    "                    # Вставка публикации\n",
    "                    cursor.execute('''INSERT INTO publication (id, title, doi, author, year) \n",
    "                        VALUES (%s, %s, %s, %s, %s);''', (\n",
    "                            p, None, None, None, None\n",
    "                        ))\n",
    "\n",
    "                # insert relation histone_has_publication\n",
    "                cursor.execute('''INSERT INTO histone_has_publication (histone_id, publication_id) \n",
    "                    VALUES (%s, %s);''', (\n",
    "                        node.name, p\n",
    "                    ))\n",
    "\n",
    "    # insert alternative name\n",
    "    alts = data[\"info\"].get(node.name, {}).get(\"alternate_names\", None)\n",
    "    \n",
    "    if alts is not None:\n",
    "        if not isinstance(alts, list):\n",
    "            print(f\"ALTS IS NOT A LIST:\\n{alts}\")\n",
    "        else:\n",
    "            for an in alts:\n",
    "                alt_name_id += 1\n",
    "                if \"name\" not in an.keys():\n",
    "                    print(f\"ALT NAME DOES NOT EXIST:\\n{an}\")\n",
    "                else:\n",
    "                    # Вставка альтернативных имен\n",
    "                    cursor.execute('''INSERT INTO alternative_name \n",
    "                        (id, name, taxonomy, gene, splice, histone) \n",
    "                        VALUES (%s, %s, %s, %s, %s, %s);''', (\n",
    "                            alt_name_id,\n",
    "                            an[\"name\"],\n",
    "                            an.get(\"taxonomy\", None),\n",
    "                            an.get(\"gene\", None),\n",
    "                            an.get(\"splice\", None),\n",
    "                            node.name\n",
    "                        ))\n",
    "\n",
    "# Сохранение изменений в базе данных\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5be27495-33e4-4e78-9a61-8e1c97afa56d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Archaeal</td>\n",
       "      <td>type</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cenH3_(Animals)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>94</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>46</td>\n",
       "      <td>H3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cenH3_(Fungi)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>95</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cenH3_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>158</td>\n",
       "      <td>cenH3_(Mammalia)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>TS H3.10</td>\n",
       "      <td>variant</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>139</td>\n",
       "      <td>H3.3-like_(Plants)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>TS_H1.6</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>19</td>\n",
       "      <td>H1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>TS_H1.7</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>20</td>\n",
       "      <td>H1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>TS_H1.9</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>21</td>\n",
       "      <td>H1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>Viral</td>\n",
       "      <td>type</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>7</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>204 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       id          level taxonomic_span taxonomic_span_id  \\\n",
       "0                Archaeal           type           null              null   \n",
       "1         cenH3_(Animals)        variant   Homo sapiens              9606   \n",
       "2         cenH3_(Eukarya)  variant_group           null              null   \n",
       "3           cenH3_(Fungi)        variant   Homo sapiens              9606   \n",
       "4    cenH3_(Homo_sapiens)        variant   Homo sapiens              9606   \n",
       "..                    ...            ...            ...               ...   \n",
       "199              TS H3.10        variant           null              null   \n",
       "200               TS_H1.6  variant_group       Mammalia             40674   \n",
       "201               TS_H1.7  variant_group       Mammalia             40674   \n",
       "202               TS_H1.9  variant_group       Mammalia             40674   \n",
       "203                 Viral           type           null              null   \n",
       "\n",
       "     description              parent  \n",
       "0              1                None  \n",
       "1             94     cenH3_(Eukarya)  \n",
       "2             46                  H3  \n",
       "3             95     cenH3_(Eukarya)  \n",
       "4            158    cenH3_(Mammalia)  \n",
       "..           ...                 ...  \n",
       "199          139  H3.3-like_(Plants)  \n",
       "200           19                  H1  \n",
       "201           20                  H1  \n",
       "202           21                  H1  \n",
       "203            7                None  \n",
       "\n",
       "[204 rows x 6 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone\"\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "474eb0b7-631e-46a9-b3d8-22ba35e88fbf",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Add sequences from histones.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "69bf02ad-9108-4b4a-b225-aee31acce0db",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAA30025.1</th>\n",
       "      <td>AAA30025.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>cH2B</td>\n",
       "      <td>cH2B_(Animals)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>161386</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7660.0</td>\n",
       "      <td>Psammechinus miliaris</td>\n",
       "      <td>Echinodermata</td>\n",
       "      <td>Echinoidea</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MAPTAQVAKKGSKKAVKAPRPSGGKKRNRKRKESYGIYIYKVLKQV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAB48832.1</th>\n",
       "      <td>AAB48832.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>CS_H2B_(Echinoidea)</td>\n",
       "      <td>CS_H2B_(Echinoidea)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7660.0</td>\n",
       "      <td>Psammechinus miliaris</td>\n",
       "      <td>Echinodermata</td>\n",
       "      <td>Echinoidea</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9032246</td>\n",
       "      <td>MPAKGAATKGEKKQAVKSKAMASSRTGDKKRRRRRLESYNIYIYKV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAC28846.1</th>\n",
       "      <td>AAC28846.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>macroH2A</td>\n",
       "      <td>macroH2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3395780</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9031.0</td>\n",
       "      <td>Gallus gallus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Aves</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22650316 26989147</td>\n",
       "      <td>MSSRGGKKKSTKTSRSAKAGVIFPVGRMLRYIKKGHPKYRIGVGAP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAC37291.1</th>\n",
       "      <td>AAC37291.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>H2A.X</td>\n",
       "      <td>H2A.X_(Protists)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>310870</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5911.0</td>\n",
       "      <td>Tetrahymena thermophila</td>\n",
       "      <td>Ciliophora</td>\n",
       "      <td>Oligohymenophorea</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22650316 26989147</td>\n",
       "      <td>MSTTGKGGKAKGKTASSKQVSRSARAGLQFPVGRISRFLKHGRYSE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAC37292.1</th>\n",
       "      <td>AAC37292.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A_(Protists)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>310872</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5911.0</td>\n",
       "      <td>Tetrahymena thermophila</td>\n",
       "      <td>Ciliophora</td>\n",
       "      <td>Oligohymenophorea</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22650316 26989147</td>\n",
       "      <td>MSTTGKGGKAKGKTASSKQVSRSARAGLQFPVGRISRFLKNGRYSE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_846259.1</th>\n",
       "      <td>XP_846259.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>H2A.Z</td>\n",
       "      <td>H2A.Z</td>\n",
       "      <td>NaN</td>\n",
       "      <td>72391930</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>185431.0</td>\n",
       "      <td>Trypanosoma brucei brucei TREU927</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22650316 26989147</td>\n",
       "      <td>MSLTGDDAVPQAPLVGGVAMSPEQASALTGGKLGGKAVGPAHGKGK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_875023.2</th>\n",
       "      <td>XP_875023.2</td>\n",
       "      <td>H2A</td>\n",
       "      <td>short_H2A</td>\n",
       "      <td>H2A.L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>119923405</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9913.0</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22650316 26989147</td>\n",
       "      <td>MSPRRHLWNCRRSRRHSLSRSTRAELQFPVSRVDRLLREGQGAYRL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_951561.1</th>\n",
       "      <td>XP_951561.1</td>\n",
       "      <td>H4</td>\n",
       "      <td>cH4</td>\n",
       "      <td>cH4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>84043542</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>185431.0</td>\n",
       "      <td>Trypanosoma brucei brucei TREU927</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MAKGKRVGESKGAQKRQKKVLRDNVRGITRGSIRRLARRAGVKRIS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_956003.1</th>\n",
       "      <td>XP_956003.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.3</td>\n",
       "      <td>H3.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>85077297</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>367110.0</td>\n",
       "      <td>Neurospora crassa OR74A</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Sordariomycetes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLASKAARKSAPSTGGVKKPHRYKPGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_966164.1</th>\n",
       "      <td>XP_966164.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.3</td>\n",
       "      <td>H3.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>86171194</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>36329.0</td>\n",
       "      <td>Plasmodium falciparum 3D7</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Aconoidasida</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLASKAARKSAPVSTGIKKPHRYRPGT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>569 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               accession type        variant_group              variant  \\\n",
       "accession                                                                 \n",
       "AAA30025.1    AAA30025.1  H2B                 cH2B       cH2B_(Animals)   \n",
       "AAB48832.1    AAB48832.1  H2B  CS_H2B_(Echinoidea)  CS_H2B_(Echinoidea)   \n",
       "AAC28846.1    AAC28846.1  H2A             macroH2A             macroH2A   \n",
       "AAC37291.1    AAC37291.1  H2A                H2A.X     H2A.X_(Protists)   \n",
       "AAC37292.1    AAC37292.1  H2A                 cH2A      cH2A_(Protists)   \n",
       "...                  ...  ...                  ...                  ...   \n",
       "XP_846259.1  XP_846259.1  H2A                H2A.Z                H2A.Z   \n",
       "XP_875023.2  XP_875023.2  H2A            short_H2A                H2A.L   \n",
       "XP_951561.1  XP_951561.1   H4                  cH4                  cH4   \n",
       "XP_956003.1  XP_956003.1   H3                 H3.3                 H3.3   \n",
       "XP_966164.1  XP_966164.1   H3                 H3.3                 H3.3   \n",
       "\n",
       "             doublet         gi  ncbi_gene_id hgnc_gene_name  taxonomy_id  \\\n",
       "accession                                                                   \n",
       "AAA30025.1       NaN     161386           NaN            NaN       7660.0   \n",
       "AAB48832.1       NaN        NaN           NaN            NaN       7660.0   \n",
       "AAC28846.1       NaN    3395780           NaN            NaN       9031.0   \n",
       "AAC37291.1       NaN     310870           NaN            NaN       5911.0   \n",
       "AAC37292.1       NaN     310872           NaN            NaN       5911.0   \n",
       "...              ...        ...           ...            ...          ...   \n",
       "XP_846259.1      NaN   72391930           NaN            NaN     185431.0   \n",
       "XP_875023.2      NaN  119923405           NaN            NaN       9913.0   \n",
       "XP_951561.1      NaN   84043542           NaN            NaN     185431.0   \n",
       "XP_956003.1      NaN   85077297           NaN            NaN     367110.0   \n",
       "XP_966164.1      NaN   86171194           NaN            NaN      36329.0   \n",
       "\n",
       "                                      organism         phylum  \\\n",
       "accession                                                       \n",
       "AAA30025.1               Psammechinus miliaris  Echinodermata   \n",
       "AAB48832.1               Psammechinus miliaris  Echinodermata   \n",
       "AAC28846.1                       Gallus gallus       Chordata   \n",
       "AAC37291.1             Tetrahymena thermophila     Ciliophora   \n",
       "AAC37292.1             Tetrahymena thermophila     Ciliophora   \n",
       "...                                        ...            ...   \n",
       "XP_846259.1  Trypanosoma brucei brucei TREU927     Euglenozoa   \n",
       "XP_875023.2                         Bos taurus       Chordata   \n",
       "XP_951561.1  Trypanosoma brucei brucei TREU927     Euglenozoa   \n",
       "XP_956003.1            Neurospora crassa OR74A     Ascomycota   \n",
       "XP_966164.1          Plasmodium falciparum 3D7    Apicomplexa   \n",
       "\n",
       "                         class taxonomy_group info         references  \\\n",
       "accession                                                               \n",
       "AAA30025.1          Echinoidea            NaN  NaN                NaN   \n",
       "AAB48832.1          Echinoidea            NaN  NaN            9032246   \n",
       "AAC28846.1                Aves            NaN  NaN  22650316 26989147   \n",
       "AAC37291.1   Oligohymenophorea            NaN  NaN  22650316 26989147   \n",
       "AAC37292.1   Oligohymenophorea            NaN  NaN  22650316 26989147   \n",
       "...                        ...            ...  ...                ...   \n",
       "XP_846259.1      Kinetoplastea            NaN  NaN  22650316 26989147   \n",
       "XP_875023.2           Mammalia            NaN  NaN  22650316 26989147   \n",
       "XP_951561.1      Kinetoplastea            NaN  NaN                NaN   \n",
       "XP_956003.1    Sordariomycetes            NaN  NaN                NaN   \n",
       "XP_966164.1       Aconoidasida            NaN  NaN                NaN   \n",
       "\n",
       "                                                      sequence  \n",
       "accession                                                       \n",
       "AAA30025.1   MAPTAQVAKKGSKKAVKAPRPSGGKKRNRKRKESYGIYIYKVLKQV...  \n",
       "AAB48832.1   MPAKGAATKGEKKQAVKSKAMASSRTGDKKRRRRRLESYNIYIYKV...  \n",
       "AAC28846.1   MSSRGGKKKSTKTSRSAKAGVIFPVGRMLRYIKKGHPKYRIGVGAP...  \n",
       "AAC37291.1   MSTTGKGGKAKGKTASSKQVSRSARAGLQFPVGRISRFLKHGRYSE...  \n",
       "AAC37292.1   MSTTGKGGKAKGKTASSKQVSRSARAGLQFPVGRISRFLKNGRYSE...  \n",
       "...                                                        ...  \n",
       "XP_846259.1  MSLTGDDAVPQAPLVGGVAMSPEQASALTGGKLGGKAVGPAHGKGK...  \n",
       "XP_875023.2  MSPRRHLWNCRRSRRHSLSRSTRAELQFPVSRVDRLLREGQGAYRL...  \n",
       "XP_951561.1  MAKGKRVGESKGAQKRQKKVLRDNVRGITRGSIRRLARRAGVKRIS...  \n",
       "XP_956003.1  MARTKQTARKSTGGKAPRKQLASKAARKSAPSTGGVKKPHRYKPGT...  \n",
       "XP_966164.1  MARTKQTARKSTGGKAPRKQLASKAARKSAPVSTGIKKPHRYRPGT...  \n",
       "\n",
       "[569 rows x 16 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_seq = pd.read_csv('histones.csv')\n",
    "df_seq.index = df_seq.accession\n",
    "df_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0904b066-7a7b-4cf6-97f0-e6043ec3be66",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sql_query_seq = []\n",
    "\n",
    "for i, row in df_seq.iterrows():\n",
    "    # insert sequence\n",
    "    cursor.execute('''INSERT INTO sequence \n",
    "        (accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration)\n",
    "        VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s);''', (\n",
    "            row[\"accession\"],\n",
    "            row[\"variant\"] if row[\"variant\"] in added_histones else None,\n",
    "            row[\"gi\"] if not pd.isna(row[\"gi\"]) else None,\n",
    "            int(row[\"ncbi_gene_id\"]) if not pd.isna(row[\"ncbi_gene_id\"]) else None,\n",
    "            row[\"hgnc_gene_name\"] if not pd.isna(row[\"hgnc_gene_name\"]) else None,\n",
    "            int(row[\"taxonomy_id\"]) if not pd.isna(row[\"taxonomy_id\"]) else None,\n",
    "            row[\"organism\"] if not pd.isna(row[\"organism\"]) else None,\n",
    "            row[\"phylum\"] if not pd.isna(row[\"phylum\"]) else None,\n",
    "            row[\"class\"] if not pd.isna(row[\"class\"]) else None,\n",
    "            row[\"taxonomy_group\"] if not pd.isna(row[\"taxonomy_group\"]) else None,\n",
    "            row[\"info\"] if not pd.isna(row[\"info\"]) else None,\n",
    "            row[\"sequence\"],\n",
    "            row[\"variant\"] if row[\"variant\"] not in added_histones else None\n",
    "        ))\n",
    "\n",
    "    # insert publication if not exist (by default all fields will be null except PK)\n",
    "    pubs_seq = row[\"references\"]\n",
    "    \n",
    "    if not pd.isna(pubs_seq) and pubs_seq != '' and pubs_seq != '?':\n",
    "        for p in pubs_seq.split():\n",
    "            p = p.replace('(?)', '')\n",
    "            if p not in pubs_unique:\n",
    "                pubs_unique.append(p)\n",
    "                # Вставка публикации\n",
    "                cursor.execute('''INSERT INTO publication \n",
    "                    (id, title, doi, author, year) \n",
    "                    VALUES (%s, %s, %s, %s, %s);''', (\n",
    "                        p, None, None, None, None\n",
    "                    ))\n",
    "\n",
    "            # insert relation sequence_has_publication\n",
    "            cursor.execute('''INSERT INTO sequence_has_publication \n",
    "                (sequence_accession, publication_id) \n",
    "                VALUES (%s, %s);''', (\n",
    "                    row[\"accession\"], p\n",
    "                ))\n",
    "\n",
    "# Сохранение изменений в базе данных\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "89f9061f-80a3-40d0-b9ed-dfe0135d766a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAA30025.1</td>\n",
       "      <td>cH2B_(Animals)</td>\n",
       "      <td>161386</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>7660.0</td>\n",
       "      <td>Psammechinus miliaris</td>\n",
       "      <td>Echinodermata</td>\n",
       "      <td>Echinoidea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAPTAQVAKKGSKKAVKAPRPSGGKKRNRKRKESYGIYIYKVLKQV...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAB48832.1</td>\n",
       "      <td>CS_H2B_(Echinoidea)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>7660.0</td>\n",
       "      <td>Psammechinus miliaris</td>\n",
       "      <td>Echinodermata</td>\n",
       "      <td>Echinoidea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MPAKGAATKGEKKQAVKSKAMASSRTGDKKRRRRRLESYNIYIYKV...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAC28846.1</td>\n",
       "      <td>macroH2A</td>\n",
       "      <td>3395780</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9031.0</td>\n",
       "      <td>Gallus gallus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Aves</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MSSRGGKKKSTKTSRSAKAGVIFPVGRMLRYIKKGHPKYRIGVGAP...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAC37291.1</td>\n",
       "      <td>H2A.X_(Protists)</td>\n",
       "      <td>310870</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5911.0</td>\n",
       "      <td>Tetrahymena thermophila</td>\n",
       "      <td>Ciliophora</td>\n",
       "      <td>Oligohymenophorea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MSTTGKGGKAKGKTASSKQVSRSARAGLQFPVGRISRFLKHGRYSE...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAC37292.1</td>\n",
       "      <td>cH2A_(Protists)</td>\n",
       "      <td>310872</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5911.0</td>\n",
       "      <td>Tetrahymena thermophila</td>\n",
       "      <td>Ciliophora</td>\n",
       "      <td>Oligohymenophorea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MSTTGKGGKAKGKTASSKQVSRSARAGLQFPVGRISRFLKNGRYSE...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564</th>\n",
       "      <td>XP_846259.1</td>\n",
       "      <td>H2A.Z</td>\n",
       "      <td>72391930</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>185431.0</td>\n",
       "      <td>Trypanosoma brucei brucei TREU927</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MSLTGDDAVPQAPLVGGVAMSPEQASALTGGKLGGKAVGPAHGKGK...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>XP_875023.2</td>\n",
       "      <td>H2A.L</td>\n",
       "      <td>119923405</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9913.0</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MSPRRHLWNCRRSRRHSLSRSTRAELQFPVSRVDRLLREGQGAYRL...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>XP_951561.1</td>\n",
       "      <td>cH4</td>\n",
       "      <td>84043542</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>185431.0</td>\n",
       "      <td>Trypanosoma brucei brucei TREU927</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAKGKRVGESKGAQKRQKKVLRDNVRGITRGSIRRLARRAGVKRIS...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567</th>\n",
       "      <td>XP_956003.1</td>\n",
       "      <td>H3.3</td>\n",
       "      <td>85077297</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>367110.0</td>\n",
       "      <td>Neurospora crassa OR74A</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Sordariomycetes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLASKAARKSAPSTGGVKKPHRYKPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>XP_966164.1</td>\n",
       "      <td>H3.3</td>\n",
       "      <td>86171194</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>36329.0</td>\n",
       "      <td>Plasmodium falciparum 3D7</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Aconoidasida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLASKAARKSAPVSTGIKKPHRYRPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>569 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession              variant         gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0     AAA30025.1       cH2B_(Animals)     161386         None           None   \n",
       "1     AAB48832.1  CS_H2B_(Echinoidea)       None         None           None   \n",
       "2     AAC28846.1             macroH2A    3395780         None           None   \n",
       "3     AAC37291.1     H2A.X_(Protists)     310870         None           None   \n",
       "4     AAC37292.1      cH2A_(Protists)     310872         None           None   \n",
       "..           ...                  ...        ...          ...            ...   \n",
       "564  XP_846259.1                H2A.Z   72391930         None           None   \n",
       "565  XP_875023.2                H2A.L  119923405         None           None   \n",
       "566  XP_951561.1                  cH4   84043542         None           None   \n",
       "567  XP_956003.1                 H3.3   85077297         None           None   \n",
       "568  XP_966164.1                 H3.3   86171194         None           None   \n",
       "\n",
       "     taxonomy_id                           organism         phylum  \\\n",
       "0         7660.0              Psammechinus miliaris  Echinodermata   \n",
       "1         7660.0              Psammechinus miliaris  Echinodermata   \n",
       "2         9031.0                      Gallus gallus       Chordata   \n",
       "3         5911.0            Tetrahymena thermophila     Ciliophora   \n",
       "4         5911.0            Tetrahymena thermophila     Ciliophora   \n",
       "..           ...                                ...            ...   \n",
       "564     185431.0  Trypanosoma brucei brucei TREU927     Euglenozoa   \n",
       "565       9913.0                         Bos taurus       Chordata   \n",
       "566     185431.0  Trypanosoma brucei brucei TREU927     Euglenozoa   \n",
       "567     367110.0            Neurospora crassa OR74A     Ascomycota   \n",
       "568      36329.0          Plasmodium falciparum 3D7    Apicomplexa   \n",
       "\n",
       "                 class taxonomy_group  info  \\\n",
       "0           Echinoidea           None  None   \n",
       "1           Echinoidea           None  None   \n",
       "2                 Aves           None  None   \n",
       "3    Oligohymenophorea           None  None   \n",
       "4    Oligohymenophorea           None  None   \n",
       "..                 ...            ...   ...   \n",
       "564      Kinetoplastea           None  None   \n",
       "565           Mammalia           None  None   \n",
       "566      Kinetoplastea           None  None   \n",
       "567    Sordariomycetes           None  None   \n",
       "568       Aconoidasida           None  None   \n",
       "\n",
       "                                              sequence  \\\n",
       "0    MAPTAQVAKKGSKKAVKAPRPSGGKKRNRKRKESYGIYIYKVLKQV...   \n",
       "1    MPAKGAATKGEKKQAVKSKAMASSRTGDKKRRRRRLESYNIYIYKV...   \n",
       "2    MSSRGGKKKSTKTSRSAKAGVIFPVGRMLRYIKKGHPKYRIGVGAP...   \n",
       "3    MSTTGKGGKAKGKTASSKQVSRSARAGLQFPVGRISRFLKHGRYSE...   \n",
       "4    MSTTGKGGKAKGKTASSKQVSRSARAGLQFPVGRISRFLKNGRYSE...   \n",
       "..                                                 ...   \n",
       "564  MSLTGDDAVPQAPLVGGVAMSPEQASALTGGKLGGKAVGPAHGKGK...   \n",
       "565  MSPRRHLWNCRRSRRHSLSRSTRAELQFPVSRVDRLLREGQGAYRL...   \n",
       "566  MAKGKRVGESKGAQKRQKKVLRDNVRGITRGSIRRLARRAGVKRIS...   \n",
       "567  MARTKQTARKSTGGKAPRKQLASKAARKSAPSTGGVKKPHRYKPGT...   \n",
       "568  MARTKQTARKSTGGKAPRKQLASKAARKSAPVSTGIKKPHRYRPGT...   \n",
       "\n",
       "    variant_under_consideration  \n",
       "0                          None  \n",
       "1                          None  \n",
       "2                          None  \n",
       "3                          None  \n",
       "4                          None  \n",
       "..                          ...  \n",
       "564                        None  \n",
       "565                        None  \n",
       "566                        None  \n",
       "567                        None  \n",
       "568                        None  \n",
       "\n",
       "[569 rows x 13 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "sequence_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96b92734-0532-41a6-8315-9d69b59d8894",
   "metadata": {},
   "source": [
    "# Saving to files and checking"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a7f0161-d00a-476a-bd26-366c4db526b9",
   "metadata": {},
   "source": [
    "## Saving to JSON"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "abb4ded5-1cf2-4c0c-923a-8db5b206832c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_tree_dict(cursor, paretnt_id):\n",
    "    cursor.execute(f\"SELECT id FROM histone WHERE parent = '{paretnt_id}'\")\n",
    "    res = cursor.fetchall()\n",
    "    if len(res) < 1:\n",
    "        return \"null\"\n",
    "    return {v: get_tree_dict(cursor, v) for v, *_ in res}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4198ddc0-b173-4a0f-977c-968d14c55438",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def dict_clean(items, default=\"null\"):\n",
    "    result = {}\n",
    "    for key, value in items.items():\n",
    "        if value is None:\n",
    "            if default is None:\n",
    "                continue\n",
    "            value = default\n",
    "        result[key] = value\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4396c1b6-21e4-4b03-ad6d-6e0d13238bb9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def replace_key_in_dict(data, old_key, new_key):\n",
    "    if isinstance(data, dict):\n",
    "        new_data = {}\n",
    "        for key, value in data.items():\n",
    "            # Если ключ совпадает с `old_key`, заменяем его на `new_key`\n",
    "            if key == old_key:\n",
    "                new_data[new_key] = replace_key_in_dict(value, old_key, new_key)\n",
    "            else:\n",
    "                new_data[key] = replace_key_in_dict(value, old_key, new_key)\n",
    "        return new_data\n",
    "    elif isinstance(data, list):\n",
    "        # Рекурсивно обрабатываем списки\n",
    "        return [replace_key_in_dict(item, old_key, new_key) for item in data]\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "44470221-f0a6-47c3-9cfc-e3b5e56741b5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 7min 40s, sys: 8.98 s, total: 7min 49s\n",
      "Wall time: 7min 27s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "query = \"SELECT id FROM histone WHERE parent IS NULL\"\n",
    "cursor.execute(query)\n",
    "types = cursor.fetchall()\n",
    "variants_tree = {}\n",
    "for t, *_ in types:\n",
    "    variants_tree[t] = get_tree_dict(cursor, t)\n",
    "\n",
    "variants_info = {}\n",
    "query = (\n",
    "    \"SELECT hdap.*, p.title, p.doi, p.author, p.year FROM ( \"\n",
    "    \"SELECT hda.*, publication_id FROM ( \"\n",
    "    \"SELECT hd.*, an.name, an.taxonomy as an_taxonomy, an.gene, an.splice FROM ( \"\n",
    "    \"SELECT h.id, \"\n",
    "    \"h.level,\"\n",
    "    \"h.taxonomic_span, \"\n",
    "    \"h.taxonomic_span_id, \"\n",
    "    \"h.parent, \"\n",
    "    \"d.summary, \"\n",
    "    \"d.taxonomy, \"\n",
    "    \"d.genes, \"\n",
    "    \"d.evolution, \"\n",
    "    \"d.expression, \"\n",
    "    \"d.knock_out, \"\n",
    "    \"d.function, \"\n",
    "    \"d.sequence, \"\n",
    "    \"d.localization, \"\n",
    "    \"d.deposition,\"\n",
    "    \"d.structure,\"\n",
    "    \"d.interactions, \"\n",
    "    \"d.disease, \"\n",
    "    \"d.caveats FROM histone h LEFT JOIN histone_description d ON h.description = d.id \"\n",
    "    \") hd LEFT JOIN alternative_name an ON hd.id = an.histone \"\n",
    "    \") hda LEFT JOIN histone_has_publication hp ON hda.id = hp.histone_id \"\n",
    "    \") hdap LEFT JOIN publication p ON hdap.publication_id = p.id;\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "variants_info_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ")\n",
    "for var_name in variants_info_df.id.unique():\n",
    "    # print(var_name)\n",
    "    level_value = variants_info_df[variants_info_df[\"id\"] == var_name].iloc[0][\"level\"]\n",
    "    taxonomic_span_value = variants_info_df[variants_info_df[\"id\"] == var_name].iloc[0][\n",
    "        \"taxonomic_span\"\n",
    "    ]\n",
    "    taxonomic_span_id_value = variants_info_df[variants_info_df[\"id\"] == var_name].iloc[\n",
    "        0\n",
    "    ][\"taxonomic_span_id\"]\n",
    "    variants_info[var_name] = {\n",
    "        \"level\": level_value if level_value else \"null\",\n",
    "        \"description\": {},\n",
    "        \"taxonomic_span\": taxonomic_span_value if taxonomic_span_value else \"null\",\n",
    "        \"taxonomic_span_id\": taxonomic_span_id_value\n",
    "        if taxonomic_span_id_value\n",
    "        else \"null\",\n",
    "        \"alternate_names\": [],\n",
    "        \"publications\": [],\n",
    "    }\n",
    "    var_publications = (\n",
    "        variants_info_df.groupby([\"id\"])[\"publication_id\"]\n",
    "        .apply(lambda x: [] if x.dropna().empty else list(x.unique()))\n",
    "        .reset_index()\n",
    "    )\n",
    "    variants_info[var_name][\"publications\"] = var_publications[\n",
    "        var_publications[\"id\"] == var_name\n",
    "    ].iloc[0][\"publication_id\"]\n",
    "    var_names = (\n",
    "        variants_info_df.groupby([\"id\"])[[\"name\", \"an_taxonomy\", \"gene\", \"splice\"]]\n",
    "        .apply(\n",
    "            lambda x: []\n",
    "            if x.drop_duplicates().dropna(how=\"all\").empty\n",
    "            else list(\n",
    "                map(\n",
    "                    lambda y: dict_clean(y, default=None),\n",
    "                    x.drop_duplicates()\n",
    "                    .rename(columns={\"an_taxonomy\": \"taxonomy\"})\n",
    "                    .to_dict(\"records\"),\n",
    "                )\n",
    "            )\n",
    "        )\n",
    "        .reset_index()\n",
    "    )\n",
    "    variants_info[var_name][\"alternate_names\"] = var_names[\n",
    "        var_names[\"id\"] == var_name\n",
    "    ].iloc[0][0]\n",
    "    default_description = {\n",
    "        \"summary\": \"null\",\n",
    "        \"taxonomy\": \"null\",\n",
    "        \"genes\": \"null\",\n",
    "        \"evolution\": \"null\",\n",
    "        \"expression\": \"null\",\n",
    "        \"knock-out\": \"null\",\n",
    "        \"function\": \"null\",\n",
    "        \"sequence\": \"null\",\n",
    "        \"localization\": \"null\",\n",
    "        \"deposition\": \"null\",\n",
    "        \"structure\": \"null\",\n",
    "        \"interactions\": \"null\",\n",
    "        \"disease\": \"null\",\n",
    "        \"caveats\": \"null\",\n",
    "    }\n",
    "    var_desc = (\n",
    "        variants_info_df.groupby([\"id\"])[\n",
    "            [\n",
    "                \"summary\",\n",
    "                \"taxonomy\",\n",
    "                \"genes\",\n",
    "                \"evolution\",\n",
    "                \"expression\",\n",
    "                \"knock_out\",\n",
    "                \"function\",\n",
    "                \"sequence\",\n",
    "                \"localization\",\n",
    "                \"deposition\",\n",
    "                \"structure\",\n",
    "                \"interactions\",\n",
    "                \"disease\",\n",
    "                \"caveats\",\n",
    "            ]\n",
    "        ]\n",
    "        .apply(\n",
    "            lambda x: default_description\n",
    "            if x.drop_duplicates().dropna(how=\"all\").empty\n",
    "            else dict_clean(x.drop_duplicates().to_dict(\"records\")[0])\n",
    "        )\n",
    "        .reset_index()\n",
    "    )\n",
    "    variants_info[var_name][\"description\"] = var_desc[var_desc[\"id\"] == var_name].iloc[\n",
    "        0\n",
    "    ][0]\n",
    "    # Заменяем ключи во всех словарях\n",
    "    for key, value in variants_info.items():\n",
    "        variants_info[key] = replace_key_in_dict(value, \"knock_out\", \"knock-out\")\n",
    "\n",
    "with open(\"curated_service/curatedDB/classification_241022.json\", \"w\") as outfile:\n",
    "    json.dump({\"tree\": variants_tree, \"info\": variants_info}, outfile, indent=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "387ab3f7-c1f4-4dd7-a2d2-b3323d77c668",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Сравним полученный JSON с classification.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "3ed4e7f9-bcfd-484e-bebc-0c1854d9dca9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open(\n",
    "    \"/home/l_singh/_scratch/hdb/project_dir/histonedb/CURATED_SET/classification.json\",\n",
    "    \"r\",\n",
    "    encoding=\"utf-8\",\n",
    ") as file:\n",
    "    data1 = json.load(file)\n",
    "\n",
    "with open(\"curated_service/curatedDB/classification_241022.json\", \"r\", encoding=\"utf-8\") as file:\n",
    "    data2 = json.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "55c378b2-9234-4ff6-80b3-d0ede2f5f157",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Функция для приведения всех значений к строкам (рекурсивно)\n",
    "def convert_values_to_str(data):\n",
    "    if isinstance(data, dict):\n",
    "        return {key: convert_values_to_str(value) for key, value in data.items()}\n",
    "    elif isinstance(data, list):\n",
    "        return [convert_values_to_str(item) for item in data]\n",
    "    elif isinstance(\n",
    "        data, (int, float, bool)\n",
    "    ):  # Преобразуем числа и логические значения в строки\n",
    "        return str(data)\n",
    "    return data  # Возвращаем неизмененные строковые значения"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "27107f3f-f7ff-4112-847c-85391baa297c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Приводим все значения к строкам для корректного сравнения\n",
    "data1 = convert_values_to_str(data1)\n",
    "data2 = convert_values_to_str(data2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "47c155e5-340d-4451-88bd-a7aded4addad",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from deepdiff import DeepDiff\n",
    "\n",
    "# Используем DeepDiff для сравнения\n",
    "differences = DeepDiff(data1, data2, ignore_order=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "8fb6d377-4d7a-4905-92e3-086f984779b6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys([])"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "differences.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7596a89-583f-42f9-95eb-456d8f2a093d",
   "metadata": {},
   "source": [
    "### Файлы идентичны!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c6be27c-9fa6-4214-845b-3a151dfe237f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Если есть различия, выводим их\n",
    "if differences:\n",
    "    print(\"Найдены различия:\")\n",
    "    # print(\"dictionary_item_added:\")\n",
    "    # print(differences[\"dictionary_item_added\"])\n",
    "    # print(\"dictionary_item_removed:\")\n",
    "    # print(differences['dictionary_item_removed'])\n",
    "    print(\"values_changed:\")\n",
    "    for k in list(differences['values_changed'].keys()):\n",
    "        # if 'level' in k or 'taxonomic_span' in k: continue\n",
    "        print(f\"{k}: {differences['values_changed'][k]}\")\n",
    "else:\n",
    "    print(\"Файлы идентичны.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "defafd96-8848-485b-91f9-dd95bae45798",
   "metadata": {},
   "source": [
    "## Saving to CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "27800874-6fe9-4d3c-bac0-b4b7e2e690a2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def find_parent_by_level(df, variant_name, target_level):\n",
    "    if variant_name == \"\":\n",
    "        return \"\"\n",
    "\n",
    "    # Ищем строку с данным названием варианта\n",
    "    current_row = df[df[\"id\"] == variant_name]\n",
    "    if current_row.empty:\n",
    "        return \"HISTONE_NOT_EXIST\"\n",
    "\n",
    "    if current_row[\"level\"].values[0] == target_level:\n",
    "        return current_row[\"id\"].values[0]\n",
    "\n",
    "    # Пока не найден родитель с нужным уровнем\n",
    "    while not current_row.empty:\n",
    "        # Получаем значение поля parent\n",
    "        parent_id = current_row[\"parent\"].values[0]\n",
    "\n",
    "        if pd.isna(parent_id):\n",
    "            print(\"here\")\n",
    "            return None  # Нет родителя\n",
    "\n",
    "        # Ищем родителя\n",
    "        parent_row = df[df[\"id\"] == parent_id]\n",
    "\n",
    "        # Проверяем уровень родителя\n",
    "        if parent_row[\"level\"].values[0] == target_level:\n",
    "            return parent_row[\"id\"].values[0]\n",
    "\n",
    "        # Продолжаем искать родителя на следующем уровне\n",
    "        current_row = parent_row\n",
    "\n",
    "    return None  # Если не найдено"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3b1d6dbb-c7f9-4ddf-83b0-d17a32fd1f98",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAA30025.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>cH2B</td>\n",
       "      <td>cH2B_(Animals)</td>\n",
       "      <td></td>\n",
       "      <td>161386</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7660</td>\n",
       "      <td>Psammechinus miliaris</td>\n",
       "      <td>Echinodermata</td>\n",
       "      <td>Echinoidea</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MAPTAQVAKKGSKKAVKAPRPSGGKKRNRKRKESYGIYIYKVLKQV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAB48832.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>CS_H2B_(Echinoidea)</td>\n",
       "      <td>CS_H2B_(Echinoidea)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7660</td>\n",
       "      <td>Psammechinus miliaris</td>\n",
       "      <td>Echinodermata</td>\n",
       "      <td>Echinoidea</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9032246</td>\n",
       "      <td>MPAKGAATKGEKKQAVKSKAMASSRTGDKKRRRRRLESYNIYIYKV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAC28846.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>macroH2A</td>\n",
       "      <td>macroH2A</td>\n",
       "      <td></td>\n",
       "      <td>3395780</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9031</td>\n",
       "      <td>Gallus gallus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Aves</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>22650316 26989147</td>\n",
       "      <td>MSSRGGKKKSTKTSRSAKAGVIFPVGRMLRYIKKGHPKYRIGVGAP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAC37291.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>H2A.X</td>\n",
       "      <td>H2A.X_(Protists)</td>\n",
       "      <td></td>\n",
       "      <td>310870</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5911</td>\n",
       "      <td>Tetrahymena thermophila</td>\n",
       "      <td>Ciliophora</td>\n",
       "      <td>Oligohymenophorea</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>22650316 26989147</td>\n",
       "      <td>MSTTGKGGKAKGKTASSKQVSRSARAGLQFPVGRISRFLKHGRYSE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAC37292.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A_(Protists)</td>\n",
       "      <td></td>\n",
       "      <td>310872</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5911</td>\n",
       "      <td>Tetrahymena thermophila</td>\n",
       "      <td>Ciliophora</td>\n",
       "      <td>Oligohymenophorea</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>22650316 26989147</td>\n",
       "      <td>MSTTGKGGKAKGKTASSKQVSRSARAGLQFPVGRISRFLKNGRYSE...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    accession type        variant_group              variant doublet       gi  \\\n",
       "0  AAA30025.1  H2B                 cH2B       cH2B_(Animals)           161386   \n",
       "1  AAB48832.1  H2B  CS_H2B_(Echinoidea)  CS_H2B_(Echinoidea)                    \n",
       "2  AAC28846.1  H2A             macroH2A             macroH2A          3395780   \n",
       "3  AAC37291.1  H2A                H2A.X     H2A.X_(Protists)           310870   \n",
       "4  AAC37292.1  H2A                 cH2A      cH2A_(Protists)           310872   \n",
       "\n",
       "  ncbi_gene_id hgnc_gene_name  taxonomy_id                 organism  \\\n",
       "0                                     7660    Psammechinus miliaris   \n",
       "1                                     7660    Psammechinus miliaris   \n",
       "2                                     9031            Gallus gallus   \n",
       "3                                     5911  Tetrahymena thermophila   \n",
       "4                                     5911  Tetrahymena thermophila   \n",
       "\n",
       "          phylum              class taxonomy_group info         references  \\\n",
       "0  Echinodermata         Echinoidea                                          \n",
       "1  Echinodermata         Echinoidea                                9032246   \n",
       "2       Chordata               Aves                      22650316 26989147   \n",
       "3     Ciliophora  Oligohymenophorea                      22650316 26989147   \n",
       "4     Ciliophora  Oligohymenophorea                      22650316 26989147   \n",
       "\n",
       "                                            sequence  \n",
       "0  MAPTAQVAKKGSKKAVKAPRPSGGKKRNRKRKESYGIYIYKVLKQV...  \n",
       "1  MPAKGAATKGEKKQAVKSKAMASSRTGDKKRRRRRLESYNIYIYKV...  \n",
       "2  MSSRGGKKKSTKTSRSAKAGVIFPVGRMLRYIKKGHPKYRIGVGAP...  \n",
       "3  MSTTGKGGKAKGKTASSKQVSRSARAGLQFPVGRISRFLKHGRYSE...  \n",
       "4  MSTTGKGGKAKGKTASSKQVSRSARAGLQFPVGRISRFLKNGRYSE...  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp ON s.accession = sp.sequence_accession\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ")\n",
    "\n",
    "sequence_df = (\n",
    "    sequence_df.groupby([\"accession\"])\n",
    "    .agg(lambda x: \"\" if x.dropna().empty else \" \".join(list(map(str, x.unique()))))\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "sequence_df = sequence_df.drop(columns=[\"sequence_accession\"])\n",
    "sequence_df[\"taxonomy_id\"] = (\n",
    "    pd.to_numeric(sequence_df[\"taxonomy_id\"], errors=\"coerce\")\n",
    "    .fillna(np.nan)\n",
    "    .astype(\"Int64\")\n",
    ")\n",
    "\n",
    "query = \"SELECT * FROM histone\"\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "sequence_df[\"variant_group\"] = sequence_df[\"variant\"].apply(\n",
    "    lambda x: find_parent_by_level(histone_df, x, \"variant_group\")\n",
    ")\n",
    "sequence_df[\"type\"] = sequence_df[\"variant\"].apply(\n",
    "    lambda x: find_parent_by_level(histone_df, x, \"type\")\n",
    ")\n",
    "\n",
    "sequence_df[\"doublet\"] = [\"\"] * sequence_df.shape[0]\n",
    "\n",
    "sequence_df = (\n",
    "    sequence_df[\n",
    "        [\n",
    "            \"accession\",\n",
    "            \"type\",\n",
    "            \"variant_group\",\n",
    "            \"variant\",\n",
    "            # \"variant_under_consideration\",\n",
    "            \"doublet\",\n",
    "            \"gi\",\n",
    "            \"ncbi_gene_id\",\n",
    "            \"hgnc_gene_name\",\n",
    "            \"taxonomy_id\",\n",
    "            \"organism\",\n",
    "            \"phylum\",\n",
    "            \"class\",\n",
    "            \"taxonomy_group\",\n",
    "            \"info\",\n",
    "            \"publication_id\",\n",
    "            \"sequence\",\n",
    "        ]\n",
    "    ]\n",
    "    .rename(columns={\"publication_id\": \"references\"})\n",
    "    .replace(\"None\", \"\")\n",
    ")\n",
    "sequence_df.to_csv('curated_service/curatedDB/histones_241022.csv', index=False)\n",
    "\n",
    "sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "a1fc6058-18f6-4fbe-9e9e-04480fecda84",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(569, 16)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ac7e5309-ec28-4aad-af71-ae1dc01c36ef",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, type, variant_group, variant, doublet, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, references, sequence]\n",
       "Index: []"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df[sequence_df[\"variant_group\"] == \"HISTONE_NOT_EXIST\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "33dd0cec-4d05-4ccf-a10a-15810af8c377",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "accession          0\n",
       "type               0\n",
       "variant_group      0\n",
       "variant            0\n",
       "doublet            0\n",
       "gi                 0\n",
       "ncbi_gene_id       0\n",
       "hgnc_gene_name     0\n",
       "taxonomy_id       20\n",
       "organism           0\n",
       "phylum             0\n",
       "class              0\n",
       "taxonomy_group     0\n",
       "info               0\n",
       "references         0\n",
       "sequence           0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df.isna().sum(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "34d279cf-f57a-45fd-9c11-e72bfee056b2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, type, variant_group, variant, doublet, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, references, sequence]\n",
       "Index: []"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df[sequence_df[\"phylum\"] == \"None\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b97d6975-8270-4ad7-ac8e-49baf2415b82",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Сравним полученную таблицу с histones.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "ef46a558-67b7-4069-b293-207170e518c2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_histonescsv = pd.read_csv(\n",
    "    \"/home/l_singh/_scratch/hdb/project_dir/histonedb/CURATED_SET/histones.csv\"\n",
    ").fillna(\"\")\n",
    "df_histonescsv[\"taxonomy_id\"] = (\n",
    "    pd.to_numeric(df_histonescsv[\"taxonomy_id\"], errors=\"coerce\")\n",
    "    .fillna(np.nan)\n",
    "    .astype(\"Int64\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "01c65549-e898-4442-a02e-11527c1a42c4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Функция для преобразования строкового представления вещественных чисел в строковые целые\n",
    "def convert_to_integer_string(value):\n",
    "    if value != \"\":  # Если значение не пустое\n",
    "        return str(\n",
    "            int(float(value))\n",
    "        )  # Преобразуем в float, затем в int, и снова в строку\n",
    "    return value  # Если значение пустое, оставляем его без изменений\n",
    "\n",
    "\n",
    "# Применяем функцию к колонке ncbi_gene_id\n",
    "df_histonescsv[\"ncbi_gene_id\"] = df_histonescsv[\"ncbi_gene_id\"].apply(\n",
    "    convert_to_integer_string\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "e0906420-2d43-4ee3-826a-ea0a77ee7145",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>HISTDB_H2A_L_0</td>\n",
       "      <td>H2A</td>\n",
       "      <td>short_H2A</td>\n",
       "      <td>H2A.L.1_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>115482714</td>\n",
       "      <td>H2AL1Q</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>29549088 35099534</td>\n",
       "      <td>MAGNKHSRSSCKPRRQCLSRSRRAELQFPVSHMERCLREGQYARHL...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         accession type variant_group                 variant doublet gi  \\\n",
       "49  HISTDB_H2A_L_0  H2A     short_H2A  H2A.L.1_(Homo_sapiens)              \n",
       "\n",
       "   ncbi_gene_id hgnc_gene_name  taxonomy_id      organism    phylum     class  \\\n",
       "49    115482714         H2AL1Q         9606  Homo sapiens  Chordata  Mammalia   \n",
       "\n",
       "   taxonomy_group info         references  \\\n",
       "49       Mammalia       29549088 35099534   \n",
       "\n",
       "                                             sequence  \n",
       "49  MAGNKHSRSSCKPRRQCLSRSRRAELQFPVSHMERCLREGQYARHL...  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_histonescsv[df_histonescsv[\"accession\"] == \"HISTDB_H2A_L_0\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "9a02613a-ccda-4844-b5a7-055f45d25de4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['accession', 'type', 'variant_group', 'variant', 'doublet', 'gi',\n",
       "       'ncbi_gene_id', 'hgnc_gene_name', 'taxonomy_id', 'organism', 'phylum',\n",
       "       'class', 'taxonomy_group', 'info', 'references', 'sequence'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_histonescsv.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "dd1a35af-aba7-4f3f-9683-986e0803ab41",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['accession', 'type', 'variant_group', 'variant', 'doublet', 'gi',\n",
       "       'ncbi_gene_id', 'hgnc_gene_name', 'taxonomy_id', 'organism', 'phylum',\n",
       "       'class', 'taxonomy_group', 'info', 'references', 'sequence'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "bb75886b-3560-44e4-a56e-804a0bd75b4f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Сначала проверим, совпадает ли набор столбцов\n",
    "set(df_histonescsv.columns) == set(sequence_df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "d3f471c9-e917-4e1f-8229-46dac9fe3416",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Функция для сортировки числовых элементов в колонке 'references', оставляя нечисловые элементы на месте\n",
    "def sort_references_column(df, column_name=\"references\"):\n",
    "    def sort_reference_string(reference_str):\n",
    "        elements = reference_str.split()  # Разделяем строку на элементы\n",
    "        numeric_elements = sorted(\n",
    "            [el for el in elements if el.isdigit()], key=int\n",
    "        )  # Сортируем числовые элементы\n",
    "        non_numeric_elements = [\n",
    "            el for el in elements if not el.isdigit()\n",
    "        ]  # Не числовые элементы остаются на месте\n",
    "        return \" \".join(\n",
    "            numeric_elements + non_numeric_elements\n",
    "        )  # Объединяем отсортированные числа и нечисловые\n",
    "\n",
    "    # Применяем функцию к каждому элементу колонки\n",
    "    df[column_name] = df[column_name].apply(sort_reference_string)\n",
    "    return df\n",
    "\n",
    "\n",
    "# Применяем сортировку к колонке 'references' в обеих таблицах\n",
    "df_histonescsv = sort_references_column(df_histonescsv)\n",
    "sequence_df = sort_references_column(sequence_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "1fc6d8e6-2aba-4230-bfce-f291b17245d1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Сортируем строки по всем столбцам\n",
    "df1_sorted = df_histonescsv.sort_values(by=list(df_histonescsv.columns)).reset_index(\n",
    "    drop=True\n",
    ")\n",
    "df2_sorted = sequence_df.sort_values(by=list(sequence_df.columns)).reset_index(\n",
    "    drop=True\n",
    ")\n",
    "\n",
    "# Сравниваем отсортированные DataFrame\n",
    "df1_sorted.equals(df2_sorted)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a905819f-654c-48c8-9f16-7f3f7cacbd0d",
   "metadata": {},
   "source": [
    "### Таблицы идентичны!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cc2f44e-fd7d-468c-a7d8-68ba7513ffb7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_histonescsv.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee3e0279-6c46-4b79-8934-d929417eb7ef",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sequence_df.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "c35d92e8-e582-41c2-b25b-70bb2f7a69d3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def find_differences(df1, df2):\n",
    "    # Используем merge с индикатором, чтобы отследить источник каждой строки\n",
    "    comparison = df1.merge(df2, how=\"outer\", indicator=True)\n",
    "\n",
    "    # Строки, которые есть только в первой таблице (df1)\n",
    "    only_in_df1 = comparison[comparison[\"_merge\"] == \"left_only\"]\n",
    "\n",
    "    # Строки, которые есть только во второй таблице (df2)\n",
    "    only_in_df2 = comparison[comparison[\"_merge\"] == \"right_only\"]\n",
    "\n",
    "    return only_in_df1, only_in_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "3645433d-b896-49ed-88ea-a3e025e0f6d3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "only_in_df1, only_in_df2 = find_differences(df_histonescsv_, sequence_df_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4e81bee-472e-4b09-a647-db64ad765910",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Записи, которые есть только в первой таблице (df_histonescsv):\n",
    "only_in_df1_ = only_in_df1.drop(columns=[\"_merge\"])\n",
    "only_in_df1_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77d09a27-7fc1-4f2a-8255-66492bc077d5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Записи, которые есть только во второй таблице (sequence_df):\n",
    "only_in_df2_ = only_in_df2.drop(columns=[\"_merge\"])\n",
    "only_in_df2_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ea66bd9-6ee1-40a7-85e5-7160d7dae8b6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_histonescsv.shape, sequence_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c302ee07-02c5-450d-b41b-54ba610e122f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "7f5f801c-f6c4-4131-bb1a-5388796e87d4",
   "metadata": {},
   "source": [
    "# Close connections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "b59fc48e-7a1b-4e7c-8ae4-976c84034123",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cursor.close()\n",
    "conn.close()\n",
    "tunnel.stop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59f8ca27-60d9-4dcb-b57c-4a3b01c3be12",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".conda-histdb_env",
   "language": "python",
   "name": "conda-env-.conda-histdb_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
