{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "713ac3f3-2233-4f7e-bce8-490dac1a0f26",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from mysql.connector import connection\n",
    "from sshtunnel import SSHTunnelForwarder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b8a89a61-5ae1-4e88-b41f-bd7ba2094d1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('db_curated_server_info.txt', 'r') as file:\n",
    "    lines = file.readlines()\n",
    "\n",
    "config = {}\n",
    "\n",
    "for line in lines:\n",
    "    line = line.strip()  \n",
    "    if line and not line.startswith('#'):  \n",
    "        key, value = line.split('=', 1)  \n",
    "        config[key] = value.strip()  \n",
    "\n",
    "server_name = config.get('server_name')\n",
    "srever_port = int(config.get('srever_port'))\n",
    "ssh_password = config.get('ssh_password')\n",
    "ssh_username = config.get('ssh_username')\n",
    "db_adress = config.get('db_adress')\n",
    "db_port = int(config.get('db_port'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8f76aa83-7659-4283-b61b-d4771339452a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43881\n"
     ]
    }
   ],
   "source": [
    "tunnel = SSHTunnelForwarder(\n",
    "    (server_name, srever_port),\n",
    "    ssh_password=ssh_password,\n",
    "    ssh_username=ssh_username,\n",
    "    remote_bind_address=(db_adress, db_port),\n",
    ")\n",
    "tunnel.start()\n",
    "print(tunnel.local_bind_port)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "194c4be6-c39c-4e6a-bd65-6265aff2b2e3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "conn = connection.MySQLConnection(\n",
    "    user=\"db_user\",\n",
    "    password=\"db_password\",\n",
    "    host=\"localhost\",\n",
    "    port=tunnel.local_bind_port,\n",
    "    database=\"db_name\",\n",
    ")\n",
    "cursor = conn.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "55e0025e-1aa7-429d-b742-7992a0807386",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('alternative_name',),\n",
       " ('histone',),\n",
       " ('histone_description',),\n",
       " ('histone_has_publication',),\n",
       " ('publication',),\n",
       " ('sequence',),\n",
       " ('sequence_has_publication',)]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SHOW TABLES;\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1b94c6f5-aa1f-4e8f-8fb5-e9fe90349ad2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'id, h.level, h.taxonomic_span, h.taxonomic_span_id, h.description, h.parent'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()\n",
    "\", h.\".join([i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3e9fc8d6-6cec-458f-bc88-c2b0dba44eac",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'id, d.summary, d.taxonomy, d.genes, d.evolution, d.expression, d.knock_out, d.function, d.sequence, d.localization, d.deposition, d.structure, d.interactions, d.disease, d.caveats'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone_description\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()\n",
    "\", d.\".join([i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1e69ddc4-1026-497c-ba5c-6052910bf0dd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'id, an.name, an.taxonomy, an.gene, an.splice, an.histone'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()\n",
    "\", an.\".join([i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "31e1c1a9-4f47-4561-8c86-f400aeaa5524",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'id, p.title, p.doi, p.author, p.year'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()\n",
    "\", p.\".join([i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d6228385-fb94-49fb-8ba9-fe7e7140446c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'accession, p.variant, p.gi, p.ncbi_gene_id, p.hgnc_gene_name, p.taxonomy_id, p.organism, p.phylum, p.class, p.taxonomy_group, p.info, p.sequence, p.variant_under_consideration'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()\n",
    "\", p.\".join([i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "43ea2683-0420-49de-bbfe-fdac3ab08566",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H3.3</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>49</td>\n",
       "      <td>H3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     id          level taxonomic_span taxonomic_span_id  description parent\n",
       "0  H3.3  variant_group     Eukaryotes              2759           49     H3"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone WHERE id='H3.3'\"\n",
    "cursor.execute(query)\n",
    "# for i in cursor:\n",
    "#     print(i)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1ddfa9a3-7e80-4455-b1b4-96e9abdb724a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('Archaeal',), ('H1',), ('H2A',), ('H2B',), ('H3',), ('H4',), ('Viral',)]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT id FROM histone WHERE parent IS NULL\"\n",
    "cursor.execute(query)\n",
    "types = cursor.fetchall()\n",
    "types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "abb4ded5-1cf2-4c0c-923a-8db5b206832c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_tree_dict(cursor, paretnt_id):\n",
    "    cursor.execute(f\"SELECT id FROM histone WHERE parent = '{paretnt_id}'\")\n",
    "    res = cursor.fetchall()\n",
    "    if len(res) < 1:\n",
    "        return \"null\"\n",
    "    return {v: get_tree_dict(cursor, v) for v, *_ in res}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4198ddc0-b173-4a0f-977c-968d14c55438",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def dict_clean(items, default=\"null\"):\n",
    "    result = {}\n",
    "    for key, value in items.items():\n",
    "        if value is None:\n",
    "            if default is None:\n",
    "                continue\n",
    "            value = default\n",
    "        result[key] = value\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4396c1b6-21e4-4b03-ad6d-6e0d13238bb9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def replace_key_in_dict(data, old_key, new_key):\n",
    "    if isinstance(data, dict):\n",
    "        new_data = {}\n",
    "        for key, value in data.items():\n",
    "            # Если ключ совпадает с `old_key`, заменяем его на `new_key`\n",
    "            if key == old_key:\n",
    "                new_data[new_key] = replace_key_in_dict(value, old_key, new_key)\n",
    "            else:\n",
    "                new_data[key] = replace_key_in_dict(value, old_key, new_key)\n",
    "        return new_data\n",
    "    elif isinstance(data, list):\n",
    "        # Рекурсивно обрабатываем списки\n",
    "        return [replace_key_in_dict(item, old_key, new_key) for item in data]\n",
    "    return data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "614d13e9-62ee-4154-8e1d-f08ed8ee1473",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# Save to JSON"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d88af2ee-a6d0-4459-bf0b-0f848689bd0b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 7min 36s, sys: 8.55 s, total: 7min 45s\n",
      "Wall time: 7min 24s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "query = \"SELECT id FROM histone WHERE parent IS NULL\"\n",
    "cursor.execute(query)\n",
    "types = cursor.fetchall()\n",
    "variants_tree = {}\n",
    "for t, *_ in types:\n",
    "    variants_tree[t] = get_tree_dict(cursor, t)\n",
    "\n",
    "variants_info = {}\n",
    "query = (\n",
    "    \"SELECT hdap.*, p.title, p.doi, p.author, p.year FROM ( \"\n",
    "    \"SELECT hda.*, publication_id FROM ( \"\n",
    "    \"SELECT hd.*, an.name, an.taxonomy as an_taxonomy, an.gene, an.splice FROM ( \"\n",
    "    \"SELECT h.id, \"\n",
    "    \"h.level,\"\n",
    "    \"h.taxonomic_span, \"\n",
    "    \"h.taxonomic_span_id, \"\n",
    "    \"h.parent, \"\n",
    "    \"d.summary, \"\n",
    "    \"d.taxonomy, \"\n",
    "    \"d.genes, \"\n",
    "    \"d.evolution, \"\n",
    "    \"d.expression, \"\n",
    "    \"d.knock_out, \"\n",
    "    \"d.function, \"\n",
    "    \"d.sequence, \"\n",
    "    \"d.localization, \"\n",
    "    \"d.deposition,\"\n",
    "    \"d.structure,\"\n",
    "    \"d.interactions, \"\n",
    "    \"d.disease, \"\n",
    "    \"d.caveats FROM histone h LEFT JOIN histone_description d ON h.description = d.id \"\n",
    "    \") hd LEFT JOIN alternative_name an ON hd.id = an.histone \"\n",
    "    \") hda LEFT JOIN histone_has_publication hp ON hda.id = hp.histone_id \"\n",
    "    \") hdap LEFT JOIN publication p ON hdap.publication_id = p.id;\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "variants_info_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ")\n",
    "for var_name in variants_info_df.id.unique():\n",
    "    # print(var_name)\n",
    "    level_value = variants_info_df[variants_info_df[\"id\"] == var_name].iloc[0][\"level\"]\n",
    "    taxonomic_span_value = variants_info_df[variants_info_df[\"id\"] == var_name].iloc[0][\n",
    "        \"taxonomic_span\"\n",
    "    ]\n",
    "    taxonomic_span_id_value = variants_info_df[variants_info_df[\"id\"] == var_name].iloc[\n",
    "        0\n",
    "    ][\"taxonomic_span_id\"]\n",
    "    variants_info[var_name] = {\n",
    "        \"level\": level_value if level_value else \"null\",\n",
    "        \"description\": {},\n",
    "        \"taxonomic_span\": taxonomic_span_value if taxonomic_span_value else \"null\",\n",
    "        \"taxonomic_span_id\": taxonomic_span_id_value\n",
    "        if taxonomic_span_id_value\n",
    "        else \"null\",\n",
    "        \"alternate_names\": [],\n",
    "        \"publications\": [],\n",
    "    }\n",
    "    var_publications = (\n",
    "        variants_info_df.groupby([\"id\"])[\"publication_id\"]\n",
    "        .apply(lambda x: [] if x.dropna().empty else list(x.unique()))\n",
    "        .reset_index()\n",
    "    )\n",
    "    variants_info[var_name][\"publications\"] = var_publications[\n",
    "        var_publications[\"id\"] == var_name\n",
    "    ].iloc[0][\"publication_id\"]\n",
    "    var_names = (\n",
    "        variants_info_df.groupby([\"id\"])[[\"name\", \"an_taxonomy\", \"gene\", \"splice\"]]\n",
    "        .apply(\n",
    "            lambda x: []\n",
    "            if x.drop_duplicates().dropna(how=\"all\").empty\n",
    "            else list(\n",
    "                map(\n",
    "                    lambda y: dict_clean(y, default=None),\n",
    "                    x.drop_duplicates()\n",
    "                    .rename(columns={\"an_taxonomy\": \"taxonomy\"})\n",
    "                    .to_dict(\"records\"),\n",
    "                )\n",
    "            )\n",
    "        )\n",
    "        .reset_index()\n",
    "    )\n",
    "    variants_info[var_name][\"alternate_names\"] = var_names[\n",
    "        var_names[\"id\"] == var_name\n",
    "    ].iloc[0][0]\n",
    "    default_description = {\n",
    "        \"summary\": \"null\",\n",
    "        \"taxonomy\": \"null\",\n",
    "        \"genes\": \"null\",\n",
    "        \"evolution\": \"null\",\n",
    "        \"expression\": \"null\",\n",
    "        \"knock-out\": \"null\",\n",
    "        \"function\": \"null\",\n",
    "        \"sequence\": \"null\",\n",
    "        \"localization\": \"null\",\n",
    "        \"deposition\": \"null\",\n",
    "        \"structure\": \"null\",\n",
    "        \"interactions\": \"null\",\n",
    "        \"disease\": \"null\",\n",
    "        \"caveats\": \"null\",\n",
    "    }\n",
    "    var_desc = (\n",
    "        variants_info_df.groupby([\"id\"])[\n",
    "            [\n",
    "                \"summary\",\n",
    "                \"taxonomy\",\n",
    "                \"genes\",\n",
    "                \"evolution\",\n",
    "                \"expression\",\n",
    "                \"knock_out\",\n",
    "                \"function\",\n",
    "                \"sequence\",\n",
    "                \"localization\",\n",
    "                \"deposition\",\n",
    "                \"structure\",\n",
    "                \"interactions\",\n",
    "                \"disease\",\n",
    "                \"caveats\",\n",
    "            ]\n",
    "        ]\n",
    "        .apply(\n",
    "            lambda x: default_description\n",
    "            if x.drop_duplicates().dropna(how=\"all\").empty\n",
    "            else dict_clean(x.drop_duplicates().to_dict(\"records\")[0])\n",
    "        )\n",
    "        .reset_index()\n",
    "    )\n",
    "    variants_info[var_name][\"description\"] = var_desc[var_desc[\"id\"] == var_name].iloc[\n",
    "        0\n",
    "    ][0]\n",
    "    # Заменяем ключи во всех словарях\n",
    "    for key, value in variants_info.items():\n",
    "        variants_info[key] = replace_key_in_dict(value, \"knock_out\", \"knock-out\")\n",
    "\n",
    "with open(\"sample.json\", \"w\") as outfile:\n",
    "    json.dump({\"tree\": variants_tree, \"info\": variants_info}, outfile, indent=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd24d376-9df8-472d-9e43-63421857f3f9",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Testing queries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 394,
   "id": "211899dd-ac63-4984-8357-4acadb858221",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Archaeal</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CS_H2B_(Echinoidea)</td>\n",
       "      <td>[{'name': 'cleavage H2B'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>H1</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>H1.0</td>\n",
       "      <td>[{'name': 'H1°'}, {'name': 'H5', 'taxonomy': '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>H1.1</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>macroH2A.2_(Homo_sapiens)</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>macroH2A.2_(Mammalia)</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>scH1</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>short_H2A</td>\n",
       "      <td>[{'name': 'sH2A'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>sperm_H2B_(Echinoidea)</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>183 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            id  \\\n",
       "0                     Archaeal   \n",
       "1          CS_H2B_(Echinoidea)   \n",
       "2                           H1   \n",
       "3                         H1.0   \n",
       "4                         H1.1   \n",
       "..                         ...   \n",
       "178  macroH2A.2_(Homo_sapiens)   \n",
       "179      macroH2A.2_(Mammalia)   \n",
       "180                       scH1   \n",
       "181                  short_H2A   \n",
       "182     sperm_H2B_(Echinoidea)   \n",
       "\n",
       "                                                     0  \n",
       "0                                                   []  \n",
       "1                           [{'name': 'cleavage H2B'}]  \n",
       "2                                                   []  \n",
       "3    [{'name': 'H1°'}, {'name': 'H5', 'taxonomy': '...  \n",
       "4                                                   []  \n",
       "..                                                 ...  \n",
       "178                                                 []  \n",
       "179                                                 []  \n",
       "180                                                 []  \n",
       "181                                 [{'name': 'sH2A'}]  \n",
       "182                                                 []  \n",
       "\n",
       "[183 rows x 2 columns]"
      ]
     },
     "execution_count": 394,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT hdap.*, p.title, p.doi, p.author, p.year FROM ( \"\n",
    "    \"SELECT hda.*, publication_id FROM ( \"\n",
    "    \"SELECT hd.*, an.name, an.taxonomy as an_taxonomy, an.gene, an.splice FROM ( \"\n",
    "    \"SELECT h.id, \"\n",
    "    \"h.level,\"\n",
    "    \"h.taxonomic_span, \"\n",
    "    \"h.taxonomic_span_id, \"\n",
    "    \"h.parent, \"\n",
    "    \"d.summary, \"\n",
    "    \"d.taxonomy, \"\n",
    "    \"d.genes, \"\n",
    "    \"d.evolution, \"\n",
    "    \"d.expression, \"\n",
    "    \"d.knock_out, \"\n",
    "    \"d.function, \"\n",
    "    \"d.sequence, \"\n",
    "    \"d.localization, \"\n",
    "    \"d.deposition,\"\n",
    "    \"d.structure,\"\n",
    "    \"d.interactions, \"\n",
    "    \"d.disease, \"\n",
    "    \"d.caveats FROM histone h LEFT JOIN histone_description d ON h.description = d.id \"\n",
    "    \") hd LEFT JOIN alternative_name an ON hd.id = an.histone \"\n",
    "    \") hda LEFT JOIN histone_has_publication hp ON hda.id = hp.histone_id \"\n",
    "    \") hdap LEFT JOIN publication p ON hdap.publication_id = p.id;\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "test_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "# test_df.columns\n",
    "# test_df.groupby([\"id\"])[\"publication_id\"].get_group(\"Archaeal\").dropna().empty\n",
    "# test_df2 = (\n",
    "#     test_df.groupby([\"id\"])[\"publication_id\"]\n",
    "#     .apply(lambda x: [] if x.dropna().empty else list(x.unique()))\n",
    "#     .reset_index()\n",
    "# )\n",
    "# test_df2[test_df2[\"id\"] == \"macroH2A\"].iloc[0][\"publication_id\"]\n",
    "\n",
    "# test_df.groupby([\"id\"])[\"summary\"].apply(lambda x: pd.unique(x)).reset_index()\n",
    "# default_description = {\n",
    "#     \"summary\": \"null\",\n",
    "#     \"taxonomy\": \"null\",\n",
    "#     \"genes\": \"null\",\n",
    "#     \"evolution\": \"null\",\n",
    "#     \"expression\": \"null\",\n",
    "#     \"knock-out:\": \"null\",\n",
    "#     \"function\": \"null\",\n",
    "#     \"sequence\": \"null\",\n",
    "#     \"localization\": \"null\",\n",
    "#     \"deposition\": \"null\",\n",
    "#     \"structure\": \"null\",\n",
    "#     \"interactions\": \"null\",\n",
    "#     \"disease\": \"null\",\n",
    "#     \"caveats\": \"null\"\n",
    "# }\n",
    "# test_df.groupby([\"id\"])[\n",
    "#     [\n",
    "#         \"summary\",\n",
    "#         \"taxonomy\",\n",
    "#         \"genes\",\n",
    "#         \"evolution\",\n",
    "#         \"expression\",\n",
    "#         \"knock_out\",\n",
    "#         \"function\",\n",
    "#         \"sequence\",\n",
    "#         \"localization\",\n",
    "#         \"deposition\",\n",
    "#         \"structure\",\n",
    "#         \"interactions\",\n",
    "#         \"disease\",\n",
    "#         \"caveats\",\n",
    "#     ]\n",
    "# ].apply(\n",
    "#     lambda x: default_description\n",
    "#     if x.drop_duplicates().dropna(how=\"all\").empty\n",
    "#     else dict_clean(x.drop_duplicates().to_dict(\"records\")[0])\n",
    "# ).reset_index()\n",
    "\n",
    "test_df.groupby([\"id\", \"name\"]).get_group((\"macroH2A\", \"macroH2A1\"))\n",
    "test_df.groupby([\"id\"])[[\"name\", \"an_taxonomy\", \"gene\", \"splice\"]].get_group(\n",
    "    \"macroH2A\"\n",
    ").drop_duplicates().dropna(how=\"all\").empty\n",
    "test_df2 = (\n",
    "    test_df.groupby([\"id\"])[[\"name\", \"an_taxonomy\", \"gene\", \"splice\"]]\n",
    "    .apply(\n",
    "        lambda x: []\n",
    "        if x.drop_duplicates().dropna(how=\"all\").empty\n",
    "        else list(\n",
    "            map(\n",
    "                lambda y: dict_clean(y, default=None),\n",
    "                x.drop_duplicates()\n",
    "                .rename(columns={\"an_taxonomy\": \"taxonomy\"})\n",
    "                .to_dict(\"records\"),\n",
    "            )\n",
    "        )\n",
    "    )\n",
    "    .reset_index()\n",
    ")\n",
    "test_df2\n",
    "# test_df2[test_df2[\"id\"]==\"Archaeal\"][0].iloc[0]\n",
    "# del test_df2, test_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "45a8f045-1e72-4729-94f6-3ecd87bef698",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['sequence_accession', 'publication_id']"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence_has_publication\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()\n",
    "[i[0] for i in cursor.description]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeeabb96-4082-4054-8cd9-40f30a34bf93",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Сравним полученный JSON с classification.json\n",
    "\n",
    "**Вывод:**\n",
    "То, что лежит в переменной `differences` под ключами `'dictionary_item_removed'` и `'values_changed'` необходимо добавить в БД и заново запустить скрипт из данного ноутбука."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "id": "bb2ae407-9ee4-4be3-a75d-372dbf7de733",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting deepdiff\n",
      "  Downloading deepdiff-8.0.1-py3-none-any.whl.metadata (8.5 kB)\n",
      "Collecting orderly-set==5.2.2 (from deepdiff)\n",
      "  Downloading orderly_set-5.2.2-py3-none-any.whl.metadata (6.3 kB)\n",
      "Downloading deepdiff-8.0.1-py3-none-any.whl (82 kB)\n",
      "Downloading orderly_set-5.2.2-py3-none-any.whl (11 kB)\n",
      "\u001b[33mWARNING: Error parsing dependencies of cycler: [Errno 2] No such file or directory: '/home/l_singh/.conda/envs/histdb_env/lib/python3.8/site-packages/cycler-0.10.0.dist-info/METADATA'\u001b[0m\u001b[33m\n",
      "\u001b[0mInstalling collected packages: orderly-set, deepdiff\n",
      "Successfully installed deepdiff-8.0.1 orderly-set-5.2.2\n"
     ]
    }
   ],
   "source": [
    "!pip install deepdiff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8a8bc26d-b94c-4dc6-aa53-790d05595054",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def transform_dict(info):\n",
    "    # Шаблон для полей description\n",
    "    default_description = {\n",
    "        \"summary\": \"null\",\n",
    "        \"taxonomy\": \"null\",\n",
    "        \"genes\": \"null\",\n",
    "        \"evolution\": \"null\",\n",
    "        \"expression\": \"null\",\n",
    "        \"knock-out\": \"null\",\n",
    "        \"function\": \"null\",\n",
    "        \"sequence\": \"null\",\n",
    "        \"localization\": \"null\",\n",
    "        \"deposition\": \"null\",\n",
    "        \"structure\": \"null\",\n",
    "        \"interactions\": \"null\",\n",
    "        \"disease\": \"null\",\n",
    "        \"caveats\": \"null\",\n",
    "    }\n",
    "\n",
    "    # Проверка, если description является строкой\n",
    "    if isinstance(info.get(\"description\"), str):\n",
    "        # Сохраняем значение summary из строки description\n",
    "        summary_text = info[\"description\"]\n",
    "        # Заменяем строку на словарь с заполненными значениями по умолчанию\n",
    "        info[\"description\"] = default_description.copy()\n",
    "        # Присваиваем значение summary из исходной строки\n",
    "        info[\"description\"][\"summary\"] = summary_text\n",
    "\n",
    "    return info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3d03eba0-253c-4060-9553-cdd6131fc9d3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#  удалим -- и заменим кавычки так, как это представлено в БД\n",
    "def clean_summary(data):\n",
    "    if isinstance(data, dict):\n",
    "        for key, value in data.items():\n",
    "            # Если ключ 'summary', выполняем замены\n",
    "            if key == \"summary\" and isinstance(value, str):\n",
    "                data[key] = value.replace(\"--\", \"\")  # .replace(\"'\", \"''\")\n",
    "            else:\n",
    "                # Рекурсивно обрабатываем вложенные словари\n",
    "                clean_summary(value)\n",
    "    elif isinstance(data, list):\n",
    "        # Рекурсивно обрабатываем списки\n",
    "        for item in data:\n",
    "            clean_summary(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f20462d2-109b-4ac4-a541-dbf4ed193a77",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# добавляем пропущенные поля в 'description'\n",
    "def add_missing_fields(info, required_fields):\n",
    "    for key, value in info.items():\n",
    "        if isinstance(value, dict):\n",
    "            # Проверяем наличие поля 'description'\n",
    "            if \"description\" in value:\n",
    "                # Проверяем наличие необходимых полей в 'description'\n",
    "                for field in required_fields:\n",
    "                    if field not in value[\"description\"]:\n",
    "                        value[\"description\"][\n",
    "                            field\n",
    "                        ] = \"null\"  # Или любое другое значение по умолчанию\n",
    "            # Рекурсивно обрабатываем вложенные словари\n",
    "            add_missing_fields(value, required_fields)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "fefb8b6e-9f39-4b47-80e7-66ec83ac3f9a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open(\n",
    "    \"/home/l_singh/_scratch/hdb/project_dir/histonedb/CURATED_SET/classification.json\",\n",
    "    \"r\",\n",
    "    encoding=\"utf-8\",\n",
    ") as file:\n",
    "    data1 = json.load(file)\n",
    "# Преобразуем каждый элемент в словаре data1['info']\n",
    "for key, value in data1[\"info\"].items():\n",
    "    data1[\"info\"][key] = transform_dict(value)\n",
    "for key, value in data1.items():\n",
    "    data1[key] = replace_key_in_dict(value, \"knock-out:\", \"knock-out\")\n",
    "clean_summary(data1[\"info\"])\n",
    "# Добавляем недостающие поля\n",
    "add_missing_fields(data1, list(default_description.keys())[1:])\n",
    "# with open(\"/home/l_singh/_scratch/hdb/project_dir/histonedb/CURATED_SET/classification.json\", 'w', encoding='utf-8') as file:\n",
    "#     json.dump(data1, file)\n",
    "\n",
    "with open(\"sample.json\", \"r\", encoding=\"utf-8\") as file:\n",
    "    data2 = json.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a7c0964b-7107-4cf5-93a9-6703a40c96d8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'level': 'variant_group',\n",
       " 'description': {'summary': 'Canonical histones are replication-dependent while histone variants are replication-independent, constitutively expressed during cell cycle. Genes encoding canonical histones are typically located within multigene clusters and use specific type of regulation at the RNA level with a stem loop structure instead of polyA tail',\n",
       "  'taxonomy': 'null',\n",
       "  'genes': 'null',\n",
       "  'evolution': 'null',\n",
       "  'expression': 'null',\n",
       "  'knock-out': 'null',\n",
       "  'function': 'null',\n",
       "  'sequence': 'null',\n",
       "  'localization': 'null',\n",
       "  'deposition': 'null',\n",
       "  'structure': 'null',\n",
       "  'interactions': 'null',\n",
       "  'disease': 'null',\n",
       "  'caveats': 'null'},\n",
       " 'taxonomic_span': 'Eukaryotes',\n",
       " 'taxonomic_span_id': '2759',\n",
       " 'alternate_names': [{'name': 'ca H4'}],\n",
       " 'publications': ['22650316']}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data1[\"info\"][\"cH4\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3d36766d-a109-43b7-9146-6ab2982d94d0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'level': 'variant_group',\n",
       " 'description': {'summary': 'null',\n",
       "  'taxonomy': 'null',\n",
       "  'genes': 'null',\n",
       "  'evolution': 'null',\n",
       "  'expression': 'null',\n",
       "  'knock-out': 'null',\n",
       "  'function': 'null',\n",
       "  'sequence': 'null',\n",
       "  'localization': 'null',\n",
       "  'deposition': 'null',\n",
       "  'structure': 'null',\n",
       "  'interactions': 'null',\n",
       "  'disease': 'null',\n",
       "  'caveats': 'null'},\n",
       " 'taxonomic_span': 'Giardia',\n",
       " 'taxonomic_span_id': '5740',\n",
       " 'alternate_names': [],\n",
       " 'publications': []}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data1[\"info\"][\"H3.B_(Giardia?)\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "13885489-b046-4919-a106-401e93815ddf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'level': 'variant_group',\n",
       " 'description': {'summary': 'null',\n",
       "  'taxonomy': 'null',\n",
       "  'genes': 'null',\n",
       "  'evolution': 'null',\n",
       "  'expression': 'null',\n",
       "  'knock-out': 'null',\n",
       "  'function': 'null',\n",
       "  'sequence': 'null',\n",
       "  'localization': 'null',\n",
       "  'deposition': 'null',\n",
       "  'structure': 'null',\n",
       "  'interactions': 'null',\n",
       "  'disease': 'null',\n",
       "  'caveats': 'null'},\n",
       " 'taxonomic_span': 'Giardia',\n",
       " 'taxonomic_span_id': '5740',\n",
       " 'alternate_names': [],\n",
       " 'publications': []}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data2[\"info\"][\"H3.B_(Giardia?)\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "23d00650-f67c-4050-b7ad-37a0db7ae8c4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'name': 'canonical H2A'}]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data2[\"info\"][\"cH2A\"][\"alternate_names\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "bcedb880-0821-4242-b07c-eb71cb217764",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data2[\"info\"][\"cH2A_(Mus_musculus)\"][\"alternate_names\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "bf7034bb-51f5-4f00-90e2-1059be570ee3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'name': 'H1°'},\n",
       " {'name': 'H5', 'taxonomy': 'aves'},\n",
       " {'name': 'H1δ'},\n",
       " {'name': 'RI H1'}]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data2[\"info\"][\"H1.0\"][\"alternate_names\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "09e1b78f-227d-4de5-a4f0-b64650d0bfec",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'summary': 'cH2A.10_(Homo_sapiens)  is an isoform (variant) of clustered (canonical) H2A histones in human endoded by H2AC21 gene.',\n",
       " 'taxonomy': 'null',\n",
       " 'genes': 'null',\n",
       " 'evolution': 'null',\n",
       " 'expression': 'null',\n",
       " 'knock-out': 'null',\n",
       " 'function': 'null',\n",
       " 'sequence': 'null',\n",
       " 'localization': 'null',\n",
       " 'deposition': 'null',\n",
       " 'structure': 'null',\n",
       " 'interactions': 'null',\n",
       " 'disease': 'null',\n",
       " 'caveats': 'null'}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data1[\"info\"][\"cH2A.10_(Homo_sapiens)\"][\"description\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "83a12e07-62bf-4fd2-9752-1eba8157faa7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Функция для приведения всех значений к строкам (рекурсивно)\n",
    "def convert_values_to_str(data):\n",
    "    if isinstance(data, dict):\n",
    "        return {key: convert_values_to_str(value) for key, value in data.items()}\n",
    "    elif isinstance(data, list):\n",
    "        return [convert_values_to_str(item) for item in data]\n",
    "    elif isinstance(\n",
    "        data, (int, float, bool)\n",
    "    ):  # Преобразуем числа и логические значения в строки\n",
    "        return str(data)\n",
    "    return data  # Возвращаем неизмененные строковые значения"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c9c30ef4-094b-4d52-a961-22bba2af2653",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Приводим все значения к строкам для корректного сравнения\n",
    "data1 = convert_values_to_str(data1)\n",
    "data2 = convert_values_to_str(data2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "91570f36-8c0d-47fb-ba3d-00e1d3758c23",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Найдены различия:\n",
      "dictionary_item_added:\n",
      "SetOrdered([\"root['tree']['H2B']['cH2B']['cH2B_(Animals)']['cH2B_(Vertebrata)']['cH2B_(Mammalia)']['cH2B_(Homo_sapiens)']['cH2B.15_(Homo_sapiens)']\", \"root['info']['Archaeal']\", \"root['info']['cenH3_(Eukarya)']\", \"root['info']['cH1']\", \"root['info']['H2A.J_(Homo_sapiens)']\", \"root['info']['H2A.X_(Animals)']\", \"root['info']['H2A.X_(Fungi)']\", \"root['info']['H2A.X_(Homo_sapiens)']\", \"root['info']['H2A.X_(Mammalia)']\", \"root['info']['H2A.X_(Mus_musculus)']\", \"root['info']['H2A.X_(Plants)']\", \"root['info']['H2A.X_(Protists)']\", \"root['info']['H2A.X_(Vertebrata)']\", \"root['info']['H3.1-like_(Plants)']\", \"root['info']['H3.3-like?']\", \"root['info']['H3.3-like_(Animals)']\", \"root['info']['H3.3-like_(Plants)']\", \"root['info']['H3.5_(Primates_or_Hominids?)']\", \"root['info']['H3.6_(Mammals?)?']\", \"root['info']['H3.7_(Mammals?)?']\", \"root['info']['H3.8_(Mammals?)?']\", \"root['info']['H3.X_(Homo_sapiens)']\", \"root['info']['H3.X_(Primates?)']\", \"root['info']['H3.Y_(Homo_sapiens)']\", \"root['info']['H3_(Lilly???)']\", \"root['info']['TS H3.10']\", \"root['info']['Viral']\", \"root['info']['cH2A_(Mus_musculus)']['alternate_names']\", \"root['info']['cH2B_(Homo_sapiens)']['taxonomic_span']\", \"root['info']['cH2B_(Homo_sapiens)']['taxonomic_span_id']\", \"root['info']['cH2B_(Homo_sapiens)']['alternate_names']\", \"root['info']['cH2B_(Mus_musculus)']['alternate_names']\", \"root['info']['H1']['taxonomic_span']\", \"root['info']['H1']['taxonomic_span_id']\", \"root['info']['H1']['alternate_names']\", \"root['info']['H1']['publications']\", \"root['info']['H2A']['alternate_names']\", \"root['info']['H3']['alternate_names']\", \"root['info']['H4']['taxonomic_span']\", \"root['info']['H4']['taxonomic_span_id']\", \"root['info']['H4']['alternate_names']\", \"root['info']['H4']['publications']\"])\n"
     ]
    }
   ],
   "source": [
    "from deepdiff import DeepDiff\n",
    "\n",
    "# Используем DeepDiff для сравнения\n",
    "differences = DeepDiff(data1, data2, ignore_order=True)\n",
    "\n",
    "# Если есть различия, выводим их\n",
    "if differences:\n",
    "    print(\"Найдены различия:\")\n",
    "    print(\"dictionary_item_added:\")\n",
    "    print(differences[\"dictionary_item_added\"])\n",
    "    # print(\"dictionary_item_removed:\")\n",
    "    # print(differences['dictionary_item_removed'])\n",
    "    # print(\"values_changed:\")\n",
    "    # for k in list(differences['values_changed'].keys()):\n",
    "    #     if 'level' in k or 'taxonomic_span' in k: continue\n",
    "    #     print(f\"{k}: {differences['values_changed'][k]}\")\n",
    "else:\n",
    "    print(\"Файлы идентичны.\")\n",
    "\n",
    "# with open(\"sample_diff.json\", \"w\") as outfile:\n",
    "#     json.dump(differences, outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "717f309e-bd80-4aae-8fac-61584226da25",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>mH2A</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>macroH2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>macroH2A1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>macroH2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>macroH2A2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>macroH2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>macroH2A1.1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>macroH2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>macroH2A1.2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>macroH2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>macroH2A2.1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>macroH2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>macroH2A2.2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>macroH2A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    id         name taxonomy  gene splice   histone\n",
       "9   10         mH2A     None  None   None  macroH2A\n",
       "10  11    macroH2A1     None  None   None  macroH2A\n",
       "11  12    macroH2A2     None  None   None  macroH2A\n",
       "12  13  macroH2A1.1     None  None   None  macroH2A\n",
       "13  14  macroH2A1.2     None  None   None  macroH2A\n",
       "14  15  macroH2A2.1     None  None   None  macroH2A\n",
       "15  16  macroH2A2.2     None  None   None  macroH2A"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name\"\n",
    "cursor.execute(query)\n",
    "alternative_name_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ")\n",
    "alternative_name_df[alternative_name_df[\"histone\"] == \"macroH2A\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 408,
   "id": "c5f37cd7-9203-486d-8fa1-302a323bf00e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'cH2B.15_(Homo_sapiens)'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[408], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mdata2\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43minfo\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mcH2B.15_(Homo_sapiens)\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'cH2B.15_(Homo_sapiens)'"
     ]
    }
   ],
   "source": [
    "data2[\"info\"][\"cH2B.15_(Homo_sapiens)\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efd468ce-0af5-4e45-99ff-4019373ce6a1",
   "metadata": {},
   "source": [
    "В некоторых описаниях classification.json нет полея taxonomy, taxonomic_span_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 436,
   "id": "788233a9-0f7b-4f23-9fb9-32fd078e61c0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['dictionary_item_added', 'dictionary_item_removed', 'values_changed'])"
      ]
     },
     "execution_count": 436,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "differences.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 508,
   "id": "86711e85-e143-4c16-9f9e-cce0d0efc7be",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[\"root['info']['cH3.1_(Mammalia)']['level']\",\n",
       " \"root['info']['cH3.2_(Mammalia)']['level']\",\n",
       " \"root['info']['cH3_(Animals)']['level']\",\n",
       " \"root['info']['cH3_(Chlorophyta)']['level']\",\n",
       " \"root['info']['cH3_(Chlorophyta)']['description']['summary']\",\n",
       " \"root['info']['cH3_(Embryophyta)']['level']\",\n",
       " \"root['info']['cH3_(Fungi)']['level']\",\n",
       " \"root['info']['cH3_(Mammalia)']['level']\",\n",
       " \"root['info']['cH3_(Plants)']['level']\",\n",
       " \"root['info']['cH3_(Plants)']['description']['summary']\",\n",
       " \"root['info']['cH3_(Protists)']['level']\",\n",
       " \"root['info']['cH3_(Vertebrata)']['level']\",\n",
       " \"root['info']['H2A.J']['description']['summary']\",\n",
       " \"root['info']['H2A.L']['description']['summary']\",\n",
       " \"root['info']['H2A.L.1_(Homo_sapiens)']['description']['summary']\",\n",
       " \"root['info']['H2A.L.3_(Homo_sapiens)']['description']['summary']\",\n",
       " \"root['info']['H2A.L_(Homo_sapiens)']['description']['summary']\",\n",
       " \"root['info']['H2A.P']['description']['summary']\",\n",
       " \"root['info']['H2A.P_(Homo_sapiens)']['description']['summary']\",\n",
       " \"root['info']['H2A.W']['description']['knock-out']\",\n",
       " \"root['info']['H2B.W.1_(Homo_sapiens)']['level']\",\n",
       " \"root['info']['H2B.W.2_(Homo_sapiens)']['level']\",\n",
       " \"root['info']['H2B.W_(Homo_sapiens)']['level']\",\n",
       " \"root['info']['H3.3_(Animals)']['level']\",\n",
       " \"root['info']['H3.3_(Animals)']['description']['knock-out']\",\n",
       " \"root['info']['H3.3_(Ascomycota)']['level']\",\n",
       " \"root['info']['H3.3_(Fungi)']['level']\",\n",
       " \"root['info']['H3.3_(Plants)']['level']\",\n",
       " \"root['info']['H3.3_(Plants)']['description']['knock-out']\",\n",
       " \"root['info']['H3.3_(Protists)']['level']\",\n",
       " \"root['info']['H3.4_(Mammalia)']['level']\",\n",
       " \"root['info']['H3.4_(Mammalia)']['description']['knock-out']\",\n",
       " \"root['info']['H3.Y_(Primates?)']['level']\"]"
      ]
     },
     "execution_count": 508,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(differences[\"values_changed\"].keys())[28:]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b0b769f-47d0-49b6-a353-8be93b67e3e9",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Save to CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "27800874-6fe9-4d3c-bac0-b4b7e2e690a2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def find_parent_by_level(df, variant_name, target_level):\n",
    "    if variant_name == \"\":\n",
    "        return \"\"\n",
    "\n",
    "    # Ищем строку с данным названием варианта\n",
    "    current_row = df[df[\"id\"] == variant_name]\n",
    "    if current_row.empty:\n",
    "        return \"HISTONE_NOT_EXIST\"\n",
    "\n",
    "    if current_row[\"level\"].values[0] == target_level:\n",
    "        return current_row[\"id\"].values[0]\n",
    "\n",
    "    # Пока не найден родитель с нужным уровнем\n",
    "    while not current_row.empty:\n",
    "        # Получаем значение поля parent\n",
    "        parent_id = current_row[\"parent\"].values[0]\n",
    "\n",
    "        if pd.isna(parent_id):\n",
    "            print(\"here\")\n",
    "            return None  # Нет родителя\n",
    "\n",
    "        # Ищем родителя\n",
    "        parent_row = df[df[\"id\"] == parent_id]\n",
    "\n",
    "        # Проверяем уровень родителя\n",
    "        if parent_row[\"level\"].values[0] == target_level:\n",
    "            return parent_row[\"id\"].values[0]\n",
    "\n",
    "        # Продолжаем искать родителя на следующем уровне\n",
    "        current_row = parent_row\n",
    "\n",
    "    return None  # Если не найдено"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3b1d6dbb-c7f9-4ddf-83b0-d17a32fd1f98",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAB48832.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>CS_H2B_(Echinoidea)</td>\n",
       "      <td>CS_H2B_(Echinoidea)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7660</td>\n",
       "      <td>Psammechinus miliaris</td>\n",
       "      <td>Echinodermata</td>\n",
       "      <td>Echinoidea</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9032246</td>\n",
       "      <td>MPAKGAATKGEKKQAVKSKAMASSRTGDKKRRRRRLESYNIYIYKV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>EED88841.1</td>\n",
       "      <td>H1</td>\n",
       "      <td>H1.0</td>\n",
       "      <td>H1.0</td>\n",
       "      <td></td>\n",
       "      <td>220970504</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>296543</td>\n",
       "      <td>Thalassiosira pseudonana CCMP1335</td>\n",
       "      <td>Bacillariophyta</td>\n",
       "      <td>Coscinodiscophyceae</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MSYKAGIAKAITELKDRTGSSSIAIKKHMQANLPADKKWMNATFLK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>525</th>\n",
       "      <td>XP_010887142.1</td>\n",
       "      <td>H1</td>\n",
       "      <td>H1.0</td>\n",
       "      <td>H1.0</td>\n",
       "      <td></td>\n",
       "      <td>742177175</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>8010</td>\n",
       "      <td>Esox lucius</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Actinopteri</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MAETVAAPAPKAKKAKAPKKPASHPKYSDMIKAAVQADKSRGGASR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ACH70944.1</td>\n",
       "      <td>H1</td>\n",
       "      <td>H1.0</td>\n",
       "      <td>H1.0</td>\n",
       "      <td></td>\n",
       "      <td>197632441</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>8030</td>\n",
       "      <td>Salmo salar</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Actinopteri</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MAETAAAPAPKAKKAKAPKKPASHPKYSDMIKAAVHADKSRGGASR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ACM08534.1</td>\n",
       "      <td>H1</td>\n",
       "      <td>H1.0</td>\n",
       "      <td>H1.0</td>\n",
       "      <td></td>\n",
       "      <td>221219746</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>8030</td>\n",
       "      <td>Salmo salar</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Actinopteri</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MIKAAVHADKSRGGASRQSVQKYIKSHYKVGDNADSQIKLSLKRMV...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession type        variant_group              variant doublet  \\\n",
       "1        AAB48832.1  H2B  CS_H2B_(Echinoidea)  CS_H2B_(Echinoidea)           \n",
       "37       EED88841.1   H1                 H1.0                 H1.0           \n",
       "525  XP_010887142.1   H1                 H1.0                 H1.0           \n",
       "10       ACH70944.1   H1                 H1.0                 H1.0           \n",
       "12       ACM08534.1   H1                 H1.0                 H1.0           \n",
       "\n",
       "            gi ncbi_gene_id hgnc_gene_name  taxonomy_id  \\\n",
       "1                                                  7660   \n",
       "37   220970504                                   296543   \n",
       "525  742177175                                     8010   \n",
       "10   197632441                                     8030   \n",
       "12   221219746                                     8030   \n",
       "\n",
       "                              organism           phylum                class  \\\n",
       "1                Psammechinus miliaris    Echinodermata           Echinoidea   \n",
       "37   Thalassiosira pseudonana CCMP1335  Bacillariophyta  Coscinodiscophyceae   \n",
       "525                        Esox lucius         Chordata          Actinopteri   \n",
       "10                         Salmo salar         Chordata          Actinopteri   \n",
       "12                         Salmo salar         Chordata          Actinopteri   \n",
       "\n",
       "    taxonomy_group info references  \\\n",
       "1                          9032246   \n",
       "37                                   \n",
       "525                                  \n",
       "10                                   \n",
       "12                                   \n",
       "\n",
       "                                              sequence  \n",
       "1    MPAKGAATKGEKKQAVKSKAMASSRTGDKKRRRRRLESYNIYIYKV...  \n",
       "37   MSYKAGIAKAITELKDRTGSSSIAIKKHMQANLPADKKWMNATFLK...  \n",
       "525  MAETVAAPAPKAKKAKAPKKPASHPKYSDMIKAAVQADKSRGGASR...  \n",
       "10   MAETAAAPAPKAKKAKAPKKPASHPKYSDMIKAAVHADKSRGGASR...  \n",
       "12   MIKAAVHADKSRGGASRQSVQKYIKSHYKVGDNADSQIKLSLKRMV...  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp ON s.accession = sp.sequence_accession\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ")\n",
    "\n",
    "sequence_df = (\n",
    "    sequence_df.groupby([\"accession\"])\n",
    "    .agg(lambda x: \"\" if x.dropna().empty else \" \".join(list(map(str, x.unique()))))\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "sequence_df = sequence_df.drop(columns=[\"sequence_accession\"])\n",
    "sequence_df[\"taxonomy_id\"] = (\n",
    "    pd.to_numeric(sequence_df[\"taxonomy_id\"], errors=\"coerce\")\n",
    "    .fillna(np.nan)\n",
    "    .astype(\"Int64\")\n",
    ")\n",
    "\n",
    "query = \"SELECT * FROM histone\"\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "# sequence_df[\"variant\"] = [\n",
    "#     row[\"variant\"] if row[\"variant\"] != \"\" else row[\"variant_under_consideration\"]\n",
    "#     for i, row in sequence_df.iterrows()\n",
    "# ]\n",
    "sequence_df[\"variant_group\"] = sequence_df[\"variant\"].apply(\n",
    "    lambda x: find_parent_by_level(histone_df, x, \"variant_group\")\n",
    ")\n",
    "sequence_df[\"type\"] = sequence_df[\"variant\"].apply(\n",
    "    lambda x: find_parent_by_level(histone_df, x, \"type\")\n",
    ")\n",
    "\n",
    "sequence_df[\"doublet\"] = [\"\"] * sequence_df.shape[0]\n",
    "\n",
    "sequence_df = (\n",
    "    sequence_df[\n",
    "        [\n",
    "            \"accession\",\n",
    "            \"type\",\n",
    "            \"variant_group\",\n",
    "            \"variant\",\n",
    "            # \"variant_under_consideration\",\n",
    "            \"doublet\",\n",
    "            \"gi\",\n",
    "            \"ncbi_gene_id\",\n",
    "            \"hgnc_gene_name\",\n",
    "            \"taxonomy_id\",\n",
    "            \"organism\",\n",
    "            \"phylum\",\n",
    "            \"class\",\n",
    "            \"taxonomy_group\",\n",
    "            \"info\",\n",
    "            \"publication_id\",\n",
    "            \"sequence\",\n",
    "        ]\n",
    "    ]\n",
    "    .rename(columns={\"publication_id\": \"references\"})\n",
    "    .replace(\"None\", \"\")\n",
    ")\n",
    "sequence_df = sequence_df.sort_values([\"variant_group\", \"variant\", \"phylum\", \"class\", \"organism\"])\n",
    "sequence_df.to_csv('histones_sample.csv', index=False)\n",
    "\n",
    "sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ac7e5309-ec28-4aad-af71-ae1dc01c36ef",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, type, variant_group, variant, doublet, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, references, sequence]\n",
       "Index: []"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df[sequence_df[\"variant_group\"] == \"HISTONE_NOT_EXIST\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "33dd0cec-4d05-4ccf-a10a-15810af8c377",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "accession          0\n",
       "type               0\n",
       "variant_group      0\n",
       "variant            0\n",
       "doublet            0\n",
       "gi                 0\n",
       "ncbi_gene_id       0\n",
       "hgnc_gene_name     0\n",
       "taxonomy_id       20\n",
       "organism           0\n",
       "phylum             0\n",
       "class              0\n",
       "taxonomy_group     0\n",
       "info               0\n",
       "references         0\n",
       "sequence           0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df.isna().sum(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "34d279cf-f57a-45fd-9c11-e72bfee056b2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, type, variant_group, variant, doublet, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, references, sequence]\n",
       "Index: []"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df[sequence_df[\"phylum\"] == \"None\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab18cf2e-325d-4461-95d9-490aaa1bf9dd",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Testing queries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "673a11dc-fd2f-47e8-a387-3b96a2b3baf3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "query = \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp ON s.accession = sp.sequence_accession\"\n",
    "cursor.execute(query)\n",
    "sequence_df = pd.DataFrame(\n",
    "    cursor.fetchall(), columns=[i[0] for i in cursor.description]\n",
    ")\n",
    "sequence_df.groupby([\"accession\"]).get_group(\"AAC28846.1\")\n",
    "\n",
    "sequence_df.groupby([\"accession\"]).agg(\n",
    "    lambda x: \"\" if x.dropna().empty else \" \".join(list(map(str, x.unique())))\n",
    ").reset_index().to_csv(\"test.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "6a53e873-696b-47f3-8be2-031e9c338e96",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Родительский вариант для macroH2A.1 с уровнем variant_group: HISTONE_NOT_EXIST\n"
     ]
    }
   ],
   "source": [
    "def find_parent_by_level(df, variant_name, target_level):\n",
    "    # Ищем строку с данным названием варианта\n",
    "    current_row = df[df[\"id\"] == variant_name]\n",
    "    if current_row.empty:\n",
    "        return \"HISTONE_NOT_EXIST\"\n",
    "\n",
    "    if current_row[\"level\"].values[0] == target_level:\n",
    "        return current_row[\"id\"].values[0]\n",
    "\n",
    "    # Пока не найден родитель с нужным уровнем\n",
    "    while not current_row.empty:\n",
    "        # Получаем значение поля parent\n",
    "        parent_id = current_row[\"parent\"].values[0]\n",
    "\n",
    "        if pd.isna(parent_id):\n",
    "            print(\"here\")\n",
    "            return None  # Нет родителя\n",
    "\n",
    "        # Ищем родителя\n",
    "        parent_row = df[df[\"id\"] == parent_id]\n",
    "\n",
    "        # Проверяем уровень родителя\n",
    "        if parent_row[\"level\"].values[0] == target_level:\n",
    "            return parent_row[\"id\"].values[0]\n",
    "\n",
    "        # Продолжаем искать родителя на следующем уровне\n",
    "        current_row = parent_row\n",
    "\n",
    "    return None  # Если не найдено\n",
    "\n",
    "\n",
    "# Пример использования функции\n",
    "\n",
    "query = \"SELECT * FROM histone\"\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "\n",
    "variant_name = \"macroH2A.1\"\n",
    "target_level = \"variant_group\"\n",
    "parent_name = find_parent_by_level(histone_df, variant_name, target_level)\n",
    "\n",
    "print(\n",
    "    f\"Родительский вариант для {variant_name} с уровнем {target_level}: {parent_name}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52644c76-e5f4-4e91-8520-de89be116159",
   "metadata": {},
   "outputs": [],
   "source": [
    "query = \"SELECT id FROM histone WHERE parent IS NULL\"\n",
    "cursor.execute(query)\n",
    "types = cursor.fetchall()\n",
    "variants_tree = {}\n",
    "for t, *_ in types:\n",
    "    variants_tree[t] = get_tree_dict(cursor, t)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc292ae0-29e4-431d-a3d5-dc05b0310362",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Сравним полученную таблицу с histones.csv\n",
    "\n",
    "**Вывод:** Будем считать сгенерированную таблицу из БД итоговой, то есть значения в поле `variant`, которые ранее содержали `__???`, будут пустыми, а также не будет `?` в поле `references`. Возможно стоит добавить итоговую сортировку так, чтобы записи вариантов были в том же порядке, что и в дереве."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "5867181f-4129-4b53-9c1a-3afe3de951b7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_histonescsv = pd.read_csv(\n",
    "    \"/home/l_singh/_scratch/hdb/project_dir/histonedb/CURATED_SET/histones.csv\"\n",
    ").fillna(\"\")\n",
    "df_histonescsv[\"taxonomy_id\"] = (\n",
    "    pd.to_numeric(df_histonescsv[\"taxonomy_id\"], errors=\"coerce\")\n",
    "    .fillna(np.nan)\n",
    "    .astype(\"Int64\")\n",
    ")\n",
    "df_histonescsv[\"references\"] = [\n",
    "    r.replace(\"?\", \"\") for r in df_histonescsv[\"references\"]\n",
    "]\n",
    "df_histonescsv[\"references\"] = [\n",
    "    r.replace(\"()\", \"\") for r in df_histonescsv[\"references\"]\n",
    "]\n",
    "df_histonescsv[\"doublet\"] = [\"\"] * df_histonescsv.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "81b3957a-5251-4888-aa2e-ae274eff2038",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Функция для преобразования строкового представления вещественных чисел в строковые целые\n",
    "def convert_to_integer_string(value):\n",
    "    if value != \"\":  # Если значение не пустое\n",
    "        return str(\n",
    "            int(float(value))\n",
    "        )  # Преобразуем в float, затем в int, и снова в строку\n",
    "    return value  # Если значение пустое, оставляем его без изменений\n",
    "\n",
    "\n",
    "# Применяем функцию к колонке ncbi_gene_id\n",
    "df_histonescsv[\"ncbi_gene_id\"] = df_histonescsv[\"ncbi_gene_id\"].apply(\n",
    "    convert_to_integer_string\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "fedd819c-acfa-4c72-b124-109a2684a733",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>HISTDB_H2A_L_0</td>\n",
       "      <td>H2A</td>\n",
       "      <td>short_H2A</td>\n",
       "      <td>H2A.L.1_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>115482714</td>\n",
       "      <td>H2AL1Q</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>29549088 35099534</td>\n",
       "      <td>MAGNKHSRSSCKPRRQCLSRSRRAELQFPVSHMERCLREGQYARHL...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession type variant_group                 variant doublet gi  \\\n",
       "153  HISTDB_H2A_L_0  H2A     short_H2A  H2A.L.1_(Homo_sapiens)              \n",
       "\n",
       "    ncbi_gene_id hgnc_gene_name  taxonomy_id      organism    phylum  \\\n",
       "153    115482714         H2AL1Q         9606  Homo sapiens  Chordata   \n",
       "\n",
       "        class taxonomy_group info         references  \\\n",
       "153  Mammalia       Mammalia       29549088 35099534   \n",
       "\n",
       "                                              sequence  \n",
       "153  MAGNKHSRSSCKPRRQCLSRSRRAELQFPVSHMERCLREGQYARHL...  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_histonescsv[df_histonescsv[\"accession\"] == \"HISTDB_H2A_L_0\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "6d0d4da4-b281-4256-8ef5-2f5385d78e98",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['accession', 'type', 'variant_group', 'variant', 'doublet', 'gi',\n",
       "       'ncbi_gene_id', 'hgnc_gene_name', 'taxonomy_id', 'organism', 'phylum',\n",
       "       'class', 'taxonomy_group', 'info', 'references', 'sequence'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_histonescsv.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "4ce0a4db-7d2d-48cf-8205-4bf63e743e2f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['accession', 'type', 'variant_group', 'variant', 'doublet', 'gi',\n",
       "       'ncbi_gene_id', 'hgnc_gene_name', 'taxonomy_id', 'organism', 'phylum',\n",
       "       'class', 'taxonomy_group', 'info', 'references', 'sequence'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "e1ea54c5-1b55-4154-b6b4-72c590a5b8ea",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Сначала проверим, совпадает ли набор столбцов\n",
    "set(df_histonescsv.columns) == set(sequence_df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "c0b1da7e-103f-437a-9625-069299306d2b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Функция для сортировки числовых элементов в колонке 'references', оставляя нечисловые элементы на месте\n",
    "def sort_references_column(df, column_name=\"references\"):\n",
    "    def sort_reference_string(reference_str):\n",
    "        elements = reference_str.split()  # Разделяем строку на элементы\n",
    "        numeric_elements = sorted(\n",
    "            [el for el in elements if el.isdigit()], key=int\n",
    "        )  # Сортируем числовые элементы\n",
    "        non_numeric_elements = [\n",
    "            el for el in elements if not el.isdigit()\n",
    "        ]  # Не числовые элементы остаются на месте\n",
    "        return \" \".join(\n",
    "            numeric_elements + non_numeric_elements\n",
    "        )  # Объединяем отсортированные числа и нечисловые\n",
    "\n",
    "    # Применяем функцию к каждому элементу колонки\n",
    "    df[column_name] = df[column_name].apply(sort_reference_string)\n",
    "    return df\n",
    "\n",
    "\n",
    "# Применяем сортировку к колонке 'references' в обеих таблицах\n",
    "df_histonescsv = sort_references_column(df_histonescsv)\n",
    "sequence_df = sort_references_column(sequence_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "226db640-274f-4e42-8602-761688af671e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Сортируем строки по всем столбцам\n",
    "df1_sorted = df_histonescsv.sort_values(by=list(df_histonescsv.columns)).reset_index(\n",
    "    drop=True\n",
    ")\n",
    "df2_sorted = sequence_df.sort_values(by=list(sequence_df.columns)).reset_index(\n",
    "    drop=True\n",
    ")\n",
    "\n",
    "# Сравниваем отсортированные DataFrame\n",
    "df1_sorted.equals(df2_sorted)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "c73e02d3-8db9-4ac3-beb3-faf7a340734e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "accession         object\n",
       "type              object\n",
       "variant_group     object\n",
       "variant           object\n",
       "doublet           object\n",
       "gi                object\n",
       "ncbi_gene_id      object\n",
       "hgnc_gene_name    object\n",
       "taxonomy_id        Int64\n",
       "organism          object\n",
       "phylum            object\n",
       "class             object\n",
       "taxonomy_group    object\n",
       "info              object\n",
       "references        object\n",
       "sequence          object\n",
       "dtype: object"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_histonescsv.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "dd2c237a-e338-403c-a3c1-b0ccdf2f3444",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "accession         object\n",
       "type              object\n",
       "variant_group     object\n",
       "variant           object\n",
       "doublet           object\n",
       "gi                object\n",
       "ncbi_gene_id      object\n",
       "hgnc_gene_name    object\n",
       "taxonomy_id        Int64\n",
       "organism          object\n",
       "phylum            object\n",
       "class             object\n",
       "taxonomy_group    object\n",
       "info              object\n",
       "references        object\n",
       "sequence          object\n",
       "dtype: object"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "441f1c46-7db7-4dac-ad4d-8aa0d1d9e4c3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# перед сравнением попробуем удалить записи из таблиц, где все значения type, variant_group и variant равны пустому значению или\n",
    "# содержат символы __???\n",
    "pattern = r\"__\\?\\?\\?\"  # Шаблон для поиска '__???'\n",
    "# df_histonescsv_ = df_histonescsv[\n",
    "#     ~df_histonescsv[\"type\"].str.contains(pattern, regex=True)\n",
    "#     & ~df_histonescsv[\"variant_group\"].str.contains(pattern, regex=True)\n",
    "#     & ~df_histonescsv[\"variant\"].str.contains(pattern, regex=True)\n",
    "# ]\n",
    "\n",
    "# Удаляем '__???' из указанных колонок\n",
    "df_histonescsv_ = df_histonescsv.copy()\n",
    "df_histonescsv_[[\"type\", \"variant_group\", \"variant\"]] = df_histonescsv_[\n",
    "    [\"type\", \"variant_group\", \"variant\"]\n",
    "].replace(r\"__\\?\\?\\?$\", \"\", regex=True)\n",
    "\n",
    "sequence_df_ = sequence_df[\n",
    "    (sequence_df[\"type\"] != \"\")\n",
    "    & (sequence_df[\"variant_group\"] != \"\")\n",
    "    & (sequence_df[\"variant\"] != \"\")\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "3974fe52-60d8-4cb8-b4a2-767962e65c47",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, type, variant_group, variant, doublet, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, references, sequence]\n",
       "Index: []"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df_[sequence_df_[\"variant\"] == \"\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "3a257a79-3d0d-423a-8b5f-36edb4315586",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, type, variant_group, variant, doublet, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, references, sequence]\n",
       "Index: []"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_histonescsv_[df_histonescsv_[\"variant\"].str.contains(pattern)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "e4065f73-89a9-42b3-a049-f3a9b8682dab",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def find_differences(df1, df2):\n",
    "    # Используем merge с индикатором, чтобы отследить источник каждой строки\n",
    "    comparison = df1.merge(df2, how=\"outer\", indicator=True)\n",
    "\n",
    "    # Строки, которые есть только в первой таблице (df1)\n",
    "    only_in_df1 = comparison[comparison[\"_merge\"] == \"left_only\"]\n",
    "\n",
    "    # Строки, которые есть только во второй таблице (df2)\n",
    "    only_in_df2 = comparison[comparison[\"_merge\"] == \"right_only\"]\n",
    "\n",
    "    return only_in_df1, only_in_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "21ba82f2-69de-49f5-9d1a-111b4197bfa6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "only_in_df1, only_in_df2 = find_differences(df_histonescsv_, sequence_df_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "3985e1e9-49d8-47bb-bc03-1d438473eb53",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>XP_005640164.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.1</td>\n",
       "      <td>cH2B_(Mammalia)</td>\n",
       "      <td></td>\n",
       "      <td>545554624</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9615</td>\n",
       "      <td>Canis lupus familiaris</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MPELTSKGTTISKKGFKRAVAKTQKKEGKKRRRCRKESYSIYIYKV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>NP_783594.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.1</td>\n",
       "      <td>cH2B_(Mammalia)</td>\n",
       "      <td></td>\n",
       "      <td>28316750</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MPEVAVKGATISKKGFKKAVTKTQKKEGRKRKRCRKESYSIYIYKV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>NP_072169.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.1</td>\n",
       "      <td>cH2B_(Mammalia)</td>\n",
       "      <td></td>\n",
       "      <td>12025524</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10116</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MPEVSAKGTTISKKGFKKAVTKTQKKEGRKRKRCREESYSIYIYKV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>NP_835502.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>24506885</td>\n",
       "      <td>MPEPAKSAPAPKKGSKKAVTKAQKKDGKKRKRSRKESYSVYVYKVL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>CAA62299.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>8672246</td>\n",
       "      <td>MPEVAVKGATISKKGFKKAVTKTQKKEGRKRKRCRKESYSIYIYKV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>Q64524.3</td>\n",
       "      <td>H2B</td>\n",
       "      <td>cH2B.E_(Mus_musculus)</td>\n",
       "      <td>cH2B.E_(Mus_musculus)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>35099534</td>\n",
       "      <td>MPELAKSAPAPKKGSKKAVTKAQKKDGKKRKRSRKESYSIYVYKVL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>NP_059141.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>cH2B(?)</td>\n",
       "      <td>cH2B(?)_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>54145</td>\n",
       "      <td>H2BC12L</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MPEPAKSAPAPKKGSKKAVTKAQKKDGRKRKRSRKESYSVYVYKVL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>NP_001358848.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.Y</td>\n",
       "      <td>H3.Y.2_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>340096</td>\n",
       "      <td>H3Y2</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>20819935</td>\n",
       "      <td>MARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>HISTDB_H3_Y_0</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.Y</td>\n",
       "      <td>H3.Y</td>\n",
       "      <td></td>\n",
       "      <td>NOGI</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9544</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ARTKQTARKATNWQAPRKPLATKAAAKRAPPRGGIKKPHRYKPGTQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464</th>\n",
       "      <td>HISTDB_H3_Y_1</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.Y</td>\n",
       "      <td>H3.Y</td>\n",
       "      <td></td>\n",
       "      <td>NOGI</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9544</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ARTKQTARKATNWQAPRKPLATKAPGKRLPPRGGIKKPHRYRPGTQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>HISTDB_H3_Y_2</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.Y</td>\n",
       "      <td>H3.Y</td>\n",
       "      <td></td>\n",
       "      <td>NOGI</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9598</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGTL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>466</th>\n",
       "      <td>HISTDB_H3_Y_3</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.Y</td>\n",
       "      <td>H3.Y</td>\n",
       "      <td></td>\n",
       "      <td>NOGI</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9598</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGTL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467</th>\n",
       "      <td>NP_001800.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1058</td>\n",
       "      <td>CENPA</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>23324462</td>\n",
       "      <td>MGPRRRSRKPEAPRRRSPSPTPTPGPSRRGPSLGASSHQHSRRRQG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468</th>\n",
       "      <td>NP_001035891.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1058</td>\n",
       "      <td>CENPA</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>23324462</td>\n",
       "      <td>MGPRRRSRKPEAPRRRSPSPTPTPGPSRRGPSLGASSHQHSRRRQG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>469</th>\n",
       "      <td>NP_001013721.2</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.5</td>\n",
       "      <td>H3.5_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>440093</td>\n",
       "      <td>H3-5</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>21274551</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSTPSTCGVKPHRYRPGTV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>470</th>\n",
       "      <td>NP_563627.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td></td>\n",
       "      <td>18378832</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MARTKHRVTRSQPRNQTDAAGASSSQAAGPTTTPTRRGGEGGDNTQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471</th>\n",
       "      <td>NP_596473.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td></td>\n",
       "      <td>19113265</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4896</td>\n",
       "      <td>Schizosaccharomyces pombe</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Schizosaccharomycetes</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MAKKSLMAEPGDPIPRPRKKRYRPGTTALREIRKYQRSTDLLIQRL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>472</th>\n",
       "      <td>NP_499128.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td></td>\n",
       "      <td>17553736</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>6239</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>Nematoda</td>\n",
       "      <td>Chromadorea</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MADDTPIIEEIAEQNESVTRIMQRLKHDMQRVTSVPGFNTSAAGVN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>473</th>\n",
       "      <td>NP_523730.2</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td></td>\n",
       "      <td>22024004</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MPRHSRAKRAPRPSANNSKSPNDDDTAFRSPEPEDGTDYGLEFTTS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>474</th>\n",
       "      <td>NP_031707.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td></td>\n",
       "      <td>6680920</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MGPRRKPQTPRRRPSSPAPGPSRQSSSVGSQTLRRRQKFMWLKEIK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>475</th>\n",
       "      <td>XP_001350068.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td></td>\n",
       "      <td>124513424</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>36329</td>\n",
       "      <td>Plasmodium falciparum 3D7</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Aconoidasida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MVRTKKNIPNHNPLNAFNRDKSFKTNKTLPNRTVHHGISSKTTNIN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476</th>\n",
       "      <td>AAK39657.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td></td>\n",
       "      <td>13794280</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>55529</td>\n",
       "      <td>Guillardia theta</td>\n",
       "      <td></td>\n",
       "      <td>Cryptophyceae</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MMKKQNLKRFKKSSNSLVDIRKFQKSTDLLIHRLPFARLVKEISLK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>477</th>\n",
       "      <td>XP_009526809.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td></td>\n",
       "      <td>695416328</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>67593</td>\n",
       "      <td>Phytophthora sojae</td>\n",
       "      <td>Oomycota</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MASPRPPALPASDASSSSVTDSGTDDSTPPNSPQPQRSPSPPFAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>478</th>\n",
       "      <td>XP_005535607.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td></td>\n",
       "      <td>544210412</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>280699</td>\n",
       "      <td>Cyanidioschyzon merolae strain 10D</td>\n",
       "      <td>Rhodophyta</td>\n",
       "      <td>Bangiophyceae</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MVRVFGNPVGGAHQYRQTQALNRDESLREQLQRTTVAEQDENIEPE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>479</th>\n",
       "      <td>XP_002287626.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td></td>\n",
       "      <td>223995905</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>296543</td>\n",
       "      <td>Thalassiosira pseudonana CCMP1335</td>\n",
       "      <td>Bacillariophyta</td>\n",
       "      <td>Coscinodiscophyceae</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MRPGEKALREIRQYQSSTSLLLRRLPFARLVREIQYGMTRQPYRWQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480</th>\n",
       "      <td>XP_001011273.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td></td>\n",
       "      <td>118356028</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>312017</td>\n",
       "      <td>Tetrahymena thermophila SB210</td>\n",
       "      <td>Ciliophora</td>\n",
       "      <td>Oligohymenophorea</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MARKAYQPKRRSNSNQNQQRSDSLKKNKQDNLRSKSAGNQQGNEKN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>481</th>\n",
       "      <td>XP_002767160.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td></td>\n",
       "      <td>294874934</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>423536</td>\n",
       "      <td>Perkinsus marinus ATCC 50983</td>\n",
       "      <td>Perkinsozoa</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MVGVENLGVGFDELLTRGGCGVRDDAVEIAFRGVEGLEDVLKDYMV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482</th>\n",
       "      <td>NP_012875.2</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td></td>\n",
       "      <td>27808712</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>559292</td>\n",
       "      <td>Saccharomyces cerevisiae S288C</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Saccharomycetes</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MSSKQQWVSSAIQSDSSGRSLSNVNRLAGDQQSINDRALSLLQRTR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483</th>\n",
       "      <td>XP_003954426.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.5</td>\n",
       "      <td>H3.5</td>\n",
       "      <td></td>\n",
       "      <td>410046862</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9598</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSTPSTXGVKKPHRYRPGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>NP_001342338.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.7(?)</td>\n",
       "      <td>H3.7(?)_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>440686</td>\n",
       "      <td>H3-7</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>12408966</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485</th>\n",
       "      <td>NP_003484.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.4</td>\n",
       "      <td>H3.4_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>8290</td>\n",
       "      <td>H3-4</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>8986613</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYRPGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>NP_001342187.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.Y</td>\n",
       "      <td>H3.Y.1_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>391769</td>\n",
       "      <td>H3Y1</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>20819935</td>\n",
       "      <td>MARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>487</th>\n",
       "      <td>XP_003804825.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>TS_H3.4</td>\n",
       "      <td>TS_H3.4</td>\n",
       "      <td></td>\n",
       "      <td>397466137</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9597</td>\n",
       "      <td>Pan paniscus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLVTKVARKSAPATGGVKKPHRYRPGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>500</th>\n",
       "      <td>NP_003538.1</td>\n",
       "      <td>H4</td>\n",
       "      <td>cH4(?)</td>\n",
       "      <td>cH4(?)_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>8369</td>\n",
       "      <td>H4C7</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>12408966</td>\n",
       "      <td>MSVRGKAGKGLGKGGAKCHRKVLSDNIQGITKCTIRRLARHGGVKR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562</th>\n",
       "      <td>NP_722575.1</td>\n",
       "      <td>H1</td>\n",
       "      <td>H1.8</td>\n",
       "      <td>H1.8_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>132243</td>\n",
       "      <td>H1-8</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "      <td>MAPGSVTSDISPSSTSTAGSSRSPESEKPGPSHGGVPPGGPSHSSL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>563</th>\n",
       "      <td>NP_001295191.1</td>\n",
       "      <td>H1</td>\n",
       "      <td>H1.8</td>\n",
       "      <td>H1.8_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>132243</td>\n",
       "      <td>H1-8</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "      <td>MAPATAPRRAGEAKGKGPKKPSEAKEDPPNVGKVKKAAKRPAKVQK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>NP_861453.1</td>\n",
       "      <td>H1</td>\n",
       "      <td>H1.7</td>\n",
       "      <td>H1.7_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>341567</td>\n",
       "      <td>H1-7</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "      <td>MEQALTGEAQSRWPRRGGSGAMAEAPGPSGESRGHSATQLPAEKTV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>570</th>\n",
       "      <td>NP_005314.2</td>\n",
       "      <td>H1</td>\n",
       "      <td>H1.6</td>\n",
       "      <td>H1.6_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3010</td>\n",
       "      <td>H1-6</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "      <td>MSETVPAASASAGVAAMEKLPTKKRGRKPAGLISASRKVPNLSVSK...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession type          variant_group                 variant  \\\n",
       "237  XP_005640164.1  H2B                  H2B.1         cH2B_(Mammalia)   \n",
       "238     NP_783594.1  H2B                  H2B.1         cH2B_(Mammalia)   \n",
       "239     NP_072169.1  H2B                  H2B.1         cH2B_(Mammalia)   \n",
       "261     NP_835502.1  H2B  cH2B.1_(Mus_musculus)   cH2B.1_(Mus_musculus)   \n",
       "262      CAA62299.1  H2B  cH2B.1_(Mus_musculus)   cH2B.1_(Mus_musculus)   \n",
       "263        Q64524.3  H2B  cH2B.E_(Mus_musculus)   cH2B.E_(Mus_musculus)   \n",
       "405     NP_059141.1  H2B                cH2B(?)  cH2B(?)_(Homo_sapiens)   \n",
       "462  NP_001358848.1   H3                   H3.Y   H3.Y.2_(Homo_sapiens)   \n",
       "463   HISTDB_H3_Y_0   H3                   H3.Y                    H3.Y   \n",
       "464   HISTDB_H3_Y_1   H3                   H3.Y                    H3.Y   \n",
       "465   HISTDB_H3_Y_2   H3                   H3.Y                    H3.Y   \n",
       "466   HISTDB_H3_Y_3   H3                   H3.Y                    H3.Y   \n",
       "467     NP_001800.1   H3                  cenH3    cenH3_(Homo_sapiens)   \n",
       "468  NP_001035891.1   H3                  cenH3    cenH3_(Homo_sapiens)   \n",
       "469  NP_001013721.2   H3                   H3.5     H3.5_(Homo_sapiens)   \n",
       "470     NP_563627.1   H3                  cenH3                   cenH3   \n",
       "471     NP_596473.1   H3                  cenH3                   cenH3   \n",
       "472     NP_499128.1   H3                  cenH3                   cenH3   \n",
       "473     NP_523730.2   H3                  cenH3                   cenH3   \n",
       "474     NP_031707.1   H3                  cenH3                   cenH3   \n",
       "475  XP_001350068.1   H3                  cenH3                   cenH3   \n",
       "476      AAK39657.1   H3                  cenH3                   cenH3   \n",
       "477  XP_009526809.1   H3                  cenH3                   cenH3   \n",
       "478  XP_005535607.1   H3                  cenH3                   cenH3   \n",
       "479  XP_002287626.1   H3                  cenH3                   cenH3   \n",
       "480  XP_001011273.1   H3                  cenH3                   cenH3   \n",
       "481  XP_002767160.1   H3                  cenH3                   cenH3   \n",
       "482     NP_012875.2   H3                  cenH3                   cenH3   \n",
       "483  XP_003954426.1   H3                   H3.5                    H3.5   \n",
       "484  NP_001342338.1   H3                H3.7(?)  H3.7(?)_(Homo_sapiens)   \n",
       "485     NP_003484.1   H3                   H3.4     H3.4_(Homo_sapiens)   \n",
       "486  NP_001342187.1   H3                   H3.Y   H3.Y.1_(Homo_sapiens)   \n",
       "487  XP_003804825.1   H3                TS_H3.4                 TS_H3.4   \n",
       "500     NP_003538.1   H4                 cH4(?)   cH4(?)_(Homo_sapiens)   \n",
       "562     NP_722575.1   H1                   H1.8     H1.8_(Homo_sapiens)   \n",
       "563  NP_001295191.1   H1                   H1.8     H1.8_(Homo_sapiens)   \n",
       "566     NP_861453.1   H1                   H1.7     H1.7_(Homo_sapiens)   \n",
       "570     NP_005314.2   H1                   H1.6     H1.6_(Homo_sapiens)   \n",
       "\n",
       "    doublet         gi ncbi_gene_id hgnc_gene_name  taxonomy_id  \\\n",
       "237          545554624                                     9615   \n",
       "238           28316750                                    10090   \n",
       "239           12025524                                    10116   \n",
       "261                                                       10090   \n",
       "262                                                       10090   \n",
       "263                                                       10090   \n",
       "405                           54145        H2BC12L         9606   \n",
       "462                          340096           H3Y2         9606   \n",
       "463               NOGI                                     9544   \n",
       "464               NOGI                                     9544   \n",
       "465               NOGI                                     9598   \n",
       "466               NOGI                                     9598   \n",
       "467                            1058          CENPA         9606   \n",
       "468                            1058          CENPA         9606   \n",
       "469                          440093           H3-5         9606   \n",
       "470           18378832                                     3702   \n",
       "471           19113265                                     4896   \n",
       "472           17553736                                     6239   \n",
       "473           22024004                                     7227   \n",
       "474            6680920                                    10090   \n",
       "475          124513424                                    36329   \n",
       "476           13794280                                    55529   \n",
       "477          695416328                                    67593   \n",
       "478          544210412                                   280699   \n",
       "479          223995905                                   296543   \n",
       "480          118356028                                   312017   \n",
       "481          294874934                                   423536   \n",
       "482           27808712                                   559292   \n",
       "483          410046862                                     9598   \n",
       "484                          440686           H3-7         9606   \n",
       "485                            8290           H3-4         9606   \n",
       "486                          391769           H3Y1         9606   \n",
       "487          397466137                                     9597   \n",
       "500                            8369           H4C7         9606   \n",
       "562                          132243           H1-8         9606   \n",
       "563                          132243           H1-8         9606   \n",
       "566                          341567           H1-7         9606   \n",
       "570                            3010           H1-6         9606   \n",
       "\n",
       "                               organism           phylum  \\\n",
       "237              Canis lupus familiaris         Chordata   \n",
       "238                        Mus musculus         Chordata   \n",
       "239                   Rattus norvegicus         Chordata   \n",
       "261                        Mus musculus         Chordata   \n",
       "262                        Mus musculus         Chordata   \n",
       "263                        Mus musculus         Chordata   \n",
       "405                        Homo sapiens         Chordata   \n",
       "462                        Homo sapiens         Chordata   \n",
       "463                      Macaca mulatta         Chordata   \n",
       "464                      Macaca mulatta         Chordata   \n",
       "465                     Pan troglodytes         Chordata   \n",
       "466                     Pan troglodytes         Chordata   \n",
       "467                        Homo sapiens         Chordata   \n",
       "468                        Homo sapiens         Chordata   \n",
       "469                        Homo sapiens         Chordata   \n",
       "470                Arabidopsis thaliana     Streptophyta   \n",
       "471           Schizosaccharomyces pombe       Ascomycota   \n",
       "472              Caenorhabditis elegans         Nematoda   \n",
       "473             Drosophila melanogaster       Arthropoda   \n",
       "474                        Mus musculus         Chordata   \n",
       "475           Plasmodium falciparum 3D7      Apicomplexa   \n",
       "476                    Guillardia theta                    \n",
       "477                  Phytophthora sojae         Oomycota   \n",
       "478  Cyanidioschyzon merolae strain 10D       Rhodophyta   \n",
       "479   Thalassiosira pseudonana CCMP1335  Bacillariophyta   \n",
       "480       Tetrahymena thermophila SB210       Ciliophora   \n",
       "481        Perkinsus marinus ATCC 50983      Perkinsozoa   \n",
       "482      Saccharomyces cerevisiae S288C       Ascomycota   \n",
       "483                     Pan troglodytes         Chordata   \n",
       "484                        Homo sapiens         Chordata   \n",
       "485                        Homo sapiens         Chordata   \n",
       "486                        Homo sapiens         Chordata   \n",
       "487                        Pan paniscus         Chordata   \n",
       "500                        Homo sapiens         Chordata   \n",
       "562                        Homo sapiens         Chordata   \n",
       "563                        Homo sapiens         Chordata   \n",
       "566                        Homo sapiens         Chordata   \n",
       "570                        Homo sapiens         Chordata   \n",
       "\n",
       "                     class taxonomy_group info references  \\\n",
       "237               Mammalia                                  \n",
       "238               Mammalia                                  \n",
       "239               Mammalia                                  \n",
       "261               Mammalia                       24506885   \n",
       "262               Mammalia                        8672246   \n",
       "263               Mammalia                       35099534   \n",
       "405               Mammalia       Mammalia                   \n",
       "462               Mammalia       Mammalia        20819935   \n",
       "463               Mammalia                                  \n",
       "464               Mammalia                                  \n",
       "465               Mammalia                                  \n",
       "466               Mammalia                                  \n",
       "467               Mammalia       Mammalia        23324462   \n",
       "468               Mammalia       Mammalia        23324462   \n",
       "469               Mammalia       Mammalia        21274551   \n",
       "470          Magnoliopsida                                  \n",
       "471  Schizosaccharomycetes                                  \n",
       "472            Chromadorea                                  \n",
       "473                Insecta                                  \n",
       "474               Mammalia                                  \n",
       "475           Aconoidasida                                  \n",
       "476          Cryptophyceae                                  \n",
       "477                                                         \n",
       "478          Bangiophyceae                                  \n",
       "479    Coscinodiscophyceae                                  \n",
       "480      Oligohymenophorea                                  \n",
       "481                                                         \n",
       "482        Saccharomycetes                                  \n",
       "483               Mammalia                                  \n",
       "484               Mammalia       Mammalia        12408966   \n",
       "485               Mammalia       Mammalia         8986613   \n",
       "486               Mammalia       Mammalia        20819935   \n",
       "487               Mammalia                                  \n",
       "500               Mammalia       Mammalia        12408966   \n",
       "562               Mammalia       Mammalia        26689747   \n",
       "563               Mammalia       Mammalia        26689747   \n",
       "566               Mammalia       Mammalia        26689747   \n",
       "570               Mammalia       Mammalia        26689747   \n",
       "\n",
       "                                              sequence  \n",
       "237  MPELTSKGTTISKKGFKRAVAKTQKKEGKKRRRCRKESYSIYIYKV...  \n",
       "238  MPEVAVKGATISKKGFKKAVTKTQKKEGRKRKRCRKESYSIYIYKV...  \n",
       "239  MPEVSAKGTTISKKGFKKAVTKTQKKEGRKRKRCREESYSIYIYKV...  \n",
       "261  MPEPAKSAPAPKKGSKKAVTKAQKKDGKKRKRSRKESYSVYVYKVL...  \n",
       "262  MPEVAVKGATISKKGFKKAVTKTQKKEGRKRKRCRKESYSIYIYKV...  \n",
       "263  MPELAKSAPAPKKGSKKAVTKAQKKDGKKRKRSRKESYSIYVYKVL...  \n",
       "405  MPEPAKSAPAPKKGSKKAVTKAQKKDGRKRKRSRKESYSVYVYKVL...  \n",
       "462  MARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGT...  \n",
       "463  ARTKQTARKATNWQAPRKPLATKAAAKRAPPRGGIKKPHRYKPGTQ...  \n",
       "464  ARTKQTARKATNWQAPRKPLATKAPGKRLPPRGGIKKPHRYRPGTQ...  \n",
       "465  ARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGTL...  \n",
       "466  ARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGTL...  \n",
       "467  MGPRRRSRKPEAPRRRSPSPTPTPGPSRRGPSLGASSHQHSRRRQG...  \n",
       "468  MGPRRRSRKPEAPRRRSPSPTPTPGPSRRGPSLGASSHQHSRRRQG...  \n",
       "469  MARTKQTARKSTGGKAPRKQLATKAARKSTPSTCGVKPHRYRPGTV...  \n",
       "470  MARTKHRVTRSQPRNQTDAAGASSSQAAGPTTTPTRRGGEGGDNTQ...  \n",
       "471  MAKKSLMAEPGDPIPRPRKKRYRPGTTALREIRKYQRSTDLLIQRL...  \n",
       "472  MADDTPIIEEIAEQNESVTRIMQRLKHDMQRVTSVPGFNTSAAGVN...  \n",
       "473  MPRHSRAKRAPRPSANNSKSPNDDDTAFRSPEPEDGTDYGLEFTTS...  \n",
       "474  MGPRRKPQTPRRRPSSPAPGPSRQSSSVGSQTLRRRQKFMWLKEIK...  \n",
       "475  MVRTKKNIPNHNPLNAFNRDKSFKTNKTLPNRTVHHGISSKTTNIN...  \n",
       "476  MMKKQNLKRFKKSSNSLVDIRKFQKSTDLLIHRLPFARLVKEISLK...  \n",
       "477  MASPRPPALPASDASSSSVTDSGTDDSTPPNSPQPQRSPSPPFAAT...  \n",
       "478  MVRVFGNPVGGAHQYRQTQALNRDESLREQLQRTTVAEQDENIEPE...  \n",
       "479  MRPGEKALREIRQYQSSTSLLLRRLPFARLVREIQYGMTRQPYRWQ...  \n",
       "480  MARKAYQPKRRSNSNQNQQRSDSLKKNKQDNLRSKSAGNQQGNEKN...  \n",
       "481  MVGVENLGVGFDELLTRGGCGVRDDAVEIAFRGVEGLEDVLKDYMV...  \n",
       "482  MSSKQQWVSSAIQSDSSGRSLSNVNRLAGDQQSINDRALSLLQRTR...  \n",
       "483  MARTKQTARKSTGGKAPRKQLATKAARKSTPSTXGVKKPHRYRPGT...  \n",
       "484  MARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGT...  \n",
       "485  MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYRPGT...  \n",
       "486  MARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGT...  \n",
       "487  MARTKQTARKSTGGKAPRKQLVTKVARKSAPATGGVKKPHRYRPGT...  \n",
       "500  MSVRGKAGKGLGKGGAKCHRKVLSDNIQGITKCTIRRLARHGGVKR...  \n",
       "562  MAPGSVTSDISPSSTSTAGSSRSPESEKPGPSHGGVPPGGPSHSSL...  \n",
       "563  MAPATAPRRAGEAKGKGPKKPSEAKEDPPNVGKVKKAAKRPAKVQK...  \n",
       "566  MEQALTGEAQSRWPRRGGSGAMAEAPGPSGESRGHSATQLPAEKTV...  \n",
       "570  MSETVPAASASAGVAAMEKLPTKKRGRKPAGLISASRKVPNLSVSK...  "
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Записи, которые есть только в первой таблице (df_histonescsv):\n",
    "only_in_df1_ = only_in_df1.drop(columns=[\"_merge\"])\n",
    "only_in_df1_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "784e1a08-1ff0-45ee-ab88-95fd94aedfb4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>572</th>\n",
       "      <td>AAK39657.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td></td>\n",
       "      <td>13794280</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>55529</td>\n",
       "      <td>Guillardia theta</td>\n",
       "      <td></td>\n",
       "      <td>Cryptophyceae</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MMKKQNLKRFKKSSNSLVDIRKFQKSTDLLIHRLPFARLVKEISLK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>573</th>\n",
       "      <td>CAA62299.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>cH2B</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>8672246</td>\n",
       "      <td>MPEVAVKGATISKKGFKKAVTKTQKKEGRKRKRCRKESYSIYIYKV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>574</th>\n",
       "      <td>HISTDB_H3_Y_0</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.3-like?</td>\n",
       "      <td>H3.Y_(Primates?)</td>\n",
       "      <td></td>\n",
       "      <td>NOGI</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9544</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ARTKQTARKATNWQAPRKPLATKAAAKRAPPRGGIKKPHRYKPGTQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>575</th>\n",
       "      <td>HISTDB_H3_Y_1</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.3-like?</td>\n",
       "      <td>H3.Y_(Primates?)</td>\n",
       "      <td></td>\n",
       "      <td>NOGI</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9544</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ARTKQTARKATNWQAPRKPLATKAPGKRLPPRGGIKKPHRYRPGTQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>576</th>\n",
       "      <td>HISTDB_H3_Y_2</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.3-like?</td>\n",
       "      <td>H3.Y_(Primates?)</td>\n",
       "      <td></td>\n",
       "      <td>NOGI</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9598</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGTL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>577</th>\n",
       "      <td>HISTDB_H3_Y_3</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.3-like?</td>\n",
       "      <td>H3.Y_(Primates?)</td>\n",
       "      <td></td>\n",
       "      <td>NOGI</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9598</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGTL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>578</th>\n",
       "      <td>NP_001013721.2</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.5_(Primates_or_Hominids?)</td>\n",
       "      <td>H3.5_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>440093</td>\n",
       "      <td>H3-5</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>21274551</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSTPSTCGVKPHRYRPGTV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>579</th>\n",
       "      <td>NP_001035891.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1058</td>\n",
       "      <td>CENPA</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>23324462</td>\n",
       "      <td>MGPRRRSRKPEAPRRRSPSPTPTPGPSRRGPSLGASSHQHSRRRQG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>NP_001295191.1</td>\n",
       "      <td>H1</td>\n",
       "      <td>OO_H1.8</td>\n",
       "      <td>H1.8_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>132243</td>\n",
       "      <td>H1-8</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "      <td>MAPATAPRRAGEAKGKGPKKPSEAKEDPPNVGKVKKAAKRPAKVQK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>581</th>\n",
       "      <td>NP_001342187.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.3-like?</td>\n",
       "      <td>H3.Y.1_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>391769</td>\n",
       "      <td>H3Y1</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>20819935</td>\n",
       "      <td>MARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>582</th>\n",
       "      <td>NP_001358848.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.3-like?</td>\n",
       "      <td>H3.Y.2_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>340096</td>\n",
       "      <td>H3Y2</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>20819935</td>\n",
       "      <td>MARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>583</th>\n",
       "      <td>NP_001800.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1058</td>\n",
       "      <td>CENPA</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>23324462</td>\n",
       "      <td>MGPRRRSRKPEAPRRRSPSPTPTPGPSRRGPSLGASSHQHSRRRQG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584</th>\n",
       "      <td>NP_003484.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>H3.4_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>8290</td>\n",
       "      <td>H3-4</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>8986613</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYRPGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>585</th>\n",
       "      <td>NP_005314.2</td>\n",
       "      <td>H1</td>\n",
       "      <td>TS_H1.6</td>\n",
       "      <td>H1.6_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3010</td>\n",
       "      <td>H1-6</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "      <td>MSETVPAASASAGVAAMEKLPTKKRGRKPAGLISASRKVPNLSVSK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>586</th>\n",
       "      <td>NP_012875.2</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3_(Fungi)</td>\n",
       "      <td></td>\n",
       "      <td>27808712</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>559292</td>\n",
       "      <td>Saccharomyces cerevisiae S288C</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Saccharomycetes</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MSSKQQWVSSAIQSDSSGRSLSNVNRLAGDQQSINDRALSLLQRTR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>587</th>\n",
       "      <td>NP_031707.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3_(Mammalia)</td>\n",
       "      <td></td>\n",
       "      <td>6680920</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MGPRRKPQTPRRRPSSPAPGPSRQSSSVGSQTLRRRQKFMWLKEIK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>588</th>\n",
       "      <td>NP_072169.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>cH2B</td>\n",
       "      <td>cH2B_(Mammalia)</td>\n",
       "      <td></td>\n",
       "      <td>12025524</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10116</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MPEVSAKGTTISKKGFKKAVTKTQKKEGRKRKRCREESYSIYIYKV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>589</th>\n",
       "      <td>NP_499128.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3_(Animals)</td>\n",
       "      <td></td>\n",
       "      <td>17553736</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>6239</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>Nematoda</td>\n",
       "      <td>Chromadorea</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MADDTPIIEEIAEQNESVTRIMQRLKHDMQRVTSVPGFNTSAAGVN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>590</th>\n",
       "      <td>NP_523730.2</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3_(Animals)</td>\n",
       "      <td></td>\n",
       "      <td>22024004</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MPRHSRAKRAPRPSANNSKSPNDDDTAFRSPEPEDGTDYGLEFTTS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>591</th>\n",
       "      <td>NP_563627.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td></td>\n",
       "      <td>18378832</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MARTKHRVTRSQPRNQTDAAGASSSQAAGPTTTPTRRGGEGGDNTQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592</th>\n",
       "      <td>NP_596473.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3_(Fungi)</td>\n",
       "      <td></td>\n",
       "      <td>19113265</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4896</td>\n",
       "      <td>Schizosaccharomyces pombe</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Schizosaccharomycetes</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MAKKSLMAEPGDPIPRPRKKRYRPGTTALREIRKYQRSTDLLIQRL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593</th>\n",
       "      <td>NP_722575.1</td>\n",
       "      <td>H1</td>\n",
       "      <td>OO_H1.8</td>\n",
       "      <td>H1.8_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>132243</td>\n",
       "      <td>H1-8</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "      <td>MAPGSVTSDISPSSTSTAGSSRSPESEKPGPSHGGVPPGGPSHSSL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>594</th>\n",
       "      <td>NP_783594.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>cH2B</td>\n",
       "      <td>cH2B_(Mammalia)</td>\n",
       "      <td></td>\n",
       "      <td>28316750</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MPEVAVKGATISKKGFKKAVTKTQKKEGRKRKRCRKESYSIYIYKV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595</th>\n",
       "      <td>NP_835502.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>cH2B</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>24506885</td>\n",
       "      <td>MPEPAKSAPAPKKGSKKAVTKAQKKDGKKRKRSRKESYSVYVYKVL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>596</th>\n",
       "      <td>NP_861453.1</td>\n",
       "      <td>H1</td>\n",
       "      <td>TS_H1.7</td>\n",
       "      <td>H1.7_(Homo_sapiens)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>341567</td>\n",
       "      <td>H1-7</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "      <td>MEQALTGEAQSRWPRRGGSGAMAEAPGPSGESRGHSATQLPAEKTV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>597</th>\n",
       "      <td>Q64524.3</td>\n",
       "      <td>H2B</td>\n",
       "      <td>cH2B</td>\n",
       "      <td>cH2B.E_(Mus_musculus)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>35099534</td>\n",
       "      <td>MPELAKSAPAPKKGSKKAVTKAQKKDGKKRKRSRKESYSIYVYKVL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>598</th>\n",
       "      <td>XP_001011273.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td></td>\n",
       "      <td>118356028</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>312017</td>\n",
       "      <td>Tetrahymena thermophila SB210</td>\n",
       "      <td>Ciliophora</td>\n",
       "      <td>Oligohymenophorea</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MARKAYQPKRRSNSNQNQQRSDSLKKNKQDNLRSKSAGNQQGNEKN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>599</th>\n",
       "      <td>XP_001350068.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td></td>\n",
       "      <td>124513424</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>36329</td>\n",
       "      <td>Plasmodium falciparum 3D7</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Aconoidasida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MVRTKKNIPNHNPLNAFNRDKSFKTNKTLPNRTVHHGISSKTTNIN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>600</th>\n",
       "      <td>XP_002287626.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td></td>\n",
       "      <td>223995905</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>296543</td>\n",
       "      <td>Thalassiosira pseudonana CCMP1335</td>\n",
       "      <td>Bacillariophyta</td>\n",
       "      <td>Coscinodiscophyceae</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MRPGEKALREIRQYQSSTSLLLRRLPFARLVREIQYGMTRQPYRWQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>601</th>\n",
       "      <td>XP_002767160.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td></td>\n",
       "      <td>294874934</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>423536</td>\n",
       "      <td>Perkinsus marinus ATCC 50983</td>\n",
       "      <td>Perkinsozoa</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MVGVENLGVGFDELLTRGGCGVRDDAVEIAFRGVEGLEDVLKDYMV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>602</th>\n",
       "      <td>XP_003804825.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td></td>\n",
       "      <td>397466137</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9597</td>\n",
       "      <td>Pan paniscus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLVTKVARKSAPATGGVKKPHRYRPGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>603</th>\n",
       "      <td>XP_003954426.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>H3.5_(Primates_or_Hominids?)</td>\n",
       "      <td>H3.5_(Primates_or_Hominids?)</td>\n",
       "      <td></td>\n",
       "      <td>410046862</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9598</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSTPSTXGVKKPHRYRPGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604</th>\n",
       "      <td>XP_005535607.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td></td>\n",
       "      <td>544210412</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>280699</td>\n",
       "      <td>Cyanidioschyzon merolae strain 10D</td>\n",
       "      <td>Rhodophyta</td>\n",
       "      <td>Bangiophyceae</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MVRVFGNPVGGAHQYRQTQALNRDESLREQLQRTTVAEQDENIEPE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>605</th>\n",
       "      <td>XP_005640164.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>cH2B</td>\n",
       "      <td>cH2B_(Mammalia)</td>\n",
       "      <td></td>\n",
       "      <td>545554624</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9615</td>\n",
       "      <td>Canis lupus familiaris</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MPELTSKGTTISKKGFKRAVAKTQKKEGKKRRRCRKESYSIYIYKV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>606</th>\n",
       "      <td>XP_009526809.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td>cenH3_(Eukarya)</td>\n",
       "      <td></td>\n",
       "      <td>695416328</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>67593</td>\n",
       "      <td>Phytophthora sojae</td>\n",
       "      <td>Oomycota</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MASPRPPALPASDASSSSVTDSGTDDSTPPNSPQPQRSPSPPFAAT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession type                 variant_group  \\\n",
       "572      AAK39657.1   H3               cenH3_(Eukarya)   \n",
       "573      CAA62299.1  H2B                          cH2B   \n",
       "574   HISTDB_H3_Y_0   H3                    H3.3-like?   \n",
       "575   HISTDB_H3_Y_1   H3                    H3.3-like?   \n",
       "576   HISTDB_H3_Y_2   H3                    H3.3-like?   \n",
       "577   HISTDB_H3_Y_3   H3                    H3.3-like?   \n",
       "578  NP_001013721.2   H3  H3.5_(Primates_or_Hominids?)   \n",
       "579  NP_001035891.1   H3               cenH3_(Eukarya)   \n",
       "580  NP_001295191.1   H1                       OO_H1.8   \n",
       "581  NP_001342187.1   H3                    H3.3-like?   \n",
       "582  NP_001358848.1   H3                    H3.3-like?   \n",
       "583     NP_001800.1   H3               cenH3_(Eukarya)   \n",
       "584     NP_003484.1   H3               H3.4_(Mammalia)   \n",
       "585     NP_005314.2   H1                       TS_H1.6   \n",
       "586     NP_012875.2   H3               cenH3_(Eukarya)   \n",
       "587     NP_031707.1   H3               cenH3_(Eukarya)   \n",
       "588     NP_072169.1  H2B                          cH2B   \n",
       "589     NP_499128.1   H3               cenH3_(Eukarya)   \n",
       "590     NP_523730.2   H3               cenH3_(Eukarya)   \n",
       "591     NP_563627.1   H3               cenH3_(Eukarya)   \n",
       "592     NP_596473.1   H3               cenH3_(Eukarya)   \n",
       "593     NP_722575.1   H1                       OO_H1.8   \n",
       "594     NP_783594.1  H2B                          cH2B   \n",
       "595     NP_835502.1  H2B                          cH2B   \n",
       "596     NP_861453.1   H1                       TS_H1.7   \n",
       "597        Q64524.3  H2B                          cH2B   \n",
       "598  XP_001011273.1   H3               cenH3_(Eukarya)   \n",
       "599  XP_001350068.1   H3               cenH3_(Eukarya)   \n",
       "600  XP_002287626.1   H3               cenH3_(Eukarya)   \n",
       "601  XP_002767160.1   H3               cenH3_(Eukarya)   \n",
       "602  XP_003804825.1   H3               H3.4_(Mammalia)   \n",
       "603  XP_003954426.1   H3  H3.5_(Primates_or_Hominids?)   \n",
       "604  XP_005535607.1   H3               cenH3_(Eukarya)   \n",
       "605  XP_005640164.1  H2B                          cH2B   \n",
       "606  XP_009526809.1   H3               cenH3_(Eukarya)   \n",
       "\n",
       "                          variant doublet         gi ncbi_gene_id  \\\n",
       "572                cenH3_(Plants)           13794280                \n",
       "573         cH2B.1_(Mus_musculus)                                   \n",
       "574              H3.Y_(Primates?)               NOGI                \n",
       "575              H3.Y_(Primates?)               NOGI                \n",
       "576              H3.Y_(Primates?)               NOGI                \n",
       "577              H3.Y_(Primates?)               NOGI                \n",
       "578           H3.5_(Homo_sapiens)                          440093   \n",
       "579          cenH3_(Homo_sapiens)                            1058   \n",
       "580           H1.8_(Homo_sapiens)                          132243   \n",
       "581         H3.Y.1_(Homo_sapiens)                          391769   \n",
       "582         H3.Y.2_(Homo_sapiens)                          340096   \n",
       "583          cenH3_(Homo_sapiens)                            1058   \n",
       "584           H3.4_(Homo_sapiens)                            8290   \n",
       "585           H1.6_(Homo_sapiens)                            3010   \n",
       "586                 cenH3_(Fungi)           27808712                \n",
       "587              cenH3_(Mammalia)            6680920                \n",
       "588               cH2B_(Mammalia)           12025524                \n",
       "589               cenH3_(Animals)           17553736                \n",
       "590               cenH3_(Animals)           22024004                \n",
       "591                cenH3_(Plants)           18378832                \n",
       "592                 cenH3_(Fungi)           19113265                \n",
       "593           H1.8_(Homo_sapiens)                          132243   \n",
       "594               cH2B_(Mammalia)           28316750                \n",
       "595         cH2B.1_(Mus_musculus)                                   \n",
       "596           H1.7_(Homo_sapiens)                          341567   \n",
       "597         cH2B.E_(Mus_musculus)                                   \n",
       "598               cenH3_(Eukarya)          118356028                \n",
       "599               cenH3_(Eukarya)          124513424                \n",
       "600               cenH3_(Eukarya)          223995905                \n",
       "601               cenH3_(Eukarya)          294874934                \n",
       "602               H3.4_(Mammalia)          397466137                \n",
       "603  H3.5_(Primates_or_Hominids?)          410046862                \n",
       "604                cenH3_(Plants)          544210412                \n",
       "605               cH2B_(Mammalia)          545554624                \n",
       "606               cenH3_(Eukarya)          695416328                \n",
       "\n",
       "    hgnc_gene_name  taxonomy_id                            organism  \\\n",
       "572                       55529                    Guillardia theta   \n",
       "573                       10090                        Mus musculus   \n",
       "574                        9544                      Macaca mulatta   \n",
       "575                        9544                      Macaca mulatta   \n",
       "576                        9598                     Pan troglodytes   \n",
       "577                        9598                     Pan troglodytes   \n",
       "578           H3-5         9606                        Homo sapiens   \n",
       "579          CENPA         9606                        Homo sapiens   \n",
       "580           H1-8         9606                        Homo sapiens   \n",
       "581           H3Y1         9606                        Homo sapiens   \n",
       "582           H3Y2         9606                        Homo sapiens   \n",
       "583          CENPA         9606                        Homo sapiens   \n",
       "584           H3-4         9606                        Homo sapiens   \n",
       "585           H1-6         9606                        Homo sapiens   \n",
       "586                      559292      Saccharomyces cerevisiae S288C   \n",
       "587                       10090                        Mus musculus   \n",
       "588                       10116                   Rattus norvegicus   \n",
       "589                        6239              Caenorhabditis elegans   \n",
       "590                        7227             Drosophila melanogaster   \n",
       "591                        3702                Arabidopsis thaliana   \n",
       "592                        4896           Schizosaccharomyces pombe   \n",
       "593           H1-8         9606                        Homo sapiens   \n",
       "594                       10090                        Mus musculus   \n",
       "595                       10090                        Mus musculus   \n",
       "596           H1-7         9606                        Homo sapiens   \n",
       "597                       10090                        Mus musculus   \n",
       "598                      312017       Tetrahymena thermophila SB210   \n",
       "599                       36329           Plasmodium falciparum 3D7   \n",
       "600                      296543   Thalassiosira pseudonana CCMP1335   \n",
       "601                      423536        Perkinsus marinus ATCC 50983   \n",
       "602                        9597                        Pan paniscus   \n",
       "603                        9598                     Pan troglodytes   \n",
       "604                      280699  Cyanidioschyzon merolae strain 10D   \n",
       "605                        9615              Canis lupus familiaris   \n",
       "606                       67593                  Phytophthora sojae   \n",
       "\n",
       "              phylum                  class taxonomy_group info references  \\\n",
       "572                           Cryptophyceae                                  \n",
       "573         Chordata               Mammalia                        8672246   \n",
       "574         Chordata               Mammalia                                  \n",
       "575         Chordata               Mammalia                                  \n",
       "576         Chordata               Mammalia                                  \n",
       "577         Chordata               Mammalia                                  \n",
       "578         Chordata               Mammalia       Mammalia        21274551   \n",
       "579         Chordata               Mammalia       Mammalia        23324462   \n",
       "580         Chordata               Mammalia       Mammalia        26689747   \n",
       "581         Chordata               Mammalia       Mammalia        20819935   \n",
       "582         Chordata               Mammalia       Mammalia        20819935   \n",
       "583         Chordata               Mammalia       Mammalia        23324462   \n",
       "584         Chordata               Mammalia       Mammalia         8986613   \n",
       "585         Chordata               Mammalia       Mammalia        26689747   \n",
       "586       Ascomycota        Saccharomycetes                                  \n",
       "587         Chordata               Mammalia                                  \n",
       "588         Chordata               Mammalia                                  \n",
       "589         Nematoda            Chromadorea                                  \n",
       "590       Arthropoda                Insecta                                  \n",
       "591     Streptophyta          Magnoliopsida                                  \n",
       "592       Ascomycota  Schizosaccharomycetes                                  \n",
       "593         Chordata               Mammalia       Mammalia        26689747   \n",
       "594         Chordata               Mammalia                                  \n",
       "595         Chordata               Mammalia                       24506885   \n",
       "596         Chordata               Mammalia       Mammalia        26689747   \n",
       "597         Chordata               Mammalia                       35099534   \n",
       "598       Ciliophora      Oligohymenophorea                                  \n",
       "599      Apicomplexa           Aconoidasida                                  \n",
       "600  Bacillariophyta    Coscinodiscophyceae                                  \n",
       "601      Perkinsozoa                                                         \n",
       "602         Chordata               Mammalia                                  \n",
       "603         Chordata               Mammalia                                  \n",
       "604       Rhodophyta          Bangiophyceae                                  \n",
       "605         Chordata               Mammalia                                  \n",
       "606         Oomycota                                                         \n",
       "\n",
       "                                              sequence  \n",
       "572  MMKKQNLKRFKKSSNSLVDIRKFQKSTDLLIHRLPFARLVKEISLK...  \n",
       "573  MPEVAVKGATISKKGFKKAVTKTQKKEGRKRKRCRKESYSIYIYKV...  \n",
       "574  ARTKQTARKATNWQAPRKPLATKAAAKRAPPRGGIKKPHRYKPGTQ...  \n",
       "575  ARTKQTARKATNWQAPRKPLATKAPGKRLPPRGGIKKPHRYRPGTQ...  \n",
       "576  ARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGTL...  \n",
       "577  ARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGTL...  \n",
       "578  MARTKQTARKSTGGKAPRKQLATKAARKSTPSTCGVKPHRYRPGTV...  \n",
       "579  MGPRRRSRKPEAPRRRSPSPTPTPGPSRRGPSLGASSHQHSRRRQG...  \n",
       "580  MAPATAPRRAGEAKGKGPKKPSEAKEDPPNVGKVKKAAKRPAKVQK...  \n",
       "581  MARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGT...  \n",
       "582  MARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGT...  \n",
       "583  MGPRRRSRKPEAPRRRSPSPTPTPGPSRRGPSLGASSHQHSRRRQG...  \n",
       "584  MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYRPGT...  \n",
       "585  MSETVPAASASAGVAAMEKLPTKKRGRKPAGLISASRKVPNLSVSK...  \n",
       "586  MSSKQQWVSSAIQSDSSGRSLSNVNRLAGDQQSINDRALSLLQRTR...  \n",
       "587  MGPRRKPQTPRRRPSSPAPGPSRQSSSVGSQTLRRRQKFMWLKEIK...  \n",
       "588  MPEVSAKGTTISKKGFKKAVTKTQKKEGRKRKRCREESYSIYIYKV...  \n",
       "589  MADDTPIIEEIAEQNESVTRIMQRLKHDMQRVTSVPGFNTSAAGVN...  \n",
       "590  MPRHSRAKRAPRPSANNSKSPNDDDTAFRSPEPEDGTDYGLEFTTS...  \n",
       "591  MARTKHRVTRSQPRNQTDAAGASSSQAAGPTTTPTRRGGEGGDNTQ...  \n",
       "592  MAKKSLMAEPGDPIPRPRKKRYRPGTTALREIRKYQRSTDLLIQRL...  \n",
       "593  MAPGSVTSDISPSSTSTAGSSRSPESEKPGPSHGGVPPGGPSHSSL...  \n",
       "594  MPEVAVKGATISKKGFKKAVTKTQKKEGRKRKRCRKESYSIYIYKV...  \n",
       "595  MPEPAKSAPAPKKGSKKAVTKAQKKDGKKRKRSRKESYSVYVYKVL...  \n",
       "596  MEQALTGEAQSRWPRRGGSGAMAEAPGPSGESRGHSATQLPAEKTV...  \n",
       "597  MPELAKSAPAPKKGSKKAVTKAQKKDGKKRKRSRKESYSIYVYKVL...  \n",
       "598  MARKAYQPKRRSNSNQNQQRSDSLKKNKQDNLRSKSAGNQQGNEKN...  \n",
       "599  MVRTKKNIPNHNPLNAFNRDKSFKTNKTLPNRTVHHGISSKTTNIN...  \n",
       "600  MRPGEKALREIRQYQSSTSLLLRRLPFARLVREIQYGMTRQPYRWQ...  \n",
       "601  MVGVENLGVGFDELLTRGGCGVRDDAVEIAFRGVEGLEDVLKDYMV...  \n",
       "602  MARTKQTARKSTGGKAPRKQLVTKVARKSAPATGGVKKPHRYRPGT...  \n",
       "603  MARTKQTARKSTGGKAPRKQLATKAARKSTPSTXGVKKPHRYRPGT...  \n",
       "604  MVRVFGNPVGGAHQYRQTQALNRDESLREQLQRTTVAEQDENIEPE...  \n",
       "605  MPELTSKGTTISKKGFKRAVAKTQKKEGKKRRRCRKESYSIYIYKV...  \n",
       "606  MASPRPPALPASDASSSSVTDSGTDDSTPPNSPQPQRSPSPPFAAT...  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Записи, которые есть только во второй таблице (sequence_df):\n",
    "only_in_df2_ = only_in_df2.drop(columns=[\"_merge\"])\n",
    "only_in_df2_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "21f95162-5f93-48cf-bbc4-8af3729db504",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((572, 16), (569, 16))"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_histonescsv.shape, sequence_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "266b1fa8-66a7-4903-b77e-eb8f3682acc4",
   "metadata": {},
   "source": [
    "# Close connection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b7719968-002f-4229-846d-e4a8f180ec6b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cursor.close()\n",
    "conn.close()\n",
    "tunnel.stop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1717af72-4e6b-4e99-9524-937a32b10a91",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3198d1da-bec4-4ae2-b5ba-e828a81729d0",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80ac59fe-375e-4e82-9b6a-81c472b6dbb1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".conda-histdb_env",
   "language": "python",
   "name": "conda-env-.conda-histdb_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
