{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "53d3ec55-7daf-4ade-8add-2830b709dc54",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import io\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import Entrez, SeqIO\n",
    "from mysql.connector import connection\n",
    "from sshtunnel import SSHTunnelForwarder\n",
    "\n",
    "# Укажите ваш email (обязательно для использования Entrez)\n",
    "Entrez.email = \"your.email@example.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "608c4bfe-b979-4b60-b2b5-b94eeb679cb5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open(\"db_curated_server_info.txt\", \"r\") as file:\n",
    "    lines = file.readlines()\n",
    "\n",
    "config = {}\n",
    "\n",
    "for line in lines:\n",
    "    line = line.strip()\n",
    "    if line and not line.startswith(\"#\"):\n",
    "        key, value = line.split(\"=\", 1)\n",
    "        config[key] = value.strip()\n",
    "\n",
    "server_name = config.get(\"server_name\")\n",
    "srever_port = int(config.get(\"srever_port\"))\n",
    "ssh_password = config.get(\"ssh_password\")\n",
    "ssh_username = config.get(\"ssh_username\")\n",
    "db_adress = config.get(\"db_adress\")\n",
    "db_port = int(config.get(\"db_port\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4e19aa1c-49a3-48c4-b67f-34c94cef608d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44147\n"
     ]
    }
   ],
   "source": [
    "tunnel = SSHTunnelForwarder(\n",
    "    (server_name, srever_port),\n",
    "    ssh_password=ssh_password,\n",
    "    ssh_username=ssh_username,\n",
    "    remote_bind_address=(db_adress, db_port),\n",
    ")\n",
    "tunnel.start()\n",
    "print(tunnel.local_bind_port)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cef2275b-12d0-48f1-ad3c-ce4cdb9aa267",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "conn = connection.MySQLConnection(\n",
    "    user=\"db_user\",\n",
    "    password=\"db_password\",\n",
    "    host=\"localhost\",\n",
    "    port=tunnel.local_bind_port,\n",
    "    database=\"db_name\",\n",
    ")\n",
    "cursor = conn.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f589ea8d-8b83-4331-a1cf-b64ed7621801",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('alternative_name',),\n",
       " ('histone',),\n",
       " ('histone_description',),\n",
       " ('histone_has_publication',),\n",
       " ('publication',),\n",
       " ('sequence',),\n",
       " ('sequence_has_publication',)]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SHOW TABLES;\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b3dd9ef8-b162-4127-9535-ac7743af8c69",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# add_histone = (\n",
    "#     \"INSERT INTO histone \"\n",
    "#     \"(id, level, taxonomic_span, taxonomic_span_id, description, parent) \"\n",
    "#     \"VALUES (%(id)s, %(level)s, %(taxonomic_span)s, %(taxonomic_span_id)s, %(description)s, %(parent)s)\"\n",
    "# )\n",
    "# add_histone_description = (\n",
    "#     \"INSERT INTO histone_description \"\n",
    "#     \"(summary, taxonomy, genes, evolution, expression, knock_out, function, sequence, localization, deposition, structure, interactions, disease, caveats) \"\n",
    "#     \"VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)\"\n",
    "# )\n",
    "add_publication = (\n",
    "    \"INSERT INTO publication \"\n",
    "    \"(id, title, doi, author, year) \"\n",
    "    \"VALUES (%(id)s, %(title)s, %(doi)s, %(author)s, %(year)s)\"\n",
    ")\n",
    "# add_sequence = (\n",
    "#     \"INSERT INTO sequence \"\n",
    "#     \"(accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration) \"\n",
    "#     \"VALUES (%(accession)s, %(variant)s, %(gi)s, %(ncbi_gene_id)s, %(hgnc_gene_name)s, %(taxonomy_id)s, %(organism)s, %(phylum)s, %(class)s, %(taxonomy_group)s, %(info)s, %(sequence)s, %(variant_under_consideration)s)\"\n",
    "# )\n",
    "# add_sequence_has_publication = (\n",
    "#     \"INSERT INTO sequence_has_publication \"\n",
    "#     \"(sequence_accession, publication_id) \"\n",
    "#     \"VALUES (%s, %s)\"\n",
    "# )\n",
    "add_alternate_names = (\n",
    "    \"INSERT INTO alternative_name \"\n",
    "    \"(name, taxonomy, gene, splice, histone) \"\n",
    "    \"VALUES (%(name)s, %(taxonomy)s, %(gene)s, %(splice)s, %(histone)s)\"\n",
    ")\n",
    "add_histone_has_publication = (\n",
    "    \"INSERT INTO histone_has_publication \"\n",
    "    \"(histone_id, publication_id) \"\n",
    "    \"VALUES (%s, %s)\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6011bb8a-bab5-4fb5-b4ae-15b83accd15b",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_histone_description = {\n",
    "    \"summary\": None,\n",
    "    \"taxonomy\": None,\n",
    "    \"genes\": None,\n",
    "    \"evolution\": None, \n",
    "    \"expression\": None,\n",
    "    \"knock_out\": None,\n",
    "    \"function\": None,\n",
    "    \"sequence\": None,\n",
    "    \"localization\": None,\n",
    "    \"deposition\": None,\n",
    "    \"structure\": None,\n",
    "    \"interactions\": None,\n",
    "    \"disease\": None,\n",
    "    \"caveats\": None,\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c1c3bb0c-180f-4fb2-9cdc-ca928f9808be",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_taxonomy_data(record):\n",
    "    import re\n",
    "    import sys\n",
    "\n",
    "    taxonomy_data = {}\n",
    "    taxonomy_data[\"organism\"] = record.annotations[\"organism\"]\n",
    "    try:\n",
    "        for a in record.features[0].qualifiers[\"db_xref\"]:\n",
    "            text = re.search(\"(\\S+):(\\S+)\", a).group(1)\n",
    "            taxid = re.search(\"(\\S+):(\\S+)\", a).group(2)\n",
    "            if text == \"taxon\":\n",
    "                print(f\"Fetched taxid from NCBI {taxid}\")\n",
    "                taxonomy_data[\"taxonomy_id\"] = int(taxid)\n",
    "            else:\n",
    "                continue\n",
    "    except:\n",
    "        print(\"!!!!!!Unable to get TAXID for this record setting it to 1\")\n",
    "        taxonomy_data[\"taxonomy_id\"] = 1  # unable to identify\n",
    "\n",
    "    lineage = dict()\n",
    "    for i in range(10):\n",
    "        try:\n",
    "            handle = Entrez.efetch(\n",
    "                id=taxonomy_data[\"taxonomy_id\"], db=\"taxonomy\", retmode=\"xml\"\n",
    "            )\n",
    "            tax_data = Entrez.read(handle)\n",
    "            lineage = {\n",
    "                d[\"Rank\"]: d[\"ScientificName\"]\n",
    "                for d in tax_data[0][\"LineageEx\"]\n",
    "                if d[\"Rank\"] in [\"class\", \"phylum\"]\n",
    "            }\n",
    "            break\n",
    "        except:\n",
    "            print(\n",
    "                \"Unexpected error: {}, Retrying, attempt {}\".format(\n",
    "                    sys.exc_info()[0], i\n",
    "                )\n",
    "            )\n",
    "            if i == 9:\n",
    "                print(\n",
    "                    f\"FATAL ERROR could not get class and phylum from NCBI after 10 attempts for taxid:{taxonomy_data['taxonomy_id']}. Will add None for class and phylum!\"\n",
    "                )\n",
    "            else:\n",
    "                continue\n",
    "    taxonomy_data[\"phylum\"] = lineage.get(\"phylum\", None)\n",
    "    taxonomy_data[\"class\"] = lineage.get(\"class\", None)\n",
    "    if taxonomy_data[\"phylum\"] is not None:\n",
    "        taxonomy_data[\"phylum\"] = str(taxonomy_data[\"phylum\"])\n",
    "    if taxonomy_data[\"class\"] is not None:\n",
    "        taxonomy_data[\"class\"] = str(taxonomy_data[\"class\"])\n",
    "    return taxonomy_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "77d7b870-3148-4cc7-be4b-edc602bf12b9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "variant\n",
       "H3-like_(Viruses)        45\n",
       "H3.3                     18\n",
       "H3.Y_(Primates)           4\n",
       "H3.3_(Homo_sapiens)       2\n",
       "H3.4_(Homo_sapiens)       1\n",
       "H3.4_(Mammalia)           1\n",
       "H3.5_(Hominidae)          1\n",
       "H3.5_(Homo_sapiens)       1\n",
       "H3.Y.1_(Homo_sapiens)     1\n",
       "H3.Y.2_(Homo_sapiens)     1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \" \"WHERE variant LIKE 'H3%'\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[\"variant\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4fbcbde4-9de2-4fe7-be94-e374b410b2da",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HISTDB_H3_Y_0</td>\n",
       "      <td>H3.Y_(Primates)</td>\n",
       "      <td>NOGI</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9544</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ARTKQTARKATNWQAPRKPLATKAAAKRAPPRGGIKKPHRYKPGTQ...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HISTDB_H3_Y_1</td>\n",
       "      <td>H3.Y_(Primates)</td>\n",
       "      <td>NOGI</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9544</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ARTKQTARKATNWQAPRKPLATKAPGKRLPPRGGIKKPHRYRPGTQ...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HISTDB_H3_Y_2</td>\n",
       "      <td>H3.Y_(Primates)</td>\n",
       "      <td>NOGI</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9598</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGTL...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HISTDB_H3_Y_3</td>\n",
       "      <td>H3.Y_(Primates)</td>\n",
       "      <td>NOGI</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9598</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGTL...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession          variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  HISTDB_H3_Y_0  H3.Y_(Primates)  NOGI         None           None   \n",
       "1  HISTDB_H3_Y_1  H3.Y_(Primates)  NOGI         None           None   \n",
       "2  HISTDB_H3_Y_2  H3.Y_(Primates)  NOGI         None           None   \n",
       "3  HISTDB_H3_Y_3  H3.Y_(Primates)  NOGI         None           None   \n",
       "\n",
       "   taxonomy_id         organism    phylum     class taxonomy_group  info  \\\n",
       "0         9544   Macaca mulatta  Chordata  Mammalia           None  None   \n",
       "1         9544   Macaca mulatta  Chordata  Mammalia           None  None   \n",
       "2         9598  Pan troglodytes  Chordata  Mammalia           None  None   \n",
       "3         9598  Pan troglodytes  Chordata  Mammalia           None  None   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  ARTKQTARKATNWQAPRKPLATKAAAKRAPPRGGIKKPHRYKPGTQ...   \n",
       "1  ARTKQTARKATNWQAPRKPLATKAPGKRLPPRGGIKKPHRYRPGTQ...   \n",
       "2  ARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGTL...   \n",
       "3  ARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGTL...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                        None  \n",
       "1                        None  \n",
       "2                        None  \n",
       "3                        None  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \" \"WHERE variant='H3.Y_(Primates)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9a15b106-5cf1-40d4-8597-2b11dc30ff33",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>65</td>\n",
       "      <td>cid</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id name taxonomy  gene splice         histone\n",
       "0  65  cid     None  None   None  cenH3_(Plants)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name \" \"WHERE name='cid'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "acf1cf14-5f3d-475a-a62d-592198b0dbe9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, title, doi, author, year, pubmed_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM publication \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].str.startswith(\"hara\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "fe3a730d-c2fc-4212-b77a-5d4844d8059c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>cenH3.6_(Repleta)</td>\n",
       "      <td>variant</td>\n",
       "      <td>repleta group</td>\n",
       "      <td>32321</td>\n",
       "      <td>245.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>245.0</td>\n",
       "      <td>cenH3.6_(Repleta)  is a recently evolved centr...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "17  cenH3.6_(Repleta)  variant  repleta group             32321        245.0   \n",
       "\n",
       "                   parent     id  \\\n",
       "17  cenH3_(Drosophilidae)  245.0   \n",
       "\n",
       "                                              summary taxonomy genes  ...  \\\n",
       "17  cenH3.6_(Repleta)  is a recently evolved centr...     None  None  ...   \n",
       "\n",
       "   knock_out function sequence localization deposition structure interactions  \\\n",
       "17      None     None     None         None       None      None         None   \n",
       "\n",
       "   disease caveats relations  \n",
       "17    None    None      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone as h LEFT JOIN histone_description as hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    # \"WHERE h.id = 'H3.6_(Homo_sapiens)' \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df.iloc[:,0].str.contains(\"H3.6\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "fb49b9fa-f514-49d9-a8ee-dc9f9283c76b",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CENP-A nucleosomes are essential for chromosome segregation. A functional kinetochore interacts with active centromeric chromatin reached with CENP-A to form the mitotic spindle. Notably, in certain holocentric organisms, CENP-A appears dispensable for meiotic chromosome segregation while remaining essential for mitosis. As demonstrated in Caenorhabditis elegans, CENP-A loading is specifically eliminated following meiosis I, and RNAi-mediated depletion of CENP-A during meiosis fails to disrupt proper chromosome segregation [monen_differential_2005]. This stands in striking contrast to mitotic divisions, where CENP-A is absolutely required for kinetochore assembly and faithful chromosome segregation. Moreover, C. elegans and C. remanei possess a second cenH3 histone gene that likely performs specialized functions distinct from canonical CENP-A roles [monen_separase_2015]. For more detailed information see the \"gene\" section.'"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"genes\"].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "5fb70ce5-8cd8-4602-be43-096516d085ec",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "query = \"SELECT * FROM histone_description \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "# df\n",
    "for s in df.apply(lambda row: ''.join(row.astype(str)), axis=1):\n",
    "    if \"henikoff_heterochromatic_2000\" in s:\n",
    "        print(s)\n",
    "        print(\"---------------\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "400c2338-9a32-4efa-9e1a-e582d4424c20",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tables_in_db_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>alternative_name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>histone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>histone_description</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>histone_has_publication</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>publication</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sequence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>sequence_has_publication</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Tables_in_db_name\n",
       "0          alternative_name\n",
       "1                   histone\n",
       "2       histone_description\n",
       "3   histone_has_publication\n",
       "4               publication\n",
       "5                  sequence\n",
       "6  sequence_has_publication"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SHOW TABLES; \"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f5b32a2a-ee60-44f2-9b3f-46a239ebaa84",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>organism</th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1407</th>\n",
       "      <td>ELK02218.1</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Pteropus alecto</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2367</th>\n",
       "      <td>HISTDB_H2A_Q_0</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Cercopithecus aethiops</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2368</th>\n",
       "      <td>HISTDB_H2A_Q_1</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Pantholops hodgsonii</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2369</th>\n",
       "      <td>HISTDB_H2A_Q_10</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Ursus maritimus</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2370</th>\n",
       "      <td>HISTDB_H2A_Q_11</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Cercocebus atys</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2371</th>\n",
       "      <td>HISTDB_H2A_Q_12</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Odobenus rosmarus divergens</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2372</th>\n",
       "      <td>HISTDB_H2A_Q_13</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Ceratotherium simum</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2373</th>\n",
       "      <td>HISTDB_H2A_Q_14</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Bos mutus</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2374</th>\n",
       "      <td>HISTDB_H2A_Q_15</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2375</th>\n",
       "      <td>HISTDB_H2A_Q_16</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2376</th>\n",
       "      <td>HISTDB_H2A_Q_17</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Equus caballus</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2377</th>\n",
       "      <td>HISTDB_H2A_Q_18</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Papio anubis</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2378</th>\n",
       "      <td>HISTDB_H2A_Q_19</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Pongo pygmaeus abelii</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2379</th>\n",
       "      <td>HISTDB_H2A_Q_2</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Ammotragus lervia</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2380</th>\n",
       "      <td>HISTDB_H2A_Q_20</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Sus scrofa</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2381</th>\n",
       "      <td>HISTDB_H2A_Q_21</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Nasalis larvatus</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2382</th>\n",
       "      <td>HISTDB_H2A_Q_22</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Nomascus leucogenys</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2383</th>\n",
       "      <td>HISTDB_H2A_Q_3</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Bison bison bison</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2384</th>\n",
       "      <td>HISTDB_H2A_Q_4</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Pan paniscus</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2385</th>\n",
       "      <td>HISTDB_H2A_Q_5</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Mandrillus leucophaeus</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2386</th>\n",
       "      <td>HISTDB_H2A_Q_6</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Mustela putorius furo</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2387</th>\n",
       "      <td>HISTDB_H2A_Q_7</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Giraffa camelopardalis tippelskirchi</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2388</th>\n",
       "      <td>HISTDB_H2A_Q_8</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Ovis aries musimon</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389</th>\n",
       "      <td>HISTDB_H2A_Q_9</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Okapia johnstoni</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6441</th>\n",
       "      <td>XP_010950849.1</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Camelus bactrianus</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6442</th>\n",
       "      <td>XP_010995375.1</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Camelus dromedarius</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6462</th>\n",
       "      <td>XP_013966888.1</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Canis lupus familiaris</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6465</th>\n",
       "      <td>XP_015107649.1</td>\n",
       "      <td>H2A.Q_(Eutheria)</td>\n",
       "      <td>Vicugna pacos</td>\n",
       "      <td>29549088</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            accession           variant                              organism  \\\n",
       "1407       ELK02218.1  H2A.Q_(Eutheria)                       Pteropus alecto   \n",
       "2367   HISTDB_H2A_Q_0  H2A.Q_(Eutheria)                Cercopithecus aethiops   \n",
       "2368   HISTDB_H2A_Q_1  H2A.Q_(Eutheria)                  Pantholops hodgsonii   \n",
       "2369  HISTDB_H2A_Q_10  H2A.Q_(Eutheria)                       Ursus maritimus   \n",
       "2370  HISTDB_H2A_Q_11  H2A.Q_(Eutheria)                       Cercocebus atys   \n",
       "2371  HISTDB_H2A_Q_12  H2A.Q_(Eutheria)           Odobenus rosmarus divergens   \n",
       "2372  HISTDB_H2A_Q_13  H2A.Q_(Eutheria)                   Ceratotherium simum   \n",
       "2373  HISTDB_H2A_Q_14  H2A.Q_(Eutheria)                             Bos mutus   \n",
       "2374  HISTDB_H2A_Q_15  H2A.Q_(Eutheria)                       Pan troglodytes   \n",
       "2375  HISTDB_H2A_Q_16  H2A.Q_(Eutheria)                            Bos taurus   \n",
       "2376  HISTDB_H2A_Q_17  H2A.Q_(Eutheria)                        Equus caballus   \n",
       "2377  HISTDB_H2A_Q_18  H2A.Q_(Eutheria)                          Papio anubis   \n",
       "2378  HISTDB_H2A_Q_19  H2A.Q_(Eutheria)                 Pongo pygmaeus abelii   \n",
       "2379   HISTDB_H2A_Q_2  H2A.Q_(Eutheria)                     Ammotragus lervia   \n",
       "2380  HISTDB_H2A_Q_20  H2A.Q_(Eutheria)                            Sus scrofa   \n",
       "2381  HISTDB_H2A_Q_21  H2A.Q_(Eutheria)                      Nasalis larvatus   \n",
       "2382  HISTDB_H2A_Q_22  H2A.Q_(Eutheria)                   Nomascus leucogenys   \n",
       "2383   HISTDB_H2A_Q_3  H2A.Q_(Eutheria)                     Bison bison bison   \n",
       "2384   HISTDB_H2A_Q_4  H2A.Q_(Eutheria)                          Pan paniscus   \n",
       "2385   HISTDB_H2A_Q_5  H2A.Q_(Eutheria)                Mandrillus leucophaeus   \n",
       "2386   HISTDB_H2A_Q_6  H2A.Q_(Eutheria)                 Mustela putorius furo   \n",
       "2387   HISTDB_H2A_Q_7  H2A.Q_(Eutheria)  Giraffa camelopardalis tippelskirchi   \n",
       "2388   HISTDB_H2A_Q_8  H2A.Q_(Eutheria)                    Ovis aries musimon   \n",
       "2389   HISTDB_H2A_Q_9  H2A.Q_(Eutheria)                      Okapia johnstoni   \n",
       "6441   XP_010950849.1  H2A.Q_(Eutheria)                    Camelus bactrianus   \n",
       "6442   XP_010995375.1  H2A.Q_(Eutheria)                   Camelus dromedarius   \n",
       "6462   XP_013966888.1  H2A.Q_(Eutheria)                Canis lupus familiaris   \n",
       "6465   XP_015107649.1  H2A.Q_(Eutheria)                         Vicugna pacos   \n",
       "\n",
       "            id title   doi author  year pubmed_id  \n",
       "1407  29549088  None  None   None  None      None  \n",
       "2367  29549088  None  None   None  None      None  \n",
       "2368  29549088  None  None   None  None      None  \n",
       "2369  29549088  None  None   None  None      None  \n",
       "2370  29549088  None  None   None  None      None  \n",
       "2371  29549088  None  None   None  None      None  \n",
       "2372  29549088  None  None   None  None      None  \n",
       "2373  29549088  None  None   None  None      None  \n",
       "2374  29549088  None  None   None  None      None  \n",
       "2375  29549088  None  None   None  None      None  \n",
       "2376  29549088  None  None   None  None      None  \n",
       "2377  29549088  None  None   None  None      None  \n",
       "2378  29549088  None  None   None  None      None  \n",
       "2379  29549088  None  None   None  None      None  \n",
       "2380  29549088  None  None   None  None      None  \n",
       "2381  29549088  None  None   None  None      None  \n",
       "2382  29549088  None  None   None  None      None  \n",
       "2383  29549088  None  None   None  None      None  \n",
       "2384  29549088  None  None   None  None      None  \n",
       "2385  29549088  None  None   None  None      None  \n",
       "2386  29549088  None  None   None  None      None  \n",
       "2387  29549088  None  None   None  None      None  \n",
       "2388  29549088  None  None   None  None      None  \n",
       "2389  29549088  None  None   None  None      None  \n",
       "6441  29549088  None  None   None  None      None  \n",
       "6442  29549088  None  None   None  None      None  \n",
       "6462  29549088  None  None   None  None      None  \n",
       "6465  29549088  None  None   None  None      None  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT s.accession, s.variant, s.organism, p.* FROM sequence s LEFT JOIN \"\n",
    "    \"sequence_has_publication sp ON s.accession = sp.sequence_accession LEFT JOIN \"\n",
    "    \"publication p ON sp.publication_id = p.id \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"variant\"].str.startswith(\"H2A.Q\", na=False)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d0d3d7e-47ed-4f2b-8e96-e61eb030872d",
   "metadata": {},
   "source": [
    "# To Do H1\n",
    "\n",
    "## <span style=\"color:black\">Update description of cH1.1_(Homo_sapiens)</span>\n",
    "\n",
    "### <span style=\"color:black\">Update summary</span>\n",
    "\n",
    "```cH1.1_(Homo sapiens) is a replication-dependent linker histone in humans that is essential during the early stages of embryogenesis [funaya_involvement_2023]. Human cH1.1 interacts with the phosphorylated form of the recombinant protein PC4 to mediate chromatin condensation [mustafi_phosphorylation-dependent_2022].```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2eec2479",
   "metadata": {},
   "source": [
    "# To Do H3\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "400863d2-b22b-42d8-8c82-39ee7ce8e8ec",
   "metadata": {},
   "source": [
    "# To Do H4\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e8bad07-ef74-40fd-a93b-e7897df0c591",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Close connections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "e5673501-5845-491a-bbb9-f5d213721d58",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cursor.close()\n",
    "conn.close()\n",
    "tunnel.stop()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".conda-histdb_env",
   "language": "python",
   "name": "conda-env-.conda-histdb_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
