{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fb32231a-293e-45b8-af7d-d996c91a4edd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import Entrez, SeqIO\n",
    "from mysql.connector import connection\n",
    "from sshtunnel import SSHTunnelForwarder\n",
    "\n",
    "Entrez.email = \"l.singh@intbio.org\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d5fa78a7-4901-4e63-9d91-d9ed584d2f62",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open(\"db_curated_server_info.txt\", \"r\") as file:\n",
    "    lines = file.readlines()\n",
    "\n",
    "config = {}\n",
    "\n",
    "for line in lines:\n",
    "    line = line.strip()\n",
    "    if line and not line.startswith(\"#\"):\n",
    "        key, value = line.split(\"=\", 1)\n",
    "        config[key] = value.strip()\n",
    "\n",
    "server_name = config.get(\"server_name\")\n",
    "srever_port = int(config.get(\"srever_port\"))\n",
    "ssh_password = config.get(\"ssh_password\")\n",
    "ssh_username = config.get(\"ssh_username\")\n",
    "db_adress = config.get(\"db_adress\")\n",
    "db_port = int(config.get(\"db_port\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ebe3d2e8-9736-4fd9-b3f1-5ec2d255567d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "36799\n"
     ]
    }
   ],
   "source": [
    "tunnel = SSHTunnelForwarder(\n",
    "    (server_name, srever_port),\n",
    "    ssh_password=ssh_password,\n",
    "    ssh_username=ssh_username,\n",
    "    remote_bind_address=(db_adress, db_port),\n",
    ")\n",
    "tunnel.start()\n",
    "print(tunnel.local_bind_port)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0de93630-7502-40f3-ad24-42edf7a11eef",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "conn = connection.MySQLConnection(\n",
    "    user=\"db_user\",\n",
    "    password=\"db_password\",\n",
    "    host=\"localhost\",\n",
    "    port=tunnel.local_bind_port,\n",
    "    database=\"db_name\",\n",
    ")\n",
    "cursor = conn.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b4767ba0-4ea1-43ae-953f-f2575d9b7a69",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('alternative_name',),\n",
       " ('histone',),\n",
       " ('histone_description',),\n",
       " ('histone_has_publication',),\n",
       " ('publication',),\n",
       " ('sequence',),\n",
       " ('sequence_has_publication',)]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SHOW TABLES;\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7d59b3d0-7801-4177-8843-fdedf36cb6d1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# add_histone = (\n",
    "#     \"INSERT INTO histone \"\n",
    "#     \"(id, level, taxonomic_span, taxonomic_span_id, description, parent) \"\n",
    "#     \"VALUES (%(id)s, %(level)s, %(taxonomic_span)s, %(taxonomic_span_id)s, %(description)s, %(parent)s)\"\n",
    "# )\n",
    "# add_histone_description = (\n",
    "#     \"INSERT INTO histone_description \"\n",
    "#     \"(summary, taxonomy, genes, evolution, expression, knock_out, function, sequence, localization, deposition, structure, interactions, disease, caveats) \"\n",
    "#     \"VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)\"\n",
    "# )\n",
    "add_publication = (\n",
    "    \"INSERT INTO publication \"\n",
    "    \"(id, title, doi, author, year) \"\n",
    "    \"VALUES (%(id)s, %(title)s, %(doi)s, %(author)s, %(year)s)\"\n",
    ")\n",
    "add_sequence = (\n",
    "    \"INSERT INTO sequence \"\n",
    "    \"(accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration) \"\n",
    "    \"VALUES (%(accession)s, %(variant)s, %(gi)s, %(ncbi_gene_id)s, %(hgnc_gene_name)s, %(taxonomy_id)s, %(organism)s, %(phylum)s, %(class)s, %(taxonomy_group)s, %(info)s, %(sequence)s, %(variant_under_consideration)s)\"\n",
    ")\n",
    "add_sequence_has_publication = (\n",
    "    \"INSERT INTO sequence_has_publication \"\n",
    "    \"(sequence_accession, publication_id) \"\n",
    "    \"VALUES (%s, %s)\"\n",
    ")\n",
    "# add_alternate_names = (\n",
    "#     \"INSERT INTO alternative_name \"\n",
    "#     \"(name, taxonomy, gene, splice, histone) \"\n",
    "#     \"VALUES (%(name)s, %(taxonomy)s, %(gene)s, %(splice)s, %(histone)s)\"\n",
    "# )\n",
    "# add_histone_has_publication = (\n",
    "#     \"INSERT INTO histone_has_publication \"\n",
    "#     \"(histone_id, publication_id) \"\n",
    "#     \"VALUES (%s, %s)\"\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "722ba070-6274-4e3c-a5cc-7dc1c956f972",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_taxonomy_data(record):\n",
    "    import re\n",
    "    import sys\n",
    "\n",
    "    taxonomy_data = {}\n",
    "    taxonomy_data[\"organism\"] = record.annotations[\"organism\"]\n",
    "    try:\n",
    "        for a in record.features[0].qualifiers[\"db_xref\"]:\n",
    "            text = re.search(\"(\\S+):(\\S+)\", a).group(1)\n",
    "            taxid = re.search(\"(\\S+):(\\S+)\", a).group(2)\n",
    "            if text == \"taxon\":\n",
    "                print(f\"Fetched taxid from NCBI {taxid}\")\n",
    "                taxonomy_data[\"taxonomy_id\"] = int(taxid)\n",
    "            else:\n",
    "                continue\n",
    "    except:\n",
    "        print(\"!!!!!!Unable to get TAXID for this record setting it to 1\")\n",
    "        taxonomy_data[\"taxonomy_id\"] = 1  # unable to identify\n",
    "\n",
    "    lineage = dict()\n",
    "    for i in range(10):\n",
    "        try:\n",
    "            handle = Entrez.efetch(\n",
    "                id=taxonomy_data[\"taxonomy_id\"], db=\"taxonomy\", retmode=\"xml\"\n",
    "            )\n",
    "            tax_data = Entrez.read(handle)\n",
    "            lineage = {\n",
    "                d[\"Rank\"]: d[\"ScientificName\"]\n",
    "                for d in tax_data[0][\"LineageEx\"]\n",
    "                if d[\"Rank\"] in [\"class\", \"phylum\"]\n",
    "            }\n",
    "            break\n",
    "        except:\n",
    "            print(\n",
    "                \"Unexpected error: {}, Retrying, attempt {}\".format(\n",
    "                    sys.exc_info()[0], i\n",
    "                )\n",
    "            )\n",
    "            if i == 9:\n",
    "                print(\n",
    "                    f\"FATAL ERROR could not get class and phylum from NCBI after 10 attempts for taxid:{taxonomy_data['taxonomy_id']}. Will add None for class and phylum!\"\n",
    "                )\n",
    "            else:\n",
    "                continue\n",
    "    taxonomy_data[\"phylum\"] = lineage.get(\"phylum\", None)\n",
    "    taxonomy_data[\"class\"] = lineage.get(\"class\", None)\n",
    "    if taxonomy_data[\"phylum\"] is not None:\n",
    "        taxonomy_data[\"phylum\"] = str(taxonomy_data[\"phylum\"])\n",
    "    if taxonomy_data[\"class\"] is not None:\n",
    "        taxonomy_data[\"class\"] = str(taxonomy_data[\"class\"])\n",
    "    return taxonomy_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "72d8d54f-f2ed-46a1-9647-2ebc4a23a706",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: WP_013413995.1\n",
      "Name: WP_013413995\n",
      "Description: histone HmfA [Methanothermus fervidus]\n",
      "Number of features: 4\n",
      "/topology=linear\n",
      "/data_file_division=BCT\n",
      "/date=22-OCT-2023\n",
      "/accessions=['WP_013413995']\n",
      "/sequence_version=1\n",
      "/keywords=['RefSeq']\n",
      "/source=Methanothermus fervidus\n",
      "/organism=Methanothermus fervidus\n",
      "/taxonomy=['Archaea', 'Euryarchaeota', 'Methanomada group', 'Methanobacteria', 'Methanobacteriales', 'Methanothermaceae', 'Methanothermus']\n",
      "/references=[Reference(title='Specific DNA binding of archaeal histones HMfA and HMfB', ...), Reference(title='Crystal structures of recombinant histones HMfA and HMfB from the hyperthermophilic archaeon Methanothermus fervidus', ...), Reference(title='HMf, a DNA-binding protein isolated from the hyperthermophilic archaeon Methanothermus fervidus, is most closely related to histones', ...)]\n",
      "/comment=REFSEQ: This record represents a single, non-redundant, protein\n",
      "sequence which may be annotated on many different RefSeq genomes\n",
      "from the same, or different, species.\n",
      "COMPLETENESS: full length.\n",
      "/structured_comment=OrderedDict([('Evidence-For-Name-Assignment', OrderedDict([('Evidence Category', 'BlastRule'), ('Evidence Accession', 'NBR016328'), ('Evidence Source', 'NCBI')]))])\n",
      "/molecule_type=protein\n",
      "Seq('MGELPIAPIGRIIKNAGAERVSDDARIALAKVLEEMGEEIASEAVKLAKHAGRK...MFK')\n",
      "MGELPIAPIGRIIKNAGAERVSDDARIALAKVLEEMGEEIASEAVKLAKHAGRKTIKAEDIELARKMFK\n"
     ]
    }
   ],
   "source": [
    "with Entrez.efetch(db=\"protein\", id=accession, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "    record = SeqIO.read(handle, \"genbank\")\n",
    "    print(record)\n",
    "print(record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "9c008420-3163-42dc-8aa9-007015d337c3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'WP_013413995.1'"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "record.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "ac5e3f52-3303-4b26-bc4b-b0c1d6729a28",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 2180\n",
      "accession WP_013413995 <class 'str'>\n",
      "variant HMfA_(Methanothermus_fervidus) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 2180 <class 'int'>\n",
      "organism Methanothermus fervidus <class 'str'>\n",
      "phylum Methanobacteriota <class 'str'>\n",
      "class Methanobacteria <class 'str'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MGELPIAPIGRIIKNAGAERVSDDARIALAKVLEEMGEEIASEAVKLAKHAGRKTIKAEDIELARKMFK <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "taxonomy_data = get_taxonomy_data(record)\n",
    "data_sequence = {\n",
    "    \"accession\": accession,\n",
    "    \"variant\": \"HMfA_(Methanothermus_fervidus)\",\n",
    "    \"gi\": None,\n",
    "    \"ncbi_gene_id\": None,\n",
    "    \"hgnc_gene_name\": None,\n",
    "    \"taxonomy_id\": None,\n",
    "    \"organism\": None,\n",
    "    \"phylum\": None,\n",
    "    \"class\": None,\n",
    "    \"taxonomy_group\": None,\n",
    "    \"info\": None,\n",
    "    \"sequence\": str(record.seq),\n",
    "    \"variant_under_consideration\": None,\n",
    "}\n",
    "data_sequence.update(taxonomy_data)\n",
    "for k, v in data_sequence.items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "53b0b155-13fd-440f-a513-3e97e0e50163",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cursor.execute(add_sequence, data_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "6f396f70-d36c-4511-b8ac-469ccdfc1f4a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "query = f\"UPDATE sequence SET accession='{record.id}' WHERE accession='WP_013413995'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "9a106228-44cc-4be5-9a8b-78f85be9cabc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5002</th>\n",
       "      <td>WP_013413995.1</td>\n",
       "      <td>HMfA_(Methanothermus_fervidus)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2180.0</td>\n",
       "      <td>Methanothermus fervidus</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Methanobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGELPIAPIGRIIKNAGAERVSDDARIALAKVLEEMGEEIASEAVK...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession                         variant    gi ncbi_gene_id  \\\n",
       "5002  WP_013413995.1  HMfA_(Methanothermus_fervidus)  None         None   \n",
       "\n",
       "     hgnc_gene_name  taxonomy_id                 organism             phylum  \\\n",
       "5002           None       2180.0  Methanothermus fervidus  Methanobacteriota   \n",
       "\n",
       "                class taxonomy_group  info  \\\n",
       "5002  Methanobacteria           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "5002  MGELPIAPIGRIIKNAGAERVSDDARIALAKVLEEMGEEIASEAVK...   \n",
       "\n",
       "     variant_under_consideration sequence_accession publication_id  \n",
       "5002                        None               None           None  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].str.contains(\"WP_013413995\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "5117c140-7e2e-4e90-b076-f62cf8971fc6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, title, doi, author, year]\n",
       "Index: []"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pid = \"henneman_structure_2018\"\n",
    "query = f\"SELECT * FROM publication WHERE id='{pid}'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "200c3930-f1ce-4826-86a4-2aa6728ad35d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_publication = {\n",
    "    \"id\": pid,\n",
    "    \"title\": \"Structure and function of archaeal histones\",\n",
    "    \"doi\": \"10.1371/journal.pgen.1007582\",\n",
    "    \"author\": None,\n",
    "    \"year\": \"2018\",\n",
    "}\n",
    "cursor.execute(add_publication, data_publication)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "8c29b8d8-c8fe-46e7-8351-6285494a91f1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cursor.execute(add_sequence_has_publication, (\"WP_013413995.1\", pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "eec5907c-b134-434f-a1a1-b44347064cb9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5002</th>\n",
       "      <td>WP_013413995.1</td>\n",
       "      <td>HMfA_(Methanothermus_fervidus)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2180.0</td>\n",
       "      <td>Methanothermus fervidus</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Methanobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGELPIAPIGRIIKNAGAERVSDDARIALAKVLEEMGEEIASEAVK...</td>\n",
       "      <td>None</td>\n",
       "      <td>WP_013413995.1</td>\n",
       "      <td>henneman_structure_2018</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession                         variant    gi ncbi_gene_id  \\\n",
       "5002  WP_013413995.1  HMfA_(Methanothermus_fervidus)  None         None   \n",
       "\n",
       "     hgnc_gene_name  taxonomy_id                 organism             phylum  \\\n",
       "5002           None       2180.0  Methanothermus fervidus  Methanobacteriota   \n",
       "\n",
       "                class taxonomy_group  info  \\\n",
       "5002  Methanobacteria           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "5002  MGELPIAPIGRIIKNAGAERVSDDARIALAKVLEEMGEEIASEAVK...   \n",
       "\n",
       "     variant_under_consideration sequence_accession           publication_id  \n",
       "5002                        None     WP_013413995.1  henneman_structure_2018  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].str.contains(\"WP_013413995\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "dc421b2d-70d9-42d6-ac1b-2402b64f6336",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64eeb69d-7b59-4630-9d56-7fd47a294be6",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# Thermococcus\n",
    "\n",
    "В базе уже добавлены 132 последовательностей данного отряда.\n",
    "\n",
    "Некоторые последовательности попали в группу FtF. Скорее всего это гистоны, похожие на бактериальные. В [статье](https://academic.oup.com/gbe/article/doi/10.1093/gbe/evab274/6459647) говорилось о том, что отряд *Thermococcus* имеет такие гистоны. Поэтому надо будет проверить.\n",
    "\n",
    "Гены hpkA (TK1413) и hpkB (TK2289) кодируют гистоны HTkA и HTkB, соответственно (см. [статью](https://academic.oup.com/gbe/article/doi/10.1093/gbe/evab274/6459647)). Accessions для HTkA: WP_011250364.1, BAA77575.1, BAD85602.1, CAT72447.1; для HTkB: WP_011251239.1, BAA77576.1, BAD86478.1, CAT71503.1.\n",
    "\n",
    "BAD85229.1 классифицирован как FtF ???"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5de9b38f-c06c-40b8-afb2-4d523aea7671",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AAB53861.1</td>\n",
       "      <td>Nucleosomal</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1151117.0</td>\n",
       "      <td>Thermococcus zilligii AN1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPIDRLIRKAGAERVSEDAAKALAEYLEEYAIEVGKKATE...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>ACJ15670.1</td>\n",
       "      <td>Nucleosomal</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>523850.0</td>\n",
       "      <td>Thermococcus onnurineus NA1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPIDRLIRKAGAERVSEEAAKVLAEYLEEYAMDLAKRAAE...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>ACJ16232.1</td>\n",
       "      <td>FtF</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>523850.0</td>\n",
       "      <td>Thermococcus onnurineus NA1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAEMIVKSKVKEAVKAIDPEMRINPEFYEALEAEIKILIEKAVKRA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>ACJ16723.1</td>\n",
       "      <td>Nucleosomal</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>523850.0</td>\n",
       "      <td>Thermococcus onnurineus NA1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAARVSEDAAKLLAEHLEEKALEIAKKAVD...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>ACS32979.1</td>\n",
       "      <td>Nucleosomal</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>593117.0</td>\n",
       "      <td>Thermococcus gammatolerans EJ3</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKAIEIAKKAVE...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3764</th>\n",
       "      <td>QEK14533.1</td>\n",
       "      <td>Nucleosomal</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2598455.0</td>\n",
       "      <td>Thermococcus aciditolerans</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPIDRLIRKAGAERVSEDAAKVLAEYLEEYAIELARKSAD...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3765</th>\n",
       "      <td>QEK15571.1</td>\n",
       "      <td>Nucleosomal</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2598455.0</td>\n",
       "      <td>Thermococcus aciditolerans</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAARVSEDAAKLLAEHLEEKAMEIARKAVD...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4414</th>\n",
       "      <td>SEV81931.1</td>\n",
       "      <td>Nucleosomal</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>277988.0</td>\n",
       "      <td>Thermococcus thioreducens</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAARVSEDAAKLLAEHLEEKALEIAKKAVD...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4415</th>\n",
       "      <td>SEV88938.1</td>\n",
       "      <td>FtF</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>277988.0</td>\n",
       "      <td>Thermococcus thioreducens</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELIVKSKVKEAVKAIEPEMRVNPEFYEALEAEIKALIEKAVKRA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4416</th>\n",
       "      <td>SEW09872.1</td>\n",
       "      <td>Nucleosomal</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>277988.0</td>\n",
       "      <td>Thermococcus thioreducens</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPIDRLIRKAGAERVSEEAAKVLAEYLEEYAIELAKKSAE...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>132 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession      variant    gi ncbi_gene_id hgnc_gene_name  taxonomy_id  \\\n",
       "11    AAB53861.1  Nucleosomal  None         None           None    1151117.0   \n",
       "93    ACJ15670.1  Nucleosomal  None         None           None     523850.0   \n",
       "94    ACJ16232.1          FtF  None         None           None     523850.0   \n",
       "95    ACJ16723.1  Nucleosomal  None         None           None     523850.0   \n",
       "105   ACS32979.1  Nucleosomal  None         None           None     593117.0   \n",
       "...          ...          ...   ...          ...            ...          ...   \n",
       "3764  QEK14533.1  Nucleosomal  None         None           None    2598455.0   \n",
       "3765  QEK15571.1  Nucleosomal  None         None           None    2598455.0   \n",
       "4414  SEV81931.1  Nucleosomal  None         None           None     277988.0   \n",
       "4415  SEV88938.1          FtF  None         None           None     277988.0   \n",
       "4416  SEW09872.1  Nucleosomal  None         None           None     277988.0   \n",
       "\n",
       "                            organism             phylum        class  \\\n",
       "11         Thermococcus zilligii AN1  Methanobacteriota  Thermococci   \n",
       "93       Thermococcus onnurineus NA1  Methanobacteriota  Thermococci   \n",
       "94       Thermococcus onnurineus NA1  Methanobacteriota  Thermococci   \n",
       "95       Thermococcus onnurineus NA1  Methanobacteriota  Thermococci   \n",
       "105   Thermococcus gammatolerans EJ3  Methanobacteriota  Thermococci   \n",
       "...                              ...                ...          ...   \n",
       "3764      Thermococcus aciditolerans  Methanobacteriota  Thermococci   \n",
       "3765      Thermococcus aciditolerans  Methanobacteriota  Thermococci   \n",
       "4414       Thermococcus thioreducens  Methanobacteriota  Thermococci   \n",
       "4415       Thermococcus thioreducens  Methanobacteriota  Thermococci   \n",
       "4416       Thermococcus thioreducens  Methanobacteriota  Thermococci   \n",
       "\n",
       "     taxonomy_group  info                                           sequence  \\\n",
       "11             None  None  MAELPIAPIDRLIRKAGAERVSEDAAKALAEYLEEYAIEVGKKATE...   \n",
       "93             None  None  MAELPIAPIDRLIRKAGAERVSEEAAKVLAEYLEEYAMDLAKRAAE...   \n",
       "94             None  None  MAEMIVKSKVKEAVKAIDPEMRINPEFYEALEAEIKILIEKAVKRA...   \n",
       "95             None  None  MAELPIAPVDRLIRKAGAARVSEDAAKLLAEHLEEKALEIAKKAVD...   \n",
       "105            None  None  MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKAIEIAKKAVE...   \n",
       "...             ...   ...                                                ...   \n",
       "3764           None  None  MAELPIAPIDRLIRKAGAERVSEDAAKVLAEYLEEYAIELARKSAD...   \n",
       "3765           None  None  MAELPIAPVDRLIRKAGAARVSEDAAKLLAEHLEEKAMEIARKAVD...   \n",
       "4414           None  None  MAELPIAPVDRLIRKAGAARVSEDAAKLLAEHLEEKALEIAKKAVD...   \n",
       "4415           None  None  MAELIVKSKVKEAVKAIEPEMRVNPEFYEALEAEIKALIEKAVKRA...   \n",
       "4416           None  None  MAELPIAPIDRLIRKAGAERVSEEAAKVLAEYLEEYAIELAKKSAE...   \n",
       "\n",
       "     variant_under_consideration  \n",
       "11                          None  \n",
       "93                          None  \n",
       "94                          None  \n",
       "95                          None  \n",
       "105                         None  \n",
       "...                          ...  \n",
       "3764                        None  \n",
       "3765                        None  \n",
       "4414                        None  \n",
       "4415                        None  \n",
       "4416                        None  \n",
       "\n",
       "[132 rows x 13 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"organism\"].str.contains(\"Thermococcus\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c642285a-41e4-4f46-b989-cb80cdf3cb14",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_therm = df[df[\"organism\"].str.contains(\"Thermococcus\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f50a99cb-da9e-4602-bb0b-c85cb4c998de",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "variant\n",
       "Nucleosomal    91\n",
       "FtF            41\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_therm[\"variant\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1ae272e0-0b83-4626-abb7-5d269da200da",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_therm[df_therm[\"organism\"].str.contains(\"TS600\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5b98882b-b5b7-4286-838c-0b9516d70810",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>477</th>\n",
       "      <td>BAA77575.1</td>\n",
       "      <td>Nucleosomal</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>478</th>\n",
       "      <td>BAA77576.1</td>\n",
       "      <td>Nucleosomal</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480</th>\n",
       "      <td>BAD85229.1</td>\n",
       "      <td>FtF</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAEMLVKSKVKEFVKSVDPEMRVSPEFYDALEAEVKALVEKAIKRA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>481</th>\n",
       "      <td>BAD85602.1</td>\n",
       "      <td>Nucleosomal</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482</th>\n",
       "      <td>BAD86478.1</td>\n",
       "      <td>Nucleosomal</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession      variant    gi ncbi_gene_id hgnc_gene_name  taxonomy_id  \\\n",
       "477  BAA77575.1  Nucleosomal  None         None           None      69014.0   \n",
       "478  BAA77576.1  Nucleosomal  None         None           None      69014.0   \n",
       "480  BAD85229.1          FtF  None         None           None      69014.0   \n",
       "481  BAD85602.1  Nucleosomal  None         None           None      69014.0   \n",
       "482  BAD86478.1  Nucleosomal  None         None           None      69014.0   \n",
       "\n",
       "                           organism             phylum        class  \\\n",
       "477  Thermococcus kodakarensis KOD1  Methanobacteriota  Thermococci   \n",
       "478  Thermococcus kodakarensis KOD1  Methanobacteriota  Thermococci   \n",
       "480  Thermococcus kodakarensis KOD1  Methanobacteriota  Thermococci   \n",
       "481  Thermococcus kodakarensis KOD1  Methanobacteriota  Thermococci   \n",
       "482  Thermococcus kodakarensis KOD1  Methanobacteriota  Thermococci   \n",
       "\n",
       "    taxonomy_group  info                                           sequence  \\\n",
       "477           None  None  MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...   \n",
       "478           None  None  MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...   \n",
       "480           None  None  MAEMLVKSKVKEFVKSVDPEMRVSPEFYDALEAEVKALVEKAIKRA...   \n",
       "481           None  None  MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...   \n",
       "482           None  None  MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...   \n",
       "\n",
       "    variant_under_consideration  \n",
       "477                        None  \n",
       "478                        None  \n",
       "480                        None  \n",
       "481                        None  \n",
       "482                        None  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_therm[df_therm[\"organism\"].str.contains(\"Thermococcus kodakarensis\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e2934c26-af79-4a13-9182-578a14c2939c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>477</th>\n",
       "      <td>BAA77575.1</td>\n",
       "      <td>Nucleosomal</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>481</th>\n",
       "      <td>BAD85602.1</td>\n",
       "      <td>Nucleosomal</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession      variant    gi ncbi_gene_id hgnc_gene_name  taxonomy_id  \\\n",
       "477  BAA77575.1  Nucleosomal  None         None           None      69014.0   \n",
       "481  BAD85602.1  Nucleosomal  None         None           None      69014.0   \n",
       "\n",
       "                           organism             phylum        class  \\\n",
       "477  Thermococcus kodakarensis KOD1  Methanobacteriota  Thermococci   \n",
       "481  Thermococcus kodakarensis KOD1  Methanobacteriota  Thermococci   \n",
       "\n",
       "    taxonomy_group  info                                           sequence  \\\n",
       "477           None  None  MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...   \n",
       "481           None  None  MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...   \n",
       "\n",
       "    variant_under_consideration  \n",
       "477                        None  \n",
       "481                        None  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "accessions = {\n",
    "    \"HTkA\": [\"WP_011250364.1\", \"BAA77575.1\", \"BAD85602.1\", \"CAT72447.1\"],\n",
    "    \"HTkB\": [\"WP_011251239.1\", \"BAA77576.1\", \"BAD86478.1\", \"CAT71503.1\"],\n",
    "}\n",
    "df_therm[df_therm[\"accession\"].isin(accessions[\"HTkA\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "bf69b5cb-58b8-42bc-bbb2-d2e2baed9b42",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>478</th>\n",
       "      <td>BAA77576.1</td>\n",
       "      <td>Nucleosomal</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482</th>\n",
       "      <td>BAD86478.1</td>\n",
       "      <td>Nucleosomal</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession      variant    gi ncbi_gene_id hgnc_gene_name  taxonomy_id  \\\n",
       "478  BAA77576.1  Nucleosomal  None         None           None      69014.0   \n",
       "482  BAD86478.1  Nucleosomal  None         None           None      69014.0   \n",
       "\n",
       "                           organism             phylum        class  \\\n",
       "478  Thermococcus kodakarensis KOD1  Methanobacteriota  Thermococci   \n",
       "482  Thermococcus kodakarensis KOD1  Methanobacteriota  Thermococci   \n",
       "\n",
       "    taxonomy_group  info                                           sequence  \\\n",
       "478           None  None  MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...   \n",
       "482           None  None  MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...   \n",
       "\n",
       "    variant_under_consideration  \n",
       "478                        None  \n",
       "482                        None  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_therm[df_therm[\"accession\"].isin(accessions[\"HTkB\"])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88871c00-c411-40ca-8add-130bca59897a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# Correct variants for BAA77575.1 and BAD85602.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b37f5f39-7ad0-45fc-b9a5-e9a16f315721",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "query = f\"UPDATE sequence SET variant='HTkA_(Thermococcus_kodakarensis)' WHERE accession='BAA77575.1'\"\n",
    "cursor.execute(query)\n",
    "\n",
    "query = f\"UPDATE sequence SET variant='HTkA_(Thermococcus_kodakarensis)' WHERE accession='BAD85602.1'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "387a5c5e-416b-478a-8262-b5b7b134699b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>483</th>\n",
       "      <td>BAA77575.1</td>\n",
       "      <td>HTkA_(Thermococcus_kodakarensis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAA77575.1</td>\n",
       "      <td>schwab_histones_2024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>487</th>\n",
       "      <td>BAD85602.1</td>\n",
       "      <td>HTkA_(Thermococcus_kodakarensis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAD85602.1</td>\n",
       "      <td>schwab_histones_2024</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession                           variant    gi ncbi_gene_id  \\\n",
       "483  BAA77575.1  HTkA_(Thermococcus_kodakarensis)  None         None   \n",
       "487  BAD85602.1  HTkA_(Thermococcus_kodakarensis)  None         None   \n",
       "\n",
       "    hgnc_gene_name  taxonomy_id                        organism  \\\n",
       "483           None      69014.0  Thermococcus kodakarensis KOD1   \n",
       "487           None      69014.0  Thermococcus kodakarensis KOD1   \n",
       "\n",
       "                phylum        class taxonomy_group  info  \\\n",
       "483  Methanobacteriota  Thermococci           None  None   \n",
       "487  Methanobacteriota  Thermococci           None  None   \n",
       "\n",
       "                                              sequence  \\\n",
       "483  MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...   \n",
       "487  MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...   \n",
       "\n",
       "    variant_under_consideration sequence_accession        publication_id  \n",
       "483                        None         BAA77575.1  schwab_histones_2024  \n",
       "487                        None         BAD85602.1  schwab_histones_2024  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions[\"HTkA\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "16655861-16de-4c21-ad04-07c329651bc2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, title, doi, author, year]\n",
       "Index: []"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pid = \"stevens_deep_2022\"\n",
    "query = f\"SELECT * FROM publication WHERE id='{pid}'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "952fc95e-c9b1-4d19-9648-b5fdb7207921",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_publication = {\n",
    "    \"id\": pid,\n",
    "    \"title\": \"Deep Conservation of Histone Variants in Thermococcales Archaea\",\n",
    "    \"doi\": \"10.1093/gbe/evab274\",\n",
    "    \"author\": None,\n",
    "    \"year\": \"2022\",\n",
    "}\n",
    "cursor.execute(add_publication, data_publication)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "951b62f2-748b-45ea-a7e2-69e43c422452",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in [\"BAA77575.1\", \"BAD85602.1\"]:\n",
    "    cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c646c9a6-04c3-4952-9283-2457fc22418c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>483</th>\n",
       "      <td>BAA77575.1</td>\n",
       "      <td>HTkA_(Thermococcus_kodakarensis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAA77575.1</td>\n",
       "      <td>schwab_histones_2024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>BAA77575.1</td>\n",
       "      <td>HTkA_(Thermococcus_kodakarensis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAA77575.1</td>\n",
       "      <td>stevens_deep_2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>488</th>\n",
       "      <td>BAD85602.1</td>\n",
       "      <td>HTkA_(Thermococcus_kodakarensis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAD85602.1</td>\n",
       "      <td>schwab_histones_2024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>489</th>\n",
       "      <td>BAD85602.1</td>\n",
       "      <td>HTkA_(Thermococcus_kodakarensis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAD85602.1</td>\n",
       "      <td>stevens_deep_2022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession                           variant    gi ncbi_gene_id  \\\n",
       "483  BAA77575.1  HTkA_(Thermococcus_kodakarensis)  None         None   \n",
       "484  BAA77575.1  HTkA_(Thermococcus_kodakarensis)  None         None   \n",
       "488  BAD85602.1  HTkA_(Thermococcus_kodakarensis)  None         None   \n",
       "489  BAD85602.1  HTkA_(Thermococcus_kodakarensis)  None         None   \n",
       "\n",
       "    hgnc_gene_name  taxonomy_id                        organism  \\\n",
       "483           None      69014.0  Thermococcus kodakarensis KOD1   \n",
       "484           None      69014.0  Thermococcus kodakarensis KOD1   \n",
       "488           None      69014.0  Thermococcus kodakarensis KOD1   \n",
       "489           None      69014.0  Thermococcus kodakarensis KOD1   \n",
       "\n",
       "                phylum        class taxonomy_group  info  \\\n",
       "483  Methanobacteriota  Thermococci           None  None   \n",
       "484  Methanobacteriota  Thermococci           None  None   \n",
       "488  Methanobacteriota  Thermococci           None  None   \n",
       "489  Methanobacteriota  Thermococci           None  None   \n",
       "\n",
       "                                              sequence  \\\n",
       "483  MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...   \n",
       "484  MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...   \n",
       "488  MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...   \n",
       "489  MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...   \n",
       "\n",
       "    variant_under_consideration sequence_accession        publication_id  \n",
       "483                        None         BAA77575.1  schwab_histones_2024  \n",
       "484                        None         BAA77575.1     stevens_deep_2022  \n",
       "488                        None         BAD85602.1  schwab_histones_2024  \n",
       "489                        None         BAD85602.1     stevens_deep_2022  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions[\"HTkA\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "3ba31147-5146-4c75-8ae9-209da0fc2960",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63ff31e5-93b8-47f0-841c-d17827cebc21",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# Correct variants for BAA77576.1 and BAD86478.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "50a10aef-00b0-43da-9e66-3c00791a76f8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "query = f\"UPDATE sequence SET variant='HTkB_(Thermococcus_kodakarensis)' WHERE accession='BAA77576.1'\"\n",
    "cursor.execute(query)\n",
    "\n",
    "query = f\"UPDATE sequence SET variant='HTkB_(Thermococcus_kodakarensis)' WHERE accession='BAD86478.1'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "8b1b2c44-15ef-45f2-af24-8f93ccc36506",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>485</th>\n",
       "      <td>BAA77576.1</td>\n",
       "      <td>HTkB_(Thermococcus_kodakarensis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAA77576.1</td>\n",
       "      <td>schwab_histones_2024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>490</th>\n",
       "      <td>BAD86478.1</td>\n",
       "      <td>HTkB_(Thermococcus_kodakarensis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAD86478.1</td>\n",
       "      <td>schwab_histones_2024</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession                           variant    gi ncbi_gene_id  \\\n",
       "485  BAA77576.1  HTkB_(Thermococcus_kodakarensis)  None         None   \n",
       "490  BAD86478.1  HTkB_(Thermococcus_kodakarensis)  None         None   \n",
       "\n",
       "    hgnc_gene_name  taxonomy_id                        organism  \\\n",
       "485           None      69014.0  Thermococcus kodakarensis KOD1   \n",
       "490           None      69014.0  Thermococcus kodakarensis KOD1   \n",
       "\n",
       "                phylum        class taxonomy_group  info  \\\n",
       "485  Methanobacteriota  Thermococci           None  None   \n",
       "490  Methanobacteriota  Thermococci           None  None   \n",
       "\n",
       "                                              sequence  \\\n",
       "485  MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...   \n",
       "490  MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...   \n",
       "\n",
       "    variant_under_consideration sequence_accession        publication_id  \n",
       "485                        None         BAA77576.1  schwab_histones_2024  \n",
       "490                        None         BAD86478.1  schwab_histones_2024  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions[\"HTkB\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "fefb3374-95c4-42b6-962e-55a818aac8a0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>stevens_deep_2022</td>\n",
       "      <td>Deep Conservation of Histone Variants in Therm...</td>\n",
       "      <td>10.1093/gbe/evab274</td>\n",
       "      <td>None</td>\n",
       "      <td>2022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  id                                              title  \\\n",
       "0  stevens_deep_2022  Deep Conservation of Histone Variants in Therm...   \n",
       "\n",
       "                   doi author  year  \n",
       "0  10.1093/gbe/evab274   None  2022  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pid = \"stevens_deep_2022\"\n",
    "query = f\"SELECT * FROM publication WHERE id='{pid}'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "9cdef31b-5101-48ba-8083-fc726efd1d1b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in [\"BAA77576.1\", \"BAD86478.1\"]:\n",
    "    cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "253ef8b9-0410-4f0e-865f-b3c215386f0d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>485</th>\n",
       "      <td>BAA77576.1</td>\n",
       "      <td>HTkB_(Thermococcus_kodakarensis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAA77576.1</td>\n",
       "      <td>schwab_histones_2024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>BAA77576.1</td>\n",
       "      <td>HTkB_(Thermococcus_kodakarensis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAA77576.1</td>\n",
       "      <td>stevens_deep_2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>491</th>\n",
       "      <td>BAD86478.1</td>\n",
       "      <td>HTkB_(Thermococcus_kodakarensis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAD86478.1</td>\n",
       "      <td>schwab_histones_2024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492</th>\n",
       "      <td>BAD86478.1</td>\n",
       "      <td>HTkB_(Thermococcus_kodakarensis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAD86478.1</td>\n",
       "      <td>stevens_deep_2022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession                           variant    gi ncbi_gene_id  \\\n",
       "485  BAA77576.1  HTkB_(Thermococcus_kodakarensis)  None         None   \n",
       "486  BAA77576.1  HTkB_(Thermococcus_kodakarensis)  None         None   \n",
       "491  BAD86478.1  HTkB_(Thermococcus_kodakarensis)  None         None   \n",
       "492  BAD86478.1  HTkB_(Thermococcus_kodakarensis)  None         None   \n",
       "\n",
       "    hgnc_gene_name  taxonomy_id                        organism  \\\n",
       "485           None      69014.0  Thermococcus kodakarensis KOD1   \n",
       "486           None      69014.0  Thermococcus kodakarensis KOD1   \n",
       "491           None      69014.0  Thermococcus kodakarensis KOD1   \n",
       "492           None      69014.0  Thermococcus kodakarensis KOD1   \n",
       "\n",
       "                phylum        class taxonomy_group  info  \\\n",
       "485  Methanobacteriota  Thermococci           None  None   \n",
       "486  Methanobacteriota  Thermococci           None  None   \n",
       "491  Methanobacteriota  Thermococci           None  None   \n",
       "492  Methanobacteriota  Thermococci           None  None   \n",
       "\n",
       "                                              sequence  \\\n",
       "485  MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...   \n",
       "486  MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...   \n",
       "491  MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...   \n",
       "492  MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...   \n",
       "\n",
       "    variant_under_consideration sequence_accession        publication_id  \n",
       "485                        None         BAA77576.1  schwab_histones_2024  \n",
       "486                        None         BAA77576.1     stevens_deep_2022  \n",
       "491                        None         BAD86478.1  schwab_histones_2024  \n",
       "492                        None         BAD86478.1     stevens_deep_2022  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions[\"HTkB\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "d0a6c103-e017-4466-b74b-396e515997b1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5f540d1-442e-498b-8ac4-c633963f41b3",
   "metadata": {},
   "source": [
    "# Add WP_011250364.1 and CAT72447.1 as HTkA_(Thermococcus_kodakarensis)\n",
    "\n",
    "These seqs from [article](https://academic.oup.com/gbe/article/doi/10.1093/gbe/evab274/6459647)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "7e21603a-f6a1-4f5b-85c0-a6e007d16562",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"WP_011250364.1\", \"CAT72447.1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "644917e1-e47b-415e-8cf3-e76697b2afea",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "9bf927a0-9204-4558-9a13-1342623357f2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 2263\n",
      "Fetched taxid from NCBI 69014\n"
     ]
    }
   ],
   "source": [
    "data_sequence = []\n",
    "for acc in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=acc, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    #     print(record)\n",
    "    # print(record.seq)\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence.append({\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"HTkA_(Thermococcus_kodakarensis)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    })\n",
    "    data_sequence[-1].update(taxonomy_data)\n",
    "    # for k, v in data_sequence[-1].items():\n",
    "    #     print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "2751d01b-91ea-4846-9495-a442f3aa940f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "4e4bb8be-5278-4b2f-9743-22200539164e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>555</th>\n",
       "      <td>CAT72447.1</td>\n",
       "      <td>HTkA_(Thermococcus_kodakarensis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5007</th>\n",
       "      <td>WP_011250364.1</td>\n",
       "      <td>HTkA_(Thermococcus_kodakarensis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2263.0</td>\n",
       "      <td>Thermococcus</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession                           variant    gi ncbi_gene_id  \\\n",
       "555       CAT72447.1  HTkA_(Thermococcus_kodakarensis)  None         None   \n",
       "5007  WP_011250364.1  HTkA_(Thermococcus_kodakarensis)  None         None   \n",
       "\n",
       "     hgnc_gene_name  taxonomy_id                        organism  \\\n",
       "555            None      69014.0  Thermococcus kodakarensis KOD1   \n",
       "5007           None       2263.0                    Thermococcus   \n",
       "\n",
       "                 phylum        class taxonomy_group  info  \\\n",
       "555   Methanobacteriota  Thermococci           None  None   \n",
       "5007  Methanobacteriota  Thermococci           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "555   MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...   \n",
       "5007  MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...   \n",
       "\n",
       "     variant_under_consideration sequence_accession publication_id  \n",
       "555                         None               None           None  \n",
       "5007                        None               None           None  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "3b16ee40-b497-4bdf-ba7c-1a9f70cda865",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>stevens_deep_2022</td>\n",
       "      <td>Deep Conservation of Histone Variants in Therm...</td>\n",
       "      <td>10.1093/gbe/evab274</td>\n",
       "      <td>None</td>\n",
       "      <td>2022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  id                                              title  \\\n",
       "0  stevens_deep_2022  Deep Conservation of Histone Variants in Therm...   \n",
       "\n",
       "                   doi author  year  \n",
       "0  10.1093/gbe/evab274   None  2022  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pid = \"stevens_deep_2022\"\n",
    "query = f\"SELECT * FROM publication WHERE id='{pid}'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "942e6d29-82c8-453c-ade0-5f3664005a26",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "a8af107a-365e-4626-b5f0-d5b45e4a3714",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>555</th>\n",
       "      <td>CAT72447.1</td>\n",
       "      <td>HTkA_(Thermococcus_kodakarensis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...</td>\n",
       "      <td>None</td>\n",
       "      <td>CAT72447.1</td>\n",
       "      <td>stevens_deep_2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5007</th>\n",
       "      <td>WP_011250364.1</td>\n",
       "      <td>HTkA_(Thermococcus_kodakarensis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2263.0</td>\n",
       "      <td>Thermococcus</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...</td>\n",
       "      <td>None</td>\n",
       "      <td>WP_011250364.1</td>\n",
       "      <td>stevens_deep_2022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession                           variant    gi ncbi_gene_id  \\\n",
       "555       CAT72447.1  HTkA_(Thermococcus_kodakarensis)  None         None   \n",
       "5007  WP_011250364.1  HTkA_(Thermococcus_kodakarensis)  None         None   \n",
       "\n",
       "     hgnc_gene_name  taxonomy_id                        organism  \\\n",
       "555            None      69014.0  Thermococcus kodakarensis KOD1   \n",
       "5007           None       2263.0                    Thermococcus   \n",
       "\n",
       "                 phylum        class taxonomy_group  info  \\\n",
       "555   Methanobacteriota  Thermococci           None  None   \n",
       "5007  Methanobacteriota  Thermococci           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "555   MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...   \n",
       "5007  MAELPIAPVDRLIRKAGAERVSEDAAKVLAEYLEEYAIELSKKAVD...   \n",
       "\n",
       "     variant_under_consideration sequence_accession     publication_id  \n",
       "555                         None         CAT72447.1  stevens_deep_2022  \n",
       "5007                        None     WP_011250364.1  stevens_deep_2022  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "092c6b1c-56d0-4747-864d-92e8ba999a48",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d330a43-c95e-4e8e-b138-247442dec53d",
   "metadata": {},
   "source": [
    "# Add WP_011251239.1 and CAT71503.1 as HTkB_(Thermococcus_kodakarensis)\n",
    "\n",
    "These seqs from [article](https://academic.oup.com/gbe/article/doi/10.1093/gbe/evab274/6459647)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "171fb4d2-2ca7-4be3-aa4d-ec24fe544d13",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"WP_011251239.1\", \"CAT71503.1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "07eaf08f-4c44-46a9-9419-7707b5bc5a2f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "24fc3b47-9ef3-42d1-beed-4c73d803f91c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 2263\n",
      "Fetched taxid from NCBI 69014\n"
     ]
    }
   ],
   "source": [
    "data_sequence = []\n",
    "for acc in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=acc, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    #     print(record)\n",
    "    # print(record.seq)\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence.append({\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"HTkB_(Thermococcus_kodakarensis)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    })\n",
    "    data_sequence[-1].update(taxonomy_data)\n",
    "    # for k, v in data_sequence[-1].items():\n",
    "    #     print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "95fc1974-454a-47b9-959b-403654e15fc6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "943da5de-afe5-46c0-b7e3-23dcee50d4a6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>555</th>\n",
       "      <td>CAT71503.1</td>\n",
       "      <td>HTkB_(Thermococcus_kodakarensis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5009</th>\n",
       "      <td>WP_011251239.1</td>\n",
       "      <td>HTkB_(Thermococcus_kodakarensis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2263.0</td>\n",
       "      <td>Thermococcus</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession                           variant    gi ncbi_gene_id  \\\n",
       "555       CAT71503.1  HTkB_(Thermococcus_kodakarensis)  None         None   \n",
       "5009  WP_011251239.1  HTkB_(Thermococcus_kodakarensis)  None         None   \n",
       "\n",
       "     hgnc_gene_name  taxonomy_id                        organism  \\\n",
       "555            None      69014.0  Thermococcus kodakarensis KOD1   \n",
       "5009           None       2263.0                    Thermococcus   \n",
       "\n",
       "                 phylum        class taxonomy_group  info  \\\n",
       "555   Methanobacteriota  Thermococci           None  None   \n",
       "5009  Methanobacteriota  Thermococci           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "555   MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...   \n",
       "5009  MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...   \n",
       "\n",
       "     variant_under_consideration sequence_accession publication_id  \n",
       "555                         None               None           None  \n",
       "5009                        None               None           None  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "a8a1a2a1-5722-46dc-a5c9-609690361f1f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>stevens_deep_2022</td>\n",
       "      <td>Deep Conservation of Histone Variants in Therm...</td>\n",
       "      <td>10.1093/gbe/evab274</td>\n",
       "      <td>None</td>\n",
       "      <td>2022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  id                                              title  \\\n",
       "0  stevens_deep_2022  Deep Conservation of Histone Variants in Therm...   \n",
       "\n",
       "                   doi author  year  \n",
       "0  10.1093/gbe/evab274   None  2022  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pid = \"stevens_deep_2022\"\n",
    "query = f\"SELECT * FROM publication WHERE id='{pid}'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "d3c2d72f-c641-49db-88b9-04bde079a01f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "0276043e-a5f8-40d3-8d5c-80a00d8c9a94",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>555</th>\n",
       "      <td>CAT71503.1</td>\n",
       "      <td>HTkB_(Thermococcus_kodakarensis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>69014.0</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...</td>\n",
       "      <td>None</td>\n",
       "      <td>CAT71503.1</td>\n",
       "      <td>stevens_deep_2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5009</th>\n",
       "      <td>WP_011251239.1</td>\n",
       "      <td>HTkB_(Thermococcus_kodakarensis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2263.0</td>\n",
       "      <td>Thermococcus</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...</td>\n",
       "      <td>None</td>\n",
       "      <td>WP_011251239.1</td>\n",
       "      <td>stevens_deep_2022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession                           variant    gi ncbi_gene_id  \\\n",
       "555       CAT71503.1  HTkB_(Thermococcus_kodakarensis)  None         None   \n",
       "5009  WP_011251239.1  HTkB_(Thermococcus_kodakarensis)  None         None   \n",
       "\n",
       "     hgnc_gene_name  taxonomy_id                        organism  \\\n",
       "555            None      69014.0  Thermococcus kodakarensis KOD1   \n",
       "5009           None       2263.0                    Thermococcus   \n",
       "\n",
       "                 phylum        class taxonomy_group  info  \\\n",
       "555   Methanobacteriota  Thermococci           None  None   \n",
       "5009  Methanobacteriota  Thermococci           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "555   MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...   \n",
       "5009  MAELPIAPVDRLIRKAGAARVSEEAAKVLAEHLEEKALEIAKKAVA...   \n",
       "\n",
       "     variant_under_consideration sequence_accession     publication_id  \n",
       "555                         None         CAT71503.1  stevens_deep_2022  \n",
       "5009                        None     WP_011251239.1  stevens_deep_2022  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "577c32a4-1b97-4b0e-982c-fa8b9a617eae",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "964f18fc-d783-4a01-a81e-4df7245f6cdf",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Add other archaeal sequences as Nucleosomal\n",
    "\n",
    "These seqs from [article](https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1007582)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "aa317e41-81a4-4626-ad3b-f2e232c10f79",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\n",
    "    \"BAI60563\"\n",
    "    \"ADC47610\"\n",
    "    \"WP_02966277\"\n",
    "    \"KYK38613\"\n",
    "    \"WP_048125684\"\n",
    "    \"WP_011973395\"\n",
    "    \"WP_010871171\"\n",
    "    \"KYH40538\"\n",
    "    \"KPV63666\"\n",
    "    \"WP_04104690\"\n",
    "    \"AIE90726\"\n",
    "    \"BAJ48508\"\n",
    "    \"KON33214\"\n",
    "    \"WP_052884954\"\n",
    "    \"WP_012186746\"\n",
    "    \"OLS16336\"\n",
    "    \"OLS15619\"\n",
    "    \"OLS12771\"\n",
    "    \"KKK44894\"\n",
    "    \"KKK41688\"\n",
    "    \"KKK45508\"\n",
    "    \"EGQ43804\"\n",
    "    \"AOV94489\"\n",
    "    \"EHK02195\"\n",
    "    \"OLS26110\"\n",
    "    \"OLS24625\"\n",
    "    \"OLS19133\"\n",
    "    \"OLS18443\"\n",
    "    \"OLS19133\"\n",
    "    \"OLS22331\"\n",
    "    \"AAC72546\"\n",
    "    \"WP_008091782\"\n",
    "    \"WP_015792102\"\n",
    "    \"AAR39136\"\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "9b9ac53b-5b60-4c71-93a1-c76317a088a9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5002</th>\n",
       "      <td>WP_013413995.1</td>\n",
       "      <td>HMfA_(Methanothermus_fervidus)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2180.0</td>\n",
       "      <td>Methanothermus fervidus</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Methanobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGELPIAPIGRIIKNAGAERVSDDARIALAKVLEEMGEEIASEAVK...</td>\n",
       "      <td>None</td>\n",
       "      <td>WP_013413995.1</td>\n",
       "      <td>henneman_structure_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5003</th>\n",
       "      <td>WP_013414263.1</td>\n",
       "      <td>HMfB_(Methanothermus_fervidus)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2180.0</td>\n",
       "      <td>Methanothermus fervidus</td>\n",
       "      <td>Methanobacteriota</td>\n",
       "      <td>Methanobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MELPIAPIGRIIKDAGAERVSDDARITLAKILEEMGRDIASEAIKL...</td>\n",
       "      <td>None</td>\n",
       "      <td>WP_013414263.1</td>\n",
       "      <td>henneman_structure_2018</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession                         variant    gi ncbi_gene_id  \\\n",
       "5002  WP_013413995.1  HMfA_(Methanothermus_fervidus)  None         None   \n",
       "5003  WP_013414263.1  HMfB_(Methanothermus_fervidus)  None         None   \n",
       "\n",
       "     hgnc_gene_name  taxonomy_id                 organism             phylum  \\\n",
       "5002           None       2180.0  Methanothermus fervidus  Methanobacteriota   \n",
       "5003           None       2180.0  Methanothermus fervidus  Methanobacteriota   \n",
       "\n",
       "                class taxonomy_group  info  \\\n",
       "5002  Methanobacteria           None  None   \n",
       "5003  Methanobacteria           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "5002  MGELPIAPIGRIIKNAGAERVSDDARIALAKVLEEMGEEIASEAVK...   \n",
       "5003  MELPIAPIGRIIKDAGAERVSDDARITLAKILEEMGRDIASEAIKL...   \n",
       "\n",
       "     variant_under_consideration sequence_accession           publication_id  \n",
       "5002                        None     WP_013413995.1  henneman_structure_2018  \n",
       "5003                        None     WP_013414263.1  henneman_structure_2018  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3c8a219-f508-46e1-936f-7d6040016e39",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Close connections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "572059d8-54f9-4ec7-a2c6-1c8066552845",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cursor.close()\n",
    "conn.close()\n",
    "tunnel.stop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "790ee020-09af-4dcc-9c36-f1b3f69ea8a7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "@deathbeds/ipydrawio": {
   "xml": ""
  },
  "kernelspec": {
   "display_name": ".conda-histdb_env",
   "language": "python",
   "name": "conda-env-.conda-histdb_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
