{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "53d3ec55-7daf-4ade-8add-2830b709dc54",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import io\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import Entrez, SeqIO\n",
    "from mysql.connector import connection\n",
    "from sshtunnel import SSHTunnelForwarder\n",
    "\n",
    "# Укажите ваш email (обязательно для использования Entrez)\n",
    "Entrez.email = \"your.email@example.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "608c4bfe-b979-4b60-b2b5-b94eeb679cb5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open(\"db_curated_server_info.txt\", \"r\") as file:\n",
    "    lines = file.readlines()\n",
    "\n",
    "config = {}\n",
    "\n",
    "for line in lines:\n",
    "    line = line.strip()\n",
    "    if line and not line.startswith(\"#\"):\n",
    "        key, value = line.split(\"=\", 1)\n",
    "        config[key] = value.strip()\n",
    "\n",
    "server_name = config.get(\"server_name\")\n",
    "srever_port = int(config.get(\"srever_port\"))\n",
    "ssh_password = config.get(\"ssh_password\")\n",
    "ssh_username = config.get(\"ssh_username\")\n",
    "db_adress = config.get(\"db_adress\")\n",
    "db_port = int(config.get(\"db_port\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4e19aa1c-49a3-48c4-b67f-34c94cef608d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "35907\n"
     ]
    }
   ],
   "source": [
    "tunnel = SSHTunnelForwarder(\n",
    "    (server_name, srever_port),\n",
    "    ssh_password=ssh_password,\n",
    "    ssh_username=ssh_username,\n",
    "    remote_bind_address=(db_adress, db_port),\n",
    ")\n",
    "tunnel.start()\n",
    "print(tunnel.local_bind_port)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cef2275b-12d0-48f1-ad3c-ce4cdb9aa267",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "conn = connection.MySQLConnection(\n",
    "    user=\"db_user\",\n",
    "    password=\"db_password\",\n",
    "    host=\"localhost\",\n",
    "    port=tunnel.local_bind_port,\n",
    "    database=\"db_name\",\n",
    ")\n",
    "cursor = conn.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f589ea8d-8b83-4331-a1cf-b64ed7621801",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('alternative_name',),\n",
       " ('histone',),\n",
       " ('histone_description',),\n",
       " ('histone_has_publication',),\n",
       " ('publication',),\n",
       " ('sequence',),\n",
       " ('sequence_has_publication',)]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SHOW TABLES;\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "b3dd9ef8-b162-4127-9535-ac7743af8c69",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# add_histone = (\n",
    "#     \"INSERT INTO histone \"\n",
    "#     \"(id, level, taxonomic_span, taxonomic_span_id, description, parent) \"\n",
    "#     \"VALUES (%(id)s, %(level)s, %(taxonomic_span)s, %(taxonomic_span_id)s, %(description)s, %(parent)s)\"\n",
    "# )\n",
    "# add_histone_description = (\n",
    "#     \"INSERT INTO histone_description \"\n",
    "#     \"(summary, taxonomy, genes, evolution, expression, knock_out, function, sequence, localization, deposition, structure, interactions, disease, caveats) \"\n",
    "#     \"VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)\"\n",
    "# )\n",
    "add_publication = (\n",
    "    \"INSERT INTO publication \"\n",
    "    \"(id, title, doi, author, year) \"\n",
    "    \"VALUES (%(id)s, %(title)s, %(doi)s, %(author)s, %(year)s)\"\n",
    ")\n",
    "add_sequence = (\n",
    "    \"INSERT INTO sequence \"\n",
    "    \"(accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration) \"\n",
    "    \"VALUES (%(accession)s, %(variant)s, %(gi)s, %(ncbi_gene_id)s, %(hgnc_gene_name)s, %(taxonomy_id)s, %(organism)s, %(phylum)s, %(class)s, %(taxonomy_group)s, %(info)s, %(sequence)s, %(variant_under_consideration)s)\"\n",
    ")\n",
    "add_sequence_has_publication = (\n",
    "    \"INSERT INTO sequence_has_publication \"\n",
    "    \"(sequence_accession, publication_id) \"\n",
    "    \"VALUES (%s, %s)\"\n",
    ")\n",
    "# add_alternate_names = (\n",
    "#     \"INSERT INTO alternative_name \"\n",
    "#     \"(name, taxonomy, gene, splice, histone) \"\n",
    "#     \"VALUES (%(name)s, %(taxonomy)s, %(gene)s, %(splice)s, %(histone)s)\"\n",
    "# )\n",
    "# add_histone_has_publication = (\n",
    "#     \"INSERT INTO histone_has_publication \"\n",
    "#     \"(histone_id, publication_id) \"\n",
    "#     \"VALUES (%s, %s)\"\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8bdab3f0-dacc-46a9-8f33-3940e8c671b1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_taxonomy_data(record):\n",
    "    import re\n",
    "    import sys\n",
    "\n",
    "    taxonomy_data = {}\n",
    "    taxonomy_data[\"organism\"] = record.annotations[\"organism\"]\n",
    "    try:\n",
    "        for a in record.features[0].qualifiers[\"db_xref\"]:\n",
    "            text = re.search(\"(\\S+):(\\S+)\", a).group(1)\n",
    "            taxid = re.search(\"(\\S+):(\\S+)\", a).group(2)\n",
    "            if text == \"taxon\":\n",
    "                print(f\"Fetched taxid from NCBI {taxid}\")\n",
    "                taxonomy_data[\"taxonomy_id\"] = int(taxid)\n",
    "            else:\n",
    "                continue\n",
    "    except:\n",
    "        print(\"!!!!!!Unable to get TAXID for this record setting it to 1\")\n",
    "        taxonomy_data[\"taxonomy_id\"] = 1  # unable to identify\n",
    "\n",
    "    lineage = dict()\n",
    "    for i in range(10):\n",
    "        try:\n",
    "            handle = Entrez.efetch(\n",
    "                id=taxonomy_data[\"taxonomy_id\"], db=\"taxonomy\", retmode=\"xml\"\n",
    "            )\n",
    "            tax_data = Entrez.read(handle)\n",
    "            lineage = {\n",
    "                d[\"Rank\"]: d[\"ScientificName\"]\n",
    "                for d in tax_data[0][\"LineageEx\"]\n",
    "                if d[\"Rank\"] in [\"class\", \"phylum\"]\n",
    "            }\n",
    "            break\n",
    "        except:\n",
    "            print(\n",
    "                \"Unexpected error: {}, Retrying, attempt {}\".format(\n",
    "                    sys.exc_info()[0], i\n",
    "                )\n",
    "            )\n",
    "            if i == 9:\n",
    "                print(\n",
    "                    f\"FATAL ERROR could not get class and phylum from NCBI after 10 attempts for taxid:{taxonomy_data['taxonomy_id']}. Will add None for class and phylum!\"\n",
    "                )\n",
    "            else:\n",
    "                continue\n",
    "    taxonomy_data[\"phylum\"] = lineage.get(\"phylum\", None)\n",
    "    taxonomy_data[\"class\"] = lineage.get(\"class\", None)\n",
    "    if taxonomy_data[\"phylum\"] is not None:\n",
    "        taxonomy_data[\"phylum\"] = str(taxonomy_data[\"phylum\"])\n",
    "    if taxonomy_data[\"class\"] is not None:\n",
    "        taxonomy_data[\"class\"] = str(taxonomy_data[\"class\"])\n",
    "    return taxonomy_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c060c047-3e7d-4095-b608-9722031af692",
   "metadata": {},
   "source": [
    "# To Do\n",
    "\n",
    "## <span style=\"color:green\">Add sequences from `neumann_centromeres_2015`</span>\n",
    "\n",
    "Accessions (CDS) from Supplementary Table S2:\n",
    "AB649144.1\n",
    "KM822210\n",
    "XM_003637685\n",
    "KM822207\n",
    "KM822209\n",
    "KM822208\n",
    "KM822212\n",
    "KM822213\n",
    "KM822214\n",
    "KM822215\n",
    "KM822216\n",
    "KM822217\n",
    "KM822218\n",
    "KM822211\n",
    "extracted from JF739989\n",
    "KM822220\n",
    "KM822219\n",
    "KM822221\n",
    "KM822222\n",
    "KM822223\n",
    "KM822224\n",
    "KM822225\n",
    "GT621186.1 KM822227\n",
    "KM822226\n",
    "extracted from JF739990\n",
    "KM822228\n",
    "KM822229\n",
    "KM822231\n",
    "KM822232\n",
    "KM822233\n",
    "KM822234\n",
    "KM822235\n",
    "KM822230\n",
    "\n",
    "## <span style=\"color:green\">Add sequences from `sanei_loss_2011`, `karimi-ashtiyani_point_2015` and `ishii_differential_2015`</span>\n",
    "Protein Accessions: AEK21394.1\n",
    "AEK21393.1\n",
    "AEK21392.1\n",
    "ADB03182.1\n",
    "\n",
    "## <span style=\"color:green\">Add sequences from `neumann_stretching_2012`</span>\n",
    "Protein Accessions: AEX31246.1\n",
    "AEX31245.1\n",
    "\n",
    "## <span style=\"color:green\">Add sequences from `kawabe_duplication_2006`</span>\n",
    "Protein Accessions:\n",
    "ABE27630.1\n",
    "ABE27629.1\n",
    "ABE27628.1\n",
    "ABE27627.1\n",
    "ABE27626.1\n",
    "ABE27625.1\n",
    "ABE27624.1\n",
    "ABE27623.1\n",
    "ABE27622.1\n",
    "ABE27621.1\n",
    "ABE27620.1\n",
    "ABE27619.1\n",
    "ABE27618.1\n",
    "ABE27617.1\n",
    "ABE27616.1\n",
    "ABE27615.1\n",
    "ABE27614.1\n",
    "ABE27613.1\n",
    "ABE27612.1\n",
    "ABE27611.1\n",
    "ABE27610.1\n",
    "ABE27609.1\n",
    "ABE27608.1\n",
    "ABE27607.1\n",
    "ABE27606.1\n",
    "ABE27605.1\n",
    "ABE27604.1\n",
    "ABE27603.1\n",
    "ABE27602.1\n",
    "ABE27601.1\n",
    "ABE27600.1\n",
    "ABE27662.1\n",
    "ABE27661.1\n",
    "ABE27660.1\n",
    "ABE27659.1\n",
    "ABE27658.1\n",
    "ABE27657.1\n",
    "ABE27656.1\n",
    "ABE27655.1\n",
    "ABE27654.1\n",
    "ABE27653.1\n",
    "ABE27652.1\n",
    "ABE27651.1\n",
    "ABE27650.1\n",
    "ABE27649.1\n",
    "ABE27648.1\n",
    "ABE27647.1\n",
    "ABE27646.1\n",
    "ABE27645.1\n",
    "ABE27644.1\n",
    "ABE27643.1\n",
    "ABE27642.1\n",
    "ABE27641.1\n",
    "ABE27640.1\n",
    "ABE27639.1\n",
    "ABE27638.1\n",
    "ABE27637.1\n",
    "ABE27636.1\n",
    "ABE27635.1\n",
    "ABE27634.1\n",
    "ABE27633.1\n",
    "ABE27632.1\n",
    "ABE27631.1\n",
    "BAC79432.1\n",
    "BAC79431.1\n",
    "BAC79430.1\n",
    "BAC79429.1\n",
    "BAC79428.1\n",
    "BAC79427.1\n",
    "    \n",
    "## <span style=\"color:green\">Add sequences from `evtushenko_conserved_2017`</span>\n",
    "\n",
    "Protein Accessions: AUN88474.1\n",
    "AUN88473.1\n",
    "AUN88472.1\n",
    "AUN88471.1\n",
    "AUN88470.1\n",
    "AUN88469.1\n",
    "AUN88468.1\n",
    "AUN88467.1\n",
    "AUN88466.1\n",
    "AUN88465.1\n",
    "AUN88464.1\n",
    "AUN88463.1\n",
    "AUN88462.1\n",
    "AUN88461.1\n",
    "AUN88460.1\n",
    "AUN88459.1\n",
    "AUN88458.1\n",
    "AUN88457.1\n",
    "AUN88456.1\n",
    "AUN88455.1\n",
    "AUN88454.1\n",
    "AUN88453.1\n",
    "AUN88452.1\n",
    "AUN88451.1\n",
    "AUN88450.1\n",
    "AUN88449.1\n",
    "    \n",
    "## <span style=\"color:black\">Add sequences from `monen_differential_2005` and `monen_separase_2015`</span>\n",
    "\n",
    "Protein Accessions: NP_499128.1 (hcp-3) NP_499073.1 (cpar-1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a5dfb39-7179-495b-bd4d-dee63217777d",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add sequences from `neumann_centromeres_2015`</span>\n",
    "\n",
    "Accessions (CDS) from Supplementary Table S2:\n",
    "AB649144.1\n",
    "KM822210\n",
    "XM_003637685\n",
    "KM822207\n",
    "KM822209\n",
    "KM822208\n",
    "KM822212\n",
    "KM822213\n",
    "KM822214\n",
    "KM822215\n",
    "KM822216\n",
    "KM822217\n",
    "KM822218\n",
    "KM822211\n",
    "extracted from JF739989\n",
    "KM822220\n",
    "KM822219\n",
    "KM822221\n",
    "KM822222\n",
    "KM822223\n",
    "KM822224\n",
    "KM822225\n",
    "GT621186.1 KM822227\n",
    "KM822226\n",
    "extracted from JF739990\n",
    "KM822228\n",
    "KM822229\n",
    "KM822231\n",
    "KM822232\n",
    "KM822233\n",
    "KM822234\n",
    "KM822235\n",
    "KM822230"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d2a4e6d9-1263-40ff-8f06-0e90cbb37bd0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cds_accessions = [\n",
    "    \"AB649144.1\",\n",
    "    \"KM822210\",\n",
    "    \"XM_003637685\",\n",
    "    \"KM822207\",\n",
    "    \"KM822209\",\n",
    "    \"KM822208\",\n",
    "    \"KM822212\",\n",
    "    \"KM822213\",\n",
    "    \"KM822214\",\n",
    "    \"KM822215\",\n",
    "    \"KM822216\",\n",
    "    \"KM822217\",\n",
    "    \"KM822218\",\n",
    "    \"KM822211\",\n",
    "    \"JF739989\",\n",
    "    \"KM822220\",\n",
    "    \"KM822219\",\n",
    "    \"KM822221\",\n",
    "    \"KM822222\",\n",
    "    \"KM822223\",\n",
    "    \"KM822224\",\n",
    "    \"KM822225\",\n",
    "    \"KM822227\",\n",
    "    \"KM822226\",\n",
    "    \"JF739990\",\n",
    "    \"KM822228\",\n",
    "    \"KM822229\",\n",
    "    \"KM822231\",\n",
    "    \"KM822232\",\n",
    "    \"KM822233\",\n",
    "    \"KM822234\",\n",
    "    \"KM822235\",\n",
    "    \"KM822230\",\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d4f7610-213c-480f-88e7-d7b2965d4fa0",
   "metadata": {},
   "source": [
    "## Get protein IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "550415f8-6042-45c8-a337-d5df3b9f181c",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### Search for AB649144.1 ...\n",
      "### Search for KM822210 ...\n",
      "### Search for XM_003637685 ...\n",
      "### Search for KM822207 ...\n",
      "### Search for KM822209 ...\n",
      "### Search for KM822208 ...\n",
      "### Search for KM822212 ...\n",
      "ID: KM822212.1\n",
      "Описание: Lathyrus clymenum isolate CenH3-1_LAC centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94118.1\n",
      "### Search for KM822213 ...\n",
      "ID: KM822213.1\n",
      "Описание: Lathyrus latifolius isolate CenH3-1_LAL centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94119.1\n",
      "### Search for KM822214 ...\n",
      "ID: KM822214.1\n",
      "Описание: Lathyrus niger isolate CenH3-1_LAN centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94120.1\n",
      "### Search for KM822215 ...\n",
      "ID: KM822215.1\n",
      "Описание: Lathyrus ochrus isolate CenH3-1_LAO centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94121.1\n",
      "### Search for KM822216 ...\n",
      "ID: KM822216.1\n",
      "Описание: Lathyrus sativus isolate CenH3-1_LAS centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94122.1\n",
      "### Search for KM822217 ...\n",
      "ID: KM822217.1\n",
      "Описание: Lathyrus sylvestris isolate CenH3-1_LASyl centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94123.1\n",
      "### Search for KM822218 ...\n",
      "ID: KM822218.1\n",
      "Описание: Lathyrus vernus isolate CenH3-1_LAV centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94124.1\n",
      "### Search for KM822211 ...\n",
      "ID: KM822211.1\n",
      "Описание: Pisum fulvum isolate CenH3-1_PFul centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94117.1\n",
      "### Search for JF739989 ...\n",
      "ID: JF739989.1\n",
      "Описание: Pisum sativum centromere-specific variant of histone H3 type 1 gene, complete cds\n",
      "Найденные идентификаторы белков: AEX31245.1\n",
      "### Search for KM822220 ...\n",
      "ID: KM822220.1\n",
      "Описание: Lathyrus clymenum isolate CenH3-2_LAC centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94126.1\n",
      "### Search for KM822219 ...\n",
      "ID: KM822219.1\n",
      "Описание: Lathyrus latifolius isolate CenH3-2_LAL centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94125.1\n",
      "### Search for KM822221 ...\n",
      "ID: KM822221.1\n",
      "Описание: Lathyrus niger isolate CenH3-2_LAN centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94127.1\n",
      "### Search for KM822222 ...\n",
      "ID: KM822222.1\n",
      "Описание: Lathyrus ochrus isolate CenH3-2_LAO centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94128.1\n",
      "### Search for KM822223 ...\n",
      "ID: KM822223.1\n",
      "Описание: Lathyrus sativus isolate CenH3-2_LAS centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94129.1\n",
      "### Search for KM822224 ...\n",
      "ID: KM822224.1\n",
      "Описание: Lathyrus sylvestris isolate CenH3-2_LASyl centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94130.1\n",
      "### Search for KM822225 ...\n",
      "ID: KM822225.1\n",
      "Описание: Lathyrus vernus isolate CenH3-2_LAV centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94131.1\n",
      "### Search for KM822227 ...\n",
      "ID: KM822227.1\n",
      "Описание: Lens culinaris isolate CenH3-2_LCul centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94133.1\n",
      "### Search for KM822226 ...\n",
      "ID: KM822226.1\n",
      "Описание: Pisum fulvum isolate CenH3-2_PFul centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94132.1\n",
      "### Search for JF739990 ...\n",
      "ID: JF739990.1\n",
      "Описание: Pisum sativum centromere-specific variant of histone H3 type 2 gene, complete cds\n",
      "Найденные идентификаторы белков: AEX31246.1\n",
      "### Search for KM822228 ...\n",
      "ID: KM822228.1\n",
      "Описание: Vicia faba isolate CenH3-2_VF centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94134.1\n",
      "### Search for KM822229 ...\n",
      "ID: KM822229.1\n",
      "Описание: Vicia lathyroides isolate CenH3-2_VL centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94135.1\n",
      "### Search for KM822231 ...\n",
      "ID: KM822231.1\n",
      "Описание: Vicia pannonica isolate CenH3-2_VPa centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94137.1\n",
      "### Search for KM822232 ...\n",
      "ID: KM822232.1\n",
      "Описание: Vicia peregrina isolate CenH3-2_VPe centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94138.1\n",
      "### Search for KM822233 ...\n",
      "ID: KM822233.1\n",
      "Описание: Vicia sativa isolate CenH3-2_VSa centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94139.1\n",
      "### Search for KM822234 ...\n",
      "ID: KM822234.1\n",
      "Описание: Vicia sepium isolate CenH3-2_VSe centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94140.1\n",
      "### Search for KM822235 ...\n",
      "ID: KM822235.1\n",
      "Описание: Vicia villosa isolate CenH3-2_VV centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94141.1\n",
      "### Search for KM822230 ...\n",
      "ID: KM822230.1\n",
      "Описание: Vicia narbonensis isolate CenH3-2_VN centromere-specific variant of histone H3 mRNA, complete cds\n",
      "Найденные идентификаторы белков: AKA94136.1\n"
     ]
    }
   ],
   "source": [
    "records_data = {}\n",
    "other_records = {}\n",
    "for accession_id in cds_accessions:\n",
    "    # ID записи в NCBI nuccore\n",
    "    print(f\"### Search for {accession_id} ...\")\n",
    "    # Загружаем запись в формате GenBank\n",
    "    with Entrez.efetch(\n",
    "        db=\"nucleotide\", id=accession_id, rettype=\"gb\", retmode=\"text\"\n",
    "    ) as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    if not (\n",
    "        record.annotations[\"organism\"].startswith(\n",
    "            (\"Pisum\", \"Lathyrus\", \"Vicia\", \"Lens\")\n",
    "        )\n",
    "    ):\n",
    "        other_records[accession_id] = {\n",
    "            \"ID\": record.id,\n",
    "            \"Description\": record.description,\n",
    "            \"Organism\": record.annotations[\"organism\"],\n",
    "        }\n",
    "        continue\n",
    "    # Выводим информацию о записи\n",
    "    print(f\"ID: {record.id}\")\n",
    "    print(f\"Описание: {record.description}\")\n",
    "    # Извлекаем идентификаторы белков\n",
    "    protein_ids = []\n",
    "    for feature in record.features:\n",
    "        if feature.type == \"CDS\":  # Ищем только кодирующие последовательности (белки)\n",
    "            if \"protein_id\" in feature.qualifiers:\n",
    "                protein_ids.append(feature.qualifiers[\"protein_id\"][0])\n",
    "            elif (\n",
    "                \"db_xref\" in feature.qualifiers\n",
    "            ):  # Ищем идентификаторы в db_xref (например, UniProt)\n",
    "                for xref in feature.qualifiers[\"db_xref\"]:\n",
    "                    if xref.startswith(\"GI:\") or xref.startswith(\n",
    "                        \"UniProt:\"\n",
    "                    ):  # Можно добавить другие базы\n",
    "                        protein_ids.append(xref)\n",
    "    # Выводим результат\n",
    "    print(\"Найденные идентификаторы белков:\", end=\" \")\n",
    "    for pid in protein_ids:\n",
    "        print(pid)\n",
    "    records_data[accession_id] = {\n",
    "        \"ID\": record.id,\n",
    "        \"Description\": record.description,\n",
    "        \"Organism\": record.annotations[\"organism\"],\n",
    "        \"Protein ID\": protein_ids[0],\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5aba848f-adf0-4363-a883-c8abc4f0525a",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'KM822212': {'ID': 'KM822212.1',\n",
       "  'Description': 'Lathyrus clymenum isolate CenH3-1_LAC centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Lathyrus clymenum',\n",
       "  'Protein ID': 'AKA94118.1'},\n",
       " 'KM822213': {'ID': 'KM822213.1',\n",
       "  'Description': 'Lathyrus latifolius isolate CenH3-1_LAL centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Lathyrus latifolius',\n",
       "  'Protein ID': 'AKA94119.1'},\n",
       " 'KM822214': {'ID': 'KM822214.1',\n",
       "  'Description': 'Lathyrus niger isolate CenH3-1_LAN centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Lathyrus niger',\n",
       "  'Protein ID': 'AKA94120.1'},\n",
       " 'KM822215': {'ID': 'KM822215.1',\n",
       "  'Description': 'Lathyrus ochrus isolate CenH3-1_LAO centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Lathyrus ochrus',\n",
       "  'Protein ID': 'AKA94121.1'},\n",
       " 'KM822216': {'ID': 'KM822216.1',\n",
       "  'Description': 'Lathyrus sativus isolate CenH3-1_LAS centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Lathyrus sativus',\n",
       "  'Protein ID': 'AKA94122.1'},\n",
       " 'KM822217': {'ID': 'KM822217.1',\n",
       "  'Description': 'Lathyrus sylvestris isolate CenH3-1_LASyl centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Lathyrus sylvestris',\n",
       "  'Protein ID': 'AKA94123.1'},\n",
       " 'KM822218': {'ID': 'KM822218.1',\n",
       "  'Description': 'Lathyrus vernus isolate CenH3-1_LAV centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Lathyrus vernus',\n",
       "  'Protein ID': 'AKA94124.1'},\n",
       " 'KM822211': {'ID': 'KM822211.1',\n",
       "  'Description': 'Pisum fulvum isolate CenH3-1_PFul centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Pisum fulvum',\n",
       "  'Protein ID': 'AKA94117.1'},\n",
       " 'JF739989': {'ID': 'JF739989.1',\n",
       "  'Description': 'Pisum sativum centromere-specific variant of histone H3 type 1 gene, complete cds',\n",
       "  'Organism': 'Pisum sativum',\n",
       "  'Protein ID': 'AEX31245.1'},\n",
       " 'KM822220': {'ID': 'KM822220.1',\n",
       "  'Description': 'Lathyrus clymenum isolate CenH3-2_LAC centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Lathyrus clymenum',\n",
       "  'Protein ID': 'AKA94126.1'},\n",
       " 'KM822219': {'ID': 'KM822219.1',\n",
       "  'Description': 'Lathyrus latifolius isolate CenH3-2_LAL centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Lathyrus latifolius',\n",
       "  'Protein ID': 'AKA94125.1'},\n",
       " 'KM822221': {'ID': 'KM822221.1',\n",
       "  'Description': 'Lathyrus niger isolate CenH3-2_LAN centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Lathyrus niger',\n",
       "  'Protein ID': 'AKA94127.1'},\n",
       " 'KM822222': {'ID': 'KM822222.1',\n",
       "  'Description': 'Lathyrus ochrus isolate CenH3-2_LAO centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Lathyrus ochrus',\n",
       "  'Protein ID': 'AKA94128.1'},\n",
       " 'KM822223': {'ID': 'KM822223.1',\n",
       "  'Description': 'Lathyrus sativus isolate CenH3-2_LAS centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Lathyrus sativus',\n",
       "  'Protein ID': 'AKA94129.1'},\n",
       " 'KM822224': {'ID': 'KM822224.1',\n",
       "  'Description': 'Lathyrus sylvestris isolate CenH3-2_LASyl centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Lathyrus sylvestris',\n",
       "  'Protein ID': 'AKA94130.1'},\n",
       " 'KM822225': {'ID': 'KM822225.1',\n",
       "  'Description': 'Lathyrus vernus isolate CenH3-2_LAV centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Lathyrus vernus',\n",
       "  'Protein ID': 'AKA94131.1'},\n",
       " 'KM822227': {'ID': 'KM822227.1',\n",
       "  'Description': 'Lens culinaris isolate CenH3-2_LCul centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Lens culinaris',\n",
       "  'Protein ID': 'AKA94133.1'},\n",
       " 'KM822226': {'ID': 'KM822226.1',\n",
       "  'Description': 'Pisum fulvum isolate CenH3-2_PFul centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Pisum fulvum',\n",
       "  'Protein ID': 'AKA94132.1'},\n",
       " 'JF739990': {'ID': 'JF739990.1',\n",
       "  'Description': 'Pisum sativum centromere-specific variant of histone H3 type 2 gene, complete cds',\n",
       "  'Organism': 'Pisum sativum',\n",
       "  'Protein ID': 'AEX31246.1'},\n",
       " 'KM822228': {'ID': 'KM822228.1',\n",
       "  'Description': 'Vicia faba isolate CenH3-2_VF centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Vicia faba',\n",
       "  'Protein ID': 'AKA94134.1'},\n",
       " 'KM822229': {'ID': 'KM822229.1',\n",
       "  'Description': 'Vicia lathyroides isolate CenH3-2_VL centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Vicia lathyroides',\n",
       "  'Protein ID': 'AKA94135.1'},\n",
       " 'KM822231': {'ID': 'KM822231.1',\n",
       "  'Description': 'Vicia pannonica isolate CenH3-2_VPa centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Vicia pannonica',\n",
       "  'Protein ID': 'AKA94137.1'},\n",
       " 'KM822232': {'ID': 'KM822232.1',\n",
       "  'Description': 'Vicia peregrina isolate CenH3-2_VPe centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Vicia peregrina',\n",
       "  'Protein ID': 'AKA94138.1'},\n",
       " 'KM822233': {'ID': 'KM822233.1',\n",
       "  'Description': 'Vicia sativa isolate CenH3-2_VSa centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Vicia sativa',\n",
       "  'Protein ID': 'AKA94139.1'},\n",
       " 'KM822234': {'ID': 'KM822234.1',\n",
       "  'Description': 'Vicia sepium isolate CenH3-2_VSe centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Vicia sepium',\n",
       "  'Protein ID': 'AKA94140.1'},\n",
       " 'KM822235': {'ID': 'KM822235.1',\n",
       "  'Description': 'Vicia villosa isolate CenH3-2_VV centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Vicia villosa',\n",
       "  'Protein ID': 'AKA94141.1'},\n",
       " 'KM822230': {'ID': 'KM822230.1',\n",
       "  'Description': 'Vicia narbonensis isolate CenH3-2_VN centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Vicia narbonensis',\n",
       "  'Protein ID': 'AKA94136.1'}}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "records_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "146ccbca-ed8b-4299-aadd-01d7afa76ea3",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AB649144.1': {'ID': 'AB649144.1',\n",
       "  'Description': 'Astragalus sinicus AsCENH3 mRNA for centromere specific histone H3 variant, complete cds',\n",
       "  'Organism': 'Astragalus sinicus'},\n",
       " 'KM822210': {'ID': 'KM822210.1',\n",
       "  'Description': 'Cicer arietinum isolate CenH3_CicA centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Cicer arietinum'},\n",
       " 'XM_003637685': {'ID': 'XM_003637685.1',\n",
       "  'Description': 'Medicago truncatula Histone H3 (MTR_100s0023) mRNA, complete cds',\n",
       "  'Organism': 'Medicago truncatula'},\n",
       " 'KM822207': {'ID': 'KM822207.1',\n",
       "  'Description': 'Melilotus albus isolate CenH3_MelA centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Melilotus albus'},\n",
       " 'KM822209': {'ID': 'KM822209.1',\n",
       "  'Description': 'Trifolium pratense isolate CenH3_TriP centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Trifolium pratense'},\n",
       " 'KM822208': {'ID': 'KM822208.1',\n",
       "  'Description': 'Trigonella foenum-graecum isolate CenH3_TFG centromere-specific variant of histone H3 mRNA, complete cds',\n",
       "  'Organism': 'Trigonella foenum-graecum'}}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "other_records"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ceca25b8-320e-4a5a-a61e-58d3a1558ba1",
   "metadata": {},
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "71b7805d-a2a1-49b7-985d-9b5fdf9f7c69",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin([v[\"Protein ID\"] for k, v in records_data.items()])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "fbf13440-3e02-4d55-8b98-bfa66a603cd4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MGRVKQFPPSKFAARNNHEKKKRRVKPGTVALREIRKFQKDVKLLIPYAPFVRCVREITTQFSSLVTRWTPEALISLQEAAEDDLIRMFEAGMLCAIHARRITLIKKDIELTRRLTGIGRLR'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with Entrez.efetch(\n",
    "    db=\"protein\", id=\"AKA94118.1\", rettype=\"gb\", retmode=\"text\"\n",
    ") as handle:\n",
    "    record = SeqIO.read(handle, \"genbank\")\n",
    "str(record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5dd3fb00-e992-47df-8dd8-63bca4832bf7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 3855\n",
      "Fetched taxid from NCBI 154494\n",
      "Fetched taxid from NCBI 313121\n",
      "Fetched taxid from NCBI 3858\n",
      "Fetched taxid from NCBI 3860\n",
      "Fetched taxid from NCBI 313117\n",
      "Fetched taxid from NCBI 313120\n",
      "Fetched taxid from NCBI 51020\n",
      "Fetched taxid from NCBI 3888\n",
      "Fetched taxid from NCBI 3855\n",
      "Fetched taxid from NCBI 154494\n",
      "Fetched taxid from NCBI 313121\n",
      "Fetched taxid from NCBI 3858\n",
      "Fetched taxid from NCBI 3860\n",
      "Fetched taxid from NCBI 313117\n",
      "Fetched taxid from NCBI 313120\n",
      "Fetched taxid from NCBI 3864\n",
      "Fetched taxid from NCBI 51020\n",
      "Fetched taxid from NCBI 3888\n",
      "Fetched taxid from NCBI 3906\n",
      "Fetched taxid from NCBI 347192\n",
      "Fetched taxid from NCBI 29753\n",
      "Fetched taxid from NCBI 233247\n",
      "Fetched taxid from NCBI 3908\n",
      "Fetched taxid from NCBI 347188\n",
      "Fetched taxid from NCBI 3911\n",
      "Fetched taxid from NCBI 3912\n",
      "accession AKA94136.1 <class 'str'>\n",
      "variant cenH3_(Plants) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 3912 <class 'int'>\n",
      "organism Vicia narbonensis <class 'str'>\n",
      "phylum Streptophyta <class 'str'>\n",
      "class Magnoliopsida <class 'str'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MARVKQTPRPAREIQERKKRRNKPGTVALREIQKMQKTFKLIIPYAPFVRVVREITNQVSSSVTRWTPGALTSLQEAAEDCLVRMFEAGWLCALHAKRVTLMKKDIELTRRLTGIGRPW <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for k, v in records_data.items():\n",
    "    with Entrez.efetch(\n",
    "        db=\"protein\", id=v[\"Protein ID\"], rettype=\"gb\", retmode=\"text\"\n",
    "    ) as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"cenH3_(Plants)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "23588d89-91bb-4371-b9bf-baf4e1a7a515",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "59748a79-135b-4774-bda5-c1caa4103b9a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>AEX31245.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3888.0</td>\n",
       "      <td>Pisum sativum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKHFPSPSKPAASDNLGKKKRRCKPGTKALREIRKFQKDVKLL...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>AEX31246.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3888.0</td>\n",
       "      <td>Pisum sativum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARENQERKKRRNKPGTVALREIKKLQKTFQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302</th>\n",
       "      <td>AKA94117.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>51020.0</td>\n",
       "      <td>Pisum fulvum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKHFPSPSKPAASDNLGKKKRRCKPGTKALREIRKFQKDVKLL...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303</th>\n",
       "      <td>AKA94118.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3855.0</td>\n",
       "      <td>Lathyrus clymenum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKQFPPSKFAARNNHEKKKRRVKPGTVALREIRKFQKDVKLLI...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304</th>\n",
       "      <td>AKA94119.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>154494.0</td>\n",
       "      <td>Lathyrus latifolius</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKHFPRPSKPAASNNHEKKKRRSKPGTKAVREIRKFQKDVKLL...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>AKA94120.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>313121.0</td>\n",
       "      <td>Lathyrus niger</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKHFPRPSKSAASDNHGKKKRRSKPGTKALREIRKFQKDVKLL...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>AKA94121.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3858.0</td>\n",
       "      <td>Lathyrus ochrus</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKHFPRPSKSAANNTHEKKKRRFKRGTALQEIRKFQKDVKLLI...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>307</th>\n",
       "      <td>AKA94122.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3860.0</td>\n",
       "      <td>Lathyrus sativus</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKHFPRPSKPAASDNQEKRKRRSKPGTKAVREIRKFQKDVKLL...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308</th>\n",
       "      <td>AKA94123.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>313117.0</td>\n",
       "      <td>Lathyrus sylvestris</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKHFPRPSKPAASNNHEKKKRRSKPGTKAVREIRKFQKDVKLL...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>AKA94124.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>313120.0</td>\n",
       "      <td>Lathyrus vernus</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKHFPRPSKPAASDNHGKKKRLSKPGTKALREIRKFQNDVKLL...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>AKA94125.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>154494.0</td>\n",
       "      <td>Lathyrus latifolius</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQKPRHARENQERKKRRNKPGTVALREIRKLQKTFQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>AKA94126.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3855.0</td>\n",
       "      <td>Lathyrus clymenum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQKPRHARENQERKKRRNKPGTVALREIRKLQKTCQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>AKA94127.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>313121.0</td>\n",
       "      <td>Lathyrus niger</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARENQERKKRRNKPGTVALREIRKLQKTFQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>AKA94128.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3858.0</td>\n",
       "      <td>Lathyrus ochrus</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKQRPRHARENQEKKKRRNKPGTVALREIRKFQKTCQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314</th>\n",
       "      <td>AKA94129.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3860.0</td>\n",
       "      <td>Lathyrus sativus</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARENQERKKRRNKPGTVALREIRKFQKTFKLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>AKA94130.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>313117.0</td>\n",
       "      <td>Lathyrus sylvestris</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQKPRHARENQERKKRRNKPGTVALREIRKLQKTFQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>AKA94131.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>313120.0</td>\n",
       "      <td>Lathyrus vernus</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARENQERKKRRNKPGTVALREIRKFQKTFQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>AKA94132.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>51020.0</td>\n",
       "      <td>Pisum fulvum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARENQERKKRRNKPGTVALREIKKLQKTFQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>AKA94133.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3864.0</td>\n",
       "      <td>Lens culinaris</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRPVLQNQERKKRRNKPGTVALREIRKLQKTFQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>AKA94134.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3906.0</td>\n",
       "      <td>Vicia faba</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARETQEKKKRRNKPGTVALREIRKMQKTFKLIIPYA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>AKA94135.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>347192.0</td>\n",
       "      <td>Vicia lathyroides</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARETQEKKKRRNKPGTVALREIRKLQKTFKLIIPYA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>AKA94136.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3912.0</td>\n",
       "      <td>Vicia narbonensis</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRPAREIQERKKRRNKPGTVALREIQKMQKTFKLIIPYA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>AKA94137.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>29753.0</td>\n",
       "      <td>Vicia pannonica</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARENQAKKKRRNKPGTVALREIRKLQKAVKLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>AKA94138.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>233247.0</td>\n",
       "      <td>Vicia peregrina</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRNARENQERKKRRNKPGTVVLHEIRKWQKAVKLIIPYA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>324</th>\n",
       "      <td>AKA94139.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3908.0</td>\n",
       "      <td>Vicia sativa</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARETQEKKKRRNKPGTVALREIRKMQKTFKLIIPYA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325</th>\n",
       "      <td>AKA94140.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>347188.0</td>\n",
       "      <td>Vicia sepium</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARETQEKKKRRNKPGTVALREIRKLQKTFKLIIPYA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326</th>\n",
       "      <td>AKA94141.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3911.0</td>\n",
       "      <td>Vicia villosa</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQKPRHARVDQEKKKRRNKPGTVALREIRKLQKSVKLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession         variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "207  AEX31245.1  cenH3_(Plants)  None         None           None   \n",
       "208  AEX31246.1  cenH3_(Plants)  None         None           None   \n",
       "302  AKA94117.1  cenH3_(Plants)  None         None           None   \n",
       "303  AKA94118.1  cenH3_(Plants)  None         None           None   \n",
       "304  AKA94119.1  cenH3_(Plants)  None         None           None   \n",
       "305  AKA94120.1  cenH3_(Plants)  None         None           None   \n",
       "306  AKA94121.1  cenH3_(Plants)  None         None           None   \n",
       "307  AKA94122.1  cenH3_(Plants)  None         None           None   \n",
       "308  AKA94123.1  cenH3_(Plants)  None         None           None   \n",
       "309  AKA94124.1  cenH3_(Plants)  None         None           None   \n",
       "310  AKA94125.1  cenH3_(Plants)  None         None           None   \n",
       "311  AKA94126.1  cenH3_(Plants)  None         None           None   \n",
       "312  AKA94127.1  cenH3_(Plants)  None         None           None   \n",
       "313  AKA94128.1  cenH3_(Plants)  None         None           None   \n",
       "314  AKA94129.1  cenH3_(Plants)  None         None           None   \n",
       "315  AKA94130.1  cenH3_(Plants)  None         None           None   \n",
       "316  AKA94131.1  cenH3_(Plants)  None         None           None   \n",
       "317  AKA94132.1  cenH3_(Plants)  None         None           None   \n",
       "318  AKA94133.1  cenH3_(Plants)  None         None           None   \n",
       "319  AKA94134.1  cenH3_(Plants)  None         None           None   \n",
       "320  AKA94135.1  cenH3_(Plants)  None         None           None   \n",
       "321  AKA94136.1  cenH3_(Plants)  None         None           None   \n",
       "322  AKA94137.1  cenH3_(Plants)  None         None           None   \n",
       "323  AKA94138.1  cenH3_(Plants)  None         None           None   \n",
       "324  AKA94139.1  cenH3_(Plants)  None         None           None   \n",
       "325  AKA94140.1  cenH3_(Plants)  None         None           None   \n",
       "326  AKA94141.1  cenH3_(Plants)  None         None           None   \n",
       "\n",
       "     taxonomy_id             organism        phylum          class  \\\n",
       "207       3888.0        Pisum sativum  Streptophyta  Magnoliopsida   \n",
       "208       3888.0        Pisum sativum  Streptophyta  Magnoliopsida   \n",
       "302      51020.0         Pisum fulvum  Streptophyta  Magnoliopsida   \n",
       "303       3855.0    Lathyrus clymenum  Streptophyta  Magnoliopsida   \n",
       "304     154494.0  Lathyrus latifolius  Streptophyta  Magnoliopsida   \n",
       "305     313121.0       Lathyrus niger  Streptophyta  Magnoliopsida   \n",
       "306       3858.0      Lathyrus ochrus  Streptophyta  Magnoliopsida   \n",
       "307       3860.0     Lathyrus sativus  Streptophyta  Magnoliopsida   \n",
       "308     313117.0  Lathyrus sylvestris  Streptophyta  Magnoliopsida   \n",
       "309     313120.0      Lathyrus vernus  Streptophyta  Magnoliopsida   \n",
       "310     154494.0  Lathyrus latifolius  Streptophyta  Magnoliopsida   \n",
       "311       3855.0    Lathyrus clymenum  Streptophyta  Magnoliopsida   \n",
       "312     313121.0       Lathyrus niger  Streptophyta  Magnoliopsida   \n",
       "313       3858.0      Lathyrus ochrus  Streptophyta  Magnoliopsida   \n",
       "314       3860.0     Lathyrus sativus  Streptophyta  Magnoliopsida   \n",
       "315     313117.0  Lathyrus sylvestris  Streptophyta  Magnoliopsida   \n",
       "316     313120.0      Lathyrus vernus  Streptophyta  Magnoliopsida   \n",
       "317      51020.0         Pisum fulvum  Streptophyta  Magnoliopsida   \n",
       "318       3864.0       Lens culinaris  Streptophyta  Magnoliopsida   \n",
       "319       3906.0           Vicia faba  Streptophyta  Magnoliopsida   \n",
       "320     347192.0    Vicia lathyroides  Streptophyta  Magnoliopsida   \n",
       "321       3912.0    Vicia narbonensis  Streptophyta  Magnoliopsida   \n",
       "322      29753.0      Vicia pannonica  Streptophyta  Magnoliopsida   \n",
       "323     233247.0      Vicia peregrina  Streptophyta  Magnoliopsida   \n",
       "324       3908.0         Vicia sativa  Streptophyta  Magnoliopsida   \n",
       "325     347188.0         Vicia sepium  Streptophyta  Magnoliopsida   \n",
       "326       3911.0        Vicia villosa  Streptophyta  Magnoliopsida   \n",
       "\n",
       "    taxonomy_group  info                                           sequence  \\\n",
       "207           None  None  MGRVKHFPSPSKPAASDNLGKKKRRCKPGTKALREIRKFQKDVKLL...   \n",
       "208           None  None  MARVKQTPRHARENQERKKRRNKPGTVALREIKKLQKTFQLLIPYA...   \n",
       "302           None  None  MGRVKHFPSPSKPAASDNLGKKKRRCKPGTKALREIRKFQKDVKLL...   \n",
       "303           None  None  MGRVKQFPPSKFAARNNHEKKKRRVKPGTVALREIRKFQKDVKLLI...   \n",
       "304           None  None  MGRVKHFPRPSKPAASNNHEKKKRRSKPGTKAVREIRKFQKDVKLL...   \n",
       "305           None  None  MGRVKHFPRPSKSAASDNHGKKKRRSKPGTKALREIRKFQKDVKLL...   \n",
       "306           None  None  MGRVKHFPRPSKSAANNTHEKKKRRFKRGTALQEIRKFQKDVKLLI...   \n",
       "307           None  None  MGRVKHFPRPSKPAASDNQEKRKRRSKPGTKAVREIRKFQKDVKLL...   \n",
       "308           None  None  MGRVKHFPRPSKPAASNNHEKKKRRSKPGTKAVREIRKFQKDVKLL...   \n",
       "309           None  None  MGRVKHFPRPSKPAASDNHGKKKRLSKPGTKALREIRKFQNDVKLL...   \n",
       "310           None  None  MARVKQKPRHARENQERKKRRNKPGTVALREIRKLQKTFQLLIPYA...   \n",
       "311           None  None  MARVKQKPRHARENQERKKRRNKPGTVALREIRKLQKTCQLLIPYA...   \n",
       "312           None  None  MARVKQTPRHARENQERKKRRNKPGTVALREIRKLQKTFQLLIPYA...   \n",
       "313           None  None  MGRVKQRPRHARENQEKKKRRNKPGTVALREIRKFQKTCQLLIPYA...   \n",
       "314           None  None  MARVKQTPRHARENQERKKRRNKPGTVALREIRKFQKTFKLLIPYA...   \n",
       "315           None  None  MARVKQKPRHARENQERKKRRNKPGTVALREIRKLQKTFQLLIPYA...   \n",
       "316           None  None  MARVKQTPRHARENQERKKRRNKPGTVALREIRKFQKTFQLLIPYA...   \n",
       "317           None  None  MARVKQTPRHARENQERKKRRNKPGTVALREIKKLQKTFQLLIPYA...   \n",
       "318           None  None  MARVKQTPRPVLQNQERKKRRNKPGTVALREIRKLQKTFQLLIPYA...   \n",
       "319           None  None  MARVKQTPRHARETQEKKKRRNKPGTVALREIRKMQKTFKLIIPYA...   \n",
       "320           None  None  MARVKQTPRHARETQEKKKRRNKPGTVALREIRKLQKTFKLIIPYA...   \n",
       "321           None  None  MARVKQTPRPAREIQERKKRRNKPGTVALREIQKMQKTFKLIIPYA...   \n",
       "322           None  None  MARVKQTPRHARENQAKKKRRNKPGTVALREIRKLQKAVKLLIPYA...   \n",
       "323           None  None  MARVKQTPRNARENQERKKRRNKPGTVVLHEIRKWQKAVKLIIPYA...   \n",
       "324           None  None  MARVKQTPRHARETQEKKKRRNKPGTVALREIRKMQKTFKLIIPYA...   \n",
       "325           None  None  MARVKQTPRHARETQEKKKRRNKPGTVALREIRKLQKTFKLIIPYA...   \n",
       "326           None  None  MARVKQKPRHARVDQEKKKRRNKPGTVALREIRKLQKSVKLLIPYA...   \n",
       "\n",
       "    variant_under_consideration  \n",
       "207                        None  \n",
       "208                        None  \n",
       "302                        None  \n",
       "303                        None  \n",
       "304                        None  \n",
       "305                        None  \n",
       "306                        None  \n",
       "307                        None  \n",
       "308                        None  \n",
       "309                        None  \n",
       "310                        None  \n",
       "311                        None  \n",
       "312                        None  \n",
       "313                        None  \n",
       "314                        None  \n",
       "315                        None  \n",
       "316                        None  \n",
       "317                        None  \n",
       "318                        None  \n",
       "319                        None  \n",
       "320                        None  \n",
       "321                        None  \n",
       "322                        None  \n",
       "323                        None  \n",
       "324                        None  \n",
       "325                        None  \n",
       "326                        None  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin([v[\"Protein ID\"] for k, v in records_data.items()])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f9b96c04-bdc9-481e-8f9e-b73a9f8f3aff",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5923b7f-0d25-4dda-9985-364134917329",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "bffde933-d1e0-4ebb-9a0d-5c88a0f4da12",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         id title   doi author  year pubmed_id\n",
       "0  neumann_centromeres_2015  None  None   None  None      None"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pid = \"neumann_centromeres_2015\"\n",
    "query = f\"SELECT * FROM publication WHERE id='{pid}'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f2f6718d-2d84-4f98-abd5-aa539e2de599",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in [v[\"Protein ID\"] for k, v in records_data.items()]:\n",
    "    cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ce3446fd-1735-47ef-9369-bd8eff48d69c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>AEX31245.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3888.0</td>\n",
       "      <td>Pisum sativum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKHFPSPSKPAASDNLGKKKRRCKPGTKALREIRKFQKDVKLL...</td>\n",
       "      <td>None</td>\n",
       "      <td>AEX31245.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>AEX31246.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3888.0</td>\n",
       "      <td>Pisum sativum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARENQERKKRRNKPGTVALREIKKLQKTFQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AEX31246.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>AKA94117.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>51020.0</td>\n",
       "      <td>Pisum fulvum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKHFPSPSKPAASDNLGKKKRRCKPGTKALREIRKFQKDVKLL...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94117.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>AKA94118.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3855.0</td>\n",
       "      <td>Lathyrus clymenum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKQFPPSKFAARNNHEKKKRRVKPGTVALREIRKFQKDVKLLI...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94118.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381</th>\n",
       "      <td>AKA94119.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>154494.0</td>\n",
       "      <td>Lathyrus latifolius</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKHFPRPSKPAASNNHEKKKRRSKPGTKAVREIRKFQKDVKLL...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94119.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>382</th>\n",
       "      <td>AKA94120.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>313121.0</td>\n",
       "      <td>Lathyrus niger</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKHFPRPSKSAASDNHGKKKRRSKPGTKALREIRKFQKDVKLL...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94120.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383</th>\n",
       "      <td>AKA94121.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3858.0</td>\n",
       "      <td>Lathyrus ochrus</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKHFPRPSKSAANNTHEKKKRRFKRGTALQEIRKFQKDVKLLI...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94121.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>384</th>\n",
       "      <td>AKA94122.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3860.0</td>\n",
       "      <td>Lathyrus sativus</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKHFPRPSKPAASDNQEKRKRRSKPGTKAVREIRKFQKDVKLL...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94122.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>385</th>\n",
       "      <td>AKA94123.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>313117.0</td>\n",
       "      <td>Lathyrus sylvestris</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKHFPRPSKPAASNNHEKKKRRSKPGTKAVREIRKFQKDVKLL...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94123.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386</th>\n",
       "      <td>AKA94124.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>313120.0</td>\n",
       "      <td>Lathyrus vernus</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKHFPRPSKPAASDNHGKKKRLSKPGTKALREIRKFQNDVKLL...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94124.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387</th>\n",
       "      <td>AKA94125.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>154494.0</td>\n",
       "      <td>Lathyrus latifolius</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQKPRHARENQERKKRRNKPGTVALREIRKLQKTFQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94125.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388</th>\n",
       "      <td>AKA94126.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3855.0</td>\n",
       "      <td>Lathyrus clymenum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQKPRHARENQERKKRRNKPGTVALREIRKLQKTCQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94126.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>389</th>\n",
       "      <td>AKA94127.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>313121.0</td>\n",
       "      <td>Lathyrus niger</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARENQERKKRRNKPGTVALREIRKLQKTFQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94127.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>390</th>\n",
       "      <td>AKA94128.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3858.0</td>\n",
       "      <td>Lathyrus ochrus</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKQRPRHARENQEKKKRRNKPGTVALREIRKFQKTCQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94128.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>391</th>\n",
       "      <td>AKA94129.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3860.0</td>\n",
       "      <td>Lathyrus sativus</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARENQERKKRRNKPGTVALREIRKFQKTFKLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94129.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>392</th>\n",
       "      <td>AKA94130.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>313117.0</td>\n",
       "      <td>Lathyrus sylvestris</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQKPRHARENQERKKRRNKPGTVALREIRKLQKTFQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94130.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>393</th>\n",
       "      <td>AKA94131.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>313120.0</td>\n",
       "      <td>Lathyrus vernus</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARENQERKKRRNKPGTVALREIRKFQKTFQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94131.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394</th>\n",
       "      <td>AKA94132.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>51020.0</td>\n",
       "      <td>Pisum fulvum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARENQERKKRRNKPGTVALREIKKLQKTFQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94132.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>AKA94133.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3864.0</td>\n",
       "      <td>Lens culinaris</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRPVLQNQERKKRRNKPGTVALREIRKLQKTFQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94133.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>AKA94134.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3906.0</td>\n",
       "      <td>Vicia faba</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARETQEKKKRRNKPGTVALREIRKMQKTFKLIIPYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94134.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>AKA94135.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>347192.0</td>\n",
       "      <td>Vicia lathyroides</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARETQEKKKRRNKPGTVALREIRKLQKTFKLIIPYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94135.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>AKA94136.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3912.0</td>\n",
       "      <td>Vicia narbonensis</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRPAREIQERKKRRNKPGTVALREIQKMQKTFKLIIPYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94136.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>AKA94137.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>29753.0</td>\n",
       "      <td>Vicia pannonica</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARENQAKKKRRNKPGTVALREIRKLQKAVKLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94137.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>400</th>\n",
       "      <td>AKA94138.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>233247.0</td>\n",
       "      <td>Vicia peregrina</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRNARENQERKKRRNKPGTVVLHEIRKWQKAVKLIIPYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94138.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>401</th>\n",
       "      <td>AKA94139.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3908.0</td>\n",
       "      <td>Vicia sativa</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARETQEKKKRRNKPGTVALREIRKMQKTFKLIIPYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94139.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>402</th>\n",
       "      <td>AKA94140.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>347188.0</td>\n",
       "      <td>Vicia sepium</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARETQEKKKRRNKPGTVALREIRKLQKTFKLIIPYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94140.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>403</th>\n",
       "      <td>AKA94141.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3911.0</td>\n",
       "      <td>Vicia villosa</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQKPRHARVDQEKKKRRNKPGTVALREIRKLQKSVKLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AKA94141.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession         variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "262  AEX31245.1  cenH3_(Plants)  None         None           None   \n",
       "263  AEX31246.1  cenH3_(Plants)  None         None           None   \n",
       "379  AKA94117.1  cenH3_(Plants)  None         None           None   \n",
       "380  AKA94118.1  cenH3_(Plants)  None         None           None   \n",
       "381  AKA94119.1  cenH3_(Plants)  None         None           None   \n",
       "382  AKA94120.1  cenH3_(Plants)  None         None           None   \n",
       "383  AKA94121.1  cenH3_(Plants)  None         None           None   \n",
       "384  AKA94122.1  cenH3_(Plants)  None         None           None   \n",
       "385  AKA94123.1  cenH3_(Plants)  None         None           None   \n",
       "386  AKA94124.1  cenH3_(Plants)  None         None           None   \n",
       "387  AKA94125.1  cenH3_(Plants)  None         None           None   \n",
       "388  AKA94126.1  cenH3_(Plants)  None         None           None   \n",
       "389  AKA94127.1  cenH3_(Plants)  None         None           None   \n",
       "390  AKA94128.1  cenH3_(Plants)  None         None           None   \n",
       "391  AKA94129.1  cenH3_(Plants)  None         None           None   \n",
       "392  AKA94130.1  cenH3_(Plants)  None         None           None   \n",
       "393  AKA94131.1  cenH3_(Plants)  None         None           None   \n",
       "394  AKA94132.1  cenH3_(Plants)  None         None           None   \n",
       "395  AKA94133.1  cenH3_(Plants)  None         None           None   \n",
       "396  AKA94134.1  cenH3_(Plants)  None         None           None   \n",
       "397  AKA94135.1  cenH3_(Plants)  None         None           None   \n",
       "398  AKA94136.1  cenH3_(Plants)  None         None           None   \n",
       "399  AKA94137.1  cenH3_(Plants)  None         None           None   \n",
       "400  AKA94138.1  cenH3_(Plants)  None         None           None   \n",
       "401  AKA94139.1  cenH3_(Plants)  None         None           None   \n",
       "402  AKA94140.1  cenH3_(Plants)  None         None           None   \n",
       "403  AKA94141.1  cenH3_(Plants)  None         None           None   \n",
       "\n",
       "     taxonomy_id             organism        phylum          class  \\\n",
       "262       3888.0        Pisum sativum  Streptophyta  Magnoliopsida   \n",
       "263       3888.0        Pisum sativum  Streptophyta  Magnoliopsida   \n",
       "379      51020.0         Pisum fulvum  Streptophyta  Magnoliopsida   \n",
       "380       3855.0    Lathyrus clymenum  Streptophyta  Magnoliopsida   \n",
       "381     154494.0  Lathyrus latifolius  Streptophyta  Magnoliopsida   \n",
       "382     313121.0       Lathyrus niger  Streptophyta  Magnoliopsida   \n",
       "383       3858.0      Lathyrus ochrus  Streptophyta  Magnoliopsida   \n",
       "384       3860.0     Lathyrus sativus  Streptophyta  Magnoliopsida   \n",
       "385     313117.0  Lathyrus sylvestris  Streptophyta  Magnoliopsida   \n",
       "386     313120.0      Lathyrus vernus  Streptophyta  Magnoliopsida   \n",
       "387     154494.0  Lathyrus latifolius  Streptophyta  Magnoliopsida   \n",
       "388       3855.0    Lathyrus clymenum  Streptophyta  Magnoliopsida   \n",
       "389     313121.0       Lathyrus niger  Streptophyta  Magnoliopsida   \n",
       "390       3858.0      Lathyrus ochrus  Streptophyta  Magnoliopsida   \n",
       "391       3860.0     Lathyrus sativus  Streptophyta  Magnoliopsida   \n",
       "392     313117.0  Lathyrus sylvestris  Streptophyta  Magnoliopsida   \n",
       "393     313120.0      Lathyrus vernus  Streptophyta  Magnoliopsida   \n",
       "394      51020.0         Pisum fulvum  Streptophyta  Magnoliopsida   \n",
       "395       3864.0       Lens culinaris  Streptophyta  Magnoliopsida   \n",
       "396       3906.0           Vicia faba  Streptophyta  Magnoliopsida   \n",
       "397     347192.0    Vicia lathyroides  Streptophyta  Magnoliopsida   \n",
       "398       3912.0    Vicia narbonensis  Streptophyta  Magnoliopsida   \n",
       "399      29753.0      Vicia pannonica  Streptophyta  Magnoliopsida   \n",
       "400     233247.0      Vicia peregrina  Streptophyta  Magnoliopsida   \n",
       "401       3908.0         Vicia sativa  Streptophyta  Magnoliopsida   \n",
       "402     347188.0         Vicia sepium  Streptophyta  Magnoliopsida   \n",
       "403       3911.0        Vicia villosa  Streptophyta  Magnoliopsida   \n",
       "\n",
       "    taxonomy_group  info                                           sequence  \\\n",
       "262           None  None  MGRVKHFPSPSKPAASDNLGKKKRRCKPGTKALREIRKFQKDVKLL...   \n",
       "263           None  None  MARVKQTPRHARENQERKKRRNKPGTVALREIKKLQKTFQLLIPYA...   \n",
       "379           None  None  MGRVKHFPSPSKPAASDNLGKKKRRCKPGTKALREIRKFQKDVKLL...   \n",
       "380           None  None  MGRVKQFPPSKFAARNNHEKKKRRVKPGTVALREIRKFQKDVKLLI...   \n",
       "381           None  None  MGRVKHFPRPSKPAASNNHEKKKRRSKPGTKAVREIRKFQKDVKLL...   \n",
       "382           None  None  MGRVKHFPRPSKSAASDNHGKKKRRSKPGTKALREIRKFQKDVKLL...   \n",
       "383           None  None  MGRVKHFPRPSKSAANNTHEKKKRRFKRGTALQEIRKFQKDVKLLI...   \n",
       "384           None  None  MGRVKHFPRPSKPAASDNQEKRKRRSKPGTKAVREIRKFQKDVKLL...   \n",
       "385           None  None  MGRVKHFPRPSKPAASNNHEKKKRRSKPGTKAVREIRKFQKDVKLL...   \n",
       "386           None  None  MGRVKHFPRPSKPAASDNHGKKKRLSKPGTKALREIRKFQNDVKLL...   \n",
       "387           None  None  MARVKQKPRHARENQERKKRRNKPGTVALREIRKLQKTFQLLIPYA...   \n",
       "388           None  None  MARVKQKPRHARENQERKKRRNKPGTVALREIRKLQKTCQLLIPYA...   \n",
       "389           None  None  MARVKQTPRHARENQERKKRRNKPGTVALREIRKLQKTFQLLIPYA...   \n",
       "390           None  None  MGRVKQRPRHARENQEKKKRRNKPGTVALREIRKFQKTCQLLIPYA...   \n",
       "391           None  None  MARVKQTPRHARENQERKKRRNKPGTVALREIRKFQKTFKLLIPYA...   \n",
       "392           None  None  MARVKQKPRHARENQERKKRRNKPGTVALREIRKLQKTFQLLIPYA...   \n",
       "393           None  None  MARVKQTPRHARENQERKKRRNKPGTVALREIRKFQKTFQLLIPYA...   \n",
       "394           None  None  MARVKQTPRHARENQERKKRRNKPGTVALREIKKLQKTFQLLIPYA...   \n",
       "395           None  None  MARVKQTPRPVLQNQERKKRRNKPGTVALREIRKLQKTFQLLIPYA...   \n",
       "396           None  None  MARVKQTPRHARETQEKKKRRNKPGTVALREIRKMQKTFKLIIPYA...   \n",
       "397           None  None  MARVKQTPRHARETQEKKKRRNKPGTVALREIRKLQKTFKLIIPYA...   \n",
       "398           None  None  MARVKQTPRPAREIQERKKRRNKPGTVALREIQKMQKTFKLIIPYA...   \n",
       "399           None  None  MARVKQTPRHARENQAKKKRRNKPGTVALREIRKLQKAVKLLIPYA...   \n",
       "400           None  None  MARVKQTPRNARENQERKKRRNKPGTVVLHEIRKWQKAVKLIIPYA...   \n",
       "401           None  None  MARVKQTPRHARETQEKKKRRNKPGTVALREIRKMQKTFKLIIPYA...   \n",
       "402           None  None  MARVKQTPRHARETQEKKKRRNKPGTVALREIRKLQKTFKLIIPYA...   \n",
       "403           None  None  MARVKQKPRHARVDQEKKKRRNKPGTVALREIRKLQKSVKLLIPYA...   \n",
       "\n",
       "    variant_under_consideration sequence_accession            publication_id  \n",
       "262                        None         AEX31245.1  neumann_centromeres_2015  \n",
       "263                        None         AEX31246.1  neumann_centromeres_2015  \n",
       "379                        None         AKA94117.1  neumann_centromeres_2015  \n",
       "380                        None         AKA94118.1  neumann_centromeres_2015  \n",
       "381                        None         AKA94119.1  neumann_centromeres_2015  \n",
       "382                        None         AKA94120.1  neumann_centromeres_2015  \n",
       "383                        None         AKA94121.1  neumann_centromeres_2015  \n",
       "384                        None         AKA94122.1  neumann_centromeres_2015  \n",
       "385                        None         AKA94123.1  neumann_centromeres_2015  \n",
       "386                        None         AKA94124.1  neumann_centromeres_2015  \n",
       "387                        None         AKA94125.1  neumann_centromeres_2015  \n",
       "388                        None         AKA94126.1  neumann_centromeres_2015  \n",
       "389                        None         AKA94127.1  neumann_centromeres_2015  \n",
       "390                        None         AKA94128.1  neumann_centromeres_2015  \n",
       "391                        None         AKA94129.1  neumann_centromeres_2015  \n",
       "392                        None         AKA94130.1  neumann_centromeres_2015  \n",
       "393                        None         AKA94131.1  neumann_centromeres_2015  \n",
       "394                        None         AKA94132.1  neumann_centromeres_2015  \n",
       "395                        None         AKA94133.1  neumann_centromeres_2015  \n",
       "396                        None         AKA94134.1  neumann_centromeres_2015  \n",
       "397                        None         AKA94135.1  neumann_centromeres_2015  \n",
       "398                        None         AKA94136.1  neumann_centromeres_2015  \n",
       "399                        None         AKA94137.1  neumann_centromeres_2015  \n",
       "400                        None         AKA94138.1  neumann_centromeres_2015  \n",
       "401                        None         AKA94139.1  neumann_centromeres_2015  \n",
       "402                        None         AKA94140.1  neumann_centromeres_2015  \n",
       "403                        None         AKA94141.1  neumann_centromeres_2015  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin([v[\"Protein ID\"] for k, v in records_data.items()])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "472eaeac-80a8-4e79-8186-ef3c084e6b79",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89b84d75-589d-4d53-8667-1630e5f17da4",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add sequences from `sanei_loss_2011`, `karimi-ashtiyani_point_2015` and `ishii_differential_2015`</span>\n",
    "Protein Accessions: AEK21394.1\n",
    "AEK21393.1\n",
    "AEK21392.1\n",
    "ADB03182.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "5821f461-680a-4dbd-a4ad-fdba33424155",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"AEK21394.1\", \"AEK21393.1\", \"AEK21392.1\", \"ADB03182.1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62f1bd69-3f86-4aba-bc0a-9c9cff92d933",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "479e3289-36f6-4f8c-a777-a9705d2eb7a1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "1475d031-747d-46c1-bb22-1e64a3d5b4df",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 4516\n",
      "Fetched taxid from NCBI 112509\n",
      "Fetched taxid from NCBI 112509\n",
      "Fetched taxid from NCBI 4516\n",
      "accession ADB03182.1 <class 'str'>\n",
      "variant cenH3_(Plants) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 4516 <class 'int'>\n",
      "organism Hordeum bulbosum <class 'str'>\n",
      "phylum Streptophyta <class 'str'>\n",
      "class Magnoliopsida <class 'str'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MARTKHPAVRKSKAPPRKKVGSARAPAAAQRRHETDGAGTSETPRRGPAPAADQGAPGEPKKRKPHRYRPGTVALREIRKYQKSVDFLIPFAPFVRLVKEVTEFYCPAISRWTPQALLAVQEAAEYHLVDVFERAHLCAIHAKRVTVMQKDIQLA <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for a in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=a, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"cenH3_(Plants)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "34694562-06cd-431a-8055-87991ef3e1be",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "934e224b-e847-48fd-939a-11a1e38aa7ea",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>ADB03182.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4516.0</td>\n",
       "      <td>Hordeum bulbosum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKSKAPPRKKVGSARAPAAAQRRHETDGAGTSETPRR...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>AEK21392.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>112509.0</td>\n",
       "      <td>Hordeum vulgare subsp. vulgare</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKSKAPPKKKIGSASSPSAAQRRQETDGAGTSETPRR...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>AEK21393.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>112509.0</td>\n",
       "      <td>Hordeum vulgare subsp. vulgare</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKKTVAAKEKRPPCSKSEPQSQPKKKEKRAYRFRPGTVALREI...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>AEK21394.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4516.0</td>\n",
       "      <td>Hordeum bulbosum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKKTVAATKRRSPRTRLEPQSQPEKKKRAHRFRPGTVALREIR...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession         variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "126  ADB03182.1  cenH3_(Plants)  None         None           None   \n",
       "202  AEK21392.1  cenH3_(Plants)  None         None           None   \n",
       "203  AEK21393.1  cenH3_(Plants)  None         None           None   \n",
       "204  AEK21394.1  cenH3_(Plants)  None         None           None   \n",
       "\n",
       "     taxonomy_id                        organism        phylum          class  \\\n",
       "126       4516.0                Hordeum bulbosum  Streptophyta  Magnoliopsida   \n",
       "202     112509.0  Hordeum vulgare subsp. vulgare  Streptophyta  Magnoliopsida   \n",
       "203     112509.0  Hordeum vulgare subsp. vulgare  Streptophyta  Magnoliopsida   \n",
       "204       4516.0                Hordeum bulbosum  Streptophyta  Magnoliopsida   \n",
       "\n",
       "    taxonomy_group  info                                           sequence  \\\n",
       "126           None  None  MARTKHPAVRKSKAPPRKKVGSARAPAAAQRRHETDGAGTSETPRR...   \n",
       "202           None  None  MARTKHPAVRKSKAPPKKKIGSASSPSAAQRRQETDGAGTSETPRR...   \n",
       "203           None  None  MARTKKTVAAKEKRPPCSKSEPQSQPKKKEKRAYRFRPGTVALREI...   \n",
       "204           None  None  MARTKKTVAATKRRSPRTRLEPQSQPEKKKRAHRFRPGTVALREIR...   \n",
       "\n",
       "    variant_under_consideration  \n",
       "126                        None  \n",
       "202                        None  \n",
       "203                        None  \n",
       "204                        None  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "3ebc5d05-19d3-4321-9b76-b2387c5ec965",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4767530-69ec-48f7-8eae-586ad4aff32d",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "aee2fa07-bb56-47c9-8fbc-8ec6470acf33",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>ishii_differential_2015</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>karimi-ashtiyani_point_2015</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>sanei_loss_2011</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              id title   doi author  year pubmed_id\n",
       "108      ishii_differential_2015  None  None   None  None      None\n",
       "114  karimi-ashtiyani_point_2015  None  None   None  None      None\n",
       "145              sanei_loss_2011  None  None   None  None      None"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"sanei_loss_2011\", \"karimi-ashtiyani_point_2015\", \"ishii_differential_2015\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "4c587c8d-9243-4a65-8247-8ca1a2f8f707",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "8599749a-dabe-46b1-9af0-8277593bca06",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>ADB03182.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4516.0</td>\n",
       "      <td>Hordeum bulbosum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKSKAPPRKKVGSARAPAAAQRRHETDGAGTSETPRR...</td>\n",
       "      <td>None</td>\n",
       "      <td>ADB03182.1</td>\n",
       "      <td>ishii_differential_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>ADB03182.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4516.0</td>\n",
       "      <td>Hordeum bulbosum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKSKAPPRKKVGSARAPAAAQRRHETDGAGTSETPRR...</td>\n",
       "      <td>None</td>\n",
       "      <td>ADB03182.1</td>\n",
       "      <td>karimi-ashtiyani_point_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>ADB03182.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4516.0</td>\n",
       "      <td>Hordeum bulbosum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKSKAPPRKKVGSARAPAAAQRRHETDGAGTSETPRR...</td>\n",
       "      <td>None</td>\n",
       "      <td>ADB03182.1</td>\n",
       "      <td>sanei_loss_2011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>AEK21392.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>112509.0</td>\n",
       "      <td>Hordeum vulgare subsp. vulgare</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKSKAPPKKKIGSASSPSAAQRRQETDGAGTSETPRR...</td>\n",
       "      <td>None</td>\n",
       "      <td>AEK21392.1</td>\n",
       "      <td>ishii_differential_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>AEK21392.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>112509.0</td>\n",
       "      <td>Hordeum vulgare subsp. vulgare</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKSKAPPKKKIGSASSPSAAQRRQETDGAGTSETPRR...</td>\n",
       "      <td>None</td>\n",
       "      <td>AEK21392.1</td>\n",
       "      <td>karimi-ashtiyani_point_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>AEK21392.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>112509.0</td>\n",
       "      <td>Hordeum vulgare subsp. vulgare</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKSKAPPKKKIGSASSPSAAQRRQETDGAGTSETPRR...</td>\n",
       "      <td>None</td>\n",
       "      <td>AEK21392.1</td>\n",
       "      <td>sanei_loss_2011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>AEK21393.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>112509.0</td>\n",
       "      <td>Hordeum vulgare subsp. vulgare</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKKTVAAKEKRPPCSKSEPQSQPKKKEKRAYRFRPGTVALREI...</td>\n",
       "      <td>None</td>\n",
       "      <td>AEK21393.1</td>\n",
       "      <td>ishii_differential_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>AEK21393.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>112509.0</td>\n",
       "      <td>Hordeum vulgare subsp. vulgare</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKKTVAAKEKRPPCSKSEPQSQPKKKEKRAYRFRPGTVALREI...</td>\n",
       "      <td>None</td>\n",
       "      <td>AEK21393.1</td>\n",
       "      <td>karimi-ashtiyani_point_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>AEK21393.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>112509.0</td>\n",
       "      <td>Hordeum vulgare subsp. vulgare</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKKTVAAKEKRPPCSKSEPQSQPKKKEKRAYRFRPGTVALREI...</td>\n",
       "      <td>None</td>\n",
       "      <td>AEK21393.1</td>\n",
       "      <td>sanei_loss_2011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>AEK21394.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4516.0</td>\n",
       "      <td>Hordeum bulbosum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKKTVAATKRRSPRTRLEPQSQPEKKKRAHRFRPGTVALREIR...</td>\n",
       "      <td>None</td>\n",
       "      <td>AEK21394.1</td>\n",
       "      <td>ishii_differential_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>AEK21394.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4516.0</td>\n",
       "      <td>Hordeum bulbosum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKKTVAATKRRSPRTRLEPQSQPEKKKRAHRFRPGTVALREIR...</td>\n",
       "      <td>None</td>\n",
       "      <td>AEK21394.1</td>\n",
       "      <td>karimi-ashtiyani_point_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>AEK21394.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4516.0</td>\n",
       "      <td>Hordeum bulbosum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKKTVAATKRRSPRTRLEPQSQPEKKKRAHRFRPGTVALREIR...</td>\n",
       "      <td>None</td>\n",
       "      <td>AEK21394.1</td>\n",
       "      <td>sanei_loss_2011</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession         variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "154  ADB03182.1  cenH3_(Plants)  None         None           None   \n",
       "155  ADB03182.1  cenH3_(Plants)  None         None           None   \n",
       "156  ADB03182.1  cenH3_(Plants)  None         None           None   \n",
       "257  AEK21392.1  cenH3_(Plants)  None         None           None   \n",
       "258  AEK21392.1  cenH3_(Plants)  None         None           None   \n",
       "259  AEK21392.1  cenH3_(Plants)  None         None           None   \n",
       "260  AEK21393.1  cenH3_(Plants)  None         None           None   \n",
       "261  AEK21393.1  cenH3_(Plants)  None         None           None   \n",
       "262  AEK21393.1  cenH3_(Plants)  None         None           None   \n",
       "263  AEK21394.1  cenH3_(Plants)  None         None           None   \n",
       "264  AEK21394.1  cenH3_(Plants)  None         None           None   \n",
       "265  AEK21394.1  cenH3_(Plants)  None         None           None   \n",
       "\n",
       "     taxonomy_id                        organism        phylum          class  \\\n",
       "154       4516.0                Hordeum bulbosum  Streptophyta  Magnoliopsida   \n",
       "155       4516.0                Hordeum bulbosum  Streptophyta  Magnoliopsida   \n",
       "156       4516.0                Hordeum bulbosum  Streptophyta  Magnoliopsida   \n",
       "257     112509.0  Hordeum vulgare subsp. vulgare  Streptophyta  Magnoliopsida   \n",
       "258     112509.0  Hordeum vulgare subsp. vulgare  Streptophyta  Magnoliopsida   \n",
       "259     112509.0  Hordeum vulgare subsp. vulgare  Streptophyta  Magnoliopsida   \n",
       "260     112509.0  Hordeum vulgare subsp. vulgare  Streptophyta  Magnoliopsida   \n",
       "261     112509.0  Hordeum vulgare subsp. vulgare  Streptophyta  Magnoliopsida   \n",
       "262     112509.0  Hordeum vulgare subsp. vulgare  Streptophyta  Magnoliopsida   \n",
       "263       4516.0                Hordeum bulbosum  Streptophyta  Magnoliopsida   \n",
       "264       4516.0                Hordeum bulbosum  Streptophyta  Magnoliopsida   \n",
       "265       4516.0                Hordeum bulbosum  Streptophyta  Magnoliopsida   \n",
       "\n",
       "    taxonomy_group  info                                           sequence  \\\n",
       "154           None  None  MARTKHPAVRKSKAPPRKKVGSARAPAAAQRRHETDGAGTSETPRR...   \n",
       "155           None  None  MARTKHPAVRKSKAPPRKKVGSARAPAAAQRRHETDGAGTSETPRR...   \n",
       "156           None  None  MARTKHPAVRKSKAPPRKKVGSARAPAAAQRRHETDGAGTSETPRR...   \n",
       "257           None  None  MARTKHPAVRKSKAPPKKKIGSASSPSAAQRRQETDGAGTSETPRR...   \n",
       "258           None  None  MARTKHPAVRKSKAPPKKKIGSASSPSAAQRRQETDGAGTSETPRR...   \n",
       "259           None  None  MARTKHPAVRKSKAPPKKKIGSASSPSAAQRRQETDGAGTSETPRR...   \n",
       "260           None  None  MARTKKTVAAKEKRPPCSKSEPQSQPKKKEKRAYRFRPGTVALREI...   \n",
       "261           None  None  MARTKKTVAAKEKRPPCSKSEPQSQPKKKEKRAYRFRPGTVALREI...   \n",
       "262           None  None  MARTKKTVAAKEKRPPCSKSEPQSQPKKKEKRAYRFRPGTVALREI...   \n",
       "263           None  None  MARTKKTVAATKRRSPRTRLEPQSQPEKKKRAHRFRPGTVALREIR...   \n",
       "264           None  None  MARTKKTVAATKRRSPRTRLEPQSQPEKKKRAHRFRPGTVALREIR...   \n",
       "265           None  None  MARTKKTVAATKRRSPRTRLEPQSQPEKKKRAHRFRPGTVALREIR...   \n",
       "\n",
       "    variant_under_consideration sequence_accession  \\\n",
       "154                        None         ADB03182.1   \n",
       "155                        None         ADB03182.1   \n",
       "156                        None         ADB03182.1   \n",
       "257                        None         AEK21392.1   \n",
       "258                        None         AEK21392.1   \n",
       "259                        None         AEK21392.1   \n",
       "260                        None         AEK21393.1   \n",
       "261                        None         AEK21393.1   \n",
       "262                        None         AEK21393.1   \n",
       "263                        None         AEK21394.1   \n",
       "264                        None         AEK21394.1   \n",
       "265                        None         AEK21394.1   \n",
       "\n",
       "                  publication_id  \n",
       "154      ishii_differential_2015  \n",
       "155  karimi-ashtiyani_point_2015  \n",
       "156              sanei_loss_2011  \n",
       "257      ishii_differential_2015  \n",
       "258  karimi-ashtiyani_point_2015  \n",
       "259              sanei_loss_2011  \n",
       "260      ishii_differential_2015  \n",
       "261  karimi-ashtiyani_point_2015  \n",
       "262              sanei_loss_2011  \n",
       "263      ishii_differential_2015  \n",
       "264  karimi-ashtiyani_point_2015  \n",
       "265              sanei_loss_2011  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "539b590a-af25-41bd-b5b2-e8c41ec7c106",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c1b8490-8296-4999-81e2-0bd42ca6c0b7",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add sequences from `neumann_stretching_2012`</span>\n",
    "Protein Accessions: AEX31246.1\n",
    "AEX31245.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "0807f2d5-868c-4ad2-99d4-2c1000e1dfe5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"AEX31246.1\", \"AEX31245.1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bba1ea81-ecb3-4da9-90e3-7f38985fac02",
   "metadata": {},
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a858d9f8-18ef-4667-90ed-41db2514f0a5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>AEX31245.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3888.0</td>\n",
       "      <td>Pisum sativum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKHFPSPSKPAASDNLGKKKRRCKPGTKALREIRKFQKDVKLL...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>AEX31246.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3888.0</td>\n",
       "      <td>Pisum sativum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARENQERKKRRNKPGTVALREIKKLQKTFQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession         variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "211  AEX31245.1  cenH3_(Plants)  None         None           None   \n",
       "212  AEX31246.1  cenH3_(Plants)  None         None           None   \n",
       "\n",
       "     taxonomy_id       organism        phylum          class taxonomy_group  \\\n",
       "211       3888.0  Pisum sativum  Streptophyta  Magnoliopsida           None   \n",
       "212       3888.0  Pisum sativum  Streptophyta  Magnoliopsida           None   \n",
       "\n",
       "     info                                           sequence  \\\n",
       "211  None  MGRVKHFPSPSKPAASDNLGKKKRRCKPGTKALREIRKFQKDVKLL...   \n",
       "212  None  MARVKQTPRHARENQERKKRRNKPGTVALREIKKLQKTFQLLIPYA...   \n",
       "\n",
       "    variant_under_consideration  \n",
       "211                        None  \n",
       "212                        None  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "47365e4e-37d8-4ef3-a5ad-0f8982276d15",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>AEX31245.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3888.0</td>\n",
       "      <td>Pisum sativum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKHFPSPSKPAASDNLGKKKRRCKPGTKALREIRKFQKDVKLL...</td>\n",
       "      <td>None</td>\n",
       "      <td>AEX31245.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>275</th>\n",
       "      <td>AEX31246.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3888.0</td>\n",
       "      <td>Pisum sativum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARENQERKKRRNKPGTVALREIKKLQKTFQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AEX31246.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession         variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "274  AEX31245.1  cenH3_(Plants)  None         None           None   \n",
       "275  AEX31246.1  cenH3_(Plants)  None         None           None   \n",
       "\n",
       "     taxonomy_id       organism        phylum          class taxonomy_group  \\\n",
       "274       3888.0  Pisum sativum  Streptophyta  Magnoliopsida           None   \n",
       "275       3888.0  Pisum sativum  Streptophyta  Magnoliopsida           None   \n",
       "\n",
       "     info                                           sequence  \\\n",
       "274  None  MGRVKHFPSPSKPAASDNLGKKKRRCKPGTKALREIRKFQKDVKLL...   \n",
       "275  None  MARVKQTPRHARENQERKKRRNKPGTVALREIKKLQKTFQLLIPYA...   \n",
       "\n",
       "    variant_under_consideration sequence_accession            publication_id  \n",
       "274                        None         AEX31245.1  neumann_centromeres_2015  \n",
       "275                        None         AEX31246.1  neumann_centromeres_2015  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e013dff-d41c-4f39-8cb7-f98710b5728d",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "1cd05b86-9882-4ea9-8637-5041349c6be2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>neumann_stretching_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        id title   doi author  year pubmed_id\n",
       "0  neumann_stretching_2012  None  None   None  None      None"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pid = \"neumann_stretching_2012\"\n",
    "query = f\"SELECT * FROM publication WHERE id='{pid}'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "e6ebfe31-9a32-4795-8531-b77f63e46690",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "f51270a5-ebb6-4d70-ab36-2eb9c93c8a6b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>AEX31245.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3888.0</td>\n",
       "      <td>Pisum sativum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKHFPSPSKPAASDNLGKKKRRCKPGTKALREIRKFQKDVKLL...</td>\n",
       "      <td>None</td>\n",
       "      <td>AEX31245.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>275</th>\n",
       "      <td>AEX31245.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3888.0</td>\n",
       "      <td>Pisum sativum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRVKHFPSPSKPAASDNLGKKKRRCKPGTKALREIRKFQKDVKLL...</td>\n",
       "      <td>None</td>\n",
       "      <td>AEX31245.1</td>\n",
       "      <td>neumann_stretching_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>AEX31246.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3888.0</td>\n",
       "      <td>Pisum sativum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARENQERKKRRNKPGTVALREIKKLQKTFQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AEX31246.1</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>AEX31246.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3888.0</td>\n",
       "      <td>Pisum sativum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARVKQTPRHARENQERKKRRNKPGTVALREIKKLQKTFQLLIPYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AEX31246.1</td>\n",
       "      <td>neumann_stretching_2012</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession         variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "274  AEX31245.1  cenH3_(Plants)  None         None           None   \n",
       "275  AEX31245.1  cenH3_(Plants)  None         None           None   \n",
       "276  AEX31246.1  cenH3_(Plants)  None         None           None   \n",
       "277  AEX31246.1  cenH3_(Plants)  None         None           None   \n",
       "\n",
       "     taxonomy_id       organism        phylum          class taxonomy_group  \\\n",
       "274       3888.0  Pisum sativum  Streptophyta  Magnoliopsida           None   \n",
       "275       3888.0  Pisum sativum  Streptophyta  Magnoliopsida           None   \n",
       "276       3888.0  Pisum sativum  Streptophyta  Magnoliopsida           None   \n",
       "277       3888.0  Pisum sativum  Streptophyta  Magnoliopsida           None   \n",
       "\n",
       "     info                                           sequence  \\\n",
       "274  None  MGRVKHFPSPSKPAASDNLGKKKRRCKPGTKALREIRKFQKDVKLL...   \n",
       "275  None  MGRVKHFPSPSKPAASDNLGKKKRRCKPGTKALREIRKFQKDVKLL...   \n",
       "276  None  MARVKQTPRHARENQERKKRRNKPGTVALREIKKLQKTFQLLIPYA...   \n",
       "277  None  MARVKQTPRHARENQERKKRRNKPGTVALREIKKLQKTFQLLIPYA...   \n",
       "\n",
       "    variant_under_consideration sequence_accession            publication_id  \n",
       "274                        None         AEX31245.1  neumann_centromeres_2015  \n",
       "275                        None         AEX31245.1   neumann_stretching_2012  \n",
       "276                        None         AEX31246.1  neumann_centromeres_2015  \n",
       "277                        None         AEX31246.1   neumann_stretching_2012  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "8d43abcb-79d2-46ee-90cf-922b19be5740",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f866cc8e-3f48-4473-b920-4a0b4460c5d5",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add sequences from `kawabe_duplication_2006`</span>\n",
    "Protein Accessions:\n",
    "ABE27630.1\n",
    "ABE27629.1\n",
    "ABE27628.1\n",
    "ABE27627.1\n",
    "ABE27626.1\n",
    "ABE27625.1\n",
    "ABE27624.1\n",
    "ABE27623.1\n",
    "ABE27622.1\n",
    "ABE27621.1\n",
    "ABE27620.1\n",
    "ABE27619.1\n",
    "ABE27618.1\n",
    "ABE27617.1\n",
    "ABE27616.1\n",
    "ABE27615.1\n",
    "ABE27614.1\n",
    "ABE27613.1\n",
    "ABE27612.1\n",
    "ABE27611.1\n",
    "ABE27610.1\n",
    "ABE27609.1\n",
    "ABE27608.1\n",
    "ABE27607.1\n",
    "ABE27606.1\n",
    "ABE27605.1\n",
    "ABE27604.1\n",
    "ABE27603.1\n",
    "ABE27602.1\n",
    "ABE27601.1\n",
    "ABE27600.1\n",
    "ABE27662.1\n",
    "ABE27661.1\n",
    "ABE27660.1\n",
    "ABE27659.1\n",
    "ABE27658.1\n",
    "ABE27657.1\n",
    "ABE27656.1\n",
    "ABE27655.1\n",
    "ABE27654.1\n",
    "ABE27653.1\n",
    "ABE27652.1\n",
    "ABE27651.1\n",
    "ABE27650.1\n",
    "ABE27649.1\n",
    "ABE27648.1\n",
    "ABE27647.1\n",
    "ABE27646.1\n",
    "ABE27645.1\n",
    "ABE27644.1\n",
    "ABE27643.1\n",
    "ABE27642.1\n",
    "ABE27641.1\n",
    "ABE27640.1\n",
    "ABE27639.1\n",
    "ABE27638.1\n",
    "ABE27637.1\n",
    "ABE27636.1\n",
    "ABE27635.1\n",
    "ABE27634.1\n",
    "ABE27633.1\n",
    "ABE27632.1\n",
    "ABE27631.1\n",
    "BAC79432.1\n",
    "BAC79431.1\n",
    "BAC79430.1\n",
    "BAC79429.1\n",
    "BAC79428.1\n",
    "BAC79427.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "265657cb-53de-4433-9563-adfcc535ceb1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\n",
    "    \"ABE27630.1\",\n",
    "    \"ABE27629.1\",\n",
    "    \"ABE27628.1\",\n",
    "    \"ABE27627.1\",\n",
    "    \"ABE27626.1\",\n",
    "    \"ABE27625.1\",\n",
    "    \"ABE27624.1\",\n",
    "    \"ABE27623.1\",\n",
    "    \"ABE27622.1\",\n",
    "    \"ABE27621.1\",\n",
    "    \"ABE27620.1\",\n",
    "    \"ABE27619.1\",\n",
    "    \"ABE27618.1\",\n",
    "    \"ABE27617.1\",\n",
    "    \"ABE27616.1\",\n",
    "    \"ABE27615.1\",\n",
    "    \"ABE27614.1\",\n",
    "    \"ABE27613.1\",\n",
    "    \"ABE27612.1\",\n",
    "    \"ABE27611.1\",\n",
    "    \"ABE27610.1\",\n",
    "    \"ABE27609.1\",\n",
    "    \"ABE27608.1\",\n",
    "    \"ABE27607.1\",\n",
    "    \"ABE27606.1\",\n",
    "    \"ABE27605.1\",\n",
    "    \"ABE27604.1\",\n",
    "    \"ABE27603.1\",\n",
    "    \"ABE27602.1\",\n",
    "    \"ABE27601.1\",\n",
    "    \"ABE27600.1\",\n",
    "    \"ABE27662.1\",\n",
    "    \"ABE27661.1\",\n",
    "    \"ABE27660.1\",\n",
    "    \"ABE27659.1\",\n",
    "    \"ABE27658.1\",\n",
    "    \"ABE27657.1\",\n",
    "    \"ABE27656.1\",\n",
    "    \"ABE27655.1\",\n",
    "    \"ABE27654.1\",\n",
    "    \"ABE27653.1\",\n",
    "    \"ABE27652.1\",\n",
    "    \"ABE27651.1\",\n",
    "    \"ABE27650.1\",\n",
    "    \"ABE27649.1\",\n",
    "    \"ABE27648.1\",\n",
    "    \"ABE27647.1\",\n",
    "    \"ABE27646.1\",\n",
    "    \"ABE27645.1\",\n",
    "    \"ABE27644.1\",\n",
    "    \"ABE27643.1\",\n",
    "    \"ABE27642.1\",\n",
    "    \"ABE27641.1\",\n",
    "    \"ABE27640.1\",\n",
    "    \"ABE27639.1\",\n",
    "    \"ABE27638.1\",\n",
    "    \"ABE27637.1\",\n",
    "    \"ABE27636.1\",\n",
    "    \"ABE27635.1\",\n",
    "    \"ABE27634.1\",\n",
    "    \"ABE27633.1\",\n",
    "    \"ABE27632.1\",\n",
    "    \"ABE27631.1\",\n",
    "    \"BAC79432.1\",\n",
    "    \"BAC79431.1\",\n",
    "    \"BAC79430.1\",\n",
    "    \"BAC79429.1\",\n",
    "    \"BAC79428.1\",\n",
    "    \"BAC79427.1\",\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c28e86cc-3ff2-498f-a513-2dd8cac6c0f4",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b09966f0-f7f9-4aff-87b0-526d06e51d48",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "83817b83-85c4-4e9d-bd3c-811ee712fc50",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 81972\n",
      "Fetched taxid from NCBI 81972\n",
      "Fetched taxid from NCBI 81972\n",
      "Fetched taxid from NCBI 81972\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 81971\n",
      "Fetched taxid from NCBI 81972\n",
      "Fetched taxid from NCBI 81972\n",
      "Fetched taxid from NCBI 81972\n",
      "Fetched taxid from NCBI 81972\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 59691\n",
      "Fetched taxid from NCBI 81971\n",
      "Fetched taxid from NCBI 81971\n",
      "Fetched taxid from NCBI 63678\n",
      "Fetched taxid from NCBI 63677\n",
      "Fetched taxid from NCBI 63677\n",
      "Fetched taxid from NCBI 63677\n",
      "Fetched taxid from NCBI 63677\n",
      "Fetched taxid from NCBI 3702\n",
      "accession BAC79427.1 <class 'str'>\n",
      "variant cenH3_(Plants) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 3702 <class 'int'>\n",
      "organism Arabidopsis thaliana <class 'str'>\n",
      "phylum Streptophyta <class 'str'>\n",
      "class Magnoliopsida <class 'str'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MARTKHRVTRSQPRNQTDAAGASSSQAAGPTTTPTRRGGEGGDNTQQTNPTTSPATGTRRGAKRSRQAMPRGSQKKSYRYRPGTVALKEIRHFQKQTNLLIPAASFIREVRSITHMLAPPQINRWTAEALVALQEAAEDYLVGLFSDSMLCAIHARRVTLMRKDFELARRLGGKGRPW <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for a in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=a, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"cenH3_(Plants)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "4f24b41a-24f0-4b08-a8ea-06e13ce192ea",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(69, 69)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(accessions), len(data_sequence_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "81b247ab-ab53-45f3-a308-13688aaf9fc0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "1eed9235-390f-4c1e-8ff0-8a11b3e08899",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>ABE27600.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>81971.0</td>\n",
       "      <td>Arabidopsis halleri subsp. halleri</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHFVTRKGSGNRTDFDANASSSQAAGPTKTPTTRGTEGGDNT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>ABE27601.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>59691.0</td>\n",
       "      <td>Arabidopsis lyrata subsp. petraea</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHFATRTGSGNRTDANASSSQAAGPTTTPTTRGSEGGDNTQQ...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>ABE27602.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>59691.0</td>\n",
       "      <td>Arabidopsis lyrata subsp. petraea</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHFATRSGSGNRTDANASSSQAAGPTTTPTTRGSEGGDNTQQ...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>ABE27603.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>59691.0</td>\n",
       "      <td>Arabidopsis lyrata subsp. petraea</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHFATRTGSGNRTDANASSSQAAGPTTTPTTRGSEGGDNTQQ...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>ABE27604.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>59691.0</td>\n",
       "      <td>Arabidopsis lyrata subsp. petraea</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHFATRSGSGNRTDANASSSQAAGPTTTPTTRGSEGGDNTQQ...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>601</th>\n",
       "      <td>BAC79428.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>63677.0</td>\n",
       "      <td>Arabidopsis halleri subsp. gemmifera</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHFVTRKGSGNRTDFDANASSSQAAGPTKTPTTRGTEGGDNT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>602</th>\n",
       "      <td>BAC79429.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>63677.0</td>\n",
       "      <td>Arabidopsis halleri subsp. gemmifera</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHFVTRKGSGNRTDFDANASSSQAAGPTKTPTTRGTEGGDNT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>603</th>\n",
       "      <td>BAC79430.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>63677.0</td>\n",
       "      <td>Arabidopsis halleri subsp. gemmifera</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHFAIKSRSGNRTDANASSSQAAGPTTTPTTRGTEGGDNTQQ...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604</th>\n",
       "      <td>BAC79431.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>63677.0</td>\n",
       "      <td>Arabidopsis halleri subsp. gemmifera</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHFAIKSRSGNRTDANASSSQAAGPTTTPTTRGTEGGDNTQQ...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>605</th>\n",
       "      <td>BAC79432.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>63678.0</td>\n",
       "      <td>Turritis glabra</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHFATRSRPRNQTDSSSQAAGPSTNPTTGGSEGGDAAQQTTP...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>69 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession         variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "62   ABE27600.1  cenH3_(Plants)  None         None           None   \n",
       "63   ABE27601.1  cenH3_(Plants)  None         None           None   \n",
       "64   ABE27602.1  cenH3_(Plants)  None         None           None   \n",
       "65   ABE27603.1  cenH3_(Plants)  None         None           None   \n",
       "66   ABE27604.1  cenH3_(Plants)  None         None           None   \n",
       "..          ...             ...   ...          ...            ...   \n",
       "601  BAC79428.1  cenH3_(Plants)  None         None           None   \n",
       "602  BAC79429.1  cenH3_(Plants)  None         None           None   \n",
       "603  BAC79430.1  cenH3_(Plants)  None         None           None   \n",
       "604  BAC79431.1  cenH3_(Plants)  None         None           None   \n",
       "605  BAC79432.1  cenH3_(Plants)  None         None           None   \n",
       "\n",
       "     taxonomy_id                              organism        phylum  \\\n",
       "62       81971.0    Arabidopsis halleri subsp. halleri  Streptophyta   \n",
       "63       59691.0     Arabidopsis lyrata subsp. petraea  Streptophyta   \n",
       "64       59691.0     Arabidopsis lyrata subsp. petraea  Streptophyta   \n",
       "65       59691.0     Arabidopsis lyrata subsp. petraea  Streptophyta   \n",
       "66       59691.0     Arabidopsis lyrata subsp. petraea  Streptophyta   \n",
       "..           ...                                   ...           ...   \n",
       "601      63677.0  Arabidopsis halleri subsp. gemmifera  Streptophyta   \n",
       "602      63677.0  Arabidopsis halleri subsp. gemmifera  Streptophyta   \n",
       "603      63677.0  Arabidopsis halleri subsp. gemmifera  Streptophyta   \n",
       "604      63677.0  Arabidopsis halleri subsp. gemmifera  Streptophyta   \n",
       "605      63678.0                       Turritis glabra  Streptophyta   \n",
       "\n",
       "             class taxonomy_group  info  \\\n",
       "62   Magnoliopsida           None  None   \n",
       "63   Magnoliopsida           None  None   \n",
       "64   Magnoliopsida           None  None   \n",
       "65   Magnoliopsida           None  None   \n",
       "66   Magnoliopsida           None  None   \n",
       "..             ...            ...   ...   \n",
       "601  Magnoliopsida           None  None   \n",
       "602  Magnoliopsida           None  None   \n",
       "603  Magnoliopsida           None  None   \n",
       "604  Magnoliopsida           None  None   \n",
       "605  Magnoliopsida           None  None   \n",
       "\n",
       "                                              sequence  \\\n",
       "62   MARTKHFVTRKGSGNRTDFDANASSSQAAGPTKTPTTRGTEGGDNT...   \n",
       "63   MARTKHFATRTGSGNRTDANASSSQAAGPTTTPTTRGSEGGDNTQQ...   \n",
       "64   MARTKHFATRSGSGNRTDANASSSQAAGPTTTPTTRGSEGGDNTQQ...   \n",
       "65   MARTKHFATRTGSGNRTDANASSSQAAGPTTTPTTRGSEGGDNTQQ...   \n",
       "66   MARTKHFATRSGSGNRTDANASSSQAAGPTTTPTTRGSEGGDNTQQ...   \n",
       "..                                                 ...   \n",
       "601  MARTKHFVTRKGSGNRTDFDANASSSQAAGPTKTPTTRGTEGGDNT...   \n",
       "602  MARTKHFVTRKGSGNRTDFDANASSSQAAGPTKTPTTRGTEGGDNT...   \n",
       "603  MARTKHFAIKSRSGNRTDANASSSQAAGPTTTPTTRGTEGGDNTQQ...   \n",
       "604  MARTKHFAIKSRSGNRTDANASSSQAAGPTTTPTTRGTEGGDNTQQ...   \n",
       "605  MARTKHFATRSRPRNQTDSSSQAAGPSTNPTTGGSEGGDAAQQTTP...   \n",
       "\n",
       "    variant_under_consideration  \n",
       "62                         None  \n",
       "63                         None  \n",
       "64                         None  \n",
       "65                         None  \n",
       "66                         None  \n",
       "..                          ...  \n",
       "601                        None  \n",
       "602                        None  \n",
       "603                        None  \n",
       "604                        None  \n",
       "605                        None  \n",
       "\n",
       "[69 rows x 13 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "0292c465-a3f4-4394-9cd6-aaa43a6bc1fd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cf1ab26-b093-441d-bd0a-2739591fa343",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "0f9a8212-2b30-4367-ad3b-5c3e9c8959c6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>kawabe_duplication_2006</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        id title   doi author  year pubmed_id\n",
       "0  kawabe_duplication_2006  None  None   None  None      None"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pid = \"kawabe_duplication_2006\"\n",
    "query = f\"SELECT * FROM publication WHERE id='{pid}'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "67deba89-3911-4545-a847-2883b74ee274",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "586d4d06-4064-48bb-9857-a62328ed70d2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>ABE27600.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>81971.0</td>\n",
       "      <td>Arabidopsis halleri subsp. halleri</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHFVTRKGSGNRTDFDANASSSQAAGPTKTPTTRGTEGGDNT...</td>\n",
       "      <td>None</td>\n",
       "      <td>ABE27600.1</td>\n",
       "      <td>kawabe_duplication_2006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>ABE27601.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>59691.0</td>\n",
       "      <td>Arabidopsis lyrata subsp. petraea</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHFATRTGSGNRTDANASSSQAAGPTTTPTTRGSEGGDNTQQ...</td>\n",
       "      <td>None</td>\n",
       "      <td>ABE27601.1</td>\n",
       "      <td>kawabe_duplication_2006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>ABE27602.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>59691.0</td>\n",
       "      <td>Arabidopsis lyrata subsp. petraea</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHFATRSGSGNRTDANASSSQAAGPTTTPTTRGSEGGDNTQQ...</td>\n",
       "      <td>None</td>\n",
       "      <td>ABE27602.1</td>\n",
       "      <td>kawabe_duplication_2006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>ABE27603.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>59691.0</td>\n",
       "      <td>Arabidopsis lyrata subsp. petraea</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHFATRTGSGNRTDANASSSQAAGPTTTPTTRGSEGGDNTQQ...</td>\n",
       "      <td>None</td>\n",
       "      <td>ABE27603.1</td>\n",
       "      <td>kawabe_duplication_2006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>ABE27604.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>59691.0</td>\n",
       "      <td>Arabidopsis lyrata subsp. petraea</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHFATRSGSGNRTDANASSSQAAGPTTTPTTRGSEGGDNTQQ...</td>\n",
       "      <td>None</td>\n",
       "      <td>ABE27604.1</td>\n",
       "      <td>kawabe_duplication_2006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>736</th>\n",
       "      <td>BAC79428.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>63677.0</td>\n",
       "      <td>Arabidopsis halleri subsp. gemmifera</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHFVTRKGSGNRTDFDANASSSQAAGPTKTPTTRGTEGGDNT...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAC79428.1</td>\n",
       "      <td>kawabe_duplication_2006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>737</th>\n",
       "      <td>BAC79429.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>63677.0</td>\n",
       "      <td>Arabidopsis halleri subsp. gemmifera</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHFVTRKGSGNRTDFDANASSSQAAGPTKTPTTRGTEGGDNT...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAC79429.1</td>\n",
       "      <td>kawabe_duplication_2006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>738</th>\n",
       "      <td>BAC79430.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>63677.0</td>\n",
       "      <td>Arabidopsis halleri subsp. gemmifera</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHFAIKSRSGNRTDANASSSQAAGPTTTPTTRGTEGGDNTQQ...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAC79430.1</td>\n",
       "      <td>kawabe_duplication_2006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>739</th>\n",
       "      <td>BAC79431.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>63677.0</td>\n",
       "      <td>Arabidopsis halleri subsp. gemmifera</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHFAIKSRSGNRTDANASSSQAAGPTTTPTTRGTEGGDNTQQ...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAC79431.1</td>\n",
       "      <td>kawabe_duplication_2006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>740</th>\n",
       "      <td>BAC79432.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>63678.0</td>\n",
       "      <td>Turritis glabra</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHFATRSRPRNQTDSSSQAAGPSTNPTTGGSEGGDAAQQTTP...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAC79432.1</td>\n",
       "      <td>kawabe_duplication_2006</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>69 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession         variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "79   ABE27600.1  cenH3_(Plants)  None         None           None   \n",
       "80   ABE27601.1  cenH3_(Plants)  None         None           None   \n",
       "81   ABE27602.1  cenH3_(Plants)  None         None           None   \n",
       "82   ABE27603.1  cenH3_(Plants)  None         None           None   \n",
       "83   ABE27604.1  cenH3_(Plants)  None         None           None   \n",
       "..          ...             ...   ...          ...            ...   \n",
       "736  BAC79428.1  cenH3_(Plants)  None         None           None   \n",
       "737  BAC79429.1  cenH3_(Plants)  None         None           None   \n",
       "738  BAC79430.1  cenH3_(Plants)  None         None           None   \n",
       "739  BAC79431.1  cenH3_(Plants)  None         None           None   \n",
       "740  BAC79432.1  cenH3_(Plants)  None         None           None   \n",
       "\n",
       "     taxonomy_id                              organism        phylum  \\\n",
       "79       81971.0    Arabidopsis halleri subsp. halleri  Streptophyta   \n",
       "80       59691.0     Arabidopsis lyrata subsp. petraea  Streptophyta   \n",
       "81       59691.0     Arabidopsis lyrata subsp. petraea  Streptophyta   \n",
       "82       59691.0     Arabidopsis lyrata subsp. petraea  Streptophyta   \n",
       "83       59691.0     Arabidopsis lyrata subsp. petraea  Streptophyta   \n",
       "..           ...                                   ...           ...   \n",
       "736      63677.0  Arabidopsis halleri subsp. gemmifera  Streptophyta   \n",
       "737      63677.0  Arabidopsis halleri subsp. gemmifera  Streptophyta   \n",
       "738      63677.0  Arabidopsis halleri subsp. gemmifera  Streptophyta   \n",
       "739      63677.0  Arabidopsis halleri subsp. gemmifera  Streptophyta   \n",
       "740      63678.0                       Turritis glabra  Streptophyta   \n",
       "\n",
       "             class taxonomy_group  info  \\\n",
       "79   Magnoliopsida           None  None   \n",
       "80   Magnoliopsida           None  None   \n",
       "81   Magnoliopsida           None  None   \n",
       "82   Magnoliopsida           None  None   \n",
       "83   Magnoliopsida           None  None   \n",
       "..             ...            ...   ...   \n",
       "736  Magnoliopsida           None  None   \n",
       "737  Magnoliopsida           None  None   \n",
       "738  Magnoliopsida           None  None   \n",
       "739  Magnoliopsida           None  None   \n",
       "740  Magnoliopsida           None  None   \n",
       "\n",
       "                                              sequence  \\\n",
       "79   MARTKHFVTRKGSGNRTDFDANASSSQAAGPTKTPTTRGTEGGDNT...   \n",
       "80   MARTKHFATRTGSGNRTDANASSSQAAGPTTTPTTRGSEGGDNTQQ...   \n",
       "81   MARTKHFATRSGSGNRTDANASSSQAAGPTTTPTTRGSEGGDNTQQ...   \n",
       "82   MARTKHFATRTGSGNRTDANASSSQAAGPTTTPTTRGSEGGDNTQQ...   \n",
       "83   MARTKHFATRSGSGNRTDANASSSQAAGPTTTPTTRGSEGGDNTQQ...   \n",
       "..                                                 ...   \n",
       "736  MARTKHFVTRKGSGNRTDFDANASSSQAAGPTKTPTTRGTEGGDNT...   \n",
       "737  MARTKHFVTRKGSGNRTDFDANASSSQAAGPTKTPTTRGTEGGDNT...   \n",
       "738  MARTKHFAIKSRSGNRTDANASSSQAAGPTTTPTTRGTEGGDNTQQ...   \n",
       "739  MARTKHFAIKSRSGNRTDANASSSQAAGPTTTPTTRGTEGGDNTQQ...   \n",
       "740  MARTKHFATRSRPRNQTDSSSQAAGPSTNPTTGGSEGGDAAQQTTP...   \n",
       "\n",
       "    variant_under_consideration sequence_accession           publication_id  \n",
       "79                         None         ABE27600.1  kawabe_duplication_2006  \n",
       "80                         None         ABE27601.1  kawabe_duplication_2006  \n",
       "81                         None         ABE27602.1  kawabe_duplication_2006  \n",
       "82                         None         ABE27603.1  kawabe_duplication_2006  \n",
       "83                         None         ABE27604.1  kawabe_duplication_2006  \n",
       "..                          ...                ...                      ...  \n",
       "736                        None         BAC79428.1  kawabe_duplication_2006  \n",
       "737                        None         BAC79429.1  kawabe_duplication_2006  \n",
       "738                        None         BAC79430.1  kawabe_duplication_2006  \n",
       "739                        None         BAC79431.1  kawabe_duplication_2006  \n",
       "740                        None         BAC79432.1  kawabe_duplication_2006  \n",
       "\n",
       "[69 rows x 15 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "36ca12cb-0873-46d3-a012-977aceac5b8e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dba544b-3619-4593-8c70-932b41a28f73",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add sequences from `evtushenko_conserved_2017`</span>\n",
    "\n",
    "Protein Accessions: AUN88474.1\n",
    "AUN88473.1\n",
    "AUN88472.1\n",
    "AUN88471.1\n",
    "AUN88470.1\n",
    "AUN88469.1\n",
    "AUN88468.1\n",
    "AUN88467.1\n",
    "AUN88466.1\n",
    "AUN88465.1\n",
    "AUN88464.1\n",
    "AUN88463.1\n",
    "AUN88462.1\n",
    "AUN88461.1\n",
    "AUN88460.1\n",
    "AUN88459.1\n",
    "AUN88458.1\n",
    "AUN88457.1\n",
    "AUN88456.1\n",
    "AUN88455.1\n",
    "AUN88454.1\n",
    "AUN88453.1\n",
    "AUN88452.1\n",
    "AUN88451.1\n",
    "AUN88450.1\n",
    "AUN88449.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "78bddd33-3a75-4693-a582-024e8b3a11b8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\n",
    "    \"AUN88474.1\",\n",
    "    \"AUN88473.1\",\n",
    "    \"AUN88472.1\",\n",
    "    \"AUN88471.1\",\n",
    "    \"AUN88470.1\",\n",
    "    \"AUN88469.1\",\n",
    "    \"AUN88468.1\",\n",
    "    \"AUN88467.1\",\n",
    "    \"AUN88466.1\",\n",
    "    \"AUN88465.1\",\n",
    "    \"AUN88464.1\",\n",
    "    \"AUN88463.1\",\n",
    "    \"AUN88462.1\",\n",
    "    \"AUN88461.1\",\n",
    "    \"AUN88460.1\",\n",
    "    \"AUN88459.1\",\n",
    "    \"AUN88458.1\",\n",
    "    \"AUN88457.1\",\n",
    "    \"AUN88456.1\",\n",
    "    \"AUN88455.1\",\n",
    "    \"AUN88454.1\",\n",
    "    \"AUN88453.1\",\n",
    "    \"AUN88452.1\",\n",
    "    \"AUN88451.1\",\n",
    "    \"AUN88450.1\",\n",
    "    \"AUN88449.1\",\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17d0fd02-b4af-4d4c-a621-c5fda8324535",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "f95c01a0-3e5e-4dcc-be9a-89721153b45a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "d3c3132c-2bdc-45b4-8b5c-8059a0a75f10",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 155090\n",
      "Fetched taxid from NCBI 155101\n",
      "Fetched taxid from NCBI 155093\n",
      "Fetched taxid from NCBI 155089\n",
      "Fetched taxid from NCBI 4553\n",
      "Fetched taxid from NCBI 4552\n",
      "Fetched taxid from NCBI 155095\n",
      "Fetched taxid from NCBI 100817\n",
      "Fetched taxid from NCBI 4550\n",
      "Fetched taxid from NCBI 37731\n",
      "Fetched taxid from NCBI 155091\n",
      "Fetched taxid from NCBI 155095\n",
      "Fetched taxid from NCBI 37731\n",
      "Fetched taxid from NCBI 155091\n",
      "Fetched taxid from NCBI 155091\n",
      "Fetched taxid from NCBI 4553\n",
      "Fetched taxid from NCBI 4550\n",
      "Fetched taxid from NCBI 155090\n",
      "Fetched taxid from NCBI 100817\n",
      "Fetched taxid from NCBI 155093\n",
      "Fetched taxid from NCBI 4552\n",
      "Fetched taxid from NCBI 155095\n",
      "Fetched taxid from NCBI 155089\n",
      "Fetched taxid from NCBI 155101\n",
      "Fetched taxid from NCBI 37731\n",
      "Fetched taxid from NCBI 4550\n",
      "accession AUN88449.1 <class 'str'>\n",
      "variant cenH3_(Plants) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 4550 <class 'int'>\n",
      "organism Secale cereale <class 'str'>\n",
      "phylum Streptophyta <class 'str'>\n",
      "class Magnoliopsida <class 'str'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MARTKHPAVRKTKAPPKKQLGPRPAQRRQETDGAGTSATPRRAGRAAAPGGAQGATGQPKQRKPHRFRPGTVALREIRRYQKSVDFLIPFAPFVRLIKEVTDFFCPEISRWTPQALVAIQEAAEYHLVDVFERANHCAIHAKRVTVMQKDIQLARRIGGRRLW <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for a in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=a, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"cenH3_(Plants)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "07ac2a0e-8e0d-479e-aa27-a1be84fae464",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(26, 26)"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(accessions), len(data_sequence_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "72e5dab4-c85c-4670-b870-df536096ab1a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "cd6f9c2c-e25d-4559-bb16-7b3552489f58",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>555</th>\n",
       "      <td>AUN88449.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4550.0</td>\n",
       "      <td>Secale cereale</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKAPPKKQLGPRPAQRRQETDGAGTSATPRRAGRA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>556</th>\n",
       "      <td>AUN88450.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>37731.0</td>\n",
       "      <td>Secale strictum subsp. africanum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>557</th>\n",
       "      <td>AUN88451.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155101.0</td>\n",
       "      <td>Secale strictum subsp. anatolicum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>558</th>\n",
       "      <td>AUN88452.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155089.0</td>\n",
       "      <td>Secale cereale subsp. ancestrale</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>559</th>\n",
       "      <td>AUN88453.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155095.0</td>\n",
       "      <td>Secale strictum subsp. strictum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>560</th>\n",
       "      <td>AUN88454.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4552.0</td>\n",
       "      <td>Secale sylvestre</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>561</th>\n",
       "      <td>AUN88455.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155093.0</td>\n",
       "      <td>Secale cereale subsp. dighoricum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562</th>\n",
       "      <td>AUN88456.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100817.0</td>\n",
       "      <td>Secale strictum subsp. kuprijanovii</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>563</th>\n",
       "      <td>AUN88457.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155090.0</td>\n",
       "      <td>Secale cereale subsp. segetale</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564</th>\n",
       "      <td>AUN88458.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4550.0</td>\n",
       "      <td>Secale cereale</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>AUN88459.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4553.0</td>\n",
       "      <td>Secale cereale subsp. vavilovii</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>AUN88460.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155091.0</td>\n",
       "      <td>Secale cereale subsp. afghanicum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567</th>\n",
       "      <td>AUN88461.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155091.0</td>\n",
       "      <td>Secale cereale subsp. afghanicum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTTTTETKKRLRFELSPRWRPPPPMRQVPPEPQP...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>AUN88462.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>37731.0</td>\n",
       "      <td>Secale strictum subsp. africanum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTTTTETKKRLRFELSPRWRPPPPMRQVPPEPQP...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>569</th>\n",
       "      <td>AUN88463.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155095.0</td>\n",
       "      <td>Secale strictum subsp. strictum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTTTTETKKRLRFELSPRWRPPPPMRQVPPEPQP...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>570</th>\n",
       "      <td>AUN88464.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155091.0</td>\n",
       "      <td>Secale cereale subsp. afghanicum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571</th>\n",
       "      <td>AUN88465.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>37731.0</td>\n",
       "      <td>Secale strictum subsp. africanum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>572</th>\n",
       "      <td>AUN88466.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4550.0</td>\n",
       "      <td>Secale cereale</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>573</th>\n",
       "      <td>AUN88467.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100817.0</td>\n",
       "      <td>Secale strictum subsp. kuprijanovii</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>574</th>\n",
       "      <td>AUN88468.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155095.0</td>\n",
       "      <td>Secale strictum subsp. strictum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>575</th>\n",
       "      <td>AUN88469.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4552.0</td>\n",
       "      <td>Secale sylvestre</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPETKKRLRFELSPRWRPPPPLRQVPPQPQPEK...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>576</th>\n",
       "      <td>AUN88470.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4553.0</td>\n",
       "      <td>Secale cereale subsp. vavilovii</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>577</th>\n",
       "      <td>AUN88471.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155089.0</td>\n",
       "      <td>Secale cereale subsp. ancestrale</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>578</th>\n",
       "      <td>AUN88472.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155093.0</td>\n",
       "      <td>Secale cereale subsp. dighoricum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>579</th>\n",
       "      <td>AUN88473.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155101.0</td>\n",
       "      <td>Secale strictum subsp. anatolicum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>AUN88474.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155090.0</td>\n",
       "      <td>Secale cereale subsp. segetale</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession         variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "555  AUN88449.1  cenH3_(Plants)  None         None           None   \n",
       "556  AUN88450.1  cenH3_(Plants)  None         None           None   \n",
       "557  AUN88451.1  cenH3_(Plants)  None         None           None   \n",
       "558  AUN88452.1  cenH3_(Plants)  None         None           None   \n",
       "559  AUN88453.1  cenH3_(Plants)  None         None           None   \n",
       "560  AUN88454.1  cenH3_(Plants)  None         None           None   \n",
       "561  AUN88455.1  cenH3_(Plants)  None         None           None   \n",
       "562  AUN88456.1  cenH3_(Plants)  None         None           None   \n",
       "563  AUN88457.1  cenH3_(Plants)  None         None           None   \n",
       "564  AUN88458.1  cenH3_(Plants)  None         None           None   \n",
       "565  AUN88459.1  cenH3_(Plants)  None         None           None   \n",
       "566  AUN88460.1  cenH3_(Plants)  None         None           None   \n",
       "567  AUN88461.1  cenH3_(Plants)  None         None           None   \n",
       "568  AUN88462.1  cenH3_(Plants)  None         None           None   \n",
       "569  AUN88463.1  cenH3_(Plants)  None         None           None   \n",
       "570  AUN88464.1  cenH3_(Plants)  None         None           None   \n",
       "571  AUN88465.1  cenH3_(Plants)  None         None           None   \n",
       "572  AUN88466.1  cenH3_(Plants)  None         None           None   \n",
       "573  AUN88467.1  cenH3_(Plants)  None         None           None   \n",
       "574  AUN88468.1  cenH3_(Plants)  None         None           None   \n",
       "575  AUN88469.1  cenH3_(Plants)  None         None           None   \n",
       "576  AUN88470.1  cenH3_(Plants)  None         None           None   \n",
       "577  AUN88471.1  cenH3_(Plants)  None         None           None   \n",
       "578  AUN88472.1  cenH3_(Plants)  None         None           None   \n",
       "579  AUN88473.1  cenH3_(Plants)  None         None           None   \n",
       "580  AUN88474.1  cenH3_(Plants)  None         None           None   \n",
       "\n",
       "     taxonomy_id                             organism        phylum  \\\n",
       "555       4550.0                       Secale cereale  Streptophyta   \n",
       "556      37731.0     Secale strictum subsp. africanum  Streptophyta   \n",
       "557     155101.0    Secale strictum subsp. anatolicum  Streptophyta   \n",
       "558     155089.0     Secale cereale subsp. ancestrale  Streptophyta   \n",
       "559     155095.0      Secale strictum subsp. strictum  Streptophyta   \n",
       "560       4552.0                     Secale sylvestre  Streptophyta   \n",
       "561     155093.0     Secale cereale subsp. dighoricum  Streptophyta   \n",
       "562     100817.0  Secale strictum subsp. kuprijanovii  Streptophyta   \n",
       "563     155090.0       Secale cereale subsp. segetale  Streptophyta   \n",
       "564       4550.0                       Secale cereale  Streptophyta   \n",
       "565       4553.0      Secale cereale subsp. vavilovii  Streptophyta   \n",
       "566     155091.0     Secale cereale subsp. afghanicum  Streptophyta   \n",
       "567     155091.0     Secale cereale subsp. afghanicum  Streptophyta   \n",
       "568      37731.0     Secale strictum subsp. africanum  Streptophyta   \n",
       "569     155095.0      Secale strictum subsp. strictum  Streptophyta   \n",
       "570     155091.0     Secale cereale subsp. afghanicum  Streptophyta   \n",
       "571      37731.0     Secale strictum subsp. africanum  Streptophyta   \n",
       "572       4550.0                       Secale cereale  Streptophyta   \n",
       "573     100817.0  Secale strictum subsp. kuprijanovii  Streptophyta   \n",
       "574     155095.0      Secale strictum subsp. strictum  Streptophyta   \n",
       "575       4552.0                     Secale sylvestre  Streptophyta   \n",
       "576       4553.0      Secale cereale subsp. vavilovii  Streptophyta   \n",
       "577     155089.0     Secale cereale subsp. ancestrale  Streptophyta   \n",
       "578     155093.0     Secale cereale subsp. dighoricum  Streptophyta   \n",
       "579     155101.0    Secale strictum subsp. anatolicum  Streptophyta   \n",
       "580     155090.0       Secale cereale subsp. segetale  Streptophyta   \n",
       "\n",
       "             class taxonomy_group  info  \\\n",
       "555  Magnoliopsida           None  None   \n",
       "556  Magnoliopsida           None  None   \n",
       "557  Magnoliopsida           None  None   \n",
       "558  Magnoliopsida           None  None   \n",
       "559  Magnoliopsida           None  None   \n",
       "560  Magnoliopsida           None  None   \n",
       "561  Magnoliopsida           None  None   \n",
       "562  Magnoliopsida           None  None   \n",
       "563  Magnoliopsida           None  None   \n",
       "564  Magnoliopsida           None  None   \n",
       "565  Magnoliopsida           None  None   \n",
       "566  Magnoliopsida           None  None   \n",
       "567  Magnoliopsida           None  None   \n",
       "568  Magnoliopsida           None  None   \n",
       "569  Magnoliopsida           None  None   \n",
       "570  Magnoliopsida           None  None   \n",
       "571  Magnoliopsida           None  None   \n",
       "572  Magnoliopsida           None  None   \n",
       "573  Magnoliopsida           None  None   \n",
       "574  Magnoliopsida           None  None   \n",
       "575  Magnoliopsida           None  None   \n",
       "576  Magnoliopsida           None  None   \n",
       "577  Magnoliopsida           None  None   \n",
       "578  Magnoliopsida           None  None   \n",
       "579  Magnoliopsida           None  None   \n",
       "580  Magnoliopsida           None  None   \n",
       "\n",
       "                                              sequence  \\\n",
       "555  MARTKHPAVRKTKAPPKKQLGPRPAQRRQETDGAGTSATPRRAGRA...   \n",
       "556  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "557  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "558  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "559  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "560  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "561  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "562  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "563  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "564  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "565  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "566  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "567  MGRTKHAVAATATTTTTETKKRLRFELSPRWRPPPPMRQVPPEPQP...   \n",
       "568  MGRTKHAVAATATTTTTETKKRLRFELSPRWRPPPPMRQVPPEPQP...   \n",
       "569  MGRTKHAVAATATTTTTETKKRLRFELSPRWRPPPPMRQVPPEPQP...   \n",
       "570  MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...   \n",
       "571  MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...   \n",
       "572  MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...   \n",
       "573  MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...   \n",
       "574  MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...   \n",
       "575  MGRTKHAVAATATTPETKKRLRFELSPRWRPPPPLRQVPPQPQPEK...   \n",
       "576  MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...   \n",
       "577  MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...   \n",
       "578  MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...   \n",
       "579  MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...   \n",
       "580  MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...   \n",
       "\n",
       "    variant_under_consideration  \n",
       "555                        None  \n",
       "556                        None  \n",
       "557                        None  \n",
       "558                        None  \n",
       "559                        None  \n",
       "560                        None  \n",
       "561                        None  \n",
       "562                        None  \n",
       "563                        None  \n",
       "564                        None  \n",
       "565                        None  \n",
       "566                        None  \n",
       "567                        None  \n",
       "568                        None  \n",
       "569                        None  \n",
       "570                        None  \n",
       "571                        None  \n",
       "572                        None  \n",
       "573                        None  \n",
       "574                        None  \n",
       "575                        None  \n",
       "576                        None  \n",
       "577                        None  \n",
       "578                        None  \n",
       "579                        None  \n",
       "580                        None  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "0614cd34-46db-4111-9e6f-8912687c6a8e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "021e59f0-9c30-46d7-a9c4-d6c018abebda",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "69336504-3ba3-4425-917e-45bf7a5f62f7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          id title   doi author  year pubmed_id\n",
       "0  evtushenko_conserved_2017  None  None   None  None      None"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pid = \"evtushenko_conserved_2017\"\n",
    "query = f\"SELECT * FROM publication WHERE id='{pid}'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "84218787-b639-4aca-a8e1-ea6d6aaff7d1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "89a53ec5-023e-4a43-96f6-3102ecc61533",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>676</th>\n",
       "      <td>AUN88449.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4550.0</td>\n",
       "      <td>Secale cereale</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKAPPKKQLGPRPAQRRQETDGAGTSATPRRAGRA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88449.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>677</th>\n",
       "      <td>AUN88450.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>37731.0</td>\n",
       "      <td>Secale strictum subsp. africanum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88450.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>678</th>\n",
       "      <td>AUN88451.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155101.0</td>\n",
       "      <td>Secale strictum subsp. anatolicum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88451.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>679</th>\n",
       "      <td>AUN88452.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155089.0</td>\n",
       "      <td>Secale cereale subsp. ancestrale</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88452.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>680</th>\n",
       "      <td>AUN88453.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155095.0</td>\n",
       "      <td>Secale strictum subsp. strictum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88453.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>681</th>\n",
       "      <td>AUN88454.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4552.0</td>\n",
       "      <td>Secale sylvestre</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88454.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>682</th>\n",
       "      <td>AUN88455.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155093.0</td>\n",
       "      <td>Secale cereale subsp. dighoricum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88455.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>683</th>\n",
       "      <td>AUN88456.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100817.0</td>\n",
       "      <td>Secale strictum subsp. kuprijanovii</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88456.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684</th>\n",
       "      <td>AUN88457.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155090.0</td>\n",
       "      <td>Secale cereale subsp. segetale</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88457.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>685</th>\n",
       "      <td>AUN88458.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4550.0</td>\n",
       "      <td>Secale cereale</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88458.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>686</th>\n",
       "      <td>AUN88459.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4553.0</td>\n",
       "      <td>Secale cereale subsp. vavilovii</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88459.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>687</th>\n",
       "      <td>AUN88460.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155091.0</td>\n",
       "      <td>Secale cereale subsp. afghanicum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88460.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688</th>\n",
       "      <td>AUN88461.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155091.0</td>\n",
       "      <td>Secale cereale subsp. afghanicum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTTTTETKKRLRFELSPRWRPPPPMRQVPPEPQP...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88461.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>689</th>\n",
       "      <td>AUN88462.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>37731.0</td>\n",
       "      <td>Secale strictum subsp. africanum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTTTTETKKRLRFELSPRWRPPPPMRQVPPEPQP...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88462.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>690</th>\n",
       "      <td>AUN88463.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155095.0</td>\n",
       "      <td>Secale strictum subsp. strictum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTTTTETKKRLRFELSPRWRPPPPMRQVPPEPQP...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88463.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>691</th>\n",
       "      <td>AUN88464.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155091.0</td>\n",
       "      <td>Secale cereale subsp. afghanicum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88464.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692</th>\n",
       "      <td>AUN88465.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>37731.0</td>\n",
       "      <td>Secale strictum subsp. africanum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88465.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>693</th>\n",
       "      <td>AUN88466.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4550.0</td>\n",
       "      <td>Secale cereale</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88466.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>694</th>\n",
       "      <td>AUN88467.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100817.0</td>\n",
       "      <td>Secale strictum subsp. kuprijanovii</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88467.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>695</th>\n",
       "      <td>AUN88468.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155095.0</td>\n",
       "      <td>Secale strictum subsp. strictum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88468.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>696</th>\n",
       "      <td>AUN88469.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4552.0</td>\n",
       "      <td>Secale sylvestre</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPETKKRLRFELSPRWRPPPPLRQVPPQPQPEK...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88469.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>697</th>\n",
       "      <td>AUN88470.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4553.0</td>\n",
       "      <td>Secale cereale subsp. vavilovii</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88470.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>698</th>\n",
       "      <td>AUN88471.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155089.0</td>\n",
       "      <td>Secale cereale subsp. ancestrale</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88471.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>699</th>\n",
       "      <td>AUN88472.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155093.0</td>\n",
       "      <td>Secale cereale subsp. dighoricum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88472.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>700</th>\n",
       "      <td>AUN88473.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155101.0</td>\n",
       "      <td>Secale strictum subsp. anatolicum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88473.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701</th>\n",
       "      <td>AUN88474.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>155090.0</td>\n",
       "      <td>Secale cereale subsp. segetale</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...</td>\n",
       "      <td>None</td>\n",
       "      <td>AUN88474.1</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession         variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "676  AUN88449.1  cenH3_(Plants)  None         None           None   \n",
       "677  AUN88450.1  cenH3_(Plants)  None         None           None   \n",
       "678  AUN88451.1  cenH3_(Plants)  None         None           None   \n",
       "679  AUN88452.1  cenH3_(Plants)  None         None           None   \n",
       "680  AUN88453.1  cenH3_(Plants)  None         None           None   \n",
       "681  AUN88454.1  cenH3_(Plants)  None         None           None   \n",
       "682  AUN88455.1  cenH3_(Plants)  None         None           None   \n",
       "683  AUN88456.1  cenH3_(Plants)  None         None           None   \n",
       "684  AUN88457.1  cenH3_(Plants)  None         None           None   \n",
       "685  AUN88458.1  cenH3_(Plants)  None         None           None   \n",
       "686  AUN88459.1  cenH3_(Plants)  None         None           None   \n",
       "687  AUN88460.1  cenH3_(Plants)  None         None           None   \n",
       "688  AUN88461.1  cenH3_(Plants)  None         None           None   \n",
       "689  AUN88462.1  cenH3_(Plants)  None         None           None   \n",
       "690  AUN88463.1  cenH3_(Plants)  None         None           None   \n",
       "691  AUN88464.1  cenH3_(Plants)  None         None           None   \n",
       "692  AUN88465.1  cenH3_(Plants)  None         None           None   \n",
       "693  AUN88466.1  cenH3_(Plants)  None         None           None   \n",
       "694  AUN88467.1  cenH3_(Plants)  None         None           None   \n",
       "695  AUN88468.1  cenH3_(Plants)  None         None           None   \n",
       "696  AUN88469.1  cenH3_(Plants)  None         None           None   \n",
       "697  AUN88470.1  cenH3_(Plants)  None         None           None   \n",
       "698  AUN88471.1  cenH3_(Plants)  None         None           None   \n",
       "699  AUN88472.1  cenH3_(Plants)  None         None           None   \n",
       "700  AUN88473.1  cenH3_(Plants)  None         None           None   \n",
       "701  AUN88474.1  cenH3_(Plants)  None         None           None   \n",
       "\n",
       "     taxonomy_id                             organism        phylum  \\\n",
       "676       4550.0                       Secale cereale  Streptophyta   \n",
       "677      37731.0     Secale strictum subsp. africanum  Streptophyta   \n",
       "678     155101.0    Secale strictum subsp. anatolicum  Streptophyta   \n",
       "679     155089.0     Secale cereale subsp. ancestrale  Streptophyta   \n",
       "680     155095.0      Secale strictum subsp. strictum  Streptophyta   \n",
       "681       4552.0                     Secale sylvestre  Streptophyta   \n",
       "682     155093.0     Secale cereale subsp. dighoricum  Streptophyta   \n",
       "683     100817.0  Secale strictum subsp. kuprijanovii  Streptophyta   \n",
       "684     155090.0       Secale cereale subsp. segetale  Streptophyta   \n",
       "685       4550.0                       Secale cereale  Streptophyta   \n",
       "686       4553.0      Secale cereale subsp. vavilovii  Streptophyta   \n",
       "687     155091.0     Secale cereale subsp. afghanicum  Streptophyta   \n",
       "688     155091.0     Secale cereale subsp. afghanicum  Streptophyta   \n",
       "689      37731.0     Secale strictum subsp. africanum  Streptophyta   \n",
       "690     155095.0      Secale strictum subsp. strictum  Streptophyta   \n",
       "691     155091.0     Secale cereale subsp. afghanicum  Streptophyta   \n",
       "692      37731.0     Secale strictum subsp. africanum  Streptophyta   \n",
       "693       4550.0                       Secale cereale  Streptophyta   \n",
       "694     100817.0  Secale strictum subsp. kuprijanovii  Streptophyta   \n",
       "695     155095.0      Secale strictum subsp. strictum  Streptophyta   \n",
       "696       4552.0                     Secale sylvestre  Streptophyta   \n",
       "697       4553.0      Secale cereale subsp. vavilovii  Streptophyta   \n",
       "698     155089.0     Secale cereale subsp. ancestrale  Streptophyta   \n",
       "699     155093.0     Secale cereale subsp. dighoricum  Streptophyta   \n",
       "700     155101.0    Secale strictum subsp. anatolicum  Streptophyta   \n",
       "701     155090.0       Secale cereale subsp. segetale  Streptophyta   \n",
       "\n",
       "             class taxonomy_group  info  \\\n",
       "676  Magnoliopsida           None  None   \n",
       "677  Magnoliopsida           None  None   \n",
       "678  Magnoliopsida           None  None   \n",
       "679  Magnoliopsida           None  None   \n",
       "680  Magnoliopsida           None  None   \n",
       "681  Magnoliopsida           None  None   \n",
       "682  Magnoliopsida           None  None   \n",
       "683  Magnoliopsida           None  None   \n",
       "684  Magnoliopsida           None  None   \n",
       "685  Magnoliopsida           None  None   \n",
       "686  Magnoliopsida           None  None   \n",
       "687  Magnoliopsida           None  None   \n",
       "688  Magnoliopsida           None  None   \n",
       "689  Magnoliopsida           None  None   \n",
       "690  Magnoliopsida           None  None   \n",
       "691  Magnoliopsida           None  None   \n",
       "692  Magnoliopsida           None  None   \n",
       "693  Magnoliopsida           None  None   \n",
       "694  Magnoliopsida           None  None   \n",
       "695  Magnoliopsida           None  None   \n",
       "696  Magnoliopsida           None  None   \n",
       "697  Magnoliopsida           None  None   \n",
       "698  Magnoliopsida           None  None   \n",
       "699  Magnoliopsida           None  None   \n",
       "700  Magnoliopsida           None  None   \n",
       "701  Magnoliopsida           None  None   \n",
       "\n",
       "                                              sequence  \\\n",
       "676  MARTKHPAVRKTKAPPKKQLGPRPAQRRQETDGAGTSATPRRAGRA...   \n",
       "677  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "678  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "679  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "680  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "681  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "682  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "683  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "684  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "685  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "686  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "687  MARTKHPAVRKTKVPPKKKLGTRPSGGTQRRQDTDGAGTSATPRRA...   \n",
       "688  MGRTKHAVAATATTTTTETKKRLRFELSPRWRPPPPMRQVPPEPQP...   \n",
       "689  MGRTKHAVAATATTTTTETKKRLRFELSPRWRPPPPMRQVPPEPQP...   \n",
       "690  MGRTKHAVAATATTTTTETKKRLRFELSPRWRPPPPMRQVPPEPQP...   \n",
       "691  MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...   \n",
       "692  MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...   \n",
       "693  MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...   \n",
       "694  MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...   \n",
       "695  MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...   \n",
       "696  MGRTKHAVAATATTPETKKRLRFELSPRWRPPPPLRQVPPQPQPEK...   \n",
       "697  MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...   \n",
       "698  MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...   \n",
       "699  MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...   \n",
       "700  MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...   \n",
       "701  MGRTKHAVAATATTPEKKKRLRFELSPRWRPPPPLRQVPPEPQPEK...   \n",
       "\n",
       "    variant_under_consideration sequence_accession             publication_id  \n",
       "676                        None         AUN88449.1  evtushenko_conserved_2017  \n",
       "677                        None         AUN88450.1  evtushenko_conserved_2017  \n",
       "678                        None         AUN88451.1  evtushenko_conserved_2017  \n",
       "679                        None         AUN88452.1  evtushenko_conserved_2017  \n",
       "680                        None         AUN88453.1  evtushenko_conserved_2017  \n",
       "681                        None         AUN88454.1  evtushenko_conserved_2017  \n",
       "682                        None         AUN88455.1  evtushenko_conserved_2017  \n",
       "683                        None         AUN88456.1  evtushenko_conserved_2017  \n",
       "684                        None         AUN88457.1  evtushenko_conserved_2017  \n",
       "685                        None         AUN88458.1  evtushenko_conserved_2017  \n",
       "686                        None         AUN88459.1  evtushenko_conserved_2017  \n",
       "687                        None         AUN88460.1  evtushenko_conserved_2017  \n",
       "688                        None         AUN88461.1  evtushenko_conserved_2017  \n",
       "689                        None         AUN88462.1  evtushenko_conserved_2017  \n",
       "690                        None         AUN88463.1  evtushenko_conserved_2017  \n",
       "691                        None         AUN88464.1  evtushenko_conserved_2017  \n",
       "692                        None         AUN88465.1  evtushenko_conserved_2017  \n",
       "693                        None         AUN88466.1  evtushenko_conserved_2017  \n",
       "694                        None         AUN88467.1  evtushenko_conserved_2017  \n",
       "695                        None         AUN88468.1  evtushenko_conserved_2017  \n",
       "696                        None         AUN88469.1  evtushenko_conserved_2017  \n",
       "697                        None         AUN88470.1  evtushenko_conserved_2017  \n",
       "698                        None         AUN88471.1  evtushenko_conserved_2017  \n",
       "699                        None         AUN88472.1  evtushenko_conserved_2017  \n",
       "700                        None         AUN88473.1  evtushenko_conserved_2017  \n",
       "701                        None         AUN88474.1  evtushenko_conserved_2017  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "5b76cb77-f924-4888-bc0a-2bab137fe386",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae2759f3-d542-4d9a-8bbc-4f7b2196ab54",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add sequences from `monen_differential_2005` and `monen_separase_2015`</span>\n",
    "\n",
    "Protein Accessions: NP_499128.1 (hcp-3) NP_499073.1 (cpar-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "6c0fec17-6e0e-497f-a80f-b586579979e6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"NP_499128.1\", \"NP_499073.1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85f9eeff-e1ea-4f9a-ab84-b1605d7aaa89",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "0bc0d49f-1131-4238-a597-e206b879ca93",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3280</th>\n",
       "      <td>NP_499128.1</td>\n",
       "      <td>cenH3_(Animals)</td>\n",
       "      <td>17553736</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>6239.0</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>Nematoda</td>\n",
       "      <td>Chromadorea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MADDTPIIEEIAEQNESVTRIMQRLKHDMQRVTSVPGFNTSAAGVN...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession          variant        gi ncbi_gene_id hgnc_gene_name  \\\n",
       "3280  NP_499128.1  cenH3_(Animals)  17553736         None           None   \n",
       "\n",
       "      taxonomy_id                organism    phylum        class  \\\n",
       "3280       6239.0  Caenorhabditis elegans  Nematoda  Chromadorea   \n",
       "\n",
       "     taxonomy_group  info                                           sequence  \\\n",
       "3280           None  None  MADDTPIIEEIAEQNESVTRIMQRLKHDMQRVTSVPGFNTSAAGVN...   \n",
       "\n",
       "     variant_under_consideration  \n",
       "3280                        None  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "72b4a147-8a58-4efb-84bd-d284071831d7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3580</th>\n",
       "      <td>NP_499128.1</td>\n",
       "      <td>cenH3_(Animals)</td>\n",
       "      <td>17553736</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>6239.0</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>Nematoda</td>\n",
       "      <td>Chromadorea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MADDTPIIEEIAEQNESVTRIMQRLKHDMQRVTSVPGFNTSAAGVN...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession          variant        gi ncbi_gene_id hgnc_gene_name  \\\n",
       "3580  NP_499128.1  cenH3_(Animals)  17553736         None           None   \n",
       "\n",
       "      taxonomy_id                organism    phylum        class  \\\n",
       "3580       6239.0  Caenorhabditis elegans  Nematoda  Chromadorea   \n",
       "\n",
       "     taxonomy_group  info                                           sequence  \\\n",
       "3580           None  None  MADDTPIIEEIAEQNESVTRIMQRLKHDMQRVTSVPGFNTSAAGVN...   \n",
       "\n",
       "     variant_under_consideration sequence_accession publication_id  \n",
       "3580                        None               None           None  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "c1df3dab-782a-408a-ba51-8a44646ef46d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 6239\n",
      "accession NP_499073.1 <class 'str'>\n",
      "variant cenH3_(Plants) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 6239 <class 'int'>\n",
      "organism Caenorhabditis elegans <class 'str'>\n",
      "phylum Nematoda <class 'str'>\n",
      "class Chromadorea <class 'str'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MADDGPIIEEIAEKNGRVARIMQRLQHDTQRVTSVPGFNTSATGYADLIALLDQYKNDLEAVGFNDLEQARRRAPSVDITVGSNSTNLVDYSHGRHDMPSHRRHDSSDEEITAANSHHQSPINVGNRNDTDGTNGRNGSRAGSSSSDRVRMIAGRNRISKTRRYRPGQKALEEIRKYQESEDLLIPKAPFARLVREIMQTSTPFSSDLRIRSDAINALQEASEALLVQMFDGSSLISAHSKRATLTTTDVQLYRRLCLPNL <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for a in accessions[1:]:\n",
    "    with Entrez.efetch(db=\"protein\", id=a, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"cenH3_(Plants)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "41c5d869-c9d2-4c95-878c-ed142ac1c8d5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2, 1)"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(accessions), len(data_sequence_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "8d2ebf81-9882-48ac-aef5-fe0a6c914d9e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "8a3a2b7b-2852-4314-8def-d4f4fe88de39",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3280</th>\n",
       "      <td>NP_499073.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>6239.0</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>Nematoda</td>\n",
       "      <td>Chromadorea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MADDGPIIEEIAEKNGRVARIMQRLQHDTQRVTSVPGFNTSATGYA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3281</th>\n",
       "      <td>NP_499128.1</td>\n",
       "      <td>cenH3_(Animals)</td>\n",
       "      <td>17553736</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>6239.0</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>Nematoda</td>\n",
       "      <td>Chromadorea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MADDTPIIEEIAEQNESVTRIMQRLKHDMQRVTSVPGFNTSAAGVN...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession          variant        gi ncbi_gene_id hgnc_gene_name  \\\n",
       "3280  NP_499073.1   cenH3_(Plants)      None         None           None   \n",
       "3281  NP_499128.1  cenH3_(Animals)  17553736         None           None   \n",
       "\n",
       "      taxonomy_id                organism    phylum        class  \\\n",
       "3280       6239.0  Caenorhabditis elegans  Nematoda  Chromadorea   \n",
       "3281       6239.0  Caenorhabditis elegans  Nematoda  Chromadorea   \n",
       "\n",
       "     taxonomy_group  info                                           sequence  \\\n",
       "3280           None  None  MADDGPIIEEIAEKNGRVARIMQRLQHDTQRVTSVPGFNTSATGYA...   \n",
       "3281           None  None  MADDTPIIEEIAEQNESVTRIMQRLKHDMQRVTSVPGFNTSAAGVN...   \n",
       "\n",
       "     variant_under_consideration  \n",
       "3280                        None  \n",
       "3281                        None  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "fa802416-925a-4685-880b-5d7c90a7393c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94ca4f1b-0bbf-482a-9d1e-97193e52031f",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "1983964b-63f9-4bee-b481-7254886460f6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, title, doi, author, year, pubmed_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"monen_differential_2005\", \"monen_separase_2015\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "d97029e3-cc83-4a1f-a3c6-01934bbda96e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_publication = [\n",
    "    {\n",
    "        \"id\": pid,\n",
    "        \"title\": None,\n",
    "        \"doi\": None,\n",
    "        \"author\": None,\n",
    "        \"year\": None,\n",
    "    }\n",
    "    for pid in pids\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "6be10be2-f050-410e-a841-42b3e552f96f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for dp in data_publication:\n",
    "    cursor.execute(add_publication, dp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "1edd2962-7d52-46aa-b8b5-58b023753ec9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>monen_differential_2005</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>monen_separase_2015</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          id title   doi author  year pubmed_id\n",
       "134  monen_differential_2005  None  None   None  None      None\n",
       "135      monen_separase_2015  None  None   None  None      None"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"monen_differential_2005\", \"monen_separase_2015\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "78ab9dd7-ffdf-4cea-9f18-5d217f45b63b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "30d3e021-dd97-42a5-820d-7b6fe4b052ae",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3580</th>\n",
       "      <td>NP_499073.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>6239.0</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>Nematoda</td>\n",
       "      <td>Chromadorea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MADDGPIIEEIAEKNGRVARIMQRLQHDTQRVTSVPGFNTSATGYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>NP_499073.1</td>\n",
       "      <td>monen_differential_2005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3581</th>\n",
       "      <td>NP_499073.1</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>6239.0</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>Nematoda</td>\n",
       "      <td>Chromadorea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MADDGPIIEEIAEKNGRVARIMQRLQHDTQRVTSVPGFNTSATGYA...</td>\n",
       "      <td>None</td>\n",
       "      <td>NP_499073.1</td>\n",
       "      <td>monen_separase_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3582</th>\n",
       "      <td>NP_499128.1</td>\n",
       "      <td>cenH3_(Animals)</td>\n",
       "      <td>17553736</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>6239.0</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>Nematoda</td>\n",
       "      <td>Chromadorea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MADDTPIIEEIAEQNESVTRIMQRLKHDMQRVTSVPGFNTSAAGVN...</td>\n",
       "      <td>None</td>\n",
       "      <td>NP_499128.1</td>\n",
       "      <td>monen_differential_2005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3583</th>\n",
       "      <td>NP_499128.1</td>\n",
       "      <td>cenH3_(Animals)</td>\n",
       "      <td>17553736</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>6239.0</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>Nematoda</td>\n",
       "      <td>Chromadorea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MADDTPIIEEIAEQNESVTRIMQRLKHDMQRVTSVPGFNTSAAGVN...</td>\n",
       "      <td>None</td>\n",
       "      <td>NP_499128.1</td>\n",
       "      <td>monen_separase_2015</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession          variant        gi ncbi_gene_id hgnc_gene_name  \\\n",
       "3580  NP_499073.1   cenH3_(Plants)      None         None           None   \n",
       "3581  NP_499073.1   cenH3_(Plants)      None         None           None   \n",
       "3582  NP_499128.1  cenH3_(Animals)  17553736         None           None   \n",
       "3583  NP_499128.1  cenH3_(Animals)  17553736         None           None   \n",
       "\n",
       "      taxonomy_id                organism    phylum        class  \\\n",
       "3580       6239.0  Caenorhabditis elegans  Nematoda  Chromadorea   \n",
       "3581       6239.0  Caenorhabditis elegans  Nematoda  Chromadorea   \n",
       "3582       6239.0  Caenorhabditis elegans  Nematoda  Chromadorea   \n",
       "3583       6239.0  Caenorhabditis elegans  Nematoda  Chromadorea   \n",
       "\n",
       "     taxonomy_group  info                                           sequence  \\\n",
       "3580           None  None  MADDGPIIEEIAEKNGRVARIMQRLQHDTQRVTSVPGFNTSATGYA...   \n",
       "3581           None  None  MADDGPIIEEIAEKNGRVARIMQRLQHDTQRVTSVPGFNTSATGYA...   \n",
       "3582           None  None  MADDTPIIEEIAEQNESVTRIMQRLKHDMQRVTSVPGFNTSAAGVN...   \n",
       "3583           None  None  MADDTPIIEEIAEQNESVTRIMQRLKHDMQRVTSVPGFNTSAAGVN...   \n",
       "\n",
       "     variant_under_consideration sequence_accession           publication_id  \n",
       "3580                        None        NP_499073.1  monen_differential_2005  \n",
       "3581                        None        NP_499073.1      monen_separase_2015  \n",
       "3582                        None        NP_499128.1  monen_differential_2005  \n",
       "3583                        None        NP_499128.1      monen_separase_2015  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "eb4168f8-81ec-4df6-b480-6a948afabffb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e8bad07-ef74-40fd-a93b-e7897df0c591",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Close connections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "e5673501-5845-491a-bbb9-f5d213721d58",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cursor.close()\n",
    "conn.close()\n",
    "tunnel.stop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03cac75c-b8d7-4773-b7f5-a1531e068ff5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".conda-histdb_env",
   "language": "python",
   "name": "conda-env-.conda-histdb_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
