{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "53d3ec55-7daf-4ade-8add-2830b709dc54",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import io\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import Entrez, SeqIO\n",
    "from mysql.connector import connection\n",
    "from sshtunnel import SSHTunnelForwarder\n",
    "\n",
    "# Укажите ваш email (обязательно для использования Entrez)\n",
    "Entrez.email = \"your.email@example.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "608c4bfe-b979-4b60-b2b5-b94eeb679cb5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open(\"db_curated_server_info.txt\", \"r\") as file:\n",
    "    lines = file.readlines()\n",
    "\n",
    "config = {}\n",
    "\n",
    "for line in lines:\n",
    "    line = line.strip()\n",
    "    if line and not line.startswith(\"#\"):\n",
    "        key, value = line.split(\"=\", 1)\n",
    "        config[key] = value.strip()\n",
    "\n",
    "server_name = config.get(\"server_name\")\n",
    "srever_port = int(config.get(\"srever_port\"))\n",
    "ssh_password = config.get(\"ssh_password\")\n",
    "ssh_username = config.get(\"ssh_username\")\n",
    "db_adress = config.get(\"db_adress\")\n",
    "db_port = int(config.get(\"db_port\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4e19aa1c-49a3-48c4-b67f-34c94cef608d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41751\n"
     ]
    }
   ],
   "source": [
    "tunnel = SSHTunnelForwarder(\n",
    "    (server_name, srever_port),\n",
    "    ssh_password=ssh_password,\n",
    "    ssh_username=ssh_username,\n",
    "    remote_bind_address=(db_adress, db_port),\n",
    ")\n",
    "tunnel.start()\n",
    "print(tunnel.local_bind_port)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cef2275b-12d0-48f1-ad3c-ce4cdb9aa267",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "conn = connection.MySQLConnection(\n",
    "    user=\"db_user\",\n",
    "    password=\"db_password\",\n",
    "    host=\"localhost\",\n",
    "    port=tunnel.local_bind_port,\n",
    "    database=\"db_name\",\n",
    ")\n",
    "cursor = conn.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f589ea8d-8b83-4331-a1cf-b64ed7621801",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('alternative_name',),\n",
       " ('histone',),\n",
       " ('histone_description',),\n",
       " ('histone_has_publication',),\n",
       " ('publication',),\n",
       " ('sequence',),\n",
       " ('sequence_has_publication',)]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SHOW TABLES;\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "b3dd9ef8-b162-4127-9535-ac7743af8c69",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "add_histone = (\n",
    "    \"INSERT INTO histone \"\n",
    "    \"(id, level, taxonomic_span, taxonomic_span_id, description, parent) \"\n",
    "    \"VALUES (%(id)s, %(level)s, %(taxonomic_span)s, %(taxonomic_span_id)s, %(description)s, %(parent)s)\"\n",
    ")\n",
    "add_histone_description = (\n",
    "    \"INSERT INTO histone_description \"\n",
    "    \"(summary, taxonomy, genes, evolution, expression, knock_out, function, sequence, localization, deposition, structure, interactions, disease, caveats) \"\n",
    "    \"VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)\"\n",
    ")\n",
    "add_publication = (\n",
    "    \"INSERT INTO publication \"\n",
    "    \"(id, title, doi, author, year) \"\n",
    "    \"VALUES (%(id)s, %(title)s, %(doi)s, %(author)s, %(year)s)\"\n",
    ")\n",
    "# add_sequence = (\n",
    "#     \"INSERT INTO sequence \"\n",
    "#     \"(accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration) \"\n",
    "#     \"VALUES (%(accession)s, %(variant)s, %(gi)s, %(ncbi_gene_id)s, %(hgnc_gene_name)s, %(taxonomy_id)s, %(organism)s, %(phylum)s, %(class)s, %(taxonomy_group)s, %(info)s, %(sequence)s, %(variant_under_consideration)s)\"\n",
    "# )\n",
    "add_sequence_has_publication = (\n",
    "    \"INSERT INTO sequence_has_publication \"\n",
    "    \"(sequence_accession, publication_id) \"\n",
    "    \"VALUES (%s, %s)\"\n",
    ")\n",
    "add_alternate_names = (\n",
    "    \"INSERT INTO alternative_name \"\n",
    "    \"(name, taxonomy, gene, splice, histone) \"\n",
    "    \"VALUES (%(name)s, %(taxonomy)s, %(gene)s, %(splice)s, %(histone)s)\"\n",
    ")\n",
    "add_histone_has_publication = (\n",
    "    \"INSERT INTO histone_has_publication \"\n",
    "    \"(histone_id, publication_id) \"\n",
    "    \"VALUES (%s, %s)\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9a15b106-5cf1-40d4-8597-2b11dc30ff33",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>65</td>\n",
       "      <td>cid</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>93</td>\n",
       "      <td>cid</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Animals)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id name taxonomy  gene splice          histone\n",
       "0  65  cid     None  None   None            cenH3\n",
       "1  93  cid     None  None   None  cenH3_(Animals)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name \" \"WHERE name='cid'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c060c047-3e7d-4095-b608-9722031af692",
   "metadata": {},
   "source": [
    "# To Do\n",
    "\n",
    "## <span style=\"color:green\">Add to gene description of cenH3_(Plants):</span>\n",
    "\n",
    "```\n",
    "It was demonstrated that the duplication of HTR12 in A. lyrata may have facilitated adaptation to multiple centromeric satellite sequences, though the underlying mechanisms of this process require further investigation [kawabe_duplication_2006]. Pisum and closely related Lathyrus species also retain two CenH3 paralogs (CenH3-1 and CenH3-2) originating from a Fabeae ancestor [neumann_centromeres_2015, neumann_stretching_2012]. Despite divergence, they colocalize in centromeres, likely enabling meta-polycentric organization with multiple functional domains. CenH3-2 is highly conserved but shows adaptive evolution in its histone fold, while CenH3-1 is more variable (especially N-terminal) without signs of positive selection [neumann_centromeres_2015]. The diploid barley species H. vulgare and H. bulbosum each encode two functional CENH3 variants (α and β) that arose before their divergence (~7 mya), are expressed and incorporated into centromeres, but display distinct interspecific incorporation patterns in hybrids [sanei_loss_2011, karimi-ashtiyani_point_2015]. Notably, in early embryos, some centromeres (9.6%) contain only βCENH3, while later stages always incorporate both variants, revealing developmental-stage-specific loading [ishii_differential_2015]. In rye, both CENH3 genes (α and β) are actively transcribed, retained in the genome, and likely functionally important for centromeric nucleosome assembly, while alternative splicing of αCENH3 generates protein diversity without affecting key DNA-binding domains (CATD) [evtushenko_conserved_2017].\n",
    "```\n",
    "\n",
    "## <span style=\"color:green\">Add to structure description of cenH3_(Plants):</span>\n",
    "```\n",
    "The L92F mutation in barley βCENH3 (and its orthologous variants in other plants) disrupts centromeric histone loading, enabling the development of haploid inducer lines for accelerated breeding, as it triggers selective elimination of mutant-derived chromosomes in hybrids with wild-type parents [karimi-ashtiyani_point_2015].\n",
    "```\n",
    "\n",
    "## <span style=\"color:green\">Add summary description of cenH3_(Mammalia):</span>\n",
    "\n",
    "```\n",
    "cenH3_(Mammalia) is a centromere-specific histone variant in mammals, often called CENP-A (Centromere Protein A), and an important component of active centromere required for chromosome segregation. For a general description see cenH3_(Animals) class.\n",
    "```\n",
    "\n",
    "## <span style=\"color:green\">Add to gene description of cenH3_(Mammalia):</span>\n",
    "\n",
    "```\n",
    "Eleven centromeric histone CENP-A genes were identified in cattle (Bos taurus), among which only CENP-A-L-1 appears to be functionally active, while the remaining members likely represent nonfunctional pseudogenes [li_identification_2008].\n",
    "```\n",
    "\n",
    "## <span style=\"color:green\">Add summary description of cenH3_(Animals):</span>\n",
    "\n",
    "```\n",
    "cenH3_(Animals) is a centromere-specific histone variant in animals (Metazoa), often called CENP-A (Centromere Protein A), and an important component of active centromere required for chromosome segregation.\n",
    "```\n",
    "\n",
    "## <span style=\"color:green\">Add to function description of cenH3_(Animals):</span>\n",
    "\n",
    "```\n",
    "CENP-A nucleosomes are essential for chromosome segregation. A functional kinetochore interacts with active centromeric chromatin reached with CENP-A to form the mitotic spindle. Notably, in certain holocentric organisms, CENP-A appears dispensable for meiotic chromosome segregation while remaining essential for mitosis. As demonstrated in Caenorhabditis elegans, CENP-A loading is specifically eliminated following meiosis I, and RNAi-mediated depletion of CENP-A during meiosis fails to disrupt proper chromosome segregation [monen_differential_2005]. This stands in striking contrast to mitotic divisions, where CENP-A is absolutely required for kinetochore assembly and faithful chromosome segregation. Moreover, C. elegans and C. remanei possess a second cenH3 histone gene that likely performs specialized functions distinct from canonical CENP-A roles [monen_separase_2015]. For more detailed information see the \"gene\" section.\n",
    "```\n",
    "\n",
    "## <span style=\"color:green\">Add to gene description of cenH3_(Animals):</span>\n",
    "\n",
    "```\n",
    "The nematodes Caenorhabditis elegans and C. remanei possess two genes encoding the centromeric histone H3 variant (cenH3): HCP-3 (also known as CeCENP-A) and CPAR-1 [monen_separase_2015, monen_differential_2005]. While HCP-3 functions as the predominant centromeric histone - being highly expressed and essential for precise chromosome segregation during mitotic divisions - CPAR-1 displays distinct characteristics: it exhibits lower expression levels but shows specific enrichment on meiotic chromosomes. Importantly, CPAR-1 undergoes separase-mediated proteolytic cleavage at the meiosis I metaphase-to-anaphase transition, implying a specialized, though not yet fully characterized, role in meiotic processes [monen_separase_2015]. \n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acc4ceb8-3969-45c6-bd9f-8c69fe3b534d",
   "metadata": {
    "tags": []
   },
   "source": [
    "# <span style=\"color:black\">Add to gene description of cenH3_(Plants):</span>\n",
    "\n",
    "```\n",
    "It was demonstrated that the duplication of HTR12 in A. lyrata may have facilitated adaptation to multiple centromeric satellite sequences, though the underlying mechanisms of this process require further investigation [kawabe_duplication_2006]. Pisum and closely related Lathyrus species also retain two CenH3 paralogs (CenH3-1 and CenH3-2) originating from a Fabeae ancestor [neumann_centromeres_2015, neumann_stretching_2012]. Despite divergence, they colocalize in centromeres, likely enabling meta-polycentric organization with multiple functional domains. CenH3-2 is highly conserved but shows adaptive evolution in its histone fold, while CenH3-1 is more variable (especially N-terminal) without signs of positive selection [neumann_centromeres_2015]. The diploid barley species H. vulgare and H. bulbosum each encode two functional CENH3 variants (α and β) that arose before their divergence (~7 mya), are expressed and incorporated into centromeres, but display distinct interspecific incorporation patterns in hybrids [sanei_loss_2011, karimi-ashtiyani_point_2015]. Notably, in early embryos, some centromeres (9.6%) contain only βCENH3, while later stages always incorporate both variants, revealing developmental-stage-specific loading [ishii_differential_2015]. In rye, both CENH3 genes (α and β) are actively transcribed, retained in the genome, and likely functionally important for centromeric nucleosome assembly, while alternative splicing of αCENH3 generates protein diversity without affecting key DNA-binding domains (CATD) [evtushenko_conserved_2017].\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "166d9b0e-8f67-4d88-ace0-b0cc0684d04a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3_(Plants) is a centromere-specific histon...</td>\n",
       "      <td>None</td>\n",
       "      <td>While most diploid eukaryotes and flowering pl...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "0  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "\n",
       "  parent   id                                            summary taxonomy  \\\n",
       "0  cenH3  233  cenH3_(Plants) is a centromere-specific histon...     None   \n",
       "\n",
       "                                               genes  ... knock_out function  \\\n",
       "0  While most diploid eukaryotes and flowering pl...  ...      None     None   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     None         None       None      None         None    None    None   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3_(Plants)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5d4bd9fe-983a-4c4a-b9bc-3ea2f5392722",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'While most diploid eukaryotes and flowering plants maintain only one CENH3 gene copy (even after whole-genome duplications), some diploid species like Arabidopsis lyrata, barley, rye, pea, and related legumes retain two functional CENH3 homologs [ishii_unequal_2020].'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"genes\"].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "044e8904-fd56-41cf-a07a-156c8907deef",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UPDATE histone_description SET genes='While most diploid eukaryotes and flowering plants maintain only one CENH3 gene copy (even after whole-genome duplications), some diploid species like Arabidopsis lyrata, barley, rye, pea, and related legumes retain two functional CENH3 homologs [ishii_unequal_2020]. It was demonstrated that the duplication of HTR12 in A. lyrata may have facilitated adaptation to multiple centromeric satellite sequences, though the underlying mechanisms of this process require further investigation [kawabe_duplication_2006]. Pisum and closely related Lathyrus species also retain two CenH3 paralogs (CenH3-1 and CenH3-2) originating from a Fabeae ancestor [neumann_centromeres_2015, neumann_stretching_2012]. Despite divergence, they colocalize in centromeres, likely enabling meta-polycentric organization with multiple functional domains. CenH3-2 is highly conserved but shows adaptive evolution in its histone fold, while CenH3-1 is more variable (especially N-terminal) without signs of positive selection [neumann_centromeres_2015]. The diploid barley species H. vulgare and H. bulbosum each encode two functional CENH3 variants (α and β) that arose before their divergence (~7 mya), are expressed and incorporated into centromeres, but display distinct interspecific incorporation patterns in hybrids [sanei_loss_2011, karimi-ashtiyani_point_2015]. Notably, in early embryos, some centromeres (9.6%) contain only βCENH3, while later stages always incorporate both variants, revealing developmental-stage-specific loading [ishii_differential_2015]. In rye, both CENH3 genes (α and β) are actively transcribed, retained in the genome, and likely functionally important for centromeric nucleosome assembly, while alternative splicing of αCENH3 generates protein diversity without affecting key DNA-binding domains (CATD) [evtushenko_conserved_2017].' WHERE id=233\n"
     ]
    }
   ],
   "source": [
    "gene_desc = (\n",
    "    df[\"genes\"].values[0]\n",
    "    + \" It was demonstrated that the duplication of HTR12 in A. lyrata may have facilitated adaptation to multiple centromeric satellite sequences, though the underlying mechanisms of this process require further investigation [kawabe_duplication_2006]. Pisum and closely related Lathyrus species also retain two CenH3 paralogs (CenH3-1 and CenH3-2) originating from a Fabeae ancestor [neumann_centromeres_2015, neumann_stretching_2012]. Despite divergence, they colocalize in centromeres, likely enabling meta-polycentric organization with multiple functional domains. CenH3-2 is highly conserved but shows adaptive evolution in its histone fold, while CenH3-1 is more variable (especially N-terminal) without signs of positive selection [neumann_centromeres_2015]. The diploid barley species H. vulgare and H. bulbosum each encode two functional CENH3 variants (α and β) that arose before their divergence (~7 mya), are expressed and incorporated into centromeres, but display distinct interspecific incorporation patterns in hybrids [sanei_loss_2011, karimi-ashtiyani_point_2015]. Notably, in early embryos, some centromeres (9.6%) contain only βCENH3, while later stages always incorporate both variants, revealing developmental-stage-specific loading [ishii_differential_2015]. In rye, both CENH3 genes (α and β) are actively transcribed, retained in the genome, and likely functionally important for centromeric nucleosome assembly, while alternative splicing of αCENH3 generates protein diversity without affecting key DNA-binding domains (CATD) [evtushenko_conserved_2017].\"\n",
    ")\n",
    "query = f\"UPDATE histone_description SET genes='{gene_desc}' WHERE id=233\"\n",
    "print(query)\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6d856598-8dfa-4ae5-9d1f-41c6074bbcd8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'While most diploid eukaryotes and flowering plants maintain only one CENH3 gene copy (even after whole-genome duplications), some diploid species like Arabidopsis lyrata, barley, rye, pea, and related legumes retain two functional CENH3 homologs [ishii_unequal_2020]. It was demonstrated that the duplication of HTR12 in A. lyrata may have facilitated adaptation to multiple centromeric satellite sequences, though the underlying mechanisms of this process require further investigation [kawabe_duplication_2006]. Pisum and closely related Lathyrus species also retain two CenH3 paralogs (CenH3-1 and CenH3-2) originating from a Fabeae ancestor [neumann_centromeres_2015, neumann_stretching_2012]. Despite divergence, they colocalize in centromeres, likely enabling meta-polycentric organization with multiple functional domains. CenH3-2 is highly conserved but shows adaptive evolution in its histone fold, while CenH3-1 is more variable (especially N-terminal) without signs of positive selection [neumann_centromeres_2015]. The diploid barley species H. vulgare and H. bulbosum each encode two functional CENH3 variants (α and β) that arose before their divergence (~7 mya), are expressed and incorporated into centromeres, but display distinct interspecific incorporation patterns in hybrids [sanei_loss_2011, karimi-ashtiyani_point_2015]. Notably, in early embryos, some centromeres (9.6%) contain only βCENH3, while later stages always incorporate both variants, revealing developmental-stage-specific loading [ishii_differential_2015]. In rye, both CENH3 genes (α and β) are actively transcribed, retained in the genome, and likely functionally important for centromeric nucleosome assembly, while alternative splicing of αCENH3 generates protein diversity without affecting key DNA-binding domains (CATD) [evtushenko_conserved_2017].'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3_(Plants)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])[\n",
    "    \"genes\"\n",
    "].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "44afbe41-bba0-4956-be03-464c60d7aeee",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "158a6890-1475-4884-b6ab-81e9949cc0c4",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add to structure description of cenH3_(Plants):</span>\n",
    "```\n",
    "The L92F mutation in barley βCENH3 (and its orthologous variants in other plants) disrupts centromeric histone loading, enabling the development of haploid inducer lines for accelerated breeding, as it triggers selective elimination of mutant-derived chromosomes in hybrids with wild-type parents [karimi-ashtiyani_point_2015].\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b039585c-e78e-4b4c-85b2-20d2a6905388",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3_(Plants) is a centromere-specific histon...</td>\n",
       "      <td>None</td>\n",
       "      <td>While most diploid eukaryotes and flowering pl...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "0  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "\n",
       "  parent   id                                            summary taxonomy  \\\n",
       "0  cenH3  233  cenH3_(Plants) is a centromere-specific histon...     None   \n",
       "\n",
       "                                               genes  ... knock_out function  \\\n",
       "0  While most diploid eukaryotes and flowering pl...  ...      None     None   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     None         None       None      None         None    None    None   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3_(Plants)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ba83faf8-a363-463f-bb96-d61cdffe7f85",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df[\"structure\"].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "22a997d4-9886-493e-b0f2-ce7fe8f39e5a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UPDATE histone_description SET structure='The L92F mutation in barley βCENH3 (and its orthologous variants in other plants) disrupts centromeric histone loading, enabling the development of haploid inducer lines for accelerated breeding, as it triggers selective elimination of mutant-derived chromosomes in hybrids with wild-type parents [karimi-ashtiyani_point_2015].' WHERE id=233\n"
     ]
    }
   ],
   "source": [
    "structure_desc = \"The L92F mutation in barley βCENH3 (and its orthologous variants in other plants) disrupts centromeric histone loading, enabling the development of haploid inducer lines for accelerated breeding, as it triggers selective elimination of mutant-derived chromosomes in hybrids with wild-type parents [karimi-ashtiyani_point_2015].\"\n",
    "query = f\"UPDATE histone_description SET structure='{structure_desc}' WHERE id=233\"\n",
    "print(query)\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "31fbeff9-ffc2-401d-b6fc-384896c0e5a4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'The L92F mutation in barley βCENH3 (and its orthologous variants in other plants) disrupts centromeric histone loading, enabling the development of haploid inducer lines for accelerated breeding, as it triggers selective elimination of mutant-derived chromosomes in hybrids with wild-type parents [karimi-ashtiyani_point_2015].'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3_(Plants)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])[\n",
    "    \"structure\"\n",
    "].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "84013480-c5d0-4910-a06a-85c860f5f9d5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'While most diploid eukaryotes and flowering plants maintain only one CENH3 gene copy (even after whole-genome duplications), some diploid species like Arabidopsis lyrata, barley, rye, pea, and related legumes retain two functional CENH3 homologs [ishii_unequal_2020]. It was demonstrated that the duplication of HTR12 in A. lyrata may have facilitated adaptation to multiple centromeric satellite sequences, though the underlying mechanisms of this process require further investigation [kawabe_duplication_2006]. Pisum and closely related Lathyrus species also retain two CenH3 paralogs (CenH3-1 and CenH3-2) originating from a Fabeae ancestor [neumann_centromeres_2015, neumann_stretching_2012]. Despite divergence, they colocalize in centromeres, likely enabling meta-polycentric organization with multiple functional domains. CenH3-2 is highly conserved but shows adaptive evolution in its histone fold, while CenH3-1 is more variable (especially N-terminal) without signs of positive selection [neumann_centromeres_2015]. The diploid barley species H. vulgare and H. bulbosum each encode two functional CENH3 variants (α and β) that arose before their divergence (~7 mya), are expressed and incorporated into centromeres, but display distinct interspecific incorporation patterns in hybrids [sanei_loss_2011, karimi-ashtiyani_point_2015]. Notably, in early embryos, some centromeres (9.6%) contain only βCENH3, while later stages always incorporate both variants, revealing developmental-stage-specific loading [ishii_differential_2015]. In rye, both CENH3 genes (α and β) are actively transcribed, retained in the genome, and likely functionally important for centromeric nucleosome assembly, while alternative splicing of αCENH3 generates protein diversity without affecting key DNA-binding domains (CATD) [evtushenko_conserved_2017].'"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3_(Plants)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])[\n",
    "    \"genes\"\n",
    "].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "fb849735-b1ae-42bf-a648-8182a3b4fc31",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e420b192-c287-4a3a-9023-1cea263db874",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add summary description of cenH3_(Mammalia):</span>\n",
    "\n",
    "```\n",
    "cenH3_(Mammalia) is a centromere-specific histone variant in mammals, often called CENP-A (Centromere Protein A), and an important component of active centromere required for chromosome segregation. For a general description see cenH3_(Animals) class.\n",
    "```\n",
    "\n",
    "# <span style=\"color:black\">Add to gene description of cenH3_(Mammalia):</span>\n",
    "\n",
    "```\n",
    "Eleven centromeric histone CENP-A genes were identified in cattle (Bos taurus), among which only CENP-A-L-1 appears to be functionally active, while the remaining members likely represent nonfunctional pseudogenes [li_identification_2008].\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "75951070-a91f-4bb1-96e0-0b9c4be3899a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3_(Mammalia)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>130</td>\n",
       "      <td>cenH3_(Animals)</td>\n",
       "      <td>130</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "0  cenH3_(Mammalia)  variant   Homo sapiens              9606          130   \n",
       "\n",
       "            parent   id summary taxonomy genes  ... knock_out function  \\\n",
       "0  cenH3_(Animals)  130    null     null  null  ...      null     null   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     null         null       null      null         null    null    null   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3_(Mammalia)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "1edf4faf-4a3f-445d-bdc0-27b994b61bc3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UPDATE histone_description SET summary='cenH3_(Mammalia) is a centromere-specific histone variant in mammals, often called CENP-A (Centromere Protein A), and an important component of active centromere required for chromosome segregation. For a general description see cenH3_(Animals) class.', genes='Eleven centromeric histone CENP-A genes were identified in cattle (Bos taurus), among which only CENP-A-L-1 appears to be functionally active, while the remaining members likely represent nonfunctional pseudogenes [li_identification_2008].' WHERE id=130\n"
     ]
    }
   ],
   "source": [
    "summary_desc = \"cenH3_(Mammalia) is a centromere-specific histone variant in mammals, often called CENP-A (Centromere Protein A), and an important component of active centromere required for chromosome segregation. For a general description see cenH3_(Animals) class.\"\n",
    "gene_desc = \"Eleven centromeric histone CENP-A genes were identified in cattle (Bos taurus), among which only CENP-A-L-1 appears to be functionally active, while the remaining members likely represent nonfunctional pseudogenes [li_identification_2008].\"\n",
    "query = f\"UPDATE histone_description SET summary='{summary_desc}', genes='{gene_desc}' WHERE id=130\"\n",
    "print(query)\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a05607a6-282e-4200-a407-322d1e9d6c86",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3_(Mammalia)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>130</td>\n",
       "      <td>cenH3_(Animals)</td>\n",
       "      <td>130</td>\n",
       "      <td>cenH3_(Mammalia) is a centromere-specific hist...</td>\n",
       "      <td>null</td>\n",
       "      <td>Eleven centromeric histone CENP-A genes were i...</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "0  cenH3_(Mammalia)  variant   Homo sapiens              9606          130   \n",
       "\n",
       "            parent   id                                            summary  \\\n",
       "0  cenH3_(Animals)  130  cenH3_(Mammalia) is a centromere-specific hist...   \n",
       "\n",
       "  taxonomy                                              genes  ... knock_out  \\\n",
       "0     null  Eleven centromeric histone CENP-A genes were i...  ...      null   \n",
       "\n",
       "  function sequence localization deposition structure interactions disease  \\\n",
       "0     null     null         null       null      null         null    null   \n",
       "\n",
       "  caveats relations  \n",
       "0    null      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3_(Mammalia)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "73f3906a-a3ef-4530-9d6b-42935eea8852",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3b7991b-b9e4-4457-8f3a-cdb8873f360c",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add summary description of cenH3_(Animals):</span>\n",
    "\n",
    "```\n",
    "cenH3_(Animals) is a centromere-specific histone variant in animals (Metazoa), often called CENP-A (Centromere Protein A), and an important component of active centromere required for chromosome segregation.\n",
    "```\n",
    "\n",
    "# <span style=\"color:black\">Add to function description of cenH3_(Animals):</span>\n",
    "\n",
    "```\n",
    "CENP-A nucleosomes are essential for chromosome segregation. A functional kinetochore interacts with active centromeric chromatin reached with CENP-A to form the mitotic spindle. Notably, in certain holocentric organisms, CENP-A appears dispensable for meiotic chromosome segregation while remaining essential for mitosis. As demonstrated in Caenorhabditis elegans, CENP-A loading is specifically eliminated following meiosis I, and RNAi-mediated depletion of CENP-A during meiosis fails to disrupt proper chromosome segregation [monen_differential_2005]. This stands in striking contrast to mitotic divisions, where CENP-A is absolutely required for kinetochore assembly and faithful chromosome segregation. Moreover, C. elegans and C. remanei possess a second cenH3 histone gene that likely performs specialized functions distinct from canonical CENP-A roles [monen_separase_2015]. For more detailed information see the \"gene\" section.\n",
    "```\n",
    "\n",
    "# <span style=\"color:black\">Add to gene description of cenH3_(Animals):</span>\n",
    "\n",
    "```\n",
    "The nematodes Caenorhabditis elegans and C. remanei possess two genes encoding the centromeric histone H3 variant (cenH3): HCP-3 (also known as CeCENP-A) and CPAR-1 [monen_separase_2015, monen_differential_2005]. While HCP-3 functions as the predominant centromeric histone - being highly expressed and essential for precise chromosome segregation during mitotic divisions - CPAR-1 displays distinct characteristics: it exhibits lower expression levels but shows specific enrichment on meiotic chromosomes. Importantly, CPAR-1 undergoes separase-mediated proteolytic cleavage at the meiosis I metaphase-to-anaphase transition, implying a specialized, though not yet fully characterized, role in meiotic processes [monen_separase_2015]. \n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "758d5033-0fed-46d4-b202-d6a786d54a8b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3_(Animals)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>94</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>94</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "0  cenH3_(Animals)  variant   Homo sapiens              9606           94   \n",
       "\n",
       "  parent  id summary taxonomy genes  ... knock_out function sequence  \\\n",
       "0  cenH3  94    null     null  null  ...      null     null     null   \n",
       "\n",
       "  localization deposition structure interactions disease caveats relations  \n",
       "0         null       null      null         null    null    null      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3_(Animals)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "d60908d8-8c45-4ae4-bf31-19881fad5196",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UPDATE histone_description SET summary='cenH3_(Animals) is a centromere-specific histone variant in animals (Metazoa), often called CENP-A (Centromere Protein A), and an important component of active centromere required for chromosome segregation.', genes='CENP-A nucleosomes are essential for chromosome segregation. A functional kinetochore interacts with active centromeric chromatin reached with CENP-A to form the mitotic spindle. Notably, in certain holocentric organisms, CENP-A appears dispensable for meiotic chromosome segregation while remaining essential for mitosis. As demonstrated in Caenorhabditis elegans, CENP-A loading is specifically eliminated following meiosis I, and RNAi-mediated depletion of CENP-A during meiosis fails to disrupt proper chromosome segregation [monen_differential_2005]. This stands in striking contrast to mitotic divisions, where CENP-A is absolutely required for kinetochore assembly and faithful chromosome segregation. Moreover, C. elegans and C. remanei possess a second cenH3 histone gene that likely performs specialized functions distinct from canonical CENP-A roles [monen_separase_2015]. For more detailed information see the \"gene\" section.', function='The nematodes Caenorhabditis elegans and C. remanei possess two genes encoding the centromeric histone H3 variant (cenH3): HCP-3 (also known as CeCENP-A) and CPAR-1 [monen_separase_2015, monen_differential_2005]. While HCP-3 functions as the predominant centromeric histone - being highly expressed and essential for precise chromosome segregation during mitotic divisions - CPAR-1 displays distinct characteristics: it exhibits lower expression levels but shows specific enrichment on meiotic chromosomes. Importantly, CPAR-1 undergoes separase-mediated proteolytic cleavage at the meiosis I metaphase-to-anaphase transition, implying a specialized, though not yet fully characterized, role in meiotic processes [monen_separase_2015].' WHERE id=94\n"
     ]
    }
   ],
   "source": [
    "summary_desc = \"cenH3_(Animals) is a centromere-specific histone variant in animals (Metazoa), often called CENP-A (Centromere Protein A), and an important component of active centromere required for chromosome segregation.\"\n",
    "gene_desc = 'CENP-A nucleosomes are essential for chromosome segregation. A functional kinetochore interacts with active centromeric chromatin reached with CENP-A to form the mitotic spindle. Notably, in certain holocentric organisms, CENP-A appears dispensable for meiotic chromosome segregation while remaining essential for mitosis. As demonstrated in Caenorhabditis elegans, CENP-A loading is specifically eliminated following meiosis I, and RNAi-mediated depletion of CENP-A during meiosis fails to disrupt proper chromosome segregation [monen_differential_2005]. This stands in striking contrast to mitotic divisions, where CENP-A is absolutely required for kinetochore assembly and faithful chromosome segregation. Moreover, C. elegans and C. remanei possess a second cenH3 histone gene that likely performs specialized functions distinct from canonical CENP-A roles [monen_separase_2015]. For more detailed information see the \"gene\" section.'\n",
    "function_desc = \"The nematodes Caenorhabditis elegans and C. remanei possess two genes encoding the centromeric histone H3 variant (cenH3): HCP-3 (also known as CeCENP-A) and CPAR-1 [monen_separase_2015, monen_differential_2005]. While HCP-3 functions as the predominant centromeric histone - being highly expressed and essential for precise chromosome segregation during mitotic divisions - CPAR-1 displays distinct characteristics: it exhibits lower expression levels but shows specific enrichment on meiotic chromosomes. Importantly, CPAR-1 undergoes separase-mediated proteolytic cleavage at the meiosis I metaphase-to-anaphase transition, implying a specialized, though not yet fully characterized, role in meiotic processes [monen_separase_2015].\"\n",
    "query = f\"UPDATE histone_description SET summary='{summary_desc}', genes='{gene_desc}', function='{function_desc}' WHERE id=94\"\n",
    "print(query)\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "dc403fe0-6c8a-4f28-b90f-ac3ec55ecb64",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3_(Animals)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>94</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>94</td>\n",
       "      <td>cenH3_(Animals) is a centromere-specific histo...</td>\n",
       "      <td>null</td>\n",
       "      <td>CENP-A nucleosomes are essential for chromosom...</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>The nematodes Caenorhabditis elegans and C. re...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "0  cenH3_(Animals)  variant   Homo sapiens              9606           94   \n",
       "\n",
       "  parent  id                                            summary taxonomy  \\\n",
       "0  cenH3  94  cenH3_(Animals) is a centromere-specific histo...     null   \n",
       "\n",
       "                                               genes  ... knock_out  \\\n",
       "0  CENP-A nucleosomes are essential for chromosom...  ...      null   \n",
       "\n",
       "                                            function sequence localization  \\\n",
       "0  The nematodes Caenorhabditis elegans and C. re...     null         null   \n",
       "\n",
       "  deposition structure interactions disease caveats relations  \n",
       "0       null      null         null    null    null      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3_(Animals)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "8b8e945a-4389-40c0-9565-d9a0eb23eef8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1780663f-fb28-4cd7-9eb4-71ef831e2660",
   "metadata": {},
   "source": [
    "# Add publications for cenH3_(Plants)\n",
    "\n",
    "kawabe_duplication_2006, neumann_centromeres_2015, neumann_stretching_2012, sanei_loss_2011, karimi-ashtiyani_point_2015, ishii_differential_2015, evtushenko_conserved_2017"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "d3c1fbe9-c8b4-4660-9b14-3bf326ac7ef9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pids = [\n",
    "    \"kawabe_duplication_2006\",\n",
    "    \"neumann_centromeres_2015\",\n",
    "    \"neumann_stretching_2012\",\n",
    "    \"sanei_loss_2011\",\n",
    "    \"karimi-ashtiyani_point_2015\",\n",
    "    \"ishii_differential_2015\",\n",
    "    \"evtushenko_conserved_2017\",\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fec842e-3243-4fde-b651-d90ea5b66af0",
   "metadata": {},
   "source": [
    "## There are some publications with id from pubmed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "4b51a629-1a18-426a-bcb0-d51ad6b1e9d5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>histone_id</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>14583738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>19766562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>21743476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>23324462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>25956076</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "0  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "1  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "2  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "3  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "4  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "5  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "\n",
       "  parent      histone_id publication_id  \n",
       "0  cenH3  cenH3_(Plants)       14583738  \n",
       "1  cenH3  cenH3_(Plants)       19766562  \n",
       "2  cenH3  cenH3_(Plants)       21743476  \n",
       "3  cenH3  cenH3_(Plants)       22650316  \n",
       "4  cenH3  cenH3_(Plants)       23324462  \n",
       "5  cenH3  cenH3_(Plants)       25956076  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_has_publication hp \"\n",
    "    \"ON h.id = hp.histone_id \"\n",
    "    \"WHERE h.id='cenH3_(Plants)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "71c597f6-cc4a-4ce6-b1a9-a72004b38cd9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>histone_id</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>14583738</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>14583738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>19766562</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>19766562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>21743476</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>21743476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH4</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>generic_H1</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H1.0</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H1.10</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H2A.L_(Eutheria)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H2A.X</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H2A.Z</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.Y_(Primates?)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>macroH2A</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>OO_H1.8</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TS_H1.6</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TS_H1.7</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TS_H1.9</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>23324462</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>23324462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>25956076</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>25956076</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          id title   doi author  year        histone_id publication_id\n",
       "24  14583738  None  None   None  None    cenH3_(Plants)       14583738\n",
       "39  19766562  None  None   None  None    cenH3_(Plants)       19766562\n",
       "47  21743476  None  None   None  None    cenH3_(Plants)       21743476\n",
       "54  22650316  None  None   None  None    cenH3_(Plants)       22650316\n",
       "55  22650316  None  None   None  None               cH4       22650316\n",
       "56  22650316  None  None   None  None        generic_H1       22650316\n",
       "57  22650316  None  None   None  None              H1.0       22650316\n",
       "58  22650316  None  None   None  None             H1.10       22650316\n",
       "59  22650316  None  None   None  None  H2A.L_(Eutheria)       22650316\n",
       "60  22650316  None  None   None  None             H2A.X       22650316\n",
       "61  22650316  None  None   None  None             H2A.Z       22650316\n",
       "62  22650316  None  None   None  None   H3.4_(Mammalia)       22650316\n",
       "63  22650316  None  None   None  None  H3.Y_(Primates?)       22650316\n",
       "64  22650316  None  None   None  None          macroH2A       22650316\n",
       "65  22650316  None  None   None  None           OO_H1.8       22650316\n",
       "66  22650316  None  None   None  None           TS_H1.6       22650316\n",
       "67  22650316  None  None   None  None           TS_H1.7       22650316\n",
       "68  22650316  None  None   None  None           TS_H1.9       22650316\n",
       "72  23324462  None  None   None  None    cenH3_(Plants)       23324462\n",
       "85  25956076  None  None   None  None    cenH3_(Plants)       25956076"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM publication p LEFT JOIN histone_has_publication ph \"\n",
    "    \"ON p.id = ph.publication_id \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "dfp = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "dfp[dfp[\"id\"].isin(df[\"publication_id\"])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57106188-8a65-4cdf-a5d2-092f2434d0b8",
   "metadata": {},
   "source": [
    "## Adding new feild pubmed_id to publication table in database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "5d822de3-7e2e-4453-a2a3-efec96194924",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "query = \"ALTER TABLE publication ADD COLUMN pubmed_id varchar(50)\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "7016d3ea-a97a-4d7a-88d6-c7306f51ef75",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>histone_id</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>14583738</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>14583738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>19766562</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>19766562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>21743476</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>21743476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH4</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>generic_H1</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H1.0</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H1.10</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H2A.L_(Eutheria)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H2A.X</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H2A.Z</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.Y_(Primates?)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>macroH2A</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>OO_H1.8</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TS_H1.6</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TS_H1.7</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TS_H1.9</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>23324462</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>23324462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>25956076</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>25956076</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          id title   doi author  year pubmed_id        histone_id  \\\n",
       "24  14583738  None  None   None  None      None    cenH3_(Plants)   \n",
       "39  19766562  None  None   None  None      None    cenH3_(Plants)   \n",
       "47  21743476  None  None   None  None      None    cenH3_(Plants)   \n",
       "54  22650316  None  None   None  None      None    cenH3_(Plants)   \n",
       "55  22650316  None  None   None  None      None               cH4   \n",
       "56  22650316  None  None   None  None      None        generic_H1   \n",
       "57  22650316  None  None   None  None      None              H1.0   \n",
       "58  22650316  None  None   None  None      None             H1.10   \n",
       "59  22650316  None  None   None  None      None  H2A.L_(Eutheria)   \n",
       "60  22650316  None  None   None  None      None             H2A.X   \n",
       "61  22650316  None  None   None  None      None             H2A.Z   \n",
       "62  22650316  None  None   None  None      None   H3.4_(Mammalia)   \n",
       "63  22650316  None  None   None  None      None  H3.Y_(Primates?)   \n",
       "64  22650316  None  None   None  None      None          macroH2A   \n",
       "65  22650316  None  None   None  None      None           OO_H1.8   \n",
       "66  22650316  None  None   None  None      None           TS_H1.6   \n",
       "67  22650316  None  None   None  None      None           TS_H1.7   \n",
       "68  22650316  None  None   None  None      None           TS_H1.9   \n",
       "72  23324462  None  None   None  None      None    cenH3_(Plants)   \n",
       "85  25956076  None  None   None  None      None    cenH3_(Plants)   \n",
       "\n",
       "   publication_id  \n",
       "24       14583738  \n",
       "39       19766562  \n",
       "47       21743476  \n",
       "54       22650316  \n",
       "55       22650316  \n",
       "56       22650316  \n",
       "57       22650316  \n",
       "58       22650316  \n",
       "59       22650316  \n",
       "60       22650316  \n",
       "61       22650316  \n",
       "62       22650316  \n",
       "63       22650316  \n",
       "64       22650316  \n",
       "65       22650316  \n",
       "66       22650316  \n",
       "67       22650316  \n",
       "68       22650316  \n",
       "72       23324462  \n",
       "85       25956076  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM publication p LEFT JOIN histone_has_publication ph \"\n",
    "    \"ON p.id = ph.publication_id \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "dfp = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "dfp[dfp[\"id\"].isin(df[\"publication_id\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "89cb0524-d59c-4743-99d3-fa099e956b83",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d5546f9-65cb-4315-b86e-026a24cfaca0",
   "metadata": {},
   "source": [
    "## Correcting all previously added publications to cenH3_(Plants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "cb87d768-28a0-4f4a-8ec3-679830c1a655",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    14583738\n",
       "1    19766562\n",
       "2    21743476\n",
       "3    22650316\n",
       "4    23324462\n",
       "5    25956076\n",
       "Name: publication_id, dtype: object"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"publication_id\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acdfff21-a62e-4ec2-9808-b60a5f348cb4",
   "metadata": {},
   "source": [
    "### Set pubmed_id to these publications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "99e96a7d-3d72-42ba-baad-fa84d9520d69",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UPDATE publication SET pubmed_id='14583738' WHERE id='14583738'\n",
      "UPDATE publication SET pubmed_id='19766562' WHERE id='19766562'\n",
      "UPDATE publication SET pubmed_id='21743476' WHERE id='21743476'\n",
      "UPDATE publication SET pubmed_id='22650316' WHERE id='22650316'\n",
      "UPDATE publication SET pubmed_id='23324462' WHERE id='23324462'\n",
      "UPDATE publication SET pubmed_id='25956076' WHERE id='25956076'\n"
     ]
    }
   ],
   "source": [
    "for pubmed_id in df[\"publication_id\"]:\n",
    "    query = f\"UPDATE publication SET pubmed_id='{pubmed_id}' WHERE id='{pubmed_id}'\"\n",
    "    print(query)\n",
    "    cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "666aca11-6b7a-48ae-9620-892b2aa19565",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>histone_id</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>14583738</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>14583738</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>14583738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>19766562</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>19766562</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>19766562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>21743476</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>21743476</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>21743476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>cH4</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>generic_H1</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>H1.0</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>H1.10</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>H2A.L_(Eutheria)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>H2A.X</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>H2A.Z</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>H3.Y_(Primates?)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>macroH2A</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>OO_H1.8</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>TS_H1.6</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>TS_H1.7</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>TS_H1.9</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>23324462</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>23324462</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>23324462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>25956076</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>25956076</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>25956076</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          id title   doi author  year pubmed_id        histone_id  \\\n",
       "24  14583738  None  None   None  None  14583738    cenH3_(Plants)   \n",
       "39  19766562  None  None   None  None  19766562    cenH3_(Plants)   \n",
       "47  21743476  None  None   None  None  21743476    cenH3_(Plants)   \n",
       "54  22650316  None  None   None  None  22650316    cenH3_(Plants)   \n",
       "55  22650316  None  None   None  None  22650316               cH4   \n",
       "56  22650316  None  None   None  None  22650316        generic_H1   \n",
       "57  22650316  None  None   None  None  22650316              H1.0   \n",
       "58  22650316  None  None   None  None  22650316             H1.10   \n",
       "59  22650316  None  None   None  None  22650316  H2A.L_(Eutheria)   \n",
       "60  22650316  None  None   None  None  22650316             H2A.X   \n",
       "61  22650316  None  None   None  None  22650316             H2A.Z   \n",
       "62  22650316  None  None   None  None  22650316   H3.4_(Mammalia)   \n",
       "63  22650316  None  None   None  None  22650316  H3.Y_(Primates?)   \n",
       "64  22650316  None  None   None  None  22650316          macroH2A   \n",
       "65  22650316  None  None   None  None  22650316           OO_H1.8   \n",
       "66  22650316  None  None   None  None  22650316           TS_H1.6   \n",
       "67  22650316  None  None   None  None  22650316           TS_H1.7   \n",
       "68  22650316  None  None   None  None  22650316           TS_H1.9   \n",
       "72  23324462  None  None   None  None  23324462    cenH3_(Plants)   \n",
       "85  25956076  None  None   None  None  25956076    cenH3_(Plants)   \n",
       "\n",
       "   publication_id  \n",
       "24       14583738  \n",
       "39       19766562  \n",
       "47       21743476  \n",
       "54       22650316  \n",
       "55       22650316  \n",
       "56       22650316  \n",
       "57       22650316  \n",
       "58       22650316  \n",
       "59       22650316  \n",
       "60       22650316  \n",
       "61       22650316  \n",
       "62       22650316  \n",
       "63       22650316  \n",
       "64       22650316  \n",
       "65       22650316  \n",
       "66       22650316  \n",
       "67       22650316  \n",
       "68       22650316  \n",
       "72       23324462  \n",
       "85       25956076  "
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM publication p LEFT JOIN histone_has_publication ph \"\n",
    "    \"ON p.id = ph.publication_id \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "dfp = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "dfp[dfp[\"id\"].isin(df[\"publication_id\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "d8dc9df5-2486-4e36-8347-96f9df68095b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>14583738</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>14583738</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>19766562</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>19766562</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>21743476</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>21743476</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>AAC28846.1</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>AAC37291.1</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>XP_846259.1</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>XP_875023.2</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>23324462</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>23324462</td>\n",
       "      <td>NP_001035891.1</td>\n",
       "      <td>23324462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>23324462</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>23324462</td>\n",
       "      <td>NP_001800.1</td>\n",
       "      <td>23324462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>25956076</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>25956076</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>147 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           id title   doi author  year pubmed_id sequence_accession  \\\n",
       "73   14583738  None  None   None  None  14583738               None   \n",
       "100  19766562  None  None   None  None  19766562               None   \n",
       "112  21743476  None  None   None  None  21743476               None   \n",
       "119  22650316  None  None   None  None  22650316         AAC28846.1   \n",
       "120  22650316  None  None   None  None  22650316         AAC37291.1   \n",
       "..        ...   ...   ...    ...   ...       ...                ...   \n",
       "258  22650316  None  None   None  None  22650316        XP_846259.1   \n",
       "259  22650316  None  None   None  None  22650316        XP_875023.2   \n",
       "262  23324462  None  None   None  None  23324462     NP_001035891.1   \n",
       "263  23324462  None  None   None  None  23324462        NP_001800.1   \n",
       "313  25956076  None  None   None  None  25956076               None   \n",
       "\n",
       "    publication_id  \n",
       "73            None  \n",
       "100           None  \n",
       "112           None  \n",
       "119       22650316  \n",
       "120       22650316  \n",
       "..             ...  \n",
       "258       22650316  \n",
       "259       22650316  \n",
       "262       23324462  \n",
       "263       23324462  \n",
       "313           None  \n",
       "\n",
       "[147 rows x 8 columns]"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM publication p LEFT JOIN sequence_has_publication sh \"\n",
    "    \"ON p.id = sh.publication_id \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "dfp = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "dfp[dfp[\"id\"].isin(df[\"publication_id\"])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf317c79-b937-4f0e-b9db-0cf97c345278",
   "metadata": {},
   "source": [
    "### Create correct info for these publications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "c9c9ca3f-35e8-4109-b836-e30d97c2d4a3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "publications_info = {\n",
    "    \"14583738\": {\n",
    "        \"id\": \"malik_phylogenomics_2003\",\n",
    "        \"title\": \"Phylogenomics of the nucleosome\",\n",
    "        \"doi\": \"10.1038/nsb996\",\n",
    "        \"author\": \"\",\n",
    "        \"year\": \"2003\",\n",
    "    },\n",
    "    \"19766562\": {\n",
    "        \"id\": \"malik_major_2009\",\n",
    "        \"title\": \"Major evolutionary transitions in centromere complexity\",\n",
    "        \"doi\": \"10.1016/j.cell.2009.08.036\",\n",
    "        \"author\": \"\",\n",
    "        \"year\": \"2009\",\n",
    "    },\n",
    "    \"21743476\": {\n",
    "        \"id\": \"tachiwana_crystal_2011\",\n",
    "        \"title\": \"Crystal structure of the human centromeric nucleosome containing CENP-A\",\n",
    "        \"doi\": \"10.1038/nature10258\",\n",
    "        \"author\": \"\",\n",
    "        \"year\": \"2011\",\n",
    "    },\n",
    "    \"22650316\": {\n",
    "        \"id\": \"talbert_unified_2012\",\n",
    "        \"title\": \"A unified phylogeny-based nomenclature for histone variants\",\n",
    "        \"doi\": \"10.1186/1756-8935-5-7\",\n",
    "        \"author\": \"\",\n",
    "        \"year\": \"2012\",\n",
    "    },\n",
    "    \"23324462\": {\n",
    "        \"id\": \"bui_cenp-nucleosome_2013\",\n",
    "        \"title\": \"The CENP-A nucleosome: a battle between Dr Jekyll and Mr Hyde\",\n",
    "        \"doi\": \"10.4161/nucl.23588\",\n",
    "        \"author\": \"\",\n",
    "        \"year\": \"2013\",\n",
    "    },\n",
    "    \"25956076\": {\n",
    "        \"id\": \"steiner_diversity_2015\",\n",
    "        \"title\": \"Diversity in the organization of centromeric chromatin\",\n",
    "        \"doi\": \"10.1016/j.gde.2015.03.010\",\n",
    "        \"author\": \"\",\n",
    "        \"year\": \"2015\",\n",
    "    },\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "730d1008-523d-4ad4-af5e-5ae65ac42705",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>histone_id</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH2B</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH3</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH3_(Fungi)</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H2A</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H2B</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3_(Fungi)</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3_(Plants)</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH3</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3_(Animals)</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3_(Plants)</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3_(Protists)</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           id title   doi author  year pubmed_id  \\\n",
       "228  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "229  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "230  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "231  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "232  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "233  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "234  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "235  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "236  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "237  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "356      talbert_unified_2012  None  None   None  None      None   \n",
       "357      talbert_unified_2012  None  None   None  None      None   \n",
       "358      talbert_unified_2012  None  None   None  None      None   \n",
       "359      talbert_unified_2012  None  None   None  None      None   \n",
       "360      talbert_unified_2012  None  None   None  None      None   \n",
       "\n",
       "          histone_id            publication_id  \n",
       "228             cH2A  malik_phylogenomics_2003  \n",
       "229             cH2B  malik_phylogenomics_2003  \n",
       "230              cH3  malik_phylogenomics_2003  \n",
       "231      cH3_(Fungi)  malik_phylogenomics_2003  \n",
       "232              H2A  malik_phylogenomics_2003  \n",
       "233              H2B  malik_phylogenomics_2003  \n",
       "234               H3  malik_phylogenomics_2003  \n",
       "235             H3.3  malik_phylogenomics_2003  \n",
       "236     H3.3_(Fungi)  malik_phylogenomics_2003  \n",
       "237    H3.3_(Plants)  malik_phylogenomics_2003  \n",
       "356              cH3      talbert_unified_2012  \n",
       "357             H3.3      talbert_unified_2012  \n",
       "358   H3.3_(Animals)      talbert_unified_2012  \n",
       "359    H3.3_(Plants)      talbert_unified_2012  \n",
       "360  H3.3_(Protists)      talbert_unified_2012  "
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM publication p LEFT JOIN histone_has_publication ph \"\n",
    "    \"ON p.id = ph.publication_id \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "dfp = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "dfp[dfp[\"id\"].isin([publications_info[pmid][\"id\"] for pmid in df[\"publication_id\"]])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3fe25e2-e0ad-4b93-90fd-fdde6eb04107",
   "metadata": {},
   "source": [
    "### Add histone relations for two of these publications that are already in DB - 14583738, 22650316"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "b37e6276-b2e2-4126-ae8b-ca00710e6308",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>histone_id</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>14583738</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>14583738</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>14583738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>cH4</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>generic_H1</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>H1.0</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>H1.10</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>H2A.L_(Eutheria)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>H2A.X</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>H2A.Z</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>H3.Y_(Primates?)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>macroH2A</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>OO_H1.8</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>TS_H1.6</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>TS_H1.7</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>TS_H1.9</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          id title   doi author  year pubmed_id        histone_id  \\\n",
       "24  14583738  None  None   None  None  14583738    cenH3_(Plants)   \n",
       "54  22650316  None  None   None  None  22650316    cenH3_(Plants)   \n",
       "55  22650316  None  None   None  None  22650316               cH4   \n",
       "56  22650316  None  None   None  None  22650316        generic_H1   \n",
       "57  22650316  None  None   None  None  22650316              H1.0   \n",
       "58  22650316  None  None   None  None  22650316             H1.10   \n",
       "59  22650316  None  None   None  None  22650316  H2A.L_(Eutheria)   \n",
       "60  22650316  None  None   None  None  22650316             H2A.X   \n",
       "61  22650316  None  None   None  None  22650316             H2A.Z   \n",
       "62  22650316  None  None   None  None  22650316   H3.4_(Mammalia)   \n",
       "63  22650316  None  None   None  None  22650316  H3.Y_(Primates?)   \n",
       "64  22650316  None  None   None  None  22650316          macroH2A   \n",
       "65  22650316  None  None   None  None  22650316           OO_H1.8   \n",
       "66  22650316  None  None   None  None  22650316           TS_H1.6   \n",
       "67  22650316  None  None   None  None  22650316           TS_H1.7   \n",
       "68  22650316  None  None   None  None  22650316           TS_H1.9   \n",
       "\n",
       "   publication_id  \n",
       "24       14583738  \n",
       "54       22650316  \n",
       "55       22650316  \n",
       "56       22650316  \n",
       "57       22650316  \n",
       "58       22650316  \n",
       "59       22650316  \n",
       "60       22650316  \n",
       "61       22650316  \n",
       "62       22650316  \n",
       "63       22650316  \n",
       "64       22650316  \n",
       "65       22650316  \n",
       "66       22650316  \n",
       "67       22650316  \n",
       "68       22650316  "
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM publication p LEFT JOIN histone_has_publication ph \"\n",
    "    \"ON p.id = ph.publication_id \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "dfp = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "dfp[dfp[\"id\"].isin([\"14583738\", \"22650316\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "73a194e3-0f7f-47c4-ac42-a5ecb7fc87df",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for i, row in dfp[dfp[\"id\"].isin([\"14583738\", \"22650316\"])].iterrows():\n",
    "    cursor.execute(\n",
    "        add_histone_has_publication,\n",
    "        (row[\"histone_id\"], publications_info[row[\"publication_id\"]][\"id\"]),\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "5a4f25aa-41b6-4bbd-be77-29a033c958af",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>14583738</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>14583738</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>AAC28846.1</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>AAC37291.1</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>AAC37292.1</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>CAA33554.1</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>XP_627120.1</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>XP_641587.1</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>XP_845905.1</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>XP_846259.1</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>XP_875023.2</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>142 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           id title   doi author  year pubmed_id sequence_accession  \\\n",
       "73   14583738  None  None   None  None  14583738               None   \n",
       "119  22650316  None  None   None  None  22650316         AAC28846.1   \n",
       "120  22650316  None  None   None  None  22650316         AAC37291.1   \n",
       "121  22650316  None  None   None  None  22650316         AAC37292.1   \n",
       "122  22650316  None  None   None  None  22650316         CAA33554.1   \n",
       "..        ...   ...   ...    ...   ...       ...                ...   \n",
       "255  22650316  None  None   None  None  22650316        XP_627120.1   \n",
       "256  22650316  None  None   None  None  22650316        XP_641587.1   \n",
       "257  22650316  None  None   None  None  22650316        XP_845905.1   \n",
       "258  22650316  None  None   None  None  22650316        XP_846259.1   \n",
       "259  22650316  None  None   None  None  22650316        XP_875023.2   \n",
       "\n",
       "    publication_id  \n",
       "73            None  \n",
       "119       22650316  \n",
       "120       22650316  \n",
       "121       22650316  \n",
       "122       22650316  \n",
       "..             ...  \n",
       "255       22650316  \n",
       "256       22650316  \n",
       "257       22650316  \n",
       "258       22650316  \n",
       "259       22650316  \n",
       "\n",
       "[142 rows x 8 columns]"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM publication p LEFT JOIN sequence_has_publication sh \"\n",
    "    \"ON p.id = sh.publication_id \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "dfp = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "dfp[dfp[\"id\"].isin([\"14583738\", \"22650316\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "199954a9-2fc4-4e6b-9758-7ff10402d788",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for i, row in dfp[dfp[\"id\"].isin([\"14583738\", \"22650316\"])].iterrows():\n",
    "    if row[\"publication_id\"] is None:\n",
    "        continue\n",
    "    cursor.execute(\n",
    "        add_sequence_has_publication,\n",
    "        (row[\"sequence_accession\"], publications_info[row[\"publication_id\"]][\"id\"]),\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "4dce193c-7172-45e8-b162-a52569a5fb8a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>histone_id</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH2B</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH3</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH3_(Fungi)</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H2A</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H2B</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3_(Fungi)</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3_(Plants)</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH3</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH4</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>generic_H1</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H1.0</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H1.10</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H2A.L_(Eutheria)</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>364</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H2A.X</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H2A.Z</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3_(Animals)</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>368</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3_(Plants)</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3_(Protists)</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>370</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.Y_(Primates?)</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>macroH2A</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>OO_H1.8</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TS_H1.6</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TS_H1.7</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TS_H1.9</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           id title   doi author  year pubmed_id  \\\n",
       "228  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "229  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "230  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "231  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "232  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "233  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "234  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "235  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "236  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "237  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "238  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "357      talbert_unified_2012  None  None   None  None      None   \n",
       "358      talbert_unified_2012  None  None   None  None      None   \n",
       "359      talbert_unified_2012  None  None   None  None      None   \n",
       "360      talbert_unified_2012  None  None   None  None      None   \n",
       "361      talbert_unified_2012  None  None   None  None      None   \n",
       "362      talbert_unified_2012  None  None   None  None      None   \n",
       "363      talbert_unified_2012  None  None   None  None      None   \n",
       "364      talbert_unified_2012  None  None   None  None      None   \n",
       "365      talbert_unified_2012  None  None   None  None      None   \n",
       "366      talbert_unified_2012  None  None   None  None      None   \n",
       "367      talbert_unified_2012  None  None   None  None      None   \n",
       "368      talbert_unified_2012  None  None   None  None      None   \n",
       "369      talbert_unified_2012  None  None   None  None      None   \n",
       "370      talbert_unified_2012  None  None   None  None      None   \n",
       "371      talbert_unified_2012  None  None   None  None      None   \n",
       "372      talbert_unified_2012  None  None   None  None      None   \n",
       "373      talbert_unified_2012  None  None   None  None      None   \n",
       "374      talbert_unified_2012  None  None   None  None      None   \n",
       "375      talbert_unified_2012  None  None   None  None      None   \n",
       "376      talbert_unified_2012  None  None   None  None      None   \n",
       "\n",
       "           histone_id            publication_id  \n",
       "228    cenH3_(Plants)  malik_phylogenomics_2003  \n",
       "229              cH2A  malik_phylogenomics_2003  \n",
       "230              cH2B  malik_phylogenomics_2003  \n",
       "231               cH3  malik_phylogenomics_2003  \n",
       "232       cH3_(Fungi)  malik_phylogenomics_2003  \n",
       "233               H2A  malik_phylogenomics_2003  \n",
       "234               H2B  malik_phylogenomics_2003  \n",
       "235                H3  malik_phylogenomics_2003  \n",
       "236              H3.3  malik_phylogenomics_2003  \n",
       "237      H3.3_(Fungi)  malik_phylogenomics_2003  \n",
       "238     H3.3_(Plants)  malik_phylogenomics_2003  \n",
       "357    cenH3_(Plants)      talbert_unified_2012  \n",
       "358               cH3      talbert_unified_2012  \n",
       "359               cH4      talbert_unified_2012  \n",
       "360        generic_H1      talbert_unified_2012  \n",
       "361              H1.0      talbert_unified_2012  \n",
       "362             H1.10      talbert_unified_2012  \n",
       "363  H2A.L_(Eutheria)      talbert_unified_2012  \n",
       "364             H2A.X      talbert_unified_2012  \n",
       "365             H2A.Z      talbert_unified_2012  \n",
       "366              H3.3      talbert_unified_2012  \n",
       "367    H3.3_(Animals)      talbert_unified_2012  \n",
       "368     H3.3_(Plants)      talbert_unified_2012  \n",
       "369   H3.3_(Protists)      talbert_unified_2012  \n",
       "370   H3.4_(Mammalia)      talbert_unified_2012  \n",
       "371  H3.Y_(Primates?)      talbert_unified_2012  \n",
       "372          macroH2A      talbert_unified_2012  \n",
       "373           OO_H1.8      talbert_unified_2012  \n",
       "374           TS_H1.6      talbert_unified_2012  \n",
       "375           TS_H1.7      talbert_unified_2012  \n",
       "376           TS_H1.9      talbert_unified_2012  "
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM publication p LEFT JOIN histone_has_publication ph \"\n",
    "    \"ON p.id = ph.publication_id \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "dfp = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "dfp[dfp[\"id\"].isin([publications_info[pmid][\"id\"] for pmid in df[\"publication_id\"]])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "789bd0f4-11ca-42da-a39d-21daf1157f06",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1040</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5890</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>AAC28846.1</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5891</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>AAC37291.1</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5892</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>AAC37292.1</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5893</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>CAA33554.1</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6026</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>XP_627120.1</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6027</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>XP_641587.1</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6028</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>XP_845905.1</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6029</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>XP_846259.1</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6030</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>XP_875023.2</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>142 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            id title   doi author  year pubmed_id  \\\n",
       "1040  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "5890      talbert_unified_2012  None  None   None  None      None   \n",
       "5891      talbert_unified_2012  None  None   None  None      None   \n",
       "5892      talbert_unified_2012  None  None   None  None      None   \n",
       "5893      talbert_unified_2012  None  None   None  None      None   \n",
       "...                        ...   ...   ...    ...   ...       ...   \n",
       "6026      talbert_unified_2012  None  None   None  None      None   \n",
       "6027      talbert_unified_2012  None  None   None  None      None   \n",
       "6028      talbert_unified_2012  None  None   None  None      None   \n",
       "6029      talbert_unified_2012  None  None   None  None      None   \n",
       "6030      talbert_unified_2012  None  None   None  None      None   \n",
       "\n",
       "     sequence_accession        publication_id  \n",
       "1040               None                  None  \n",
       "5890         AAC28846.1  talbert_unified_2012  \n",
       "5891         AAC37291.1  talbert_unified_2012  \n",
       "5892         AAC37292.1  talbert_unified_2012  \n",
       "5893         CAA33554.1  talbert_unified_2012  \n",
       "...                 ...                   ...  \n",
       "6026        XP_627120.1  talbert_unified_2012  \n",
       "6027        XP_641587.1  talbert_unified_2012  \n",
       "6028        XP_845905.1  talbert_unified_2012  \n",
       "6029        XP_846259.1  talbert_unified_2012  \n",
       "6030        XP_875023.2  talbert_unified_2012  \n",
       "\n",
       "[142 rows x 8 columns]"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM publication p LEFT JOIN sequence_has_publication sh \"\n",
    "    \"ON p.id = sh.publication_id \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "dfp = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "dfp[dfp[\"id\"].isin([publications_info[pmid][\"id\"] for pmid in df[\"publication_id\"]])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62b62e23-e809-40e9-b2a5-0fdd86e8049b",
   "metadata": {},
   "source": [
    "### Delete two relations for these publications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "729bae71-e284-4041-a8b3-4074aeeeba2f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>histone_id</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>14583738</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>14583738</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>14583738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>cH4</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>generic_H1</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>H1.0</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>H1.10</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>H2A.L_(Eutheria)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>H2A.X</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>H2A.Z</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>H3.Y_(Primates?)</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>macroH2A</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>OO_H1.8</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>TS_H1.6</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>TS_H1.7</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>22650316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>22650316</td>\n",
       "      <td>TS_H1.9</td>\n",
       "      <td>22650316</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          id title   doi author  year pubmed_id        histone_id  \\\n",
       "24  14583738  None  None   None  None  14583738    cenH3_(Plants)   \n",
       "54  22650316  None  None   None  None  22650316    cenH3_(Plants)   \n",
       "55  22650316  None  None   None  None  22650316               cH4   \n",
       "56  22650316  None  None   None  None  22650316        generic_H1   \n",
       "57  22650316  None  None   None  None  22650316              H1.0   \n",
       "58  22650316  None  None   None  None  22650316             H1.10   \n",
       "59  22650316  None  None   None  None  22650316  H2A.L_(Eutheria)   \n",
       "60  22650316  None  None   None  None  22650316             H2A.X   \n",
       "61  22650316  None  None   None  None  22650316             H2A.Z   \n",
       "62  22650316  None  None   None  None  22650316   H3.4_(Mammalia)   \n",
       "63  22650316  None  None   None  None  22650316  H3.Y_(Primates?)   \n",
       "64  22650316  None  None   None  None  22650316          macroH2A   \n",
       "65  22650316  None  None   None  None  22650316           OO_H1.8   \n",
       "66  22650316  None  None   None  None  22650316           TS_H1.6   \n",
       "67  22650316  None  None   None  None  22650316           TS_H1.7   \n",
       "68  22650316  None  None   None  None  22650316           TS_H1.9   \n",
       "\n",
       "   publication_id  \n",
       "24       14583738  \n",
       "54       22650316  \n",
       "55       22650316  \n",
       "56       22650316  \n",
       "57       22650316  \n",
       "58       22650316  \n",
       "59       22650316  \n",
       "60       22650316  \n",
       "61       22650316  \n",
       "62       22650316  \n",
       "63       22650316  \n",
       "64       22650316  \n",
       "65       22650316  \n",
       "66       22650316  \n",
       "67       22650316  \n",
       "68       22650316  "
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM publication p LEFT JOIN histone_has_publication ph \"\n",
    "    \"ON p.id = ph.publication_id \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "dfp = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "dfp[dfp[\"id\"].isin([\"14583738\", \"22650316\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "f44ec558-19d0-4197-8061-62bff859c81e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for del_id in [\"14583738\", \"22650316\"]:\n",
    "    query = f\"DELETE FROM histone_has_publication WHERE publication_id='{del_id}' \"\n",
    "    cursor.execute(query)\n",
    "    query = f\"DELETE FROM sequence_has_publication WHERE publication_id='{del_id}' \"\n",
    "    cursor.execute(query)\n",
    "    query = f\"DELETE FROM publication WHERE id='{del_id}' \"\n",
    "    cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "16455023-5a8f-4ffa-8f50-8b4ee136d839",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>histone_id</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, title, doi, author, year, pubmed_id, histone_id, publication_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM publication p LEFT JOIN histone_has_publication ph \"\n",
    "    \"ON p.id = ph.publication_id \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "dfp = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "dfp[dfp[\"id\"].isin([\"14583738\", \"22650316\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "bc4d3647-b505-4e93-98d6-a6b50889bba9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>histone_id</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH2B</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH3</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH3_(Fungi)</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H2A</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H2B</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3_(Fungi)</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3_(Plants)</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH3</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>343</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH4</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>generic_H1</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H1.0</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H1.10</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H2A.L_(Eutheria)</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>348</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H2A.X</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H2A.Z</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3_(Animals)</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3_(Plants)</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.3_(Protists)</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.Y_(Primates?)</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>macroH2A</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>OO_H1.8</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TS_H1.6</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TS_H1.7</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>talbert_unified_2012</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TS_H1.9</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           id title   doi author  year pubmed_id  \\\n",
       "212  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "213  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "214  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "215  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "216  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "217  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "218  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "219  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "220  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "221  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "222  malik_phylogenomics_2003  None  None   None  None      None   \n",
       "341      talbert_unified_2012  None  None   None  None      None   \n",
       "342      talbert_unified_2012  None  None   None  None      None   \n",
       "343      talbert_unified_2012  None  None   None  None      None   \n",
       "344      talbert_unified_2012  None  None   None  None      None   \n",
       "345      talbert_unified_2012  None  None   None  None      None   \n",
       "346      talbert_unified_2012  None  None   None  None      None   \n",
       "347      talbert_unified_2012  None  None   None  None      None   \n",
       "348      talbert_unified_2012  None  None   None  None      None   \n",
       "349      talbert_unified_2012  None  None   None  None      None   \n",
       "350      talbert_unified_2012  None  None   None  None      None   \n",
       "351      talbert_unified_2012  None  None   None  None      None   \n",
       "352      talbert_unified_2012  None  None   None  None      None   \n",
       "353      talbert_unified_2012  None  None   None  None      None   \n",
       "354      talbert_unified_2012  None  None   None  None      None   \n",
       "355      talbert_unified_2012  None  None   None  None      None   \n",
       "356      talbert_unified_2012  None  None   None  None      None   \n",
       "357      talbert_unified_2012  None  None   None  None      None   \n",
       "358      talbert_unified_2012  None  None   None  None      None   \n",
       "359      talbert_unified_2012  None  None   None  None      None   \n",
       "360      talbert_unified_2012  None  None   None  None      None   \n",
       "\n",
       "           histone_id            publication_id  \n",
       "212    cenH3_(Plants)  malik_phylogenomics_2003  \n",
       "213              cH2A  malik_phylogenomics_2003  \n",
       "214              cH2B  malik_phylogenomics_2003  \n",
       "215               cH3  malik_phylogenomics_2003  \n",
       "216       cH3_(Fungi)  malik_phylogenomics_2003  \n",
       "217               H2A  malik_phylogenomics_2003  \n",
       "218               H2B  malik_phylogenomics_2003  \n",
       "219                H3  malik_phylogenomics_2003  \n",
       "220              H3.3  malik_phylogenomics_2003  \n",
       "221      H3.3_(Fungi)  malik_phylogenomics_2003  \n",
       "222     H3.3_(Plants)  malik_phylogenomics_2003  \n",
       "341    cenH3_(Plants)      talbert_unified_2012  \n",
       "342               cH3      talbert_unified_2012  \n",
       "343               cH4      talbert_unified_2012  \n",
       "344        generic_H1      talbert_unified_2012  \n",
       "345              H1.0      talbert_unified_2012  \n",
       "346             H1.10      talbert_unified_2012  \n",
       "347  H2A.L_(Eutheria)      talbert_unified_2012  \n",
       "348             H2A.X      talbert_unified_2012  \n",
       "349             H2A.Z      talbert_unified_2012  \n",
       "350              H3.3      talbert_unified_2012  \n",
       "351    H3.3_(Animals)      talbert_unified_2012  \n",
       "352     H3.3_(Plants)      talbert_unified_2012  \n",
       "353   H3.3_(Protists)      talbert_unified_2012  \n",
       "354   H3.4_(Mammalia)      talbert_unified_2012  \n",
       "355  H3.Y_(Primates?)      talbert_unified_2012  \n",
       "356          macroH2A      talbert_unified_2012  \n",
       "357           OO_H1.8      talbert_unified_2012  \n",
       "358           TS_H1.6      talbert_unified_2012  \n",
       "359           TS_H1.7      talbert_unified_2012  \n",
       "360           TS_H1.9      talbert_unified_2012  "
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM publication p LEFT JOIN histone_has_publication ph \"\n",
    "    \"ON p.id = ph.publication_id \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "dfp = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "dfp[dfp[\"id\"].isin([publications_info[pmid][\"id\"] for pmid in df[\"publication_id\"]])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "363069ed-087c-4fae-9615-cd87add0c4a4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98eafd7f-2d53-44c1-ba88-c8233ffeb092",
   "metadata": {},
   "source": [
    "### Change ids for other publications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "d13a1242-6edb-4172-bb53-f9aa7225819e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>histone_id</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>19766562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>21743476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>23324462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>25956076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "0  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "1  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "2  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "3  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "4  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "5  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "\n",
       "  parent      histone_id            publication_id  \n",
       "0  cenH3  cenH3_(Plants)                  19766562  \n",
       "1  cenH3  cenH3_(Plants)                  21743476  \n",
       "2  cenH3  cenH3_(Plants)                  23324462  \n",
       "3  cenH3  cenH3_(Plants)                  25956076  \n",
       "4  cenH3  cenH3_(Plants)  malik_phylogenomics_2003  \n",
       "5  cenH3  cenH3_(Plants)      talbert_unified_2012  "
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_has_publication hp \"\n",
    "    \"ON h.id = hp.histone_id \"\n",
    "    \"WHERE h.id='cenH3_(Plants)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "21f3d245-cb22-45b0-a642-3bf0a8b0ae3a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    14583738\n",
       "1    19766562\n",
       "2    21743476\n",
       "3    22650316\n",
       "4    23324462\n",
       "5    25956076\n",
       "Name: publication_id, dtype: object"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"publication_id\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "ad2e7a26-566b-4882-9408-97dc2ca2bea8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>19766562</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>19766562</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>21743476</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>21743476</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>23324462</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>23324462</td>\n",
       "      <td>NP_001035891.1</td>\n",
       "      <td>23324462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>23324462</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>23324462</td>\n",
       "      <td>NP_001800.1</td>\n",
       "      <td>23324462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>25956076</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>25956076</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           id title   doi author  year pubmed_id sequence_accession  \\\n",
       "99   19766562  None  None   None  None  19766562               None   \n",
       "111  21743476  None  None   None  None  21743476               None   \n",
       "120  23324462  None  None   None  None  23324462     NP_001035891.1   \n",
       "121  23324462  None  None   None  None  23324462        NP_001800.1   \n",
       "171  25956076  None  None   None  None  25956076               None   \n",
       "\n",
       "    publication_id  \n",
       "99            None  \n",
       "111           None  \n",
       "120       23324462  \n",
       "121       23324462  \n",
       "171           None  "
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM publication p LEFT JOIN sequence_has_publication sh \"\n",
    "    \"ON p.id = sh.publication_id \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "dfp = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "dfp[dfp[\"id\"].isin(df[\"publication_id\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "b4136abe-c195-46a4-b2ca-7073d9c6e9f4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_001035891.1</td>\n",
       "      <td>23324462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_001800.1</td>\n",
       "      <td>23324462</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sequence_accession publication_id\n",
       "0     NP_001035891.1       23324462\n",
       "1        NP_001800.1       23324462"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence_has_publication \" \"WHERE publication_id = '23324462' \"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "03d7d293-2525-4a3d-a013-a2d6f30b9cf5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "two_accessions = [\"NP_001035891.1\", \"NP_001800.1\"]\n",
    "query = f\"DELETE FROM sequence_has_publication WHERE publication_id='23324462' \"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "dfe6d011-c4dc-46db-b7f7-57a924fce4c0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UPDATE publication SET id='malik_major_2009', title='Major evolutionary transitions in centromere complexity', doi='10.1016/j.cell.2009.08.036', year='2009' WHERE id='19766562'\n",
      "UPDATE publication SET id='tachiwana_crystal_2011', title='Crystal structure of the human centromeric nucleosome containing CENP-A', doi='10.1038/nature10258', year='2011' WHERE id='21743476'\n",
      "UPDATE publication SET id='bui_cenp-nucleosome_2013', title='The CENP-A nucleosome: a battle between Dr Jekyll and Mr Hyde', doi='10.4161/nucl.23588', year='2013' WHERE id='23324462'\n",
      "UPDATE publication SET id='steiner_diversity_2015', title='Diversity in the organization of centromeric chromatin', doi='10.1016/j.gde.2015.03.010', year='2015' WHERE id='25956076'\n"
     ]
    }
   ],
   "source": [
    "for pubmed_id in df[\"publication_id\"]:\n",
    "    if pubmed_id in [\"14583738\", \"22650316\"]:\n",
    "        continue\n",
    "    query = f\"UPDATE publication SET id='{publications_info[pubmed_id]['id']}', title='{publications_info[pubmed_id]['title']}', doi='{publications_info[pubmed_id]['doi']}', year='{publications_info[pubmed_id]['year']}' WHERE id='{pubmed_id}'\"\n",
    "    print(query)\n",
    "    cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "95f6a9f5-3c7d-4b4a-8526-1a844ed2e056",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>histone_id</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>bui_cenp-nucleosome_2013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>malik_major_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>steiner_diversity_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>tachiwana_crystal_2011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "0  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "1  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "2  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "3  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "4  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "5  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "\n",
       "  parent      histone_id            publication_id  \n",
       "0  cenH3  cenH3_(Plants)  bui_cenp-nucleosome_2013  \n",
       "1  cenH3  cenH3_(Plants)          malik_major_2009  \n",
       "2  cenH3  cenH3_(Plants)  malik_phylogenomics_2003  \n",
       "3  cenH3  cenH3_(Plants)    steiner_diversity_2015  \n",
       "4  cenH3  cenH3_(Plants)    tachiwana_crystal_2011  \n",
       "5  cenH3  cenH3_(Plants)      talbert_unified_2012  "
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_has_publication hp \"\n",
    "    \"ON h.id = hp.histone_id \"\n",
    "    \"WHERE h.id='cenH3_(Plants)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "4a35402d-4208-461d-b3c8-a88f3e8bd8e6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for a in two_accessions:\n",
    "    cursor.execute(\n",
    "        add_sequence_has_publication,\n",
    "        (a, publications_info[\"23324462\"][\"id\"]),\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "a739f053-6c69-4e5c-b89c-0bef103682fa",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_001035891.1</td>\n",
       "      <td>bui_cenp-nucleosome_2013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_001800.1</td>\n",
       "      <td>bui_cenp-nucleosome_2013</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sequence_accession            publication_id\n",
       "0     NP_001035891.1  bui_cenp-nucleosome_2013\n",
       "1        NP_001800.1  bui_cenp-nucleosome_2013"
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence_has_publication \"\n",
    "    f\"WHERE publication_id = '{publications_info['23324462']['id']}' \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "d9caf85b-e282-4b87-949a-c9615063d000",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10dd42b5-4459-4153-a7c7-d1dc87ace78a",
   "metadata": {},
   "source": [
    "## Add publications for cenH3_(Plants)\n",
    "\n",
    "kawabe_duplication_2006, neumann_centromeres_2015, neumann_stretching_2012, sanei_loss_2011, karimi-ashtiyani_point_2015, ishii_differential_2015, evtushenko_conserved_2017"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "7c7e9025-83b1-465f-b758-23216e26fc73",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['kawabe_duplication_2006',\n",
       " 'neumann_centromeres_2015',\n",
       " 'neumann_stretching_2012',\n",
       " 'sanei_loss_2011',\n",
       " 'karimi-ashtiyani_point_2015',\n",
       " 'ishii_differential_2015',\n",
       " 'evtushenko_conserved_2017']"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "bec16eb8-45e2-406c-be0c-91f97aafa318",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, title, doi, author, year, pubmed_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "0fab58c1-ca21-4623-881b-0e71141e8d9c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_publication = [\n",
    "    {\n",
    "        \"id\": pid,\n",
    "        \"title\": None,\n",
    "        \"doi\": None,\n",
    "        \"author\": None,\n",
    "        \"year\": None,\n",
    "    }\n",
    "    for pid in pids\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "eecc416f-1015-4864-b836-3e603e14a261",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for dp in data_publication:\n",
    "    cursor.execute(add_publication, dp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "5484aec4-fd50-47e2-8d30-6ac32676a6e8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for pid in pids:\n",
    "    cursor.execute(add_histone_has_publication, (\"cenH3_(Plants)\", pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "72487731-eba0-4928-8f0c-10a11482c648",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>histone_id</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>bui_cenp-nucleosome_2013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>evtushenko_conserved_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>ishii_differential_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>karimi-ashtiyani_point_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>kawabe_duplication_2006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>malik_major_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>malik_phylogenomics_2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>neumann_centromeres_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>neumann_stretching_2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>sanei_loss_2011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>steiner_diversity_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>tachiwana_crystal_2011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>233</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3_(Plants)</td>\n",
       "      <td>talbert_unified_2012</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "0   cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "1   cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "2   cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "3   cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "4   cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "5   cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "6   cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "7   cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "8   cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "9   cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "10  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "11  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "12  cenH3_(Plants)  variant     Eukaryotes              2759          233   \n",
       "\n",
       "   parent      histone_id               publication_id  \n",
       "0   cenH3  cenH3_(Plants)     bui_cenp-nucleosome_2013  \n",
       "1   cenH3  cenH3_(Plants)    evtushenko_conserved_2017  \n",
       "2   cenH3  cenH3_(Plants)      ishii_differential_2015  \n",
       "3   cenH3  cenH3_(Plants)  karimi-ashtiyani_point_2015  \n",
       "4   cenH3  cenH3_(Plants)      kawabe_duplication_2006  \n",
       "5   cenH3  cenH3_(Plants)             malik_major_2009  \n",
       "6   cenH3  cenH3_(Plants)     malik_phylogenomics_2003  \n",
       "7   cenH3  cenH3_(Plants)     neumann_centromeres_2015  \n",
       "8   cenH3  cenH3_(Plants)      neumann_stretching_2012  \n",
       "9   cenH3  cenH3_(Plants)              sanei_loss_2011  \n",
       "10  cenH3  cenH3_(Plants)       steiner_diversity_2015  \n",
       "11  cenH3  cenH3_(Plants)       tachiwana_crystal_2011  \n",
       "12  cenH3  cenH3_(Plants)         talbert_unified_2012  "
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_has_publication hp \"\n",
    "    \"ON h.id = hp.histone_id \"\n",
    "    \"WHERE h.id='cenH3_(Plants)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "2aeec71c-81e7-40d7-a726-b0fd16512dc1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e8bad07-ef74-40fd-a93b-e7897df0c591",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Close connections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "e5673501-5845-491a-bbb9-f5d213721d58",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cursor.close()\n",
    "conn.close()\n",
    "tunnel.stop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39b93075-4c7c-468f-bb04-e9750f311dd9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".conda-histdb_env",
   "language": "python",
   "name": "conda-env-.conda-histdb_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
