{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "53d3ec55-7daf-4ade-8add-2830b709dc54",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import io\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import Entrez, SeqIO\n",
    "from mysql.connector import connection\n",
    "from sshtunnel import SSHTunnelForwarder\n",
    "\n",
    "# Укажите ваш email (обязательно для использования Entrez)\n",
    "Entrez.email = \"your.email@example.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "608c4bfe-b979-4b60-b2b5-b94eeb679cb5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open(\"db_curated_server_info.txt\", \"r\") as file:\n",
    "    lines = file.readlines()\n",
    "\n",
    "config = {}\n",
    "\n",
    "for line in lines:\n",
    "    line = line.strip()\n",
    "    if line and not line.startswith(\"#\"):\n",
    "        key, value = line.split(\"=\", 1)\n",
    "        config[key] = value.strip()\n",
    "\n",
    "server_name = config.get(\"server_name\")\n",
    "srever_port = int(config.get(\"srever_port\"))\n",
    "ssh_password = config.get(\"ssh_password\")\n",
    "ssh_username = config.get(\"ssh_username\")\n",
    "db_adress = config.get(\"db_adress\")\n",
    "db_port = int(config.get(\"db_port\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4e19aa1c-49a3-48c4-b67f-34c94cef608d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "38721\n"
     ]
    }
   ],
   "source": [
    "tunnel = SSHTunnelForwarder(\n",
    "    (server_name, srever_port),\n",
    "    ssh_password=ssh_password,\n",
    "    ssh_username=ssh_username,\n",
    "    remote_bind_address=(db_adress, db_port),\n",
    ")\n",
    "tunnel.start()\n",
    "print(tunnel.local_bind_port)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cef2275b-12d0-48f1-ad3c-ce4cdb9aa267",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "conn = connection.MySQLConnection(\n",
    "    user=\"db_user\",\n",
    "    password=\"db_password\",\n",
    "    host=\"localhost\",\n",
    "    port=tunnel.local_bind_port,\n",
    "    database=\"db_name\",\n",
    ")\n",
    "cursor = conn.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f589ea8d-8b83-4331-a1cf-b64ed7621801",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('alternative_name',),\n",
       " ('histone',),\n",
       " ('histone_description',),\n",
       " ('histone_has_publication',),\n",
       " ('publication',),\n",
       " ('sequence',),\n",
       " ('sequence_has_publication',)]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SHOW TABLES;\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b3dd9ef8-b162-4127-9535-ac7743af8c69",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# add_histone = (\n",
    "#     \"INSERT INTO histone \"\n",
    "#     \"(id, level, taxonomic_span, taxonomic_span_id, description, parent) \"\n",
    "#     \"VALUES (%(id)s, %(level)s, %(taxonomic_span)s, %(taxonomic_span_id)s, %(description)s, %(parent)s)\"\n",
    "# )\n",
    "# add_histone_description = (\n",
    "#     \"INSERT INTO histone_description \"\n",
    "#     \"(summary, taxonomy, genes, evolution, expression, knock_out, function, sequence, localization, deposition, structure, interactions, disease, caveats) \"\n",
    "#     \"VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)\"\n",
    "# )\n",
    "# add_publication = (\n",
    "#     \"INSERT INTO publication \"\n",
    "#     \"(id, title, doi, author, year) \"\n",
    "#     \"VALUES (%(id)s, %(title)s, %(doi)s, %(author)s, %(year)s)\"\n",
    "# )\n",
    "# add_sequence = (\n",
    "#     \"INSERT INTO sequence \"\n",
    "#     \"(accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration) \"\n",
    "#     \"VALUES (%(accession)s, %(variant)s, %(gi)s, %(ncbi_gene_id)s, %(hgnc_gene_name)s, %(taxonomy_id)s, %(organism)s, %(phylum)s, %(class)s, %(taxonomy_group)s, %(info)s, %(sequence)s, %(variant_under_consideration)s)\"\n",
    "# )\n",
    "# add_sequence_has_publication = (\n",
    "#     \"INSERT INTO sequence_has_publication \"\n",
    "#     \"(sequence_accession, publication_id) \"\n",
    "#     \"VALUES (%s, %s)\"\n",
    "# )\n",
    "# add_alternate_names = (\n",
    "#     \"INSERT INTO alternative_name \"\n",
    "#     \"(name, taxonomy, gene, splice, histone) \"\n",
    "#     \"VALUES (%(name)s, %(taxonomy)s, %(gene)s, %(splice)s, %(histone)s)\"\n",
    "# )\n",
    "add_histone_has_publication = (\n",
    "    \"INSERT INTO histone_has_publication \"\n",
    "    \"(histone_id, publication_id) \"\n",
    "    \"VALUES (%s, %s)\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1c3bb0c-180f-4fb2-9cdc-ca928f9808be",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_taxonomy_data(record):\n",
    "    import re\n",
    "    import sys\n",
    "\n",
    "    taxonomy_data = {}\n",
    "    taxonomy_data[\"organism\"] = record.annotations[\"organism\"]\n",
    "    try:\n",
    "        for a in record.features[0].qualifiers[\"db_xref\"]:\n",
    "            text = re.search(\"(\\S+):(\\S+)\", a).group(1)\n",
    "            taxid = re.search(\"(\\S+):(\\S+)\", a).group(2)\n",
    "            if text == \"taxon\":\n",
    "                print(f\"Fetched taxid from NCBI {taxid}\")\n",
    "                taxonomy_data[\"taxonomy_id\"] = int(taxid)\n",
    "            else:\n",
    "                continue\n",
    "    except:\n",
    "        print(\"!!!!!!Unable to get TAXID for this record setting it to 1\")\n",
    "        taxonomy_data[\"taxonomy_id\"] = 1  # unable to identify\n",
    "\n",
    "    lineage = dict()\n",
    "    for i in range(10):\n",
    "        try:\n",
    "            handle = Entrez.efetch(\n",
    "                id=taxonomy_data[\"taxonomy_id\"], db=\"taxonomy\", retmode=\"xml\"\n",
    "            )\n",
    "            tax_data = Entrez.read(handle)\n",
    "            lineage = {\n",
    "                d[\"Rank\"]: d[\"ScientificName\"]\n",
    "                for d in tax_data[0][\"LineageEx\"]\n",
    "                if d[\"Rank\"] in [\"class\", \"phylum\"]\n",
    "            }\n",
    "            break\n",
    "        except:\n",
    "            print(\n",
    "                \"Unexpected error: {}, Retrying, attempt {}\".format(\n",
    "                    sys.exc_info()[0], i\n",
    "                )\n",
    "            )\n",
    "            if i == 9:\n",
    "                print(\n",
    "                    f\"FATAL ERROR could not get class and phylum from NCBI after 10 attempts for taxid:{taxonomy_data['taxonomy_id']}. Will add None for class and phylum!\"\n",
    "                )\n",
    "            else:\n",
    "                continue\n",
    "    taxonomy_data[\"phylum\"] = lineage.get(\"phylum\", None)\n",
    "    taxonomy_data[\"class\"] = lineage.get(\"class\", None)\n",
    "    if taxonomy_data[\"phylum\"] is not None:\n",
    "        taxonomy_data[\"phylum\"] = str(taxonomy_data[\"phylum\"])\n",
    "    if taxonomy_data[\"class\"] is not None:\n",
    "        taxonomy_data[\"class\"] = str(taxonomy_data[\"class\"])\n",
    "    return taxonomy_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c060c047-3e7d-4095-b608-9722031af692",
   "metadata": {},
   "source": [
    "# To Do\n",
    "    \n",
    "## <span style=\"color:black\">Add publicationes for added histone nodes</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7484ea7-b0b9-460f-ab25-bf8ef6d091bf",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add publicationes for added histone nodes</span>\n",
    "\n",
    "**New *Drosophilidae* nodes:** cenH3.1_(Drosophilidae), cenH3.2_(Drosophila_eugracilis), cenH3.3_(Montium), cenH3.4_(Montium), cenH3.5_(Drosophila), cenH3.6_(Repleta)\n",
    "\n",
    "**Publications:** kursel_recurrent_2017, kursel_gametic_2021, teixeira_concurrent_2018\n",
    "\n",
    "**New *Culicidae* nodes:** cenH3.1_(Culicidae), cenH3.2_(Culicidae), cenH3.3_(Aedes)\n",
    "\n",
    "**Publications:** kursel_ancient_2020"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64f1c212-34db-43c9-a391-44954804ac5f",
   "metadata": {},
   "source": [
    "### Add Drosophilidae histones publications\n",
    "\n",
    "**New *Drosophilidae* nodes:** cenH3.1_(Drosophilidae), cenH3.2_(Drosophila_eugracilis), cenH3.3_(Montium), cenH3.4_(Montium), cenH3.5_(Drosophila), cenH3.6_(Repleta)\n",
    "\n",
    "**Publications:** kursel_recurrent_2017, kursel_gametic_2021, teixeira_concurrent_2018"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1983964b-63f9-4bee-b481-7254886460f6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>kursel_gametic_2021</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>kursel_recurrent_2017</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>teixeira_concurrent_2018</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           id title   doi author  year pubmed_id\n",
       "118       kursel_gametic_2021  None  None   None  None      None\n",
       "119     kursel_recurrent_2017  None  None   None  None      None\n",
       "173  teixeira_concurrent_2018  None  None   None  None      None"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"kursel_recurrent_2017\", \"kursel_gametic_2021\", \"teixeira_concurrent_2018\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "42461515-4499-474e-8e88-22d8b7b4b2cc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "variants = [\n",
    "    \"cenH3.1_(Drosophilidae)\",\n",
    "    \"cenH3.2_(Drosophila_eugracilis)\",\n",
    "    \"cenH3.3_(Montium)\",\n",
    "    \"cenH3.4_(Montium)\",\n",
    "    \"cenH3.5_(Drosophila)\",\n",
    "    \"cenH3.6_(Repleta)\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "672e4322-8d7f-485c-8345-3b7d0abaef5f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>histone_id</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>cenH3.1_(Drosophilidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophilidae</td>\n",
       "      <td>7214</td>\n",
       "      <td>240.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>cenH3.2_(Drosophila_eugracilis)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophila eugracilis</td>\n",
       "      <td>29029</td>\n",
       "      <td>241.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>cenH3.3_(Montium)</td>\n",
       "      <td>variant</td>\n",
       "      <td>montium subgroup</td>\n",
       "      <td>32352</td>\n",
       "      <td>242.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>cenH3.4_(Montium)</td>\n",
       "      <td>variant</td>\n",
       "      <td>montium subgroup</td>\n",
       "      <td>32352</td>\n",
       "      <td>243.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>cenH3.5_(Drosophila)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophila</td>\n",
       "      <td>32281</td>\n",
       "      <td>244.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>cenH3.6_(Repleta)</td>\n",
       "      <td>variant</td>\n",
       "      <td>repleta group</td>\n",
       "      <td>32321</td>\n",
       "      <td>245.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 id    level         taxonomic_span  \\\n",
       "6           cenH3.1_(Drosophilidae)  variant          Drosophilidae   \n",
       "10  cenH3.2_(Drosophila_eugracilis)  variant  Drosophila eugracilis   \n",
       "14                cenH3.3_(Montium)  variant       montium subgroup   \n",
       "15                cenH3.4_(Montium)  variant       montium subgroup   \n",
       "16             cenH3.5_(Drosophila)  variant             Drosophila   \n",
       "17                cenH3.6_(Repleta)  variant          repleta group   \n",
       "\n",
       "   taxonomic_span_id  description                 parent histone_id  \\\n",
       "6               7214        240.0  cenH3_(Drosophilidae)       None   \n",
       "10             29029        241.0  cenH3_(Drosophilidae)       None   \n",
       "14             32352        242.0  cenH3_(Drosophilidae)       None   \n",
       "15             32352        243.0  cenH3_(Drosophilidae)       None   \n",
       "16             32281        244.0  cenH3_(Drosophilidae)       None   \n",
       "17             32321        245.0  cenH3_(Drosophilidae)       None   \n",
       "\n",
       "   publication_id  \n",
       "6            None  \n",
       "10           None  \n",
       "14           None  \n",
       "15           None  \n",
       "16           None  \n",
       "17           None  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_has_publication hp \"\n",
    "    \"ON h.id = hp.histone_id \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(variants)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "78ab9dd7-ffdf-4cea-9f18-5d217f45b63b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for v in variants:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_histone_has_publication, (v, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "30d3e021-dd97-42a5-820d-7b6fe4b052ae",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>histone_id</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>cenH3.1_(Drosophilidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophilidae</td>\n",
       "      <td>7214</td>\n",
       "      <td>240.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>cenH3.1_(Drosophilidae)</td>\n",
       "      <td>kursel_gametic_2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>cenH3.1_(Drosophilidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophilidae</td>\n",
       "      <td>7214</td>\n",
       "      <td>240.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>cenH3.1_(Drosophilidae)</td>\n",
       "      <td>kursel_recurrent_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>cenH3.1_(Drosophilidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophilidae</td>\n",
       "      <td>7214</td>\n",
       "      <td>240.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>cenH3.1_(Drosophilidae)</td>\n",
       "      <td>teixeira_concurrent_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>cenH3.2_(Drosophila_eugracilis)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophila eugracilis</td>\n",
       "      <td>29029</td>\n",
       "      <td>241.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>cenH3.2_(Drosophila_eugracilis)</td>\n",
       "      <td>kursel_gametic_2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>cenH3.2_(Drosophila_eugracilis)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophila eugracilis</td>\n",
       "      <td>29029</td>\n",
       "      <td>241.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>cenH3.2_(Drosophila_eugracilis)</td>\n",
       "      <td>kursel_recurrent_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>cenH3.2_(Drosophila_eugracilis)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophila eugracilis</td>\n",
       "      <td>29029</td>\n",
       "      <td>241.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>cenH3.2_(Drosophila_eugracilis)</td>\n",
       "      <td>teixeira_concurrent_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>cenH3.3_(Montium)</td>\n",
       "      <td>variant</td>\n",
       "      <td>montium subgroup</td>\n",
       "      <td>32352</td>\n",
       "      <td>242.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>cenH3.3_(Montium)</td>\n",
       "      <td>kursel_gametic_2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>cenH3.3_(Montium)</td>\n",
       "      <td>variant</td>\n",
       "      <td>montium subgroup</td>\n",
       "      <td>32352</td>\n",
       "      <td>242.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>cenH3.3_(Montium)</td>\n",
       "      <td>kursel_recurrent_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>cenH3.3_(Montium)</td>\n",
       "      <td>variant</td>\n",
       "      <td>montium subgroup</td>\n",
       "      <td>32352</td>\n",
       "      <td>242.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>cenH3.3_(Montium)</td>\n",
       "      <td>teixeira_concurrent_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>cenH3.4_(Montium)</td>\n",
       "      <td>variant</td>\n",
       "      <td>montium subgroup</td>\n",
       "      <td>32352</td>\n",
       "      <td>243.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>cenH3.4_(Montium)</td>\n",
       "      <td>kursel_gametic_2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>cenH3.4_(Montium)</td>\n",
       "      <td>variant</td>\n",
       "      <td>montium subgroup</td>\n",
       "      <td>32352</td>\n",
       "      <td>243.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>cenH3.4_(Montium)</td>\n",
       "      <td>kursel_recurrent_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>cenH3.4_(Montium)</td>\n",
       "      <td>variant</td>\n",
       "      <td>montium subgroup</td>\n",
       "      <td>32352</td>\n",
       "      <td>243.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>cenH3.4_(Montium)</td>\n",
       "      <td>teixeira_concurrent_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>cenH3.5_(Drosophila)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophila</td>\n",
       "      <td>32281</td>\n",
       "      <td>244.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>cenH3.5_(Drosophila)</td>\n",
       "      <td>kursel_gametic_2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>cenH3.5_(Drosophila)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophila</td>\n",
       "      <td>32281</td>\n",
       "      <td>244.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>cenH3.5_(Drosophila)</td>\n",
       "      <td>kursel_recurrent_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>cenH3.5_(Drosophila)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophila</td>\n",
       "      <td>32281</td>\n",
       "      <td>244.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>cenH3.5_(Drosophila)</td>\n",
       "      <td>teixeira_concurrent_2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>cenH3.6_(Repleta)</td>\n",
       "      <td>variant</td>\n",
       "      <td>repleta group</td>\n",
       "      <td>32321</td>\n",
       "      <td>245.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>cenH3.6_(Repleta)</td>\n",
       "      <td>kursel_gametic_2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>cenH3.6_(Repleta)</td>\n",
       "      <td>variant</td>\n",
       "      <td>repleta group</td>\n",
       "      <td>32321</td>\n",
       "      <td>245.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>cenH3.6_(Repleta)</td>\n",
       "      <td>kursel_recurrent_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>cenH3.6_(Repleta)</td>\n",
       "      <td>variant</td>\n",
       "      <td>repleta group</td>\n",
       "      <td>32321</td>\n",
       "      <td>245.0</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>cenH3.6_(Repleta)</td>\n",
       "      <td>teixeira_concurrent_2018</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 id    level         taxonomic_span  \\\n",
       "6           cenH3.1_(Drosophilidae)  variant          Drosophilidae   \n",
       "7           cenH3.1_(Drosophilidae)  variant          Drosophilidae   \n",
       "8           cenH3.1_(Drosophilidae)  variant          Drosophilidae   \n",
       "12  cenH3.2_(Drosophila_eugracilis)  variant  Drosophila eugracilis   \n",
       "13  cenH3.2_(Drosophila_eugracilis)  variant  Drosophila eugracilis   \n",
       "14  cenH3.2_(Drosophila_eugracilis)  variant  Drosophila eugracilis   \n",
       "18                cenH3.3_(Montium)  variant       montium subgroup   \n",
       "19                cenH3.3_(Montium)  variant       montium subgroup   \n",
       "20                cenH3.3_(Montium)  variant       montium subgroup   \n",
       "21                cenH3.4_(Montium)  variant       montium subgroup   \n",
       "22                cenH3.4_(Montium)  variant       montium subgroup   \n",
       "23                cenH3.4_(Montium)  variant       montium subgroup   \n",
       "24             cenH3.5_(Drosophila)  variant             Drosophila   \n",
       "25             cenH3.5_(Drosophila)  variant             Drosophila   \n",
       "26             cenH3.5_(Drosophila)  variant             Drosophila   \n",
       "27                cenH3.6_(Repleta)  variant          repleta group   \n",
       "28                cenH3.6_(Repleta)  variant          repleta group   \n",
       "29                cenH3.6_(Repleta)  variant          repleta group   \n",
       "\n",
       "   taxonomic_span_id  description                 parent  \\\n",
       "6               7214        240.0  cenH3_(Drosophilidae)   \n",
       "7               7214        240.0  cenH3_(Drosophilidae)   \n",
       "8               7214        240.0  cenH3_(Drosophilidae)   \n",
       "12             29029        241.0  cenH3_(Drosophilidae)   \n",
       "13             29029        241.0  cenH3_(Drosophilidae)   \n",
       "14             29029        241.0  cenH3_(Drosophilidae)   \n",
       "18             32352        242.0  cenH3_(Drosophilidae)   \n",
       "19             32352        242.0  cenH3_(Drosophilidae)   \n",
       "20             32352        242.0  cenH3_(Drosophilidae)   \n",
       "21             32352        243.0  cenH3_(Drosophilidae)   \n",
       "22             32352        243.0  cenH3_(Drosophilidae)   \n",
       "23             32352        243.0  cenH3_(Drosophilidae)   \n",
       "24             32281        244.0  cenH3_(Drosophilidae)   \n",
       "25             32281        244.0  cenH3_(Drosophilidae)   \n",
       "26             32281        244.0  cenH3_(Drosophilidae)   \n",
       "27             32321        245.0  cenH3_(Drosophilidae)   \n",
       "28             32321        245.0  cenH3_(Drosophilidae)   \n",
       "29             32321        245.0  cenH3_(Drosophilidae)   \n",
       "\n",
       "                         histone_id            publication_id  \n",
       "6           cenH3.1_(Drosophilidae)       kursel_gametic_2021  \n",
       "7           cenH3.1_(Drosophilidae)     kursel_recurrent_2017  \n",
       "8           cenH3.1_(Drosophilidae)  teixeira_concurrent_2018  \n",
       "12  cenH3.2_(Drosophila_eugracilis)       kursel_gametic_2021  \n",
       "13  cenH3.2_(Drosophila_eugracilis)     kursel_recurrent_2017  \n",
       "14  cenH3.2_(Drosophila_eugracilis)  teixeira_concurrent_2018  \n",
       "18                cenH3.3_(Montium)       kursel_gametic_2021  \n",
       "19                cenH3.3_(Montium)     kursel_recurrent_2017  \n",
       "20                cenH3.3_(Montium)  teixeira_concurrent_2018  \n",
       "21                cenH3.4_(Montium)       kursel_gametic_2021  \n",
       "22                cenH3.4_(Montium)     kursel_recurrent_2017  \n",
       "23                cenH3.4_(Montium)  teixeira_concurrent_2018  \n",
       "24             cenH3.5_(Drosophila)       kursel_gametic_2021  \n",
       "25             cenH3.5_(Drosophila)     kursel_recurrent_2017  \n",
       "26             cenH3.5_(Drosophila)  teixeira_concurrent_2018  \n",
       "27                cenH3.6_(Repleta)       kursel_gametic_2021  \n",
       "28                cenH3.6_(Repleta)     kursel_recurrent_2017  \n",
       "29                cenH3.6_(Repleta)  teixeira_concurrent_2018  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_has_publication hp \"\n",
    "    \"ON h.id = hp.histone_id \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(variants)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2623f923-6c9c-492f-8947-4ad6700bf97a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "publication_id\n",
       "kursel_gametic_2021         6\n",
       "kursel_recurrent_2017       6\n",
       "teixeira_concurrent_2018    6\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df[\"id\"].isin(variants)][\"publication_id\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "eb4168f8-81ec-4df6-b480-6a948afabffb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa333cb4-70c4-4007-aa4d-82f0ee94e77c",
   "metadata": {},
   "source": [
    "### Add Culicidae histones publications\n",
    "\n",
    "**New *Culicidae* nodes:** cenH3.1_(Culicidae), cenH3.2_(Culicidae), cenH3.3_(Aedes)\n",
    "\n",
    "**Publications:** kursel_ancient_2020"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "12173bd8-5fa7-419d-be10-dda19e8d3996",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>kursel_ancient_2020</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      id title   doi author  year pubmed_id\n",
       "117  kursel_ancient_2020  None  None   None  None      None"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"kursel_ancient_2020\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2a72dff4-925f-443e-a0e6-d9a1643f7013",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "variants = [\"cenH3.1_(Culicidae)\", \"cenH3.2_(Culicidae)\", \"cenH3.3_(Aedes)\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d6fb902a-8c27-4a4a-a262-5036349478c3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>histone_id</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>cenH3.1_(Culicidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Culicidae</td>\n",
       "      <td>7157</td>\n",
       "      <td>246.0</td>\n",
       "      <td>cenH3_(Culicidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>cenH3.2_(Culicidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Culicidae</td>\n",
       "      <td>7157</td>\n",
       "      <td>247.0</td>\n",
       "      <td>cenH3_(Culicidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>cenH3.3_(Aedes)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Aedes</td>\n",
       "      <td>7158</td>\n",
       "      <td>248.0</td>\n",
       "      <td>cenH3_(Culicidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     id    level taxonomic_span taxonomic_span_id  \\\n",
       "5   cenH3.1_(Culicidae)  variant      Culicidae              7157   \n",
       "11  cenH3.2_(Culicidae)  variant      Culicidae              7157   \n",
       "17      cenH3.3_(Aedes)  variant          Aedes              7158   \n",
       "\n",
       "    description             parent histone_id publication_id  \n",
       "5         246.0  cenH3_(Culicidae)       None           None  \n",
       "11        247.0  cenH3_(Culicidae)       None           None  \n",
       "17        248.0  cenH3_(Culicidae)       None           None  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_has_publication hp \"\n",
    "    \"ON h.id = hp.histone_id \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(variants)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "14907898-aee3-487d-a656-f29d6b78078b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for v in variants:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_histone_has_publication, (v, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3d1847ec-3c79-47f7-952c-0ec034d18f0b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>histone_id</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>cenH3.1_(Culicidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Culicidae</td>\n",
       "      <td>7157</td>\n",
       "      <td>246.0</td>\n",
       "      <td>cenH3_(Culicidae)</td>\n",
       "      <td>cenH3.1_(Culicidae)</td>\n",
       "      <td>kursel_ancient_2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>cenH3.2_(Culicidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Culicidae</td>\n",
       "      <td>7157</td>\n",
       "      <td>247.0</td>\n",
       "      <td>cenH3_(Culicidae)</td>\n",
       "      <td>cenH3.2_(Culicidae)</td>\n",
       "      <td>kursel_ancient_2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>cenH3.3_(Aedes)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Aedes</td>\n",
       "      <td>7158</td>\n",
       "      <td>248.0</td>\n",
       "      <td>cenH3_(Culicidae)</td>\n",
       "      <td>cenH3.3_(Aedes)</td>\n",
       "      <td>kursel_ancient_2020</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     id    level taxonomic_span taxonomic_span_id  \\\n",
       "5   cenH3.1_(Culicidae)  variant      Culicidae              7157   \n",
       "11  cenH3.2_(Culicidae)  variant      Culicidae              7157   \n",
       "17      cenH3.3_(Aedes)  variant          Aedes              7158   \n",
       "\n",
       "    description             parent           histone_id       publication_id  \n",
       "5         246.0  cenH3_(Culicidae)  cenH3.1_(Culicidae)  kursel_ancient_2020  \n",
       "11        247.0  cenH3_(Culicidae)  cenH3.2_(Culicidae)  kursel_ancient_2020  \n",
       "17        248.0  cenH3_(Culicidae)      cenH3.3_(Aedes)  kursel_ancient_2020  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_has_publication hp \"\n",
    "    \"ON h.id = hp.histone_id \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(variants)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "3ad5bd6e-4abb-4b19-ae6a-56362a76d62d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "publication_id\n",
       "kursel_ancient_2020    3\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df[\"id\"].isin(variants)][\"publication_id\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "6d7adf1f-8687-4f0f-bb46-a417b7434225",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e8bad07-ef74-40fd-a93b-e7897df0c591",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Close connections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e5673501-5845-491a-bbb9-f5d213721d58",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cursor.close()\n",
    "conn.close()\n",
    "tunnel.stop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f7b2e3c-e0b8-4016-80eb-a0dd7afbf1a2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".conda-histdb_env",
   "language": "python",
   "name": "conda-env-.conda-histdb_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
