{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "53d3ec55-7daf-4ade-8add-2830b709dc54",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import io\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import Entrez, SeqIO\n",
    "from mysql.connector import connection\n",
    "from sshtunnel import SSHTunnelForwarder\n",
    "\n",
    "# Укажите ваш email (обязательно для использования Entrez)\n",
    "Entrez.email = \"your.email@example.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "608c4bfe-b979-4b60-b2b5-b94eeb679cb5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open(\"db_curated_server_info.txt\", \"r\") as file:\n",
    "    lines = file.readlines()\n",
    "\n",
    "config = {}\n",
    "\n",
    "for line in lines:\n",
    "    line = line.strip()\n",
    "    if line and not line.startswith(\"#\"):\n",
    "        key, value = line.split(\"=\", 1)\n",
    "        config[key] = value.strip()\n",
    "\n",
    "server_name = config.get(\"server_name\")\n",
    "srever_port = int(config.get(\"srever_port\"))\n",
    "ssh_password = config.get(\"ssh_password\")\n",
    "ssh_username = config.get(\"ssh_username\")\n",
    "db_adress = config.get(\"db_adress\")\n",
    "db_port = int(config.get(\"db_port\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4e19aa1c-49a3-48c4-b67f-34c94cef608d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46107\n"
     ]
    }
   ],
   "source": [
    "tunnel = SSHTunnelForwarder(\n",
    "    (server_name, srever_port),\n",
    "    ssh_password=ssh_password,\n",
    "    ssh_username=ssh_username,\n",
    "    remote_bind_address=(db_adress, db_port),\n",
    ")\n",
    "tunnel.start()\n",
    "print(tunnel.local_bind_port)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cef2275b-12d0-48f1-ad3c-ce4cdb9aa267",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "conn = connection.MySQLConnection(\n",
    "    user=\"db_user\",\n",
    "    password=\"db_password\",\n",
    "    host=\"localhost\",\n",
    "    port=tunnel.local_bind_port,\n",
    "    database=\"db_name\",\n",
    ")\n",
    "cursor = conn.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f589ea8d-8b83-4331-a1cf-b64ed7621801",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('alternative_name',),\n",
       " ('histone',),\n",
       " ('histone_description',),\n",
       " ('histone_has_publication',),\n",
       " ('publication',),\n",
       " ('sequence',),\n",
       " ('sequence_has_publication',)]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SHOW TABLES;\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b3dd9ef8-b162-4127-9535-ac7743af8c69",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "add_histone = (\n",
    "    \"INSERT INTO histone \"\n",
    "    \"(id, level, taxonomic_span, taxonomic_span_id, description, parent) \"\n",
    "    \"VALUES (%(id)s, %(level)s, %(taxonomic_span)s, %(taxonomic_span_id)s, %(description)s, %(parent)s)\"\n",
    ")\n",
    "add_histone_description = (\n",
    "    \"INSERT INTO histone_description \"\n",
    "    \"(summary, taxonomy, genes, evolution, expression, knock_out, function, sequence, localization, deposition, structure, interactions, disease, caveats) \"\n",
    "    \"VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)\"\n",
    ")\n",
    "# add_publication = (\n",
    "#     \"INSERT INTO publication \"\n",
    "#     \"(id, title, doi, author, year) \"\n",
    "#     \"VALUES (%(id)s, %(title)s, %(doi)s, %(author)s, %(year)s)\"\n",
    "# )\n",
    "# add_sequence = (\n",
    "#     \"INSERT INTO sequence \"\n",
    "#     \"(accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration) \"\n",
    "#     \"VALUES (%(accession)s, %(variant)s, %(gi)s, %(ncbi_gene_id)s, %(hgnc_gene_name)s, %(taxonomy_id)s, %(organism)s, %(phylum)s, %(class)s, %(taxonomy_group)s, %(info)s, %(sequence)s, %(variant_under_consideration)s)\"\n",
    "# )\n",
    "# add_sequence_has_publication = (\n",
    "#     \"INSERT INTO sequence_has_publication \"\n",
    "#     \"(sequence_accession, publication_id) \"\n",
    "#     \"VALUES (%s, %s)\"\n",
    "# )\n",
    "add_alternate_names = (\n",
    "    \"INSERT INTO alternative_name \"\n",
    "    \"(name, taxonomy, gene, splice, histone) \"\n",
    "    \"VALUES (%(name)s, %(taxonomy)s, %(gene)s, %(splice)s, %(histone)s)\"\n",
    ")\n",
    "# add_histone_has_publication = (\n",
    "#     \"INSERT INTO histone_has_publication \"\n",
    "#     \"(histone_id, publication_id) \"\n",
    "#     \"VALUES (%s, %s)\"\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6011bb8a-bab5-4fb5-b4ae-15b83accd15b",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_histone_description = {\n",
    "    \"summary\": None,\n",
    "    \"taxonomy\": None,\n",
    "    \"genes\": None,\n",
    "    \"evolution\": None,\n",
    "    \"expression\": None,\n",
    "    \"knock_out\": None,\n",
    "    \"function\": None,\n",
    "    \"sequence\": None,\n",
    "    \"localization\": None,\n",
    "    \"deposition\": None,\n",
    "    \"structure\": None,\n",
    "    \"interactions\": None,\n",
    "    \"disease\": None,\n",
    "    \"caveats\": None,\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d0d3d7e-47ed-4f2b-8e96-e61eb030872d",
   "metadata": {},
   "source": [
    "# To Do H4\n",
    "\n",
    "## <span style=\"color:green\">Add a new node H4.G_(Hominidae) to H4</span>\n",
    "\n",
    "## <span style=\"color:green\">Add description to H4.G_(Hominidae)</span>\n",
    "\n",
    "### <span style=\"color:green\">Add summary</span>\n",
    "\n",
    "```H4.G_(Hominidae) is a hominid-specific histone H4 variant, also known as H4G or H4.7, which promotes breast cancer proliferation by destabilizing nucleolar chromatin [pang_histone_2020, talbert_histone_2021, lai_roles_2024]. H4.G is encode by clustered gene H4C7. Since the expression of this gene has been described only in tumor cells and its product, H4.G, is incapable of forming stable nucleosomes in vitro, there is no compelling evidence that H4.G is a bona fide histone variant in normal cells [seal_standardized_2022].```\n",
    "\n",
    "### <span style=\"color:green\">Add gene</span>\n",
    "\n",
    "```H4.G is encode by clustered gene H4C7.```\n",
    "\n",
    "### <span style=\"color:green\">Add sequence</span>\n",
    "\n",
    "```H4.G shares 85% identity with canonical histone H4, differing by 15 amino acid residues located predominantly near the N-terminal α1, α2, and α3 regions, and possessing a C-terminus shortened by five residues [long_novel_2019].```\n",
    "\n",
    "### <span style=\"color:green\">Add expression</span>\n",
    "\n",
    "```Although H4.G expression is observed at low levels in various human cell lines, its significant upregulation is found primarily in breast cancer, as well as in T-cell prolymphocytic leukemia, HuR-silenced thyroid carcinoma, and endometrioid carcinoma [pang_histone_2020, lai_roles_2024].```\n",
    "\n",
    "### <span style=\"color:green\">Add localization</span>\n",
    "\n",
    "```It is predominantly localized to the nucleolus which contains the nucleolar chromatin [lai_roles_2024].```\n",
    "\n",
    "### <span style=\"color:green\">Add function</span>\n",
    "\n",
    "```The functional role of H4.G in cells is likely to loosen the structure of nucleolar chromatin by forming destabilized nucleosome-like structures, thereby enhancing ribosomal RNA (rRNA) transcription and ribosome biogenesis, which promotes tumor cell growth [long_novel_2019, lai_roles_2024].```\n",
    "\n",
    "### <span style=\"color:green\">Add deposition</span>\n",
    "\n",
    "```H4.G is not stably incorporated into nucleolar chromatin even with the assistance of the chaperone NPM1; instead, the H4.G-NPM1 complex forms unstable nucleosome-like structures that rapidly dissociate, and this very property appears to facilitate chromatin loosening [pang_histone_2020].```\n",
    "\n",
    "### <span style=\"color:green\">Add knock_out</span>\n",
    "\n",
    "```H4.G knockdown in MCF7 breast cancer cells significantly impairs cell growth, proliferation, rRNA synthesis, and protein synthesis [long_novel_2019, pang_histone_2020].```\n",
    "\n",
    "### <span style=\"color:green\">Add structure</span>\n",
    "\n",
    "```H4.G is unable to form stable nucleosomes in vitro, even in the presence of the chaperone NPM1; instead, it likely forms transient nucleosome-like structures that rapidly dissociate [pang_histone_2020].```\n",
    "\n",
    "### <span style=\"color:green\">Add disease</span>\n",
    "\n",
    "```H4.G is overexpressed in a number of cancers, and its expression level positively correlates with cancer stage progression, as demonstrated in breast cancer, where it promotes tumor growth by facilitating rRNA and protein synthesis [lai_roles_2024].```\n",
    "\n",
    "## <span style=\"color:green\">Update summary of H4</span>\n",
    "\n",
    "### <span style=\"color:green\">Update summary</span>\n",
    "\n",
    "```H4 is one of the core histones. Two H4s and two H3s form H3-H4 tetramer via \"hand shake\" and \"four helix bundle\" motifs. Two H2A-H2B dimers then associate with H3-H4 tetramer to form complete nucleosome core. H4 is the most conservative histone type and has very few known variants. Structure of H4 has a histone fold domain and a flexible N-terminal tail. H4 provides sites for H2B interaction via \"four-helix bundle\" and forms a small β-sheet with H2Aa. Acetylation of the H4 tail within the basic patch region (K5, K8, K12, K16, K20) modulates its conformational ensemble through electrostatic effects, alterations in contact patterns, and subtle changes in the protonation of key residues, thereby contributing to the regulation of chromatin architecture [dewing_acetylation-dependent_2024].```\n",
    "\n",
    "## <span style=\"color:green\">Add description to cH4_(Vertebrata)</span>\n",
    "\n",
    "### <span style=\"color:green\">Add summary</span>\n",
    "\n",
    "```cH4_(Vertebrata) clustered H4 histones in vertebrates, often called canonical, replication-dependent, replication-coupled or \"bulk\" H4 histones. This is a subclass of cH4_(Animals) (see it for a detailed description). Asymmetric dimethylation of histone H4 at arginine 3 (H4R3me2a), catalyzed by PRMT1 in vertebrates, constitutes a key epigenetic mark for the establishment and maintenance of open and transcriptionally active chromatin at regulatory regions of the vertebrate genome, particularly at enhancers and super-enhancers [sudhakar_histone_2024].```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fff36921-fd76-441e-bf69-ca6711090761",
   "metadata": {},
   "source": [
    "# DONE\n",
    "\n",
    "## <span style=\"color:black\">Add a new node H4.G_(Hominidae) to H4</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ebedbabb-f773-475e-8415-1bc88a0ff115",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, level, taxonomic_span, taxonomic_span_id, description, parent, id, summary, taxonomy, genes, evolution, expression, knock_out, function, sequence, localization, deposition, structure, interactions, disease, caveats, relations]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 22 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='H4.G_(Hominidae)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cbf17b5a-517d-4dcc-a4ea-19e3bad4e5a3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"H4.G_(Hominidae)\",\n",
    "    \"level\": \"variant_group\",\n",
    "    \"taxonomic_span\": \"Hominidae\",\n",
    "    \"taxonomic_span_id\": \"9604\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"H4\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "66eae759-5984-4993-93f0-dc6f0060b948",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H4.G_(Hominidae)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Hominidae</td>\n",
       "      <td>9604</td>\n",
       "      <td>None</td>\n",
       "      <td>H4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 id          level taxonomic_span taxonomic_span_id  \\\n",
       "0  H4.G_(Hominidae)  variant_group      Hominidae              9604   \n",
       "\n",
       "  description parent    id summary taxonomy genes  ... knock_out function  \\\n",
       "0        None     H4  None    None     None  None  ...      None     None   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     None         None       None      None         None    None    None   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='H4.G_(Hominidae)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8ff1b4d4-3884-46be-9cc5-77680aa10881",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c41419fb-2e9b-4c01-8e8e-d47dc9471bf3",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add description to H4.G_(Hominidae)</span>\n",
    "\n",
    "### <span style=\"color:black\">Add summary</span>\n",
    "\n",
    "```H4.G_(Hominidae) is a hominid-specific histone H4 variant, also known as H4G or H4.7, which promotes breast cancer proliferation by destabilizing nucleolar chromatin [pang_histone_2020, talbert_histone_2021, lai_roles_2024]. H4.G is encode by clustered gene H4C7. Since the expression of this gene has been described only in tumor cells and its product, H4.G, is incapable of forming stable nucleosomes in vitro, there is no compelling evidence that H4.G is a bona fide histone variant in normal cells [seal_standardized_2022].```\n",
    "\n",
    "### <span style=\"color:black\">Add gene</span>\n",
    "\n",
    "```H4.G is encode by clustered gene H4C7.```\n",
    "\n",
    "### <span style=\"color:black\">Add sequence</span>\n",
    "\n",
    "```H4.G shares 85% identity with canonical histone H4, differing by 15 amino acid residues located predominantly near the N-terminal α1, α2, and α3 regions, and possessing a C-terminus shortened by five residues [long_novel_2019].```\n",
    "\n",
    "### <span style=\"color:black\">Add expression</span>\n",
    "\n",
    "```Although H4.G expression is observed at low levels in various human cell lines, its significant upregulation is found primarily in breast cancer, as well as in T-cell prolymphocytic leukemia, HuR-silenced thyroid carcinoma, and endometrioid carcinoma [pang_histone_2020, lai_roles_2024].```\n",
    "\n",
    "### <span style=\"color:black\">Add localization</span>\n",
    "\n",
    "```It is predominantly localized to the nucleolus which contains the nucleolar chromatin [lai_roles_2024].```\n",
    "\n",
    "### <span style=\"color:black\">Add function</span>\n",
    "\n",
    "```The functional role of H4.G in cells is likely to loosen the structure of nucleolar chromatin by forming destabilized nucleosome-like structures, thereby enhancing ribosomal RNA (rRNA) transcription and ribosome biogenesis, which promotes tumor cell growth [long_novel_2019, lai_roles_2024].```\n",
    "\n",
    "### <span style=\"color:black\">Add deposition</span>\n",
    "\n",
    "```H4.G is not stably incorporated into nucleolar chromatin even with the assistance of the chaperone NPM1; instead, the H4.G-NPM1 complex forms unstable nucleosome-like structures that rapidly dissociate, and this very property appears to facilitate chromatin loosening [pang_histone_2020].```\n",
    "\n",
    "### <span style=\"color:black\">Add knock_out</span>\n",
    "\n",
    "```H4.G knockdown in MCF7 breast cancer cells significantly impairs cell growth, proliferation, rRNA synthesis, and protein synthesis [long_novel_2019, pang_histone_2020].```\n",
    "\n",
    "### <span style=\"color:black\">Add structure</span>\n",
    "\n",
    "```H4.G is unable to form stable nucleosomes in vitro, even in the presence of the chaperone NPM1; instead, it likely forms transient nucleosome-like structures that rapidly dissociate [pang_histone_2020].```\n",
    "\n",
    "### <span style=\"color:black\">Add disease</span>\n",
    "\n",
    "```H4.G is overexpressed in a number of cancers, and its expression level positively correlates with cancer stage progression, as demonstrated in breast cancer, where it promotes tumor growth by facilitating rRNA and protein synthesis [lai_roles_2024].```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3d8868af-3b15-4f72-8d8d-67ca10baf007",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H4.G_(Hominidae)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Hominidae</td>\n",
       "      <td>9604</td>\n",
       "      <td>None</td>\n",
       "      <td>H4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 id          level taxonomic_span taxonomic_span_id  \\\n",
       "0  H4.G_(Hominidae)  variant_group      Hominidae              9604   \n",
       "\n",
       "  description parent    id summary taxonomy genes  ... knock_out function  \\\n",
       "0        None     H4  None    None     None  None  ...      None     None   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     None         None       None      None         None    None    None   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='H4.G_(Hominidae)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "40dd4997-97d7-44c0-a79c-76af9d1f9533",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_histone_description = {\n",
    "    \"summary\": \"H4.G_(Hominidae) is a hominid-specific histone H4 variant, also known as H4G or H4.7, which promotes breast cancer proliferation by destabilizing nucleolar chromatin [pang_histone_2020, talbert_histone_2021, lai_roles_2024]. H4.G is encode by clustered gene H4C7. Since the expression of this gene has been described only in tumor cells and its product, H4.G, is incapable of forming stable nucleosomes in vitro, there is no compelling evidence that H4.G is a bona fide histone variant in normal cells [seal_standardized_2022].\",\n",
    "    \"taxonomy\": None,\n",
    "    \"genes\": \"H4.G is encode by clustered gene H4C7.\",\n",
    "    \"evolution\": None,\n",
    "    \"expression\": \"Although H4.G expression is observed at low levels in various human cell lines, its significant upregulation is found primarily in breast cancer, as well as in T-cell prolymphocytic leukemia, HuR-silenced thyroid carcinoma, and endometrioid carcinoma [pang_histone_2020, lai_roles_2024].\",\n",
    "    \"knock_out\": \"H4.G knockdown in MCF7 breast cancer cells significantly impairs cell growth, proliferation, rRNA synthesis, and protein synthesis [long_novel_2019, pang_histone_2020].\",\n",
    "    \"function\": \"The functional role of H4.G in cells is likely to loosen the structure of nucleolar chromatin by forming destabilized nucleosome-like structures, thereby enhancing ribosomal RNA (rRNA) transcription and ribosome biogenesis, which promotes tumor cell growth [long_novel_2019, lai_roles_2024].\",\n",
    "    \"sequence\": \"H4.G shares 85% identity with canonical histone H4, differing by 15 amino acid residues located predominantly near the N-terminal α1, α2, and α3 regions, and possessing a C-terminus shortened by five residues [long_novel_2019].\",\n",
    "    \"localization\": \"It is predominantly localized to the nucleolus which contains the nucleolar chromatin [lai_roles_2024].\",\n",
    "    \"deposition\": \"H4.G is not stably incorporated into nucleolar chromatin even with the assistance of the chaperone NPM1; instead, the H4.G-NPM1 complex forms unstable nucleosome-like structures that rapidly dissociate, and this very property appears to facilitate chromatin loosening [pang_histone_2020].\",\n",
    "    \"structure\": \"H4.G is unable to form stable nucleosomes in vitro, even in the presence of the chaperone NPM1; instead, it likely forms transient nucleosome-like structures that rapidly dissociate [pang_histone_2020].\",\n",
    "    \"interactions\": None,\n",
    "    \"disease\": \"H4.G is overexpressed in a number of cancers, and its expression level positively correlates with cancer stage progression, as demonstrated in breast cancer, where it promotes tumor growth by facilitating rRNA and protein synthesis [lai_roles_2024].\",\n",
    "    \"caveats\": None,\n",
    "}\n",
    "cursor.execute(add_histone_description, tuple(data_histone_description.values()))\n",
    "\n",
    "histone_description_id = cursor.lastrowid\n",
    "query = f\"UPDATE histone SET description={histone_description_id} WHERE id='H4.G_(Hominidae)'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f0b3b1be-a49e-4256-80d1-fc7230ddee75",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H4.G_(Hominidae)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Hominidae</td>\n",
       "      <td>9604</td>\n",
       "      <td>258</td>\n",
       "      <td>H4</td>\n",
       "      <td>258</td>\n",
       "      <td>H4.G_(Hominidae) is a hominid-specific histone...</td>\n",
       "      <td>None</td>\n",
       "      <td>H4.G is encode by clustered gene H4C7.</td>\n",
       "      <td>...</td>\n",
       "      <td>H4.G knockdown in MCF7 breast cancer cells sig...</td>\n",
       "      <td>The functional role of H4.G in cells is likely...</td>\n",
       "      <td>H4.G shares 85% identity with canonical histon...</td>\n",
       "      <td>It is predominantly localized to the nucleolus...</td>\n",
       "      <td>H4.G is not stably incorporated into nucleolar...</td>\n",
       "      <td>H4.G is unable to form stable nucleosomes in v...</td>\n",
       "      <td>None</td>\n",
       "      <td>H4.G is overexpressed in a number of cancers, ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 id          level taxonomic_span taxonomic_span_id  \\\n",
       "0  H4.G_(Hominidae)  variant_group      Hominidae              9604   \n",
       "\n",
       "   description parent   id                                            summary  \\\n",
       "0          258     H4  258  H4.G_(Hominidae) is a hominid-specific histone...   \n",
       "\n",
       "  taxonomy                                   genes  ...  \\\n",
       "0     None  H4.G is encode by clustered gene H4C7.  ...   \n",
       "\n",
       "                                           knock_out  \\\n",
       "0  H4.G knockdown in MCF7 breast cancer cells sig...   \n",
       "\n",
       "                                            function  \\\n",
       "0  The functional role of H4.G in cells is likely...   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  H4.G shares 85% identity with canonical histon...   \n",
       "\n",
       "                                        localization  \\\n",
       "0  It is predominantly localized to the nucleolus...   \n",
       "\n",
       "                                          deposition  \\\n",
       "0  H4.G is not stably incorporated into nucleolar...   \n",
       "\n",
       "                                           structure interactions  \\\n",
       "0  H4.G is unable to form stable nucleosomes in v...         None   \n",
       "\n",
       "                                             disease caveats relations  \n",
       "0  H4.G is overexpressed in a number of cancers, ...    None      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='H4.G_(Hominidae)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "817ebbbf-7cc6-41e0-8545-8056fe31691e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f6f98e7-5ea6-47d3-b00e-cfa6075c6e06",
   "metadata": {
    "tags": []
   },
   "source": [
    "## <span style=\"color:black\">Add a alternate name H4.7 to H4.G_(Hominidae)</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e5519ef2-e6fc-4777-adae-c3b7f22f60cd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, name, taxonomy, gene, splice, histone]\n",
       "Index: []"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name \" \"WHERE histone='H4.G_(Hominidae)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "dd3a6da8-e2cf-413e-836f-27d79dcc04a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_alternate_name = {\n",
    "    \"name\": \"H4.7\",\n",
    "    \"taxonomy\": None,\n",
    "    \"gene\": None,\n",
    "    \"splice\": None,\n",
    "    \"histone\": \"H4.G_(Hominidae)\",\n",
    "}\n",
    "cursor.execute(add_alternate_names, data_alternate_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6c2d8456-157e-44c7-b3e8-c615bc423d0a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>122</td>\n",
       "      <td>H4.7</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H4.G_(Hominidae)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    id  name taxonomy  gene splice           histone\n",
       "0  122  H4.7     None  None   None  H4.G_(Hominidae)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name \" \"WHERE histone='H4.G_(Hominidae)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "7574f83d-316c-47e0-a8d8-03ef2f9ef5f1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22560d72-d4d7-4d89-a5a5-2e8f570c4a8a",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Update summary of H4</span>\n",
    "\n",
    "### <span style=\"color:black\">Update summary</span>\n",
    "\n",
    "```H4 is one of the core histones. Two H4s and two H3s form H3-H4 tetramer via \"hand shake\" and \"four helix bundle\" motifs. Two H2A-H2B dimers then associate with H3-H4 tetramer to form complete nucleosome core. H4 is the most conservative histone type and has very few known variants. Structure of H4 has a histone fold domain and a flexible N-terminal tail. H4 provides sites for H2B interaction via \"four-helix bundle\" and forms a small β-sheet with H2Aa. Acetylation of the H4 tail within the basic patch region (K5, K8, K12, K16, K20) modulates its conformational ensemble through electrostatic effects, alterations in contact patterns, and subtle changes in the protonation of key residues, thereby contributing to the regulation of chromatin architecture [dewing_acetylation-dependent_2024].```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d6ef62e8-f242-4311-a8f8-a883071beb6d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H4</td>\n",
       "      <td>type</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>6</td>\n",
       "      <td>None</td>\n",
       "      <td>6</td>\n",
       "      <td>H4 is one of the core histones. Two H4s and tw...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   id level taxonomic_span taxonomic_span_id  description parent  id  \\\n",
       "0  H4  type           null              null            6   None   6   \n",
       "\n",
       "                                             summary taxonomy genes  ...  \\\n",
       "0  H4 is one of the core histones. Two H4s and tw...     null  null  ...   \n",
       "\n",
       "  knock_out function sequence localization deposition structure interactions  \\\n",
       "0      null     null     null         null       null      null         null   \n",
       "\n",
       "  disease caveats relations  \n",
       "0    null    null      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='H4'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "4065a577-ece0-45f7-b6e2-a453fe1c1cce",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UPDATE histone_description SET summary='H4 is one of the core histones. Two H4s and two H3s form H3-H4 tetramer via \"hand shake\" and \"four helix bundle\" motifs. Two H2A-H2B dimers then associate with H3-H4 tetramer to form complete nucleosome core. H4 is the most conservative histone type and has very few known variants. Structure of H4 has a histone fold domain and a flexible N-terminal tail. H4 provides sites for H2B interaction via \"four-helix bundle\" and forms a small β-sheet with H2Aa. Acetylation of the H4 tail within the basic patch region (K5, K8, K12, K16, K20) modulates its conformational ensemble through electrostatic effects, alterations in contact patterns, and subtle changes in the protonation of key residues, thereby contributing to the regulation of chromatin architecture [dewing_acetylation-dependent_2024].' WHERE id=6\n"
     ]
    }
   ],
   "source": [
    "summary_desc = 'H4 is one of the core histones. Two H4s and two H3s form H3-H4 tetramer via \"hand shake\" and \"four helix bundle\" motifs. Two H2A-H2B dimers then associate with H3-H4 tetramer to form complete nucleosome core. H4 is the most conservative histone type and has very few known variants. Structure of H4 has a histone fold domain and a flexible N-terminal tail. H4 provides sites for H2B interaction via \"four-helix bundle\" and forms a small β-sheet with H2Aa. Acetylation of the H4 tail within the basic patch region (K5, K8, K12, K16, K20) modulates its conformational ensemble through electrostatic effects, alterations in contact patterns, and subtle changes in the protonation of key residues, thereby contributing to the regulation of chromatin architecture [dewing_acetylation-dependent_2024].'\n",
    "query = f\"UPDATE histone_description SET summary='{summary_desc}' WHERE id=6\"\n",
    "print(query)\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c5e15687-cc19-453f-91a9-1c4cb0524812",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'H4 is one of the core histones. Two H4s and two H3s form H3-H4 tetramer via \"hand shake\" and \"four helix bundle\" motifs. Two H2A-H2B dimers then associate with H3-H4 tetramer to form complete nucleosome core. H4 is the most conservative histone type and has very few known variants. Structure of H4 has a histone fold domain and a flexible N-terminal tail. H4 provides sites for H2B interaction via \"four-helix bundle\" and forms a small β-sheet with H2Aa. Acetylation of the H4 tail within the basic patch region (K5, K8, K12, K16, K20) modulates its conformational ensemble through electrostatic effects, alterations in contact patterns, and subtle changes in the protonation of key residues, thereby contributing to the regulation of chromatin architecture [dewing_acetylation-dependent_2024].'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='H4'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[\"summary\"].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "41c6dc12-9932-4821-a947-a1679400f1a7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ef5274e-ddf6-4601-9e68-05f4fba5d56f",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add description to cH4_(Vertebrata)</span>\n",
    "\n",
    "### <span style=\"color:black\">Add summary</span>\n",
    "\n",
    "```cH4_(Vertebrata) clustered H4 histones in vertebrates, often called canonical, replication-dependent, replication-coupled or \"bulk\" H4 histones. This is a subclass of cH4_(Animals) (see it for a detailed description). Asymmetric dimethylation of histone H4 at arginine 3 (H4R3me2a), catalyzed by PRMT1 in vertebrates, constitutes a key epigenetic mark for the establishment and maintenance of open and transcriptionally active chromatin at regulatory regions of the vertebrate genome, particularly at enhancers and super-enhancers [sudhakar_histone_2024].```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "d85dfef5-4607-4e37-ba34-fc4a5b55c03e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH4_(Vertebrata)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Vertebrata</td>\n",
       "      <td>7742</td>\n",
       "      <td>None</td>\n",
       "      <td>cH4_(Animals)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  cH4_(Vertebrata)  variant     Vertebrata              7742        None   \n",
       "\n",
       "          parent    id summary taxonomy genes  ... knock_out function  \\\n",
       "0  cH4_(Animals)  None    None     None  None  ...      None     None   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     None         None       None      None         None    None    None   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cH4_(Vertebrata)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "1841193c-b25e-4431-814a-5cc3e4979380",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone_description = {\n",
    "    \"summary\": 'cH4_(Vertebrata) clustered H4 histones in vertebrates, often called canonical, replication-dependent, replication-coupled or \"bulk\" H4 histones. This is a subclass of cH4_(Animals) (see it for a detailed description). Asymmetric dimethylation of histone H4 at arginine 3 (H4R3me2a), catalyzed by PRMT1 in vertebrates, constitutes a key epigenetic mark for the establishment and maintenance of open and transcriptionally active chromatin at regulatory regions of the vertebrate genome, particularly at enhancers and super-enhancers [sudhakar_histone_2024].',\n",
    "    \"taxonomy\": None,\n",
    "    \"genes\": None,\n",
    "    \"evolution\": None,\n",
    "    \"expression\": None,\n",
    "    \"knock_out\": None,\n",
    "    \"function\": None,\n",
    "    \"sequence\": None,\n",
    "    \"localization\": None,\n",
    "    \"deposition\": None,\n",
    "    \"structure\": None,\n",
    "    \"interactions\": None,\n",
    "    \"disease\": None,\n",
    "    \"caveats\": None,\n",
    "}\n",
    "cursor.execute(add_histone_description, tuple(data_histone_description.values()))\n",
    "\n",
    "histone_description_id = cursor.lastrowid\n",
    "query = f\"UPDATE histone SET description={histone_description_id} WHERE id='cH4_(Vertebrata)'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "af400355-368a-4619-bf64-100a027e660f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'cH4_(Vertebrata) clustered H4 histones in vertebrates, often called canonical, replication-dependent, replication-coupled or \"bulk\" H4 histones. This is a subclass of cH4_(Animals) (see it for a detailed description). Asymmetric dimethylation of histone H4 at arginine 3 (H4R3me2a), catalyzed by PRMT1 in vertebrates, constitutes a key epigenetic mark for the establishment and maintenance of open and transcriptionally active chromatin at regulatory regions of the vertebrate genome, particularly at enhancers and super-enhancers [sudhakar_histone_2024].'"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cH4_(Vertebrata)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[\"summary\"].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "799377a7-0c80-4acf-aa50-f6be124857ab",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e8bad07-ef74-40fd-a93b-e7897df0c591",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Close connections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e5673501-5845-491a-bbb9-f5d213721d58",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cursor.close()\n",
    "conn.close()\n",
    "tunnel.stop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dce175b6-cf41-4087-ac24-b84582423870",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".conda-histdb_env",
   "language": "python",
   "name": "conda-env-.conda-histdb_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
