{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "53d3ec55-7daf-4ade-8add-2830b709dc54",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import io\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import Entrez, SeqIO\n",
    "from mysql.connector import connection\n",
    "from sshtunnel import SSHTunnelForwarder\n",
    "\n",
    "# Укажите ваш email (обязательно для использования Entrez)\n",
    "Entrez.email = \"your.email@example.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "608c4bfe-b979-4b60-b2b5-b94eeb679cb5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open(\"db_curated_server_info.txt\", \"r\") as file:\n",
    "    lines = file.readlines()\n",
    "\n",
    "config = {}\n",
    "\n",
    "for line in lines:\n",
    "    line = line.strip()\n",
    "    if line and not line.startswith(\"#\"):\n",
    "        key, value = line.split(\"=\", 1)\n",
    "        config[key] = value.strip()\n",
    "\n",
    "server_name = config.get(\"server_name\")\n",
    "srever_port = int(config.get(\"srever_port\"))\n",
    "ssh_password = config.get(\"ssh_password\")\n",
    "ssh_username = config.get(\"ssh_username\")\n",
    "db_adress = config.get(\"db_adress\")\n",
    "db_port = int(config.get(\"db_port\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4e19aa1c-49a3-48c4-b67f-34c94cef608d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44121\n"
     ]
    }
   ],
   "source": [
    "tunnel = SSHTunnelForwarder(\n",
    "    (server_name, srever_port),\n",
    "    ssh_password=ssh_password,\n",
    "    ssh_username=ssh_username,\n",
    "    remote_bind_address=(db_adress, db_port),\n",
    ")\n",
    "tunnel.start()\n",
    "print(tunnel.local_bind_port)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cef2275b-12d0-48f1-ad3c-ce4cdb9aa267",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "conn = connection.MySQLConnection(\n",
    "    user=\"db_user\",\n",
    "    password=\"db_password\",\n",
    "    host=\"localhost\",\n",
    "    port=tunnel.local_bind_port,\n",
    "    database=\"db_name\",\n",
    ")\n",
    "cursor = conn.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f589ea8d-8b83-4331-a1cf-b64ed7621801",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('alternative_name',),\n",
       " ('histone',),\n",
       " ('histone_description',),\n",
       " ('histone_has_publication',),\n",
       " ('publication',),\n",
       " ('sequence',),\n",
       " ('sequence_has_publication',)]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SHOW TABLES;\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b3dd9ef8-b162-4127-9535-ac7743af8c69",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "add_histone = (\n",
    "    \"INSERT INTO histone \"\n",
    "    \"(id, level, taxonomic_span, taxonomic_span_id, description, parent) \"\n",
    "    \"VALUES (%(id)s, %(level)s, %(taxonomic_span)s, %(taxonomic_span_id)s, %(description)s, %(parent)s)\"\n",
    ")\n",
    "add_histone_description = (\n",
    "    \"INSERT INTO histone_description \"\n",
    "    \"(summary, taxonomy, genes, evolution, expression, knock_out, function, sequence, localization, deposition, structure, interactions, disease, caveats) \"\n",
    "    \"VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)\"\n",
    ")\n",
    "add_publication = (\n",
    "    \"INSERT INTO publication \"\n",
    "    \"(id, title, doi, author, year) \"\n",
    "    \"VALUES (%(id)s, %(title)s, %(doi)s, %(author)s, %(year)s)\"\n",
    ")\n",
    "add_sequence = (\n",
    "    \"INSERT INTO sequence \"\n",
    "    \"(accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration) \"\n",
    "    \"VALUES (%(accession)s, %(variant)s, %(gi)s, %(ncbi_gene_id)s, %(hgnc_gene_name)s, %(taxonomy_id)s, %(organism)s, %(phylum)s, %(class)s, %(taxonomy_group)s, %(info)s, %(sequence)s, %(variant_under_consideration)s)\"\n",
    ")\n",
    "add_sequence_has_publication = (\n",
    "    \"INSERT INTO sequence_has_publication \"\n",
    "    \"(sequence_accession, publication_id) \"\n",
    "    \"VALUES (%s, %s)\"\n",
    ")\n",
    "add_alternate_names = (\n",
    "    \"INSERT INTO alternative_name \"\n",
    "    \"(name, taxonomy, gene, splice, histone) \"\n",
    "    \"VALUES (%(name)s, %(taxonomy)s, %(gene)s, %(splice)s, %(histone)s)\"\n",
    ")\n",
    "add_histone_has_publication = (\n",
    "    \"INSERT INTO histone_has_publication \"\n",
    "    \"(histone_id, publication_id) \"\n",
    "    \"VALUES (%s, %s)\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c1c3bb0c-180f-4fb2-9cdc-ca928f9808be",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_taxonomy_data(record):\n",
    "    import re\n",
    "    import sys\n",
    "\n",
    "    taxonomy_data = {}\n",
    "    taxonomy_data[\"organism\"] = record.annotations[\"organism\"]\n",
    "    try:\n",
    "        for a in record.features[0].qualifiers[\"db_xref\"]:\n",
    "            text = re.search(\"(\\S+):(\\S+)\", a).group(1)\n",
    "            taxid = re.search(\"(\\S+):(\\S+)\", a).group(2)\n",
    "            if text == \"taxon\":\n",
    "                print(f\"Fetched taxid from NCBI {taxid}\")\n",
    "                taxonomy_data[\"taxonomy_id\"] = int(taxid)\n",
    "            else:\n",
    "                continue\n",
    "    except:\n",
    "        print(\"!!!!!!Unable to get TAXID for this record setting it to 1\")\n",
    "        taxonomy_data[\"taxonomy_id\"] = 1  # unable to identify\n",
    "\n",
    "    lineage = dict()\n",
    "    for i in range(10):\n",
    "        try:\n",
    "            handle = Entrez.efetch(\n",
    "                id=taxonomy_data[\"taxonomy_id\"], db=\"taxonomy\", retmode=\"xml\"\n",
    "            )\n",
    "            tax_data = Entrez.read(handle)\n",
    "            lineage = {\n",
    "                d[\"Rank\"]: d[\"ScientificName\"]\n",
    "                for d in tax_data[0][\"LineageEx\"]\n",
    "                if d[\"Rank\"] in [\"class\", \"phylum\"]\n",
    "            }\n",
    "            break\n",
    "        except:\n",
    "            print(\n",
    "                \"Unexpected error: {}, Retrying, attempt {}\".format(\n",
    "                    sys.exc_info()[0], i\n",
    "                )\n",
    "            )\n",
    "            if i == 9:\n",
    "                print(\n",
    "                    f\"FATAL ERROR could not get class and phylum from NCBI after 10 attempts for taxid:{taxonomy_data['taxonomy_id']}. Will add None for class and phylum!\"\n",
    "                )\n",
    "            else:\n",
    "                continue\n",
    "    taxonomy_data[\"phylum\"] = lineage.get(\"phylum\", None)\n",
    "    taxonomy_data[\"class\"] = lineage.get(\"class\", None)\n",
    "    if taxonomy_data[\"phylum\"] is not None:\n",
    "        taxonomy_data[\"phylum\"] = str(taxonomy_data[\"phylum\"])\n",
    "    if taxonomy_data[\"class\"] is not None:\n",
    "        taxonomy_data[\"class\"] = str(taxonomy_data[\"class\"])\n",
    "    return taxonomy_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9a15b106-5cf1-40d4-8597-2b11dc30ff33",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>65</td>\n",
       "      <td>cid</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>93</td>\n",
       "      <td>cid</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Animals)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id name taxonomy  gene splice          histone\n",
       "0  65  cid     None  None   None            cenH3\n",
       "1  93  cid     None  None   None  cenH3_(Animals)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name \" \"WHERE name='cid'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c060c047-3e7d-4095-b608-9722031af692",
   "metadata": {},
   "source": [
    "# To Do\n",
    "\n",
    "## <span style=\"color:green\">Add a new node cenH3_(Insecta) to cenH3_(Animals)</span>\n",
    "\n",
    "## <span style=\"color:green\">Add a summary description to node cenH3_(Insecta)</span>\n",
    "\n",
    "```cenH3_(Insecta) is a centromere-specific histone variant in insects, functionally analogous to CENP-A (Centromere Protein A) in other eukaryotes. It is essential for centromere specification and proper chromosome segregation during cell division. Notably, cenH3 is absent in four holocentric insect clades, suggesting alternative centromere organization mechanisms in these lineages [talbert_histone_2021, senaratne_formation_2021, sridhar_kinetochore_2022]. Some insect families, such as Drosophilidae and Culicidae, possess multiple cenH3 paralogs with evidence of functional specialization. For more overview of these families, see the cenH3_(Drosophilidae) and cenH3_(Culicidae) classes.```\n",
    "\n",
    "## <span style=\"color:green\">Add a gene description to node cenH3_(Insecta)</span>\n",
    "\n",
    "```In some insect lineages, cenH3 has undergone gene duplication, leading to paralogs with distinct functional roles. For detailed lineage-specific information, see the cenH3_(Drosophilidae) and cenH3_(Culicidae) classes.```\n",
    "\n",
    "\n",
    "## <span style=\"color:green\">Add new nodes cenH3_(Drosophilidae), cenH3_(Culicidae) to cenH3_(Insecta)</span>\n",
    "\n",
    "## <span style=\"color:green\">Add a summary description to node cenH3_(Drosophilidae)</span>\n",
    "\n",
    "```cenH3_(Drosophilidae) is a centromere-specific histone variant in Drosophilidae family, functionally analogous to CENP-A in mammals.```\n",
    "\n",
    "## <span style=\"color:green\">Add a gene description to node cenH3_(Drosophilidae)</span>\n",
    "\n",
    "```In Drosophila, six cenH3 paralogs (Cid1–Cid6) have been identified: Cid1, Cid4, and Cid6 (and likely Cid2) are essential for mitotic chromosome segregation and are ubiquitously expressed in somatic cells, while the remaining paralogs are germline-specific and may play a role in suppressing centromere drive [kursel_gametic_2021, kursel_recurrent_2017].```\n",
    "\n",
    "## <span style=\"color:green\">Add a summary description to node cenH3_(Culicidae)</span>\n",
    "\n",
    "```cenH3_(Culicidae) is a centromere-specific histone variant in Culicidae family (mosquitos), functionally analogous to CENP-A in mammals```\n",
    "\n",
    "## <span style=\"color:green\">Add a gene description to node cenH3_(Culicidae)</span>\n",
    "\n",
    "```In mosquitoes (Culicidae), three cenH3 paralogs exist: the evolutionarily ancient mosqCid1 and mosqCid2, which have been conserved for over 150 million years, and mosqCid3, a more recent paralog resulting from an independent duplication of mosqCid1 in Aedes aegypti and Aedes albopictus [kursel_ancient_2020].```\n",
    "\n",
    "## <span style=\"color:green\">Add alternate names cid, cid1, cid2, cid3, cid4, cid5, cid6, mosqCid1, mosqCid2, mosqCid3 to cenH3_(Insecta)</span>\n",
    "\n",
    "## <span style=\"color:green\">Add alternate names cid, cid1, cid2, cid3, cid4, cid5, cid6 to cenH3_(Drosophilidae)</span>\n",
    "\n",
    "## <span style=\"color:green\">Add alternate names mosqCid1, mosqCid2, mosqCid3 to cenH3_(Culicidae)</span>\n",
    "\n",
    "## <span style=\"color:green\">Add new nodes cenH3.1_(Drosophilidae), cenH3.2_(Drosophila_eugracilis), cenH3.3_(Montium), cenH3.4_(Montium), cenH3.5_(Drosophila), cenH3.6_(Repleta) to cenH3_(Drosophilidae)</span>\n",
    "\n",
    "## <span style=\"color:green\">Add a summary description to node cenH3.1</span>\n",
    "\n",
    "```CenH3.1_(Drosophilidae) is a centromere-specific histone variant encoded by the Cid1 (also known as Cid) gene found in most Drosophila species except D. eugracilis, where it was pseudogenized and replaced by Cid2, and D. buzzatii and D. seriema, where Cid1 degenerated due to transposon insertions and was replaced by a new copy - Cid6 [teixeira_concurrent_2018, kursel_recurrent_2017]. As the ancestral centromeric histone in Drosophilidae, Cid1  represents the canonical centromeric histone in this family. It is essential for chromosome segregation during mitosis and meiosis, maintaining conserved histone-fold domains (HFD) critical for nucleosome assembly at centromeres. Unlike its paralogs (Cid3-Cid5), Cid1 is ubiquitously expressed in both somatic and germline tissues, though its expression is highest in gonads [kursel_recurrent_2017]. The N-terminal tail of Cid1 contains four conserved motifs. Remarkably, Cid1 persists throughout oogenesis in females but is specifically degraded during male meiosis I, disappearing from post-meiotic spermatids and sperm [kursel_gametic_2021].```\n",
    "\n",
    "## <span style=\"color:green\">Add a summary description to node cenH3.2_(Drosophila_eugracilis)</span>\n",
    "\n",
    "```cenH3.2_(Drosophila_eugracilis) is a centromere-specific histone variant encoded by the Cid2 gene, which has been identified exclusively in Drosophila eugracilis as a rare replacement for the ancestral Cid1 gene [kursel_recurrent_2017]. Unlike other Cid paralogs that coexist with Cid1, Cid2 uniquely serves as the sole centromeric histone in this species following the pseudogenization of Cid1 due to a frameshift-causing 2-bp deletion. While Cid2's precise functional specialization remains unstudied, its singular presence in D. eugracilis suggests it has assumed the core centromeric functions of the lost Cid1, though potential neofunctionalization cannot be ruled out given the lack of comparative data with other Cid paralogs [kursel_recurrent_2017].```\n",
    "\n",
    "## <span style=\"color:green\">Add a summary description to node cenH3.3_(Montium)</span>\n",
    "\n",
    "```cenH3.3_(Montium) is a centromere-specific histone variant encoded by the Cid3 gene, found exclusively in the montium subgroup of Drosophila (e.g., D. kikkawai, D. auraria) [kursel_recurrent_2017]. Unlike the ubiquitously expressed Cid1, Cid3 exhibits male germline-specific expression, suggesting specialized meiotic functions, potentially related to suppressing centromere drive during spermatogenesis [kursel_recurrent_2017]. The histone variant retains conserved histone-fold domains through gene conversion with Cid1 but has a divergent N-terminal tail. Evolving under positive selection, Cid3 shows adaptive changes in DNA-contact regions, suggesting conflict-driven evolution. This paralog persists alongside Cid1 and Cid4 in a triplicate system, maintaining non-redundant meiotic functions for 15 million years [kursel_recurrent_2017].```\n",
    "\n",
    "## <span style=\"color:green\">Add a summary description to node cenH3.4_(Montium)</span>\n",
    "\n",
    "```cenH3.4_(Montium) is a centromere-specific histone variant encoded by the Cid4 gene, uniquely present in the montium subgroup of Drosophila. Unlike its paralogs, Cid4 shows broad somatic expression and has effectively replaced Cid1 as the dominant centromeric histone in mitotic cells [kursel_recurrent_2017]. The histone variant shows positive selection in its DNA-binding Loop 1 region and retains all four conserved N-terminal motifs found in Cid1, while acquiring a novel motif that may enable specialized interactions [kursel_recurrent_2017]. As part of the montium's three-gene system with Cid1 and Cid3, Cid4 specializes in maintaining mitotic centromere function through 15 million years of evolution.```\n",
    "\n",
    "## <span style=\"color:green\">Add a summary description to node cenH3.5_(Drosophila)</span>\n",
    "\n",
    "```cenH3.5_(Drosophila) is a divergent centromeric histone paralog found exclusively in the Drosophila subgenus, originating from an ancient duplication of Cid1 [kursel_gametic_2021, kursel_recurrent_2017]. Unlike the ubiquitously expressed Cid1, Cid5 exhibits germline-restricted expression and has functionally specialized for male gametogenesis. It is absent in somatic cells but co-expressed with Cid1 in early germ cells of both sexes [kursel_gametic_2021].  Cid5 replaces Cid1 during male meiosis I, becoming the sole centromeric histone in sperm, but is lost during late oogenesis in females [kursel_gametic_2021]. After fertilization, paternally inherited Cid5 is rapidly replaced by maternal Cid1 in the embryo. The histone variant lacks conserved N-terminal motifs present in Cid1 and features a divergent histone-fold domain, potentially adapting it for sperm chromatin [kursel_gametic_2021]. Notably, Cid5 co-exists with CENP-C2 (male-biased CENP-C paralog) in the Drosophila subgenus, potentially forming a distinct kinetochore configuration during spermatogenesis [teixeira_concurrent_2018]. While Cid5 maintains key centromere localization motifs, its rapid evolution and testis-specific expression pattern indicate subfunctionalization or neofunctionalization relative to Cid1.```\n",
    "\n",
    "## <span style=\"color:green\">Add a summary description to node cenH3.6_(Repleta)</span>\n",
    "\n",
    "```cenH3.6_(Repleta)  is a recently evolved centromeric histone variant found specifically in Drosophila buzzatii and D. seriema (repleta group), resulting from an interchromosomal duplication of Cid1 that occurred 4.6-11.3 million years ago [teixeira_concurrent_2018]. This paralog shows ~80% amino acid identity with ancestral Cid1 from closely related species like D. mojavensis, compared to only ~40% identity with Cid5. Unlike the male germline-restricted Cid5, Cid6 displays constitutive expression across all developmental stages, suggesting it functionally replaced the degenerated Cid1 copy in these species [teixeira_concurrent_2018]. While Cid6 maintains core centromere functions, it shows no evidence of positive selection observed in Cid5, indicating preservation of ancestral Cid1 roles rather than neofunctionalization [teixeira_concurrent_2018].```\n",
    "\n",
    "## <span style=\"color:green\">Add new nodes cenH3.1_(Culicidae), cenH3.2_(Culicidae), cenH3.3_(Aedes) to cenH3_(Culicidae)</span>\n",
    "\n",
    "## <span style=\"color:green\">Add a summary description to node cenH3.1_(Culicidae)</span>\n",
    "\n",
    "```cenH3.1_(Culicidae) is a centromeric histone variant encoded by the mosqCid1 gene, found in most studied mosquito species (Anopheles, Aedes, Culex), with the exception of An. albimanus and An. darlingi where it has been lost [kursel_ancient_2020]. This gene emerged over 150 million years ago and has been retained in most species, acquiring specialized functions. Unlike its paralog mosqCid2, which shows high expression in ovaries and early embryos, mosqCid1 exhibits low expression levels in both somatic and germline tissues [kursel_ancient_2020]. Notably, mosqCid1 evolves under positive selection, suggesting its potential role in suppressing genetic conflicts such as centromere drive [kursel_ancient_2020]. Its N-terminal tail contains unique motifs absent in mosqCid2, likely reflecting differences in protein interactions.```\n",
    "\n",
    "## <span style=\"color:green\">Add a summary description to node cenH3.2_(Culicidae)</span>\n",
    "\n",
    "```cenH3.2_(Culicidae) is the ancestral (emerged over 150 My) centromeric histone variant in mosquitoes, conserved across Anopheles, Aedes, and Culex species [kursel_ancient_2020]. Unlike its paralog mosqCid1, mosqCid2 specializes in germline processes, showing dramatic upregulation (6-10 fold) in ovaries after blood feeding and during early embryogenesis [kursel_ancient_2020]. It contains four unique N-terminal motifs absent in mosqCid1, reflecting functional divergence. Remarkably, mosqCid2 serves as the sole centromeric histone in An. albimanus and An. darlingi following mosqCid1 loss, demonstrating its fundamental role in chromosome segregation [kursel_ancient_2020].```\n",
    "\n",
    "## <span style=\"color:green\">Add a summary description to node cenH3.3_(Aedes)</span>\n",
    "\n",
    "```cenH3.3_(Aedes) is a recently evolved centromeric histone variant unique to Aedes mosquitoes, arising from a lineage-specific duplication of mosqCid1 [kursel_ancient_2020]. This paralog demonstrates specialized temporal expression during embryogenesis, suggesting a developmental-stage specific role in early development. Unlike the germline-focused mosqCid2 or the ubiquitously expressed mosqCid1, mosqCid3 shows no significant expression in adult tissues or ovaries, indicating its exclusive embryonic function [kursel_ancient_2020].```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a60f3124-6ff9-46ad-954f-ff736bba9fb2",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add a new node cenH3_(Insecta) to cenH3_(Animals)</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "832e94de-4ab7-42d0-993f-3f85a4ed53c3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"cenH3_(Insecta)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Insecta\",\n",
    "    \"taxonomic_span_id\": \"50557 \",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cenH3_(Animals)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "024c28cb-49cb-4d6b-8acb-6c474ddc2254",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>cenH3_(Insecta)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>50557</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cenH3_(Animals)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "12  cenH3_(Insecta)  variant        Insecta            50557           NaN   \n",
       "\n",
       "             parent  \n",
       "12  cenH3_(Animals)  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone\"\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df[histone_df[\"id\"].isin([\"cenH3_(Insecta)\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b0a2236a-7739-429a-8806-f7e3bd3688b1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59a7ddcc-c394-424a-b4de-a5d90586a315",
   "metadata": {
    "tags": []
   },
   "source": [
    "# <span style=\"color:black\">Add a summary description to node cenH3_(Insecta)</span>\n",
    "\n",
    "```cenH3_(Insecta) is a centromere-specific histone variant in insects, functionally analogous to CENP-A (Centromere Protein A) in other eukaryotes. It is essential for centromere specification and proper chromosome segregation during cell division. Notably, cenH3 is absent in four holocentric insect clades, suggesting alternative centromere organization mechanisms in these lineages [talbert_histone_2021, senaratne_formation_2021, sridhar_kinetochore_2022]. Some insect families, such as Drosophilidae and Culicidae, possess multiple cenH3 paralogs with evidence of functional specialization. For more overview of these families, see the cenH3_(Drosophilidae) and cenH3_(Culicidae) classes.```\n",
    "\n",
    "# <span style=\"color:black\">Add a gene description to node cenH3_(Insecta)</span>\n",
    "\n",
    "```In some insect lineages, cenH3 has undergone gene duplication, leading to paralogs with distinct functional roles. For detailed lineage-specific information, see the cenH3_(Drosophilidae) and cenH3_(Culicidae) classes.```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "5f04a49e-1bc3-4351-a84e-47ec74933f4a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3_(Insecta)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>50557</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Animals)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  cenH3_(Insecta)  variant        Insecta            50557         None   \n",
       "\n",
       "            parent    id summary taxonomy genes  ... knock_out function  \\\n",
       "0  cenH3_(Animals)  None    None     None  None  ...      None     None   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     None         None       None      None         None    None    None   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3_(Insecta)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7ce6b4b0-c51d-4e69-8ef2-764158fd4bec",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "histone_description_summary = \"cenH3_(Insecta) is a centromere-specific histone variant in insects, functionally analogous to CENP-A (Centromere Protein A) in other eukaryotes. It is essential for centromere specification and proper chromosome segregation during cell division. Notably, cenH3 is absent in four holocentric insect clades, suggesting alternative centromere organization mechanisms in these lineages [talbert_histone_2021, senaratne_formation_2021, sridhar_kinetochore_2022]. Some insect families, such as Drosophilidae and Culicidae, possess multiple cenH3 paralogs with evidence of functional specialization. For more overview of these families, see the cenH3_(Drosophilidae) and cenH3_(Culicidae) classes.\"\n",
    "histone_description_genes = \"In some insect lineages, cenH3 has undergone gene duplication, leading to paralogs with distinct functional roles. For detailed lineage-specific information, see the cenH3_(Drosophilidae) and cenH3_(Culicidae) classes.\"\n",
    "data_histone_description = (\n",
    "    histone_description_summary,\n",
    "    None,\n",
    "    histone_description_genes,\n",
    ") + (None,) * 11\n",
    "cursor.execute(add_histone_description, data_histone_description)\n",
    "\n",
    "histone_description_id = cursor.lastrowid\n",
    "query = f\"UPDATE histone SET description={histone_description_id} WHERE id='cenH3_(Insecta)'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e811dc97-4595-4931-b722-200d03decaa7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3_(Insecta)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>50557</td>\n",
       "      <td>237</td>\n",
       "      <td>cenH3_(Animals)</td>\n",
       "      <td>237</td>\n",
       "      <td>cenH3_(Insecta) is a centromere-specific histo...</td>\n",
       "      <td>None</td>\n",
       "      <td>In some insect lineages, cenH3 has undergone g...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "0  cenH3_(Insecta)  variant        Insecta            50557           237   \n",
       "\n",
       "            parent   id                                            summary  \\\n",
       "0  cenH3_(Animals)  237  cenH3_(Insecta) is a centromere-specific histo...   \n",
       "\n",
       "  taxonomy                                              genes  ... knock_out  \\\n",
       "0     None  In some insect lineages, cenH3 has undergone g...  ...      None   \n",
       "\n",
       "  function sequence localization deposition structure interactions disease  \\\n",
       "0     None     None         None       None      None         None    None   \n",
       "\n",
       "  caveats relations  \n",
       "0    None      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3_(Insecta)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "071131cb-2b1a-4d5e-b9e4-ca04270cec4b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9dfc8b14-e5cd-4b09-8744-98e913c176ee",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add new nodes cenH3_(Drosophilidae), cenH3_(Culicidae) to cenH3_(Insecta)</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a67c272c-34fe-4ba7-9339-cf180252e878",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"cenH3_(Drosophilidae)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Drosophilidae\",\n",
    "    \"taxonomic_span_id\": \"7214\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cenH3_(Insecta)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "data_histone = {\n",
    "    \"id\": \"cenH3_(Culicidae)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Culicidae\",\n",
    "    \"taxonomic_span_id\": \"7157\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cenH3_(Insecta)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "570e59ed-53ab-4b8b-a83a-a128705d62ba",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>cenH3_(Culicidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Culicidae</td>\n",
       "      <td>7157</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cenH3_(Insecta)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophilidae</td>\n",
       "      <td>7214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cenH3_(Insecta)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       id    level taxonomic_span taxonomic_span_id  \\\n",
       "10      cenH3_(Culicidae)  variant      Culicidae              7157   \n",
       "11  cenH3_(Drosophilidae)  variant  Drosophilidae              7214   \n",
       "\n",
       "    description           parent  \n",
       "10          NaN  cenH3_(Insecta)  \n",
       "11          NaN  cenH3_(Insecta)  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone\"\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df[histone_df[\"id\"].isin([\"cenH3_(Drosophilidae)\", \"cenH3_(Culicidae)\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "14b2438f-8352-4dd2-955a-11bfe0920a9c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27d20383-57b1-4604-a790-9dc7e48cb31d",
   "metadata": {
    "tags": []
   },
   "source": [
    "# <span style=\"color:black\">Add a summary description to node cenH3_(Drosophilidae)</span>\n",
    "\n",
    "```cenH3_(Drosophilidae) is a centromere-specific histone variant in Drosophilidae family, functionally analogous to CENP-A in mammals.```\n",
    "\n",
    "# <span style=\"color:black\">Add a gene description to node cenH3_(Drosophilidae)</span>\n",
    "\n",
    "```In Drosophila, six cenH3 paralogs (Cid1–Cid6) have been identified: Cid1, Cid4, and Cid6 (and likely Cid2) are essential for mitotic chromosome segregation and are ubiquitously expressed in somatic cells, while the remaining paralogs are germline-specific and may play a role in suppressing centromere drive [kursel_gametic_2021, kursel_recurrent_2017].```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c8e84ef0-b62f-4039-a315-f236ac4c120e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophilidae</td>\n",
       "      <td>7214</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Insecta)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      id    level taxonomic_span taxonomic_span_id  \\\n",
       "0  cenH3_(Drosophilidae)  variant  Drosophilidae              7214   \n",
       "\n",
       "  description           parent    id summary taxonomy genes  ... knock_out  \\\n",
       "0        None  cenH3_(Insecta)  None    None     None  None  ...      None   \n",
       "\n",
       "  function sequence localization deposition structure interactions disease  \\\n",
       "0     None     None         None       None      None         None    None   \n",
       "\n",
       "  caveats relations  \n",
       "0    None      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3_(Drosophilidae)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "28726ff8-1428-4b9e-ab66-b53cba002e1a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "histone_description_summary = \"cenH3_(Drosophilidae) is a centromere-specific histone variant in Drosophilidae family, functionally analogous to CENP-A in mammals.\"\n",
    "histone_description_genes = \"In Drosophila, six cenH3 paralogs (Cid1–Cid6) have been identified: Cid1, Cid4, and Cid6 (and likely Cid2) are essential for mitotic chromosome segregation and are ubiquitously expressed in somatic cells, while the remaining paralogs are germline-specific and may play a role in suppressing centromere drive [kursel_gametic_2021, kursel_recurrent_2017].\"\n",
    "data_histone_description = (\n",
    "    histone_description_summary,\n",
    "    None,\n",
    "    histone_description_genes,\n",
    ") + (None,) * 11\n",
    "cursor.execute(add_histone_description, data_histone_description)\n",
    "\n",
    "histone_description_id = cursor.lastrowid\n",
    "query = f\"UPDATE histone SET description={histone_description_id} WHERE id='cenH3_(Drosophilidae)'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e31e83ca-59b1-4999-ae34-40bde1eb65e4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophilidae</td>\n",
       "      <td>7214</td>\n",
       "      <td>238</td>\n",
       "      <td>cenH3_(Insecta)</td>\n",
       "      <td>238</td>\n",
       "      <td>cenH3_(Drosophilidae) is a centromere-specific...</td>\n",
       "      <td>None</td>\n",
       "      <td>In Drosophila, six cenH3 paralogs (Cid1–Cid6) ...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      id    level taxonomic_span taxonomic_span_id  \\\n",
       "0  cenH3_(Drosophilidae)  variant  Drosophilidae              7214   \n",
       "\n",
       "   description           parent   id  \\\n",
       "0          238  cenH3_(Insecta)  238   \n",
       "\n",
       "                                             summary taxonomy  \\\n",
       "0  cenH3_(Drosophilidae) is a centromere-specific...     None   \n",
       "\n",
       "                                               genes  ... knock_out function  \\\n",
       "0  In Drosophila, six cenH3 paralogs (Cid1–Cid6) ...  ...      None     None   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     None         None       None      None         None    None    None   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3_(Drosophilidae)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d2f1a4d6-b3d0-41c8-8f3d-540f04a49b35",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa88c544-4dfd-42fc-a1ce-3eea58dc91b3",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add a summary description to node cenH3_(Culicidae)</span>\n",
    "\n",
    "```cenH3_(Culicidae) is a centromere-specific histone variant in Culicidae family (mosquitos), functionally analogous to CENP-A in mammals```\n",
    "\n",
    "# <span style=\"color:black\">Add a gene description to node cenH3_(Culicidae)</span>\n",
    "\n",
    "```In mosquitoes (Culicidae), three cenH3 paralogs exist: the evolutionarily ancient mosqCid1 and mosqCid2, which have been conserved for over 150 million years, and mosqCid3, a more recent paralog resulting from an independent duplication of mosqCid1 in Aedes aegypti and Aedes albopictus [kursel_ancient_2020].```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0b3ac354-b971-4eb8-aee6-c58ef7be145e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3_(Culicidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Culicidae</td>\n",
       "      <td>7157</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Insecta)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  cenH3_(Culicidae)  variant      Culicidae              7157        None   \n",
       "\n",
       "            parent    id summary taxonomy genes  ... knock_out function  \\\n",
       "0  cenH3_(Insecta)  None    None     None  None  ...      None     None   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     None         None       None      None         None    None    None   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3_(Culicidae)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "c317e330-7aec-45a9-8fdf-6ecbef4fa735",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "histone_description_summary = \"cenH3_(Culicidae) is a centromere-specific histone variant in Culicidae family (mosquitos), functionally analogous to CENP-A in mammals.\"\n",
    "histone_description_genes = \"In mosquitoes (Culicidae), three cenH3 paralogs exist: the evolutionarily ancient mosqCid1 and mosqCid2, which have been conserved for over 150 million years, and mosqCid3, a more recent paralog resulting from an independent duplication of mosqCid1 in Aedes aegypti and Aedes albopictus [kursel_ancient_2020].\"\n",
    "data_histone_description = (\n",
    "    histone_description_summary,\n",
    "    None,\n",
    "    histone_description_genes,\n",
    ") + (None,) * 11\n",
    "cursor.execute(add_histone_description, data_histone_description)\n",
    "\n",
    "histone_description_id = cursor.lastrowid\n",
    "query = f\"UPDATE histone SET description={histone_description_id} WHERE id='cenH3_(Culicidae)'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3c39352a-3da1-4ffd-87c3-368a4ae35911",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3_(Culicidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Culicidae</td>\n",
       "      <td>7157</td>\n",
       "      <td>239</td>\n",
       "      <td>cenH3_(Insecta)</td>\n",
       "      <td>239</td>\n",
       "      <td>cenH3_(Culicidae) is a centromere-specific his...</td>\n",
       "      <td>None</td>\n",
       "      <td>In mosquitoes (Culicidae), three cenH3 paralog...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "0  cenH3_(Culicidae)  variant      Culicidae              7157          239   \n",
       "\n",
       "            parent   id                                            summary  \\\n",
       "0  cenH3_(Insecta)  239  cenH3_(Culicidae) is a centromere-specific his...   \n",
       "\n",
       "  taxonomy                                              genes  ... knock_out  \\\n",
       "0     None  In mosquitoes (Culicidae), three cenH3 paralog...  ...      None   \n",
       "\n",
       "  function sequence localization deposition structure interactions disease  \\\n",
       "0     None     None         None       None      None         None    None   \n",
       "\n",
       "  caveats relations  \n",
       "0    None      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3_(Culicidae)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "c95f439f-34c1-494f-b414-7b201cac11b2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca1e37e6-b8ca-48d3-b59c-0cebd30f6452",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add alternate names cid, cid1, cid2, cid3, cid4, cid5, cid6, mosqCid1, mosqCid2, mosqCid3 to cenH3_(Insecta)</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0705df05-0e42-4b8b-b5eb-719c93e53fe3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, name, taxonomy, gene, splice, histone]\n",
       "Index: []"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name \" \"WHERE histone='cenH3_(Insecta)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "e058355b-209f-4594-99df-3ef560f568c4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['cid',\n",
       " 'cid1',\n",
       " 'cid2',\n",
       " 'cid3',\n",
       " 'cid4',\n",
       " 'cid5',\n",
       " 'cid6',\n",
       " 'mosqCid1',\n",
       " 'mosqCid2',\n",
       " 'mosqCid3']"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alternate_names = (\n",
    "    [\"cid\"] + [f\"cid{i}\" for i in range(1, 7)] + [f\"mosqCid{i}\" for i in range(1, 4)]\n",
    ")\n",
    "alternate_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "d98e4f10-0dce-4f3d-bcae-2b56c4bec13c",
   "metadata": {},
   "outputs": [],
   "source": [
    "for an in alternate_names:\n",
    "    data_alternate_name = {\n",
    "        \"name\": an,\n",
    "        \"taxonomy\": None,\n",
    "        \"gene\": None,\n",
    "        \"splice\": None,\n",
    "        \"histone\": \"cenH3_(Insecta)\",\n",
    "    }\n",
    "    cursor.execute(add_alternate_names, data_alternate_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "057c93af-1e1b-480c-8a0d-499c619ff44d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>99</td>\n",
       "      <td>cid</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Insecta)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100</td>\n",
       "      <td>cid1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Insecta)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>101</td>\n",
       "      <td>cid2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Insecta)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>102</td>\n",
       "      <td>cid3</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Insecta)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>103</td>\n",
       "      <td>cid4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Insecta)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>104</td>\n",
       "      <td>cid5</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Insecta)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>105</td>\n",
       "      <td>cid6</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Insecta)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>106</td>\n",
       "      <td>mosqCid1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Insecta)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>107</td>\n",
       "      <td>mosqCid2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Insecta)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>108</td>\n",
       "      <td>mosqCid3</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Insecta)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    id      name taxonomy  gene splice          histone\n",
       "0   99       cid     None  None   None  cenH3_(Insecta)\n",
       "1  100      cid1     None  None   None  cenH3_(Insecta)\n",
       "2  101      cid2     None  None   None  cenH3_(Insecta)\n",
       "3  102      cid3     None  None   None  cenH3_(Insecta)\n",
       "4  103      cid4     None  None   None  cenH3_(Insecta)\n",
       "5  104      cid5     None  None   None  cenH3_(Insecta)\n",
       "6  105      cid6     None  None   None  cenH3_(Insecta)\n",
       "7  106  mosqCid1     None  None   None  cenH3_(Insecta)\n",
       "8  107  mosqCid2     None  None   None  cenH3_(Insecta)\n",
       "9  108  mosqCid3     None  None   None  cenH3_(Insecta)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name \" \"WHERE histone='cenH3_(Insecta)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "5b38e565-0add-4b3a-8831-f4a966cf791a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "075e433c-5374-464b-a213-d98e5c4ed03f",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add alternate names cid, cid1, cid2, cid3, cid4, cid5, cid6 to cenH3_(Drosophilidae)</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "77285b29-6642-47fd-b986-145509b279a5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, name, taxonomy, gene, splice, histone]\n",
       "Index: []"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name \" \"WHERE histone='cenH3_(Drosophilidae)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "e758c8f6-3a9c-47b2-bbd2-5161fd1bf7ed",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['cid', 'cid1', 'cid2', 'cid3', 'cid4', 'cid5', 'cid6']"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alternate_names = [\"cid\"] + [f\"cid{i}\" for i in range(1, 7)]\n",
    "alternate_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "07cd7307-4fe3-48f0-b2d2-7537a3891955",
   "metadata": {},
   "outputs": [],
   "source": [
    "for an in alternate_names:\n",
    "    data_alternate_name = {\n",
    "        \"name\": an,\n",
    "        \"taxonomy\": None,\n",
    "        \"gene\": None,\n",
    "        \"splice\": None,\n",
    "        \"histone\": \"cenH3_(Drosophilidae)\",\n",
    "    }\n",
    "    cursor.execute(add_alternate_names, data_alternate_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "1f703e0f-9842-442f-a410-00fe006da62c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>109</td>\n",
       "      <td>cid</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>110</td>\n",
       "      <td>cid1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>111</td>\n",
       "      <td>cid2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>112</td>\n",
       "      <td>cid3</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>113</td>\n",
       "      <td>cid4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>114</td>\n",
       "      <td>cid5</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>115</td>\n",
       "      <td>cid6</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    id  name taxonomy  gene splice                histone\n",
       "0  109   cid     None  None   None  cenH3_(Drosophilidae)\n",
       "1  110  cid1     None  None   None  cenH3_(Drosophilidae)\n",
       "2  111  cid2     None  None   None  cenH3_(Drosophilidae)\n",
       "3  112  cid3     None  None   None  cenH3_(Drosophilidae)\n",
       "4  113  cid4     None  None   None  cenH3_(Drosophilidae)\n",
       "5  114  cid5     None  None   None  cenH3_(Drosophilidae)\n",
       "6  115  cid6     None  None   None  cenH3_(Drosophilidae)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name \" \"WHERE histone='cenH3_(Drosophilidae)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "89667ca6-332a-4fa1-b76f-971d69abd0a2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d40991d3-7326-4f21-a3d5-7f20a087bc5c",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add alternate names mosqCid1, mosqCid2, mosqCid3 to cenH3_(Culicidae)</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "2d4d583c-9cf6-47d3-a4dd-fe734922a540",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, name, taxonomy, gene, splice, histone]\n",
       "Index: []"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name \" \"WHERE histone='cenH3_(Culicidae)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "c541b6e3-c049-49fb-bc53-4b63ae545007",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['mosqCid1', 'mosqCid2', 'mosqCid3']"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alternate_names = [f\"mosqCid{i}\" for i in range(1, 4)]\n",
    "alternate_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "3c5bcba1-a8fb-4041-97d3-520947aa1e07",
   "metadata": {},
   "outputs": [],
   "source": [
    "for an in alternate_names:\n",
    "    data_alternate_name = {\n",
    "        \"name\": an,\n",
    "        \"taxonomy\": None,\n",
    "        \"gene\": None,\n",
    "        \"splice\": None,\n",
    "        \"histone\": \"cenH3_(Culicidae)\",\n",
    "    }\n",
    "    cursor.execute(add_alternate_names, data_alternate_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "ed5ed2ae-ed85-4358-9b4b-00deaae39bd1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>116</td>\n",
       "      <td>mosqCid1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Culicidae)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>117</td>\n",
       "      <td>mosqCid2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Culicidae)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>118</td>\n",
       "      <td>mosqCid3</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Culicidae)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    id      name taxonomy  gene splice            histone\n",
       "0  116  mosqCid1     None  None   None  cenH3_(Culicidae)\n",
       "1  117  mosqCid2     None  None   None  cenH3_(Culicidae)\n",
       "2  118  mosqCid3     None  None   None  cenH3_(Culicidae)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name \" \"WHERE histone='cenH3_(Culicidae)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "aede68c6-3444-4f8e-8b4c-3eeccdf0d1c7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3887ac20-eb64-4890-ba33-9a0864b80a6a",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add new nodes cenH3.1_(Drosophilidae), cenH3.2_(Drosophila_eugracilis), cenH3.3_(Montium), cenH3.4_(Montium), cenH3.5_(Drosophila), cenH3.6_(Repleta) to cenH3_(Drosophilidae)</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "971056f5-3441-4a3d-a7be-0a91db9a2e85",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "histone_nodes = [\n",
    "    \"cenH3.1_(Drosophilidae)\",\n",
    "    \"cenH3.2_(Drosophila_eugracilis)\",\n",
    "    \"cenH3.3_(Montium)\",\n",
    "    \"cenH3.4_(Montium)\",\n",
    "    \"cenH3.5_(Drosophila)\",\n",
    "    \"cenH3.6_(Repleta)\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "0cbf9933-4806-4799-8981-cbeed5ed1145",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"cenH3.1_(Drosophilidae)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Drosophilidae\",\n",
    "    \"taxonomic_span_id\": \"7214\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cenH3_(Drosophilidae)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "data_histone = {\n",
    "    \"id\": \"cenH3.2_(Drosophila_eugracilis)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Drosophila eugracilis\",\n",
    "    \"taxonomic_span_id\": \"29029\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cenH3_(Drosophilidae)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "data_histone = {\n",
    "    \"id\": \"cenH3.3_(Montium)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"montium subgroup\",\n",
    "    \"taxonomic_span_id\": \"32352\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cenH3_(Drosophilidae)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "data_histone = {\n",
    "    \"id\": \"cenH3.4_(Montium)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"montium subgroup\",\n",
    "    \"taxonomic_span_id\": \"32352\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cenH3_(Drosophilidae)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "data_histone = {\n",
    "    \"id\": \"cenH3.5_(Drosophila)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Drosophila\",\n",
    "    \"taxonomic_span_id\": \"32281\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cenH3_(Drosophilidae)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "data_histone = {\n",
    "    \"id\": \"cenH3.6_(Repleta)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"repleta group\",\n",
    "    \"taxonomic_span_id\": \"32321\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cenH3_(Drosophilidae)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "eb65d8b9-6a78-45fc-9be8-1a47943caa4e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>cenH3.1_(Drosophilidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophilidae</td>\n",
       "      <td>7214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>cenH3.2_(Drosophila_eugracilis)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophila eugracilis</td>\n",
       "      <td>29029</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>cenH3.3_(Montium)</td>\n",
       "      <td>variant</td>\n",
       "      <td>montium subgroup</td>\n",
       "      <td>32352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>cenH3.4_(Montium)</td>\n",
       "      <td>variant</td>\n",
       "      <td>montium subgroup</td>\n",
       "      <td>32352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>cenH3.5_(Drosophila)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophila</td>\n",
       "      <td>32281</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>cenH3.6_(Repleta)</td>\n",
       "      <td>variant</td>\n",
       "      <td>repleta group</td>\n",
       "      <td>32321</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 id    level         taxonomic_span  \\\n",
       "5           cenH3.1_(Drosophilidae)  variant          Drosophilidae   \n",
       "8   cenH3.2_(Drosophila_eugracilis)  variant  Drosophila eugracilis   \n",
       "11                cenH3.3_(Montium)  variant       montium subgroup   \n",
       "12                cenH3.4_(Montium)  variant       montium subgroup   \n",
       "13             cenH3.5_(Drosophila)  variant             Drosophila   \n",
       "14                cenH3.6_(Repleta)  variant          repleta group   \n",
       "\n",
       "   taxonomic_span_id  description                 parent  \n",
       "5               7214          NaN  cenH3_(Drosophilidae)  \n",
       "8              29029          NaN  cenH3_(Drosophilidae)  \n",
       "11             32352          NaN  cenH3_(Drosophilidae)  \n",
       "12             32352          NaN  cenH3_(Drosophilidae)  \n",
       "13             32281          NaN  cenH3_(Drosophilidae)  \n",
       "14             32321          NaN  cenH3_(Drosophilidae)  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone\"\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df[histone_df[\"id\"].isin(histone_nodes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "2b2c5e3e-7ca5-40ce-936b-a6c62a9f295a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a57b8964-6fa9-4615-a6b3-e07f9005b8b4",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add a summary description to node cenH3.1</span>\n",
    "\n",
    "```CenH3.1_(Drosophilidae) is a centromere-specific histone variant encoded by the Cid1 (also known as Cid) gene found in most Drosophila species except D. eugracilis, where it was pseudogenized and replaced by Cid2, and D. buzzatii and D. seriema, where Cid1 degenerated due to transposon insertions and was replaced by a new copy - Cid6 [teixeira_concurrent_2018, kursel_recurrent_2017]. As the ancestral centromeric histone in Drosophilidae, Cid1  represents the canonical centromeric histone in this family. It is essential for chromosome segregation during mitosis and meiosis, maintaining conserved histone-fold domains (HFD) critical for nucleosome assembly at centromeres. Unlike its paralogs (Cid3-Cid5), Cid1 is ubiquitously expressed in both somatic and germline tissues, though its expression is highest in gonads [kursel_recurrent_2017]. The N-terminal tail of Cid1 contains four conserved motifs. Remarkably, Cid1 persists throughout oogenesis in females but is specifically degraded during male meiosis I, disappearing from post-meiotic spermatids and sperm [kursel_gametic_2021].```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "92c43c66-4c7c-467a-adea-94d92fa64ccb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3.1_(Drosophilidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophilidae</td>\n",
       "      <td>7214</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        id    level taxonomic_span taxonomic_span_id  \\\n",
       "0  cenH3.1_(Drosophilidae)  variant  Drosophilidae              7214   \n",
       "\n",
       "  description                 parent    id summary taxonomy genes  ...  \\\n",
       "0        None  cenH3_(Drosophilidae)  None    None     None  None  ...   \n",
       "\n",
       "  knock_out function sequence localization deposition structure interactions  \\\n",
       "0      None     None     None         None       None      None         None   \n",
       "\n",
       "  disease caveats relations  \n",
       "0    None    None      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3.1_(Drosophilidae)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "7505d183-d7a6-4204-b91c-c78b602e0b9b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "histone_description_summary = \"CenH3.1_(Drosophilidae) is a centromere-specific histone variant encoded by the Cid1 (also known as Cid) gene found in most Drosophila species except D. eugracilis, where it was pseudogenized and replaced by Cid2, and D. buzzatii and D. seriema, where Cid1 degenerated due to transposon insertions and was replaced by a new copy - Cid6 [teixeira_concurrent_2018, kursel_recurrent_2017]. As the ancestral centromeric histone in Drosophilidae, Cid1  represents the canonical centromeric histone in this family. It is essential for chromosome segregation during mitosis and meiosis, maintaining conserved histone-fold domains (HFD) critical for nucleosome assembly at centromeres. Unlike its paralogs (Cid3-Cid5), Cid1 is ubiquitously expressed in both somatic and germline tissues, though its expression is highest in gonads [kursel_recurrent_2017]. The N-terminal tail of Cid1 contains four conserved motifs. Remarkably, Cid1 persists throughout oogenesis in females but is specifically degraded during male meiosis I, disappearing from post-meiotic spermatids and sperm [kursel_gametic_2021].\"\n",
    "data_histone_description = (histone_description_summary,) + (None,) * 13\n",
    "cursor.execute(add_histone_description, data_histone_description)\n",
    "\n",
    "histone_description_id = cursor.lastrowid\n",
    "query = f\"UPDATE histone SET description={histone_description_id} WHERE id='cenH3.1_(Drosophilidae)'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "081c82cb-dd9f-4b9f-8fc6-a8f82dc35456",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3.1_(Drosophilidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophilidae</td>\n",
       "      <td>7214</td>\n",
       "      <td>240</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>240</td>\n",
       "      <td>CenH3.1_(Drosophilidae) is a centromere-specif...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        id    level taxonomic_span taxonomic_span_id  \\\n",
       "0  cenH3.1_(Drosophilidae)  variant  Drosophilidae              7214   \n",
       "\n",
       "   description                 parent   id  \\\n",
       "0          240  cenH3_(Drosophilidae)  240   \n",
       "\n",
       "                                             summary taxonomy genes  ...  \\\n",
       "0  CenH3.1_(Drosophilidae) is a centromere-specif...     None  None  ...   \n",
       "\n",
       "  knock_out function sequence localization deposition structure interactions  \\\n",
       "0      None     None     None         None       None      None         None   \n",
       "\n",
       "  disease caveats relations  \n",
       "0    None    None      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3.1_(Drosophilidae)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "cac7ca1c-9991-42ed-b01e-b6f1a4c2f3fa",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a857aff8-bcfd-4aee-9223-200ff5cb1c1c",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add a summary description to node cenH3.2_(Drosophila_eugracilis)</span>\n",
    "\n",
    "```cenH3.2_(Drosophila_eugracilis) is a centromere-specific histone variant encoded by the Cid2 gene, which has been identified exclusively in Drosophila eugracilis as a rare replacement for the ancestral Cid1 gene [kursel_recurrent_2017]. Unlike other Cid paralogs that coexist with Cid1, Cid2 uniquely serves as the sole centromeric histone in this species following the pseudogenization of Cid1 due to a frameshift-causing 2-bp deletion. While Cid2's precise functional specialization remains unstudied, its singular presence in D. eugracilis suggests it has assumed the core centromeric functions of the lost Cid1, though potential neofunctionalization cannot be ruled out given the lack of comparative data with other Cid paralogs [kursel_recurrent_2017].```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "2718b53b-274d-42db-afc0-cad86261e3d9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3.2_(Drosophila_eugracilis)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophila eugracilis</td>\n",
       "      <td>29029</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                id    level         taxonomic_span  \\\n",
       "0  cenH3.2_(Drosophila_eugracilis)  variant  Drosophila eugracilis   \n",
       "\n",
       "  taxonomic_span_id description                 parent    id summary taxonomy  \\\n",
       "0             29029        None  cenH3_(Drosophilidae)  None    None     None   \n",
       "\n",
       "  genes  ... knock_out function sequence localization deposition structure  \\\n",
       "0  None  ...      None     None     None         None       None      None   \n",
       "\n",
       "  interactions disease caveats relations  \n",
       "0         None    None    None      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3.2_(Drosophila_eugracilis)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "1528005b-3690-41b1-a407-4f4adb5ed21e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "histone_description_summary = \"cenH3.2_(Drosophila_eugracilis) is a centromere-specific histone variant encoded by the Cid2 gene, which has been identified exclusively in Drosophila eugracilis as a rare replacement for the ancestral Cid1 gene [kursel_recurrent_2017]. Unlike other Cid paralogs that coexist with Cid1, Cid2 uniquely serves as the sole centromeric histone in this species following the pseudogenization of Cid1 due to a frameshift-causing 2-bp deletion. While Cid2's precise functional specialization remains unstudied, its singular presence in D. eugracilis suggests it has assumed the core centromeric functions of the lost Cid1, though potential neofunctionalization cannot be ruled out given the lack of comparative data with other Cid paralogs [kursel_recurrent_2017].\"\n",
    "data_histone_description = (histone_description_summary,) + (None,) * 13\n",
    "cursor.execute(add_histone_description, data_histone_description)\n",
    "\n",
    "histone_description_id = cursor.lastrowid\n",
    "query = f\"UPDATE histone SET description={histone_description_id} WHERE id='cenH3.2_(Drosophila_eugracilis)'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "830ec9a6-8a47-45c1-9848-8bce3fd09370",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3.2_(Drosophila_eugracilis)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophila eugracilis</td>\n",
       "      <td>29029</td>\n",
       "      <td>241</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>241</td>\n",
       "      <td>cenH3.2_(Drosophila_eugracilis) is a centromer...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                id    level         taxonomic_span  \\\n",
       "0  cenH3.2_(Drosophila_eugracilis)  variant  Drosophila eugracilis   \n",
       "\n",
       "  taxonomic_span_id  description                 parent   id  \\\n",
       "0             29029          241  cenH3_(Drosophilidae)  241   \n",
       "\n",
       "                                             summary taxonomy genes  ...  \\\n",
       "0  cenH3.2_(Drosophila_eugracilis) is a centromer...     None  None  ...   \n",
       "\n",
       "  knock_out function sequence localization deposition structure interactions  \\\n",
       "0      None     None     None         None       None      None         None   \n",
       "\n",
       "  disease caveats relations  \n",
       "0    None    None      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3.2_(Drosophila_eugracilis)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "46257032-3915-49ba-8109-6d306985e8e5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85ef7e24-1d94-4fda-99bc-a9ec87d6c486",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add a summary description to node cenH3.3_(Montium)</span>\n",
    "\n",
    "```cenH3.3_(Montium) is a centromere-specific histone variant encoded by the Cid3 gene, found exclusively in the montium subgroup of Drosophila (e.g., D. kikkawai, D. auraria) [kursel_recurrent_2017]. Unlike the ubiquitously expressed Cid1, Cid3 exhibits male germline-specific expression, suggesting specialized meiotic functions, potentially related to suppressing centromere drive during spermatogenesis [kursel_recurrent_2017]. The histone variant retains conserved histone-fold domains through gene conversion with Cid1 but has a divergent N-terminal tail. Evolving under positive selection, Cid3 shows adaptive changes in DNA-contact regions, suggesting conflict-driven evolution. This paralog persists alongside Cid1 and Cid4 in a triplicate system, maintaining non-redundant meiotic functions for 15 million years [kursel_recurrent_2017].```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "1854e017-8571-444f-b97b-c0f9eb9f1249",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3.3_(Montium)</td>\n",
       "      <td>variant</td>\n",
       "      <td>montium subgroup</td>\n",
       "      <td>32352</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  id    level    taxonomic_span taxonomic_span_id description  \\\n",
       "0  cenH3.3_(Montium)  variant  montium subgroup             32352        None   \n",
       "\n",
       "                  parent    id summary taxonomy genes  ... knock_out function  \\\n",
       "0  cenH3_(Drosophilidae)  None    None     None  None  ...      None     None   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     None         None       None      None         None    None    None   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3.3_(Montium)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "737fb4de-f330-4031-9d2b-75cb4c17ac42",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "histone_description_summary = \"cenH3.3_(Montium) is a centromere-specific histone variant encoded by the Cid3 gene, found exclusively in the montium subgroup of Drosophila (e.g., D. kikkawai, D. auraria) [kursel_recurrent_2017]. Unlike the ubiquitously expressed Cid1, Cid3 exhibits male germline-specific expression, suggesting specialized meiotic functions, potentially related to suppressing centromere drive during spermatogenesis [kursel_recurrent_2017]. The histone variant retains conserved histone-fold domains through gene conversion with Cid1 but has a divergent N-terminal tail. Evolving under positive selection, Cid3 shows adaptive changes in DNA-contact regions, suggesting conflict-driven evolution. This paralog persists alongside Cid1 and Cid4 in a triplicate system, maintaining non-redundant meiotic functions for 15 million years [kursel_recurrent_2017].\"\n",
    "data_histone_description = (histone_description_summary,) + (None,) * 13\n",
    "cursor.execute(add_histone_description, data_histone_description)\n",
    "\n",
    "histone_description_id = cursor.lastrowid\n",
    "query = f\"UPDATE histone SET description={histone_description_id} WHERE id='cenH3.3_(Montium)'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "73639f4c-956c-438b-a254-72a6118dc4c2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3.3_(Montium)</td>\n",
       "      <td>variant</td>\n",
       "      <td>montium subgroup</td>\n",
       "      <td>32352</td>\n",
       "      <td>242</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>242</td>\n",
       "      <td>cenH3.3_(Montium) is a centromere-specific his...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  id    level    taxonomic_span taxonomic_span_id  \\\n",
       "0  cenH3.3_(Montium)  variant  montium subgroup             32352   \n",
       "\n",
       "   description                 parent   id  \\\n",
       "0          242  cenH3_(Drosophilidae)  242   \n",
       "\n",
       "                                             summary taxonomy genes  ...  \\\n",
       "0  cenH3.3_(Montium) is a centromere-specific his...     None  None  ...   \n",
       "\n",
       "  knock_out function sequence localization deposition structure interactions  \\\n",
       "0      None     None     None         None       None      None         None   \n",
       "\n",
       "  disease caveats relations  \n",
       "0    None    None      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3.3_(Montium)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "7ec46584-f40b-43d4-b507-688d47260b23",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9749e31-bf11-4d68-a74a-9c023293a16b",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add a summary description to node cenH3.4_(Montium)</span>\n",
    "\n",
    "```cenH3.4_(Montium) is a centromere-specific histone variant encoded by the Cid4 gene, uniquely present in the montium subgroup of Drosophila. Unlike its paralogs, Cid4 shows broad somatic expression and has effectively replaced Cid1 as the dominant centromeric histone in mitotic cells [kursel_recurrent_2017]. The histone variant shows positive selection in its DNA-binding Loop 1 region and retains all four conserved N-terminal motifs found in Cid1, while acquiring a novel motif that may enable specialized interactions [kursel_recurrent_2017]. As part of the montium's three-gene system with Cid1 and Cid3, Cid4 specializes in maintaining mitotic centromere function through 15 million years of evolution.```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "5696e56b-724c-473c-80e7-8bff66fb1f1f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3.4_(Montium)</td>\n",
       "      <td>variant</td>\n",
       "      <td>montium subgroup</td>\n",
       "      <td>32352</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  id    level    taxonomic_span taxonomic_span_id description  \\\n",
       "0  cenH3.4_(Montium)  variant  montium subgroup             32352        None   \n",
       "\n",
       "                  parent    id summary taxonomy genes  ... knock_out function  \\\n",
       "0  cenH3_(Drosophilidae)  None    None     None  None  ...      None     None   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     None         None       None      None         None    None    None   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3.4_(Montium)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "02dc6620-aa09-4c52-9c1d-559f90b550f2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "histone_description_summary = \"cenH3.4_(Montium) is a centromere-specific histone variant encoded by the Cid4 gene, uniquely present in the montium subgroup of Drosophila. Unlike its paralogs, Cid4 shows broad somatic expression and has effectively replaced Cid1 as the dominant centromeric histone in mitotic cells [kursel_recurrent_2017]. The histone variant shows positive selection in its DNA-binding Loop 1 region and retains all four conserved N-terminal motifs found in Cid1, while acquiring a novel motif that may enable specialized interactions [kursel_recurrent_2017]. As part of the montium's three-gene system with Cid1 and Cid3, Cid4 specializes in maintaining mitotic centromere function through 15 million years of evolution.\"\n",
    "data_histone_description = (histone_description_summary,) + (None,) * 13\n",
    "cursor.execute(add_histone_description, data_histone_description)\n",
    "\n",
    "histone_description_id = cursor.lastrowid\n",
    "query = f\"UPDATE histone SET description={histone_description_id} WHERE id='cenH3.4_(Montium)'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "acf40373-7c42-453c-bb76-2d2c8c989db1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3.4_(Montium)</td>\n",
       "      <td>variant</td>\n",
       "      <td>montium subgroup</td>\n",
       "      <td>32352</td>\n",
       "      <td>243</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>243</td>\n",
       "      <td>cenH3.4_(Montium) is a centromere-specific his...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  id    level    taxonomic_span taxonomic_span_id  \\\n",
       "0  cenH3.4_(Montium)  variant  montium subgroup             32352   \n",
       "\n",
       "   description                 parent   id  \\\n",
       "0          243  cenH3_(Drosophilidae)  243   \n",
       "\n",
       "                                             summary taxonomy genes  ...  \\\n",
       "0  cenH3.4_(Montium) is a centromere-specific his...     None  None  ...   \n",
       "\n",
       "  knock_out function sequence localization deposition structure interactions  \\\n",
       "0      None     None     None         None       None      None         None   \n",
       "\n",
       "  disease caveats relations  \n",
       "0    None    None      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3.4_(Montium)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "c099c50d-8543-4712-bec1-1a2a8d19a19c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e202a8f-ffcf-40ab-b5e3-ffd8523e2432",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add a summary description to node cenH3.5_(Drosophila)</span>\n",
    "\n",
    "```cenH3.5_(Drosophila) is a divergent centromeric histone paralog found exclusively in the Drosophila subgenus, originating from an ancient duplication of Cid1 [kursel_gametic_2021, kursel_recurrent_2017]. Unlike the ubiquitously expressed Cid1, Cid5 exhibits germline-restricted expression and has functionally specialized for male gametogenesis. It is absent in somatic cells but co-expressed with Cid1 in early germ cells of both sexes [kursel_gametic_2021].  Cid5 replaces Cid1 during male meiosis I, becoming the sole centromeric histone in sperm, but is lost during late oogenesis in females [kursel_gametic_2021]. After fertilization, paternally inherited Cid5 is rapidly replaced by maternal Cid1 in the embryo. The histone variant lacks conserved N-terminal motifs present in Cid1 and features a divergent histone-fold domain, potentially adapting it for sperm chromatin [kursel_gametic_2021]. Notably, Cid5 co-exists with CENP-C2 (male-biased CENP-C paralog) in the Drosophila subgenus, potentially forming a distinct kinetochore configuration during spermatogenesis [teixeira_concurrent_2018]. While Cid5 maintains key centromere localization motifs, its rapid evolution and testis-specific expression pattern indicate subfunctionalization or neofunctionalization relative to Cid1.```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "eb673951-7aaa-40e3-ad6c-1135501f6dc9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3.5_(Drosophila)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophila</td>\n",
       "      <td>32281</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  cenH3.5_(Drosophila)  variant     Drosophila             32281        None   \n",
       "\n",
       "                  parent    id summary taxonomy genes  ... knock_out function  \\\n",
       "0  cenH3_(Drosophilidae)  None    None     None  None  ...      None     None   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     None         None       None      None         None    None    None   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3.5_(Drosophila)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "313eccb9-897c-444a-863d-799e5e5fe68a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "histone_description_summary = \"cenH3.5_(Drosophila) is a divergent centromeric histone paralog found exclusively in the Drosophila subgenus, originating from an ancient duplication of Cid1 [kursel_gametic_2021, kursel_recurrent_2017]. Unlike the ubiquitously expressed Cid1, Cid5 exhibits germline-restricted expression and has functionally specialized for male gametogenesis. It is absent in somatic cells but co-expressed with Cid1 in early germ cells of both sexes [kursel_gametic_2021].  Cid5 replaces Cid1 during male meiosis I, becoming the sole centromeric histone in sperm, but is lost during late oogenesis in females [kursel_gametic_2021]. After fertilization, paternally inherited Cid5 is rapidly replaced by maternal Cid1 in the embryo. The histone variant lacks conserved N-terminal motifs present in Cid1 and features a divergent histone-fold domain, potentially adapting it for sperm chromatin [kursel_gametic_2021]. Notably, Cid5 co-exists with CENP-C2 (male-biased CENP-C paralog) in the Drosophila subgenus, potentially forming a distinct kinetochore configuration during spermatogenesis [teixeira_concurrent_2018]. While Cid5 maintains key centromere localization motifs, its rapid evolution and testis-specific expression pattern indicate subfunctionalization or neofunctionalization relative to Cid1.\"\n",
    "data_histone_description = (histone_description_summary,) + (None,) * 13\n",
    "cursor.execute(add_histone_description, data_histone_description)\n",
    "\n",
    "histone_description_id = cursor.lastrowid\n",
    "query = f\"UPDATE histone SET description={histone_description_id} WHERE id='cenH3.5_(Drosophila)'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "e4ebacb6-6ed0-4490-9852-633cec8594fb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3.5_(Drosophila)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Drosophila</td>\n",
       "      <td>32281</td>\n",
       "      <td>244</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>244</td>\n",
       "      <td>cenH3.5_(Drosophila) is a divergent centromeri...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     id    level taxonomic_span taxonomic_span_id  \\\n",
       "0  cenH3.5_(Drosophila)  variant     Drosophila             32281   \n",
       "\n",
       "   description                 parent   id  \\\n",
       "0          244  cenH3_(Drosophilidae)  244   \n",
       "\n",
       "                                             summary taxonomy genes  ...  \\\n",
       "0  cenH3.5_(Drosophila) is a divergent centromeri...     None  None  ...   \n",
       "\n",
       "  knock_out function sequence localization deposition structure interactions  \\\n",
       "0      None     None     None         None       None      None         None   \n",
       "\n",
       "  disease caveats relations  \n",
       "0    None    None      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3.5_(Drosophila)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "d0331a9e-0349-4b8d-81c4-526661e6a0b2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e51ef75a-dc8c-4b1d-9815-01c15dcacdad",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add a summary description to node cenH3.6_(Repleta)</span>\n",
    "\n",
    "```cenH3.6_(Repleta)  is a recently evolved centromeric histone variant found specifically in Drosophila buzzatii and D. seriema (repleta group), resulting from an interchromosomal duplication of Cid1 that occurred 4.6-11.3 million years ago [teixeira_concurrent_2018]. This paralog shows ~80% amino acid identity with ancestral Cid1 from closely related species like D. mojavensis, compared to only ~40% identity with Cid5. Unlike the male germline-restricted Cid5, Cid6 displays constitutive expression across all developmental stages, suggesting it functionally replaced the degenerated Cid1 copy in these species [teixeira_concurrent_2018]. While Cid6 maintains core centromere functions, it shows no evidence of positive selection observed in Cid5, indicating preservation of ancestral Cid1 roles rather than neofunctionalization [teixeira_concurrent_2018].```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "eebc2db2-6a0b-43ef-880f-a09833290ae7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3.6_(Repleta)</td>\n",
       "      <td>variant</td>\n",
       "      <td>repleta group</td>\n",
       "      <td>32321</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  cenH3.6_(Repleta)  variant  repleta group             32321        None   \n",
       "\n",
       "                  parent    id summary taxonomy genes  ... knock_out function  \\\n",
       "0  cenH3_(Drosophilidae)  None    None     None  None  ...      None     None   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     None         None       None      None         None    None    None   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3.6_(Repleta)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "cae5facf-2876-4273-8f70-28ad16756872",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "histone_description_summary = \"cenH3.6_(Repleta)  is a recently evolved centromeric histone variant found specifically in Drosophila buzzatii and D. seriema (repleta group), resulting from an interchromosomal duplication of Cid1 that occurred 4.6-11.3 million years ago [teixeira_concurrent_2018]. This paralog shows ~80% amino acid identity with ancestral Cid1 from closely related species like D. mojavensis, compared to only ~40% identity with Cid5. Unlike the male germline-restricted Cid5, Cid6 displays constitutive expression across all developmental stages, suggesting it functionally replaced the degenerated Cid1 copy in these species [teixeira_concurrent_2018]. While Cid6 maintains core centromere functions, it shows no evidence of positive selection observed in Cid5, indicating preservation of ancestral Cid1 roles rather than neofunctionalization [teixeira_concurrent_2018].\"\n",
    "data_histone_description = (histone_description_summary,) + (None,) * 13\n",
    "cursor.execute(add_histone_description, data_histone_description)\n",
    "\n",
    "histone_description_id = cursor.lastrowid\n",
    "query = f\"UPDATE histone SET description={histone_description_id} WHERE id='cenH3.6_(Repleta)'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "49bd2783-1068-49a8-b524-3ebddbf88199",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3.6_(Repleta)</td>\n",
       "      <td>variant</td>\n",
       "      <td>repleta group</td>\n",
       "      <td>32321</td>\n",
       "      <td>245</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "      <td>245</td>\n",
       "      <td>cenH3.6_(Repleta)  is a recently evolved centr...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "0  cenH3.6_(Repleta)  variant  repleta group             32321          245   \n",
       "\n",
       "                  parent   id  \\\n",
       "0  cenH3_(Drosophilidae)  245   \n",
       "\n",
       "                                             summary taxonomy genes  ...  \\\n",
       "0  cenH3.6_(Repleta)  is a recently evolved centr...     None  None  ...   \n",
       "\n",
       "  knock_out function sequence localization deposition structure interactions  \\\n",
       "0      None     None     None         None       None      None         None   \n",
       "\n",
       "  disease caveats relations  \n",
       "0    None    None      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3.6_(Repleta)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "e488f5e8-86ed-4cbd-a2f8-6f8aecc11c24",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4986f47c-c3aa-4960-96ce-c00873763e5a",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add new nodes cenH3.1_(Culicidae), cenH3.2_(Culicidae), cenH3.3_(Aedes) to cenH3_(Culicidae)</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "db2f790d-fad8-499f-9101-f2cc9526eb14",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "histone_nodes = [\n",
    "    \"cenH3.1_(Culicidae)\",\n",
    "    \"cenH3.2_(Culicidae)\",\n",
    "    \"cenH3.3_(Aedes)\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "eb238e8b-bb14-4c76-94a3-baeb4d9be47f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"cenH3.1_(Culicidae)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Culicidae\",\n",
    "    \"taxonomic_span_id\": \"7157\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cenH3_(Culicidae)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "data_histone = {\n",
    "    \"id\": \"cenH3.2_(Culicidae)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Culicidae\",\n",
    "    \"taxonomic_span_id\": \"7157\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cenH3_(Culicidae)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "data_histone = {\n",
    "    \"id\": \"cenH3.3_(Aedes)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Aedes\",\n",
    "    \"taxonomic_span_id\": \"7158\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cenH3_(Culicidae)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "12663920-fc0f-48d2-818d-b035c4803546",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>cenH3.1_(Culicidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Culicidae</td>\n",
       "      <td>7157</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cenH3_(Culicidae)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>cenH3.2_(Culicidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Culicidae</td>\n",
       "      <td>7157</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cenH3_(Culicidae)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>cenH3.3_(Aedes)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Aedes</td>\n",
       "      <td>7158</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cenH3_(Culicidae)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     id    level taxonomic_span taxonomic_span_id  \\\n",
       "5   cenH3.1_(Culicidae)  variant      Culicidae              7157   \n",
       "9   cenH3.2_(Culicidae)  variant      Culicidae              7157   \n",
       "13      cenH3.3_(Aedes)  variant          Aedes              7158   \n",
       "\n",
       "    description             parent  \n",
       "5           NaN  cenH3_(Culicidae)  \n",
       "9           NaN  cenH3_(Culicidae)  \n",
       "13          NaN  cenH3_(Culicidae)  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM histone\"\n",
    "cursor.execute(query)\n",
    "histone_df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "histone_df[histone_df[\"id\"].isin(histone_nodes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "684173e0-20ea-4218-950c-ab63c10d95bf",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e67f5b6-8110-4545-b429-dce6ebcd1d52",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add a summary description to node cenH3.1_(Culicidae)</span>\n",
    "\n",
    "```cenH3.1_(Culicidae) is a centromeric histone variant encoded by the mosqCid1 gene, found in most studied mosquito species (Anopheles, Aedes, Culex), with the exception of An. albimanus and An. darlingi where it has been lost [kursel_ancient_2020]. This gene emerged over 150 million years ago and has been retained in most species, acquiring specialized functions. Unlike its paralog mosqCid2, which shows high expression in ovaries and early embryos, mosqCid1 exhibits low expression levels in both somatic and germline tissues [kursel_ancient_2020]. Notably, mosqCid1 evolves under positive selection, suggesting its potential role in suppressing genetic conflicts such as centromere drive [kursel_ancient_2020]. Its N-terminal tail contains unique motifs absent in mosqCid2, likely reflecting differences in protein interactions.```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "281ee89c-6159-4b94-8163-410c54fda753",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3.1_(Culicidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Culicidae</td>\n",
       "      <td>7157</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Culicidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  cenH3.1_(Culicidae)  variant      Culicidae              7157        None   \n",
       "\n",
       "              parent    id summary taxonomy genes  ... knock_out function  \\\n",
       "0  cenH3_(Culicidae)  None    None     None  None  ...      None     None   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     None         None       None      None         None    None    None   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3.1_(Culicidae)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "45428d51-17d9-4d29-ba3f-f32c4dc07e98",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "histone_description_summary = \"cenH3.1_(Culicidae) is a centromeric histone variant encoded by the mosqCid1 gene, found in most studied mosquito species (Anopheles, Aedes, Culex), with the exception of An. albimanus and An. darlingi where it has been lost [kursel_ancient_2020]. This gene emerged over 150 million years ago and has been retained in most species, acquiring specialized functions. Unlike its paralog mosqCid2, which shows high expression in ovaries and early embryos, mosqCid1 exhibits low expression levels in both somatic and germline tissues [kursel_ancient_2020]. Notably, mosqCid1 evolves under positive selection, suggesting its potential role in suppressing genetic conflicts such as centromere drive [kursel_ancient_2020]. Its N-terminal tail contains unique motifs absent in mosqCid2, likely reflecting differences in protein interactions.\"\n",
    "data_histone_description = (histone_description_summary,) + (None,) * 13\n",
    "cursor.execute(add_histone_description, data_histone_description)\n",
    "\n",
    "histone_description_id = cursor.lastrowid\n",
    "query = f\"UPDATE histone SET description={histone_description_id} WHERE id='cenH3.1_(Culicidae)'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "cf60634f-8ec0-4889-ba1f-e89406724eb9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3.1_(Culicidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Culicidae</td>\n",
       "      <td>7157</td>\n",
       "      <td>246</td>\n",
       "      <td>cenH3_(Culicidae)</td>\n",
       "      <td>246</td>\n",
       "      <td>cenH3.1_(Culicidae) is a centromeric histone v...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "0  cenH3.1_(Culicidae)  variant      Culicidae              7157          246   \n",
       "\n",
       "              parent   id                                            summary  \\\n",
       "0  cenH3_(Culicidae)  246  cenH3.1_(Culicidae) is a centromeric histone v...   \n",
       "\n",
       "  taxonomy genes  ... knock_out function sequence localization deposition  \\\n",
       "0     None  None  ...      None     None     None         None       None   \n",
       "\n",
       "  structure interactions disease caveats relations  \n",
       "0      None         None    None    None      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3.1_(Culicidae)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "f6ab920a-c06e-482d-ad9c-755f066d9180",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5af7e2e5-c39e-45f4-9eba-fc49eb8d9c12",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add a summary description to node cenH3.2_(Culicidae)</span>\n",
    "\n",
    "```cenH3.2_(Culicidae) is the ancestral (emerged over 150 My) centromeric histone variant in mosquitoes, conserved across Anopheles, Aedes, and Culex species [kursel_ancient_2020]. Unlike its paralog mosqCid1, mosqCid2 specializes in germline processes, showing dramatic upregulation (6-10 fold) in ovaries after blood feeding and during early embryogenesis [kursel_ancient_2020]. It contains four unique N-terminal motifs absent in mosqCid1, reflecting functional divergence. Remarkably, mosqCid2 serves as the sole centromeric histone in An. albimanus and An. darlingi following mosqCid1 loss, demonstrating its fundamental role in chromosome segregation [kursel_ancient_2020].```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "d318b4f0-7191-4d8f-b9a3-e76aaea90d47",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3.2_(Culicidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Culicidae</td>\n",
       "      <td>7157</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Culicidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  cenH3.2_(Culicidae)  variant      Culicidae              7157        None   \n",
       "\n",
       "              parent    id summary taxonomy genes  ... knock_out function  \\\n",
       "0  cenH3_(Culicidae)  None    None     None  None  ...      None     None   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     None         None       None      None         None    None    None   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3.2_(Culicidae)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "9c1ad033-6593-4015-b3e1-f0f5dd937d61",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "histone_description_summary = \"cenH3.2_(Culicidae) is the ancestral (emerged over 150 My) centromeric histone variant in mosquitoes, conserved across Anopheles, Aedes, and Culex species [kursel_ancient_2020]. Unlike its paralog mosqCid1, mosqCid2 specializes in germline processes, showing dramatic upregulation (6-10 fold) in ovaries after blood feeding and during early embryogenesis [kursel_ancient_2020]. It contains four unique N-terminal motifs absent in mosqCid1, reflecting functional divergence. Remarkably, mosqCid2 serves as the sole centromeric histone in An. albimanus and An. darlingi following mosqCid1 loss, demonstrating its fundamental role in chromosome segregation [kursel_ancient_2020].\"\n",
    "data_histone_description = (histone_description_summary,) + (None,) * 13\n",
    "cursor.execute(add_histone_description, data_histone_description)\n",
    "\n",
    "histone_description_id = cursor.lastrowid\n",
    "query = f\"UPDATE histone SET description={histone_description_id} WHERE id='cenH3.2_(Culicidae)'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "f8fa6656-5eb7-47f7-9588-402fa54cc44d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3.2_(Culicidae)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Culicidae</td>\n",
       "      <td>7157</td>\n",
       "      <td>247</td>\n",
       "      <td>cenH3_(Culicidae)</td>\n",
       "      <td>247</td>\n",
       "      <td>cenH3.2_(Culicidae) is the ancestral (emerged ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "0  cenH3.2_(Culicidae)  variant      Culicidae              7157          247   \n",
       "\n",
       "              parent   id                                            summary  \\\n",
       "0  cenH3_(Culicidae)  247  cenH3.2_(Culicidae) is the ancestral (emerged ...   \n",
       "\n",
       "  taxonomy genes  ... knock_out function sequence localization deposition  \\\n",
       "0     None  None  ...      None     None     None         None       None   \n",
       "\n",
       "  structure interactions disease caveats relations  \n",
       "0      None         None    None    None      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3.2_(Culicidae)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "001caf39-f7d8-44d8-a755-409d452a82f8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "598b1ea4-20e3-42ef-9d5a-8ff7bdb930fc",
   "metadata": {},
   "source": [
    "# <span style=\"color:black\">Add a summary description to node cenH3.3_(Aedes)</span>\n",
    "\n",
    "```cenH3.3_(Aedes) is a recently evolved centromeric histone variant unique to Aedes mosquitoes, arising from a lineage-specific duplication of mosqCid1 [kursel_ancient_2020]. This paralog demonstrates specialized temporal expression during embryogenesis, suggesting a developmental-stage specific role in early development. Unlike the germline-focused mosqCid2 or the ubiquitously expressed mosqCid1, mosqCid3 shows no significant expression in adult tissues or ovaries, indicating its exclusive embryonic function [kursel_ancient_2020].```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "b9d64d51-9aef-4875-b825-b9f47c0c0991",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3.3_(Aedes)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Aedes</td>\n",
       "      <td>7158</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Culicidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  cenH3.3_(Aedes)  variant          Aedes              7158        None   \n",
       "\n",
       "              parent    id summary taxonomy genes  ... knock_out function  \\\n",
       "0  cenH3_(Culicidae)  None    None     None  None  ...      None     None   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     None         None       None      None         None    None    None   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3.3_(Aedes)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "a197514a-283c-466e-9291-943d12e84ced",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "histone_description_summary = \"cenH3.3_(Aedes) is a recently evolved centromeric histone variant unique to Aedes mosquitoes, arising from a lineage-specific duplication of mosqCid1 [kursel_ancient_2020]. This paralog demonstrates specialized temporal expression during embryogenesis, suggesting a developmental-stage specific role in early development. Unlike the germline-focused mosqCid2 or the ubiquitously expressed mosqCid1, mosqCid3 shows no significant expression in adult tissues or ovaries, indicating its exclusive embryonic function [kursel_ancient_2020].\"\n",
    "data_histone_description = (histone_description_summary,) + (None,) * 13\n",
    "cursor.execute(add_histone_description, data_histone_description)\n",
    "\n",
    "histone_description_id = cursor.lastrowid\n",
    "query = f\"UPDATE histone SET description={histone_description_id} WHERE id='cenH3.3_(Aedes)'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "3d7d9162-1591-482e-aab9-cbc5ee2f99c8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3.3_(Aedes)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Aedes</td>\n",
       "      <td>7158</td>\n",
       "      <td>248</td>\n",
       "      <td>cenH3_(Culicidae)</td>\n",
       "      <td>248</td>\n",
       "      <td>cenH3.3_(Aedes) is a recently evolved centrome...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "0  cenH3.3_(Aedes)  variant          Aedes              7158          248   \n",
       "\n",
       "              parent   id                                            summary  \\\n",
       "0  cenH3_(Culicidae)  248  cenH3.3_(Aedes) is a recently evolved centrome...   \n",
       "\n",
       "  taxonomy genes  ... knock_out function sequence localization deposition  \\\n",
       "0     None  None  ...      None     None     None         None       None   \n",
       "\n",
       "  structure interactions disease caveats relations  \n",
       "0      None         None    None    None      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3.3_(Aedes)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "5e3d8691-3fd6-4ba0-ae98-d6b61c0780a4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e8bad07-ef74-40fd-a93b-e7897df0c591",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Close connections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "e5673501-5845-491a-bbb9-f5d213721d58",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cursor.close()\n",
    "conn.close()\n",
    "tunnel.stop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1103e496-2fdb-4c8a-952f-b38cbdd615b1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".conda-histdb_env",
   "language": "python",
   "name": "conda-env-.conda-histdb_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
