{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "53d3ec55-7daf-4ade-8add-2830b709dc54",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import io\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import Entrez, SeqIO\n",
    "from mysql.connector import connection\n",
    "from sshtunnel import SSHTunnelForwarder\n",
    "\n",
    "# Укажите ваш email (обязательно для использования Entrez)\n",
    "Entrez.email = \"your.email@example.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "608c4bfe-b979-4b60-b2b5-b94eeb679cb5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open(\"db_curated_server_info.txt\", \"r\") as file:\n",
    "    lines = file.readlines()\n",
    "\n",
    "config = {}\n",
    "\n",
    "for line in lines:\n",
    "    line = line.strip()\n",
    "    if line and not line.startswith(\"#\"):\n",
    "        key, value = line.split(\"=\", 1)\n",
    "        config[key] = value.strip()\n",
    "\n",
    "server_name = config.get(\"server_name\")\n",
    "srever_port = int(config.get(\"srever_port\"))\n",
    "ssh_password = config.get(\"ssh_password\")\n",
    "ssh_username = config.get(\"ssh_username\")\n",
    "db_adress = config.get(\"db_adress\")\n",
    "db_port = int(config.get(\"db_port\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4e19aa1c-49a3-48c4-b67f-34c94cef608d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45073\n"
     ]
    }
   ],
   "source": [
    "tunnel = SSHTunnelForwarder(\n",
    "    (server_name, srever_port),\n",
    "    ssh_password=ssh_password,\n",
    "    ssh_username=ssh_username,\n",
    "    remote_bind_address=(db_adress, db_port),\n",
    ")\n",
    "tunnel.start()\n",
    "print(tunnel.local_bind_port)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cef2275b-12d0-48f1-ad3c-ce4cdb9aa267",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "conn = connection.MySQLConnection(\n",
    "    user=\"db_user\",\n",
    "    password=\"db_password\",\n",
    "    host=\"localhost\",\n",
    "    port=tunnel.local_bind_port,\n",
    "    database=\"db_name\",\n",
    ")\n",
    "cursor = conn.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f589ea8d-8b83-4331-a1cf-b64ed7621801",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('alternative_name',),\n",
       " ('histone',),\n",
       " ('histone_description',),\n",
       " ('histone_has_publication',),\n",
       " ('publication',),\n",
       " ('sequence',),\n",
       " ('sequence_has_publication',)]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SHOW TABLES;\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "b3dd9ef8-b162-4127-9535-ac7743af8c69",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# add_histone = (\n",
    "#     \"INSERT INTO histone \"\n",
    "#     \"(id, level, taxonomic_span, taxonomic_span_id, description, parent) \"\n",
    "#     \"VALUES (%(id)s, %(level)s, %(taxonomic_span)s, %(taxonomic_span_id)s, %(description)s, %(parent)s)\"\n",
    "# )\n",
    "# add_histone_description = (\n",
    "#     \"INSERT INTO histone_description \"\n",
    "#     \"(summary, taxonomy, genes, evolution, expression, knock_out, function, sequence, localization, deposition, structure, interactions, disease, caveats) \"\n",
    "#     \"VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)\"\n",
    "# )\n",
    "add_publication = (\n",
    "    \"INSERT INTO publication \"\n",
    "    \"(id, title, doi, author, year) \"\n",
    "    \"VALUES (%(id)s, %(title)s, %(doi)s, %(author)s, %(year)s)\"\n",
    ")\n",
    "add_sequence = (\n",
    "    \"INSERT INTO sequence \"\n",
    "    \"(accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration) \"\n",
    "    \"VALUES (%(accession)s, %(variant)s, %(gi)s, %(ncbi_gene_id)s, %(hgnc_gene_name)s, %(taxonomy_id)s, %(organism)s, %(phylum)s, %(class)s, %(taxonomy_group)s, %(info)s, %(sequence)s, %(variant_under_consideration)s)\"\n",
    ")\n",
    "add_sequence_has_publication = (\n",
    "    \"INSERT INTO sequence_has_publication \"\n",
    "    \"(sequence_accession, publication_id) \"\n",
    "    \"VALUES (%s, %s)\"\n",
    ")\n",
    "# add_alternate_names = (\n",
    "#     \"INSERT INTO alternative_name \"\n",
    "#     \"(name, taxonomy, gene, splice, histone) \"\n",
    "#     \"VALUES (%(name)s, %(taxonomy)s, %(gene)s, %(splice)s, %(histone)s)\"\n",
    "# )\n",
    "# add_histone_has_publication = (\n",
    "#     \"INSERT INTO histone_has_publication \"\n",
    "#     \"(histone_id, publication_id) \"\n",
    "#     \"VALUES (%s, %s)\"\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "f59dab92-6fd1-431d-9b47-a6c579454aaa",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_taxonomy_data(record):\n",
    "    import re\n",
    "    import sys\n",
    "\n",
    "    taxonomy_data = {}\n",
    "    taxonomy_data[\"organism\"] = record.annotations[\"organism\"]\n",
    "    try:\n",
    "        for a in record.features[0].qualifiers[\"db_xref\"]:\n",
    "            text = re.search(\"(\\S+):(\\S+)\", a).group(1)\n",
    "            taxid = re.search(\"(\\S+):(\\S+)\", a).group(2)\n",
    "            if text == \"taxon\":\n",
    "                print(f\"Fetched taxid from NCBI {taxid}\")\n",
    "                taxonomy_data[\"taxonomy_id\"] = int(taxid)\n",
    "            else:\n",
    "                continue\n",
    "    except:\n",
    "        print(\"!!!!!!Unable to get TAXID for this record setting it to 1\")\n",
    "        taxonomy_data[\"taxonomy_id\"] = 1  # unable to identify\n",
    "\n",
    "    lineage = dict()\n",
    "    for i in range(10):\n",
    "        try:\n",
    "            handle = Entrez.efetch(\n",
    "                id=taxonomy_data[\"taxonomy_id\"], db=\"taxonomy\", retmode=\"xml\"\n",
    "            )\n",
    "            tax_data = Entrez.read(handle)\n",
    "            lineage = {\n",
    "                d[\"Rank\"]: d[\"ScientificName\"]\n",
    "                for d in tax_data[0][\"LineageEx\"]\n",
    "                if d[\"Rank\"] in [\"class\", \"phylum\"]\n",
    "            }\n",
    "            break\n",
    "        except:\n",
    "            print(\n",
    "                \"Unexpected error: {}, Retrying, attempt {}\".format(\n",
    "                    sys.exc_info()[0], i\n",
    "                )\n",
    "            )\n",
    "            if i == 9:\n",
    "                print(\n",
    "                    f\"FATAL ERROR could not get class and phylum from NCBI after 10 attempts for taxid:{taxonomy_data['taxonomy_id']}. Will add None for class and phylum!\"\n",
    "                )\n",
    "            else:\n",
    "                continue\n",
    "    taxonomy_data[\"phylum\"] = lineage.get(\"phylum\", None)\n",
    "    taxonomy_data[\"class\"] = lineage.get(\"class\", None)\n",
    "    if taxonomy_data[\"phylum\"] is not None:\n",
    "        taxonomy_data[\"phylum\"] = str(taxonomy_data[\"phylum\"])\n",
    "    if taxonomy_data[\"class\"] is not None:\n",
    "        taxonomy_data[\"class\"] = str(taxonomy_data[\"class\"])\n",
    "    return taxonomy_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6011bb8a-bab5-4fb5-b4ae-15b83accd15b",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_histone_description = {\n",
    "    \"summary\": None,\n",
    "    \"taxonomy\": None,\n",
    "    \"genes\": None,\n",
    "    \"evolution\": None,\n",
    "    \"expression\": None,\n",
    "    \"knock_out\": None,\n",
    "    \"function\": None,\n",
    "    \"sequence\": None,\n",
    "    \"localization\": None,\n",
    "    \"deposition\": None,\n",
    "    \"structure\": None,\n",
    "    \"interactions\": None,\n",
    "    \"disease\": None,\n",
    "    \"caveats\": None,\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c248e27c",
   "metadata": {},
   "source": [
    "# To Do H2B\n",
    "\n",
    "## <span style=\"color:green\">Update description of H2B.V_(Trypanosoma)</span>\n",
    "\n",
    "### <span style=\"color:green\">Update expression</span>\n",
    "\n",
    "```H2B.V transcript levels peak during the S phase of the cell cycle, similarly to canonical histones, whereas protein abundance progressively accumulates as the cycle progresses. A significant increase in H2B.V levels is observed during the initial stages of metacyclogenesis, with maximum abundance detected in metacyclic trypomastigotes [roson_new_2025].```\n",
    "\n",
    "### <span style=\"color:green\">Update function</span>\n",
    "\n",
    "```The role of H2B.V appears to involve the regulation of parasite differentiation and virulence processes [roson_new_2025].```\n",
    "\n",
    "### <span style=\"color:green\">Update localization</span>\n",
    "\n",
    "```H2B.V, together with H4.V, marks new transcription initiation and termination sites. H2B.V localizes to divergent strand-switch regions, which demarcate transcription initiation sites, as well as near certain tRNA loci [roson_new_2025].```\n",
    "\n",
    "### <span style=\"color:green\">Update knock_out</span>\n",
    "\n",
    "```Parasites with heterozygous knockout of H2B.V exhibit increased differentiation into metacyclic forms and enhanced invasion of mammalian host cells [roson_new_2025].```\n",
    "\n",
    "### <span style=\"color:green\">Update interections</span>\n",
    "\n",
    "```H2B.V dimerizes with the histone variant H2A.Z. The H2B.V–H2A.Z dimers are more unstable compared to canonical H2B–H2A dimers, thereby promoting a more accessible chromatin configuration permissive for transcription initiation [roson_new_2025].```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f849284",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# To Do H3\n",
    "\n",
    "## <span style=\"color:green\">Update description of H3.V_(Trypanosoma)</span>\n",
    "\n",
    "### <span style=\"color:green\">Update summary</span>\n",
    "\n",
    "```H3.V is a histone variant characterized so far in some trypanosomatids. It shares ~45-60% sequence identity with canonical H3. H3.V is not essential for viability. Although H3.V associated with the mitotic spindle and enriched at telomeres it is not essential for mini-chromosome segregation, telomere maintenance or transcriptional silencing at the telomere-proximal expression sites [lowell_variant_2004, anderson_kinetoplastid-specific_2013]. While the general structure of H3.V resembles that of CenH3, it does not possess the specific sequence features required to be classified as a CenH3 variant [lowell_variant_2004].```\n",
    "\n",
    "### <span style=\"color:green\">Update function</span>\n",
    "\n",
    "```H3.V plays a critical role in the regulation of antigenic variation. Together with H4.V they are involved in the positioning and expression of variant surface glycoprotein (VSG) genes, which the parasite Trypanosoma brucei switches to avoid immune detection [muller_genome_2018, schulz_base_2016]. H3.V is important for transcription termination and expression of downstream genes. The trypanosomatid-specific hyper-modified DNA base J and H3.V collaboratively influence the termination process resulting in read-through transcription and increased expression of downstream genes [siegel_four_2009, reynolds_histone_2016, schulz_base_2016, mcdonald_localization_2022].```\n",
    "\n",
    "### <span style=\"color:green\">Update localization</span>\n",
    "\n",
    "```In Trypanosoma brucei, H3.V, together with H4.V, is enriched at transcription termination sites and in telomeric regions [anderson_kinetoplastid-specific_2013, roson_new_2025].```\n",
    "\n",
    "### <span style=\"color:green\">Update knock_out</span>\n",
    "\n",
    "```While H3.V is dispensable for parasite viability, its deletion impairs variant surface glycoprotein (VSG) silencing. However, simultaneous deletion of H3.V, H4.V, and DNA base J leads to critical alterations in cell growth and replication [roson_new_2025].```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33bdc63d-8fce-407f-b649-528873a7cfde",
   "metadata": {},
   "source": [
    "# To Do H1\n",
    "\n",
    "## <span style=\"color:green\">Update description of cH1.5</span>\n",
    "\n",
    "### <span style=\"color:green\">Update summary</span>\n",
    "\n",
    "```cH1.5 is a replication dependent linker histone located within large histone gene clusters in mammals. cH1.5 plays a critical role in chromatin organization, thereby influencing cellular development and homeostasis [behrends_linker_2020].```\n",
    "\n",
    "### <span style=\"color:green\">Update expression</span>\n",
    "\n",
    "```cH1.5 is present across various cell types. As demonstrated on human cells, its expression level changes during differentiation: for instance, it decreases during the differentiation of NT2 cells into neural cells but increases during the dedifferentiation of keratinocytes into induced pluripotent stem cells [behrends_linker_2020].```\n",
    "\n",
    "## <span style=\"color:green\">Update description of cH1.5_(Homo_sapiens)</span>\n",
    "\n",
    "### <span style=\"color:green\">Update summary</span>\n",
    "\n",
    "```cH1.5_(Homo_sapiens) is a replication dependent linker histone in human. сH1.5 plays a critical role in maintaining centromere integrity and is essential for sustaining appropriate transcriptional programs in differentiated cell types in human cells [saha_linker_2025, li_dynamic_2012].```\n",
    "\n",
    "### <span style=\"color:green\">Update expression</span>\n",
    "\n",
    "```cH1.5 is present across various human cell types. Notably, it serves as the predominant H1 variant in astrocyte-lineage SVGp12 cells and in glioblastoma multiforme cells [saha_linker_2025]. cH1.5 expression is downregulated during cellular differentiation. Compared to normal differentiated somatic cells, H1.5 exhibits higher expression levels in both pluripotent cells and fibroblasts [li_dynamic_2012].```\n",
    "\n",
    "### <span style=\"color:green\">Update function</span>\n",
    "\n",
    "```cH1.5 is essential for the regulation of mitotic integrity and for sustaining appropriate transcriptional programs in differentiated cell types [saha_linker_2025, li_dynamic_2012].```\n",
    "\n",
    "### <span style=\"color:green\">Update localization</span>\n",
    "\n",
    "```In differentiated cells, cH1.5 forms enrichment blocks in both genic and intergenic regions, preferentially associating with genes encoding membrane and membrane-associated proteins [li_dynamic_2012]. Moreover, cH1.5 is enriched at centromeric regions and co‑localizes with the centromeric histone CENP‑A in human, as confirmed by immunofluorescence and chromatin immunoprecipitation assays [saha_linker_2025]. cH1.5 binding is also associated with gene repression, chromatin compaction and promoting the inclusion of alternatively spliced exons into the mature mRNA [li_dynamic_2012, behrends_linker_2020].```\n",
    "\n",
    ">It is also enriched around splicing sites, particularly on genes with alternative splicing, where its binding promotes the inclusion of alternatively spliced exons into the mature mRNA (see behrends_linker_2020).\n",
    "\n",
    "### <span style=\"color:green\">Update knock_out</span>\n",
    "\n",
    "```cH1.5 depletion leads to decreased cell growth, increased chromatin accessibility and deregulation of gene expression [li_dynamic_2012, behrends_linker_2020]. Additionally, knockdown of cH1.5 results in the loss of centromeric α‑satellite DNA transcription, a reduction in de novo CENP‑A loading, and the accumulation of mitotic defects [saha_linker_2025].```\n",
    "\n",
    "### <span style=\"color:green\">Update structure</span>\n",
    "\n",
    "```cH1.5 binds to CENP‑A nucleosomes in a non‑canonical, dyad‑independent manner [saha_linker_2025].```\n",
    "\n",
    "### <span style=\"color:green\">Update interections</span>\n",
    "\n",
    "```cH1.5 directly interacts with CENP‑A‑containing nucleosomes both in vitro and in vivo [saha_linker_2025].```\n",
    "\n",
    "### <span style=\"color:green\">Update disease</span>\n",
    "\n",
    "```cH1.5 can serve as a valuable prognostic marker in cancer patients. Its expression level is decreased in many cancer cell lines, while being highly overexpressed in glioblastoma multiforme [saha_linker_2025, behrends_linker_2020, li_dynamic_2012]. Furthermore, cH1.5 mutations are associated with follicular lymphoma and altered chromatin states. Impairment of cH1.5 function may contribute to chromosomal instability and the development of malignancies, particularly in brain cells[saha_linker_2025].```\n",
    "\n",
    ">human H1.5 bind right on the center of the dyad symmetrically [wu_binding_2021, bednar_structure_2017].\n",
    "\n",
    "## <span style=\"color:green\">Update description of cH1.1</span>\n",
    "\n",
    "### <span style=\"color:green\">Update summary</span>\n",
    "\n",
    "```cH1.1 is a replication dependent linker histone located within large histone gene clusters in mammals. cH1.1 plays an important role in the early stages of embryogenesis [funaya_involvement_2023].```\n",
    "\n",
    "### <span style=\"color:green\">Update function</span>\n",
    "\n",
    "```cH1.1 plays a critical role in the early stages of embryogenesis. cH1.1 is absent in human differentiated keratinocytes and fibroblasts but is re-expressed upon reprogramming, suggesting that it is specific to pluripotent cells [terme_histone_2011]. In mice, it is essential for preimplantation development, particularly at the morula and blastocyst stages [funaya_involvement_2023].```\n",
    "\n",
    "### <span style=\"color:green\">Update expression</span>\n",
    "\n",
    "```cH1.1 expression predominates in early embryonic cells compared to other replication-dependent linker histone variants. In mice, cH1.1—alongside H1.8—exhibits the highest expression levels at the one- and two-cell stages of embryonic development [funaya_involvement_2023, funaya_linker_2018]. Although its expression declines by the four-cell stage, it remains elevated through the blastocyst stage. In human cells, сH1.1 has also been shown to be highly expressed in pluripotent cells—including embryonic and induced pluripotent stem cells—and absent in differentiated keratinocytes and fibroblasts [terme_histone_2011].```\n",
    "\n",
    "### <span style=\"color:green\">Update localization</span>\n",
    "\n",
    "```In mice, cH1.1 is associated with actively transcribed genes, whereas the somatic linker histone variants сH1.2–5 are excluded from genomic regions occupied by H1a, indicating that H1a exhibits a distinct and non-overlapping genomic localization [funaya_involvement_2023].```\n",
    "\n",
    "### <span style=\"color:green\">Update knock_out</span>\n",
    "\n",
    "```Maternal knockout of the cH1.1 gene in mice results in reduced litter size and impaired embryonic development during the morula-to-blastocyst transition [funaya_involvement_2023].```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "915b4727-c732-4acd-b106-b2dc86c99665",
   "metadata": {},
   "source": [
    "# To Do H3\n",
    "\n",
    "## <span style=\"color:green\">Add sequences to H3-like_(Plants)</span>\n",
    "\n",
    "Accessions: NP_172794.1 (HTR6), NP_201338.1 (HTR11),  NP_196795.1 (HTR15) [alvarez-venegas_canonical_2019, okada_analysis_2005]\n",
    "\n",
    "## <span style=\"color:green\">Add sequences to H3-like_(Plants)</span>\n",
    "\n",
    "Accessions: ABA97899.1 (HTR714), ABA97902.1 (HTR715) [alvarez-venegas_canonical_2019, hu_identification_2015]\n",
    "\n",
    "## <span style=\"color:green\">Add sequences to H3-like_(Plants)</span>\n",
    "\n",
    "Accessions: BAE48427.1 (gcH3), BAA96098.1 (gH3), BAE48431.1 (leH3), BAE48433.1 (soH3-1), BAE48435.1 (soH3-2) [alvarez-venegas_canonical_2019, probst_similar_2020]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d8707d5-5ef0-4645-a0ba-3ecfd864cdac",
   "metadata": {},
   "source": [
    "# DONE H2B\n",
    "\n",
    "## <span style=\"color:black\">Update description of H2B.V_(Trypanosoma)</span>\n",
    "\n",
    "### <span style=\"color:black\">Update expression</span>\n",
    "\n",
    "```H2B.V transcript levels peak during the S phase of the cell cycle, similarly to canonical histones, whereas protein abundance progressively accumulates as the cycle progresses. A significant increase in H2B.V levels is observed during the initial stages of metacyclogenesis, with maximum abundance detected in metacyclic trypomastigotes [roson_new_2025].```\n",
    "\n",
    "### <span style=\"color:black\">Update function</span>\n",
    "\n",
    "```The role of H2B.V appears to involve the regulation of parasite differentiation and virulence processes [roson_new_2025].```\n",
    "\n",
    "### <span style=\"color:black\">Update localization</span>\n",
    "\n",
    "```H2B.V, together with H4.V, marks new transcription initiation and termination sites. H2B.V localizes to divergent strand-switch regions, which demarcate transcription initiation sites, as well as near certain tRNA loci [roson_new_2025].```\n",
    "\n",
    "### <span style=\"color:black\">Update knock_out</span>\n",
    "\n",
    "```Parasites with heterozygous knockout of H2B.V exhibit increased differentiation into metacyclic forms and enhanced invasion of mammalian host cells [roson_new_2025].```\n",
    "\n",
    "### <span style=\"color:black\">Update interections</span>\n",
    "\n",
    "```H2B.V dimerizes with the histone variant H2A.Z. The H2B.V–H2A.Z dimers are more unstable compared to canonical H2B–H2A dimers, thereby promoting a more accessible chromatin configuration permissive for transcription initiation [roson_new_2025].```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "16177120-e1fa-4ae7-9302-033f610118bf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': {0: 41},\n",
       " 'level': {0: 'variant_group'},\n",
       " 'taxonomic_span': {0: 'Trypanosoma'},\n",
       " 'taxonomic_span_id': {0: '5690'},\n",
       " 'description': {0: 41},\n",
       " 'parent': {0: 'H2B'},\n",
       " 'summary': {0: 'H2B.V_(Trypanosoma) is a histone variant characterized so far in Trypanosoma brucei. It shares ~38% sequence identity with major H2B. H2B.V is essential for viability. H2A.Z and H2B.V colocalize throughout the cell cycle and exhibit nearly identical genomic distribution. Data strongly suggest that H2A.Z and H2B.V function together within a single nucleosome [lowell_histone_2005]. H2BV possibly regulates H3 K4 and K76 trimethylation in Trypanosoma brucei [mandava_trypanosome_2008].'},\n",
       " 'taxonomy': {0: 'Trypanosoma'},\n",
       " 'genes': {0: 'null'},\n",
       " 'evolution': {0: 'null'},\n",
       " 'expression': {0: 'null'},\n",
       " 'knock_out': {0: 'null'},\n",
       " 'function': {0: 'null'},\n",
       " 'sequence': {0: 'null'},\n",
       " 'localization': {0: 'null'},\n",
       " 'deposition': {0: 'null'},\n",
       " 'structure': {0: 'null'},\n",
       " 'interactions': {0: 'null'},\n",
       " 'disease': {0: 'null'},\n",
       " 'caveats': {0: 'null'},\n",
       " 'relations': {0: None}}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='H2B.V_(Trypanosoma)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "99aeec3e-f095-420f-8456-aa507db707d3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UPDATE histone_description SET expression='H2B.V transcript levels peak during the S phase of the cell cycle, similarly to canonical histones, whereas protein abundance progressively accumulates as the cycle progresses. A significant increase in H2B.V levels is observed during the initial stages of metacyclogenesis, with maximum abundance detected in metacyclic trypomastigotes [roson_new_2025].', knock_out='Parasites with heterozygous knockout of H2B.V exhibit increased differentiation into metacyclic forms and enhanced invasion of mammalian host cells [roson_new_2025].', function='The role of H2B.V appears to involve the regulation of parasite differentiation and virulence processes [roson_new_2025].', localization='H2B.V, together with H4.V, marks new transcription initiation and termination sites. H2B.V localizes to divergent strand-switch regions, which demarcate transcription initiation sites, as well as near certain tRNA loci [roson_new_2025].', interactions='H2B.V dimerizes with the histone variant H2A.Z. The H2B.V–H2A.Z dimers are more unstable compared to canonical H2B–H2A dimers, thereby promoting a more accessible chromatin configuration permissive for transcription initiation [roson_new_2025].' WHERE id=41\n"
     ]
    }
   ],
   "source": [
    "expression_desc = \"H2B.V transcript levels peak during the S phase of the cell cycle, similarly to canonical histones, whereas protein abundance progressively accumulates as the cycle progresses. A significant increase in H2B.V levels is observed during the initial stages of metacyclogenesis, with maximum abundance detected in metacyclic trypomastigotes [roson_new_2025].\"\n",
    "knock_out_desc = \"Parasites with heterozygous knockout of H2B.V exhibit increased differentiation into metacyclic forms and enhanced invasion of mammalian host cells [roson_new_2025].\"\n",
    "function_desc = \"The role of H2B.V appears to involve the regulation of parasite differentiation and virulence processes [roson_new_2025].\"\n",
    "localization_desc = \"H2B.V, together with H4.V, marks new transcription initiation and termination sites. H2B.V localizes to divergent strand-switch regions, which demarcate transcription initiation sites, as well as near certain tRNA loci [roson_new_2025].\"\n",
    "interactions_desc = \"H2B.V dimerizes with the histone variant H2A.Z. The H2B.V–H2A.Z dimers are more unstable compared to canonical H2B–H2A dimers, thereby promoting a more accessible chromatin configuration permissive for transcription initiation [roson_new_2025].\"\n",
    "query = f\"UPDATE histone_description SET expression='{expression_desc}', knock_out='{knock_out_desc}', function='{function_desc}', localization='{localization_desc}', interactions='{interactions_desc}' WHERE id=41\"\n",
    "print(query)\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ff458561-01b1-492a-9b30-a0bdb2ef199a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': {0: 41},\n",
       " 'level': {0: 'variant_group'},\n",
       " 'taxonomic_span': {0: 'Trypanosoma'},\n",
       " 'taxonomic_span_id': {0: '5690'},\n",
       " 'description': {0: 41},\n",
       " 'parent': {0: 'H2B'},\n",
       " 'summary': {0: 'H2B.V_(Trypanosoma) is a histone variant characterized so far in Trypanosoma brucei. It shares ~38% sequence identity with major H2B. H2B.V is essential for viability. H2A.Z and H2B.V colocalize throughout the cell cycle and exhibit nearly identical genomic distribution. Data strongly suggest that H2A.Z and H2B.V function together within a single nucleosome [lowell_histone_2005]. H2BV possibly regulates H3 K4 and K76 trimethylation in Trypanosoma brucei [mandava_trypanosome_2008].'},\n",
       " 'taxonomy': {0: 'Trypanosoma'},\n",
       " 'genes': {0: 'null'},\n",
       " 'evolution': {0: 'null'},\n",
       " 'expression': {0: 'H2B.V transcript levels peak during the S phase of the cell cycle, similarly to canonical histones, whereas protein abundance progressively accumulates as the cycle progresses. A significant increase in H2B.V levels is observed during the initial stages of metacyclogenesis, with maximum abundance detected in metacyclic trypomastigotes [roson_new_2025].'},\n",
       " 'knock_out': {0: 'Parasites with heterozygous knockout of H2B.V exhibit increased differentiation into metacyclic forms and enhanced invasion of mammalian host cells [roson_new_2025].'},\n",
       " 'function': {0: 'The role of H2B.V appears to involve the regulation of parasite differentiation and virulence processes [roson_new_2025].'},\n",
       " 'sequence': {0: 'null'},\n",
       " 'localization': {0: 'H2B.V, together with H4.V, marks new transcription initiation and termination sites. H2B.V localizes to divergent strand-switch regions, which demarcate transcription initiation sites, as well as near certain tRNA loci [roson_new_2025].'},\n",
       " 'deposition': {0: 'null'},\n",
       " 'structure': {0: 'null'},\n",
       " 'interactions': {0: 'H2B.V dimerizes with the histone variant H2A.Z. The H2B.V–H2A.Z dimers are more unstable compared to canonical H2B–H2A dimers, thereby promoting a more accessible chromatin configuration permissive for transcription initiation [roson_new_2025].'},\n",
       " 'disease': {0: 'null'},\n",
       " 'caveats': {0: 'null'},\n",
       " 'relations': {0: None}}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='H2B.V_(Trypanosoma)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3c67a43e-6afe-42d0-815c-f1b8134a3c43",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6737cd3-ca30-4fac-b1be-ebca85e8b7e6",
   "metadata": {},
   "source": [
    "# DONE H3\n",
    "\n",
    "## <span style=\"color:black\">Update description of H3.V_(Trypanosoma)</span>\n",
    "\n",
    "### <span style=\"color:black\">Update summary</span>\n",
    "\n",
    "```H3.V is a histone variant characterized so far in some trypanosomatids. It shares ~45-60% sequence identity with canonical H3. H3.V is not essential for viability. Although H3.V associated with the mitotic spindle and enriched at telomeres it is not essential for mini-chromosome segregation, telomere maintenance or transcriptional silencing at the telomere-proximal expression sites [lowell_variant_2004, anderson_kinetoplastid-specific_2013]. While the general structure of H3.V resembles that of CenH3, it does not possess the specific sequence features required to be classified as a CenH3 variant [lowell_variant_2004].```\n",
    "\n",
    "### <span style=\"color:black\">Update function</span>\n",
    "\n",
    "```H3.V plays a critical role in the regulation of antigenic variation. Together with H4.V they are involved in the positioning and expression of variant surface glycoprotein (VSG) genes, which the parasite Trypanosoma brucei switches to avoid immune detection [muller_genome_2018, schulz_base_2016]. H3.V is important for transcription termination and expression of downstream genes. The trypanosomatid-specific hyper-modified DNA base J and H3.V collaboratively influence the termination process resulting in read-through transcription and increased expression of downstream genes [siegel_four_2009, reynolds_histone_2016, schulz_base_2016, mcdonald_localization_2022].```\n",
    "\n",
    "### <span style=\"color:black\">Update localization</span>\n",
    "\n",
    "```In Trypanosoma brucei, H3.V, together with H4.V, is enriched at transcription termination sites and in telomeric regions [anderson_kinetoplastid-specific_2013, roson_new_2025].```\n",
    "\n",
    "### <span style=\"color:black\">Update knock_out</span>\n",
    "\n",
    "```While H3.V is dispensable for parasite viability, its deletion impairs variant surface glycoprotein (VSG) silencing. However, simultaneous deletion of H3.V, H4.V, and DNA base J leads to critical alterations in cell growth and replication [roson_new_2025].```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5b980a6c-46f6-4814-8d6d-601caafe4257",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H3.V_(Trypanosomatidae)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Trypanosomes</td>\n",
       "      <td>93954</td>\n",
       "      <td>205</td>\n",
       "      <td>H3</td>\n",
       "      <td>205</td>\n",
       "      <td>H3.V is a histone variant characterized so far...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        id          level taxonomic_span taxonomic_span_id  \\\n",
       "0  H3.V_(Trypanosomatidae)  variant_group   Trypanosomes             93954   \n",
       "\n",
       "   description parent   id                                            summary  \\\n",
       "0          205     H3  205  H3.V is a histone variant characterized so far...   \n",
       "\n",
       "  taxonomy genes  ... knock_out function sequence localization deposition  \\\n",
       "0     None  None  ...      None     None     None         None       None   \n",
       "\n",
       "  structure interactions disease caveats relations  \n",
       "0      None         None    None    None      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='H3.V_(Trypanosomatidae)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "07de79ff-b979-4005-bb41-a4416f0eeeb1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': {0: 205},\n",
       " 'level': {0: 'variant_group'},\n",
       " 'taxonomic_span': {0: 'Trypanosomes'},\n",
       " 'taxonomic_span_id': {0: '93954'},\n",
       " 'description': {0: 205},\n",
       " 'parent': {0: 'H3'},\n",
       " 'summary': {0: 'H3.V is a histone variant characterized so far in some trypanosomatids. It shares ~45-60% sequence identity with canonical H3. H3.V is not essential for viability. Although H3.V associated with the mitotic spindle and enriched at telomeres it is not essential for mini-chromosome segregation, telomere maintenance or transcriptional silencing at the telomere-proximal expression sites [lowell_variant_2004,  anderson_kinetoplastid-specific_2013]. While the general structure of H3.V resembles that of CenH3, it does not possess the specific sequence features required to be classified as a CenH3 variant [lowell_variant_2004]. H3.V plays a critical role in the regulation of antigenic variation. Together with H4.V they are involved in the positioning and expression of variant surface glycoprotein (VSG) genes, which the parasite Trypanosoma brucei switches to avoid immune detection [muller_genome_2018, schulz_base_2016]. H3.V is important for transcription termination and expression of downstream genes. The trypanosomatid-specific hyper-modified DNA base J  and H3.V collaboratively influence the termination process resulting in read-through transcription and increased expression of downstream genes [siegel_four_2009, reynolds_histone_2016, schulz_base_2016, mcdonald_localization_2022].'},\n",
       " 'taxonomy': {0: None},\n",
       " 'genes': {0: None},\n",
       " 'evolution': {0: None},\n",
       " 'expression': {0: None},\n",
       " 'knock_out': {0: None},\n",
       " 'function': {0: None},\n",
       " 'sequence': {0: None},\n",
       " 'localization': {0: None},\n",
       " 'deposition': {0: None},\n",
       " 'structure': {0: None},\n",
       " 'interactions': {0: None},\n",
       " 'disease': {0: None},\n",
       " 'caveats': {0: None},\n",
       " 'relations': {0: None}}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ed788acb-32d3-4d21-b5e6-336d8047932b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UPDATE histone_description SET summary='H3.V is a histone variant characterized so far in some trypanosomatids. It shares ~45-60% sequence identity with canonical H3. H3.V is not essential for viability. Although H3.V associated with the mitotic spindle and enriched at telomeres it is not essential for mini-chromosome segregation, telomere maintenance or transcriptional silencing at the telomere-proximal expression sites [lowell_variant_2004, anderson_kinetoplastid-specific_2013]. While the general structure of H3.V resembles that of CenH3, it does not possess the specific sequence features required to be classified as a CenH3 variant [lowell_variant_2004].', knock_out='While H3.V is dispensable for parasite viability, its deletion impairs variant surface glycoprotein (VSG) silencing. However, simultaneous deletion of H3.V, H4.V, and DNA base J leads to critical alterations in cell growth and replication [roson_new_2025].', function='H3.V plays a critical role in the regulation of antigenic variation. Together with H4.V they are involved in the positioning and expression of variant surface glycoprotein (VSG) genes, which the parasite Trypanosoma brucei switches to avoid immune detection [muller_genome_2018, schulz_base_2016]. H3.V is important for transcription termination and expression of downstream genes. The trypanosomatid-specific hyper-modified DNA base J and H3.V collaboratively influence the termination process resulting in read-through transcription and increased expression of downstream genes [siegel_four_2009, reynolds_histone_2016, schulz_base_2016, mcdonald_localization_2022].', localization='In Trypanosoma brucei, H3.V, together with H4.V, is enriched at transcription termination sites and in telomeric regions [anderson_kinetoplastid-specific_2013, roson_new_2025].' WHERE id=205\n"
     ]
    }
   ],
   "source": [
    "summary_desc = \"H3.V is a histone variant characterized so far in some trypanosomatids. It shares ~45-60% sequence identity with canonical H3. H3.V is not essential for viability. Although H3.V associated with the mitotic spindle and enriched at telomeres it is not essential for mini-chromosome segregation, telomere maintenance or transcriptional silencing at the telomere-proximal expression sites [lowell_variant_2004, anderson_kinetoplastid-specific_2013]. While the general structure of H3.V resembles that of CenH3, it does not possess the specific sequence features required to be classified as a CenH3 variant [lowell_variant_2004].\"\n",
    "knock_out_desc = \"While H3.V is dispensable for parasite viability, its deletion impairs variant surface glycoprotein (VSG) silencing. However, simultaneous deletion of H3.V, H4.V, and DNA base J leads to critical alterations in cell growth and replication [roson_new_2025].\"\n",
    "function_desc = \"H3.V plays a critical role in the regulation of antigenic variation. Together with H4.V they are involved in the positioning and expression of variant surface glycoprotein (VSG) genes, which the parasite Trypanosoma brucei switches to avoid immune detection [muller_genome_2018, schulz_base_2016]. H3.V is important for transcription termination and expression of downstream genes. The trypanosomatid-specific hyper-modified DNA base J and H3.V collaboratively influence the termination process resulting in read-through transcription and increased expression of downstream genes [siegel_four_2009, reynolds_histone_2016, schulz_base_2016, mcdonald_localization_2022].\"\n",
    "localization_desc = \"In Trypanosoma brucei, H3.V, together with H4.V, is enriched at transcription termination sites and in telomeric regions [anderson_kinetoplastid-specific_2013, roson_new_2025].\"\n",
    "query = f\"UPDATE histone_description SET summary='{summary_desc}', knock_out='{knock_out_desc}', function='{function_desc}', localization='{localization_desc}' WHERE id=205\"\n",
    "print(query)\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "74e47e70-34a1-42f4-9f7a-c834ba086dbe",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': {0: 205},\n",
       " 'level': {0: 'variant_group'},\n",
       " 'taxonomic_span': {0: 'Trypanosomes'},\n",
       " 'taxonomic_span_id': {0: '93954'},\n",
       " 'description': {0: 205},\n",
       " 'parent': {0: 'H3'},\n",
       " 'summary': {0: 'H3.V is a histone variant characterized so far in some trypanosomatids. It shares ~45-60% sequence identity with canonical H3. H3.V is not essential for viability. Although H3.V associated with the mitotic spindle and enriched at telomeres it is not essential for mini-chromosome segregation, telomere maintenance or transcriptional silencing at the telomere-proximal expression sites [lowell_variant_2004, anderson_kinetoplastid-specific_2013]. While the general structure of H3.V resembles that of CenH3, it does not possess the specific sequence features required to be classified as a CenH3 variant [lowell_variant_2004].'},\n",
       " 'taxonomy': {0: None},\n",
       " 'genes': {0: None},\n",
       " 'evolution': {0: None},\n",
       " 'expression': {0: None},\n",
       " 'knock_out': {0: 'While H3.V is dispensable for parasite viability, its deletion impairs variant surface glycoprotein (VSG) silencing. However, simultaneous deletion of H3.V, H4.V, and DNA base J leads to critical alterations in cell growth and replication [roson_new_2025].'},\n",
       " 'function': {0: 'H3.V plays a critical role in the regulation of antigenic variation. Together with H4.V they are involved in the positioning and expression of variant surface glycoprotein (VSG) genes, which the parasite Trypanosoma brucei switches to avoid immune detection [muller_genome_2018, schulz_base_2016]. H3.V is important for transcription termination and expression of downstream genes. The trypanosomatid-specific hyper-modified DNA base J and H3.V collaboratively influence the termination process resulting in read-through transcription and increased expression of downstream genes [siegel_four_2009, reynolds_histone_2016, schulz_base_2016, mcdonald_localization_2022].'},\n",
       " 'sequence': {0: None},\n",
       " 'localization': {0: 'In Trypanosoma brucei, H3.V, together with H4.V, is enriched at transcription termination sites and in telomeric regions [anderson_kinetoplastid-specific_2013, roson_new_2025].'},\n",
       " 'deposition': {0: None},\n",
       " 'structure': {0: None},\n",
       " 'interactions': {0: None},\n",
       " 'disease': {0: None},\n",
       " 'caveats': {0: None},\n",
       " 'relations': {0: None}}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='H3.V_(Trypanosomatidae)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4b991f2a-a0e2-48e9-8321-7db18804d65d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3bb3f666-1424-4f9d-acdf-afa254db8c42",
   "metadata": {},
   "source": [
    "# DONE H1\n",
    "\n",
    "## <span style=\"color:black\">Update description of cH1.5</span>\n",
    "\n",
    "### <span style=\"color:black\">Update summary</span>\n",
    "\n",
    "```cH1.5 is a replication dependent linker histone located within large histone gene clusters in mammals. cH1.5 plays a critical role in chromatin organization, thereby influencing cellular development and homeostasis [behrends_linker_2020].```\n",
    "\n",
    "### <span style=\"color:black\">Update expression</span>\n",
    "\n",
    "```cH1.5 is present across various cell types. As demonstrated on human cells, its expression level changes during differentiation: for instance, it decreases during the differentiation of NT2 cells into neural cells but increases during the dedifferentiation of keratinocytes into induced pluripotent stem cells [behrends_linker_2020].```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "60517148-2639-41b4-9d78-29e9bf49c1e9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH1.5</td>\n",
       "      <td>variant</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>16</td>\n",
       "      <td>cH1</td>\n",
       "      <td>16</td>\n",
       "      <td>cH1.5 is a replication dependent linker histon...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      id    level taxonomic_span taxonomic_span_id  description parent  id  \\\n",
       "0  cH1.5  variant       Mammalia             40674           16    cH1  16   \n",
       "\n",
       "                                             summary taxonomy genes  ...  \\\n",
       "0  cH1.5 is a replication dependent linker histon...     null  null  ...   \n",
       "\n",
       "  knock_out function sequence localization deposition structure interactions  \\\n",
       "0      null     null     null         null       null      null         null   \n",
       "\n",
       "  disease caveats relations  \n",
       "0    null    null      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cH1.5'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "c2b37fa0-2b53-4ff1-9dc9-2d346b3f4f72",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': {0: 16},\n",
       " 'level': {0: 'variant'},\n",
       " 'taxonomic_span': {0: 'Mammalia'},\n",
       " 'taxonomic_span_id': {0: '40674'},\n",
       " 'description': {0: 16},\n",
       " 'parent': {0: 'cH1'},\n",
       " 'summary': {0: 'cH1.5 is a replication dependent linker histone located within large histone gene clusters in mammals [happel_histone_2009, talbert_histone_2021].'},\n",
       " 'taxonomy': {0: 'null'},\n",
       " 'genes': {0: 'null'},\n",
       " 'evolution': {0: 'null'},\n",
       " 'expression': {0: 'null'},\n",
       " 'knock_out': {0: 'null'},\n",
       " 'function': {0: 'null'},\n",
       " 'sequence': {0: 'null'},\n",
       " 'localization': {0: 'null'},\n",
       " 'deposition': {0: 'null'},\n",
       " 'structure': {0: 'null'},\n",
       " 'interactions': {0: 'null'},\n",
       " 'disease': {0: 'null'},\n",
       " 'caveats': {0: 'null'},\n",
       " 'relations': {0: None}}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e1c6f973-4db4-421b-b085-f44352b61ba9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UPDATE histone_description SET summary='cH1.5 is a replication dependent linker histone located within large histone gene clusters in mammals. cH1.5 plays a critical role in chromatin organization, thereby influencing cellular development and homeostasis [behrends_linker_2020].', expression='cH1.5 is present across various cell types. As demonstrated on human cells, its expression level changes during differentiation: for instance, it decreases during the differentiation of NT2 cells into neural cells but increases during the dedifferentiation of keratinocytes into induced pluripotent stem cells [behrends_linker_2020].' WHERE id=16\n"
     ]
    }
   ],
   "source": [
    "summary_desc = \"cH1.5 is a replication dependent linker histone located within large histone gene clusters in mammals. cH1.5 plays a critical role in chromatin organization, thereby influencing cellular development and homeostasis [behrends_linker_2020].\"\n",
    "expression_desc = \"cH1.5 is present across various cell types. As demonstrated on human cells, its expression level changes during differentiation: for instance, it decreases during the differentiation of NT2 cells into neural cells but increases during the dedifferentiation of keratinocytes into induced pluripotent stem cells [behrends_linker_2020].\"\n",
    "query = f\"UPDATE histone_description SET summary='{summary_desc}', expression='{expression_desc}' WHERE id=16\"\n",
    "print(query)\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a5c4d37c-e6a8-4c05-b0f2-c468832ab1f2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': {0: 16},\n",
       " 'level': {0: 'variant'},\n",
       " 'taxonomic_span': {0: 'Mammalia'},\n",
       " 'taxonomic_span_id': {0: '40674'},\n",
       " 'description': {0: 16},\n",
       " 'parent': {0: 'cH1'},\n",
       " 'summary': {0: 'cH1.5 is a replication dependent linker histone located within large histone gene clusters in mammals. cH1.5 plays a critical role in chromatin organization, thereby influencing cellular development and homeostasis [behrends_linker_2020].'},\n",
       " 'taxonomy': {0: 'null'},\n",
       " 'genes': {0: 'null'},\n",
       " 'evolution': {0: 'null'},\n",
       " 'expression': {0: 'cH1.5 is present across various cell types. As demonstrated on human cells, its expression level changes during differentiation: for instance, it decreases during the differentiation of NT2 cells into neural cells but increases during the dedifferentiation of keratinocytes into induced pluripotent stem cells [behrends_linker_2020].'},\n",
       " 'knock_out': {0: 'null'},\n",
       " 'function': {0: 'null'},\n",
       " 'sequence': {0: 'null'},\n",
       " 'localization': {0: 'null'},\n",
       " 'deposition': {0: 'null'},\n",
       " 'structure': {0: 'null'},\n",
       " 'interactions': {0: 'null'},\n",
       " 'disease': {0: 'null'},\n",
       " 'caveats': {0: 'null'},\n",
       " 'relations': {0: None}}"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cH1.5'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "83d56245-fcfc-4a2c-bc61-4e1cc01b9115",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8df84fc2-56a6-4697-8286-9ce0b5ea2262",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Update description of cH1.5_(Homo_sapiens)</span>\n",
    "\n",
    "### <span style=\"color:black\">Update summary</span>\n",
    "\n",
    "```cH1.5_(Homo_sapiens) is a replication dependent linker histone in human. сH1.5 plays a critical role in maintaining centromere integrity and is essential for sustaining appropriate transcriptional programs in differentiated cell types in human cells [saha_linker_2025, li_dynamic_2012].```\n",
    "\n",
    "### <span style=\"color:black\">Update expression</span>\n",
    "\n",
    "```cH1.5 is present across various human cell types. Notably, it serves as the predominant H1 variant in astrocyte-lineage SVGp12 cells and in glioblastoma multiforme cells [saha_linker_2025]. cH1.5 expression is downregulated during cellular differentiation. Compared to normal differentiated somatic cells, H1.5 exhibits higher expression levels in both pluripotent cells and fibroblasts [li_dynamic_2012].```\n",
    "\n",
    "### <span style=\"color:black\">Update function</span>\n",
    "\n",
    "```cH1.5 is essential for the regulation of mitotic integrity and for sustaining appropriate transcriptional programs in differentiated cell types [saha_linker_2025, li_dynamic_2012].```\n",
    "\n",
    "### <span style=\"color:black\">Update localization</span>\n",
    "\n",
    "```In differentiated cells, cH1.5 forms enrichment blocks in both genic and intergenic regions, preferentially associating with genes encoding membrane and membrane-associated proteins [li_dynamic_2012]. Moreover, cH1.5 is enriched at centromeric regions and co‑localizes with the centromeric histone CENP‑A in human, as confirmed by immunofluorescence and chromatin immunoprecipitation assays [saha_linker_2025]. cH1.5 binding is also associated with gene repression, chromatin compaction and promoting the inclusion of alternatively spliced exons into the mature mRNA [li_dynamic_2012, behrends_linker_2020].```\n",
    "\n",
    ">It is also enriched around splicing sites, particularly on genes with alternative splicing, where its binding promotes the inclusion of alternatively spliced exons into the mature mRNA (see behrends_linker_2020).\n",
    "\n",
    "### <span style=\"color:black\">Update knock_out</span>\n",
    "\n",
    "```cH1.5 depletion leads to decreased cell growth, increased chromatin accessibility and deregulation of gene expression [li_dynamic_2012, behrends_linker_2020]. Additionally, knockdown of cH1.5 results in the loss of centromeric α‑satellite DNA transcription, a reduction in de novo CENP‑A loading, and the accumulation of mitotic defects [saha_linker_2025].```\n",
    "\n",
    "### <span style=\"color:black\">Update structure</span>\n",
    "\n",
    "```cH1.5 binds to CENP‑A nucleosomes in a non‑canonical, dyad‑independent manner [saha_linker_2025].```\n",
    "\n",
    "### <span style=\"color:black\">Update interections</span>\n",
    "\n",
    "```cH1.5 directly interacts with CENP‑A‑containing nucleosomes both in vitro and in vivo [saha_linker_2025].```\n",
    "\n",
    "### <span style=\"color:black\">Update disease</span>\n",
    "\n",
    "```cH1.5 can serve as a valuable prognostic marker in cancer patients. Its expression level is decreased in many cancer cell lines, while being highly overexpressed in glioblastoma multiforme [saha_linker_2025, behrends_linker_2020, li_dynamic_2012]. Furthermore, cH1.5 mutations are associated with follicular lymphoma and altered chromatin states. Impairment of cH1.5 function may contribute to chromosomal instability and the development of malignancies, particularly in brain cells[saha_linker_2025].```\n",
    "\n",
    ">human H1.5 bind right on the center of the dyad symmetrically [wu_binding_2021, bednar_structure_2017]."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "c57be998-980e-4143-a73a-782dd4f7c649",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH1.5_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>66</td>\n",
       "      <td>cH1.5</td>\n",
       "      <td>66</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     id    level taxonomic_span taxonomic_span_id  \\\n",
       "0  cH1.5_(Homo_sapiens)  variant   Homo sapiens              9606   \n",
       "\n",
       "   description parent  id summary taxonomy genes  ... knock_out function  \\\n",
       "0           66  cH1.5  66    null     null  null  ...      null     null   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     null         null       null      null         null    null    null   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cH1.5_(Homo_sapiens)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "6100dff0-34cb-4eff-958a-29e0696b180f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': {0: 66},\n",
       " 'level': {0: 'variant'},\n",
       " 'taxonomic_span': {0: 'Homo sapiens'},\n",
       " 'taxonomic_span_id': {0: '9606'},\n",
       " 'description': {0: 66},\n",
       " 'parent': {0: 'cH1.5'},\n",
       " 'summary': {0: 'null'},\n",
       " 'taxonomy': {0: 'null'},\n",
       " 'genes': {0: 'null'},\n",
       " 'evolution': {0: 'null'},\n",
       " 'expression': {0: 'null'},\n",
       " 'knock_out': {0: 'null'},\n",
       " 'function': {0: 'null'},\n",
       " 'sequence': {0: 'null'},\n",
       " 'localization': {0: 'null'},\n",
       " 'deposition': {0: 'null'},\n",
       " 'structure': {0: 'null'},\n",
       " 'interactions': {0: 'null'},\n",
       " 'disease': {0: 'null'},\n",
       " 'caveats': {0: 'null'},\n",
       " 'relations': {0: None}}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f94be216-0a96-41ed-af4b-790b32cb160f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UPDATE histone_description SET summary='cH1.5_(Homo_sapiens) is a replication dependent linker histone in human. сH1.5 plays a critical role in maintaining centromere integrity and is essential for sustaining appropriate transcriptional programs in differentiated cell types in human cells [saha_linker_2025, li_dynamic_2012].', expression='cH1.5 is present across various human cell types. Notably, it serves as the predominant H1 variant in astrocyte-lineage SVGp12 cells and in glioblastoma multiforme cells [saha_linker_2025]. cH1.5 expression is downregulated during cellular differentiation. Compared to normal differentiated somatic cells, H1.5 exhibits higher expression levels in both pluripotent cells and fibroblasts [li_dynamic_2012].', knock_out='cH1.5 depletion leads to decreased cell growth, increased chromatin accessibility and deregulation of gene expression [li_dynamic_2012, behrends_linker_2020]. Additionally, knockdown of cH1.5 results in the loss of centromeric α‑satellite DNA transcription, a reduction in de novo CENP‑A loading, and the accumulation of mitotic defects [saha_linker_2025].', function='cH1.5 is essential for the regulation of mitotic integrity and for sustaining appropriate transcriptional programs in differentiated cell types [saha_linker_2025, li_dynamic_2012].', localization='In differentiated cells, cH1.5 forms enrichment blocks in both genic and intergenic regions, preferentially associating with genes encoding membrane and membrane-associated proteins [li_dynamic_2012]. Moreover, cH1.5 is enriched at centromeric regions and co‑localizes with the centromeric histone CENP‑A in human, as confirmed by immunofluorescence and chromatin immunoprecipitation assays [saha_linker_2025]. cH1.5 binding is also associated with gene repression, chromatin compaction and promoting the inclusion of alternatively spliced exons into the mature mRNA [li_dynamic_2012, behrends_linker_2020].', structure='cH1.5 binds to CENP‑A nucleosomes in a non‑canonical, dyad‑independent manner [saha_linker_2025].', interactions='cH1.5 directly interacts with CENP‑A‑containing nucleosomes both in vitro and in vivo [saha_linker_2025].', disease='cH1.5 can serve as a valuable prognostic marker in cancer patients. Its expression level is decreased in many cancer cell lines, while being highly overexpressed in glioblastoma multiforme [saha_linker_2025, behrends_linker_2020, li_dynamic_2012]. Furthermore, cH1.5 mutations are associated with follicular lymphoma and altered chromatin states. Impairment of cH1.5 function may contribute to chromosomal instability and the development of malignancies, particularly in brain cells[saha_linker_2025].' WHERE id=66\n"
     ]
    }
   ],
   "source": [
    "summary_desc = \"cH1.5_(Homo_sapiens) is a replication dependent linker histone in human. сH1.5 plays a critical role in maintaining centromere integrity and is essential for sustaining appropriate transcriptional programs in differentiated cell types in human cells [saha_linker_2025, li_dynamic_2012].\"\n",
    "expression_desc = \"cH1.5 is present across various human cell types. Notably, it serves as the predominant H1 variant in astrocyte-lineage SVGp12 cells and in glioblastoma multiforme cells [saha_linker_2025]. cH1.5 expression is downregulated during cellular differentiation. Compared to normal differentiated somatic cells, H1.5 exhibits higher expression levels in both pluripotent cells and fibroblasts [li_dynamic_2012].\"\n",
    "knock_out_desc = \"cH1.5 depletion leads to decreased cell growth, increased chromatin accessibility and deregulation of gene expression [li_dynamic_2012, behrends_linker_2020]. Additionally, knockdown of cH1.5 results in the loss of centromeric α‑satellite DNA transcription, a reduction in de novo CENP‑A loading, and the accumulation of mitotic defects [saha_linker_2025].\"\n",
    "function_desc = \"cH1.5 is essential for the regulation of mitotic integrity and for sustaining appropriate transcriptional programs in differentiated cell types [saha_linker_2025, li_dynamic_2012].\"\n",
    "localization_desc = \"In differentiated cells, cH1.5 forms enrichment blocks in both genic and intergenic regions, preferentially associating with genes encoding membrane and membrane-associated proteins [li_dynamic_2012]. Moreover, cH1.5 is enriched at centromeric regions and co‑localizes with the centromeric histone CENP‑A in human, as confirmed by immunofluorescence and chromatin immunoprecipitation assays [saha_linker_2025]. cH1.5 binding is also associated with gene repression, chromatin compaction and promoting the inclusion of alternatively spliced exons into the mature mRNA [li_dynamic_2012, behrends_linker_2020].\"\n",
    "structure_desc = \"cH1.5 binds to CENP‑A nucleosomes in a non‑canonical, dyad‑independent manner [saha_linker_2025].\"\n",
    "interactions_desc = \"cH1.5 directly interacts with CENP‑A‑containing nucleosomes both in vitro and in vivo [saha_linker_2025].\"\n",
    "disease_desc = \"cH1.5 can serve as a valuable prognostic marker in cancer patients. Its expression level is decreased in many cancer cell lines, while being highly overexpressed in glioblastoma multiforme [saha_linker_2025, behrends_linker_2020, li_dynamic_2012]. Furthermore, cH1.5 mutations are associated with follicular lymphoma and altered chromatin states. Impairment of cH1.5 function may contribute to chromosomal instability and the development of malignancies, particularly in brain cells[saha_linker_2025].\"\n",
    "query = f\"UPDATE histone_description SET summary='{summary_desc}', expression='{expression_desc}', knock_out='{knock_out_desc}', function='{function_desc}', localization='{localization_desc}', structure='{structure_desc}', interactions='{interactions_desc}', disease='{disease_desc}' WHERE id=66\"\n",
    "print(query)\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0714e025-df04-45c8-8722-211f1091e6fa",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': {0: 66},\n",
       " 'level': {0: 'variant'},\n",
       " 'taxonomic_span': {0: 'Homo sapiens'},\n",
       " 'taxonomic_span_id': {0: '9606'},\n",
       " 'description': {0: 66},\n",
       " 'parent': {0: 'cH1.5'},\n",
       " 'summary': {0: 'cH1.5_(Homo_sapiens) is a replication dependent linker histone in human. сH1.5 plays a critical role in maintaining centromere integrity and is essential for sustaining appropriate transcriptional programs in differentiated cell types in human cells [saha_linker_2025, li_dynamic_2012].'},\n",
       " 'taxonomy': {0: 'null'},\n",
       " 'genes': {0: 'null'},\n",
       " 'evolution': {0: 'null'},\n",
       " 'expression': {0: 'cH1.5 is present across various human cell types. Notably, it serves as the predominant H1 variant in astrocyte-lineage SVGp12 cells and in glioblastoma multiforme cells [saha_linker_2025]. cH1.5 expression is downregulated during cellular differentiation. Compared to normal differentiated somatic cells, H1.5 exhibits higher expression levels in both pluripotent cells and fibroblasts [li_dynamic_2012].'},\n",
       " 'knock_out': {0: 'cH1.5 depletion leads to decreased cell growth, increased chromatin accessibility and deregulation of gene expression [li_dynamic_2012, behrends_linker_2020]. Additionally, knockdown of cH1.5 results in the loss of centromeric α‑satellite DNA transcription, a reduction in de novo CENP‑A loading, and the accumulation of mitotic defects [saha_linker_2025].'},\n",
       " 'function': {0: 'cH1.5 is essential for the regulation of mitotic integrity and for sustaining appropriate transcriptional programs in differentiated cell types [saha_linker_2025, li_dynamic_2012].'},\n",
       " 'sequence': {0: 'null'},\n",
       " 'localization': {0: 'In differentiated cells, cH1.5 forms enrichment blocks in both genic and intergenic regions, preferentially associating with genes encoding membrane and membrane-associated proteins [li_dynamic_2012]. Moreover, cH1.5 is enriched at centromeric regions and co‑localizes with the centromeric histone CENP‑A in human, as confirmed by immunofluorescence and chromatin immunoprecipitation assays [saha_linker_2025]. cH1.5 binding is also associated with gene repression, chromatin compaction and promoting the inclusion of alternatively spliced exons into the mature mRNA [li_dynamic_2012, behrends_linker_2020].'},\n",
       " 'deposition': {0: 'null'},\n",
       " 'structure': {0: 'cH1.5 binds to CENP‑A nucleosomes in a non‑canonical, dyad‑independent manner [saha_linker_2025].'},\n",
       " 'interactions': {0: 'cH1.5 directly interacts with CENP‑A‑containing nucleosomes both in vitro and in vivo [saha_linker_2025].'},\n",
       " 'disease': {0: 'cH1.5 can serve as a valuable prognostic marker in cancer patients. Its expression level is decreased in many cancer cell lines, while being highly overexpressed in glioblastoma multiforme [saha_linker_2025, behrends_linker_2020, li_dynamic_2012]. Furthermore, cH1.5 mutations are associated with follicular lymphoma and altered chromatin states. Impairment of cH1.5 function may contribute to chromosomal instability and the development of malignancies, particularly in brain cells[saha_linker_2025].'},\n",
       " 'caveats': {0: 'null'},\n",
       " 'relations': {0: None}}"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cH1.5_(Homo_sapiens)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "464b0e93-cf3d-4e0a-9d00-0feb8d4a85f9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d7e3cce-a8cc-435b-892f-d52a2dba6b6e",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Update description of cH1.1</span>\n",
    "\n",
    "### <span style=\"color:black\">Update summary</span>\n",
    "\n",
    "```cH1.1 is a replication dependent linker histone located within large histone gene clusters in mammals. cH1.1 plays an important role in the early stages of embryogenesis [funaya_involvement_2023].```\n",
    "\n",
    "### <span style=\"color:black\">Update function</span>\n",
    "\n",
    "```cH1.1 plays a critical role in the early stages of embryogenesis. cH1.1 is absent in human differentiated keratinocytes and fibroblasts but is re-expressed upon reprogramming, suggesting that it is specific to pluripotent cells [terme_histone_2011]. In mice, it is essential for preimplantation development, particularly at the morula and blastocyst stages [funaya_involvement_2023].```\n",
    "\n",
    "### <span style=\"color:black\">Update expression</span>\n",
    "\n",
    "```cH1.1 expression predominates in early embryonic cells compared to other replication-dependent linker histone variants. In mice, cH1.1—alongside H1.8—exhibits the highest expression levels at the one- and two-cell stages of embryonic development [funaya_involvement_2023, funaya_linker_2018]. Although its expression declines by the four-cell stage, it remains elevated through the blastocyst stage. In human cells, сH1.1 has also been shown to be highly expressed in pluripotent cells—including embryonic and induced pluripotent stem cells—and absent in differentiated keratinocytes and fibroblasts [terme_histone_2011].```\n",
    "\n",
    "### <span style=\"color:black\">Update localization</span>\n",
    "\n",
    "```In mice, cH1.1 is associated with actively transcribed genes, whereas the somatic linker histone variants сH1.2–5 are excluded from genomic regions occupied by H1a, indicating that H1a exhibits a distinct and non-overlapping genomic localization [funaya_involvement_2023].```\n",
    "\n",
    "### <span style=\"color:black\">Update knock_out</span>\n",
    "\n",
    "```Maternal knockout of the cH1.1 gene in mice results in reduced litter size and impaired embryonic development during the morula-to-blastocyst transition [funaya_involvement_2023].```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "180b6042-4cf2-4b00-a6ed-3b635c5bc4ce",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH1.1</td>\n",
       "      <td>variant</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>11</td>\n",
       "      <td>cH1</td>\n",
       "      <td>11</td>\n",
       "      <td>cH1.1 is a replication dependent linker histon...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      id    level taxonomic_span taxonomic_span_id  description parent  id  \\\n",
       "0  cH1.1  variant       Mammalia             40674           11    cH1  11   \n",
       "\n",
       "                                             summary taxonomy genes  ...  \\\n",
       "0  cH1.1 is a replication dependent linker histon...     null  null  ...   \n",
       "\n",
       "  knock_out function sequence localization deposition structure interactions  \\\n",
       "0      null     null     null         null       null      null         null   \n",
       "\n",
       "  disease caveats relations  \n",
       "0    null    null      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cH1.1'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "34198d51-b4b0-4b86-bd1f-e8718c73bad6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': {0: 11},\n",
       " 'level': {0: 'variant'},\n",
       " 'taxonomic_span': {0: 'Mammalia'},\n",
       " 'taxonomic_span_id': {0: '40674'},\n",
       " 'description': {0: 11},\n",
       " 'parent': {0: 'cH1'},\n",
       " 'summary': {0: 'cH1.1 is a replication dependent linker histone located within large histone gene clusters in mammals [happel_histone_2009, talbert_histone_2021].'},\n",
       " 'taxonomy': {0: 'null'},\n",
       " 'genes': {0: 'null'},\n",
       " 'evolution': {0: 'null'},\n",
       " 'expression': {0: 'null'},\n",
       " 'knock_out': {0: 'null'},\n",
       " 'function': {0: 'null'},\n",
       " 'sequence': {0: 'null'},\n",
       " 'localization': {0: 'null'},\n",
       " 'deposition': {0: 'null'},\n",
       " 'structure': {0: 'null'},\n",
       " 'interactions': {0: 'null'},\n",
       " 'disease': {0: 'null'},\n",
       " 'caveats': {0: 'null'},\n",
       " 'relations': {0: None}}"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "aaf0e0d4-2be7-4182-8306-93104c00220a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UPDATE histone_description SET summary='cH1.1 is a replication dependent linker histone located within large histone gene clusters in mammals. cH1.1 plays an important role in the early stages of embryogenesis [funaya_involvement_2023].', expression='cH1.1 expression predominates in early embryonic cells compared to other replication-dependent linker histone variants. In mice, cH1.1—alongside H1.8—exhibits the highest expression levels at the one- and two-cell stages of embryonic development [funaya_involvement_2023, funaya_linker_2018]. Although its expression declines by the four-cell stage, it remains elevated through the blastocyst stage. In human cells, сH1.1 has also been shown to be highly expressed in pluripotent cells—including embryonic and induced pluripotent stem cells—and absent in differentiated keratinocytes and fibroblasts [terme_histone_2011].', knock_out='Maternal knockout of the cH1.1 gene in mice results in reduced litter size and impaired embryonic development during the morula-to-blastocyst transition [funaya_involvement_2023].', function='cH1.1 plays a critical role in the early stages of embryogenesis. cH1.1 is absent in human differentiated keratinocytes and fibroblasts but is re-expressed upon reprogramming, suggesting that it is specific to pluripotent cells [terme_histone_2011]. In mice, it is essential for preimplantation development, particularly at the morula and blastocyst stages [funaya_involvement_2023].', localization='In mice, cH1.1 is associated with actively transcribed genes, whereas the somatic linker histone variants сH1.2–5 are excluded from genomic regions occupied by H1a, indicating that H1a exhibits a distinct and non-overlapping genomic localization [funaya_involvement_2023].' WHERE id=11\n"
     ]
    }
   ],
   "source": [
    "summary_desc = \"cH1.1 is a replication dependent linker histone located within large histone gene clusters in mammals. cH1.1 plays an important role in the early stages of embryogenesis [funaya_involvement_2023].\"\n",
    "expression_desc = \"cH1.1 expression predominates in early embryonic cells compared to other replication-dependent linker histone variants. In mice, cH1.1—alongside H1.8—exhibits the highest expression levels at the one- and two-cell stages of embryonic development [funaya_involvement_2023, funaya_linker_2018]. Although its expression declines by the four-cell stage, it remains elevated through the blastocyst stage. In human cells, сH1.1 has also been shown to be highly expressed in pluripotent cells—including embryonic and induced pluripotent stem cells—and absent in differentiated keratinocytes and fibroblasts [terme_histone_2011].\"\n",
    "knock_out_desc = \"Maternal knockout of the cH1.1 gene in mice results in reduced litter size and impaired embryonic development during the morula-to-blastocyst transition [funaya_involvement_2023].\"\n",
    "function_desc = \"cH1.1 plays a critical role in the early stages of embryogenesis. cH1.1 is absent in human differentiated keratinocytes and fibroblasts but is re-expressed upon reprogramming, suggesting that it is specific to pluripotent cells [terme_histone_2011]. In mice, it is essential for preimplantation development, particularly at the morula and blastocyst stages [funaya_involvement_2023].\"\n",
    "localization_desc = \"In mice, cH1.1 is associated with actively transcribed genes, whereas the somatic linker histone variants сH1.2–5 are excluded from genomic regions occupied by H1a, indicating that H1a exhibits a distinct and non-overlapping genomic localization [funaya_involvement_2023].\"\n",
    "query = f\"UPDATE histone_description SET summary='{summary_desc}', expression='{expression_desc}', knock_out='{knock_out_desc}', function='{function_desc}', localization='{localization_desc}' WHERE id=11\"\n",
    "print(query)\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "a6f677f3-dd21-4b9e-a58a-db31a6a3c947",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': {0: 11},\n",
       " 'level': {0: 'variant'},\n",
       " 'taxonomic_span': {0: 'Mammalia'},\n",
       " 'taxonomic_span_id': {0: '40674'},\n",
       " 'description': {0: 11},\n",
       " 'parent': {0: 'cH1'},\n",
       " 'summary': {0: 'cH1.1 is a replication dependent linker histone located within large histone gene clusters in mammals. cH1.1 plays an important role in the early stages of embryogenesis [funaya_involvement_2023].'},\n",
       " 'taxonomy': {0: 'null'},\n",
       " 'genes': {0: 'null'},\n",
       " 'evolution': {0: 'null'},\n",
       " 'expression': {0: 'cH1.1 expression predominates in early embryonic cells compared to other replication-dependent linker histone variants. In mice, cH1.1—alongside H1.8—exhibits the highest expression levels at the one- and two-cell stages of embryonic development [funaya_involvement_2023, funaya_linker_2018]. Although its expression declines by the four-cell stage, it remains elevated through the blastocyst stage. In human cells, сH1.1 has also been shown to be highly expressed in pluripotent cells—including embryonic and induced pluripotent stem cells—and absent in differentiated keratinocytes and fibroblasts [terme_histone_2011].'},\n",
       " 'knock_out': {0: 'Maternal knockout of the cH1.1 gene in mice results in reduced litter size and impaired embryonic development during the morula-to-blastocyst transition [funaya_involvement_2023].'},\n",
       " 'function': {0: 'cH1.1 plays a critical role in the early stages of embryogenesis. cH1.1 is absent in human differentiated keratinocytes and fibroblasts but is re-expressed upon reprogramming, suggesting that it is specific to pluripotent cells [terme_histone_2011]. In mice, it is essential for preimplantation development, particularly at the morula and blastocyst stages [funaya_involvement_2023].'},\n",
       " 'sequence': {0: 'null'},\n",
       " 'localization': {0: 'In mice, cH1.1 is associated with actively transcribed genes, whereas the somatic linker histone variants сH1.2–5 are excluded from genomic regions occupied by H1a, indicating that H1a exhibits a distinct and non-overlapping genomic localization [funaya_involvement_2023].'},\n",
       " 'deposition': {0: 'null'},\n",
       " 'structure': {0: 'null'},\n",
       " 'interactions': {0: 'null'},\n",
       " 'disease': {0: 'null'},\n",
       " 'caveats': {0: 'null'},\n",
       " 'relations': {0: None}}"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cH1.1'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "54ff41fe-547e-41ca-a1c4-65eb9219fc88",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7592245e-6de3-4b20-8351-583d611a99d6",
   "metadata": {},
   "source": [
    "# DONE H3\n",
    "\n",
    "## Add sequences to H3-like_(Plants)\n",
    "\n",
    "Accessions: NP_172794.1 (HTR6), NP_201338.1 (HTR11),  NP_196795.1 (HTR15) [alvarez-venegas_canonical_2019, okada_analysis_2005]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "a16c5f7d-d564-4dd6-8490-174651c227fa",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"NP_172794.1\", \"NP_201338.1\", \"NP_196795.1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1cc4008-5d56-4654-9bf0-0893ffc7df4c",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "4aca6092-0dea-4697-83c5-61e5241b5ac6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "88045c74-f56f-4f41-b515-3e54b3ad5f2a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 3702\n",
      "Fetched taxid from NCBI 3702\n",
      "Fetched taxid from NCBI 3702\n",
      "accession NP_196795.1 <class 'str'>\n",
      "variant H3-like_(Plants) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 3702 <class 'int'>\n",
      "organism Arabidopsis thaliana <class 'str'>\n",
      "phylum Streptophyta <class 'str'>\n",
      "class Magnoliopsida <class 'str'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MARSNQTARKATGGKAPHFAMRVWQHSTPPLKKPYRYKPGTVALREIRKYQKTTDLVIRKLPFQRLVKEIAQSLKADLRFQTGAVSALQEAAEAFMVGMFEDTNLCAMHAKRSTIMPKDIQLAKRLRGDRV <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for a in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=a, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"H3-like_(Plants)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "aa8f5543-02c0-4688-88ba-d304ad26358a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "ac938a47-09bc-45a0-82a5-fceec45c244d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3557</th>\n",
       "      <td>NP_172794.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702.0</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQSARKSHGGKAPTKQLATKAARKSAPTTGGVKKPHRFRPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3568</th>\n",
       "      <td>NP_196795.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702.0</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARSNQTARKATGGKAPHFAMRVWQHSTPPLKKPYRYKPGTVALRE...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3572</th>\n",
       "      <td>NP_201338.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702.0</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARISTGGKAPRKQLAPKAARQSAPATGGVKKPHRFRPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession           variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "3557  NP_172794.1  H3-like_(Plants)  None         None           None   \n",
       "3568  NP_196795.1  H3-like_(Plants)  None         None           None   \n",
       "3572  NP_201338.1  H3-like_(Plants)  None         None           None   \n",
       "\n",
       "      taxonomy_id              organism        phylum          class  \\\n",
       "3557       3702.0  Arabidopsis thaliana  Streptophyta  Magnoliopsida   \n",
       "3568       3702.0  Arabidopsis thaliana  Streptophyta  Magnoliopsida   \n",
       "3572       3702.0  Arabidopsis thaliana  Streptophyta  Magnoliopsida   \n",
       "\n",
       "     taxonomy_group  info                                           sequence  \\\n",
       "3557           None  None  MARTKQSARKSHGGKAPTKQLATKAARKSAPTTGGVKKPHRFRPGT...   \n",
       "3568           None  None  MARSNQTARKATGGKAPHFAMRVWQHSTPPLKKPYRYKPGTVALRE...   \n",
       "3572           None  None  MARTKQTARISTGGKAPRKQLAPKAARQSAPATGGVKKPHRFRPGT...   \n",
       "\n",
       "     variant_under_consideration  \n",
       "3557                        None  \n",
       "3568                        None  \n",
       "3572                        None  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "e1dda38a-497a-4879-98f6-362b52013ea6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc7efe72-330d-4989-82c2-805a9725e07f",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "7dfef0cb-923d-4f42-9b00-aac0d915d16c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                id title   doi author  year pubmed_id\n",
       "74  alvarez-venegas_canonical_2019  None  None   None  None      None"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"alvarez-venegas_canonical_2019\", \"okada_analysis_2005\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "8ae92b6f-49b2-4ac3-a71f-45bb164aafdd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_publication = {\n",
    "    \"id\": \"okada_analysis_2005\",\n",
    "    \"title\": None,\n",
    "    \"doi\": None,\n",
    "    \"author\": None,\n",
    "    \"year\": None,\n",
    "}\n",
    "cursor.execute(add_publication, data_publication)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "7f7f2d3c-08bc-45c3-948e-72fcd96c000b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>okada_analysis_2005</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 id title   doi author  year pubmed_id\n",
       "74   alvarez-venegas_canonical_2019  None  None   None  None      None\n",
       "144             okada_analysis_2005  None  None   None  None      None"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"alvarez-venegas_canonical_2019\", \"okada_analysis_2005\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "b909ea1a-270b-400a-88fc-e7c1ce9c0e33",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "21d03565-952d-4095-99f7-00759efd8934",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3979</th>\n",
       "      <td>NP_172794.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702.0</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQSARKSHGGKAPTKQLATKAARKSAPTTGGVKKPHRFRPGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>NP_172794.1</td>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3980</th>\n",
       "      <td>NP_172794.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702.0</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQSARKSHGGKAPTKQLATKAARKSAPTTGGVKKPHRFRPGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>NP_172794.1</td>\n",
       "      <td>okada_analysis_2005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3998</th>\n",
       "      <td>NP_196795.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702.0</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARSNQTARKATGGKAPHFAMRVWQHSTPPLKKPYRYKPGTVALRE...</td>\n",
       "      <td>None</td>\n",
       "      <td>NP_196795.1</td>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3999</th>\n",
       "      <td>NP_196795.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702.0</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARSNQTARKATGGKAPHFAMRVWQHSTPPLKKPYRYKPGTVALRE...</td>\n",
       "      <td>None</td>\n",
       "      <td>NP_196795.1</td>\n",
       "      <td>okada_analysis_2005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4006</th>\n",
       "      <td>NP_201338.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702.0</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARISTGGKAPRKQLAPKAARQSAPATGGVKKPHRFRPGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>NP_201338.1</td>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4007</th>\n",
       "      <td>NP_201338.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702.0</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARISTGGKAPRKQLAPKAARQSAPATGGVKKPHRFRPGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>NP_201338.1</td>\n",
       "      <td>okada_analysis_2005</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession           variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "3979  NP_172794.1  H3-like_(Plants)  None         None           None   \n",
       "3980  NP_172794.1  H3-like_(Plants)  None         None           None   \n",
       "3998  NP_196795.1  H3-like_(Plants)  None         None           None   \n",
       "3999  NP_196795.1  H3-like_(Plants)  None         None           None   \n",
       "4006  NP_201338.1  H3-like_(Plants)  None         None           None   \n",
       "4007  NP_201338.1  H3-like_(Plants)  None         None           None   \n",
       "\n",
       "      taxonomy_id              organism        phylum          class  \\\n",
       "3979       3702.0  Arabidopsis thaliana  Streptophyta  Magnoliopsida   \n",
       "3980       3702.0  Arabidopsis thaliana  Streptophyta  Magnoliopsida   \n",
       "3998       3702.0  Arabidopsis thaliana  Streptophyta  Magnoliopsida   \n",
       "3999       3702.0  Arabidopsis thaliana  Streptophyta  Magnoliopsida   \n",
       "4006       3702.0  Arabidopsis thaliana  Streptophyta  Magnoliopsida   \n",
       "4007       3702.0  Arabidopsis thaliana  Streptophyta  Magnoliopsida   \n",
       "\n",
       "     taxonomy_group  info                                           sequence  \\\n",
       "3979           None  None  MARTKQSARKSHGGKAPTKQLATKAARKSAPTTGGVKKPHRFRPGT...   \n",
       "3980           None  None  MARTKQSARKSHGGKAPTKQLATKAARKSAPTTGGVKKPHRFRPGT...   \n",
       "3998           None  None  MARSNQTARKATGGKAPHFAMRVWQHSTPPLKKPYRYKPGTVALRE...   \n",
       "3999           None  None  MARSNQTARKATGGKAPHFAMRVWQHSTPPLKKPYRYKPGTVALRE...   \n",
       "4006           None  None  MARTKQTARISTGGKAPRKQLAPKAARQSAPATGGVKKPHRFRPGT...   \n",
       "4007           None  None  MARTKQTARISTGGKAPRKQLAPKAARQSAPATGGVKKPHRFRPGT...   \n",
       "\n",
       "     variant_under_consideration sequence_accession  \\\n",
       "3979                        None        NP_172794.1   \n",
       "3980                        None        NP_172794.1   \n",
       "3998                        None        NP_196795.1   \n",
       "3999                        None        NP_196795.1   \n",
       "4006                        None        NP_201338.1   \n",
       "4007                        None        NP_201338.1   \n",
       "\n",
       "                      publication_id  \n",
       "3979  alvarez-venegas_canonical_2019  \n",
       "3980             okada_analysis_2005  \n",
       "3998  alvarez-venegas_canonical_2019  \n",
       "3999             okada_analysis_2005  \n",
       "4006  alvarez-venegas_canonical_2019  \n",
       "4007             okada_analysis_2005  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "9bb22363-87bc-43f6-a5d9-8e2a826e451a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2d36d78-e5cd-4954-898c-166b63c10e8e",
   "metadata": {},
   "source": [
    "## Add sequences to H3-like_(Plants)\n",
    "\n",
    "Accessions: ABA97899.1 (HTR714), ABA97902.1 (HTR715) [alvarez-venegas_canonical_2019, hu_identification_2015]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "136ca526-08b0-4ea3-8bde-b80de6c24978",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"ABA97899.1\", \"ABA97902.1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "798658c4-cb65-4276-9539-d7d734b35372",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "182f84dc-32e7-4883-a50e-0760e9752f5e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "1ec5e0f2-3cac-4fd3-97e9-5678b951308f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 39947\n",
      "Fetched taxid from NCBI 39947\n",
      "accession ABA97902.1 <class 'str'>\n",
      "variant H3-like_(Plants) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 39947 <class 'int'>\n",
      "organism Oryza sativa Japonica Group <class 'str'>\n",
      "phylum Streptophyta <class 'str'>\n",
      "class Magnoliopsida <class 'str'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MARTKQTAKKSTASNVPRKLLVMKVARKSAPTMAGLKKPHRFKPGTVALREIRTYQKSTELLIRKLPFQRLVQEIAQDVRSYVRFQSSAVVALQEAAETYLVGLFKDTNLCVIHAKRVTIMPKDIQLARRIRGEKA <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for a in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=a, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"H3-like_(Plants)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "fea77b35-929e-4531-bfcb-d57cab3499f6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "ae9f44d1-e985-4743-93ec-0ac6f5a82fab",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>ABA97899.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>39947.0</td>\n",
       "      <td>Oryza sativa Japonica Group</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTAKKSTAGNVPRKLLVMKVARKSAPMMAGLKKPHRFNPWI...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>ABA97902.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>39947.0</td>\n",
       "      <td>Oryza sativa Japonica Group</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTAKKSTASNVPRKLLVMKVARKSAPTMAGLKKPHRFKPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession           variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "50  ABA97899.1  H3-like_(Plants)  None         None           None   \n",
       "51  ABA97902.1  H3-like_(Plants)  None         None           None   \n",
       "\n",
       "    taxonomy_id                     organism        phylum          class  \\\n",
       "50      39947.0  Oryza sativa Japonica Group  Streptophyta  Magnoliopsida   \n",
       "51      39947.0  Oryza sativa Japonica Group  Streptophyta  Magnoliopsida   \n",
       "\n",
       "   taxonomy_group  info                                           sequence  \\\n",
       "50           None  None  MARTKQTAKKSTAGNVPRKLLVMKVARKSAPMMAGLKKPHRFNPWI...   \n",
       "51           None  None  MARTKQTAKKSTASNVPRKLLVMKVARKSAPTMAGLKKPHRFKPGT...   \n",
       "\n",
       "   variant_under_consideration  \n",
       "50                        None  \n",
       "51                        None  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "85e9a101-b5d6-4573-9265-739b87441e1c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "478eb050-8cce-43bd-b85b-95566d79a436",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "cf7b427c-0656-46dd-9f48-fc2ae5f6b3cd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                id title   doi author  year pubmed_id\n",
       "74  alvarez-venegas_canonical_2019  None  None   None  None      None"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"alvarez-venegas_canonical_2019\", \"hu_identification_2015\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "1a5095f8-e917-404b-9a7f-16b68cb5e6bd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_publication = {\n",
    "    \"id\": \"hu_identification_2015\",\n",
    "    \"title\": None,\n",
    "    \"doi\": None,\n",
    "    \"author\": None,\n",
    "    \"year\": None,\n",
    "}\n",
    "cursor.execute(add_publication, data_publication)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "279703f7-82c9-4ad1-b48e-7b19b1f8808f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>hu_identification_2015</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 id title   doi author  year pubmed_id\n",
       "74   alvarez-venegas_canonical_2019  None  None   None  None      None\n",
       "106          hu_identification_2015  None  None   None  None      None"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"alvarez-venegas_canonical_2019\", \"hu_identification_2015\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "30497be9-0ffb-4e62-8947-7f058aa66c7a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "6c03f417-699e-4035-91f6-ca6181cbbe3e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>ABA97899.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>39947.0</td>\n",
       "      <td>Oryza sativa Japonica Group</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTAKKSTAGNVPRKLLVMKVARKSAPMMAGLKKPHRFNPWI...</td>\n",
       "      <td>None</td>\n",
       "      <td>ABA97899.1</td>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>ABA97899.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>39947.0</td>\n",
       "      <td>Oryza sativa Japonica Group</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTAKKSTAGNVPRKLLVMKVARKSAPMMAGLKKPHRFNPWI...</td>\n",
       "      <td>None</td>\n",
       "      <td>ABA97899.1</td>\n",
       "      <td>hu_identification_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>ABA97902.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>39947.0</td>\n",
       "      <td>Oryza sativa Japonica Group</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTAKKSTASNVPRKLLVMKVARKSAPTMAGLKKPHRFKPGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>ABA97902.1</td>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>ABA97902.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>39947.0</td>\n",
       "      <td>Oryza sativa Japonica Group</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTAKKSTASNVPRKLLVMKVARKSAPTMAGLKKPHRFKPGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>ABA97902.1</td>\n",
       "      <td>hu_identification_2015</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession           variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "62  ABA97899.1  H3-like_(Plants)  None         None           None   \n",
       "63  ABA97899.1  H3-like_(Plants)  None         None           None   \n",
       "64  ABA97902.1  H3-like_(Plants)  None         None           None   \n",
       "65  ABA97902.1  H3-like_(Plants)  None         None           None   \n",
       "\n",
       "    taxonomy_id                     organism        phylum          class  \\\n",
       "62      39947.0  Oryza sativa Japonica Group  Streptophyta  Magnoliopsida   \n",
       "63      39947.0  Oryza sativa Japonica Group  Streptophyta  Magnoliopsida   \n",
       "64      39947.0  Oryza sativa Japonica Group  Streptophyta  Magnoliopsida   \n",
       "65      39947.0  Oryza sativa Japonica Group  Streptophyta  Magnoliopsida   \n",
       "\n",
       "   taxonomy_group  info                                           sequence  \\\n",
       "62           None  None  MARTKQTAKKSTAGNVPRKLLVMKVARKSAPMMAGLKKPHRFNPWI...   \n",
       "63           None  None  MARTKQTAKKSTAGNVPRKLLVMKVARKSAPMMAGLKKPHRFNPWI...   \n",
       "64           None  None  MARTKQTAKKSTASNVPRKLLVMKVARKSAPTMAGLKKPHRFKPGT...   \n",
       "65           None  None  MARTKQTAKKSTASNVPRKLLVMKVARKSAPTMAGLKKPHRFKPGT...   \n",
       "\n",
       "   variant_under_consideration sequence_accession  \\\n",
       "62                        None         ABA97899.1   \n",
       "63                        None         ABA97899.1   \n",
       "64                        None         ABA97902.1   \n",
       "65                        None         ABA97902.1   \n",
       "\n",
       "                    publication_id  \n",
       "62  alvarez-venegas_canonical_2019  \n",
       "63          hu_identification_2015  \n",
       "64  alvarez-venegas_canonical_2019  \n",
       "65          hu_identification_2015  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "99d79941-8d46-43d3-a10c-575c922c3bfe",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13c75cb9-990c-471b-b57e-9338b6cc5b8e",
   "metadata": {},
   "source": [
    "## Add sequences to H3-like_(Plants)\n",
    "\n",
    "Accessions: BAE48427.1 (gcH3), BAA96098.1 (gH3), BAE48431.1 (leH3), BAE48433.1 (soH3-1), BAE48435.1 (soH3-2) [alvarez-venegas_canonical_2019, probst_similar_2020]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "b7e94b3c-da96-4320-9cb1-d1b74bc625bd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"BAE48427.1\", \"BAA96098.1\", \"BAE48431.1\", \"BAE48433.1\", \"BAE48435.1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94b50c1e-a9b4-4475-ad88-657403a94b25",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "7d490792-3049-4261-abd5-8d40bc9ba103",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "8f186087-eacb-4128-a79d-c3e3a143ffb9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 4690\n",
      "Fetched taxid from NCBI 4690\n",
      "Fetched taxid from NCBI 4690\n",
      "Fetched taxid from NCBI 4690\n",
      "Fetched taxid from NCBI 4690\n",
      "accession BAE48435.1 <class 'str'>\n",
      "variant H3-like_(Plants) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 4690 <class 'int'>\n",
      "organism Lilium longiflorum <class 'str'>\n",
      "phylum Streptophyta <class 'str'>\n",
      "class Magnoliopsida <class 'str'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MARTKQTARKSTGGKAPRKQLATKAARKSAPTTGGVKKPHRYRPGTVALREIRKYQKSTDLLIRKLPFQRLVREIAQDYKADLRFQSHAVLALQEAAEAYLVGLFEDTNLCAIHAKRVTIMPKDIQLARRIRGERA <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for a in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=a, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"H3-like_(Plants)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "46322f85-978a-462c-b302-b31c1ea30482",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "557ccac9-8f90-4de6-b8a6-db6122d7afd7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>868</th>\n",
       "      <td>BAA96098.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690.0</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARPRKEAPQRNLDRDENARQQPTEEPQDEAPRNQGRQQQQQRPPA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>880</th>\n",
       "      <td>BAE48427.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690.0</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARMKHTARMSTGGKAPRKQLASKALRKAPPPPTKGVKQPHHYHLR...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>881</th>\n",
       "      <td>BAE48431.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690.0</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARMKLNARMSTGGKAPRKQLAYKAVRKAAPPTIGVKLPNSYRPGD...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>882</th>\n",
       "      <td>BAE48433.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690.0</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSIPTGMGGMKRPRRYRPG...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>883</th>\n",
       "      <td>BAE48435.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690.0</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSAPTTGGVKKPHRYRPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession           variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "868  BAA96098.1  H3-like_(Plants)  None         None           None   \n",
       "880  BAE48427.1  H3-like_(Plants)  None         None           None   \n",
       "881  BAE48431.1  H3-like_(Plants)  None         None           None   \n",
       "882  BAE48433.1  H3-like_(Plants)  None         None           None   \n",
       "883  BAE48435.1  H3-like_(Plants)  None         None           None   \n",
       "\n",
       "     taxonomy_id            organism        phylum          class  \\\n",
       "868       4690.0  Lilium longiflorum  Streptophyta  Magnoliopsida   \n",
       "880       4690.0  Lilium longiflorum  Streptophyta  Magnoliopsida   \n",
       "881       4690.0  Lilium longiflorum  Streptophyta  Magnoliopsida   \n",
       "882       4690.0  Lilium longiflorum  Streptophyta  Magnoliopsida   \n",
       "883       4690.0  Lilium longiflorum  Streptophyta  Magnoliopsida   \n",
       "\n",
       "    taxonomy_group  info                                           sequence  \\\n",
       "868           None  None  MARPRKEAPQRNLDRDENARQQPTEEPQDEAPRNQGRQQQQQRPPA...   \n",
       "880           None  None  MARMKHTARMSTGGKAPRKQLASKALRKAPPPPTKGVKQPHHYHLR...   \n",
       "881           None  None  MARMKLNARMSTGGKAPRKQLAYKAVRKAAPPTIGVKLPNSYRPGD...   \n",
       "882           None  None  MARTKQTARKSTGGKAPRKQLATKAARKSIPTGMGGMKRPRRYRPG...   \n",
       "883           None  None  MARTKQTARKSTGGKAPRKQLATKAARKSAPTTGGVKKPHRYRPGT...   \n",
       "\n",
       "    variant_under_consideration  \n",
       "868                        None  \n",
       "880                        None  \n",
       "881                        None  \n",
       "882                        None  \n",
       "883                        None  "
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "c1b5f779-ed19-47a1-a969-5a5d5ca35e9c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ab16d64-8203-4a8a-bbd8-86923d83df7f",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "912ee3cf-e108-4f93-94c1-a35f910416a1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                id title   doi author  year pubmed_id\n",
       "74  alvarez-venegas_canonical_2019  None  None   None  None      None"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"alvarez-venegas_canonical_2019\", \"probst_similar_2020\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "9c887214-2331-431c-8e95-28503f5b3717",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_publication = {\n",
    "    \"id\": \"probst_similar_2020\",\n",
    "    \"title\": None,\n",
    "    \"doi\": None,\n",
    "    \"author\": None,\n",
    "    \"year\": None,\n",
    "}\n",
    "cursor.execute(add_publication, data_publication)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "7ec8544d-cc3c-4b62-8915-b67d2e8fdb26",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>probst_similar_2020</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 id title   doi author  year pubmed_id\n",
       "74   alvarez-venegas_canonical_2019  None  None   None  None      None\n",
       "150             probst_similar_2020  None  None   None  None      None"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"alvarez-venegas_canonical_2019\", \"probst_similar_2020\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "7f16db3a-9d85-4524-8afa-f3e183eda979",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "71a7f6c7-3d9c-4d99-82bc-065f0bdd8443",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1137</th>\n",
       "      <td>BAA96098.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690.0</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARPRKEAPQRNLDRDENARQQPTEEPQDEAPRNQGRQQQQQRPPA...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAA96098.1</td>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1138</th>\n",
       "      <td>BAA96098.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690.0</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARPRKEAPQRNLDRDENARQQPTEEPQDEAPRNQGRQQQQQRPPA...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAA96098.1</td>\n",
       "      <td>probst_similar_2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1153</th>\n",
       "      <td>BAE48427.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690.0</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARMKHTARMSTGGKAPRKQLASKALRKAPPPPTKGVKQPHHYHLR...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAE48427.1</td>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1154</th>\n",
       "      <td>BAE48427.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690.0</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARMKHTARMSTGGKAPRKQLASKALRKAPPPPTKGVKQPHHYHLR...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAE48427.1</td>\n",
       "      <td>probst_similar_2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1155</th>\n",
       "      <td>BAE48431.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690.0</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARMKLNARMSTGGKAPRKQLAYKAVRKAAPPTIGVKLPNSYRPGD...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAE48431.1</td>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1156</th>\n",
       "      <td>BAE48431.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690.0</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARMKLNARMSTGGKAPRKQLAYKAVRKAAPPTIGVKLPNSYRPGD...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAE48431.1</td>\n",
       "      <td>probst_similar_2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1157</th>\n",
       "      <td>BAE48433.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690.0</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSIPTGMGGMKRPRRYRPG...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAE48433.1</td>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1158</th>\n",
       "      <td>BAE48433.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690.0</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSIPTGMGGMKRPRRYRPG...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAE48433.1</td>\n",
       "      <td>probst_similar_2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1159</th>\n",
       "      <td>BAE48435.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690.0</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSAPTTGGVKKPHRYRPGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAE48435.1</td>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1160</th>\n",
       "      <td>BAE48435.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690.0</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSAPTTGGVKKPHRYRPGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>BAE48435.1</td>\n",
       "      <td>probst_similar_2020</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession           variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "1137  BAA96098.1  H3-like_(Plants)  None         None           None   \n",
       "1138  BAA96098.1  H3-like_(Plants)  None         None           None   \n",
       "1153  BAE48427.1  H3-like_(Plants)  None         None           None   \n",
       "1154  BAE48427.1  H3-like_(Plants)  None         None           None   \n",
       "1155  BAE48431.1  H3-like_(Plants)  None         None           None   \n",
       "1156  BAE48431.1  H3-like_(Plants)  None         None           None   \n",
       "1157  BAE48433.1  H3-like_(Plants)  None         None           None   \n",
       "1158  BAE48433.1  H3-like_(Plants)  None         None           None   \n",
       "1159  BAE48435.1  H3-like_(Plants)  None         None           None   \n",
       "1160  BAE48435.1  H3-like_(Plants)  None         None           None   \n",
       "\n",
       "      taxonomy_id            organism        phylum          class  \\\n",
       "1137       4690.0  Lilium longiflorum  Streptophyta  Magnoliopsida   \n",
       "1138       4690.0  Lilium longiflorum  Streptophyta  Magnoliopsida   \n",
       "1153       4690.0  Lilium longiflorum  Streptophyta  Magnoliopsida   \n",
       "1154       4690.0  Lilium longiflorum  Streptophyta  Magnoliopsida   \n",
       "1155       4690.0  Lilium longiflorum  Streptophyta  Magnoliopsida   \n",
       "1156       4690.0  Lilium longiflorum  Streptophyta  Magnoliopsida   \n",
       "1157       4690.0  Lilium longiflorum  Streptophyta  Magnoliopsida   \n",
       "1158       4690.0  Lilium longiflorum  Streptophyta  Magnoliopsida   \n",
       "1159       4690.0  Lilium longiflorum  Streptophyta  Magnoliopsida   \n",
       "1160       4690.0  Lilium longiflorum  Streptophyta  Magnoliopsida   \n",
       "\n",
       "     taxonomy_group  info                                           sequence  \\\n",
       "1137           None  None  MARPRKEAPQRNLDRDENARQQPTEEPQDEAPRNQGRQQQQQRPPA...   \n",
       "1138           None  None  MARPRKEAPQRNLDRDENARQQPTEEPQDEAPRNQGRQQQQQRPPA...   \n",
       "1153           None  None  MARMKHTARMSTGGKAPRKQLASKALRKAPPPPTKGVKQPHHYHLR...   \n",
       "1154           None  None  MARMKHTARMSTGGKAPRKQLASKALRKAPPPPTKGVKQPHHYHLR...   \n",
       "1155           None  None  MARMKLNARMSTGGKAPRKQLAYKAVRKAAPPTIGVKLPNSYRPGD...   \n",
       "1156           None  None  MARMKLNARMSTGGKAPRKQLAYKAVRKAAPPTIGVKLPNSYRPGD...   \n",
       "1157           None  None  MARTKQTARKSTGGKAPRKQLATKAARKSIPTGMGGMKRPRRYRPG...   \n",
       "1158           None  None  MARTKQTARKSTGGKAPRKQLATKAARKSIPTGMGGMKRPRRYRPG...   \n",
       "1159           None  None  MARTKQTARKSTGGKAPRKQLATKAARKSAPTTGGVKKPHRYRPGT...   \n",
       "1160           None  None  MARTKQTARKSTGGKAPRKQLATKAARKSAPTTGGVKKPHRYRPGT...   \n",
       "\n",
       "     variant_under_consideration sequence_accession  \\\n",
       "1137                        None         BAA96098.1   \n",
       "1138                        None         BAA96098.1   \n",
       "1153                        None         BAE48427.1   \n",
       "1154                        None         BAE48427.1   \n",
       "1155                        None         BAE48431.1   \n",
       "1156                        None         BAE48431.1   \n",
       "1157                        None         BAE48433.1   \n",
       "1158                        None         BAE48433.1   \n",
       "1159                        None         BAE48435.1   \n",
       "1160                        None         BAE48435.1   \n",
       "\n",
       "                      publication_id  \n",
       "1137  alvarez-venegas_canonical_2019  \n",
       "1138             probst_similar_2020  \n",
       "1153  alvarez-venegas_canonical_2019  \n",
       "1154             probst_similar_2020  \n",
       "1155  alvarez-venegas_canonical_2019  \n",
       "1156             probst_similar_2020  \n",
       "1157  alvarez-venegas_canonical_2019  \n",
       "1158             probst_similar_2020  \n",
       "1159  alvarez-venegas_canonical_2019  \n",
       "1160             probst_similar_2020  "
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "4151141b-aecc-43e7-9f52-de17c66ec293",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e8bad07-ef74-40fd-a93b-e7897df0c591",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Close connections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "e5673501-5845-491a-bbb9-f5d213721d58",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cursor.close()\n",
    "conn.close()\n",
    "tunnel.stop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "778f8bc6-63f6-4c76-aa4b-7bdffa839e52",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".conda-histdb_env",
   "language": "python",
   "name": "conda-env-.conda-histdb_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
