{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "53d3ec55-7daf-4ade-8add-2830b709dc54",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import io\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import Entrez, SeqIO\n",
    "from mysql.connector import connection\n",
    "from sshtunnel import SSHTunnelForwarder\n",
    "\n",
    "# Укажите ваш email (обязательно для использования Entrez)\n",
    "Entrez.email = \"your.email@example.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "608c4bfe-b979-4b60-b2b5-b94eeb679cb5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open(\"db_curated_server_info.txt\", \"r\") as file:\n",
    "    lines = file.readlines()\n",
    "\n",
    "config = {}\n",
    "\n",
    "for line in lines:\n",
    "    line = line.strip()\n",
    "    if line and not line.startswith(\"#\"):\n",
    "        key, value = line.split(\"=\", 1)\n",
    "        config[key] = value.strip()\n",
    "\n",
    "server_name = config.get(\"server_name\")\n",
    "srever_port = int(config.get(\"srever_port\"))\n",
    "ssh_password = config.get(\"ssh_password\")\n",
    "ssh_username = config.get(\"ssh_username\")\n",
    "db_adress = config.get(\"db_adress\")\n",
    "db_port = int(config.get(\"db_port\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4e19aa1c-49a3-48c4-b67f-34c94cef608d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46373\n"
     ]
    }
   ],
   "source": [
    "tunnel = SSHTunnelForwarder(\n",
    "    (server_name, srever_port),\n",
    "    ssh_password=ssh_password,\n",
    "    ssh_username=ssh_username,\n",
    "    remote_bind_address=(db_adress, db_port),\n",
    ")\n",
    "tunnel.start()\n",
    "print(tunnel.local_bind_port)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cef2275b-12d0-48f1-ad3c-ce4cdb9aa267",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "conn = connection.MySQLConnection(\n",
    "    user=\"db_user\",\n",
    "    password=\"db_password\",\n",
    "    host=\"localhost\",\n",
    "    port=tunnel.local_bind_port,\n",
    "    database=\"db_name\",\n",
    ")\n",
    "cursor = conn.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f589ea8d-8b83-4331-a1cf-b64ed7621801",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('alternative_name',),\n",
       " ('histone',),\n",
       " ('histone_description',),\n",
       " ('histone_has_publication',),\n",
       " ('publication',),\n",
       " ('sequence',),\n",
       " ('sequence_has_publication',)]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SHOW TABLES;\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b3dd9ef8-b162-4127-9535-ac7743af8c69",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "add_histone = (\n",
    "    \"INSERT INTO histone \"\n",
    "    \"(id, level, taxonomic_span, taxonomic_span_id, description, parent) \"\n",
    "    \"VALUES (%(id)s, %(level)s, %(taxonomic_span)s, %(taxonomic_span_id)s, %(description)s, %(parent)s)\"\n",
    ")\n",
    "add_histone_description = (\n",
    "    \"INSERT INTO histone_description \"\n",
    "    \"(summary, taxonomy, genes, evolution, expression, knock_out, function, sequence, localization, deposition, structure, interactions, disease, caveats) \"\n",
    "    \"VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)\"\n",
    ")\n",
    "# add_publication = (\n",
    "#     \"INSERT INTO publication \"\n",
    "#     \"(id, title, doi, author, year) \"\n",
    "#     \"VALUES (%(id)s, %(title)s, %(doi)s, %(author)s, %(year)s)\"\n",
    "# )\n",
    "# add_sequence = (\n",
    "#     \"INSERT INTO sequence \"\n",
    "#     \"(accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration) \"\n",
    "#     \"VALUES (%(accession)s, %(variant)s, %(gi)s, %(ncbi_gene_id)s, %(hgnc_gene_name)s, %(taxonomy_id)s, %(organism)s, %(phylum)s, %(class)s, %(taxonomy_group)s, %(info)s, %(sequence)s, %(variant_under_consideration)s)\"\n",
    "# )\n",
    "# add_sequence_has_publication = (\n",
    "#     \"INSERT INTO sequence_has_publication \"\n",
    "#     \"(sequence_accession, publication_id) \"\n",
    "#     \"VALUES (%s, %s)\"\n",
    "# )\n",
    "add_alternate_names = (\n",
    "    \"INSERT INTO alternative_name \"\n",
    "    \"(name, taxonomy, gene, splice, histone) \"\n",
    "    \"VALUES (%(name)s, %(taxonomy)s, %(gene)s, %(splice)s, %(histone)s)\"\n",
    ")\n",
    "# add_histone_has_publication = (\n",
    "#     \"INSERT INTO histone_has_publication \"\n",
    "#     \"(histone_id, publication_id) \"\n",
    "#     \"VALUES (%s, %s)\"\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6011bb8a-bab5-4fb5-b4ae-15b83accd15b",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_histone_description = {\n",
    "    \"summary\": \"\",\n",
    "    \"taxonomy\": \"\",\n",
    "    \"genes\": \"\",\n",
    "    \"evolution\": \"\",\n",
    "    \"expression\": \"\",\n",
    "    \"knock_out\": \"\",\n",
    "    \"function\": \"\",\n",
    "    \"sequence\": \"\",\n",
    "    \"localization\": \"\",\n",
    "    \"deposition\": \"\",\n",
    "    \"structure\": \"\",\n",
    "    \"interactions\": \"\",\n",
    "    \"disease\": \"\",\n",
    "    \"caveats\": \"\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9a15b106-5cf1-40d4-8597-2b11dc30ff33",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>65</td>\n",
       "      <td>cid</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>93</td>\n",
       "      <td>cid</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Animals)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>99</td>\n",
       "      <td>cid</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Insecta)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>109</td>\n",
       "      <td>cid</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3_(Drosophilidae)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    id name taxonomy  gene splice                histone\n",
       "0   65  cid     None  None   None                  cenH3\n",
       "1   93  cid     None  None   None        cenH3_(Animals)\n",
       "2   99  cid     None  None   None        cenH3_(Insecta)\n",
       "3  109  cid     None  None   None  cenH3_(Drosophilidae)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name \" \"WHERE name='cid'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d0d3d7e-47ed-4f2b-8e96-e61eb030872d",
   "metadata": {},
   "source": [
    "# To Do H4\n",
    "\n",
    "## <span style=\"color:green\">Add a new nodes cH4_(Animals), cH4_(Plants), cH4_(Fungi), cH4_(Protists) to cH4</span>\n",
    "\n",
    "## <span style=\"color:green\">Add a new node cH4_(Vertebrata) to cH4_(Animals)</span>\n",
    "\n",
    "## <span style=\"color:green\">Add a new node cH4_(Mammalia) to cH4_(Vertebrata)</span>\n",
    "\n",
    "## <span style=\"color:green\">Change parent node of cH4_(Homo_sapiens) to cH4_(Mammalia)</span>\n",
    "\n",
    "## <span style=\"color:green\">Add a new node H4.V_(Trypanosomatidae) to H4</span>\n",
    "\n",
    "## <span style=\"color:green\">Add description to H4.V_(Trypanosomatidae)</span>\n",
    "\n",
    "### <span style=\"color:green\">Add summary</span>\n",
    "\n",
    "```H4.V_(Trypanosomatidae) is a histone variant characterized so far in some trypanosomatids (T. brucei and T. cruzi). It shares ~98% sequence identity with canonical H4. H4.V co-localizes with the histone variant H3.V and DNA base J at transcription termination sites (TTSs), where it is a key factor in ensuring efficient transcription termination [kim_genetic_2021].```\n",
    "\n",
    "### <span style=\"color:green\">Add gene</span>\n",
    "\n",
    "```H4.V encoded by a single-copy gene located apart from the typical  30 tandem arrays of canonical histone H4 [roson_new_2025].```\n",
    "\n",
    "### <span style=\"color:green\">Add function</span>\n",
    "\n",
    "```H4.V is the major signal for RNA polymerase II transcription termination at TTSs and, together with H3.V, plays a role in repressing transcription of telomeric variant surface glycoprotein genes [kim_genetic_2021]. The enrichment of H4.V at telomeres and TTSs, coupled with its increased expression in metacyclic forms, suggests its role in regulating the expression of virulence genes and adaptation to different host environments [roson_new_2025].```\n",
    "\n",
    "### <span style=\"color:green\">Add knock_out</span>\n",
    "\n",
    "```The H4.V knockout in T. brucei is viable, and the knockout cells grow normally without morphological abnormalities [kim_genetic_2021]. However, these mutants exhibit a significant increase in antisense transcription. The simultaneous knockout of H4.V and H3.V leads to severe defects in growth, DNA replication, and cell-cycle progression, which are exacerbated in the triple knockout with base J, indicating a synthetic lethal genetic interaction among these three epigenetic marks [kim_genetic_2021].```\n",
    "\n",
    "### <span style=\"color:green\">Add expression</span>\n",
    "\n",
    "```H4.V expression analysis in T. cruzi demonstrated that H4.V transcript levels remain stable throughout the cell cycle, while protein abundance increases during the G2/M phase [roson_new_2025]. H4.V is more abundant in the nuclei of metacyclic trypomastigotes but is barely detectable in amastigotes and bloodstream trypomastigotes. During metacyclogenesis, its expression progressively increases, suggesting a role in parasite differentiation [roson_new_2025].```\n",
    "\n",
    "### <span style=\"color:green\">Add localization</span>\n",
    "\n",
    "```While the core function of H4.V in marking transcription termination sites is conserved in trypanosomes, its genomic localization exhibits significant species-specificity between T. brucei and T. cruzi. In T. brucei, H4.V is specifically localized at transcription termination sites that demarcate polycistronic transcription units and is absent from telomeric repeats, where the histone variant H3.V and base J are instead enriched [kim_genetic_2021]. In contrast, T. cruzi H4.V displays a unique localization pattern, occupying not only convergent strand-switch regions but also telomeric regions and specific tDNA loci [roson_new_2025].```\n",
    "\n",
    "## <span style=\"color:green\">Add description to cH4_(Animals)</span>\n",
    "\n",
    "### <span style=\"color:green\">Add summary</span>\n",
    "\n",
    "```cH4_(Animals) clustered H4 histones in animals (Metazoa), often called canonical, replication-dependent, replication-coupled or \"bulk\" H4 histones are a major class of H4 histones in animals with high sequence similarity across all animal species.```\n",
    "\n",
    "### <span style=\"color:green\">Add sequence</span>\n",
    "\n",
    "```The threonine 71 position of histone H4 is highly conserved across numerous animal species and can be subjected to O-GlcNAcylation (H4T71Gc). This modification plays a critical role, in concert with H3K9me3, in the progression and maintenance of condensed chromatin [hayakawa_novel_2025].```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "258affd6-3be1-4674-995a-b2e25fe4b6cd",
   "metadata": {},
   "source": [
    "# To Do H2B\n",
    "\n",
    "## <span style=\"color:green\">Add a alternate name H2B.E to cH2B.E_(Mus_musculus)</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cb7cf91-36b2-48fc-a393-7d202f9a15bb",
   "metadata": {},
   "source": [
    "# DONE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8938343d-5532-4936-b961-f090af75cb3a",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add a new nodes cH4_(Animals), cH4_(Plants), cH4_(Fungi), cH4_(Protists) to cH4</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "03d6f644-f0ab-4a68-b7a8-22b4e93e0801",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH4</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Eukaryotes</td>\n",
       "      <td>2759</td>\n",
       "      <td>59</td>\n",
       "      <td>H4</td>\n",
       "      <td>59</td>\n",
       "      <td>Canonical histones are replication-dependent w...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    id          level taxonomic_span taxonomic_span_id  description parent  \\\n",
       "0  cH4  variant_group     Eukaryotes              2759           59     H4   \n",
       "\n",
       "   id                                            summary taxonomy genes  ...  \\\n",
       "0  59  Canonical histones are replication-dependent w...     null  null  ...   \n",
       "\n",
       "  knock_out function sequence localization deposition structure interactions  \\\n",
       "0      null     null     null         null       null      null         null   \n",
       "\n",
       "  disease caveats relations  \n",
       "0    null    null      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cH4'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2cf55af1-9a51-4ac1-bc7d-e4ab516b8abb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH4_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>109</td>\n",
       "      <td>cH4</td>\n",
       "      <td>109</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "0  cH4_(Homo_sapiens)  variant   Homo sapiens              9606          109   \n",
       "\n",
       "  parent   id summary taxonomy genes  ... knock_out function sequence  \\\n",
       "0    cH4  109    null     null  null  ...      null     null     null   \n",
       "\n",
       "  localization deposition structure interactions disease caveats relations  \n",
       "0         null       null      null         null    null    null      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.parent='cH4'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "884955f5-f904-422a-ad2e-783881d539a0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"cH4_(Animals)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Metazoa\",\n",
    "    \"taxonomic_span_id\": \"33208\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cH4\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "data_histone = {\n",
    "    \"id\": \"cH4_(Plants)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Viridiplantae\",\n",
    "    \"taxonomic_span_id\": \"33090\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cH4\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "data_histone = {\n",
    "    \"id\": \"cH4_(Fungi)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Fungi\",\n",
    "    \"taxonomic_span_id\": \"4751\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cH4\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)\n",
    "\n",
    "data_histone = {\n",
    "    \"id\": \"cH4_(Protists)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"SAR,Metamonada,Discoba,Amoebozoa\",\n",
    "    \"taxonomic_span_id\": \"2698737,2611341,2611352,554915\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cH4\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "292e09cb-f885-4ef5-8aea-11caac5cc951",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH4_(Animals)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>33208</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cH4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cH4_(Fungi)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>4751</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cH4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cH4_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>109.0</td>\n",
       "      <td>cH4</td>\n",
       "      <td>109.0</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cH4_(Plants)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Viridiplantae</td>\n",
       "      <td>33090</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cH4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cH4_(Protists)</td>\n",
       "      <td>variant</td>\n",
       "      <td>SAR,Metamonada,Discoba,Amoebozoa</td>\n",
       "      <td>2698737,2611341,2611352,554915</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cH4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   id    level                    taxonomic_span  \\\n",
       "0       cH4_(Animals)  variant                           Metazoa   \n",
       "1         cH4_(Fungi)  variant                             Fungi   \n",
       "2  cH4_(Homo_sapiens)  variant                      Homo sapiens   \n",
       "3        cH4_(Plants)  variant                     Viridiplantae   \n",
       "4      cH4_(Protists)  variant  SAR,Metamonada,Discoba,Amoebozoa   \n",
       "\n",
       "                taxonomic_span_id  description parent     id summary taxonomy  \\\n",
       "0                           33208          NaN    cH4    NaN    None     None   \n",
       "1                            4751          NaN    cH4    NaN    None     None   \n",
       "2                            9606        109.0    cH4  109.0    null     null   \n",
       "3                           33090          NaN    cH4    NaN    None     None   \n",
       "4  2698737,2611341,2611352,554915          NaN    cH4    NaN    None     None   \n",
       "\n",
       "  genes  ... knock_out function sequence localization deposition structure  \\\n",
       "0  None  ...      None     None     None         None       None      None   \n",
       "1  None  ...      None     None     None         None       None      None   \n",
       "2  null  ...      null     null     null         null       null      null   \n",
       "3  None  ...      None     None     None         None       None      None   \n",
       "4  None  ...      None     None     None         None       None      None   \n",
       "\n",
       "  interactions disease caveats relations  \n",
       "0         None    None    None      None  \n",
       "1         None    None    None      None  \n",
       "2         null    null    null      None  \n",
       "3         None    None    None      None  \n",
       "4         None    None    None      None  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.parent='cH4'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9ac5da99-2996-4a63-9bc4-5ff8c129c3d9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3230cf6-ca1e-4f5c-a19f-257c198c8e2a",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add a new node cH4_(Vertebrata) to cH4_(Animals)</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7fb588f0-cccd-4321-a9e1-98c5bc7a1cad",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, level, taxonomic_span, taxonomic_span_id, description, parent, id, summary, taxonomy, genes, evolution, expression, knock_out, function, sequence, localization, deposition, structure, interactions, disease, caveats, relations]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 22 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.parent='cH4_(Animals)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "334534bb-d5f5-4524-9ec3-7d6e3c0bcd24",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"cH4_(Vertebrata)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Vertebrata\",\n",
    "    \"taxonomic_span_id\": \"7742\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cH4_(Animals)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6b9ff4d7-5b96-4585-880e-9b8e641cfa57",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH4_(Vertebrata)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Vertebrata</td>\n",
       "      <td>7742</td>\n",
       "      <td>None</td>\n",
       "      <td>cH4_(Animals)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  cH4_(Vertebrata)  variant     Vertebrata              7742        None   \n",
       "\n",
       "          parent    id summary taxonomy genes  ... knock_out function  \\\n",
       "0  cH4_(Animals)  None    None     None  None  ...      None     None   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     None         None       None      None         None    None    None   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.parent='cH4_(Animals)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b281bf26-2c4d-4230-a83b-6682c9e65813",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "968ba481-4bdb-4a22-8ac4-33215eead424",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add a new node cH4_(Mammalia) to cH4_(Vertebrata)</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "58355a48-cc74-48c8-ac2b-977387d42fe9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, level, taxonomic_span, taxonomic_span_id, description, parent, id, summary, taxonomy, genes, evolution, expression, knock_out, function, sequence, localization, deposition, structure, interactions, disease, caveats, relations]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 22 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.parent='cH4_(Vertebrata)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "fd1f5a7f-b1c6-410d-9d28-03815e950fee",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"cH4_(Mammalia)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"Mammalia\",\n",
    "    \"taxonomic_span_id\": \"40674\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cH4_(Vertebrata)\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "b35a9801-d1ba-4d58-afba-97eb291c8281",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH4_(Mammalia)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>None</td>\n",
       "      <td>cH4_(Vertebrata)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               id    level taxonomic_span taxonomic_span_id description  \\\n",
       "0  cH4_(Mammalia)  variant       Mammalia             40674        None   \n",
       "\n",
       "             parent    id summary taxonomy genes  ... knock_out function  \\\n",
       "0  cH4_(Vertebrata)  None    None     None  None  ...      None     None   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     None         None       None      None         None    None    None   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.parent='cH4_(Vertebrata)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "9212e7e4-a9a5-47b7-b4f0-95ab9d1620e6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d402b183-a43d-445a-81a3-ea47d44e1524",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Change parent node of cH4_(Homo_sapiens) to cH4_(Mammalia)</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "adb633e7-4e8f-45ec-ab32-2eb552295396",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, level, taxonomic_span, taxonomic_span_id, description, parent, id, summary, taxonomy, genes, evolution, expression, knock_out, function, sequence, localization, deposition, structure, interactions, disease, caveats, relations]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 22 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.parent='cH4_(Mammalia)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "0d939691-5fa3-421d-b717-a077ee4504c5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH4_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>109</td>\n",
       "      <td>cH4</td>\n",
       "      <td>109</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "0  cH4_(Homo_sapiens)  variant   Homo sapiens              9606          109   \n",
       "\n",
       "  parent   id summary taxonomy genes  ... knock_out function sequence  \\\n",
       "0    cH4  109    null     null  null  ...      null     null     null   \n",
       "\n",
       "  localization deposition structure interactions disease caveats relations  \n",
       "0         null       null      null         null    null    null      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cH4_(Homo_sapiens)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e4648a24-3ca0-4688-ba3a-8dbc47dfb0e1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "query = f\"UPDATE histone SET parent='cH4_(Mammalia)' WHERE id='cH4_(Homo_sapiens)'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4096b594-49a8-476a-9b90-f27c14197d29",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH4_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>109</td>\n",
       "      <td>cH4_(Mammalia)</td>\n",
       "      <td>109</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "0  cH4_(Homo_sapiens)  variant   Homo sapiens              9606          109   \n",
       "\n",
       "           parent   id summary taxonomy genes  ... knock_out function  \\\n",
       "0  cH4_(Mammalia)  109    null     null  null  ...      null     null   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     null         null       null      null         null    null    null   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.parent='cH4_(Mammalia)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "b142c7d5-30f9-4e44-9380-5f17f9d43627",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "431b7893-8a24-4867-915b-f8f7a3c3dda5",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add a new node H4.V_(Trypanosomatidae) to H4</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "f7004fd8-57c0-49c0-af06-69cd39bd6cd1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, level, taxonomic_span, taxonomic_span_id, description, parent, id, summary, taxonomy, genes, evolution, expression, knock_out, function, sequence, localization, deposition, structure, interactions, disease, caveats, relations]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 22 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='H4.V_(Trypanosomatidae)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "f8fe1aac-6992-45b2-b5fc-7829117f09c1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"H4.V_(Trypanosomatidae)\",\n",
    "    \"level\": \"variant_group\",\n",
    "    \"taxonomic_span\": \"Trypanosomes\",\n",
    "    \"taxonomic_span_id\": \"93954\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"H4\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "d0a880e5-7d4f-4e78-9490-5d8864d03f77",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H4.V_(Trypanosomatidae)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Trypanosomes</td>\n",
       "      <td>93954</td>\n",
       "      <td>None</td>\n",
       "      <td>H4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        id          level taxonomic_span taxonomic_span_id  \\\n",
       "0  H4.V_(Trypanosomatidae)  variant_group   Trypanosomes             93954   \n",
       "\n",
       "  description parent    id summary taxonomy genes  ... knock_out function  \\\n",
       "0        None     H4  None    None     None  None  ...      None     None   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     None         None       None      None         None    None    None   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='H4.V_(Trypanosomatidae)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "ec7f13a6-d615-4c54-bfea-d05aeeda8e39",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2818a97-40f3-47e5-bf7a-4bd6e929a89b",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add description to H4.V_(Trypanosomatidae)</span>\n",
    "\n",
    "### <span style=\"color:black\">Add summary</span>\n",
    "\n",
    "```H4.V_(Trypanosomatidae) is a histone variant characterized so far in some trypanosomatids (T. brucei and T. cruzi). It shares ~98% sequence identity with canonical H4. H4.V co-localizes with the histone variant H3.V and DNA base J at transcription termination sites (TTSs), where it is a key factor in ensuring efficient transcription termination [kim_genetic_2021].```\n",
    "\n",
    "### <span style=\"color:black\">Add gene</span>\n",
    "\n",
    "```H4.V encoded by a single-copy gene located apart from the typical  30 tandem arrays of canonical histone H4 [roson_new_2025].```\n",
    "\n",
    "### <span style=\"color:black\">Add function</span>\n",
    "\n",
    "```H4.V is the major signal for RNA polymerase II transcription termination at TTSs and, together with H3.V, plays a role in repressing transcription of telomeric variant surface glycoprotein genes [kim_genetic_2021]. The enrichment of H4.V at telomeres and TTSs, coupled with its increased expression in metacyclic forms, suggests its role in regulating the expression of virulence genes and adaptation to different host environments [roson_new_2025].```\n",
    "\n",
    "### <span style=\"color:black\">Add knock_out</span>\n",
    "\n",
    "```The H4.V knockout in T. brucei is viable, and the knockout cells grow normally without morphological abnormalities [kim_genetic_2021]. However, these mutants exhibit a significant increase in antisense transcription. The simultaneous knockout of H4.V and H3.V leads to severe defects in growth, DNA replication, and cell-cycle progression, which are exacerbated in the triple knockout with base J, indicating a synthetic lethal genetic interaction among these three epigenetic marks [kim_genetic_2021].```\n",
    "\n",
    "### <span style=\"color:black\">Add expression</span>\n",
    "\n",
    "```H4.V expression analysis in T. cruzi demonstrated that H4.V transcript levels remain stable throughout the cell cycle, while protein abundance increases during the G2/M phase [roson_new_2025]. H4.V is more abundant in the nuclei of metacyclic trypomastigotes but is barely detectable in amastigotes and bloodstream trypomastigotes. During metacyclogenesis, its expression progressively increases, suggesting a role in parasite differentiation [roson_new_2025].```\n",
    "\n",
    "### <span style=\"color:black\">Add localization</span>\n",
    "\n",
    "```While the core function of H4.V in marking transcription termination sites is conserved in trypanosomes, its genomic localization exhibits significant species-specificity between T. brucei and T. cruzi. In T. brucei, H4.V is specifically localized at transcription termination sites that demarcate polycistronic transcription units and is absent from telomeric repeats, where the histone variant H3.V and base J are instead enriched [kim_genetic_2021]. In contrast, T. cruzi H4.V displays a unique localization pattern, occupying not only convergent strand-switch regions but also telomeric regions and specific tDNA loci [roson_new_2025].```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6e69dc25-f32e-446f-8ab9-76ede1110bcc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H4.V_(Trypanosomatidae)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Trypanosomes</td>\n",
       "      <td>93954</td>\n",
       "      <td>None</td>\n",
       "      <td>H4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        id          level taxonomic_span taxonomic_span_id  \\\n",
       "0  H4.V_(Trypanosomatidae)  variant_group   Trypanosomes             93954   \n",
       "\n",
       "  description parent    id summary taxonomy genes  ... knock_out function  \\\n",
       "0        None     H4  None    None     None  None  ...      None     None   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     None         None       None      None         None    None    None   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='H4.V_(Trypanosomatidae)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "56f321be-d052-48b7-aa66-12ab656e7f45",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_histone_description = {\n",
    "    \"summary\": \"H4.V_(Trypanosomatidae) is a histone variant characterized so far in some trypanosomatids (T. brucei and T. cruzi). It shares ~98% sequence identity with canonical H4. H4.V co-localizes with the histone variant H3.V and DNA base J at transcription termination sites (TTSs), where it is a key factor in ensuring efficient transcription termination [kim_genetic_2021].\",\n",
    "    \"taxonomy\": None,\n",
    "    \"genes\": \"H4.V encoded by a single-copy gene located apart from the typical  30 tandem arrays of canonical histone H4 [roson_new_2025].\",\n",
    "    \"evolution\": None,\n",
    "    \"expression\": \"H4.V expression analysis in T. cruzi demonstrated that H4.V transcript levels remain stable throughout the cell cycle, while protein abundance increases during the G2/M phase [roson_new_2025]. H4.V is more abundant in the nuclei of metacyclic trypomastigotes but is barely detectable in amastigotes and bloodstream trypomastigotes. During metacyclogenesis, its expression progressively increases, suggesting a role in parasite differentiation [roson_new_2025].\",\n",
    "    \"knock_out\": \"The H4.V knockout in T. brucei is viable, and the knockout cells grow normally without morphological abnormalities [kim_genetic_2021]. However, these mutants exhibit a significant increase in antisense transcription. The simultaneous knockout of H4.V and H3.V leads to severe defects in growth, DNA replication, and cell-cycle progression, which are exacerbated in the triple knockout with base J, indicating a synthetic lethal genetic interaction among these three epigenetic marks [kim_genetic_2021].\",\n",
    "    \"function\": \"H4.V is the major signal for RNA polymerase II transcription termination at TTSs and, together with H3.V, plays a role in repressing transcription of telomeric variant surface glycoprotein genes [kim_genetic_2021]. The enrichment of H4.V at telomeres and TTSs, coupled with its increased expression in metacyclic forms, suggests its role in regulating the expression of virulence genes and adaptation to different host environments [roson_new_2025].\",\n",
    "    \"sequence\": None,\n",
    "    \"localization\": \"While the core function of H4.V in marking transcription termination sites is conserved in trypanosomes, its genomic localization exhibits significant species-specificity between T. brucei and T. cruzi. In T. brucei, H4.V is specifically localized at transcription termination sites that demarcate polycistronic transcription units and is absent from telomeric repeats, where the histone variant H3.V and base J are instead enriched [kim_genetic_2021]. In contrast, T. cruzi H4.V displays a unique localization pattern, occupying not only convergent strand-switch regions but also telomeric regions and specific tDNA loci [roson_new_2025].\",\n",
    "    \"deposition\": None,\n",
    "    \"structure\": None,\n",
    "    \"interactions\": None,\n",
    "    \"disease\": None,\n",
    "    \"caveats\": None,\n",
    "}\n",
    "cursor.execute(add_histone_description, tuple(data_histone_description.values()))\n",
    "\n",
    "histone_description_id = cursor.lastrowid\n",
    "query = (\n",
    "    f\"UPDATE histone SET description={histone_description_id} WHERE id='H4.V_(Trypanosomatidae)'\"\n",
    ")\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "6a8f091d-ce7d-4dba-a0b9-90fe83c2d39c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H4.V_(Trypanosomatidae)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Trypanosomes</td>\n",
       "      <td>93954</td>\n",
       "      <td>256</td>\n",
       "      <td>H4</td>\n",
       "      <td>256</td>\n",
       "      <td>H4.V_(Trypanosomatidae) is a histone variant c...</td>\n",
       "      <td>None</td>\n",
       "      <td>H4.V encoded by a single-copy gene located apa...</td>\n",
       "      <td>...</td>\n",
       "      <td>The H4.V knockout in T. brucei is viable, and ...</td>\n",
       "      <td>H4.V is the major signal for RNA polymerase II...</td>\n",
       "      <td>None</td>\n",
       "      <td>While the core function of H4.V in marking tra...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        id          level taxonomic_span taxonomic_span_id  \\\n",
       "0  H4.V_(Trypanosomatidae)  variant_group   Trypanosomes             93954   \n",
       "\n",
       "   description parent   id                                            summary  \\\n",
       "0          256     H4  256  H4.V_(Trypanosomatidae) is a histone variant c...   \n",
       "\n",
       "  taxonomy                                              genes  ...  \\\n",
       "0     None  H4.V encoded by a single-copy gene located apa...  ...   \n",
       "\n",
       "                                           knock_out  \\\n",
       "0  The H4.V knockout in T. brucei is viable, and ...   \n",
       "\n",
       "                                            function sequence  \\\n",
       "0  H4.V is the major signal for RNA polymerase II...     None   \n",
       "\n",
       "                                        localization deposition structure  \\\n",
       "0  While the core function of H4.V in marking tra...       None      None   \n",
       "\n",
       "  interactions disease caveats relations  \n",
       "0         None    None    None      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='H4.V_(Trypanosomatidae)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "00efd5cb-84d0-4f58-877f-cbb0b76ecd67",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ed3eaf9-c577-437f-b424-1ced35ef4c38",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add description to cH4_(Animals)</span>\n",
    "\n",
    "### <span style=\"color:black\">Add summary</span>\n",
    "\n",
    "```cH4_(Animals) clustered H4 histones in animals (Metazoa), often called canonical, replication-dependent, replication-coupled or \"bulk\" H4 histones are a major class of H4 histones in animals with high sequence similarity across all animal species.```\n",
    "\n",
    "### <span style=\"color:black\">Add sequence</span>\n",
    "\n",
    "```The threonine 71 position of histone H4 is highly conserved across numerous animal species and can be subjected to O-GlcNAcylation (H4T71Gc). This modification plays a critical role, in concert with H3K9me3, in the progression and maintenance of condensed chromatin [hayakawa_novel_2025].```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "f79ff3f6-fb4b-4208-95c5-c17d7faaa780",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH4_(Animals)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>33208</td>\n",
       "      <td>None</td>\n",
       "      <td>cH4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              id    level taxonomic_span taxonomic_span_id description parent  \\\n",
       "0  cH4_(Animals)  variant        Metazoa             33208        None    cH4   \n",
       "\n",
       "     id summary taxonomy genes  ... knock_out function sequence localization  \\\n",
       "0  None    None     None  None  ...      None     None     None         None   \n",
       "\n",
       "  deposition structure interactions disease caveats relations  \n",
       "0       None      None         None    None    None      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cH4_(Animals)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "e5483ac6-d66a-4ed1-95f8-34298f6057e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_histone_description = {\n",
    "    \"summary\": 'cH4_(Animals) clustered H4 histones in animals (Metazoa), often called canonical, replication-dependent, replication-coupled or \"bulk\" H4 histones are a major class of H4 histones in animals with high sequence similarity across all animal species.',\n",
    "    \"taxonomy\": None,\n",
    "    \"genes\": None,\n",
    "    \"evolution\": None,\n",
    "    \"expression\": None,\n",
    "    \"knock_out\": None,\n",
    "    \"function\": None,\n",
    "    \"sequence\": \"The threonine 71 position of histone H4 is highly conserved across numerous animal species and can be subjected to O-GlcNAcylation (H4T71Gc). This modification plays a critical role, in concert with H3K9me3, in the progression and maintenance of condensed chromatin [hayakawa_novel_2025].\",\n",
    "    \"localization\": None,\n",
    "    \"deposition\": None,\n",
    "    \"structure\": None,\n",
    "    \"interactions\": None,\n",
    "    \"disease\": None,\n",
    "    \"caveats\": None,\n",
    "}\n",
    "cursor.execute(add_histone_description, tuple(data_histone_description.values()))\n",
    "\n",
    "histone_description_id = cursor.lastrowid\n",
    "query = (\n",
    "    f\"UPDATE histone SET description={histone_description_id} WHERE id='cH4_(Animals)'\"\n",
    ")\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "998bde46-30ba-43b3-b2a4-490651522763",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH4_(Animals)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>33208</td>\n",
       "      <td>257</td>\n",
       "      <td>cH4</td>\n",
       "      <td>257</td>\n",
       "      <td>cH4_(Animals) clustered H4 histones in animals...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>The threonine 71 position of histone H4 is hig...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              id    level taxonomic_span taxonomic_span_id  description  \\\n",
       "0  cH4_(Animals)  variant        Metazoa             33208          257   \n",
       "\n",
       "  parent   id                                            summary taxonomy  \\\n",
       "0    cH4  257  cH4_(Animals) clustered H4 histones in animals...     None   \n",
       "\n",
       "  genes  ... knock_out function  \\\n",
       "0  None  ...      None     None   \n",
       "\n",
       "                                            sequence localization deposition  \\\n",
       "0  The threonine 71 position of histone H4 is hig...         None       None   \n",
       "\n",
       "  structure interactions disease caveats relations  \n",
       "0      None         None    None    None      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cH4_(Animals)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "429e67a9-19e6-4472-a5bc-32c545dac26e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d0b367a-188e-41cd-9b34-2ed6196a3067",
   "metadata": {
    "tags": []
   },
   "source": [
    "# To Do H2B\n",
    "\n",
    "## <span style=\"color:black\">Add a alternate name H2B.E to cH2B.E_(Mus_musculus)</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "89f810c4-1059-4ec1-9d1e-850f730025d6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>89</td>\n",
       "      <td>H2B.21</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH2B.E_(Mus_musculus)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id    name taxonomy  gene splice                histone\n",
       "0  89  H2B.21     None  None   None  cH2B.E_(Mus_musculus)"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM alternative_name \"\n",
    "    \"WHERE histone='cH2B.E_(Mus_musculus)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "0e9c347a-9f57-4e22-90cc-a2a826b826b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_alternate_name = {\n",
    "    \"name\": \"H2B.E\",\n",
    "    \"taxonomy\": None,\n",
    "    \"gene\": None,\n",
    "    \"splice\": None,\n",
    "    \"histone\": \"cH2B.E_(Mus_musculus)\",\n",
    "}\n",
    "cursor.execute(add_alternate_names, data_alternate_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "6aa48161-3ace-4172-a5ce-45bdfa5907b4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>89</td>\n",
       "      <td>H2B.21</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH2B.E_(Mus_musculus)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>120</td>\n",
       "      <td>H2B.E</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH2B.E_(Mus_musculus)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    id    name taxonomy  gene splice                histone\n",
       "0   89  H2B.21     None  None   None  cH2B.E_(Mus_musculus)\n",
       "1  120   H2B.E     None  None   None  cH2B.E_(Mus_musculus)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM alternative_name \"\n",
    "    \"WHERE histone='cH2B.E_(Mus_musculus)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "b121780c-f39a-47d8-8937-fbf7b1f38b47",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e8bad07-ef74-40fd-a93b-e7897df0c591",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Close connections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "e5673501-5845-491a-bbb9-f5d213721d58",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cursor.close()\n",
    "conn.close()\n",
    "tunnel.stop()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".conda-histdb_env",
   "language": "python",
   "name": "conda-env-.conda-histdb_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
