{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "53d3ec55-7daf-4ade-8add-2830b709dc54",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import io\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import Entrez, SeqIO\n",
    "from mysql.connector import connection\n",
    "from sshtunnel import SSHTunnelForwarder\n",
    "\n",
    "# Укажите ваш email (обязательно для использования Entrez)\n",
    "Entrez.email = \"your.email@example.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "608c4bfe-b979-4b60-b2b5-b94eeb679cb5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open(\"db_curated_server_info.txt\", \"r\") as file:\n",
    "    lines = file.readlines()\n",
    "\n",
    "config = {}\n",
    "\n",
    "for line in lines:\n",
    "    line = line.strip()\n",
    "    if line and not line.startswith(\"#\"):\n",
    "        key, value = line.split(\"=\", 1)\n",
    "        config[key] = value.strip()\n",
    "\n",
    "server_name = config.get(\"server_name\")\n",
    "srever_port = int(config.get(\"srever_port\"))\n",
    "ssh_password = config.get(\"ssh_password\")\n",
    "ssh_username = config.get(\"ssh_username\")\n",
    "db_adress = config.get(\"db_adress\")\n",
    "db_port = int(config.get(\"db_port\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4e19aa1c-49a3-48c4-b67f-34c94cef608d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "36237\n"
     ]
    }
   ],
   "source": [
    "tunnel = SSHTunnelForwarder(\n",
    "    (server_name, srever_port),\n",
    "    ssh_password=ssh_password,\n",
    "    ssh_username=ssh_username,\n",
    "    remote_bind_address=(db_adress, db_port),\n",
    ")\n",
    "tunnel.start()\n",
    "print(tunnel.local_bind_port)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cef2275b-12d0-48f1-ad3c-ce4cdb9aa267",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "conn = connection.MySQLConnection(\n",
    "    user=\"db_user\",\n",
    "    password=\"db_password\",\n",
    "    host=\"localhost\",\n",
    "    port=tunnel.local_bind_port,\n",
    "    database=\"db_name\",\n",
    ")\n",
    "cursor = conn.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f589ea8d-8b83-4331-a1cf-b64ed7621801",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('alternative_name',),\n",
       " ('histone',),\n",
       " ('histone_description',),\n",
       " ('histone_has_publication',),\n",
       " ('publication',),\n",
       " ('sequence',),\n",
       " ('sequence_has_publication',)]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SHOW TABLES;\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "b3dd9ef8-b162-4127-9535-ac7743af8c69",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "add_histone = (\n",
    "    \"INSERT INTO histone \"\n",
    "    \"(id, level, taxonomic_span, taxonomic_span_id, description, parent) \"\n",
    "    \"VALUES (%(id)s, %(level)s, %(taxonomic_span)s, %(taxonomic_span_id)s, %(description)s, %(parent)s)\"\n",
    ")\n",
    "# add_histone_description = (\n",
    "#     \"INSERT INTO histone_description \"\n",
    "#     \"(summary, taxonomy, genes, evolution, expression, knock_out, function, sequence, localization, deposition, structure, interactions, disease, caveats) \"\n",
    "#     \"VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)\"\n",
    "# )\n",
    "add_publication = (\n",
    "    \"INSERT INTO publication \"\n",
    "    \"(id, title, doi, author, year) \"\n",
    "    \"VALUES (%(id)s, %(title)s, %(doi)s, %(author)s, %(year)s)\"\n",
    ")\n",
    "add_sequence = (\n",
    "    \"INSERT INTO sequence \"\n",
    "    \"(accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration) \"\n",
    "    \"VALUES (%(accession)s, %(variant)s, %(gi)s, %(ncbi_gene_id)s, %(hgnc_gene_name)s, %(taxonomy_id)s, %(organism)s, %(phylum)s, %(class)s, %(taxonomy_group)s, %(info)s, %(sequence)s, %(variant_under_consideration)s)\"\n",
    ")\n",
    "add_sequence_has_publication = (\n",
    "    \"INSERT INTO sequence_has_publication \"\n",
    "    \"(sequence_accession, publication_id) \"\n",
    "    \"VALUES (%s, %s)\"\n",
    ")\n",
    "add_alternate_names = (\n",
    "    \"INSERT INTO alternative_name \"\n",
    "    \"(name, taxonomy, gene, splice, histone) \"\n",
    "    \"VALUES (%(name)s, %(taxonomy)s, %(gene)s, %(splice)s, %(histone)s)\"\n",
    ")\n",
    "# add_histone_has_publication = (\n",
    "#     \"INSERT INTO histone_has_publication \"\n",
    "#     \"(histone_id, publication_id) \"\n",
    "#     \"VALUES (%s, %s)\"\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6011bb8a-bab5-4fb5-b4ae-15b83accd15b",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_histone_description = {\n",
    "    \"summary\": None,\n",
    "    \"taxonomy\": None,\n",
    "    \"genes\": None,\n",
    "    \"evolution\": None, \n",
    "    \"expression\": None,\n",
    "    \"knock_out\": None,\n",
    "    \"function\": None,\n",
    "    \"sequence\": None,\n",
    "    \"localization\": None,\n",
    "    \"deposition\": None,\n",
    "    \"structure\": None,\n",
    "    \"interactions\": None,\n",
    "    \"disease\": None,\n",
    "    \"caveats\": None,\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c1c3bb0c-180f-4fb2-9cdc-ca928f9808be",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_taxonomy_data(record):\n",
    "    import re\n",
    "    import sys\n",
    "\n",
    "    taxonomy_data = {}\n",
    "    taxonomy_data[\"organism\"] = record.annotations[\"organism\"]\n",
    "    try:\n",
    "        for a in record.features[0].qualifiers[\"db_xref\"]:\n",
    "            text = re.search(\"(\\S+):(\\S+)\", a).group(1)\n",
    "            taxid = re.search(\"(\\S+):(\\S+)\", a).group(2)\n",
    "            if text == \"taxon\":\n",
    "                print(f\"Fetched taxid from NCBI {taxid}\")\n",
    "                taxonomy_data[\"taxonomy_id\"] = int(taxid)\n",
    "            else:\n",
    "                continue\n",
    "    except:\n",
    "        print(\"!!!!!!Unable to get TAXID for this record setting it to 1\")\n",
    "        taxonomy_data[\"taxonomy_id\"] = 1  # unable to identify\n",
    "\n",
    "    lineage = dict()\n",
    "    for i in range(10):\n",
    "        try:\n",
    "            handle = Entrez.efetch(\n",
    "                id=taxonomy_data[\"taxonomy_id\"], db=\"taxonomy\", retmode=\"xml\"\n",
    "            )\n",
    "            tax_data = Entrez.read(handle)\n",
    "            lineage = {\n",
    "                d[\"Rank\"]: d[\"ScientificName\"]\n",
    "                for d in tax_data[0][\"LineageEx\"]\n",
    "                if d[\"Rank\"] in [\"class\", \"phylum\"]\n",
    "            }\n",
    "            break\n",
    "        except:\n",
    "            print(\n",
    "                \"Unexpected error: {}, Retrying, attempt {}\".format(\n",
    "                    sys.exc_info()[0], i\n",
    "                )\n",
    "            )\n",
    "            if i == 9:\n",
    "                print(\n",
    "                    f\"FATAL ERROR could not get class and phylum from NCBI after 10 attempts for taxid:{taxonomy_data['taxonomy_id']}. Will add None for class and phylum!\"\n",
    "                )\n",
    "            else:\n",
    "                continue\n",
    "    taxonomy_data[\"phylum\"] = lineage.get(\"phylum\", None)\n",
    "    taxonomy_data[\"class\"] = lineage.get(\"class\", None)\n",
    "    if taxonomy_data[\"phylum\"] is not None:\n",
    "        taxonomy_data[\"phylum\"] = str(taxonomy_data[\"phylum\"])\n",
    "    if taxonomy_data[\"class\"] is not None:\n",
    "        taxonomy_data[\"class\"] = str(taxonomy_data[\"class\"])\n",
    "    return taxonomy_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2eec2479",
   "metadata": {},
   "source": [
    "# To Do H3\n",
    "\n",
    "## <span style=\"color:green\">Add sequences to H3-like_(Plants)</span>\n",
    "\n",
    "Sequences fasta [alvarez-venegas_canonical_2019, hu_identification_2015]:\n",
    "\n",
    "```fasta\n",
    ">HTR701 LOC_Os06g06480.1\n",
    "MRKGEEGGLDTDPHEADGVQVRDSRKQLATKATCKSAPATGGVKKPHRFRPGTVALREIR\n",
    "KYQKSTELLIRKLPFQRLVREIAQDFKTDLRFQSSAVAALQEAAEAYLVGLFEDTNLCAI\n",
    "HAKRVTIMPKDIQLARRIRGERA\n",
    "```\n",
    "> поправила последовательность из [Rice Genome Annotation Project database](https://rice.uga.edu) так, чтобы она соответствовала данным из статьи (см. рис. 3) [hu_identification_2015]\n",
    "\n",
    "```fasta\n",
    ">HTR709 LOC_Os02g25910.1\n",
    "MARTKQTARKSTGGKAPRKPLRAIAAVMPAPSSRGVARKSVPFIGVKKPRRHRPGTVALR\n",
    "EIRKYQKNTELLIRKLPFQRLVREIAQHFKHDMRFQSHAVLALQEAAEAYLVGLFEDTNL\n",
    "CAIHSKRVTIMSKDVQLARRIRGERL\n",
    "```\n",
    "> поправила последовательность из [Rice Genome Annotation Project database](https://rice.uga.edu) так, чтобы она соответствовала данным из статьи (см. рис. 3) [hu_identification_2015]\n",
    "\n",
    "## <span style=\"color:green\">Add sequences to H3.Y_(Primates)</span>\n",
    "\n",
    "### <span style=\"color:green\">Delete Macaque sequences to HISTDB_H3_Y_0 and HISTDB_H3_Y_1 [seal_standardized_2022]</span>\n",
    "\n",
    "### <span style=\"color:green\">Add Macaque sequence XP_001110711.1 as H3.Y.1_(Primates) [seal_standardized_2022]</span>\n",
    "\n",
    "### <span style=\"color:green\">Change sequence identificator and variant: HISTDB_H3_Y_2 -> XP_024212320.2 -> H3.Y.2_(Primates) [seal_standardized_2022]</span>\n",
    "\n",
    "### <span style=\"color:green\">Change sequence identificator and variant: HISTDB_H3_Y_3 -> HISTDB_H3_Y_0 -> H3.Y.1_(Primates) [wiedemann_identification_2010]</span>\n",
    "\n",
    "## <span style=\"color:green\">Add sequences to H3.4_(Mammalia)</span>\n",
    "\n",
    "### <span style=\"color:green\">Add sequence NP_001304932.1 as H3.4_(Mammalia) [ueda_testis-specific_2017]</span>\n",
    "\n",
    "### <span style=\"color:green\">Add fasta sequences to H3.4_(Mammalia)</span>\n",
    "\n",
    "Sequences fasta [dong_structural_2020]:\n",
    "\n",
    "```fasta\n",
    ">uncharacterized protein [rat]\n",
    "MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYHPGTVALREIRRYQKSTELLIRKLPFQR\n",
    "LVREIAQDFKTDLRFQSSAVMALQEACESYLVGLFEDTNLCAIHAKRVTIMPKDIQLARRIRGERA\n",
    "```\n",
    "\n",
    "## <span style=\"color:green\">Add sequences to H3.5_(Hominidae)</span>\n",
    "\n",
    "### <span style=\"color:green\">Add orangutan sequence XP_002823134.1, gorilla sequence ADW85799.1 as H3.5_(Hominidae) [schenk_h35_2011]</span>\n",
    "\n",
    "### <span style=\"color:green\">Delete chimpanzee sequence XP_003954426.1 (it is pseudogene [seal_standardized_2022])</span>\n",
    "\n",
    "### <span style=\"color:green\">Add alternate name H3.3C to H3.5_(Homo_sapiens)</span>\n",
    "\n",
    "## <span style=\"color:green\">Add sequences to H3.7_(Homo_sapiens)</span>\n",
    "\n",
    "### <span style=\"color:green\">Add sequence NP_001359034.1 to H3.7_(Homo_sapiens) [taguchi_crystal_2017, seal_standardized_2022]</span>\n",
    "\n",
    "## <span style=\"color:green\">Add sequences to H3.B_(Giardia)</span>\n",
    "\n",
    "### <span style=\"color:green\">Add sequence XP_767393.1 to H3.B_(Giardia) [dawson_cenh3_2007]</span>\n",
    "\n",
    "## <span style=\"color:green\">Add new node to cenH3 to cenH3_(Protists)</span>\n",
    "\n",
    "## <span style=\"color:green\">Add sequences to cenH3_(Protists)</span>\n",
    "\n",
    "### <span style=\"color:green\">Add sequence XP_771620.1 to cenH3_(Protists) [dawson_cenh3_2007]</span>\n",
    "\n",
    "## <span style=\"color:green\">Add sequences to cH3_(Protists)</span>\n",
    "\n",
    "### <span style=\"color:green\">Add sequence AAF00592.1 to cH3_(Protists) [dawson_cenh3_2007]</span>\n",
    "\n",
    "## <span style=\"color:green\">Add sequences to H3.P_(Euplotes_crassus)</span>\n",
    "\n",
    "### <span style=\"color:green\">Add sequence AAC47441.1 to H3.P_(Euplotes_crassus) [jahn_unusual_1997]</span>\n",
    "\n",
    "## <span style=\"color:green\">Add sequences to H3.V_(Trypanosomatidae)</span>\n",
    "\n",
    "### <span style=\"color:green\">Add sequence AAO24601.1 to H3.V_(Trypanosomatidae) [lowell_variant_2004]</span>\n",
    "\n",
    "### <span style=\"color:green\">Add fasta sequences to H3.V_(Trypanosomatidae)</span>\n",
    "\n",
    "Sequences fasta [lowell_variant_2004]:\n",
    "\n",
    "```fasta\n",
    ">ACC histone H3 variant [Trypanosoma cruzi]\n",
    "MGSLKKVASVEKHSAFSSESKLPKPRKNLASRKIHTEGRIVAKKESASAGTRKKHRWRPGTVVLREVRRYQSSTEFLIAKAP\n",
    "FRRLVREIVSNLKDSFRMSATCVEALQESTELYVTSVLADANLCTLHANRVTVYPKDIQLALKLRGERL\n",
    ">Q9U196 UniPrac [Leishmania major]\n",
    "MAGITKAAVVASHPKKNVASRKMNKKSRSIAKKEAKAMRADSAGAKSRRWRPGTVALREVRKYQRSTELLIARTPFRRLVKE\n",
    "IMSTFKDTMHMRHSALEAMQDATESYLVSLLCDANLCTIHAKRVTLYPKDLQLALRLRGERT\n",
    "```\n",
    "\n",
    "### <span style=\"color:green\">Add sequence XP_828007.1 to H3.V_(Trypanosomatidae) [schulz_base_2016]</span>\n",
    "\n",
    "### <span style=\"color:green\">Add fasta sequences to H3.V_(Trypanosomatidae)</span>\n",
    "\n",
    "Sequences fasta [siegel_four_2009]:\n",
    "\n",
    "```fasta\n",
    ">ACC histone H3 variant [Trypanosoma brucei]\n",
    "MAQMKITPRPVRPKSVASRPIQSVARAPVKKVENTPPQKRHHRWRPGTVALREIRRLQSSTDFLIQRAPFRRFLREVVSNLK\n",
    "DSYRMSAACVDAIQEATETYITSVFMDANLCTLHANRVTLFPKDIQLALKLRGERN\n",
    "```\n",
    "\n",
    "## <span style=\"color:green\">Add sequences to cH3_(Protists)</span>\n",
    "\n",
    "### <span style=\"color:green\">Add fasta sequences to cH3_(Protists)</span>\n",
    "\n",
    "Sequences fasta [lowell_variant_2004]:\n",
    "\n",
    "```fasta\n",
    ">ACC histone H3 variant [Trypanosoma brucei]\n",
    "MSRTKETARTKKTITSKKSKKASKGSDAASGVKTAQRRWRPGTVALREIROFQRSTDLLLQKAPFQRLVREVSGAQKEGLRF\n",
    "QSSAILAAQEATESYIVSLLADTNRACIHSGRVTIQPKDIHLALCLRGERA\n",
    "```\n",
    "\n",
    "## <span style=\"color:green\">Add sequences to H3.10_(Arabidopsis)</span>\n",
    "\n",
    "### <span style=\"color:green\">Add sequence NP_173418.1 to H3.10_(Arabidopsis) [borg_targeted_2020, alvarez-venegas_canonical_2019, okada_analysis_2005]</span>\n",
    "\n",
    "## <span style=\"color:green\">Add sequences to H3.14_(Arabidopsis)</span>\n",
    "\n",
    "### <span style=\"color:green\">Add sequence NP_177690.1 to H3.14_(Arabidopsis) [alvarez-venegas_canonical_2019, nunez-vazquez_histone_2025]</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "400863d2-b22b-42d8-8c82-39ee7ce8e8ec",
   "metadata": {},
   "source": [
    "# To Do H4\n",
    "\n",
    "## <span style=\"color:green\">Add sequences to H4.G_(Hominidae)</span>\n",
    "\n",
    "### <span style=\"color:green\">Add sequence NP_003538.1 as H4.G_(Hominidae) [ding_primate-specific_2021]</span>\n",
    "\n",
    "## <span style=\"color:green\">Add sequences to H4.V_(Trypanosomatidae)</span>\n",
    "\n",
    "### <span style=\"color:green\">Add fasta sequences to H4.V_(Trypanosomatidae)</span>\n",
    "\n",
    "Sequences fasta [siegel_four_2009]:\n",
    "\n",
    "```fasta\n",
    ">ACC histone H4 variant [Trypanosoma brucei]\n",
    "MAKGKRVGESKGAQKRQKKVLRDNVRGITRGSIRRLARRAGVKRISGVIYDEVRGVLKTFVESIVRDAGAYTEYSRKKTVTAAHVVFALRKRGKVLYGYD\n",
    "```\n",
    "\n",
    "## <span style=\"color:green\">Add sequences to cH4_(Protists)</span>\n",
    "\n",
    "### <span style=\"color:green\">Add fasta sequences to cH4_(Protists)</span>\n",
    "\n",
    "Sequences fasta [siegel_four_2009]:\n",
    "\n",
    "```fasta\n",
    ">ACC histone H4 variant [Trypanosoma brucei]\n",
    "MAKGKKSGEAKGSQKRQKKVLRENVRGITRGSIRRLARRGGVKRISGVIYDEVRGVLKSFVEGVVRDATAYTEYSRKKTVTAVDVVNALRKRGKILYGYA\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa1c96dd-dd11-43cd-8713-9ce54254b3e9",
   "metadata": {},
   "source": [
    "# DONE H3\n",
    "\n",
    "## <span style=\"color:black\">Add sequences to H3-like_(Plants)</span>\n",
    "\n",
    "Sequences fasta [alvarez-venegas_canonical_2019, hu_identification_2015]:\n",
    "\n",
    "```fasta\n",
    ">HISTDB_H3_like_Plants_0 HTR701 LOC_Os06g06480.1 [Oryza sativa]\n",
    "MRKGEEGGLDTDPHEADGVQVRDSRKQLATKATCKSAPATGGVKKPHRFRPGTVALREIR\n",
    "KYQKSTELLIRKLPFQRLVREIAQDFKTDLRFQSSAVAALQEAAEAYLVGLFEDTNLCAI\n",
    "HAKRVTIMPKDIQLARRIRGERA\n",
    "```\n",
    "> поправила последовательность из [Rice Genome Annotation Project database](https://rice.uga.edu) так, чтобы она соответствовала данным из статьи (см. рис. 3) [hu_identification_2015]\n",
    "\n",
    "```fasta\n",
    ">HISTDB_H3_like_Plants_1 HTR709 LOC_Os02g25910.1 [Oryza sativa]\n",
    "MARTKQTARKSTGGKAPRKPLRAIAAVMPAPSSRGVARKSVPFIGVKKPRRHRPGTVALR\n",
    "EIRKYQKNTELLIRKLPFQRLVREIAQHFKHDMRFQSHAVLALQEAAEAYLVGLFEDTNL\n",
    "CAIHSKRVTIMSKDVQLARRIRGERL\n",
    "```\n",
    "> поправила последовательность из [Rice Genome Annotation Project database](https://rice.uga.edu) так, чтобы она соответствовала данным из статьи (см. рис. 3) [hu_identification_2015]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "eb5e4f02-b6d2-4306-bb6c-f7b1759be612",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABA97899.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>39947</td>\n",
       "      <td>Oryza sativa Japonica Group</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTAKKSTAGNVPRKLLVMKVARKSAPMMAGLKKPHRFNPWI...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABA97902.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>39947</td>\n",
       "      <td>Oryza sativa Japonica Group</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTAKKSTASNVPRKLLVMKVARKSAPTMAGLKKPHRFKPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BAA96098.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARPRKEAPQRNLDRDENARQQPTEEPQDEAPRNQGRQQQQQRPPA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BAE48427.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARMKHTARMSTGGKAPRKQLASKALRKAPPPPTKGVKQPHHYHLR...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BAE48431.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARMKLNARMSTGGKAPRKQLAYKAVRKAAPPTIGVKLPNSYRPGD...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BAE48433.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSIPTGMGGMKRPRRYRPG...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>BAE48435.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSAPTTGGVKKPHRYRPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NP_172794.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQSARKSHGGKAPTKQLATKAARKSAPTTGGVKKPHRFRPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NP_196795.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARSNQTARKATGGKAPHFAMRVWQHSTPPLKKPYRYKPGTVALRE...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NP_201338.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARISTGGKAPRKQLAPKAARQSAPATGGVKKPHRFRPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession           variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0   ABA97899.1  H3-like_(Plants)  None         None           None   \n",
       "1   ABA97902.1  H3-like_(Plants)  None         None           None   \n",
       "2   BAA96098.1  H3-like_(Plants)  None         None           None   \n",
       "3   BAE48427.1  H3-like_(Plants)  None         None           None   \n",
       "4   BAE48431.1  H3-like_(Plants)  None         None           None   \n",
       "5   BAE48433.1  H3-like_(Plants)  None         None           None   \n",
       "6   BAE48435.1  H3-like_(Plants)  None         None           None   \n",
       "7  NP_172794.1  H3-like_(Plants)  None         None           None   \n",
       "8  NP_196795.1  H3-like_(Plants)  None         None           None   \n",
       "9  NP_201338.1  H3-like_(Plants)  None         None           None   \n",
       "\n",
       "   taxonomy_id                     organism        phylum          class  \\\n",
       "0        39947  Oryza sativa Japonica Group  Streptophyta  Magnoliopsida   \n",
       "1        39947  Oryza sativa Japonica Group  Streptophyta  Magnoliopsida   \n",
       "2         4690           Lilium longiflorum  Streptophyta  Magnoliopsida   \n",
       "3         4690           Lilium longiflorum  Streptophyta  Magnoliopsida   \n",
       "4         4690           Lilium longiflorum  Streptophyta  Magnoliopsida   \n",
       "5         4690           Lilium longiflorum  Streptophyta  Magnoliopsida   \n",
       "6         4690           Lilium longiflorum  Streptophyta  Magnoliopsida   \n",
       "7         3702         Arabidopsis thaliana  Streptophyta  Magnoliopsida   \n",
       "8         3702         Arabidopsis thaliana  Streptophyta  Magnoliopsida   \n",
       "9         3702         Arabidopsis thaliana  Streptophyta  Magnoliopsida   \n",
       "\n",
       "  taxonomy_group  info                                           sequence  \\\n",
       "0           None  None  MARTKQTAKKSTAGNVPRKLLVMKVARKSAPMMAGLKKPHRFNPWI...   \n",
       "1           None  None  MARTKQTAKKSTASNVPRKLLVMKVARKSAPTMAGLKKPHRFKPGT...   \n",
       "2           None  None  MARPRKEAPQRNLDRDENARQQPTEEPQDEAPRNQGRQQQQQRPPA...   \n",
       "3           None  None  MARMKHTARMSTGGKAPRKQLASKALRKAPPPPTKGVKQPHHYHLR...   \n",
       "4           None  None  MARMKLNARMSTGGKAPRKQLAYKAVRKAAPPTIGVKLPNSYRPGD...   \n",
       "5           None  None  MARTKQTARKSTGGKAPRKQLATKAARKSIPTGMGGMKRPRRYRPG...   \n",
       "6           None  None  MARTKQTARKSTGGKAPRKQLATKAARKSAPTTGGVKKPHRYRPGT...   \n",
       "7           None  None  MARTKQSARKSHGGKAPTKQLATKAARKSAPTTGGVKKPHRFRPGT...   \n",
       "8           None  None  MARSNQTARKATGGKAPHFAMRVWQHSTPPLKKPYRYKPGTVALRE...   \n",
       "9           None  None  MARTKQTARISTGGKAPRKQLAPKAARQSAPATGGVKKPHRFRPGT...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                        None  \n",
       "1                        None  \n",
       "2                        None  \n",
       "3                        None  \n",
       "4                        None  \n",
       "5                        None  \n",
       "6                        None  \n",
       "7                        None  \n",
       "8                        None  \n",
       "9                        None  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE variant='H3-like_(Plants)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ae5976c6-9768-4128-97eb-5904af26d605",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_sequence = {\n",
    "    \"accession\": \"HISTDB_H3_like_Plants_0\",\n",
    "    \"variant\": \"H3-like_(Plants)\",\n",
    "    \"gi\": None,\n",
    "    \"ncbi_gene_id\": None,\n",
    "    \"hgnc_gene_name\": None,\n",
    "    \"taxonomy_id\": 4530,\n",
    "    \"organism\": \"Oryza sativa\",\n",
    "    \"phylum\": \"Streptophyta\",\n",
    "    \"class\": \"Magnoliopsida\",\n",
    "    \"taxonomy_group\": None,\n",
    "    \"info\": None,\n",
    "    \"sequence\": \"MRKGEEGGLDTDPHEADGVQVRDSRKQLATKATCKSAPATGGVKKPHRFRPGTVALREIRKYQKSTELLIRKLPFQRLVREIAQDFKTDLRFQSSAVAALQEAAEAYLVGLFEDTNLCAIHAKRVTIMPKDIQLARRIRGERA\",\n",
    "    \"variant_under_consideration\": None,\n",
    "}\n",
    "cursor.execute(add_sequence, data_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9b687d56-f853-4b8c-a42f-4694482bd193",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_sequence = {\n",
    "    \"accession\": \"HISTDB_H3_like_Plants_1\",\n",
    "    \"variant\": \"H3-like_(Plants)\",\n",
    "    \"gi\": None,\n",
    "    \"ncbi_gene_id\": None,\n",
    "    \"hgnc_gene_name\": None,\n",
    "    \"taxonomy_id\": 4530,\n",
    "    \"organism\": \"Oryza sativa\",\n",
    "    \"phylum\": \"Streptophyta\",\n",
    "    \"class\": \"Magnoliopsida\",\n",
    "    \"taxonomy_group\": None,\n",
    "    \"info\": None,\n",
    "    \"sequence\": \"MARTKQTARKSTGGKAPRKPLRAIAAVMPAPSSRGVARKSVPFIGVKKPRRHRPGTVALREIRKYQKNTELLIRKLPFQRLVREIAQHFKHDMRFQSHAVLALQEAAEAYLVGLFEDTNLCAIHSKRVTIMSKDVQLARRIRGERL\",\n",
    "    \"variant_under_consideration\": None,\n",
    "}\n",
    "cursor.execute(add_sequence, data_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "68201732-8e44-4e2e-8d5c-ebe2b98b9c33",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABA97899.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>39947</td>\n",
       "      <td>Oryza sativa Japonica Group</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTAKKSTAGNVPRKLLVMKVARKSAPMMAGLKKPHRFNPWI...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABA97902.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>39947</td>\n",
       "      <td>Oryza sativa Japonica Group</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTAKKSTASNVPRKLLVMKVARKSAPTMAGLKKPHRFKPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BAA96098.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARPRKEAPQRNLDRDENARQQPTEEPQDEAPRNQGRQQQQQRPPA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BAE48427.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARMKHTARMSTGGKAPRKQLASKALRKAPPPPTKGVKQPHHYHLR...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BAE48431.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARMKLNARMSTGGKAPRKQLAYKAVRKAAPPTIGVKLPNSYRPGD...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BAE48433.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSIPTGMGGMKRPRRYRPG...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>BAE48435.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4690</td>\n",
       "      <td>Lilium longiflorum</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSAPTTGGVKKPHRYRPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HISTDB_H3_like_Plants_0</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4530</td>\n",
       "      <td>Oryza sativa</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MRKGEEGGLDTDPHEADGVQVRDSRKQLATKATCKSAPATGGVKKP...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HISTDB_H3_like_Plants_1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4530</td>\n",
       "      <td>Oryza sativa</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKPLRAIAAVMPAPSSRGVARKSVPFIGV...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NP_172794.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQSARKSHGGKAPTKQLATKAARKSAPTTGGVKKPHRFRPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NP_196795.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARSNQTARKATGGKAPHFAMRVWQHSTPPLKKPYRYKPGTVALRE...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NP_201338.1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARISTGGKAPRKQLAPKAARQSAPATGGVKKPHRFRPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  accession           variant    gi ncbi_gene_id  \\\n",
       "0                ABA97899.1  H3-like_(Plants)  None         None   \n",
       "1                ABA97902.1  H3-like_(Plants)  None         None   \n",
       "2                BAA96098.1  H3-like_(Plants)  None         None   \n",
       "3                BAE48427.1  H3-like_(Plants)  None         None   \n",
       "4                BAE48431.1  H3-like_(Plants)  None         None   \n",
       "5                BAE48433.1  H3-like_(Plants)  None         None   \n",
       "6                BAE48435.1  H3-like_(Plants)  None         None   \n",
       "7   HISTDB_H3_like_Plants_0  H3-like_(Plants)  None         None   \n",
       "8   HISTDB_H3_like_Plants_1  H3-like_(Plants)  None         None   \n",
       "9               NP_172794.1  H3-like_(Plants)  None         None   \n",
       "10              NP_196795.1  H3-like_(Plants)  None         None   \n",
       "11              NP_201338.1  H3-like_(Plants)  None         None   \n",
       "\n",
       "   hgnc_gene_name  taxonomy_id                     organism        phylum  \\\n",
       "0            None        39947  Oryza sativa Japonica Group  Streptophyta   \n",
       "1            None        39947  Oryza sativa Japonica Group  Streptophyta   \n",
       "2            None         4690           Lilium longiflorum  Streptophyta   \n",
       "3            None         4690           Lilium longiflorum  Streptophyta   \n",
       "4            None         4690           Lilium longiflorum  Streptophyta   \n",
       "5            None         4690           Lilium longiflorum  Streptophyta   \n",
       "6            None         4690           Lilium longiflorum  Streptophyta   \n",
       "7            None         4530                 Oryza sativa  Streptophyta   \n",
       "8            None         4530                 Oryza sativa  Streptophyta   \n",
       "9            None         3702         Arabidopsis thaliana  Streptophyta   \n",
       "10           None         3702         Arabidopsis thaliana  Streptophyta   \n",
       "11           None         3702         Arabidopsis thaliana  Streptophyta   \n",
       "\n",
       "            class taxonomy_group  info  \\\n",
       "0   Magnoliopsida           None  None   \n",
       "1   Magnoliopsida           None  None   \n",
       "2   Magnoliopsida           None  None   \n",
       "3   Magnoliopsida           None  None   \n",
       "4   Magnoliopsida           None  None   \n",
       "5   Magnoliopsida           None  None   \n",
       "6   Magnoliopsida           None  None   \n",
       "7   Magnoliopsida           None  None   \n",
       "8   Magnoliopsida           None  None   \n",
       "9   Magnoliopsida           None  None   \n",
       "10  Magnoliopsida           None  None   \n",
       "11  Magnoliopsida           None  None   \n",
       "\n",
       "                                             sequence  \\\n",
       "0   MARTKQTAKKSTAGNVPRKLLVMKVARKSAPMMAGLKKPHRFNPWI...   \n",
       "1   MARTKQTAKKSTASNVPRKLLVMKVARKSAPTMAGLKKPHRFKPGT...   \n",
       "2   MARPRKEAPQRNLDRDENARQQPTEEPQDEAPRNQGRQQQQQRPPA...   \n",
       "3   MARMKHTARMSTGGKAPRKQLASKALRKAPPPPTKGVKQPHHYHLR...   \n",
       "4   MARMKLNARMSTGGKAPRKQLAYKAVRKAAPPTIGVKLPNSYRPGD...   \n",
       "5   MARTKQTARKSTGGKAPRKQLATKAARKSIPTGMGGMKRPRRYRPG...   \n",
       "6   MARTKQTARKSTGGKAPRKQLATKAARKSAPTTGGVKKPHRYRPGT...   \n",
       "7   MRKGEEGGLDTDPHEADGVQVRDSRKQLATKATCKSAPATGGVKKP...   \n",
       "8   MARTKQTARKSTGGKAPRKPLRAIAAVMPAPSSRGVARKSVPFIGV...   \n",
       "9   MARTKQSARKSHGGKAPTKQLATKAARKSAPTTGGVKKPHRFRPGT...   \n",
       "10  MARSNQTARKATGGKAPHFAMRVWQHSTPPLKKPYRYKPGTVALRE...   \n",
       "11  MARTKQTARISTGGKAPRKQLAPKAARQSAPATGGVKKPHRFRPGT...   \n",
       "\n",
       "   variant_under_consideration  \n",
       "0                         None  \n",
       "1                         None  \n",
       "2                         None  \n",
       "3                         None  \n",
       "4                         None  \n",
       "5                         None  \n",
       "6                         None  \n",
       "7                         None  \n",
       "8                         None  \n",
       "9                         None  \n",
       "10                        None  \n",
       "11                        None  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE variant='H3-like_(Plants)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "eb9b6217-141a-465d-9d6a-ef8a42db06a6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c53eb54-49a8-422b-80b1-36c57cd4105e",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7dfef0cb-923d-4f42-9b00-aac0d915d16c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>hu_identification_2015</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 id title   doi author  year pubmed_id\n",
       "74   alvarez-venegas_canonical_2019  None  None   None  None      None\n",
       "106          hu_identification_2015  None  None   None  None      None"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"alvarez-venegas_canonical_2019\", \"hu_identification_2015\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b909ea1a-270b-400a-88fc-e7c1ce9c0e33",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in [\"HISTDB_H3_like_Plants_0\", \"HISTDB_H3_like_Plants_1\"]:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "21d03565-952d-4095-99f7-00759efd8934",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2575</th>\n",
       "      <td>HISTDB_H3_like_Plants_0</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4530.0</td>\n",
       "      <td>Oryza sativa</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MRKGEEGGLDTDPHEADGVQVRDSRKQLATKATCKSAPATGGVKKP...</td>\n",
       "      <td>None</td>\n",
       "      <td>HISTDB_H3_like_Plants_0</td>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2576</th>\n",
       "      <td>HISTDB_H3_like_Plants_0</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4530.0</td>\n",
       "      <td>Oryza sativa</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MRKGEEGGLDTDPHEADGVQVRDSRKQLATKATCKSAPATGGVKKP...</td>\n",
       "      <td>None</td>\n",
       "      <td>HISTDB_H3_like_Plants_0</td>\n",
       "      <td>hu_identification_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2577</th>\n",
       "      <td>HISTDB_H3_like_Plants_1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4530.0</td>\n",
       "      <td>Oryza sativa</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKPLRAIAAVMPAPSSRGVARKSVPFIGV...</td>\n",
       "      <td>None</td>\n",
       "      <td>HISTDB_H3_like_Plants_1</td>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2578</th>\n",
       "      <td>HISTDB_H3_like_Plants_1</td>\n",
       "      <td>H3-like_(Plants)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4530.0</td>\n",
       "      <td>Oryza sativa</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKPLRAIAAVMPAPSSRGVARKSVPFIGV...</td>\n",
       "      <td>None</td>\n",
       "      <td>HISTDB_H3_like_Plants_1</td>\n",
       "      <td>hu_identification_2015</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    accession           variant    gi ncbi_gene_id  \\\n",
       "2575  HISTDB_H3_like_Plants_0  H3-like_(Plants)  None         None   \n",
       "2576  HISTDB_H3_like_Plants_0  H3-like_(Plants)  None         None   \n",
       "2577  HISTDB_H3_like_Plants_1  H3-like_(Plants)  None         None   \n",
       "2578  HISTDB_H3_like_Plants_1  H3-like_(Plants)  None         None   \n",
       "\n",
       "     hgnc_gene_name  taxonomy_id      organism        phylum          class  \\\n",
       "2575           None       4530.0  Oryza sativa  Streptophyta  Magnoliopsida   \n",
       "2576           None       4530.0  Oryza sativa  Streptophyta  Magnoliopsida   \n",
       "2577           None       4530.0  Oryza sativa  Streptophyta  Magnoliopsida   \n",
       "2578           None       4530.0  Oryza sativa  Streptophyta  Magnoliopsida   \n",
       "\n",
       "     taxonomy_group  info                                           sequence  \\\n",
       "2575           None  None  MRKGEEGGLDTDPHEADGVQVRDSRKQLATKATCKSAPATGGVKKP...   \n",
       "2576           None  None  MRKGEEGGLDTDPHEADGVQVRDSRKQLATKATCKSAPATGGVKKP...   \n",
       "2577           None  None  MARTKQTARKSTGGKAPRKPLRAIAAVMPAPSSRGVARKSVPFIGV...   \n",
       "2578           None  None  MARTKQTARKSTGGKAPRKPLRAIAAVMPAPSSRGVARKSVPFIGV...   \n",
       "\n",
       "     variant_under_consideration       sequence_accession  \\\n",
       "2575                        None  HISTDB_H3_like_Plants_0   \n",
       "2576                        None  HISTDB_H3_like_Plants_0   \n",
       "2577                        None  HISTDB_H3_like_Plants_1   \n",
       "2578                        None  HISTDB_H3_like_Plants_1   \n",
       "\n",
       "                      publication_id  \n",
       "2575  alvarez-venegas_canonical_2019  \n",
       "2576          hu_identification_2015  \n",
       "2577  alvarez-venegas_canonical_2019  \n",
       "2578          hu_identification_2015  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin([\"HISTDB_H3_like_Plants_0\", \"HISTDB_H3_like_Plants_1\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9bb22363-87bc-43f6-a5d9-8e2a826e451a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "339bfc2b-aa96-4b28-918a-8b919cbdb89a",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add sequences to H3.Y_(Primates)</span>\n",
    "\n",
    "### <span style=\"color:black\">Delete Macaque sequences to HISTDB_H3_Y_0 and HISTDB_H3_Y_1 [seal_standardized_2022]</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "32c99ed3-232d-4666-9f22-a375df38a359",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2255</th>\n",
       "      <td>HISTDB_H3_Y_0</td>\n",
       "      <td>H3.Y_(Primates)</td>\n",
       "      <td>NOGI</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9544.0</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ARTKQTARKATNWQAPRKPLATKAAAKRAPPRGGIKKPHRYKPGTQ...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2256</th>\n",
       "      <td>HISTDB_H3_Y_1</td>\n",
       "      <td>H3.Y_(Primates)</td>\n",
       "      <td>NOGI</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9544.0</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ARTKQTARKATNWQAPRKPLATKAPGKRLPPRGGIKKPHRYRPGTQ...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession          variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "2255  HISTDB_H3_Y_0  H3.Y_(Primates)  NOGI         None           None   \n",
       "2256  HISTDB_H3_Y_1  H3.Y_(Primates)  NOGI         None           None   \n",
       "\n",
       "      taxonomy_id        organism    phylum     class taxonomy_group  info  \\\n",
       "2255       9544.0  Macaca mulatta  Chordata  Mammalia           None  None   \n",
       "2256       9544.0  Macaca mulatta  Chordata  Mammalia           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "2255  ARTKQTARKATNWQAPRKPLATKAAAKRAPPRGGIKKPHRYKPGTQ...   \n",
       "2256  ARTKQTARKATNWQAPRKPLATKAPGKRLPPRGGIKKPHRYRPGTQ...   \n",
       "\n",
       "     variant_under_consideration  \n",
       "2255                        None  \n",
       "2256                        None  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin([\"HISTDB_H3_Y_0\", \"HISTDB_H3_Y_1\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c7d61711-9137-4b20-aa8e-70a05b95ee1d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "query = \"DELETE FROM sequence WHERE accession='HISTDB_H3_Y_0'\"\n",
    "cursor.execute(query)\n",
    "query = \"DELETE FROM sequence WHERE accession='HISTDB_H3_Y_1'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4340f4e5-147c-47d2-a0dd-bfbd9026bd5a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin([\"HISTDB_H3_Y_0\", \"HISTDB_H3_Y_1\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "91b1c4fe-ad23-4d0a-b286-2052a8007423",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "898856da-2493-4300-ba57-06f1bf780285",
   "metadata": {},
   "source": [
    "### <span style=\"color:black\">Add Macaque sequence XP_001110711.1 as H3.Y.1_(Primates) [seal_standardized_2022]</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "826877cb-d59b-4985-831a-59ea866308b2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"XP_001110711.1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ed180b7-92b6-433d-87d6-7f824a7ad023",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "bc127863-0369-45e6-940c-ac13f5e7d9b3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "484c49ef-b7e2-4644-9d90-9e98ee4c0d9d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 9544\n",
      "accession XP_001110711.1 <class 'str'>\n",
      "variant H3.Y.1_(Primates) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 9544 <class 'int'>\n",
      "organism Macaca mulatta <class 'str'>\n",
      "phylum Chordata <class 'str'>\n",
      "class Mammalia <class 'str'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MARTKQTARKATNWQAPRKPLATKAPGKRLPPRGGIKKPHRYRPGTQALREIRKYQKSTQLLLRKLPFQRLVREIAQAISPDLRFQSAAIGALQEASEAYLVNLFEDTNLCAIHARRVTIMPRDMQLARRIRGEGA <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for a in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=a, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"H3.Y.1_(Primates)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "02f65c9f-594e-4623-aaf7-ac658f5c2053",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f96b9518-5dd4-4f74-967b-41418b8256f9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5701</th>\n",
       "      <td>XP_001110711.1</td>\n",
       "      <td>H3.Y.1_(Primates)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9544.0</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKATNWQAPRKPLATKAPGKRLPPRGGIKKPHRYRPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession            variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "5701  XP_001110711.1  H3.Y.1_(Primates)  None         None           None   \n",
       "\n",
       "      taxonomy_id        organism    phylum     class taxonomy_group  info  \\\n",
       "5701       9544.0  Macaca mulatta  Chordata  Mammalia           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "5701  MARTKQTARKATNWQAPRKPLATKAPGKRLPPRGGIKKPHRYRPGT...   \n",
       "\n",
       "     variant_under_consideration  \n",
       "5701                        None  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "75767572-2104-4be7-b307-ff2d4cbf142f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31179ec5-183c-466b-b933-2a79fca617cd",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "380a669c-1c01-4358-b85f-4ce05b95e6f4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, title, doi, author, year, pubmed_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"seal_standardized_2022\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "765865ff-681d-48da-9f3e-f22ee8aab7a1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_publication = {\n",
    "    \"id\": \"seal_standardized_2022\",\n",
    "    \"title\": None,\n",
    "    \"doi\": None,\n",
    "    \"author\": None,\n",
    "    \"year\": None,\n",
    "}\n",
    "cursor.execute(add_publication, data_publication)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "17b682bf-9bd1-4ad1-bfa3-0216ad395fa2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>seal_standardized_2022</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         id title   doi author  year pubmed_id\n",
       "157  seal_standardized_2022  None  None   None  None      None"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"seal_standardized_2022\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "15282dca-78cf-428b-9a0f-f7bc52d8ef2c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "6bbabd92-b067-441e-9412-e23fccebe3bb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6259</th>\n",
       "      <td>XP_001110711.1</td>\n",
       "      <td>H3.Y.1_(Primates)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9544.0</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKATNWQAPRKPLATKAPGKRLPPRGGIKKPHRYRPGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>XP_001110711.1</td>\n",
       "      <td>seal_standardized_2022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession            variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "6259  XP_001110711.1  H3.Y.1_(Primates)  None         None           None   \n",
       "\n",
       "      taxonomy_id        organism    phylum     class taxonomy_group  info  \\\n",
       "6259       9544.0  Macaca mulatta  Chordata  Mammalia           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "6259  MARTKQTARKATNWQAPRKPLATKAPGKRLPPRGGIKKPHRYRPGT...   \n",
       "\n",
       "     variant_under_consideration sequence_accession          publication_id  \n",
       "6259                        None     XP_001110711.1  seal_standardized_2022  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "2a596a19-8187-4f77-af37-23d970f51f50",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08364049-6938-42d3-9385-3331a6dff26a",
   "metadata": {},
   "source": [
    "### <span style=\"color:black\">Change sequence identificator and variant: HISTDB_H3_Y_2 -> XP_024212320.2 -> H3.Y.2_(Primates) [seal_standardized_2022]</span>\n",
    "\n",
    "### <span style=\"color:black\">Change sequence identificator and variant: HISTDB_H3_Y_3 -> HISTDB_H3_Y_0 -> H3.Y.1_(Primates) [wiedemann_identification_2010]</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "4359c89b-e25d-4cd1-ba6e-5d16e6807f42",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2255</th>\n",
       "      <td>HISTDB_H3_Y_2</td>\n",
       "      <td>H3.Y_(Primates)</td>\n",
       "      <td>NOGI</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9598.0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGTL...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2256</th>\n",
       "      <td>HISTDB_H3_Y_3</td>\n",
       "      <td>H3.Y_(Primates)</td>\n",
       "      <td>NOGI</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9598.0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGTL...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession          variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "2255  HISTDB_H3_Y_2  H3.Y_(Primates)  NOGI         None           None   \n",
       "2256  HISTDB_H3_Y_3  H3.Y_(Primates)  NOGI         None           None   \n",
       "\n",
       "      taxonomy_id         organism    phylum     class taxonomy_group  info  \\\n",
       "2255       9598.0  Pan troglodytes  Chordata  Mammalia           None  None   \n",
       "2256       9598.0  Pan troglodytes  Chordata  Mammalia           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "2255  ARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGTL...   \n",
       "2256  ARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGTL...   \n",
       "\n",
       "     variant_under_consideration  \n",
       "2255                        None  \n",
       "2256                        None  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin([\"HISTDB_H3_Y_2\", \"HISTDB_H3_Y_3\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "e289bed7-6e57-4b80-a6fc-9ad7e6d62ca0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"XP_024212320.2\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9e5410e-50cd-471f-ab0b-fdc2716aee4d",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "1792d27b-5193-4969-944e-f01d3d0024f1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "4f269254-66ab-4e90-bad4-0c35af9c97ef",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 9598\n",
      "accession XP_024212320.2 <class 'str'>\n",
      "variant H3.Y.2_(Primates) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 9598 <class 'int'>\n",
      "organism Pan troglodytes <class 'str'>\n",
      "phylum Chordata <class 'str'>\n",
      "class Mammalia <class 'str'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGTLALREIRKYQKSTQLLLRKLPFQRLVREIAQAISPDLRFQSAAIGALQEASEAYLVQLFEDTNLCAIHARRVTIMPRDMQLARRLRREGP <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for a in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=a, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"H3.Y.2_(Primates)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "8fb85b84-119f-4dbc-8b46-b81e36b126f2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "0dbc7e2c-7709-4265-80cb-5116c885744c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2255</th>\n",
       "      <td>HISTDB_H3_Y_2</td>\n",
       "      <td>H3.Y_(Primates)</td>\n",
       "      <td>NOGI</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9598.0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGTL...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2256</th>\n",
       "      <td>HISTDB_H3_Y_3</td>\n",
       "      <td>H3.Y_(Primates)</td>\n",
       "      <td>NOGI</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9598.0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGTL...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5884</th>\n",
       "      <td>XP_024212320.2</td>\n",
       "      <td>H3.Y.2_(Primates)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9598.0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession            variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "2255   HISTDB_H3_Y_2    H3.Y_(Primates)  NOGI         None           None   \n",
       "2256   HISTDB_H3_Y_3    H3.Y_(Primates)  NOGI         None           None   \n",
       "5884  XP_024212320.2  H3.Y.2_(Primates)  None         None           None   \n",
       "\n",
       "      taxonomy_id         organism    phylum     class taxonomy_group  info  \\\n",
       "2255       9598.0  Pan troglodytes  Chordata  Mammalia           None  None   \n",
       "2256       9598.0  Pan troglodytes  Chordata  Mammalia           None  None   \n",
       "5884       9598.0  Pan troglodytes  Chordata  Mammalia           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "2255  ARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGTL...   \n",
       "2256  ARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGTL...   \n",
       "5884  MARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGT...   \n",
       "\n",
       "     variant_under_consideration  \n",
       "2255                        None  \n",
       "2256                        None  \n",
       "5884                        None  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions+[\"HISTDB_H3_Y_2\", \"HISTDB_H3_Y_3\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "bed764d5-4bad-46bf-86bf-80b8a051fb6f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "query = \"DELETE FROM sequence WHERE accession='HISTDB_H3_Y_2'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "af365b40-d0c8-4f6a-b466-f4b96f6049df",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2255</th>\n",
       "      <td>HISTDB_H3_Y_3</td>\n",
       "      <td>H3.Y_(Primates)</td>\n",
       "      <td>NOGI</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9598.0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGTL...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5883</th>\n",
       "      <td>XP_024212320.2</td>\n",
       "      <td>H3.Y.2_(Primates)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9598.0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession            variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "2255   HISTDB_H3_Y_3    H3.Y_(Primates)  NOGI         None           None   \n",
       "5883  XP_024212320.2  H3.Y.2_(Primates)  None         None           None   \n",
       "\n",
       "      taxonomy_id         organism    phylum     class taxonomy_group  info  \\\n",
       "2255       9598.0  Pan troglodytes  Chordata  Mammalia           None  None   \n",
       "5883       9598.0  Pan troglodytes  Chordata  Mammalia           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "2255  ARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGTL...   \n",
       "5883  MARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGT...   \n",
       "\n",
       "     variant_under_consideration  \n",
       "2255                        None  \n",
       "5883                        None  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions+[\"HISTDB_H3_Y_2\", \"HISTDB_H3_Y_3\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "cda0d2c6-736d-4797-a8e4-786e76cfac02",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e59af5e6-16ab-46e3-9873-782fa72492e4",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "416a4a95-80d3-4e51-8792-1bffb87a8d7a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [sequence_accession, publication_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence_has_publication \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"sequence_accession\"].isin([\"HISTDB_H3_Y_0\", \"HISTDB_H3_Y_1\", \"HISTDB_H3_Y_2\", \"HISTDB_H3_Y_3\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "e917449c-69ff-4c82-9979-bd91de04caea",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>seal_standardized_2022</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         id title   doi author  year pubmed_id\n",
       "157  seal_standardized_2022  None  None   None  None      None"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"seal_standardized_2022\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "d3022fb4-96ba-454f-9ce2-1928e21518b1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "14dec7f3-cf2c-4b96-992c-10be596671a7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6498</th>\n",
       "      <td>XP_024212320.2</td>\n",
       "      <td>H3.Y.2_(Primates)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9598.0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>XP_024212320.2</td>\n",
       "      <td>seal_standardized_2022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession            variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "6498  XP_024212320.2  H3.Y.2_(Primates)  None         None           None   \n",
       "\n",
       "      taxonomy_id         organism    phylum     class taxonomy_group  info  \\\n",
       "6498       9598.0  Pan troglodytes  Chordata  Mammalia           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "6498  MARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGT...   \n",
       "\n",
       "     variant_under_consideration sequence_accession          publication_id  \n",
       "6498                        None     XP_024212320.2  seal_standardized_2022  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "1cdd66c6-795f-4ca5-a40b-2e1cfb89505f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "a97a8c6b-8ded-40ad-b3fe-0fd56de7c225",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "query = \"UPDATE sequence SET accession='HISTDB_H3_Y_0', variant='H3.Y.1_(Primates)' WHERE accession='HISTDB_H3_Y_3'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "534e66a8-01d4-4c93-984d-8d1a725da094",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2255</th>\n",
       "      <td>HISTDB_H3_Y_0</td>\n",
       "      <td>H3.Y.1_(Primates)</td>\n",
       "      <td>NOGI</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9598.0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGTL...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5883</th>\n",
       "      <td>XP_024212320.2</td>\n",
       "      <td>H3.Y.2_(Primates)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9598.0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession            variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "2255   HISTDB_H3_Y_0  H3.Y.1_(Primates)  NOGI         None           None   \n",
       "5883  XP_024212320.2  H3.Y.2_(Primates)  None         None           None   \n",
       "\n",
       "      taxonomy_id         organism    phylum     class taxonomy_group  info  \\\n",
       "2255       9598.0  Pan troglodytes  Chordata  Mammalia           None  None   \n",
       "5883       9598.0  Pan troglodytes  Chordata  Mammalia           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "2255  ARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGTL...   \n",
       "5883  MARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGT...   \n",
       "\n",
       "     variant_under_consideration  \n",
       "2255                        None  \n",
       "5883                        None  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions+[\"HISTDB_H3_Y_0\", \"HISTDB_H3_Y_2\", \"HISTDB_H3_Y_3\"])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9712a9ba-7858-45d8-a35e-5e28b795c2c1",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "21175aae-e6d0-416d-b545-862610b6b0d0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, title, doi, author, year, pubmed_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"wiedemann_identification_2010\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "6c47a947-0d05-490a-a2fe-bc97c6286346",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_publication = {\n",
    "    \"id\": \"wiedemann_identification_2010\",\n",
    "    \"title\": None,\n",
    "    \"doi\": None,\n",
    "    \"author\": None,\n",
    "    \"year\": None,\n",
    "}\n",
    "cursor.execute(add_publication, data_publication)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "6113e30a-696b-45f6-a87d-83671de20466",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>wiedemann_identification_2010</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                id title   doi author  year pubmed_id\n",
       "183  wiedemann_identification_2010  None  None   None  None      None"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"wiedemann_identification_2010\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "4d8fde53-0ff6-4ea9-afc6-90ff55cf7f2e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for pid in pids:\n",
    "    cursor.execute(add_sequence_has_publication, (\"HISTDB_H3_Y_0\", pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "f5fd29ee-5926-4e28-b042-1c4ea4db1399",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2579</th>\n",
       "      <td>HISTDB_H3_Y_0</td>\n",
       "      <td>H3.Y.1_(Primates)</td>\n",
       "      <td>NOGI</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9598.0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGTL...</td>\n",
       "      <td>None</td>\n",
       "      <td>HISTDB_H3_Y_0</td>\n",
       "      <td>wiedemann_identification_2010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6498</th>\n",
       "      <td>XP_024212320.2</td>\n",
       "      <td>H3.Y.2_(Primates)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9598.0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>XP_024212320.2</td>\n",
       "      <td>seal_standardized_2022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession            variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "2579   HISTDB_H3_Y_0  H3.Y.1_(Primates)  NOGI         None           None   \n",
       "6498  XP_024212320.2  H3.Y.2_(Primates)  None         None           None   \n",
       "\n",
       "      taxonomy_id         organism    phylum     class taxonomy_group  info  \\\n",
       "2579       9598.0  Pan troglodytes  Chordata  Mammalia           None  None   \n",
       "6498       9598.0  Pan troglodytes  Chordata  Mammalia           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "2579  ARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGTL...   \n",
       "6498  MARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGT...   \n",
       "\n",
       "     variant_under_consideration sequence_accession  \\\n",
       "2579                        None      HISTDB_H3_Y_0   \n",
       "6498                        None     XP_024212320.2   \n",
       "\n",
       "                     publication_id  \n",
       "2579  wiedemann_identification_2010  \n",
       "6498         seal_standardized_2022  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions+[\"HISTDB_H3_Y_0\", \"HISTDB_H3_Y_2\", \"HISTDB_H3_Y_3\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "23cc3c6f-bd61-480a-a30c-180b1226b390",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "c04ac060-b8d2-45c1-b4b9-5987c6dcbae4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2579</th>\n",
       "      <td>HISTDB_H3_Y_0</td>\n",
       "      <td>H3.Y.1_(Primates)</td>\n",
       "      <td>NOGI</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9598.0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGTL...</td>\n",
       "      <td>None</td>\n",
       "      <td>HISTDB_H3_Y_0</td>\n",
       "      <td>wiedemann_identification_2010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6258</th>\n",
       "      <td>XP_001110711.1</td>\n",
       "      <td>H3.Y.1_(Primates)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9544.0</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKATNWQAPRKPLATKAPGKRLPPRGGIKKPHRYRPGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>XP_001110711.1</td>\n",
       "      <td>seal_standardized_2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6498</th>\n",
       "      <td>XP_024212320.2</td>\n",
       "      <td>H3.Y.2_(Primates)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9598.0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>XP_024212320.2</td>\n",
       "      <td>seal_standardized_2022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession            variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "2579   HISTDB_H3_Y_0  H3.Y.1_(Primates)  NOGI         None           None   \n",
       "6258  XP_001110711.1  H3.Y.1_(Primates)  None         None           None   \n",
       "6498  XP_024212320.2  H3.Y.2_(Primates)  None         None           None   \n",
       "\n",
       "      taxonomy_id         organism    phylum     class taxonomy_group  info  \\\n",
       "2579       9598.0  Pan troglodytes  Chordata  Mammalia           None  None   \n",
       "6258       9544.0   Macaca mulatta  Chordata  Mammalia           None  None   \n",
       "6498       9598.0  Pan troglodytes  Chordata  Mammalia           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "2579  ARTKQTARKATAWQAPRKPLATKAARKRASPTGGIKKPHRYKPGTL...   \n",
       "6258  MARTKQTARKATNWQAPRKPLATKAPGKRLPPRGGIKKPHRYRPGT...   \n",
       "6498  MARTKQTARKATAWQAPRKPLATKAAGKRAPPTGGIKKPHRYKPGT...   \n",
       "\n",
       "     variant_under_consideration sequence_accession  \\\n",
       "2579                        None      HISTDB_H3_Y_0   \n",
       "6258                        None     XP_001110711.1   \n",
       "6498                        None     XP_024212320.2   \n",
       "\n",
       "                     publication_id  \n",
       "2579  wiedemann_identification_2010  \n",
       "6258         seal_standardized_2022  \n",
       "6498         seal_standardized_2022  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"variant\"].isin([\"H3.Y.1_(Primates)\", \"H3.Y.2_(Primates)\"])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fa2ad05-4aa0-4c37-8b15-66f3a2f37433",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add sequences to H3.4_(Mammalia)</span>\n",
    "\n",
    "### <span style=\"color:black\">Add sequence NP_001304932.1 as H3.4_(Mammalia) [ueda_testis-specific_2017]</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "500c2a49-cbec-46c4-b94a-56cd786fee2d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"NP_001304932.1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3be15475-4a6b-4b12-b194-69d7c7bebc3a",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "ae145d68-77bc-4767-bfd6-9de12020b869",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "f22421db-6aa9-44f3-9f54-803cfaf63e51",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 10090\n",
      "accession NP_001304932.1 <class 'str'>\n",
      "variant H3.4_(Mammalia) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 10090 <class 'int'>\n",
      "organism Mus musculus <class 'str'>\n",
      "phylum Chordata <class 'str'>\n",
      "class Mammalia <class 'str'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYHPGTVALREIRRYQKSTELLIRKLPFQRLVREIAQDFKTDLRFQSSAVMALQEACESYLVGLFEDTNLCAIHAKRVTIMPKDIQLARRIRGERA <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for a in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=a, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"H3.4_(Mammalia)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "f743cbc8-fe6e-452e-9563-576aefabbb1e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "53c06738-3e58-4dfd-82a9-7aad9cad7442",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3461</th>\n",
       "      <td>NP_001304932.1</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>10090.0</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYHPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession          variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "3461  NP_001304932.1  H3.4_(Mammalia)  None         None           None   \n",
       "\n",
       "      taxonomy_id      organism    phylum     class taxonomy_group  info  \\\n",
       "3461      10090.0  Mus musculus  Chordata  Mammalia           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "3461  MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYHPGT...   \n",
       "\n",
       "     variant_under_consideration  \n",
       "3461                        None  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "fcbddcf6-fed5-4c16-a10b-114cda912e0f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a535b717-bdb0-4355-bbb6-aa8e7362603b",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "5b88d400-7c9e-4b78-9ba6-6efadc3196fd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>ueda_testis-specific_2017</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            id title   doi author  year pubmed_id\n",
       "181  ueda_testis-specific_2017  None  None   None  None      None"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"ueda_testis-specific_2017\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "44594a40-d727-489b-8db2-da4e3842cfc6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "1c6a8fba-b31c-4899-8bd8-87065994d4af",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3846</th>\n",
       "      <td>NP_001304932.1</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>10090.0</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYHPGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>NP_001304932.1</td>\n",
       "      <td>ueda_testis-specific_2017</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession          variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "3846  NP_001304932.1  H3.4_(Mammalia)  None         None           None   \n",
       "\n",
       "      taxonomy_id      organism    phylum     class taxonomy_group  info  \\\n",
       "3846      10090.0  Mus musculus  Chordata  Mammalia           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "3846  MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYHPGT...   \n",
       "\n",
       "     variant_under_consideration sequence_accession             publication_id  \n",
       "3846                        None     NP_001304932.1  ueda_testis-specific_2017  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "c5ab57b9-804a-4808-894f-bb13b3a8b0bd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1882b88-88d4-4834-a2e0-b84be46b941f",
   "metadata": {},
   "source": [
    "### <span style=\"color:black\">Add fasta sequences to H3.4_(Mammalia)</span>\n",
    "\n",
    "Sequences fasta [dong_structural_2020]:\n",
    "\n",
    "```fasta\n",
    ">uncharacterized protein [rat]\n",
    "MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYHPGTVALREIRRYQKSTELLIRKLPFQR\n",
    "LVREIAQDFKTDLRFQSSAVMALQEACESYLVGLFEDTNLCAIHAKRVTIMPKDIQLARRIRGERA\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "a63e5ed7-e19b-4f73-8e51-1700cd2b0e2f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_001304932.1</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYHPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XP_003804825.1</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>397466137</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9597</td>\n",
       "      <td>Pan paniscus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLVTKVARKSAPATGGVKKPHRYRPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession          variant         gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  NP_001304932.1  H3.4_(Mammalia)       None         None           None   \n",
       "1  XP_003804825.1  H3.4_(Mammalia)  397466137         None           None   \n",
       "\n",
       "   taxonomy_id      organism    phylum     class taxonomy_group  info  \\\n",
       "0        10090  Mus musculus  Chordata  Mammalia           None  None   \n",
       "1         9597  Pan paniscus  Chordata  Mammalia           None  None   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYHPGT...   \n",
       "1  MARTKQTARKSTGGKAPRKQLVTKVARKSAPATGGVKKPHRYRPGT...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                        None  \n",
       "1                        None  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE variant='H3.4_(Mammalia)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "841b454f-95ae-44cf-a8c4-8ef50f64b072",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_sequence = {\n",
    "    \"accession\": \"HISTDB_H3_4_Mammalia_0\",\n",
    "    \"variant\": \"H3.4_(Mammalia)\",\n",
    "    \"gi\": None,\n",
    "    \"ncbi_gene_id\": None,\n",
    "    \"hgnc_gene_name\": None,\n",
    "    \"taxonomy_id\": 10114,\n",
    "    \"organism\": \"Rattus\",\n",
    "    \"phylum\": \" Chordata\",\n",
    "    \"class\": \"Mammalia\",\n",
    "    \"taxonomy_group\": None,\n",
    "    \"info\": None,\n",
    "    \"sequence\": \"MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYHPGTVALREIRRYQKSTELLIRKLPFQRLVREIAQDFKTDLRFQSSAVMALQEACESYLVGLFEDTNLCAIHAKRVTIMPKDIQLARRIRGERA\",\n",
    "    \"variant_under_consideration\": None,\n",
    "}\n",
    "cursor.execute(add_sequence, data_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "f9d9d4aa-b8d6-4447-8310-1b1e4947ec41",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HISTDB_H3_4_Mammalia_0</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>10114</td>\n",
       "      <td>Rattus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYHPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_001304932.1</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYHPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XP_003804825.1</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>397466137</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9597</td>\n",
       "      <td>Pan paniscus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLVTKVARKSAPATGGVKKPHRYRPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                accession          variant         gi ncbi_gene_id  \\\n",
       "0  HISTDB_H3_4_Mammalia_0  H3.4_(Mammalia)       None         None   \n",
       "1          NP_001304932.1  H3.4_(Mammalia)       None         None   \n",
       "2          XP_003804825.1  H3.4_(Mammalia)  397466137         None   \n",
       "\n",
       "  hgnc_gene_name  taxonomy_id      organism     phylum     class  \\\n",
       "0           None        10114        Rattus   Chordata  Mammalia   \n",
       "1           None        10090  Mus musculus   Chordata  Mammalia   \n",
       "2           None         9597  Pan paniscus   Chordata  Mammalia   \n",
       "\n",
       "  taxonomy_group  info                                           sequence  \\\n",
       "0           None  None  MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYHPGT...   \n",
       "1           None  None  MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYHPGT...   \n",
       "2           None  None  MARTKQTARKSTGGKAPRKQLVTKVARKSAPATGGVKKPHRYRPGT...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                        None  \n",
       "1                        None  \n",
       "2                        None  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE variant='H3.4_(Mammalia)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "575bad74-f6ac-4251-b9fc-023c484f7950",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e1e5ff3-8c37-4327-ba7a-11d23e3af383",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "4b61173f-3268-4b5e-859b-13f5bee13360",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>dong_structural_2020</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      id title   doi author  year pubmed_id\n",
       "90  dong_structural_2020  None  None   None  None      None"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"dong_structural_2020\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "c24c73f6-d7a2-480c-ab2d-7a01f3240d70",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for pid in pids:\n",
    "    cursor.execute(add_sequence_has_publication, (\"HISTDB_H3_4_Mammalia_0\", pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "50856cd2-29a2-4620-a2db-86e12199ed19",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2530</th>\n",
       "      <td>HISTDB_H3_4_Mammalia_0</td>\n",
       "      <td>H3.4_(Mammalia)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>10114.0</td>\n",
       "      <td>Rattus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYHPGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>HISTDB_H3_4_Mammalia_0</td>\n",
       "      <td>dong_structural_2020</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   accession          variant    gi ncbi_gene_id  \\\n",
       "2530  HISTDB_H3_4_Mammalia_0  H3.4_(Mammalia)  None         None   \n",
       "\n",
       "     hgnc_gene_name  taxonomy_id organism     phylum     class taxonomy_group  \\\n",
       "2530           None      10114.0   Rattus   Chordata  Mammalia           None   \n",
       "\n",
       "      info                                           sequence  \\\n",
       "2530  None  MARTKQTARKSTGGKAPRKQLATKVARKSAPATGGVKKPHRYHPGT...   \n",
       "\n",
       "     variant_under_consideration      sequence_accession        publication_id  \n",
       "2530                        None  HISTDB_H3_4_Mammalia_0  dong_structural_2020  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin([\"HISTDB_H3_4_Mammalia_0\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "6b0ba895-6e7a-46d9-917c-12532bd5ae2a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "721a0e23-1ecd-47fa-897f-494c0b25425a",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add sequences to H3.5_(Hominidae)</span>\n",
    "\n",
    "### <span style=\"color:black\">Add orangutan sequence XP_002823134.1, gorilla sequence ADW85799.1 as H3.5_(Hominidae) [schenk_h35_2011]</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "cf247d46-6c9c-41b7-8e5f-78a71638afda",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"XP_002823134.1\", \"ADW85799.1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09b4f797-0eea-4973-86f4-41a159157a38",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "d2d3eb07-a96c-43b7-b494-b837891c5d53",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "1ffb6db6-ef99-42bd-94c8-86eefb50b3e3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 9601\n",
      "Fetched taxid from NCBI 9593\n",
      "accession ADW85799.1 <class 'str'>\n",
      "variant H3.5_(Hominidae) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 9593 <class 'int'>\n",
      "organism Gorilla gorilla <class 'str'>\n",
      "phylum Chordata <class 'str'>\n",
      "class Mammalia <class 'str'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MARTKQTARKSTGGKAPRKQLATKAARKSTPSTCGVKKPHRYRPGTVALREIRRYQKSTELLIRKLPFQRLVREIAQDFNTGLSFQSAAIGALQEASEAYLVGLLEDTNLCAIHAKRVTIMPKDIQLARRIRGERA <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for a in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=a, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"H3.5_(Hominidae)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "7414f251-4c8b-45f0-b031-4da9b20f7ac2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "f6bb0ec2-3795-4ce0-b3c2-208dea73d2d7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>ADW85799.1</td>\n",
       "      <td>H3.5_(Hominidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9593.0</td>\n",
       "      <td>Gorilla gorilla</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSTPSTCGVKKPHRYRPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5763</th>\n",
       "      <td>XP_002823134.1</td>\n",
       "      <td>H3.5_(Hominidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9601.0</td>\n",
       "      <td>Pongo abelii</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSTPSTGGVKKPHRHGPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession           variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "238       ADW85799.1  H3.5_(Hominidae)  None         None           None   \n",
       "5763  XP_002823134.1  H3.5_(Hominidae)  None         None           None   \n",
       "\n",
       "      taxonomy_id         organism    phylum     class taxonomy_group  info  \\\n",
       "238        9593.0  Gorilla gorilla  Chordata  Mammalia           None  None   \n",
       "5763       9601.0     Pongo abelii  Chordata  Mammalia           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "238   MARTKQTARKSTGGKAPRKQLATKAARKSTPSTCGVKKPHRYRPGT...   \n",
       "5763  MARTKQTARKSTGGKAPRKQLATKAARKSTPSTGGVKKPHRHGPGT...   \n",
       "\n",
       "     variant_under_consideration  \n",
       "238                         None  \n",
       "5763                        None  "
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "ee55406c-910e-45e2-bd1b-928564700333",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "289398fd-03bc-4ee0-ae2c-00530338f27e",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "d4a011df-98d3-466a-811f-8129e84f63a7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, title, doi, author, year, pubmed_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"schenk_h35_2011\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "e6ca2fec-fb4f-4ce6-b0c8-739c7a1ba064",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_publication = {\n",
    "    \"id\": \"schenk_h35_2011\",\n",
    "    \"title\": None,\n",
    "    \"doi\": None,\n",
    "    \"author\": None,\n",
    "    \"year\": None,\n",
    "}\n",
    "cursor.execute(add_publication, data_publication)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "8e34a280-6b0d-4b0e-a594-11d8f31bc762",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>schenk_h35_2011</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  id title   doi author  year pubmed_id\n",
       "155  schenk_h35_2011  None  None   None  None      None"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"schenk_h35_2011\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "f8b45d60-f0de-4376-9c49-c70303f629d9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "a6b149c0-0160-45f3-a5cd-88738d168f65",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>ADW85799.1</td>\n",
       "      <td>H3.5_(Hominidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9593.0</td>\n",
       "      <td>Gorilla gorilla</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSTPSTCGVKKPHRYRPGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>ADW85799.1</td>\n",
       "      <td>schenk_h35_2011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6342</th>\n",
       "      <td>XP_002823134.1</td>\n",
       "      <td>H3.5_(Hominidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9601.0</td>\n",
       "      <td>Pongo abelii</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSTPSTGGVKKPHRHGPGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>XP_002823134.1</td>\n",
       "      <td>schenk_h35_2011</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession           variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "285       ADW85799.1  H3.5_(Hominidae)  None         None           None   \n",
       "6342  XP_002823134.1  H3.5_(Hominidae)  None         None           None   \n",
       "\n",
       "      taxonomy_id         organism    phylum     class taxonomy_group  info  \\\n",
       "285        9593.0  Gorilla gorilla  Chordata  Mammalia           None  None   \n",
       "6342       9601.0     Pongo abelii  Chordata  Mammalia           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "285   MARTKQTARKSTGGKAPRKQLATKAARKSTPSTCGVKKPHRYRPGT...   \n",
       "6342  MARTKQTARKSTGGKAPRKQLATKAARKSTPSTGGVKKPHRHGPGT...   \n",
       "\n",
       "     variant_under_consideration sequence_accession   publication_id  \n",
       "285                         None         ADW85799.1  schenk_h35_2011  \n",
       "6342                        None     XP_002823134.1  schenk_h35_2011  "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "52b259d3-0574-4e2e-866b-22b598ca4ead",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc0b6c7e-6588-42f4-81d2-9dcf1291d258",
   "metadata": {},
   "source": [
    "### <span style=\"color:black\">Delete chimpanzee sequence XP_003954426.1 (it is pseudogene [seal_standardized_2022])</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "b8285ef9-d37d-4cf6-9f9f-c12b584705b2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5810</th>\n",
       "      <td>XP_003954426.1</td>\n",
       "      <td>H3.5_(Hominidae)</td>\n",
       "      <td>410046862</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9598.0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSTPSTXGVKKPHRYRPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession           variant         gi ncbi_gene_id hgnc_gene_name  \\\n",
       "5810  XP_003954426.1  H3.5_(Hominidae)  410046862         None           None   \n",
       "\n",
       "      taxonomy_id         organism    phylum     class taxonomy_group  info  \\\n",
       "5810       9598.0  Pan troglodytes  Chordata  Mammalia           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "5810  MARTKQTARKSTGGKAPRKQLATKAARKSTPSTXGVKKPHRYRPGT...   \n",
       "\n",
       "     variant_under_consideration  \n",
       "5810                        None  "
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin([\"XP_003954426.1\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "c211c54d-79dc-4973-8fc4-d3afac6fe4cd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "query = \"DELETE FROM sequence WHERE accession='XP_003954426.1'\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "aa97057a-944e-44e7-8b81-6a72348658c4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin([\"XP_003954426.1\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "3af67892-51e3-4a30-93a0-9a7b56be648a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cdaae89-e1ae-4206-aebf-5f3a6f86420b",
   "metadata": {},
   "source": [
    "### <span style=\"color:black\">Add alternate name H3.3C to H3.5_(Homo_sapiens)</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "11042070-3534-420c-856a-a2b16f43de7d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, name, taxonomy, gene, splice, histone]\n",
       "Index: []"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name \" \"WHERE histone='H3.5_(Homo_sapiens)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "aa631802-5e34-477f-87a7-442315f7d0cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_alternate_name = {\n",
    "    \"name\": \"H3.3C\",\n",
    "    \"taxonomy\": None,\n",
    "    \"gene\": None,\n",
    "    \"splice\": None,\n",
    "    \"histone\": \"H3.5_(Homo_sapiens)\",\n",
    "}\n",
    "cursor.execute(add_alternate_names, data_alternate_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "a1f630b8-3223-46fd-a2db-5b25845ccfe0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>124</td>\n",
       "      <td>H3.3C</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>H3.5_(Homo_sapiens)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    id   name taxonomy  gene splice              histone\n",
       "0  124  H3.3C     None  None   None  H3.5_(Homo_sapiens)"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name \" \"WHERE histone='H3.5_(Homo_sapiens)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "911089ed-c749-494a-883c-9f534ade9a2e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9983bf9d-7749-411e-ba94-3949defee1e6",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add sequences to H3.7_(Homo_sapiens)</span>\n",
    "\n",
    "### <span style=\"color:black\">Add sequence NP_001359034.1 to H3.7_(Homo_sapiens) [taguchi_crystal_2017, seal_standardized_2022]</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "d3c85fc4-3a68-41c9-aa03-e2f4aa5b93ed",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"NP_001359034.1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b8c0af0-09b2-4784-8f7e-668bb711c287",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "41c50e8e-a45e-4d3c-a02c-b95d3f6b3354",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "c2205579-b231-4e3f-8559-f1797eeb5e1a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 9606\n",
      "accession NP_001359034.1 <class 'str'>\n",
      "variant H3.7_(Homo_sapiens) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 9606 <class 'int'>\n",
      "organism Homo sapiens <class 'str'>\n",
      "phylum Chordata <class 'str'>\n",
      "class Mammalia <class 'str'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGTVALREIRRYQKSTELLIRKLPFQRLVREIAQEFKTDLRFQSSAVMALQEAREAYLVGLFEDTNLCAIHAKRVTIMPKDIQLVSRIRGERA <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for a in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=a, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"H3.7_(Homo_sapiens)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "8899373d-7921-4f12-a2a6-8a006d2ae8b6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "277b2ebb-f431-45ee-a5ed-c738ab6404d0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3467</th>\n",
       "      <td>NP_001359034.1</td>\n",
       "      <td>H3.7_(Homo_sapiens)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession              variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "3467  NP_001359034.1  H3.7_(Homo_sapiens)  None         None           None   \n",
       "\n",
       "      taxonomy_id      organism    phylum     class taxonomy_group  info  \\\n",
       "3467       9606.0  Homo sapiens  Chordata  Mammalia           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "3467  MARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGT...   \n",
       "\n",
       "     variant_under_consideration  \n",
       "3467                        None  "
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "bbf87c2c-1e50-479f-9859-96645dde8fae",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a5554c8-98d9-4aa8-82cd-7ecb794322f8",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "5771bdea-366e-4e3a-a1ea-e92aea8733cc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>seal_standardized_2022</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>taguchi_crystal_2017</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         id title   doi author  year pubmed_id\n",
       "158  seal_standardized_2022  None  None   None  None      None\n",
       "174    taguchi_crystal_2017  None  None   None  None      None"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"taguchi_crystal_2017\", \"seal_standardized_2022\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "8ca038b1-c6e8-4759-b075-f7707bfe2e88",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "6e3e5967-85cd-4834-846e-3dc940225857",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3853</th>\n",
       "      <td>NP_001359034.1</td>\n",
       "      <td>H3.7_(Homo_sapiens)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>NP_001359034.1</td>\n",
       "      <td>seal_standardized_2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3854</th>\n",
       "      <td>NP_001359034.1</td>\n",
       "      <td>H3.7_(Homo_sapiens)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>NP_001359034.1</td>\n",
       "      <td>taguchi_crystal_2017</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession              variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "3853  NP_001359034.1  H3.7_(Homo_sapiens)  None         None           None   \n",
       "3854  NP_001359034.1  H3.7_(Homo_sapiens)  None         None           None   \n",
       "\n",
       "      taxonomy_id      organism    phylum     class taxonomy_group  info  \\\n",
       "3853       9606.0  Homo sapiens  Chordata  Mammalia           None  None   \n",
       "3854       9606.0  Homo sapiens  Chordata  Mammalia           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "3853  MARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGT...   \n",
       "3854  MARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGT...   \n",
       "\n",
       "     variant_under_consideration sequence_accession          publication_id  \n",
       "3853                        None     NP_001359034.1  seal_standardized_2022  \n",
       "3854                        None     NP_001359034.1    taguchi_crystal_2017  "
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "28b83adf-e318-4256-99dc-cb38335aef93",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a576d2d-ff14-4d3e-bfd5-3a8f0561279d",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add sequences to H3.B_(Giardia)</span>\n",
    "\n",
    "### <span style=\"color:black\">Add sequence XP_767393.1 to H3.B_(Giardia) [dawson_cenh3_2007]</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "d5e16087-933c-4943-9bf7-482582e87d12",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"XP_767393.1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f34201e-7138-48d6-9115-20d3ebaff701",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "a7acd790-e31d-4509-9783-5734dd814251",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "0998e7ee-3a93-497c-a7e3-36ef8d4ecf83",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 184922\n",
      "accession XP_767393.1 <class 'str'>\n",
      "variant H3.B_(Giardia) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 184922 <class 'int'>\n",
      "organism Giardia lamblia ATCC 50803 <class 'str'>\n",
      "phylum Fornicata <class 'str'>\n",
      "class None <class 'NoneType'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MARTKNTAMDRSKSVHVNTARKGQHAPRKTILSKKTVARKAISKSEKAVTRRARPGSQVRKEITNMQRRVTSVIPIACFQRLVRDITCSLPSGGNEIRFQAQAIGALQEASEAMLSQVLGDCQILANHAHRVTIMDKDIQIYMRIVRPPWMNGIHGSML <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for a in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=a, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"H3.B_(Giardia)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "c4c12560-a4bd-4df7-9072-5ebe56bb04ba",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "7724c464-b4f7-4e3a-963b-c7a347c2319f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5925</th>\n",
       "      <td>XP_767393.1</td>\n",
       "      <td>H3.B_(Giardia)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>184922.0</td>\n",
       "      <td>Giardia lamblia ATCC 50803</td>\n",
       "      <td>Fornicata</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKNTAMDRSKSVHVNTARKGQHAPRKTILSKKTVARKAISKSE...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession         variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "5925  XP_767393.1  H3.B_(Giardia)  None         None           None   \n",
       "\n",
       "      taxonomy_id                    organism     phylum class taxonomy_group  \\\n",
       "5925     184922.0  Giardia lamblia ATCC 50803  Fornicata  None           None   \n",
       "\n",
       "      info                                           sequence  \\\n",
       "5925  None  MARTKNTAMDRSKSVHVNTARKGQHAPRKTILSKKTVARKAISKSE...   \n",
       "\n",
       "     variant_under_consideration  \n",
       "5925                        None  "
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "5275e4b3-bb41-430b-bcbe-f13e7af4d7e0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f106592-6fc4-48b8-8bc2-cde4528cbfc3",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "e87b4176-254f-4192-a34f-fb25c3fda574",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, title, doi, author, year, pubmed_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"dawson_cenh3_2007\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "55e8363e-00e0-4f9c-aa45-c85fbc55edf9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_publication = {\n",
    "    \"id\": \"dawson_cenh3_2007\",\n",
    "    \"title\": None,\n",
    "    \"doi\": None,\n",
    "    \"author\": None,\n",
    "    \"year\": None,\n",
    "}\n",
    "cursor.execute(add_publication, data_publication)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "ebd7b894-c920-46f6-a57c-e0e48c90da07",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>dawson_cenh3_2007</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   id title   doi author  year pubmed_id\n",
       "89  dawson_cenh3_2007  None  None   None  None      None"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"dawson_cenh3_2007\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "4de9a71d-4011-4b93-9538-137662d12836",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "55a55e9c-5a6e-4eeb-bde6-2750b5879852",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6550</th>\n",
       "      <td>XP_767393.1</td>\n",
       "      <td>H3.B_(Giardia)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>184922.0</td>\n",
       "      <td>Giardia lamblia ATCC 50803</td>\n",
       "      <td>Fornicata</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKNTAMDRSKSVHVNTARKGQHAPRKTILSKKTVARKAISKSE...</td>\n",
       "      <td>None</td>\n",
       "      <td>XP_767393.1</td>\n",
       "      <td>dawson_cenh3_2007</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession         variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "6550  XP_767393.1  H3.B_(Giardia)  None         None           None   \n",
       "\n",
       "      taxonomy_id                    organism     phylum class taxonomy_group  \\\n",
       "6550     184922.0  Giardia lamblia ATCC 50803  Fornicata  None           None   \n",
       "\n",
       "      info                                           sequence  \\\n",
       "6550  None  MARTKNTAMDRSKSVHVNTARKGQHAPRKTILSKKTVARKAISKSE...   \n",
       "\n",
       "     variant_under_consideration sequence_accession     publication_id  \n",
       "6550                        None        XP_767393.1  dawson_cenh3_2007  "
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "a09f0b15-241d-46dd-bd40-99a2020c312f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b913739c-3291-4876-9238-d7d0fa1d1f25",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add new node to cenH3 to cenH3_(Protists)</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "a997f0ff-6e17-41ac-ac41-915c4f438790",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, level, taxonomic_span, taxonomic_span_id, description, parent, id, summary, taxonomy, genes, evolution, expression, knock_out, function, sequence, localization, deposition, structure, interactions, disease, caveats, relations]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 22 columns]"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3_(Protists)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "5e3e8611-eadb-473c-aef1-826c2ea08e62",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_histone = {\n",
    "    \"id\": \"cenH3_(Protists)\",\n",
    "    \"level\": \"variant\",\n",
    "    \"taxonomic_span\": \"SAR,Metamonada,Discoba,Amoebozoa\",\n",
    "    \"taxonomic_span_id\": \"2698737,2611341,2611352,554915\",\n",
    "    \"description\": None,\n",
    "    \"parent\": \"cenH3\",\n",
    "}\n",
    "cursor.execute(add_histone, data_histone)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "d15596ec-f07f-4435-806a-214d102d256d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cenH3_(Protists)</td>\n",
       "      <td>variant</td>\n",
       "      <td>SAR,Metamonada,Discoba,Amoebozoa</td>\n",
       "      <td>2698737,2611341,2611352,554915</td>\n",
       "      <td>None</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 id    level                    taxonomic_span  \\\n",
       "0  cenH3_(Protists)  variant  SAR,Metamonada,Discoba,Amoebozoa   \n",
       "\n",
       "                taxonomic_span_id description parent    id summary taxonomy  \\\n",
       "0  2698737,2611341,2611352,554915        None  cenH3  None    None     None   \n",
       "\n",
       "  genes  ... knock_out function sequence localization deposition structure  \\\n",
       "0  None  ...      None     None     None         None       None      None   \n",
       "\n",
       "  interactions disease caveats relations  \n",
       "0         None    None    None      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cenH3_(Protists)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "e3043f37-42a3-4900-8de3-a93038898edf",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2c99e1a-b732-472f-aa48-a052679a9ed0",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add sequences to cenH3_(Protists)</span>\n",
    "\n",
    "### <span style=\"color:black\">Add sequence XP_771620.1 to cenH3_(Protists) [dawson_cenh3_2007]</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "69add2de-f77d-4635-9104-2467e7620841",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"XP_771620.1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0966553-93ee-4179-834d-6bde658c7bd9",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "04ec3dc2-5045-4b49-ab5f-3750e48ee391",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "b0be9a8e-f6fa-40db-8271-2cb72e5ac722",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 184922\n",
      "accession XP_771620.1 <class 'str'>\n",
      "variant cenH3_(Protists) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 184922 <class 'int'>\n",
      "organism Giardia lamblia ATCC 50803 <class 'str'>\n",
      "phylum Fornicata <class 'str'>\n",
      "class None <class 'NoneType'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MSGGSRSQVARNTGHRRREISGRNMIPGVVVNARQSRSKLSSDPFSSVPRRPARVSHMEREIYHYQHNVDTLIQKLPFARLVQELVEQIAQRDGSKGPYRFQGMAMEALQSATEEYIVELFSTALLATYHANRVTLMSKDILLVLRIQQRNLNSLR <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for a in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=a, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"cenH3_(Protists)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "ad9afe2b-b56d-4430-a076-b97e9a505233",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "c75af844-03c9-42a6-a36c-f358af0bf37d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5926</th>\n",
       "      <td>XP_771620.1</td>\n",
       "      <td>cenH3_(Protists)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>184922.0</td>\n",
       "      <td>Giardia lamblia ATCC 50803</td>\n",
       "      <td>Fornicata</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MSGGSRSQVARNTGHRRREISGRNMIPGVVVNARQSRSKLSSDPFS...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession           variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "5926  XP_771620.1  cenH3_(Protists)  None         None           None   \n",
       "\n",
       "      taxonomy_id                    organism     phylum class taxonomy_group  \\\n",
       "5926     184922.0  Giardia lamblia ATCC 50803  Fornicata  None           None   \n",
       "\n",
       "      info                                           sequence  \\\n",
       "5926  None  MSGGSRSQVARNTGHRRREISGRNMIPGVVVNARQSRSKLSSDPFS...   \n",
       "\n",
       "     variant_under_consideration  \n",
       "5926                        None  "
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "121907a5-6a2d-47bc-a837-20e8f8fd7e92",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e085baa5-ece1-4aeb-bcc5-42f3e421a556",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "b23ad20b-8042-4af0-8b32-4a3e0ecab31b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>dawson_cenh3_2007</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   id title   doi author  year pubmed_id\n",
       "89  dawson_cenh3_2007  None  None   None  None      None"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"dawson_cenh3_2007\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "3efe4085-980d-4cab-9782-6484c7d9fc70",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "906bad4b-1491-4ad6-a524-afbc110325f7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6551</th>\n",
       "      <td>XP_771620.1</td>\n",
       "      <td>cenH3_(Protists)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>184922.0</td>\n",
       "      <td>Giardia lamblia ATCC 50803</td>\n",
       "      <td>Fornicata</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MSGGSRSQVARNTGHRRREISGRNMIPGVVVNARQSRSKLSSDPFS...</td>\n",
       "      <td>None</td>\n",
       "      <td>XP_771620.1</td>\n",
       "      <td>dawson_cenh3_2007</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession           variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "6551  XP_771620.1  cenH3_(Protists)  None         None           None   \n",
       "\n",
       "      taxonomy_id                    organism     phylum class taxonomy_group  \\\n",
       "6551     184922.0  Giardia lamblia ATCC 50803  Fornicata  None           None   \n",
       "\n",
       "      info                                           sequence  \\\n",
       "6551  None  MSGGSRSQVARNTGHRRREISGRNMIPGVVVNARQSRSKLSSDPFS...   \n",
       "\n",
       "     variant_under_consideration sequence_accession     publication_id  \n",
       "6551                        None        XP_771620.1  dawson_cenh3_2007  "
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "687bd188-6f6c-4d28-a980-b77b20b25bd0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5e01516-4121-41b7-91da-5ee13f00524b",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add sequences to cH3_(Protists)</span>\n",
    "\n",
    "### <span style=\"color:black\">Add sequence AAF00592.1 to cH3_(Protists) [dawson_cenh3_2007]</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "f1c29bb9-a425-421e-a077-262569582b9c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"AAF00592.1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b010ea1-d678-4d31-b4ff-8d288c4561ff",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "f9556e83-69ef-48a9-9c92-d60b22f18598",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "9665418d-6508-40b9-b170-c0c3a3f4c4cd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 5741\n",
      "accession AAF00592.1 <class 'str'>\n",
      "variant cH3_(Protists) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 5741 <class 'int'>\n",
      "organism Giardia duodenalis <class 'str'>\n",
      "phylum Fornicata <class 'str'>\n",
      "class None <class 'NoneType'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MARTKHTARKTTSATKAPRKTIARKAARKTASSTSGIKKTGRKKQGMVAVKEIKKYQKSTDLLIRKLPFSKLVRDIVTSGLSKSDIRFQGAAVEALQESAENYIISLFVDTQLCAEHAKRVTIMKPDMELATRIGKRIEPEYRKGK <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for a in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=a, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"cH3_(Protists)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "4c8524e8-43da-488a-98f7-36ff3bcc03f7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "63557936-79a7-4c67-a43f-f36c31fce6da",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>AAF00592.1</td>\n",
       "      <td>cH3_(Protists)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5741.0</td>\n",
       "      <td>Giardia duodenalis</td>\n",
       "      <td>Fornicata</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHTARKTTSATKAPRKTIARKAARKTASSTSGIKKTGRKKQG...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession         variant    gi ncbi_gene_id hgnc_gene_name  taxonomy_id  \\\n",
       "27  AAF00592.1  cH3_(Protists)  None         None           None       5741.0   \n",
       "\n",
       "              organism     phylum class taxonomy_group  info  \\\n",
       "27  Giardia duodenalis  Fornicata  None           None  None   \n",
       "\n",
       "                                             sequence  \\\n",
       "27  MARTKHTARKTTSATKAPRKTIARKAARKTASSTSGIKKTGRKKQG...   \n",
       "\n",
       "   variant_under_consideration  \n",
       "27                        None  "
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "bc2bb89c-4414-49aa-8aeb-26f47fecc5d8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a24d5e2-2e5e-41c8-94d3-5a64ffa26f38",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "0730a8f2-b644-4e77-95bc-f004cae97efe",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>dawson_cenh3_2007</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   id title   doi author  year pubmed_id\n",
       "89  dawson_cenh3_2007  None  None   None  None      None"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"dawson_cenh3_2007\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "4970fffd-b540-48c3-ae70-3619e1888a25",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "5089778e-b827-4ad5-b51a-f4914ff665d7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>AAF00592.1</td>\n",
       "      <td>cH3_(Protists)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5741.0</td>\n",
       "      <td>Giardia duodenalis</td>\n",
       "      <td>Fornicata</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHTARKTTSATKAPRKTIARKAARKTASSTSGIKKTGRKKQG...</td>\n",
       "      <td>None</td>\n",
       "      <td>AAF00592.1</td>\n",
       "      <td>dawson_cenh3_2007</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession         variant    gi ncbi_gene_id hgnc_gene_name  taxonomy_id  \\\n",
       "35  AAF00592.1  cH3_(Protists)  None         None           None       5741.0   \n",
       "\n",
       "              organism     phylum class taxonomy_group  info  \\\n",
       "35  Giardia duodenalis  Fornicata  None           None  None   \n",
       "\n",
       "                                             sequence  \\\n",
       "35  MARTKHTARKTTSATKAPRKTIARKAARKTASSTSGIKKTGRKKQG...   \n",
       "\n",
       "   variant_under_consideration sequence_accession     publication_id  \n",
       "35                        None         AAF00592.1  dawson_cenh3_2007  "
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "329ebb4d-1867-4a6d-9ac5-b4c77db007e0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f7514fd-aeee-49f0-a17b-17edd611baa9",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add sequences to H3.P_(Euplotes_crassus)</span>\n",
    "\n",
    "### <span style=\"color:black\">Add sequence AAC47441.1 to H3.P_(Euplotes_crassus) [jahn_unusual_1997]</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "813086f5-a60f-4ee2-ad97-c119147ba24f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"AAC47441.1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35a73fee-01ce-4f8f-b037-0fb951caed14",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "b8da6c19-0073-4407-8d4f-04a0cf521046",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "8a241302-2506-42f4-ad96-a0f85f99d830",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 5936\n",
      "accession AAC47441.1 <class 'str'>\n",
      "variant H3.P_(Euplotes_crassus) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 5936 <class 'int'>\n",
      "organism Moneuplotes crassus <class 'str'>\n",
      "phylum Ciliophora <class 'str'>\n",
      "class Spirotrichea <class 'str'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MARTKQTARKTTGQKAPRKSVGGSKAPIGAGKSVVKASRKNVPSIIAKQAIKKPHRFRPGTVALREIRKFQKSTDLLIRKLPFQRLVREIATEYKSDLRFQSQAVLALQEATEAYMVSLFEDTNLCAIHAKRVTIMPKDIHLARRIRGERS <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for a in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=a, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"H3.P_(Euplotes_crassus)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "8b65053e-8a3d-49ec-b520-594a3b1538a2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "a339d3dc-cbd4-4466-b6b1-fffb12c7c36f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>AAC47441.1</td>\n",
       "      <td>H3.P_(Euplotes_crassus)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5936.0</td>\n",
       "      <td>Moneuplotes crassus</td>\n",
       "      <td>Ciliophora</td>\n",
       "      <td>Spirotrichea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKTTGQKAPRKSVGGSKAPIGAGKSVVKASRKNVPSII...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession                  variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "26  AAC47441.1  H3.P_(Euplotes_crassus)  None         None           None   \n",
       "\n",
       "    taxonomy_id             organism      phylum         class taxonomy_group  \\\n",
       "26       5936.0  Moneuplotes crassus  Ciliophora  Spirotrichea           None   \n",
       "\n",
       "    info                                           sequence  \\\n",
       "26  None  MARTKQTARKTTGQKAPRKSVGGSKAPIGAGKSVVKASRKNVPSII...   \n",
       "\n",
       "   variant_under_consideration  \n",
       "26                        None  "
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "749ef98e-8808-41d4-b180-e75ec8e1bd81",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "304a8f9e-9988-48ec-b6e7-43246ed1b1cc",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "aa1dbd17-beef-4bb4-8eda-f99889597a27",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, title, doi, author, year, pubmed_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"jahn_unusual_1997\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "e1538503-d369-4a93-8a7d-5c90c4d33264",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_publication = {\n",
    "    \"id\": \"jahn_unusual_1997\",\n",
    "    \"title\": None,\n",
    "    \"doi\": None,\n",
    "    \"author\": None,\n",
    "    \"year\": None,\n",
    "}\n",
    "cursor.execute(add_publication, data_publication)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "909b1a59-f24d-4897-bf82-f77b38e6af98",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>jahn_unusual_1997</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id title   doi author  year pubmed_id\n",
       "113  jahn_unusual_1997  None  None   None  None      None"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"jahn_unusual_1997\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "ec1b1b07-d881-4eb7-89b2-884bf20faf8d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "c6abf33b-eeea-4b3a-ab3f-6fb87641d381",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>AAC47441.1</td>\n",
       "      <td>H3.P_(Euplotes_crassus)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5936.0</td>\n",
       "      <td>Moneuplotes crassus</td>\n",
       "      <td>Ciliophora</td>\n",
       "      <td>Spirotrichea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKTTGQKAPRKSVGGSKAPIGAGKSVVKASRKNVPSII...</td>\n",
       "      <td>None</td>\n",
       "      <td>AAC47441.1</td>\n",
       "      <td>jahn_unusual_1997</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession                  variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "34  AAC47441.1  H3.P_(Euplotes_crassus)  None         None           None   \n",
       "\n",
       "    taxonomy_id             organism      phylum         class taxonomy_group  \\\n",
       "34       5936.0  Moneuplotes crassus  Ciliophora  Spirotrichea           None   \n",
       "\n",
       "    info                                           sequence  \\\n",
       "34  None  MARTKQTARKTTGQKAPRKSVGGSKAPIGAGKSVVKASRKNVPSII...   \n",
       "\n",
       "   variant_under_consideration sequence_accession     publication_id  \n",
       "34                        None         AAC47441.1  jahn_unusual_1997  "
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "95c5e177-7270-47eb-855c-1eb25fdde516",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02888f1f-a5f3-4388-92fb-2bcfeb6b6cf2",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add sequences to H3.V_(Trypanosomatidae)</span>\n",
    "\n",
    "### <span style=\"color:black\">Add sequence AAO24601.1 to H3.V_(Trypanosomatidae) [lowell_variant_2004]</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "34e86058-58ae-4155-a2af-2abf031f0ea3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"AAO24601.1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82b95adc-30ca-4608-8c27-27852cd8c44a",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "1eb3c838-750b-471b-ad39-720ec810e5f4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "2864b2c0-1dd2-4ee6-ae80-218f352393d7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 5702\n",
      "accession AAO24601.1 <class 'str'>\n",
      "variant H3.V_(Trypanosomatidae) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 5702 <class 'int'>\n",
      "organism Trypanosoma brucei brucei <class 'str'>\n",
      "phylum Euglenozoa <class 'str'>\n",
      "class Kinetoplastea <class 'str'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MAQMKKITPRPVRPKSVASRPIQAVARAPVKKVENTPPQKRHHRWRPGTVALREIRRLQSSTDFLIQRAPFRRFLREVVSNLKDSYRMSAACVDAIQEATETYITSVFMDANLCTLHANRVTLFPKDIQLALKLRGERN <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for a in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=a, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"H3.V_(Trypanosomatidae)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "5b840ab9-c2d7-42f0-ac57-5150cae4a7c9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "d6ea1c33-2be8-4303-9bb5-cfb56154032b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>AAO24601.1</td>\n",
       "      <td>H3.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5702.0</td>\n",
       "      <td>Trypanosoma brucei brucei</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAQMKKITPRPVRPKSVASRPIQAVARAPVKKVENTPPQKRHHRWR...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession                  variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "42  AAO24601.1  H3.V_(Trypanosomatidae)  None         None           None   \n",
       "\n",
       "    taxonomy_id                   organism      phylum          class  \\\n",
       "42       5702.0  Trypanosoma brucei brucei  Euglenozoa  Kinetoplastea   \n",
       "\n",
       "   taxonomy_group  info                                           sequence  \\\n",
       "42           None  None  MAQMKKITPRPVRPKSVASRPIQAVARAPVKKVENTPPQKRHHRWR...   \n",
       "\n",
       "   variant_under_consideration  \n",
       "42                        None  "
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "25618abf-a513-4582-8f38-087cf8c1a8b0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59507fb8-f282-4f82-a94e-f8a0501f8771",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "97563116-f99d-4941-874e-d8ab3309281d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, title, doi, author, year, pubmed_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"lowell_variant_2004\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "7f58806c-6603-4a68-a6b4-397a7480ec2d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_publication = {\n",
    "    \"id\": \"lowell_variant_2004\",\n",
    "    \"title\": None,\n",
    "    \"doi\": None,\n",
    "    \"author\": None,\n",
    "    \"year\": None,\n",
    "}\n",
    "cursor.execute(add_publication, data_publication)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "52fc1850-c2e8-4345-98a7-f7e5a6975e4c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>lowell_variant_2004</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      id title   doi author  year pubmed_id\n",
       "128  lowell_variant_2004  None  None   None  None      None"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"lowell_variant_2004\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "8a39ed25-8380-4995-a314-c82519553bff",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "725dd0af-0d3a-43b3-9c8e-434238775a86",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>AAO24601.1</td>\n",
       "      <td>H3.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5702.0</td>\n",
       "      <td>Trypanosoma brucei brucei</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAQMKKITPRPVRPKSVASRPIQAVARAPVKKVENTPPQKRHHRWR...</td>\n",
       "      <td>None</td>\n",
       "      <td>AAO24601.1</td>\n",
       "      <td>lowell_variant_2004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession                  variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "52  AAO24601.1  H3.V_(Trypanosomatidae)  None         None           None   \n",
       "\n",
       "    taxonomy_id                   organism      phylum          class  \\\n",
       "52       5702.0  Trypanosoma brucei brucei  Euglenozoa  Kinetoplastea   \n",
       "\n",
       "   taxonomy_group  info                                           sequence  \\\n",
       "52           None  None  MAQMKKITPRPVRPKSVASRPIQAVARAPVKKVENTPPQKRHHRWR...   \n",
       "\n",
       "   variant_under_consideration sequence_accession       publication_id  \n",
       "52                        None         AAO24601.1  lowell_variant_2004  "
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "4f19e952-a5b9-4374-b833-5203cfbf9e39",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17965753-c133-456f-8f72-4494eff01ae9",
   "metadata": {},
   "source": [
    "### <span style=\"color:black\">Add fasta sequences to H3.V_(Trypanosomatidae)</span>\n",
    "\n",
    "Sequences fasta [lowell_variant_2004]:\n",
    "\n",
    "```fasta\n",
    ">ACC histone H3 variant [Trypanosoma cruzi]\n",
    "MGSLKKVASVEKHSAFSSESKLPKPRKNLASRKIHTEGRIVAKKESASAGTRKKHRWRPGTVVLREVRRYQSSTEFLIAKAP\n",
    "FRRLVREIVSNLKDSFRMSATCVEALQESTELYVTSVLADANLCTLHANRVTVYPKDIQLALKLRGERL\n",
    ">Q9U196 UniPrac [Leishmania major]\n",
    "MAGITKAAVVASHPKKNVASRKMNKKSRSIAKKEAKAMRADSAGAKSRRWRPGTVALREVRKYQRSTELLIARTPFRRLVKE\n",
    "IMSTFKDTMHMRHSALEAMQDATESYLVSLLCDANLCTIHAKRVTLYPKDLQLALRLRGERT\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "fa3899ba-a1cd-4711-b53e-14bfc2d34768",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAO24601.1</td>\n",
       "      <td>H3.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5702</td>\n",
       "      <td>Trypanosoma brucei brucei</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAQMKKITPRPVRPKSVASRPIQAVARAPVKKVENTPPQKRHHRWR...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    accession                  variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  AAO24601.1  H3.V_(Trypanosomatidae)  None         None           None   \n",
       "\n",
       "   taxonomy_id                   organism      phylum          class  \\\n",
       "0         5702  Trypanosoma brucei brucei  Euglenozoa  Kinetoplastea   \n",
       "\n",
       "  taxonomy_group  info                                           sequence  \\\n",
       "0           None  None  MAQMKKITPRPVRPKSVASRPIQAVARAPVKKVENTPPQKRHHRWR...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                        None  "
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE variant='H3.V_(Trypanosomatidae)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "bc8e668a-5dc0-4420-8832-9d744009d54d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_sequence = {\n",
    "    \"accession\": \"HISTDB_H3_V_Trypanosomatidae_0\",\n",
    "    \"variant\": \"H3.V_(Trypanosomatidae)\",\n",
    "    \"gi\": None,\n",
    "    \"ncbi_gene_id\": None,\n",
    "    \"hgnc_gene_name\": None,\n",
    "    \"taxonomy_id\": 5693,\n",
    "    \"organism\": \"Trypanosoma cruzi\",\n",
    "    \"phylum\": \"Euglenozoa\",\n",
    "    \"class\": \"Kinetoplastea\",\n",
    "    \"taxonomy_group\": None,\n",
    "    \"info\": None,\n",
    "    \"sequence\": \"MGSLKKVASVEKHSAFSSESKLPKPRKNLASRKIHTEGRIVAKKESASAGTRKKHRWRPGTVVLREVRRYQSSTEFLIAKAPFRRLVREIVSNLKDSFRMSATCVEALQESTELYVTSVLADANLCTLHANRVTVYPKDIQLALKLRGERL\",\n",
    "    \"variant_under_consideration\": None,\n",
    "}\n",
    "cursor.execute(add_sequence, data_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "a18aab8f-6737-48b7-b7ec-b7fc067519c6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_sequence = {\n",
    "    \"accession\": \"HISTDB_H3_V_Trypanosomatidae_1\",\n",
    "    \"variant\": \"H3.V_(Trypanosomatidae)\",\n",
    "    \"gi\": None,\n",
    "    \"ncbi_gene_id\": None,\n",
    "    \"hgnc_gene_name\": None,\n",
    "    \"taxonomy_id\": 5664,\n",
    "    \"organism\": \"Leishmania major\",\n",
    "    \"phylum\": \"Euglenozoa\",\n",
    "    \"class\": \" Kinetoplastea\",\n",
    "    \"taxonomy_group\": None,\n",
    "    \"info\": None,\n",
    "    \"sequence\": \"MAGITKAAVVASHPKKNVASRKMNKKSRSIAKKEAKAMRADSAGAKSRRWRPGTVALREVRKYQRSTELLIARTPFRRLVKEIMSTFKDTMHMRHSALEAMQDATESYLVSLLCDANLCTIHAKRVTLYPKDLQLALRLRGERT\",\n",
    "    \"variant_under_consideration\": None,\n",
    "}\n",
    "cursor.execute(add_sequence, data_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "386c8a63-6991-49f6-a257-237d3a689eae",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAO24601.1</td>\n",
       "      <td>H3.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5702</td>\n",
       "      <td>Trypanosoma brucei brucei</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAQMKKITPRPVRPKSVASRPIQAVARAPVKKVENTPPQKRHHRWR...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HISTDB_H3_V_Trypanosomatidae_0</td>\n",
       "      <td>H3.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5693</td>\n",
       "      <td>Trypanosoma cruzi</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGSLKKVASVEKHSAFSSESKLPKPRKNLASRKIHTEGRIVAKKES...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HISTDB_H3_V_Trypanosomatidae_1</td>\n",
       "      <td>H3.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5664</td>\n",
       "      <td>Leishmania major</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAGITKAAVVASHPKKNVASRKMNKKSRSIAKKEAKAMRADSAGAK...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        accession                  variant    gi ncbi_gene_id  \\\n",
       "0                      AAO24601.1  H3.V_(Trypanosomatidae)  None         None   \n",
       "1  HISTDB_H3_V_Trypanosomatidae_0  H3.V_(Trypanosomatidae)  None         None   \n",
       "2  HISTDB_H3_V_Trypanosomatidae_1  H3.V_(Trypanosomatidae)  None         None   \n",
       "\n",
       "  hgnc_gene_name  taxonomy_id                   organism      phylum  \\\n",
       "0           None         5702  Trypanosoma brucei brucei  Euglenozoa   \n",
       "1           None         5693          Trypanosoma cruzi  Euglenozoa   \n",
       "2           None         5664           Leishmania major  Euglenozoa   \n",
       "\n",
       "            class taxonomy_group  info  \\\n",
       "0   Kinetoplastea           None  None   \n",
       "1   Kinetoplastea           None  None   \n",
       "2   Kinetoplastea           None  None   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MAQMKKITPRPVRPKSVASRPIQAVARAPVKKVENTPPQKRHHRWR...   \n",
       "1  MGSLKKVASVEKHSAFSSESKLPKPRKNLASRKIHTEGRIVAKKES...   \n",
       "2  MAGITKAAVVASHPKKNVASRKMNKKSRSIAKKEAKAMRADSAGAK...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                        None  \n",
       "1                        None  \n",
       "2                        None  "
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE variant='H3.V_(Trypanosomatidae)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "7509b302-8f03-4bf3-b5f2-dc3a687b8f6e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c2ab6f3-36b2-483b-9403-b4d4b5d04e34",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "a699d00d-6a9e-4550-a5c0-2c81ab55d48a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>lowell_variant_2004</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      id title   doi author  year pubmed_id\n",
       "128  lowell_variant_2004  None  None   None  None      None"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"lowell_variant_2004\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "6833c79a-69f8-4b11-a56b-93c737898798",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in [\"HISTDB_H3_V_Trypanosomatidae_0\", \"HISTDB_H3_V_Trypanosomatidae_1\"]:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "fd60c0fc-0f65-430f-84fd-1c5da8f2a420",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2584</th>\n",
       "      <td>HISTDB_H3_V_Trypanosomatidae_0</td>\n",
       "      <td>H3.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5693.0</td>\n",
       "      <td>Trypanosoma cruzi</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGSLKKVASVEKHSAFSSESKLPKPRKNLASRKIHTEGRIVAKKES...</td>\n",
       "      <td>None</td>\n",
       "      <td>HISTDB_H3_V_Trypanosomatidae_0</td>\n",
       "      <td>lowell_variant_2004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2585</th>\n",
       "      <td>HISTDB_H3_V_Trypanosomatidae_1</td>\n",
       "      <td>H3.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5664.0</td>\n",
       "      <td>Leishmania major</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAGITKAAVVASHPKKNVASRKMNKKSRSIAKKEAKAMRADSAGAK...</td>\n",
       "      <td>None</td>\n",
       "      <td>HISTDB_H3_V_Trypanosomatidae_1</td>\n",
       "      <td>lowell_variant_2004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           accession                  variant    gi  \\\n",
       "2584  HISTDB_H3_V_Trypanosomatidae_0  H3.V_(Trypanosomatidae)  None   \n",
       "2585  HISTDB_H3_V_Trypanosomatidae_1  H3.V_(Trypanosomatidae)  None   \n",
       "\n",
       "     ncbi_gene_id hgnc_gene_name  taxonomy_id           organism      phylum  \\\n",
       "2584         None           None       5693.0  Trypanosoma cruzi  Euglenozoa   \n",
       "2585         None           None       5664.0   Leishmania major  Euglenozoa   \n",
       "\n",
       "               class taxonomy_group  info  \\\n",
       "2584   Kinetoplastea           None  None   \n",
       "2585   Kinetoplastea           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "2584  MGSLKKVASVEKHSAFSSESKLPKPRKNLASRKIHTEGRIVAKKES...   \n",
       "2585  MAGITKAAVVASHPKKNVASRKMNKKSRSIAKKEAKAMRADSAGAK...   \n",
       "\n",
       "     variant_under_consideration              sequence_accession  \\\n",
       "2584                        None  HISTDB_H3_V_Trypanosomatidae_0   \n",
       "2585                        None  HISTDB_H3_V_Trypanosomatidae_1   \n",
       "\n",
       "           publication_id  \n",
       "2584  lowell_variant_2004  \n",
       "2585  lowell_variant_2004  "
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin([\"HISTDB_H3_V_Trypanosomatidae_0\", \"HISTDB_H3_V_Trypanosomatidae_1\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "a77a3687-ed17-4514-9f2d-4ec772643fe0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c26d484-0b6a-4746-9f65-229240910af2",
   "metadata": {},
   "source": [
    "### <span style=\"color:black\">Add sequence XP_828007.1 to H3.V_(Trypanosomatidae) [schulz_base_2016]</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "89ac64d6-0825-45bc-b8bf-c67c036a0586",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"XP_828007.1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42a77af3-efaa-4726-9c5a-ffeb0f147b6a",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "cd085d4b-4088-48d8-84bd-99ec5353d656",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "a4ec4d01-9c97-4296-bbea-f0b9289c70da",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 185431\n",
      "accession XP_828007.1 <class 'str'>\n",
      "variant H3.V_(Trypanosomatidae) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 185431 <class 'int'>\n",
      "organism Trypanosoma brucei brucei TREU927 <class 'str'>\n",
      "phylum Euglenozoa <class 'str'>\n",
      "class Kinetoplastea <class 'str'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MAQMKKITPRPVRPKSVASRPIQSVARAPVKKVENTPPQKRHHRWRPGTVALREIRRLQSSTDFLIQRAPFRRFLREVVSNLKDSYRMSAACVDAIQEATETYITSVFMDANLCTLHANRVTLFPKDIQLALKLRGERN <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for a in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=a, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"H3.V_(Trypanosomatidae)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "d9f04455-b181-4085-a231-1f4a51e1acfd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "f11e52a2-7667-48ad-8a86-de1c2867c76a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5933</th>\n",
       "      <td>XP_828007.1</td>\n",
       "      <td>H3.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>185431.0</td>\n",
       "      <td>Trypanosoma brucei brucei TREU927</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAQMKKITPRPVRPKSVASRPIQSVARAPVKKVENTPPQKRHHRWR...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession                  variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "5933  XP_828007.1  H3.V_(Trypanosomatidae)  None         None           None   \n",
       "\n",
       "      taxonomy_id                           organism      phylum  \\\n",
       "5933     185431.0  Trypanosoma brucei brucei TREU927  Euglenozoa   \n",
       "\n",
       "              class taxonomy_group  info  \\\n",
       "5933  Kinetoplastea           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "5933  MAQMKKITPRPVRPKSVASRPIQSVARAPVKKVENTPPQKRHHRWR...   \n",
       "\n",
       "     variant_under_consideration  \n",
       "5933                        None  "
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "985cdfff-658b-4700-a0ef-4c339570a04a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ff1730c-0c38-4981-b505-75978c5ab0d8",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "c19b56d2-d774-4dd7-8146-c839ba6efafa",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, title, doi, author, year, pubmed_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"schulz_base_2016\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "id": "44b24d6f-843b-4565-8a8a-ac62d8fc38f9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_publication = {\n",
    "    \"id\": \"schulz_base_2016\",\n",
    "    \"title\": None,\n",
    "    \"doi\": None,\n",
    "    \"author\": None,\n",
    "    \"year\": None,\n",
    "}\n",
    "cursor.execute(add_publication, data_publication)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "id": "57fbb622-a61f-4acb-90d8-4219a25c3877",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>schulz_base_2016</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   id title   doi author  year pubmed_id\n",
       "159  schulz_base_2016  None  None   None  None      None"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"schulz_base_2016\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "fa680f9a-562e-4186-80b3-cc7025400485",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "c1714cb6-b7f3-48d2-82a4-75db5b9cd194",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6558</th>\n",
       "      <td>XP_828007.1</td>\n",
       "      <td>H3.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>185431.0</td>\n",
       "      <td>Trypanosoma brucei brucei TREU927</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAQMKKITPRPVRPKSVASRPIQSVARAPVKKVENTPPQKRHHRWR...</td>\n",
       "      <td>None</td>\n",
       "      <td>XP_828007.1</td>\n",
       "      <td>schulz_base_2016</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession                  variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "6558  XP_828007.1  H3.V_(Trypanosomatidae)  None         None           None   \n",
       "\n",
       "      taxonomy_id                           organism      phylum  \\\n",
       "6558     185431.0  Trypanosoma brucei brucei TREU927  Euglenozoa   \n",
       "\n",
       "              class taxonomy_group  info  \\\n",
       "6558  Kinetoplastea           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "6558  MAQMKKITPRPVRPKSVASRPIQSVARAPVKKVENTPPQKRHHRWR...   \n",
       "\n",
       "     variant_under_consideration sequence_accession    publication_id  \n",
       "6558                        None        XP_828007.1  schulz_base_2016  "
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "id": "a6c39f5c-0ed0-4dae-90e9-2d085b3e28f5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48439846-6b0b-4bd2-bc38-27f6449d7f81",
   "metadata": {},
   "source": [
    "### <span style=\"color:black\">Add fasta sequences to H3.V_(Trypanosomatidae)</span>\n",
    "\n",
    "Sequences fasta [siegel_four_2009]:\n",
    "\n",
    "```fasta\n",
    ">ACC histone H3 variant [Trypanosoma brucei]\n",
    "MAQMKITPRPVRPKSVASRPIQSVARAPVKKVENTPPQKRHHRWRPGTVALREIRRLQSSTDFLIQRAPFRRFLREVVSNLK\n",
    "DSYRMSAACVDAIQEATETYITSVFMDANLCTLHANRVTLFPKDIQLALKLRGERN\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "id": "dd3f24eb-af56-4a2a-9832-505a034455f7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAO24601.1</td>\n",
       "      <td>H3.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5702</td>\n",
       "      <td>Trypanosoma brucei brucei</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAQMKKITPRPVRPKSVASRPIQAVARAPVKKVENTPPQKRHHRWR...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HISTDB_H3_V_Trypanosomatidae_0</td>\n",
       "      <td>H3.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5693</td>\n",
       "      <td>Trypanosoma cruzi</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGSLKKVASVEKHSAFSSESKLPKPRKNLASRKIHTEGRIVAKKES...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HISTDB_H3_V_Trypanosomatidae_1</td>\n",
       "      <td>H3.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5664</td>\n",
       "      <td>Leishmania major</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAGITKAAVVASHPKKNVASRKMNKKSRSIAKKEAKAMRADSAGAK...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XP_828007.1</td>\n",
       "      <td>H3.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>185431</td>\n",
       "      <td>Trypanosoma brucei brucei TREU927</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAQMKKITPRPVRPKSVASRPIQSVARAPVKKVENTPPQKRHHRWR...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        accession                  variant    gi ncbi_gene_id  \\\n",
       "0                      AAO24601.1  H3.V_(Trypanosomatidae)  None         None   \n",
       "1  HISTDB_H3_V_Trypanosomatidae_0  H3.V_(Trypanosomatidae)  None         None   \n",
       "2  HISTDB_H3_V_Trypanosomatidae_1  H3.V_(Trypanosomatidae)  None         None   \n",
       "3                     XP_828007.1  H3.V_(Trypanosomatidae)  None         None   \n",
       "\n",
       "  hgnc_gene_name  taxonomy_id                           organism      phylum  \\\n",
       "0           None         5702          Trypanosoma brucei brucei  Euglenozoa   \n",
       "1           None         5693                  Trypanosoma cruzi  Euglenozoa   \n",
       "2           None         5664                   Leishmania major  Euglenozoa   \n",
       "3           None       185431  Trypanosoma brucei brucei TREU927  Euglenozoa   \n",
       "\n",
       "            class taxonomy_group  info  \\\n",
       "0   Kinetoplastea           None  None   \n",
       "1   Kinetoplastea           None  None   \n",
       "2   Kinetoplastea           None  None   \n",
       "3   Kinetoplastea           None  None   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MAQMKKITPRPVRPKSVASRPIQAVARAPVKKVENTPPQKRHHRWR...   \n",
       "1  MGSLKKVASVEKHSAFSSESKLPKPRKNLASRKIHTEGRIVAKKES...   \n",
       "2  MAGITKAAVVASHPKKNVASRKMNKKSRSIAKKEAKAMRADSAGAK...   \n",
       "3  MAQMKKITPRPVRPKSVASRPIQSVARAPVKKVENTPPQKRHHRWR...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                        None  \n",
       "1                        None  \n",
       "2                        None  \n",
       "3                        None  "
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE variant='H3.V_(Trypanosomatidae)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "id": "f916ab90-981d-40bb-aada-5bbdc73e6fe9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_sequence = {\n",
    "    \"accession\": \"HISTDB_H3_V_Trypanosomatidae_2\",\n",
    "    \"variant\": \"H3.V_(Trypanosomatidae)\",\n",
    "    \"gi\": None,\n",
    "    \"ncbi_gene_id\": None,\n",
    "    \"hgnc_gene_name\": None,\n",
    "    \"taxonomy_id\": 5691,\n",
    "    \"organism\": \"Trypanosoma brucei\",\n",
    "    \"phylum\": \"Euglenozoa\",\n",
    "    \"class\": \"Kinetoplastea\",\n",
    "    \"taxonomy_group\": None,\n",
    "    \"info\": None,\n",
    "    \"sequence\": \"MAQMKITPRPVRPKSVASRPIQSVARAPVKKVENTPPQKRHHRWRPGTVALREIRRLQSSTDFLIQRAPFRRFLREVVSNLKDSYRMSAACVDAIQEATETYITSVFMDANLCTLHANRVTLFPKDIQLALKLRGERN\",\n",
    "    \"variant_under_consideration\": None,\n",
    "}\n",
    "cursor.execute(add_sequence, data_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "26af769e-01dd-422d-8e17-f322cac4c95c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAO24601.1</td>\n",
       "      <td>H3.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5702</td>\n",
       "      <td>Trypanosoma brucei brucei</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAQMKKITPRPVRPKSVASRPIQAVARAPVKKVENTPPQKRHHRWR...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HISTDB_H3_V_Trypanosomatidae_0</td>\n",
       "      <td>H3.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5693</td>\n",
       "      <td>Trypanosoma cruzi</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MGSLKKVASVEKHSAFSSESKLPKPRKNLASRKIHTEGRIVAKKES...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HISTDB_H3_V_Trypanosomatidae_1</td>\n",
       "      <td>H3.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5664</td>\n",
       "      <td>Leishmania major</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAGITKAAVVASHPKKNVASRKMNKKSRSIAKKEAKAMRADSAGAK...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HISTDB_H3_V_Trypanosomatidae_2</td>\n",
       "      <td>H3.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5691</td>\n",
       "      <td>Trypanosoma brucei</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAQMKITPRPVRPKSVASRPIQSVARAPVKKVENTPPQKRHHRWRP...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XP_828007.1</td>\n",
       "      <td>H3.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>185431</td>\n",
       "      <td>Trypanosoma brucei brucei TREU927</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAQMKKITPRPVRPKSVASRPIQSVARAPVKKVENTPPQKRHHRWR...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        accession                  variant    gi ncbi_gene_id  \\\n",
       "0                      AAO24601.1  H3.V_(Trypanosomatidae)  None         None   \n",
       "1  HISTDB_H3_V_Trypanosomatidae_0  H3.V_(Trypanosomatidae)  None         None   \n",
       "2  HISTDB_H3_V_Trypanosomatidae_1  H3.V_(Trypanosomatidae)  None         None   \n",
       "3  HISTDB_H3_V_Trypanosomatidae_2  H3.V_(Trypanosomatidae)  None         None   \n",
       "4                     XP_828007.1  H3.V_(Trypanosomatidae)  None         None   \n",
       "\n",
       "  hgnc_gene_name  taxonomy_id                           organism      phylum  \\\n",
       "0           None         5702          Trypanosoma brucei brucei  Euglenozoa   \n",
       "1           None         5693                  Trypanosoma cruzi  Euglenozoa   \n",
       "2           None         5664                   Leishmania major  Euglenozoa   \n",
       "3           None         5691                 Trypanosoma brucei  Euglenozoa   \n",
       "4           None       185431  Trypanosoma brucei brucei TREU927  Euglenozoa   \n",
       "\n",
       "            class taxonomy_group  info  \\\n",
       "0   Kinetoplastea           None  None   \n",
       "1   Kinetoplastea           None  None   \n",
       "2   Kinetoplastea           None  None   \n",
       "3   Kinetoplastea           None  None   \n",
       "4   Kinetoplastea           None  None   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MAQMKKITPRPVRPKSVASRPIQAVARAPVKKVENTPPQKRHHRWR...   \n",
       "1  MGSLKKVASVEKHSAFSSESKLPKPRKNLASRKIHTEGRIVAKKES...   \n",
       "2  MAGITKAAVVASHPKKNVASRKMNKKSRSIAKKEAKAMRADSAGAK...   \n",
       "3  MAQMKITPRPVRPKSVASRPIQSVARAPVKKVENTPPQKRHHRWRP...   \n",
       "4  MAQMKKITPRPVRPKSVASRPIQSVARAPVKKVENTPPQKRHHRWR...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                        None  \n",
       "1                        None  \n",
       "2                        None  \n",
       "3                        None  \n",
       "4                        None  "
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE variant='H3.V_(Trypanosomatidae)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "80a636d4-522c-47c6-b003-7ba224a8af92",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2bfc693-dd6a-4f09-bca0-c4a50cc2539a",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "badfc79b-0579-41f4-a338-c3fd6419668e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, title, doi, author, year, pubmed_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 194,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"siegel_four_2009\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "id": "3f1ae212-c9f5-47b9-b95d-bbecb2e1542b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_publication = {\n",
    "    \"id\": \"siegel_four_2009\",\n",
    "    \"title\": None,\n",
    "    \"doi\": None,\n",
    "    \"author\": None,\n",
    "    \"year\": None,\n",
    "}\n",
    "cursor.execute(add_publication, data_publication)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "id": "db1ad4db-d5a6-46f5-affb-7bd04aaded71",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>siegel_four_2009</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   id title   doi author  year pubmed_id\n",
       "166  siegel_four_2009  None  None   None  None      None"
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"siegel_four_2009\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "d0f2a2ea-f525-49e8-b412-f7cc6ac3d8d1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in [\"HISTDB_H3_V_Trypanosomatidae_2\"]:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "id": "72ca4ac0-a7b3-440b-80fe-7a34da49f2f6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2586</th>\n",
       "      <td>HISTDB_H3_V_Trypanosomatidae_2</td>\n",
       "      <td>H3.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5691.0</td>\n",
       "      <td>Trypanosoma brucei</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAQMKITPRPVRPKSVASRPIQSVARAPVKKVENTPPQKRHHRWRP...</td>\n",
       "      <td>None</td>\n",
       "      <td>HISTDB_H3_V_Trypanosomatidae_2</td>\n",
       "      <td>siegel_four_2009</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           accession                  variant    gi  \\\n",
       "2586  HISTDB_H3_V_Trypanosomatidae_2  H3.V_(Trypanosomatidae)  None   \n",
       "\n",
       "     ncbi_gene_id hgnc_gene_name  taxonomy_id            organism      phylum  \\\n",
       "2586         None           None       5691.0  Trypanosoma brucei  Euglenozoa   \n",
       "\n",
       "              class taxonomy_group  info  \\\n",
       "2586  Kinetoplastea           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "2586  MAQMKITPRPVRPKSVASRPIQSVARAPVKKVENTPPQKRHHRWRP...   \n",
       "\n",
       "     variant_under_consideration              sequence_accession  \\\n",
       "2586                        None  HISTDB_H3_V_Trypanosomatidae_2   \n",
       "\n",
       "        publication_id  \n",
       "2586  siegel_four_2009  "
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin([\"HISTDB_H3_V_Trypanosomatidae_2\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "26d6a18a-e5da-46a0-8bcc-1afe6d1317a2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a97b416-8083-4b77-b591-d6bd82a2649d",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add sequences to cH3_(Protists)</span>\n",
    "\n",
    "### <span style=\"color:black\">Add fasta sequences to cH3_(Protists)</span>\n",
    "\n",
    "Sequences fasta [lowell_variant_2004]:\n",
    "\n",
    "```fasta\n",
    ">ACC histone H3 variant [Trypanosoma brucei]\n",
    "MSRTKETARTKKTITSKKSKKASKGSDAASGVKTAQRRWRPGTVALREIROFQRSTDLLLQKAPFQRLVREVSGAQKEGLRF\n",
    "QSSAILAAQEATESYIVSLLADTNRACIHSGRVTIQPKDIHLALCLRGERA\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "id": "117e7b98-1024-4784-8746-3b5a9d4d9cdd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAF00592.1</td>\n",
       "      <td>cH3_(Protists)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5741</td>\n",
       "      <td>Giardia duodenalis</td>\n",
       "      <td>Fornicata</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHTARKTTSATKAPRKTIARKAARKTASSTSGIKKTGRKKQG...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    accession         variant    gi ncbi_gene_id hgnc_gene_name  taxonomy_id  \\\n",
       "0  AAF00592.1  cH3_(Protists)  None         None           None         5741   \n",
       "\n",
       "             organism     phylum class taxonomy_group  info  \\\n",
       "0  Giardia duodenalis  Fornicata  None           None  None   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MARTKHTARKTTSATKAPRKTIARKAARKTASSTSGIKKTGRKKQG...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                        None  "
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE variant='cH3_(Protists)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "3c548ff1-0d53-4970-9873-aa4a05c69262",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_sequence = {\n",
    "    \"accession\": \"HISTDB_cH3_Protists_0\",\n",
    "    \"variant\": \"cH3_(Protists)\",\n",
    "    \"gi\": None,\n",
    "    \"ncbi_gene_id\": None,\n",
    "    \"hgnc_gene_name\": None,\n",
    "    \"taxonomy_id\": 5691,\n",
    "    \"organism\": \"Trypanosoma brucei\",\n",
    "    \"phylum\": \"Euglenozoa\",\n",
    "    \"class\": \"Kinetoplastea\",\n",
    "    \"taxonomy_group\": None,\n",
    "    \"info\": None,\n",
    "    \"sequence\": \"MSRTKETARTKKTITSKKSKKASKGSDAASGVKTAQRRWRPGTVALREIROFQRSTDLLLQKAPFQRLVREVSGAQKEGLRFQSSAILAAQEATESYIVSLLADTNRACIHSGRVTIQPKDIHLALCLRGERA\",\n",
    "    \"variant_under_consideration\": None,\n",
    "}\n",
    "cursor.execute(add_sequence, data_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "a8d2ac74-0dc9-4e47-b2cc-1714c651a1d0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAF00592.1</td>\n",
       "      <td>cH3_(Protists)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5741</td>\n",
       "      <td>Giardia duodenalis</td>\n",
       "      <td>Fornicata</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKHTARKTTSATKAPRKTIARKAARKTASSTSGIKKTGRKKQG...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HISTDB_cH3_Protists_0</td>\n",
       "      <td>cH3_(Protists)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5691</td>\n",
       "      <td>Trypanosoma brucei</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MSRTKETARTKKTITSKKSKKASKGSDAASGVKTAQRRWRPGTVAL...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               accession         variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0             AAF00592.1  cH3_(Protists)  None         None           None   \n",
       "1  HISTDB_cH3_Protists_0  cH3_(Protists)  None         None           None   \n",
       "\n",
       "   taxonomy_id            organism      phylum          class taxonomy_group  \\\n",
       "0         5741  Giardia duodenalis   Fornicata           None           None   \n",
       "1         5691  Trypanosoma brucei  Euglenozoa  Kinetoplastea           None   \n",
       "\n",
       "   info                                           sequence  \\\n",
       "0  None  MARTKHTARKTTSATKAPRKTIARKAARKTASSTSGIKKTGRKKQG...   \n",
       "1  None  MSRTKETARTKKTITSKKSKKASKGSDAASGVKTAQRRWRPGTVAL...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                        None  \n",
       "1                        None  "
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE variant='cH3_(Protists)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "49873cd2-3aa4-4c64-9317-c7092f9311e9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a98739b-1319-4fc8-8471-479556f9273d",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "5d513c60-7ab1-4f15-9765-c5ec0df4e44f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>lowell_variant_2004</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      id title   doi author  year pubmed_id\n",
       "128  lowell_variant_2004  None  None   None  None      None"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"lowell_variant_2004\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "ca375942-7ea1-42b0-9ca6-594aa560c476",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in [\"HISTDB_cH3_Protists_0\"]:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "id": "4b1ee685-dea9-458c-9207-318fce86c7bc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2316</th>\n",
       "      <td>HISTDB_cH3_Protists_0</td>\n",
       "      <td>cH3_(Protists)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5691.0</td>\n",
       "      <td>Trypanosoma brucei</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MSRTKETARTKKTITSKKSKKASKGSDAASGVKTAQRRWRPGTVAL...</td>\n",
       "      <td>None</td>\n",
       "      <td>HISTDB_cH3_Protists_0</td>\n",
       "      <td>lowell_variant_2004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  accession         variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "2316  HISTDB_cH3_Protists_0  cH3_(Protists)  None         None           None   \n",
       "\n",
       "      taxonomy_id            organism      phylum          class  \\\n",
       "2316       5691.0  Trypanosoma brucei  Euglenozoa  Kinetoplastea   \n",
       "\n",
       "     taxonomy_group  info                                           sequence  \\\n",
       "2316           None  None  MSRTKETARTKKTITSKKSKKASKGSDAASGVKTAQRRWRPGTVAL...   \n",
       "\n",
       "     variant_under_consideration     sequence_accession       publication_id  \n",
       "2316                        None  HISTDB_cH3_Protists_0  lowell_variant_2004  "
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin([\"HISTDB_cH3_Protists_0\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "434196fb-5f4f-49cf-8bf0-be971f90e651",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c2f1655-5c33-4192-ad7d-fc3786bf02c3",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add sequences to H3.10_(Arabidopsis)</span>\n",
    "\n",
    "### <span style=\"color:black\">Add sequence NP_173418.1 to H3.10_(Arabidopsis) [borg_targeted_2020, alvarez-venegas_canonical_2019, okada_analysis_2005]</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "id": "f2cdac73-945a-4007-a51c-7c2d85402818",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"NP_173418.1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c6413bd-ddb1-40e3-9bff-5a7f580ec9af",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "id": "6b4ebb69-8fa5-495a-a704-b87b56b9d671",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "b456539e-d13a-4c5a-a742-30e762bb57c8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 3702\n",
      "accession NP_173418.1 <class 'str'>\n",
      "variant H3.10_(Arabidopsis) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 3702 <class 'int'>\n",
      "organism Arabidopsis thaliana <class 'str'>\n",
      "phylum Streptophyta <class 'str'>\n",
      "class Magnoliopsida <class 'str'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MARTKQTARKSTGGKGPRKELATKAARKTRRPYRGGVKRAHRFRPGTVALREIRKYQKSTDLLIRKLPFQRLVREIAQDFKVDLRFQSHAVLALQEAAEAYLVGLFEDTNLCAIHAKRVTIMSKDIQLARRIRGERA <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for a in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=a, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"H3.10_(Arabidopsis)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "e9f04439-1d99-4735-8736-a5a9ba827674",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "id": "320b2fcd-0f69-4b58-a129-f7a8a4a8cad4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3575</th>\n",
       "      <td>NP_173418.1</td>\n",
       "      <td>H3.10_(Arabidopsis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702.0</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKGPRKELATKAARKTRRPYRGGVKRAHRFRPG...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession              variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "3575  NP_173418.1  H3.10_(Arabidopsis)  None         None           None   \n",
       "\n",
       "      taxonomy_id              organism        phylum          class  \\\n",
       "3575       3702.0  Arabidopsis thaliana  Streptophyta  Magnoliopsida   \n",
       "\n",
       "     taxonomy_group  info                                           sequence  \\\n",
       "3575           None  None  MARTKQTARKSTGGKGPRKELATKAARKTRRPYRGGVKRAHRFRPG...   \n",
       "\n",
       "     variant_under_consideration  \n",
       "3575                        None  "
      ]
     },
     "execution_count": 213,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "5f18ca79-c34c-4f01-9cec-27108f03f63e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c33b96a9-e06e-4792-9f22-361dcc127f12",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "id": "9dd86506-c7fa-4d7e-8c77-e684446aed30",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>okada_analysis_2005</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 id title   doi author  year pubmed_id\n",
       "74   alvarez-venegas_canonical_2019  None  None   None  None      None\n",
       "148             okada_analysis_2005  None  None   None  None      None"
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"alvarez-venegas_canonical_2019\", \"okada_analysis_2005\", \"borg_targeted_2020\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "id": "bf8376d3-1282-4ba6-bcfa-f8abad173829",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_publication = {\n",
    "    \"id\": \"borg_targeted_2020\",\n",
    "    \"title\": None,\n",
    "    \"doi\": None,\n",
    "    \"author\": None,\n",
    "    \"year\": None,\n",
    "}\n",
    "cursor.execute(add_publication, data_publication)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "id": "9a0390fd-80ba-41ed-bfa7-33d7622cb311",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>borg_targeted_2020</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>okada_analysis_2005</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 id title   doi author  year pubmed_id\n",
       "74   alvarez-venegas_canonical_2019  None  None   None  None      None\n",
       "80               borg_targeted_2020  None  None   None  None      None\n",
       "149             okada_analysis_2005  None  None   None  None      None"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"alvarez-venegas_canonical_2019\", \"okada_analysis_2005\", \"borg_targeted_2020\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "79509cfe-79dc-41ca-a14e-21f7deaca743",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "26019340-738e-49fc-8b54-da5021ceda0e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4008</th>\n",
       "      <td>NP_173418.1</td>\n",
       "      <td>H3.10_(Arabidopsis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702.0</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKGPRKELATKAARKTRRPYRGGVKRAHRFRPG...</td>\n",
       "      <td>None</td>\n",
       "      <td>NP_173418.1</td>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4009</th>\n",
       "      <td>NP_173418.1</td>\n",
       "      <td>H3.10_(Arabidopsis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702.0</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKGPRKELATKAARKTRRPYRGGVKRAHRFRPG...</td>\n",
       "      <td>None</td>\n",
       "      <td>NP_173418.1</td>\n",
       "      <td>borg_targeted_2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4010</th>\n",
       "      <td>NP_173418.1</td>\n",
       "      <td>H3.10_(Arabidopsis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702.0</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSTGGKGPRKELATKAARKTRRPYRGGVKRAHRFRPG...</td>\n",
       "      <td>None</td>\n",
       "      <td>NP_173418.1</td>\n",
       "      <td>okada_analysis_2005</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession              variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "4008  NP_173418.1  H3.10_(Arabidopsis)  None         None           None   \n",
       "4009  NP_173418.1  H3.10_(Arabidopsis)  None         None           None   \n",
       "4010  NP_173418.1  H3.10_(Arabidopsis)  None         None           None   \n",
       "\n",
       "      taxonomy_id              organism        phylum          class  \\\n",
       "4008       3702.0  Arabidopsis thaliana  Streptophyta  Magnoliopsida   \n",
       "4009       3702.0  Arabidopsis thaliana  Streptophyta  Magnoliopsida   \n",
       "4010       3702.0  Arabidopsis thaliana  Streptophyta  Magnoliopsida   \n",
       "\n",
       "     taxonomy_group  info                                           sequence  \\\n",
       "4008           None  None  MARTKQTARKSTGGKGPRKELATKAARKTRRPYRGGVKRAHRFRPG...   \n",
       "4009           None  None  MARTKQTARKSTGGKGPRKELATKAARKTRRPYRGGVKRAHRFRPG...   \n",
       "4010           None  None  MARTKQTARKSTGGKGPRKELATKAARKTRRPYRGGVKRAHRFRPG...   \n",
       "\n",
       "     variant_under_consideration sequence_accession  \\\n",
       "4008                        None        NP_173418.1   \n",
       "4009                        None        NP_173418.1   \n",
       "4010                        None        NP_173418.1   \n",
       "\n",
       "                      publication_id  \n",
       "4008  alvarez-venegas_canonical_2019  \n",
       "4009              borg_targeted_2020  \n",
       "4010             okada_analysis_2005  "
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "5843d313-a236-4e9c-aa83-646d18885cb6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8967b29b-b8ae-42b9-a2ce-5c6e5d4177fe",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add sequences to H3.14_(Arabidopsis)</span>\n",
    "\n",
    "### <span style=\"color:black\">Add sequence NP_177690.1 to H3.14_(Arabidopsis) [alvarez-venegas_canonical_2019, nunez-vazquez_histone_2025]</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "id": "b27c020c-f454-4fbc-8e86-faf02ad03584",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"NP_177690.1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e32f0f8-c552-4aa7-ab74-d3dc4a8de1fe",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "9e3b424d-c072-4298-9589-bd819dd41075",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "e8b88408-eba0-4884-be05-4659d9b7a7a5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 3702\n",
      "accession NP_177690.1 <class 'str'>\n",
      "variant H3.14_(Arabidopsis) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 3702 <class 'int'>\n",
      "organism Arabidopsis thaliana <class 'str'>\n",
      "phylum Streptophyta <class 'str'>\n",
      "class Magnoliopsida <class 'str'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MARTKQTARKSHGGKAPRTLLATKAARKSAPTTGGVKKPHRYRPGTVALREIRKYQKSTELLIRKLPFQRLVREIAQDYKTDLRFQSHAVLALQEAAEAYLVGLFEDTNLCAIHAKRVTIMPKDVQLARRIRGERA <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for a in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=a, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"H3.14_(Arabidopsis)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "b419224e-6e0c-48b8-8d33-a8f6809f81e7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "a3ea18bb-03a9-4778-a5df-48556ceb8da2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3579</th>\n",
       "      <td>NP_177690.1</td>\n",
       "      <td>H3.14_(Arabidopsis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702.0</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSHGGKAPRTLLATKAARKSAPTTGGVKKPHRYRPGT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession              variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "3579  NP_177690.1  H3.14_(Arabidopsis)  None         None           None   \n",
       "\n",
       "      taxonomy_id              organism        phylum          class  \\\n",
       "3579       3702.0  Arabidopsis thaliana  Streptophyta  Magnoliopsida   \n",
       "\n",
       "     taxonomy_group  info                                           sequence  \\\n",
       "3579           None  None  MARTKQTARKSHGGKAPRTLLATKAARKSAPTTGGVKKPHRYRPGT...   \n",
       "\n",
       "     variant_under_consideration  \n",
       "3579                        None  "
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "378f9ab4-5cb2-4d81-ae43-9873021e1bda",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9dfd93f7-935e-4d2b-a96b-ad3186ca74e1",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "d36fda6e-242f-4e66-8d45-5c8eb8f22536",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                id title   doi author  year pubmed_id\n",
       "74  alvarez-venegas_canonical_2019  None  None   None  None      None"
      ]
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"alvarez-venegas_canonical_2019\", \"nunez-vazquez_histone_2025\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "id": "936f9d77-7eac-4661-a5df-34b14f8df0fe",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_publication = {\n",
    "    \"id\": \"nunez-vazquez_histone_2025\",\n",
    "    \"title\": None,\n",
    "    \"doi\": None,\n",
    "    \"author\": None,\n",
    "    \"year\": None,\n",
    "}\n",
    "cursor.execute(add_publication, data_publication)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "444e24c8-213e-4ed7-9385-7835a406d450",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>nunez-vazquez_histone_2025</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 id title   doi author  year pubmed_id\n",
       "74   alvarez-venegas_canonical_2019  None  None   None  None      None\n",
       "148      nunez-vazquez_histone_2025  None  None   None  None      None"
      ]
     },
     "execution_count": 230,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"alvarez-venegas_canonical_2019\", \"nunez-vazquez_histone_2025\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "b0f40cdc-1a29-489d-aa95-f139657278db",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "id": "54c7f37e-4c01-40f9-afdd-82d8d56c4f43",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4017</th>\n",
       "      <td>NP_177690.1</td>\n",
       "      <td>H3.14_(Arabidopsis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702.0</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSHGGKAPRTLLATKAARKSAPTTGGVKKPHRYRPGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>NP_177690.1</td>\n",
       "      <td>alvarez-venegas_canonical_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4018</th>\n",
       "      <td>NP_177690.1</td>\n",
       "      <td>H3.14_(Arabidopsis)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3702.0</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MARTKQTARKSHGGKAPRTLLATKAARKSAPTTGGVKKPHRYRPGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>NP_177690.1</td>\n",
       "      <td>nunez-vazquez_histone_2025</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession              variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "4017  NP_177690.1  H3.14_(Arabidopsis)  None         None           None   \n",
       "4018  NP_177690.1  H3.14_(Arabidopsis)  None         None           None   \n",
       "\n",
       "      taxonomy_id              organism        phylum          class  \\\n",
       "4017       3702.0  Arabidopsis thaliana  Streptophyta  Magnoliopsida   \n",
       "4018       3702.0  Arabidopsis thaliana  Streptophyta  Magnoliopsida   \n",
       "\n",
       "     taxonomy_group  info                                           sequence  \\\n",
       "4017           None  None  MARTKQTARKSHGGKAPRTLLATKAARKSAPTTGGVKKPHRYRPGT...   \n",
       "4018           None  None  MARTKQTARKSHGGKAPRTLLATKAARKSAPTTGGVKKPHRYRPGT...   \n",
       "\n",
       "     variant_under_consideration sequence_accession  \\\n",
       "4017                        None        NP_177690.1   \n",
       "4018                        None        NP_177690.1   \n",
       "\n",
       "                      publication_id  \n",
       "4017  alvarez-venegas_canonical_2019  \n",
       "4018      nunez-vazquez_histone_2025  "
      ]
     },
     "execution_count": 232,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "id": "b5de725a-6299-4836-a387-5283f63d720e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80bb5c82-1f88-4873-9e1b-d07ee7368f38",
   "metadata": {},
   "source": [
    "# DONE H4\n",
    "\n",
    "## <span style=\"color:black\">Add sequences to H4.G_(Hominidae)</span>\n",
    "\n",
    "### <span style=\"color:black\">Add sequence NP_003538.1 as H4.G_(Hominidae) [ding_primate-specific_2021]</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "id": "e730a3d5-7694-4188-9be4-975050ab1c6d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "accessions = [\"NP_003538.1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "231c61cb-877e-467f-8353-db787eb946b0",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add sequences to curatedDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "id": "c8a2e3c7-56bb-4209-a21d-505629f5e42a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 235,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "id": "429e5e13-eea9-4741-9152-9afc0e1724be",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 9606\n",
      "accession NP_003538.1 <class 'str'>\n",
      "variant H4.G_(Hominidae) <class 'str'>\n",
      "gi None <class 'NoneType'>\n",
      "ncbi_gene_id None <class 'NoneType'>\n",
      "hgnc_gene_name None <class 'NoneType'>\n",
      "taxonomy_id 9606 <class 'int'>\n",
      "organism Homo sapiens <class 'str'>\n",
      "phylum Chordata <class 'str'>\n",
      "class Mammalia <class 'str'>\n",
      "taxonomy_group None <class 'NoneType'>\n",
      "info None <class 'NoneType'>\n",
      "sequence MSVRGKAGKGLGKGGAKCHRKVLSDNIQGITKCTIRRLARHGGVKRILGLIYEETRRVFKVFLENVIWYAVTNTEHAKRKTVTAMAVVYVLKRQGRTL <class 'str'>\n",
      "variant_under_consideration None <class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "data_sequence_list = []\n",
    "for a in accessions:\n",
    "    with Entrez.efetch(db=\"protein\", id=a, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "    taxonomy_data = get_taxonomy_data(record)\n",
    "    data_sequence = {\n",
    "        \"accession\": record.id,\n",
    "        \"variant\": \"H4.G_(Hominidae)\",\n",
    "        \"gi\": None,\n",
    "        \"ncbi_gene_id\": None,\n",
    "        \"hgnc_gene_name\": None,\n",
    "        \"taxonomy_id\": None,\n",
    "        \"organism\": None,\n",
    "        \"phylum\": None,\n",
    "        \"class\": None,\n",
    "        \"taxonomy_group\": None,\n",
    "        \"info\": None,\n",
    "        \"sequence\": str(record.seq),\n",
    "        \"variant_under_consideration\": None,\n",
    "    }\n",
    "    data_sequence.update(taxonomy_data)\n",
    "    data_sequence_list.append(data_sequence)\n",
    "for k, v in data_sequence_list[-1].items():\n",
    "    print(k, v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "id": "041eb728-9046-4825-9b3a-424af3f370f1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for ds in data_sequence_list:\n",
    "    cursor.execute(add_sequence, ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "id": "f7d1960b-1fe7-4169-be71-2e0aa054e543",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3520</th>\n",
       "      <td>NP_003538.1</td>\n",
       "      <td>H4.G_(Hominidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MSVRGKAGKGLGKGGAKCHRKVLSDNIQGITKCTIRRLARHGGVKR...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession           variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "3520  NP_003538.1  H4.G_(Hominidae)  None         None           None   \n",
       "\n",
       "      taxonomy_id      organism    phylum     class taxonomy_group  info  \\\n",
       "3520       9606.0  Homo sapiens  Chordata  Mammalia           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "3520  MSVRGKAGKGLGKGGAKCHRKVLSDNIQGITKCTIRRLARHGGVKR...   \n",
       "\n",
       "     variant_under_consideration  \n",
       "3520                        None  "
      ]
     },
     "execution_count": 238,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence \"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "id": "3aaa67ef-bfdb-4486-9108-880d13c09154",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2da06c15-d551-4dc8-9af2-834659f52884",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "id": "1f89c4d6-3655-44d7-b9e8-327a63f690b4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, title, doi, author, year, pubmed_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 240,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"ding_primate-specific_2021\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "id": "915a53b4-669c-4a1b-b67b-7526e5920470",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_publication = {\n",
    "    \"id\": \"ding_primate-specific_2021\",\n",
    "    \"title\": None,\n",
    "    \"doi\": None,\n",
    "    \"author\": None,\n",
    "    \"year\": None,\n",
    "}\n",
    "cursor.execute(add_publication, data_publication)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "id": "6a262917-9a8a-41e8-9f61-e1157c8a7f04",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>ding_primate-specific_2021</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            id title   doi author  year pubmed_id\n",
       "91  ding_primate-specific_2021  None  None   None  None      None"
      ]
     },
     "execution_count": 242,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"ding_primate-specific_2021\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "id": "b8054ea6-6a94-481e-a204-f5bb45c2c08f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in accessions:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "id": "eb8c585d-011e-440d-a8e4-2bcbdd1564e5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3933</th>\n",
       "      <td>NP_003538.1</td>\n",
       "      <td>H4.G_(Hominidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MSVRGKAGKGLGKGGAKCHRKVLSDNIQGITKCTIRRLARHGGVKR...</td>\n",
       "      <td>None</td>\n",
       "      <td>NP_003538.1</td>\n",
       "      <td>ding_primate-specific_2021</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession           variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "3933  NP_003538.1  H4.G_(Hominidae)  None         None           None   \n",
       "\n",
       "      taxonomy_id      organism    phylum     class taxonomy_group  info  \\\n",
       "3933       9606.0  Homo sapiens  Chordata  Mammalia           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "3933  MSVRGKAGKGLGKGGAKCHRKVLSDNIQGITKCTIRRLARHGGVKR...   \n",
       "\n",
       "     variant_under_consideration sequence_accession  \\\n",
       "3933                        None        NP_003538.1   \n",
       "\n",
       "                  publication_id  \n",
       "3933  ding_primate-specific_2021  "
      ]
     },
     "execution_count": 244,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin(accessions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "id": "16dfdedb-d3eb-455d-b333-a81d959402bd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8c29425-3d6e-47db-ba1e-af9555e16f9f",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add sequences to H4.V_(Trypanosomatidae)</span>\n",
    "\n",
    "### <span style=\"color:black\">Add fasta sequences to H4.V_(Trypanosomatidae)</span>\n",
    "\n",
    "Sequences fasta [siegel_four_2009]:\n",
    "\n",
    "```fasta\n",
    ">ACC histone H4 variant [Trypanosoma brucei]\n",
    "MAKGKRVGESKGAQKRQKKVLRDNVRGITRGSIRRLARRAGVKRISGVIYDEVRGVLKTFVESIVRDAGAYTEYSRKKTVTAAHVVFALRKRGKVLYGYD\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "id": "fca9e3ef-a4de-4ceb-b8d8-9a77277931e5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE variant='H4.V_(Trypanosomatidae)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "id": "8e7f3641-8839-46af-a186-33974492703f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_sequence = {\n",
    "    \"accession\": \"HISTDB_H4_V_Trypanosomatidae_0\",\n",
    "    \"variant\": \"H4.V_(Trypanosomatidae)\",\n",
    "    \"gi\": None,\n",
    "    \"ncbi_gene_id\": None,\n",
    "    \"hgnc_gene_name\": None,\n",
    "    \"taxonomy_id\": 5691,\n",
    "    \"organism\": \"Trypanosoma brucei\",\n",
    "    \"phylum\": \"Euglenozoa\",\n",
    "    \"class\": \"Kinetoplastea\",\n",
    "    \"taxonomy_group\": None,\n",
    "    \"info\": None,\n",
    "    \"sequence\": \"MAKGKRVGESKGAQKRQKKVLRDNVRGITRGSIRRLARRAGVKRISGVIYDEVRGVLKTFVESIVRDAGAYTEYSRKKTVTAAHVVFALRKRGKVLYGYD\",\n",
    "    \"variant_under_consideration\": None,\n",
    "}\n",
    "cursor.execute(add_sequence, data_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "id": "2d640e0c-af92-45fc-a0e1-cbcc95d7d15d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HISTDB_H4_V_Trypanosomatidae_0</td>\n",
       "      <td>H4.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5691</td>\n",
       "      <td>Trypanosoma brucei</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAKGKRVGESKGAQKRQKKVLRDNVRGITRGSIRRLARRAGVKRIS...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        accession                  variant    gi ncbi_gene_id  \\\n",
       "0  HISTDB_H4_V_Trypanosomatidae_0  H4.V_(Trypanosomatidae)  None         None   \n",
       "\n",
       "  hgnc_gene_name  taxonomy_id            organism      phylum          class  \\\n",
       "0           None         5691  Trypanosoma brucei  Euglenozoa  Kinetoplastea   \n",
       "\n",
       "  taxonomy_group  info                                           sequence  \\\n",
       "0           None  None  MAKGKRVGESKGAQKRQKKVLRDNVRGITRGSIRRLARRAGVKRIS...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                        None  "
      ]
     },
     "execution_count": 248,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE variant='H4.V_(Trypanosomatidae)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "id": "2fb4b0ee-5848-4ac4-be46-961b7aee7aff",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "531e925b-5588-4dc8-acaa-1bcbc160683a",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "id": "f1b19e48-bf3c-440d-81d4-c8a6c8df5ab9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>siegel_four_2009</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   id title   doi author  year pubmed_id\n",
       "169  siegel_four_2009  None  None   None  None      None"
      ]
     },
     "execution_count": 250,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"siegel_four_2009\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "299d6ea2-4851-437f-8d95-df74f179b085",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in [\"HISTDB_H4_V_Trypanosomatidae_0\"]:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "id": "b3c00f2e-9d7f-4a65-8131-f206e0771647",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2675</th>\n",
       "      <td>HISTDB_H4_V_Trypanosomatidae_0</td>\n",
       "      <td>H4.V_(Trypanosomatidae)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5691.0</td>\n",
       "      <td>Trypanosoma brucei</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAKGKRVGESKGAQKRQKKVLRDNVRGITRGSIRRLARRAGVKRIS...</td>\n",
       "      <td>None</td>\n",
       "      <td>HISTDB_H4_V_Trypanosomatidae_0</td>\n",
       "      <td>siegel_four_2009</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           accession                  variant    gi  \\\n",
       "2675  HISTDB_H4_V_Trypanosomatidae_0  H4.V_(Trypanosomatidae)  None   \n",
       "\n",
       "     ncbi_gene_id hgnc_gene_name  taxonomy_id            organism      phylum  \\\n",
       "2675         None           None       5691.0  Trypanosoma brucei  Euglenozoa   \n",
       "\n",
       "              class taxonomy_group  info  \\\n",
       "2675  Kinetoplastea           None  None   \n",
       "\n",
       "                                               sequence  \\\n",
       "2675  MAKGKRVGESKGAQKRQKKVLRDNVRGITRGSIRRLARRAGVKRIS...   \n",
       "\n",
       "     variant_under_consideration              sequence_accession  \\\n",
       "2675                        None  HISTDB_H4_V_Trypanosomatidae_0   \n",
       "\n",
       "        publication_id  \n",
       "2675  siegel_four_2009  "
      ]
     },
     "execution_count": 252,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin([\"HISTDB_H4_V_Trypanosomatidae_0\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "id": "a1dcb5c4-bbf0-46a7-bca3-a5142de375a5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2b91411-a18b-4c4f-a473-384d1d0be5b5",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add sequences to cH4_(Protists)</span>\n",
    "\n",
    "### <span style=\"color:black\">Add fasta sequences to cH4_(Protists)</span>\n",
    "\n",
    "Sequences fasta [siegel_four_2009]:\n",
    "\n",
    "```fasta\n",
    ">ACC histone H4 variant [Trypanosoma brucei]\n",
    "MAKGKKSGEAKGSQKRQKKVLRENVRGITRGSIRRLARRGGVKRISGVIYDEVRGVLKSFVEGVVRDATAYTEYSRKKTVTAVDVVNALRKRGKILYGYA\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "id": "c99debb2-7753-4366-8df5-e4aee3d9edf3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration]\n",
       "Index: []"
      ]
     },
     "execution_count": 254,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE variant='cH4_(Protists)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "id": "30bcdaaf-5100-4386-8da3-011eb0255410",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_sequence = {\n",
    "    \"accession\": \"HISTDB_cH4_Protists_0\",\n",
    "    \"variant\": \"cH4_(Protists)\",\n",
    "    \"gi\": None,\n",
    "    \"ncbi_gene_id\": None,\n",
    "    \"hgnc_gene_name\": None,\n",
    "    \"taxonomy_id\": 5691,\n",
    "    \"organism\": \"Trypanosoma brucei\",\n",
    "    \"phylum\": \"Euglenozoa\",\n",
    "    \"class\": \"Kinetoplastea\",\n",
    "    \"taxonomy_group\": None,\n",
    "    \"info\": None,\n",
    "    \"sequence\": \"MAKGKKSGEAKGSQKRQKKVLRENVRGITRGSIRRLARRGGVKRISGVIYDEVRGVLKSFVEGVVRDATAYTEYSRKKTVTAVDVVNALRKRGKILYGYA\",\n",
    "    \"variant_under_consideration\": None,\n",
    "}\n",
    "cursor.execute(add_sequence, data_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "id": "c25cc472-5f18-4292-8b47-326b6e9ce71a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HISTDB_cH4_Protists_0</td>\n",
       "      <td>cH4_(Protists)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5691</td>\n",
       "      <td>Trypanosoma brucei</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAKGKKSGEAKGSQKRQKKVLRENVRGITRGSIRRLARRGGVKRIS...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               accession         variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "0  HISTDB_cH4_Protists_0  cH4_(Protists)  None         None           None   \n",
       "\n",
       "   taxonomy_id            organism      phylum          class taxonomy_group  \\\n",
       "0         5691  Trypanosoma brucei  Euglenozoa  Kinetoplastea           None   \n",
       "\n",
       "   info                                           sequence  \\\n",
       "0  None  MAKGKKSGEAKGSQKRQKKVLRENVRGITRGSIRRLARRGGVKRIS...   \n",
       "\n",
       "  variant_under_consideration  \n",
       "0                        None  "
      ]
     },
     "execution_count": 256,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM sequence WHERE variant='cH4_(Protists)'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "id": "6719b27b-be74-4832-b00d-4ec70fe5e90d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f96b3973-a0b0-4c51-95ad-f8014d4cc871",
   "metadata": {},
   "source": [
    "## Add sequence publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "id": "cba169b8-a6d5-4fab-85a3-e03069b1dad9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>author</th>\n",
       "      <th>year</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>siegel_four_2009</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   id title   doi author  year pubmed_id\n",
       "169  siegel_four_2009  None  None   None  None      None"
      ]
     },
     "execution_count": 258,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pids = [\"siegel_four_2009\"]\n",
    "query = f\"SELECT * FROM publication\"\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(pids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "id": "64b3b1c6-5a58-4fd3-9cc3-8b6c4a906848",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for acc in [\"HISTDB_cH4_Protists_0\"]:\n",
    "    for pid in pids:\n",
    "        cursor.execute(add_sequence_has_publication, (acc, pid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "id": "dc9e6d8a-0d47-495d-a26c-94b0bb2b312a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>variant</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>sequence</th>\n",
       "      <th>variant_under_consideration</th>\n",
       "      <th>sequence_accession</th>\n",
       "      <th>publication_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2317</th>\n",
       "      <td>HISTDB_cH4_Protists_0</td>\n",
       "      <td>cH4_(Protists)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5691.0</td>\n",
       "      <td>Trypanosoma brucei</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>Kinetoplastea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MAKGKKSGEAKGSQKRQKKVLRENVRGITRGSIRRLARRGGVKRIS...</td>\n",
       "      <td>None</td>\n",
       "      <td>HISTDB_cH4_Protists_0</td>\n",
       "      <td>siegel_four_2009</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  accession         variant    gi ncbi_gene_id hgnc_gene_name  \\\n",
       "2317  HISTDB_cH4_Protists_0  cH4_(Protists)  None         None           None   \n",
       "\n",
       "      taxonomy_id            organism      phylum          class  \\\n",
       "2317       5691.0  Trypanosoma brucei  Euglenozoa  Kinetoplastea   \n",
       "\n",
       "     taxonomy_group  info                                           sequence  \\\n",
       "2317           None  None  MAKGKKSGEAKGSQKRQKKVLRENVRGITRGSIRRLARRGGVKRIS...   \n",
       "\n",
       "     variant_under_consideration     sequence_accession    publication_id  \n",
       "2317                        None  HISTDB_cH4_Protists_0  siegel_four_2009  "
      ]
     },
     "execution_count": 260,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM sequence s LEFT JOIN sequence_has_publication sp \"\n",
    "    \"ON s.accession = sp.sequence_accession \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"accession\"].isin([\"HISTDB_cH4_Protists_0\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "id": "e5c50a82-7fa5-4279-b102-4c1fd9389942",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e8bad07-ef74-40fd-a93b-e7897df0c591",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Close connections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "id": "e5673501-5845-491a-bbb9-f5d213721d58",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cursor.close()\n",
    "conn.close()\n",
    "tunnel.stop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f404689d-f369-48ba-917d-0451b6e297a0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".conda-histdb_env",
   "language": "python",
   "name": "conda-env-.conda-histdb_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
