{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "53d3ec55-7daf-4ade-8add-2830b709dc54",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import io\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import Entrez, SeqIO\n",
    "from mysql.connector import connection\n",
    "from sshtunnel import SSHTunnelForwarder\n",
    "\n",
    "# Укажите ваш email (обязательно для использования Entrez)\n",
    "Entrez.email = \"your.email@example.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "608c4bfe-b979-4b60-b2b5-b94eeb679cb5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open(\"db_curated_server_info.txt\", \"r\") as file:\n",
    "    lines = file.readlines()\n",
    "\n",
    "config = {}\n",
    "\n",
    "for line in lines:\n",
    "    line = line.strip()\n",
    "    if line and not line.startswith(\"#\"):\n",
    "        key, value = line.split(\"=\", 1)\n",
    "        config[key] = value.strip()\n",
    "\n",
    "server_name = config.get(\"server_name\")\n",
    "srever_port = int(config.get(\"srever_port\"))\n",
    "ssh_password = config.get(\"ssh_password\")\n",
    "ssh_username = config.get(\"ssh_username\")\n",
    "db_adress = config.get(\"db_adress\")\n",
    "db_port = int(config.get(\"db_port\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4e19aa1c-49a3-48c4-b67f-34c94cef608d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33495\n"
     ]
    }
   ],
   "source": [
    "tunnel = SSHTunnelForwarder(\n",
    "    (server_name, srever_port),\n",
    "    ssh_password=ssh_password,\n",
    "    ssh_username=ssh_username,\n",
    "    remote_bind_address=(db_adress, db_port),\n",
    ")\n",
    "tunnel.start()\n",
    "print(tunnel.local_bind_port)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cef2275b-12d0-48f1-ad3c-ce4cdb9aa267",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "conn = connection.MySQLConnection(\n",
    "    user=\"db_user\",\n",
    "    password=\"db_password\",\n",
    "    host=\"localhost\",\n",
    "    port=tunnel.local_bind_port,\n",
    "    database=\"db_name\",\n",
    ")\n",
    "cursor = conn.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f589ea8d-8b83-4331-a1cf-b64ed7621801",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('alternative_name',),\n",
       " ('histone',),\n",
       " ('histone_description',),\n",
       " ('histone_has_publication',),\n",
       " ('publication',),\n",
       " ('sequence',),\n",
       " ('sequence_has_publication',)]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SHOW TABLES;\"\n",
    "cursor.execute(query)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b3dd9ef8-b162-4127-9535-ac7743af8c69",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "add_histone = (\n",
    "    \"INSERT INTO histone \"\n",
    "    \"(id, level, taxonomic_span, taxonomic_span_id, description, parent) \"\n",
    "    \"VALUES (%(id)s, %(level)s, %(taxonomic_span)s, %(taxonomic_span_id)s, %(description)s, %(parent)s)\"\n",
    ")\n",
    "add_histone_description = (\n",
    "    \"INSERT INTO histone_description \"\n",
    "    \"(summary, taxonomy, genes, evolution, expression, knock_out, function, sequence, localization, deposition, structure, interactions, disease, caveats) \"\n",
    "    \"VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)\"\n",
    ")\n",
    "# add_publication = (\n",
    "#     \"INSERT INTO publication \"\n",
    "#     \"(id, title, doi, author, year) \"\n",
    "#     \"VALUES (%(id)s, %(title)s, %(doi)s, %(author)s, %(year)s)\"\n",
    "# )\n",
    "# add_sequence = (\n",
    "#     \"INSERT INTO sequence \"\n",
    "#     \"(accession, variant, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, sequence, variant_under_consideration) \"\n",
    "#     \"VALUES (%(accession)s, %(variant)s, %(gi)s, %(ncbi_gene_id)s, %(hgnc_gene_name)s, %(taxonomy_id)s, %(organism)s, %(phylum)s, %(class)s, %(taxonomy_group)s, %(info)s, %(sequence)s, %(variant_under_consideration)s)\"\n",
    "# )\n",
    "# add_sequence_has_publication = (\n",
    "#     \"INSERT INTO sequence_has_publication \"\n",
    "#     \"(sequence_accession, publication_id) \"\n",
    "#     \"VALUES (%s, %s)\"\n",
    "# )\n",
    "add_alternate_names = (\n",
    "    \"INSERT INTO alternative_name \"\n",
    "    \"(name, taxonomy, gene, splice, histone) \"\n",
    "    \"VALUES (%(name)s, %(taxonomy)s, %(gene)s, %(splice)s, %(histone)s)\"\n",
    ")\n",
    "# add_histone_has_publication = (\n",
    "#     \"INSERT INTO histone_has_publication \"\n",
    "#     \"(histone_id, publication_id) \"\n",
    "#     \"VALUES (%s, %s)\"\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6011bb8a-bab5-4fb5-b4ae-15b83accd15b",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_histone_description = {\n",
    "    \"summary\": None,\n",
    "    \"taxonomy\": None,\n",
    "    \"genes\": None,\n",
    "    \"evolution\": None, \n",
    "    \"expression\": None,\n",
    "    \"knock_out\": None,\n",
    "    \"function\": None,\n",
    "    \"sequence\": None,\n",
    "    \"localization\": None,\n",
    "    \"deposition\": None,\n",
    "    \"structure\": None,\n",
    "    \"interactions\": None,\n",
    "    \"disease\": None,\n",
    "    \"caveats\": None,\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d0d3d7e-47ed-4f2b-8e96-e61eb030872d",
   "metadata": {},
   "source": [
    "# To Do H1\n",
    "\n",
    "## DEPRECATED <span style=\"color:red\">Rename cH1 variants to cH1_(Mammalia?)</span>\n",
    "\n",
    "## <span style=\"color:green\">Update summary of cH1_(Mammalia?)</span>\n",
    "\n",
    "### <span style=\"color:green\">Update summary</span>\n",
    "\n",
    "```cH1_(Mammalia?) is a group of replication dependent linker histones located within large histone gene clusters in mammals [happel_histone_2009, talbert_histone_2021].```\n",
    "\n",
    "## <span style=\"color:green\">Change parent of H1.1-5 variants to cH1</span>\n",
    "\n",
    "## <span style=\"color:green\">Rename H1.1-5 variants to cH1.1-5</span>\n",
    "\n",
    "## <span style=\"color:green\">Update summary of cH1.1-5 variants</span>\n",
    "\n",
    "### <span style=\"color:green\">Update summary</span>\n",
    "\n",
    "```cH1.X is a replication dependent linker histone located within large histone gene clusters [happel_histone_2009, talbert_histone_2021].```\n",
    "\n",
    "## <span style=\"color:green\">Change taxonomic span of cH1.1-5 variants to Mammalia</span>\n",
    "\n",
    "## <span style=\"color:green\">Change parent of TS_H1.6 variant to cH1</span>\n",
    "\n",
    "## <span style=\"color:green\">Rename TS_H1.6 variant to cH1.6</span>\n",
    "\n",
    "## <span style=\"color:green\">Add alternate name TS_H1.6 of cH1.6</span>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9c684d8-bba7-4c8d-b275-ec7b5e25f05c",
   "metadata": {},
   "source": [
    "# DONE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3615724a-f483-405d-9cb9-05717129d9be",
   "metadata": {},
   "source": [
    "## DEPRECATED <span style=\"color:red\">Rename cH1 variants to cH1_(Mammalia?)</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "47016bbc-8e9f-47bc-bdef-aa0d19378ea0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH1</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>8</td>\n",
       "      <td>H1</td>\n",
       "      <td>8</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    id          level taxonomic_span taxonomic_span_id  description parent  \\\n",
       "0  cH1  variant_group           null              null            8     H1   \n",
       "\n",
       "   id summary taxonomy genes  ... knock_out function sequence localization  \\\n",
       "0   8    null     null  null  ...      null     null     null         null   \n",
       "\n",
       "  deposition structure interactions disease caveats relations  \n",
       "0       null      null         null    null    null      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cH1'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f41ff6e6-622b-4c28-bd5e-2969e81ae25e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UPDATE histone SET id='cH1_(Mammalia?)' WHERE id='cH1'\n"
     ]
    }
   ],
   "source": [
    "query = f\"UPDATE histone SET id='cH1_(Mammalia?)' WHERE id='cH1'\"\n",
    "print(query)\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e42c7827-02a9-45b1-b579-7cf95e7066c2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, level, taxonomic_span, taxonomic_span_id, description, parent, id, summary, taxonomy, genes, evolution, expression, knock_out, function, sequence, localization, deposition, structure, interactions, disease, caveats, relations]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 22 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cH1'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ec9c4ee0-89a9-45c2-a058-c125cf6fe9ed",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH1_(Mammalia?)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>8</td>\n",
       "      <td>H1</td>\n",
       "      <td>8</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                id          level taxonomic_span taxonomic_span_id  \\\n",
       "0  cH1_(Mammalia?)  variant_group           null              null   \n",
       "\n",
       "   description parent  id summary taxonomy genes  ... knock_out function  \\\n",
       "0            8     H1   8    null     null  null  ...      null     null   \n",
       "\n",
       "  sequence localization deposition structure interactions disease caveats  \\\n",
       "0     null         null       null      null         null    null    null   \n",
       "\n",
       "  relations  \n",
       "0      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cH1_(Mammalia?)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6fcac61f-742a-4152-86d6-bba094cd3687",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac44a47d-c96b-4c7a-90f3-249afb085da7",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Update summary of cH1_(Mammalia?)</span>\n",
    "\n",
    "### <span style=\"color:black\">Update summary</span>\n",
    "\n",
    "```cH1_(Mammalia?) is a group of replication dependent linker histones located within large histone gene clusters in mammals [happel_histone_2009, talbert_histone_2021].```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "7c6837da-77fd-4f12-8592-05d974ea3069",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH1_(Mammalia?)</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>8</td>\n",
       "      <td>H1</td>\n",
       "      <td>8</td>\n",
       "      <td>cH1_(Mammalia?) is a group of replication depe...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                id          level taxonomic_span taxonomic_span_id  \\\n",
       "0  cH1_(Mammalia?)  variant_group           null              null   \n",
       "\n",
       "   description parent  id                                            summary  \\\n",
       "0            8     H1   8  cH1_(Mammalia?) is a group of replication depe...   \n",
       "\n",
       "  taxonomy genes  ... knock_out function sequence localization deposition  \\\n",
       "0     null  null  ...      null     null     null         null       null   \n",
       "\n",
       "  structure interactions disease caveats relations  \n",
       "0      null         null    null    null      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cH1_(Mammalia?)'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "39b945cd-04f2-4f5a-846e-e3f5d04f0275",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UPDATE histone_description SET summary='cH1 is a group of replication dependent linker histones located within large histone gene clusters in mammals [happel_histone_2009, talbert_histone_2021].' WHERE id=8\n"
     ]
    }
   ],
   "source": [
    "summary_desc = \"cH1 is a group of replication dependent linker histones located within large histone gene clusters in mammals [happel_histone_2009, talbert_histone_2021].\"\n",
    "query = f\"UPDATE histone_description SET summary='{summary_desc}' WHERE id=8\"\n",
    "print(query)\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "abc7f220-1aa4-43a3-aef0-42685b9d8b95",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UPDATE histone SET id='cH1' WHERE id='cH1_(Mammalia?)'\n"
     ]
    }
   ],
   "source": [
    "query = f\"UPDATE histone SET id='cH1' WHERE id='cH1_(Mammalia?)'\"\n",
    "print(query)\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "bb205c1d-0f7d-495a-aac9-325dcdac68ef",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH1</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>8</td>\n",
       "      <td>H1</td>\n",
       "      <td>8</td>\n",
       "      <td>cH1 is a group of replication dependent linker...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    id          level taxonomic_span taxonomic_span_id  description parent  \\\n",
       "0  cH1  variant_group           null              null            8     H1   \n",
       "\n",
       "   id                                            summary taxonomy genes  ...  \\\n",
       "0   8  cH1 is a group of replication dependent linker...     null  null  ...   \n",
       "\n",
       "  knock_out function sequence localization deposition structure interactions  \\\n",
       "0      null     null     null         null       null      null         null   \n",
       "\n",
       "  disease caveats relations  \n",
       "0    null    null      None  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id='cH1'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "35640332-fe46-4d01-becf-a613bfb29fa7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bc3717e-bddc-40b4-b6d3-0a933d2108ce",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Change parent of H1.1-5 variants to cH1</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a43e2516-fea9-4d2d-9870-8c5cd70f9ad8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "variants = [\"H1.1\", \"H1.2\", \"H1.3\", \"H1.4\", \"H1.5\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "688e9d73-559e-41b9-921b-58ab6d15241b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>H1.1</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>11.0</td>\n",
       "      <td>H1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>H1.2</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>13.0</td>\n",
       "      <td>H1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>H1.3</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>14.0</td>\n",
       "      <td>H1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>H1.4</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>15.0</td>\n",
       "      <td>H1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>H1.5</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>16.0</td>\n",
       "      <td>H1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id          level taxonomic_span taxonomic_span_id  description parent\n",
       "115  H1.1  variant_group           Homo              9605         11.0     H1\n",
       "119  H1.2  variant_group           Homo              9605         13.0     H1\n",
       "121  H1.3  variant_group           Homo              9605         14.0     H1\n",
       "123  H1.4  variant_group           Homo              9605         15.0     H1\n",
       "125  H1.5  variant_group           Homo              9605         16.0     H1"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(variants)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "5d620773-4b01-411d-bb92-7a4852fda9d1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UPDATE histone SET parent='cH1', level='variant' WHERE id='H1.1'\n",
      "UPDATE histone SET parent='cH1', level='variant' WHERE id='H1.2'\n",
      "UPDATE histone SET parent='cH1', level='variant' WHERE id='H1.3'\n",
      "UPDATE histone SET parent='cH1', level='variant' WHERE id='H1.4'\n",
      "UPDATE histone SET parent='cH1', level='variant' WHERE id='H1.5'\n"
     ]
    }
   ],
   "source": [
    "for v in variants:\n",
    "    query = f\"UPDATE histone SET parent='cH1', level='variant' WHERE id='{v}'\"\n",
    "    print(query)\n",
    "    cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "711859c2-936f-4319-a8a4-59a20f39af94",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>H1.1</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>11.0</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>H1.2</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>13.0</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>H1.3</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>14.0</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>H1.4</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>15.0</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>H1.5</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>16.0</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id    level taxonomic_span taxonomic_span_id  description parent\n",
       "115  H1.1  variant           Homo              9605         11.0    cH1\n",
       "119  H1.2  variant           Homo              9605         13.0    cH1\n",
       "121  H1.3  variant           Homo              9605         14.0    cH1\n",
       "123  H1.4  variant           Homo              9605         15.0    cH1\n",
       "125  H1.5  variant           Homo              9605         16.0    cH1"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(variants)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcf7b54c-3145-40e6-96cc-6090a26e93e6",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Rename H1.1-5 variants to cH1.1-5</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "10ec843d-8fee-4671-844a-7a16fd47f0f3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>H1.1_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>61.0</td>\n",
       "      <td>H1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>H1.2_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>63.0</td>\n",
       "      <td>H1.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>H1.3_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>64.0</td>\n",
       "      <td>H1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>H1.4_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>65.0</td>\n",
       "      <td>H1.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>H1.5_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>66.0</td>\n",
       "      <td>H1.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      id    level taxonomic_span taxonomic_span_id  \\\n",
       "118  H1.1_(Homo_sapiens)  variant   Homo sapiens              9606   \n",
       "120  H1.2_(Homo_sapiens)  variant   Homo sapiens              9606   \n",
       "122  H1.3_(Homo_sapiens)  variant   Homo sapiens              9606   \n",
       "124  H1.4_(Homo_sapiens)  variant   Homo sapiens              9606   \n",
       "126  H1.5_(Homo_sapiens)  variant   Homo sapiens              9606   \n",
       "\n",
       "     description parent  \n",
       "118         61.0   H1.1  \n",
       "120         63.0   H1.2  \n",
       "122         64.0   H1.3  \n",
       "124         65.0   H1.4  \n",
       "126         66.0   H1.5  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"parent\"].isin(variants)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "8bb76dc3-40d1-48fe-9d3c-f92ab4712571",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UPDATE histone SET parent=null WHERE parent='H1.1'\n",
      "UPDATE histone SET id='cH1.1' WHERE id='H1.1'\n",
      "UPDATE histone SET id='cH1.1_(Homo_sapiens)', parent='cH1.1' WHERE id='H1.1_(Homo_sapiens)'\n",
      "UPDATE histone SET parent=null WHERE parent='H1.2'\n",
      "UPDATE histone SET id='cH1.2' WHERE id='H1.2'\n",
      "UPDATE histone SET id='cH1.2_(Homo_sapiens)', parent='cH1.2' WHERE id='H1.2_(Homo_sapiens)'\n",
      "UPDATE histone SET parent=null WHERE parent='H1.3'\n",
      "UPDATE histone SET id='cH1.3' WHERE id='H1.3'\n",
      "UPDATE histone SET id='cH1.3_(Homo_sapiens)', parent='cH1.3' WHERE id='H1.3_(Homo_sapiens)'\n",
      "UPDATE histone SET parent=null WHERE parent='H1.4'\n",
      "UPDATE histone SET id='cH1.4' WHERE id='H1.4'\n",
      "UPDATE histone SET id='cH1.4_(Homo_sapiens)', parent='cH1.4' WHERE id='H1.4_(Homo_sapiens)'\n",
      "UPDATE histone SET parent=null WHERE parent='H1.5'\n",
      "UPDATE histone SET id='cH1.5' WHERE id='H1.5'\n",
      "UPDATE histone SET id='cH1.5_(Homo_sapiens)', parent='cH1.5' WHERE id='H1.5_(Homo_sapiens)'\n"
     ]
    }
   ],
   "source": [
    "for v in variants:\n",
    "    query = f\"UPDATE histone SET parent=null WHERE parent='{v}'\"\n",
    "    print(query)\n",
    "    cursor.execute(query)\n",
    "    query = f\"UPDATE histone SET id='c{v}' WHERE id='{v}'\"\n",
    "    print(query)\n",
    "    cursor.execute(query)\n",
    "    query = f\"UPDATE histone SET id='c{v}_(Homo_sapiens)', parent='c{v}' WHERE id='{v}_(Homo_sapiens)'\"\n",
    "    print(query)\n",
    "    cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "75c0fdb6-e9ef-4df1-bcb9-fa3d6324f1d5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, level, taxonomic_span, taxonomic_span_id, description, parent]\n",
       "Index: []"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin(variants)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "7210cf40-b3fc-47bb-b29d-07ff59496969",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>cH1.1</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>11.0</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>cH1.2</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>13.0</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>cH1.3</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>14.0</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>cH1.4</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>15.0</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>cH1.5</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>16.0</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id    level taxonomic_span taxonomic_span_id  description parent\n",
       "29  cH1.1  variant           Homo              9605         11.0    cH1\n",
       "31  cH1.2  variant           Homo              9605         13.0    cH1\n",
       "33  cH1.3  variant           Homo              9605         14.0    cH1\n",
       "35  cH1.4  variant           Homo              9605         15.0    cH1\n",
       "37  cH1.5  variant           Homo              9605         16.0    cH1"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"id\"].isin([f\"c{v}\" for v in variants])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "5dbcc66d-bc4d-4a3e-9bc1-7166ffe77759",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>cH1.1_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>61.0</td>\n",
       "      <td>cH1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>cH1.2_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>63.0</td>\n",
       "      <td>cH1.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>cH1.3_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>64.0</td>\n",
       "      <td>cH1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>cH1.4_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>65.0</td>\n",
       "      <td>cH1.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>cH1.5_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>66.0</td>\n",
       "      <td>cH1.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      id    level taxonomic_span taxonomic_span_id  \\\n",
       "30  cH1.1_(Homo_sapiens)  variant   Homo sapiens              9606   \n",
       "32  cH1.2_(Homo_sapiens)  variant   Homo sapiens              9606   \n",
       "34  cH1.3_(Homo_sapiens)  variant   Homo sapiens              9606   \n",
       "36  cH1.4_(Homo_sapiens)  variant   Homo sapiens              9606   \n",
       "38  cH1.5_(Homo_sapiens)  variant   Homo sapiens              9606   \n",
       "\n",
       "    description parent  \n",
       "30         61.0  cH1.1  \n",
       "32         63.0  cH1.2  \n",
       "34         64.0  cH1.3  \n",
       "36         65.0  cH1.4  \n",
       "38         66.0  cH1.5  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"parent\"].isin([f\"c{v}\" for v in variants])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "6cd567d7-7150-488c-9e4f-7a4b19b1b050",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3129d511-7d55-4238-850d-30b939931635",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Update summary of cH1.1-5 variants</span>\n",
    "\n",
    "### <span style=\"color:black\">Update summary</span>\n",
    "\n",
    "```cH1.X is a replication dependent linker histone located within large histone gene clusters [happel_histone_2009, talbert_histone_2021].```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "83533fa7-a5e0-4fdc-90a1-bacdd161e629",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH1.1</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>11</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cH1.2</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>13</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cH1.3</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>14</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cH1.4</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>15</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cH1.5</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>16</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id    level taxonomic_span taxonomic_span_id  description parent\n",
       "0  cH1.1  variant           Homo              9605           11    cH1\n",
       "1  cH1.2  variant           Homo              9605           13    cH1\n",
       "2  cH1.3  variant           Homo              9605           14    cH1\n",
       "3  cH1.4  variant           Homo              9605           15    cH1\n",
       "4  cH1.5  variant           Homo              9605           16    cH1"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone \"\n",
    "    \"WHERE id LIKE 'cH1._'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "60bf6756-bcf6-47a9-9f30-aba05333760a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UPDATE histone_description SET summary='cH1.1 is a replication dependent linker histone located within large histone gene clusters in mammals [happel_histone_2009, talbert_histone_2021].' WHERE id=11\n",
      "UPDATE histone_description SET summary='cH1.2 is a replication dependent linker histone located within large histone gene clusters in mammals [happel_histone_2009, talbert_histone_2021].' WHERE id=13\n",
      "UPDATE histone_description SET summary='cH1.3 is a replication dependent linker histone located within large histone gene clusters in mammals [happel_histone_2009, talbert_histone_2021].' WHERE id=14\n",
      "UPDATE histone_description SET summary='cH1.4 is a replication dependent linker histone located within large histone gene clusters in mammals [happel_histone_2009, talbert_histone_2021].' WHERE id=15\n",
      "UPDATE histone_description SET summary='cH1.5 is a replication dependent linker histone located within large histone gene clusters in mammals [happel_histone_2009, talbert_histone_2021].' WHERE id=16\n"
     ]
    }
   ],
   "source": [
    "for i, row in df.iterrows():\n",
    "    summary_desc = f\"{row['id']} is a replication dependent linker histone located within large histone gene clusters in mammals [happel_histone_2009, talbert_histone_2021].\"\n",
    "    query = f\"UPDATE histone_description SET summary='{summary_desc}' WHERE id={row['description']}\"\n",
    "    print(query)\n",
    "    cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "04e8bdaf-c52b-442b-8dd2-2ac9ae151aba",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "      <th>id</th>\n",
       "      <th>summary</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>knock_out</th>\n",
       "      <th>function</th>\n",
       "      <th>sequence</th>\n",
       "      <th>localization</th>\n",
       "      <th>deposition</th>\n",
       "      <th>structure</th>\n",
       "      <th>interactions</th>\n",
       "      <th>disease</th>\n",
       "      <th>caveats</th>\n",
       "      <th>relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH1.1</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>11</td>\n",
       "      <td>cH1</td>\n",
       "      <td>11</td>\n",
       "      <td>cH1.1 is a replication dependent linker histon...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cH1.2</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>13</td>\n",
       "      <td>cH1</td>\n",
       "      <td>13</td>\n",
       "      <td>cH1.2 is a replication dependent linker histon...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cH1.3</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>14</td>\n",
       "      <td>cH1</td>\n",
       "      <td>14</td>\n",
       "      <td>cH1.3 is a replication dependent linker histon...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cH1.4</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>15</td>\n",
       "      <td>cH1</td>\n",
       "      <td>15</td>\n",
       "      <td>cH1.4 is a replication dependent linker histon...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cH1.5</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>16</td>\n",
       "      <td>cH1</td>\n",
       "      <td>16</td>\n",
       "      <td>cH1.5 is a replication dependent linker histon...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>...</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>null</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      id    level taxonomic_span taxonomic_span_id  description parent  id  \\\n",
       "0  cH1.1  variant           Homo              9605           11    cH1  11   \n",
       "1  cH1.2  variant           Homo              9605           13    cH1  13   \n",
       "2  cH1.3  variant           Homo              9605           14    cH1  14   \n",
       "3  cH1.4  variant           Homo              9605           15    cH1  15   \n",
       "4  cH1.5  variant           Homo              9605           16    cH1  16   \n",
       "\n",
       "                                             summary taxonomy genes  ...  \\\n",
       "0  cH1.1 is a replication dependent linker histon...     null  null  ...   \n",
       "1  cH1.2 is a replication dependent linker histon...     null  null  ...   \n",
       "2  cH1.3 is a replication dependent linker histon...     null  null  ...   \n",
       "3  cH1.4 is a replication dependent linker histon...     null  null  ...   \n",
       "4  cH1.5 is a replication dependent linker histon...     null  null  ...   \n",
       "\n",
       "  knock_out function sequence localization deposition structure interactions  \\\n",
       "0      null     null     null         null       null      null         null   \n",
       "1      null     null     null         null       null      null         null   \n",
       "2      null     null     null         null       null      null         null   \n",
       "3      null     null     null         null       null      null         null   \n",
       "4      null     null     null         null       null      null         null   \n",
       "\n",
       "  disease caveats relations  \n",
       "0    null    null      None  \n",
       "1    null    null      None  \n",
       "2    null    null      None  \n",
       "3    null    null      None  \n",
       "4    null    null      None  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone h LEFT JOIN histone_description hd \"\n",
    "    \"ON h.description = hd.id \"\n",
    "    \"WHERE h.id LIKE 'cH1._'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "aa169e8c-3de3-4096-8514-3dd9a307145b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd183989-499c-438e-98f8-20e98885447d",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Change taxonomic span of cH1.1-5 variants to Mammalia</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "3f97cf77-5bd5-40b4-9d62-67493c986c16",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "variants = [\"cH1.1\", \"cH1.2\", \"cH1.3\", \"cH1.4\", \"cH1.5\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "14a79a84-1a52-4a6c-a694-784e01dade1c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH1.1</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>11</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cH1.2</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>13</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cH1.3</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>14</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cH1.4</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>15</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cH1.5</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>16</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id    level taxonomic_span taxonomic_span_id  description parent\n",
       "0  cH1.1  variant           Homo              9605           11    cH1\n",
       "1  cH1.2  variant           Homo              9605           13    cH1\n",
       "2  cH1.3  variant           Homo              9605           14    cH1\n",
       "3  cH1.4  variant           Homo              9605           15    cH1\n",
       "4  cH1.5  variant           Homo              9605           16    cH1"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone \"\n",
    "    \"WHERE id LIKE 'cH1._'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "945c45f6-57d7-4bc4-a4f9-8eb61665df6d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UPDATE histone SET taxonomic_span='Mammalia', taxonomic_span_id='40674' WHERE id='cH1.1'\n",
      "UPDATE histone SET taxonomic_span='Mammalia', taxonomic_span_id='40674' WHERE id='cH1.2'\n",
      "UPDATE histone SET taxonomic_span='Mammalia', taxonomic_span_id='40674' WHERE id='cH1.3'\n",
      "UPDATE histone SET taxonomic_span='Mammalia', taxonomic_span_id='40674' WHERE id='cH1.4'\n",
      "UPDATE histone SET taxonomic_span='Mammalia', taxonomic_span_id='40674' WHERE id='cH1.5'\n"
     ]
    }
   ],
   "source": [
    "for v in variants:\n",
    "    query = f\"UPDATE histone SET taxonomic_span='Mammalia', taxonomic_span_id='40674' WHERE id='{v}'\"\n",
    "    print(query)\n",
    "    cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "dfb3b036-86d9-46ef-8472-e3bd80e684a9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cH1.1</td>\n",
       "      <td>variant</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>11</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cH1.2</td>\n",
       "      <td>variant</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>13</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cH1.3</td>\n",
       "      <td>variant</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>14</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cH1.4</td>\n",
       "      <td>variant</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>15</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cH1.5</td>\n",
       "      <td>variant</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>16</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id    level taxonomic_span taxonomic_span_id  description parent\n",
       "0  cH1.1  variant       Mammalia             40674           11    cH1\n",
       "1  cH1.2  variant       Mammalia             40674           13    cH1\n",
       "2  cH1.3  variant       Mammalia             40674           14    cH1\n",
       "3  cH1.4  variant       Mammalia             40674           15    cH1\n",
       "4  cH1.5  variant       Mammalia             40674           16    cH1"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone \"\n",
    "    \"WHERE id LIKE 'cH1._'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "f153aa90-a7b1-4cae-842e-07e174465c00",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cc295fa-a82f-425b-aca6-c6d62b1f2f92",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Change parent of TS_H1.6 variant to cH1</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "17e5c991-3a7c-4168-9674-b896aaf0efba",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TS_H1.6</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>19</td>\n",
       "      <td>H1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        id          level taxonomic_span taxonomic_span_id  description parent\n",
       "0  TS_H1.6  variant_group       Mammalia             40674           19     H1"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone \"\n",
    "    \"WHERE id='TS_H1.6'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "4967becc-fddd-4cc0-a41f-6e18b1aef04c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UPDATE histone SET parent='cH1' WHERE id='TS_H1.6'\n"
     ]
    }
   ],
   "source": [
    "query = f\"UPDATE histone SET parent='cH1' WHERE id='TS_H1.6'\"\n",
    "print(query)\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "52ce6137-1881-4892-91ce-767fbceccee6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TS_H1.6</td>\n",
       "      <td>variant_group</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>40674</td>\n",
       "      <td>19</td>\n",
       "      <td>cH1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        id          level taxonomic_span taxonomic_span_id  description parent\n",
       "0  TS_H1.6  variant_group       Mammalia             40674           19    cH1"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone \"\n",
    "    \"WHERE id='TS_H1.6'\"\n",
    ")\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "c155bb45-6585-4df0-9938-59bf6a02c691",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af62b251-ebe6-4d6a-b628-7b2568c150a2",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Rename TS_H1.6 variant to cH1.6</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "ce39aa19-c97e-4e03-91de-ad96874495ef",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>H1.6_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>68.0</td>\n",
       "      <td>TS_H1.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      id    level taxonomic_span taxonomic_span_id  \\\n",
       "127  H1.6_(Homo_sapiens)  variant   Homo sapiens              9606   \n",
       "\n",
       "     description   parent  \n",
       "127         68.0  TS_H1.6  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"parent\"]==\"TS_H1.6\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "304d6250-a2ce-4938-914a-c9f44717ac09",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UPDATE histone SET parent=null WHERE parent='TS_H1.6'\n",
      "UPDATE histone SET id='cH1.6' WHERE id='TS_H1.6'\n",
      "UPDATE histone SET id='cH1.6_(Homo_sapiens)', parent='cH1.6' WHERE id='H1.6_(Homo_sapiens)'\n"
     ]
    }
   ],
   "source": [
    "query = f\"UPDATE histone SET parent=null WHERE parent='TS_H1.6'\"\n",
    "print(query)\n",
    "cursor.execute(query)\n",
    "query = f\"UPDATE histone SET id='cH1.6' WHERE id='TS_H1.6'\"\n",
    "print(query)\n",
    "cursor.execute(query)\n",
    "query = f\"UPDATE histone SET id='cH1.6_(Homo_sapiens)', parent='cH1.6' WHERE id='H1.6_(Homo_sapiens)'\"\n",
    "print(query)\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "712aa02b-1525-47b5-9b73-9d76c35fdf89",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, level, taxonomic_span, taxonomic_span_id, description, parent]\n",
       "Index: []"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"parent\"]==\"TS_H1.6\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "d581882a-f1c2-48e0-995c-917e999b4439",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>level</th>\n",
       "      <th>taxonomic_span</th>\n",
       "      <th>taxonomic_span_id</th>\n",
       "      <th>description</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>cH1.6_(Homo_sapiens)</td>\n",
       "      <td>variant</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>68.0</td>\n",
       "      <td>cH1.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      id    level taxonomic_span taxonomic_span_id  \\\n",
       "40  cH1.6_(Homo_sapiens)  variant   Homo sapiens              9606   \n",
       "\n",
       "    description parent  \n",
       "40         68.0  cH1.6  "
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = (\n",
    "    \"SELECT * FROM histone \"\n",
    ")\n",
    "cursor.execute(query)\n",
    "df = pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])\n",
    "df[df[\"parent\"]==\"cH1.6\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "8e7b30fd-fa55-46bb-a079-787786c7aa98",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20642893-338b-4ab7-81e4-c73eb4a88f97",
   "metadata": {},
   "source": [
    "## <span style=\"color:black\">Add alternate name TS_H1.6 of cH1.6</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "5a2d93ce-a4ce-4260-b3ca-3d04bdc4ddb6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>H1t</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH1.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id name taxonomy  gene splice histone\n",
       "0   8  H1t     None  None   None   cH1.6"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name \" \"WHERE histone='cH1.6'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "4dcef455-5ef2-4111-b18a-68ae774f3ae4",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_alternate_name = {\n",
    "    \"name\": \"TS_H1.6\",\n",
    "    \"taxonomy\": None,\n",
    "    \"gene\": None,\n",
    "    \"splice\": None,\n",
    "    \"histone\": \"cH1.6\",\n",
    "}\n",
    "cursor.execute(add_alternate_names, data_alternate_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "2669a724-0fa4-45e9-90ac-05a6c12330dd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice</th>\n",
       "      <th>histone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>H1t</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH1.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>123</td>\n",
       "      <td>TS_H1.6</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>cH1.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    id     name taxonomy  gene splice histone\n",
       "0    8      H1t     None  None   None   cH1.6\n",
       "1  123  TS_H1.6     None  None   None   cH1.6"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * FROM alternative_name \" \"WHERE histone='cH1.6'\"\n",
    "cursor.execute(query)\n",
    "pd.DataFrame(cursor.fetchall(), columns=[i[0] for i in cursor.description])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "52298069-81ac-46d6-bb45-d9a94808ebd0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make sure data is committed to the database\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e8bad07-ef74-40fd-a93b-e7897df0c591",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Close connections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "e5673501-5845-491a-bbb9-f5d213721d58",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cursor.close()\n",
    "conn.close()\n",
    "tunnel.stop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cef5e157-5dae-466b-ba4a-613df32ff889",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".conda-histdb_env",
   "language": "python",
   "name": "conda-env-.conda-histdb_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
