{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%cd '..'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import inspect\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import Entrez, SeqIO\n",
    "from curated_set_services import CuratedSet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((571, 16),\n",
       " Index(['accession', 'type', 'variant_group', 'variant', 'doublet', 'gi',\n",
       "        'ncbi_gene_id', 'hgnc_gene_name', 'taxonomy_id', 'organism', 'phylum',\n",
       "        'class', 'taxonomy_group', 'info', 'references', 'sequence'],\n",
       "       dtype='object'))"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load data from histones.csv\n",
    "curated_set = CuratedSet()\n",
    "cs = curated_set\n",
    "\n",
    "cs.data.shape, cs.data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(cs.has_duplicates())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NP_835502.1</th>\n",
       "      <td>NP_835502.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>24506885</td>\n",
       "      <td>MPEPAKSAPAPKKGSKKAVTKAQKKDGKKRKRSRKESYSVYVYKVL...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               accession type          variant_group                variant  \\\n",
       "NP_835502.1  NP_835502.1  H2B  cH2B.1_(Mus_musculus)  cH2B.1_(Mus_musculus)   \n",
       "\n",
       "            doublet gi ncbi_gene_id hgnc_gene_name taxonomy_id      organism  \\\n",
       "NP_835502.1                                              10090  Mus musculus   \n",
       "\n",
       "               phylum     class taxonomy_group info references  \\\n",
       "NP_835502.1  Chordata  Mammalia                       24506885   \n",
       "\n",
       "                                                      sequence  \n",
       "NP_835502.1  MPEPAKSAPAPKKGSKKAVTKAQKKDGKKRKRSRKESYSVYVYKVL...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data[cs.data[\"accession\"] == \"NP_835502.1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NP_505463.1</th>\n",
       "      <td>NP_505463.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A_(Animals)</td>\n",
       "      <td></td>\n",
       "      <td>17562014</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>6239</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>Nematoda</td>\n",
       "      <td>Chromadorea</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26989147 22650316(?)</td>\n",
       "      <td>MSGRGKGGKAKTGGKAKSRSSRAGLQFPVGRLHRILRKGNYAQRVG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EEC09557.1</th>\n",
       "      <td>EEC09557.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A_(Animals)</td>\n",
       "      <td></td>\n",
       "      <td>215500063</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>6945</td>\n",
       "      <td>Ixodes scapularis</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Arachnida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26989147 22650316(?)</td>\n",
       "      <td>MSGRGKGGKVKGKSKTRSSRAGLQFPVGRIHRLLRKGNYAERVGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_724343.1</th>\n",
       "      <td>NP_724343.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A_(Animals)</td>\n",
       "      <td></td>\n",
       "      <td>24585673</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26989147 22650316(?)</td>\n",
       "      <td>MSGRGKGGKVKGKAKSRSNRAGLQFPVGRIHRLLRKGNYAERVGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_001119899.1</th>\n",
       "      <td>XP_001119899.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A_(Animals)</td>\n",
       "      <td></td>\n",
       "      <td>110764935</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7460</td>\n",
       "      <td>Apis mellifera</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26989147 22650316(?)</td>\n",
       "      <td>MSGRGKGGKAKAKAKSRSNRAGLQFPVGRIHRLLRKGNYAERVGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EDO48405.1</th>\n",
       "      <td>EDO48405.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A_(Animals)</td>\n",
       "      <td></td>\n",
       "      <td>156227602</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>45351</td>\n",
       "      <td>Nematostella vectensis</td>\n",
       "      <td>Cnidaria</td>\n",
       "      <td>Anthozoa</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26989147 22650316(?)</td>\n",
       "      <td>MSGRGKGKAKGTKSKTRSSRAGLQFPVGRIHRHLRKGNYAERVGAG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     accession type variant_group         variant doublet  \\\n",
       "NP_505463.1        NP_505463.1  H2A          cH2A  cH2A_(Animals)           \n",
       "EEC09557.1          EEC09557.1  H2A          cH2A  cH2A_(Animals)           \n",
       "NP_724343.1        NP_724343.1  H2A          cH2A  cH2A_(Animals)           \n",
       "XP_001119899.1  XP_001119899.1  H2A          cH2A  cH2A_(Animals)           \n",
       "EDO48405.1          EDO48405.1  H2A          cH2A  cH2A_(Animals)           \n",
       "\n",
       "                       gi ncbi_gene_id hgnc_gene_name taxonomy_id  \\\n",
       "NP_505463.1      17562014                                    6239   \n",
       "EEC09557.1      215500063                                    6945   \n",
       "NP_724343.1      24585673                                    7227   \n",
       "XP_001119899.1  110764935                                    7460   \n",
       "EDO48405.1      156227602                                   45351   \n",
       "\n",
       "                               organism      phylum        class  \\\n",
       "NP_505463.1      Caenorhabditis elegans    Nematoda  Chromadorea   \n",
       "EEC09557.1            Ixodes scapularis  Arthropoda    Arachnida   \n",
       "NP_724343.1     Drosophila melanogaster  Arthropoda      Insecta   \n",
       "XP_001119899.1           Apis mellifera  Arthropoda      Insecta   \n",
       "EDO48405.1       Nematostella vectensis    Cnidaria     Anthozoa   \n",
       "\n",
       "               taxonomy_group info            references  \\\n",
       "NP_505463.1                         26989147 22650316(?)   \n",
       "EEC09557.1                          26989147 22650316(?)   \n",
       "NP_724343.1                         26989147 22650316(?)   \n",
       "XP_001119899.1                      26989147 22650316(?)   \n",
       "EDO48405.1                          26989147 22650316(?)   \n",
       "\n",
       "                                                         sequence  \n",
       "NP_505463.1     MSGRGKGGKAKTGGKAKSRSSRAGLQFPVGRLHRILRKGNYAQRVG...  \n",
       "EEC09557.1      MSGRGKGGKVKGKSKTRSSRAGLQFPVGRIHRLLRKGNYAERVGAG...  \n",
       "NP_724343.1     MSGRGKGGKVKGKAKSRSNRAGLQFPVGRIHRLLRKGNYAERVGAG...  \n",
       "XP_001119899.1  MSGRGKGGKAKAKAKSRSNRAGLQFPVGRIHRLLRKGNYAERVGAG...  \n",
       "EDO48405.1      MSGRGKGKAKGTKSKTRSSRAGLQFPVGRIHRHLRKGNYAERVGAG...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(570, 16)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cH2B1_accessions = [\n",
    "    \"NP_835502.1\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((1, 5),\n",
       " Index(['accession', 'type', 'variant_group', 'variant', 'references'], dtype='object'))"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(\n",
    "    {\n",
    "        \"accession\": cH2B1_accessions,\n",
    "        \"type\": [\"H2B\"] * len(cH2B1_accessions),\n",
    "        \"variant_group\": [\"cH2B.1_(Mus_musculus)\"] * len(cH2B1_accessions),\n",
    "        \"variant\": [\"cH2B.1_(Mus_musculus)\"] * len(cH2B1_accessions),\n",
    "        \"references\": [\"24506885\"] * len(cH2B1_accessions),\n",
    "    }\n",
    ")\n",
    "df.index = df.accession\n",
    "df.shape, df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>references</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NP_835502.1</th>\n",
       "      <td>NP_835502.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td>24506885</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               accession type          variant_group                variant  \\\n",
       "accession                                                                     \n",
       "NP_835502.1  NP_835502.1  H2B  cH2B.1_(Mus_musculus)  cH2B.1_(Mus_musculus)   \n",
       "\n",
       "            references  \n",
       "accession               \n",
       "NP_835502.1   24506885  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((571, 16), [])"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data = pd.concat([cs.data, df]).fillna(\"\")\n",
    "cs.data.shape, list(cs.has_duplicates())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NP_005311.1</th>\n",
       "      <td>NP_005311.1</td>\n",
       "      <td>H1</td>\n",
       "      <td>H1.3</td>\n",
       "      <td>H1.3_(Homo_sapiens)__???</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3007.0</td>\n",
       "      <td>H1-3</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "      <td>MSETAPLAPTIPAPAEKTPVKKKAKKAGATAGKRKASGPPVSELIT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_005316.1</th>\n",
       "      <td>NP_005316.1</td>\n",
       "      <td>H1</td>\n",
       "      <td>H1.1</td>\n",
       "      <td>H1.1_(Homo_sapiens)__???</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3024.0</td>\n",
       "      <td>H1-1</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "      <td>MSETVPPAPAASAAPEKPLAGKKAKKPAKAAAASKKKPAGPSVSEL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_005314.2</th>\n",
       "      <td>NP_005314.2</td>\n",
       "      <td>H1</td>\n",
       "      <td>H1.6</td>\n",
       "      <td>H1.6_(Homo_sapiens)__???</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3010.0</td>\n",
       "      <td>H1-6</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "      <td>MSETVPAASASAGVAAMEKLPTKKRGRKPAGLISASRKVPNLSVSK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_005310.1</th>\n",
       "      <td>NP_005310.1</td>\n",
       "      <td>H1</td>\n",
       "      <td>H1.2</td>\n",
       "      <td>H1.2_(Homo_sapiens)__???</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3006.0</td>\n",
       "      <td>H1-2</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td>26689747</td>\n",
       "      <td>MSETAPAAPAAAPPAEKAPVKKKAAKKAGGTPRKASGPPVSELITK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_835502.1</th>\n",
       "      <td>NP_835502.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>24506885</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               accession type          variant_group  \\\n",
       "NP_005311.1  NP_005311.1   H1                   H1.3   \n",
       "NP_005316.1  NP_005316.1   H1                   H1.1   \n",
       "NP_005314.2  NP_005314.2   H1                   H1.6   \n",
       "NP_005310.1  NP_005310.1   H1                   H1.2   \n",
       "NP_835502.1  NP_835502.1  H2B  cH2B.1_(Mus_musculus)   \n",
       "\n",
       "                              variant doublet gi ncbi_gene_id hgnc_gene_name  \\\n",
       "NP_005311.1  H1.3_(Homo_sapiens)__???                  3007.0           H1-3   \n",
       "NP_005316.1  H1.1_(Homo_sapiens)__???                  3024.0           H1-1   \n",
       "NP_005314.2  H1.6_(Homo_sapiens)__???                  3010.0           H1-6   \n",
       "NP_005310.1  H1.2_(Homo_sapiens)__???                  3006.0           H1-2   \n",
       "NP_835502.1     cH2B.1_(Mus_musculus)                                          \n",
       "\n",
       "            taxonomy_id      organism    phylum     class taxonomy_group info  \\\n",
       "NP_005311.1        9606  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "NP_005316.1        9606  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "NP_005314.2        9606  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "NP_005310.1        9606  Homo sapiens  Chordata  Mammalia       Mammalia        \n",
       "NP_835502.1                                                                     \n",
       "\n",
       "            references                                           sequence  \n",
       "NP_005311.1   26689747  MSETAPLAPTIPAPAEKTPVKKKAKKAGATAGKRKASGPPVSELIT...  \n",
       "NP_005316.1   26689747  MSETVPPAPAASAAPEKPLAGKKAKKPAKAAAASKKKPAGPSVSEL...  \n",
       "NP_005314.2   26689747  MSETVPAASASAGVAAMEKLPTKKRGRKPAGLISASRKVPNLSVSK...  \n",
       "NP_005310.1   26689747  MSETAPAAPAAAPPAEKAPVKKKAAKKAGGTPRKASGPPVSELITK...  \n",
       "NP_835502.1   24506885                                                     "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2, 16)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data[cs.data[\"variant\"] == \"cH2B.1_(Mus_musculus)\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/l_singh/.conda/envs/histdb_env/lib/python3.8/site-packages/Bio/Entrez/__init__.py:658: UserWarning: \n",
      "Email address is not specified.\n",
      "\n",
      "To make use of NCBI's E-utilities, NCBI requires you to specify your\n",
      "email address with each request.  As an example, if your email address\n",
      "is A.N.Other@example.com, you can specify it as follows:\n",
      "   from Bio import Entrez\n",
      "   Entrez.email = 'A.N.Other@example.com'\n",
      "In case of excessive usage of the E-utilities, NCBI will attempt to contact\n",
      "a user at the email address provided before blocking access to the\n",
      "E-utilities.\n",
      "  warnings.warn(\n",
      "/home/l_singh/.conda/envs/histdb_env/lib/python3.8/site-packages/Bio/GenBank/__init__.py:1143: BiopythonParserWarning: Dropping bond qualifier in feature location\n",
      "  warnings.warn(\n",
      "/mnt/scratch/v_kulikov/histonedb_Vladimir/CURATED_SET/curated_set_services.py:283: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  updating_data['accession'] = new_accessions\n"
     ]
    }
   ],
   "source": [
    "cs.update_accession_version()\n",
    "cs.data = cs.data.set_index(cs.data.accession.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2, 16)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data[cs.data[\"variant\"] == \"cH2B.1_(Mus_musculus)\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 10090\n",
      " changes to 10090\n",
      " changes to Mus musculus\n",
      " changes to Chordata\n",
      " changes to Mammalia\n"
     ]
    }
   ],
   "source": [
    "curated_set.update_taxids(blank_data=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2, 16)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data[cs.data[\"variant\"] == \"cH2B.1_(Mus_musculus)\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NP_835502.1</th>\n",
       "      <td>NP_835502.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>24506885</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAA62299.1</th>\n",
       "      <td>CAA62299.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>8672246</td>\n",
       "      <td>MPEVAVKGATISKKGFKKAVTKTQKKEGRKRKRCRKESYSIYIYKV...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               accession type          variant_group                variant  \\\n",
       "NP_835502.1  NP_835502.1  H2B  cH2B.1_(Mus_musculus)  cH2B.1_(Mus_musculus)   \n",
       "CAA62299.1    CAA62299.1  H2B  cH2B.1_(Mus_musculus)  cH2B.1_(Mus_musculus)   \n",
       "\n",
       "            doublet gi ncbi_gene_id hgnc_gene_name taxonomy_id      organism  \\\n",
       "NP_835502.1                                              10090  Mus musculus   \n",
       "CAA62299.1                                               10090  Mus musculus   \n",
       "\n",
       "               phylum     class taxonomy_group info references  \\\n",
       "NP_835502.1  Chordata  Mammalia                       24506885   \n",
       "CAA62299.1   Chordata  Mammalia                        8672246   \n",
       "\n",
       "                                                      sequence  \n",
       "NP_835502.1                                                     \n",
       "CAA62299.1   MPEVAVKGATISKKGFKKAVTKTQKKEGRKRKRCRKESYSIYIYKV...  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data[cs.data[\"variant\"] == \"cH2B.1_(Mus_musculus)\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Updating sequences for late_H2B_(Echinoidea)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NP_835502.1</th>\n",
       "      <td>NP_835502.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>24506885</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               accession type          variant_group                variant  \\\n",
       "NP_835502.1  NP_835502.1  H2B  cH2B.1_(Mus_musculus)  cH2B.1_(Mus_musculus)   \n",
       "\n",
       "            doublet gi ncbi_gene_id hgnc_gene_name taxonomy_id      organism  \\\n",
       "NP_835502.1                                              10090  Mus musculus   \n",
       "\n",
       "               phylum     class taxonomy_group info references sequence  \n",
       "NP_835502.1  Chordata  Mammalia                       24506885           "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data[cs.data['sequence'] == '']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading FASTA SeqRecords by ACCESSIONs from NCBI\n",
      "Fetching 1 seqs\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/l_singh/.conda/envs/histdb_env/lib/python3.8/site-packages/Bio/Entrez/__init__.py:658: UserWarning: \n",
      "Email address is not specified.\n",
      "\n",
      "To make use of NCBI's E-utilities, NCBI requires you to specify your\n",
      "email address with each request.  As an example, if your email address\n",
      "is A.N.Other@example.com, you can specify it as follows:\n",
      "   from Bio import Entrez\n",
      "   Entrez.email = 'A.N.Other@example.com'\n",
      "In case of excessive usage of the E-utilities, NCBI will attempt to contact\n",
      "a user at the email address provided before blocking access to the\n",
      "E-utilities.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence for NP_835502.1 changes from  to MPEPAKSAPAPKKGSKKAVTKAQKKDGKKRKRSRKESYSVYVYKVLKQVHPDTGISSKAMGIMNSFVNDIFERIASEASRLAHYNKRSTITSREIQTAVRLLLPGELAKHAVSEGTKAVTKYTSSK\n",
      "Sequences updated: 1\n"
     ]
    }
   ],
   "source": [
    "cs.update_sequence(blank_data=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NP_835502.1</th>\n",
       "      <td>NP_835502.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>24506885</td>\n",
       "      <td>MPEPAKSAPAPKKGSKKAVTKAQKKDGKKRKRSRKESYSVYVYKVL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAA62299.1</th>\n",
       "      <td>CAA62299.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td>cH2B.1_(Mus_musculus)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>8672246</td>\n",
       "      <td>MPEVAVKGATISKKGFKKAVTKTQKKEGRKRKRCRKESYSIYIYKV...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               accession type          variant_group                variant  \\\n",
       "NP_835502.1  NP_835502.1  H2B  cH2B.1_(Mus_musculus)  cH2B.1_(Mus_musculus)   \n",
       "CAA62299.1    CAA62299.1  H2B  cH2B.1_(Mus_musculus)  cH2B.1_(Mus_musculus)   \n",
       "\n",
       "            doublet gi ncbi_gene_id hgnc_gene_name taxonomy_id      organism  \\\n",
       "NP_835502.1                                              10090  Mus musculus   \n",
       "CAA62299.1                                               10090  Mus musculus   \n",
       "\n",
       "               phylum     class taxonomy_group info references  \\\n",
       "NP_835502.1  Chordata  Mammalia                       24506885   \n",
       "CAA62299.1   Chordata  Mammalia                        8672246   \n",
       "\n",
       "                                                      sequence  \n",
       "NP_835502.1  MPEPAKSAPAPKKGSKKAVTKAQKKDGKKRKRSRKESYSVYVYKVL...  \n",
       "CAA62299.1   MPEVAVKGATISKKGFKKAVTKTQKKEGRKRKRCRKESYSIYIYKV...  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data[cs.data[\"variant\"] == \"cH2B.1_(Mus_musculus)\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "New_df = pd.DataFrame(\n",
    "    {\n",
    "        \"accession\": [\"HISTDB_cH2B_1_0\"],\n",
    "        \"type\": [\"H2B\"],\n",
    "        \"variant_group\": [\"cH2B\"],\n",
    "        \"variant\": [\"cH2B_(Mammalia)\"],\n",
    "        \"taxonomy_id\": [10116],\n",
    "        \"organism\": [\"Rattus norvegicus\"],\n",
    "        \"phylum\": [\"Chordata\"],\n",
    "        \"class\": [\"Mammalia\"],\n",
    "        \"references\": [12213818],\n",
    "        \"sequence\": [\"MPEVSAKGTTISKKGFKKAVTKTQKKEGRKRKRCREESYSIYIYKVLKQVHPDTGISSKAMSIMNSFVTDIFERIAGEASRLAHYNKRSTITSREIQTAVRLLLPGELAKHAVSEGTKAVTKYTSSK\"],\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "cs.data = pd.concat([cs.data, New_df], ignore_index=True).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Added sequence with accession 0\n",
      "Added sequence with accession 1\n",
      "Added sequence with accession 2\n",
      "Added sequence with accession 3\n",
      "Added sequence with accession 4\n",
      "Added sequence with accession 5\n",
      "Added sequence with accession 6\n",
      "Added sequence with accession 7\n",
      "Added sequence with accession 8\n",
      "Added sequence with accession 9\n",
      "Added sequence with accession 10\n",
      "Added sequence with accession 11\n",
      "Added sequence with accession 12\n",
      "Added sequence with accession 13\n",
      "Added sequence with accession 14\n",
      "Added sequence with accession 15\n",
      "Added sequence with accession 16\n",
      "Added sequence with accession 17\n",
      "Added sequence with accession 18\n",
      "Added sequence with accession 19\n",
      "Added sequence with accession 20\n",
      "Added sequence with accession 21\n",
      "Added sequence with accession 22\n",
      "Added sequence with accession 23\n",
      "Added sequence with accession 24\n",
      "Added sequence with accession 25\n",
      "Added sequence with accession 26\n",
      "Added sequence with accession 27\n",
      "Added sequence with accession 28\n",
      "Added sequence with accession 29\n",
      "Added sequence with accession 30\n",
      "Added sequence with accession 31\n",
      "Added sequence with accession 32\n",
      "Added sequence with accession 33\n",
      "Added sequence with accession 34\n",
      "Added sequence with accession 35\n",
      "Added sequence with accession 36\n",
      "Added sequence with accession 37\n",
      "Added sequence with accession 38\n",
      "Added sequence with accession 39\n",
      "Added sequence with accession 40\n",
      "Added sequence with accession 41\n",
      "Added sequence with accession 42\n",
      "Added sequence with accession 43\n",
      "Added sequence with accession 44\n",
      "Added sequence with accession 45\n",
      "Added sequence with accession 46\n",
      "Added sequence with accession 47\n",
      "Added sequence with accession 48\n",
      "Added sequence with accession 49\n",
      "Added sequence with accession 50\n",
      "Added sequence with accession 51\n",
      "Added sequence with accession 52\n",
      "Added sequence with accession 53\n",
      "Added sequence with accession 54\n",
      "Added sequence with accession 55\n",
      "Added sequence with accession 56\n",
      "Added sequence with accession 57\n",
      "Added sequence with accession 58\n",
      "Added sequence with accession 59\n",
      "Added sequence with accession 60\n",
      "Added sequence with accession 61\n",
      "Added sequence with accession 62\n",
      "Added sequence with accession 63\n",
      "Added sequence with accession 64\n",
      "Added sequence with accession 65\n",
      "Added sequence with accession 66\n",
      "Added sequence with accession 67\n",
      "Added sequence with accession 68\n",
      "Added sequence with accession 69\n",
      "Added sequence with accession 70\n",
      "Added sequence with accession 71\n",
      "Added sequence with accession 72\n",
      "Added sequence with accession 73\n",
      "Added sequence with accession 74\n",
      "Added sequence with accession 75\n",
      "Added sequence with accession 76\n",
      "Added sequence with accession 77\n",
      "Added sequence with accession 78\n",
      "Added sequence with accession 79\n",
      "Added sequence with accession 80\n",
      "Added sequence with accession 81\n",
      "Added sequence with accession 82\n",
      "Added sequence with accession 83\n",
      "Added sequence with accession 84\n",
      "Added sequence with accession 85\n",
      "Added sequence with accession 86\n",
      "Added sequence with accession 87\n",
      "Added sequence with accession 88\n",
      "Added sequence with accession 89\n",
      "Added sequence with accession 90\n",
      "Added sequence with accession 91\n",
      "Added sequence with accession 92\n",
      "Added sequence with accession 93\n",
      "Added sequence with accession 94\n",
      "Added sequence with accession 95\n",
      "Added sequence with accession 96\n",
      "Added sequence with accession 97\n",
      "Added sequence with accession 98\n",
      "Added sequence with accession 99\n",
      "Added sequence with accession 100\n",
      "Added sequence with accession 101\n",
      "Added sequence with accession 102\n",
      "Added sequence with accession 103\n",
      "Added sequence with accession 104\n",
      "Added sequence with accession 105\n",
      "Added sequence with accession 106\n",
      "Added sequence with accession 107\n",
      "Added sequence with accession 108\n",
      "Added sequence with accession 109\n",
      "Added sequence with accession 110\n",
      "Added sequence with accession 111\n",
      "Added sequence with accession 112\n",
      "Added sequence with accession 113\n",
      "Added sequence with accession 114\n",
      "Added sequence with accession 115\n",
      "Added sequence with accession 116\n",
      "Added sequence with accession 117\n",
      "Added sequence with accession 118\n",
      "Added sequence with accession 119\n",
      "Added sequence with accession 120\n",
      "Added sequence with accession 121\n",
      "Added sequence with accession 122\n",
      "Added sequence with accession 123\n",
      "Added sequence with accession 124\n",
      "Added sequence with accession 125\n",
      "Added sequence with accession 126\n",
      "Added sequence with accession 127\n",
      "Added sequence with accession 128\n",
      "Added sequence with accession 129\n",
      "Added sequence with accession 130\n",
      "Added sequence with accession 131\n",
      "Added sequence with accession 132\n",
      "Added sequence with accession 133\n",
      "Added sequence with accession 134\n",
      "Added sequence with accession 135\n",
      "Added sequence with accession 136\n",
      "Added sequence with accession 137\n",
      "Added sequence with accession 138\n",
      "Added sequence with accession 139\n",
      "Added sequence with accession 140\n",
      "Added sequence with accession 141\n",
      "Added sequence with accession 142\n",
      "Added sequence with accession 143\n",
      "Added sequence with accession 144\n",
      "Added sequence with accession 145\n",
      "Added sequence with accession 146\n",
      "Added sequence with accession 147\n",
      "Added sequence with accession 148\n",
      "Added sequence with accession 149\n",
      "Added sequence with accession 150\n",
      "Added sequence with accession 151\n",
      "Added sequence with accession 152\n",
      "Added sequence with accession 153\n",
      "Added sequence with accession 154\n",
      "Added sequence with accession 155\n",
      "Added sequence with accession 156\n",
      "Added sequence with accession 157\n",
      "Added sequence with accession 158\n",
      "Added sequence with accession 159\n",
      "Added sequence with accession 160\n",
      "Added sequence with accession 161\n",
      "Added sequence with accession 162\n",
      "Added sequence with accession 163\n",
      "Added sequence with accession 164\n",
      "Added sequence with accession 165\n",
      "Added sequence with accession 166\n",
      "Added sequence with accession 167\n",
      "Added sequence with accession 168\n",
      "Added sequence with accession 169\n",
      "Added sequence with accession 170\n",
      "Added sequence with accession 171\n",
      "Added sequence with accession 172\n",
      "Added sequence with accession 173\n",
      "Added sequence with accession 174\n",
      "Added sequence with accession 175\n",
      "Added sequence with accession 176\n",
      "Added sequence with accession 177\n",
      "Added sequence with accession 178\n",
      "Added sequence with accession 179\n",
      "Added sequence with accession 180\n",
      "Added sequence with accession 181\n",
      "Added sequence with accession 182\n",
      "Added sequence with accession 183\n",
      "Added sequence with accession 184\n",
      "Added sequence with accession 185\n",
      "Added sequence with accession 186\n",
      "Added sequence with accession 187\n",
      "Added sequence with accession 188\n",
      "Added sequence with accession 189\n",
      "Added sequence with accession 190\n",
      "Added sequence with accession 191\n",
      "Added sequence with accession 192\n",
      "Added sequence with accession 193\n",
      "Added sequence with accession 194\n",
      "Added sequence with accession 195\n",
      "Added sequence with accession 196\n",
      "Added sequence with accession 197\n",
      "Added sequence with accession 198\n",
      "Added sequence with accession 199\n",
      "Added sequence with accession 200\n",
      "Added sequence with accession 201\n",
      "Added sequence with accession 202\n",
      "Added sequence with accession 203\n",
      "Added sequence with accession 204\n",
      "Added sequence with accession 205\n",
      "Added sequence with accession 206\n",
      "Added sequence with accession 207\n",
      "Added sequence with accession 208\n",
      "Added sequence with accession 209\n",
      "Added sequence with accession 210\n",
      "Added sequence with accession 211\n",
      "Added sequence with accession 212\n",
      "Added sequence with accession 213\n",
      "Added sequence with accession 214\n",
      "Added sequence with accession 215\n",
      "Added sequence with accession 216\n",
      "Added sequence with accession 217\n",
      "Added sequence with accession 218\n",
      "Added sequence with accession 219\n",
      "Added sequence with accession 220\n",
      "Added sequence with accession 221\n",
      "Added sequence with accession 222\n",
      "Added sequence with accession 223\n",
      "Added sequence with accession 224\n",
      "Added sequence with accession 225\n",
      "Added sequence with accession 226\n",
      "Added sequence with accession 227\n",
      "Added sequence with accession 228\n",
      "Added sequence with accession 229\n",
      "Added sequence with accession 230\n",
      "Added sequence with accession 231\n",
      "Added sequence with accession 232\n",
      "Added sequence with accession 233\n",
      "Added sequence with accession 234\n",
      "Added sequence with accession 235\n",
      "Added sequence with accession 236\n",
      "Added sequence with accession 237\n",
      "Added sequence with accession 238\n",
      "Added sequence with accession 239\n",
      "Added sequence with accession 240\n",
      "Added sequence with accession 241\n",
      "Added sequence with accession 242\n",
      "Added sequence with accession 243\n",
      "Added sequence with accession 244\n",
      "Added sequence with accession 245\n",
      "Added sequence with accession 246\n",
      "Added sequence with accession 247\n",
      "Added sequence with accession 248\n",
      "Added sequence with accession 249\n",
      "Added sequence with accession 250\n",
      "Added sequence with accession 251\n",
      "Added sequence with accession 252\n",
      "Added sequence with accession 253\n",
      "Added sequence with accession 254\n",
      "Added sequence with accession 255\n",
      "Added sequence with accession 256\n",
      "Added sequence with accession 257\n",
      "Added sequence with accession 258\n",
      "Added sequence with accession 259\n",
      "Added sequence with accession 260\n",
      "Added sequence with accession 261\n",
      "Added sequence with accession 262\n",
      "Added sequence with accession 263\n",
      "Added sequence with accession 264\n",
      "Added sequence with accession 265\n",
      "Added sequence with accession 266\n",
      "Added sequence with accession 267\n",
      "Added sequence with accession 268\n",
      "Added sequence with accession 269\n",
      "Added sequence with accession 270\n",
      "Added sequence with accession 271\n",
      "Added sequence with accession 272\n",
      "Added sequence with accession 273\n",
      "Added sequence with accession 274\n",
      "Added sequence with accession 275\n",
      "Added sequence with accession 276\n",
      "Added sequence with accession 277\n",
      "Added sequence with accession 278\n",
      "Added sequence with accession 279\n",
      "Added sequence with accession 280\n",
      "Added sequence with accession 281\n",
      "Added sequence with accession 282\n",
      "Added sequence with accession 283\n",
      "Added sequence with accession 284\n",
      "Added sequence with accession 285\n",
      "Added sequence with accession 286\n",
      "Added sequence with accession 287\n",
      "Added sequence with accession 288\n",
      "Added sequence with accession 289\n",
      "Added sequence with accession 290\n",
      "Added sequence with accession 291\n",
      "Added sequence with accession 292\n",
      "Added sequence with accession 293\n",
      "Added sequence with accession 294\n",
      "Added sequence with accession 295\n",
      "Added sequence with accession 296\n",
      "Added sequence with accession 297\n",
      "Added sequence with accession 298\n",
      "Added sequence with accession 299\n",
      "Added sequence with accession 300\n",
      "Added sequence with accession 301\n",
      "Added sequence with accession 302\n",
      "Added sequence with accession 303\n",
      "Added sequence with accession 304\n",
      "Added sequence with accession 305\n",
      "Added sequence with accession 306\n",
      "Added sequence with accession 307\n",
      "Added sequence with accession 308\n",
      "Added sequence with accession 309\n",
      "Added sequence with accession 310\n",
      "Added sequence with accession 311\n",
      "Added sequence with accession 312\n",
      "Added sequence with accession 313\n",
      "Added sequence with accession 314\n",
      "Added sequence with accession 315\n",
      "Added sequence with accession 316\n",
      "Added sequence with accession 317\n",
      "Added sequence with accession 318\n",
      "Added sequence with accession 319\n",
      "Added sequence with accession 320\n",
      "Added sequence with accession 321\n",
      "Added sequence with accession 322\n",
      "Added sequence with accession 323\n",
      "Added sequence with accession 324\n",
      "Added sequence with accession 325\n",
      "Added sequence with accession 326\n",
      "Added sequence with accession 327\n",
      "Added sequence with accession 328\n",
      "Added sequence with accession 329\n",
      "Added sequence with accession 330\n",
      "Added sequence with accession 331\n",
      "Added sequence with accession 332\n",
      "Added sequence with accession 333\n",
      "Added sequence with accession 334\n",
      "Added sequence with accession 335\n",
      "Added sequence with accession 336\n",
      "Added sequence with accession 337\n",
      "Added sequence with accession 338\n",
      "Added sequence with accession 339\n",
      "Added sequence with accession 340\n",
      "Added sequence with accession 341\n",
      "Added sequence with accession 342\n",
      "Added sequence with accession 343\n",
      "Added sequence with accession 344\n",
      "Added sequence with accession 345\n",
      "Added sequence with accession 346\n",
      "Added sequence with accession 347\n",
      "Added sequence with accession 348\n",
      "Added sequence with accession 349\n",
      "Added sequence with accession 350\n",
      "Added sequence with accession 351\n",
      "Added sequence with accession 352\n",
      "Added sequence with accession 353\n",
      "Added sequence with accession 354\n",
      "Added sequence with accession 355\n",
      "Added sequence with accession 356\n",
      "Added sequence with accession 357\n",
      "Added sequence with accession 358\n",
      "Added sequence with accession 359\n",
      "Added sequence with accession 360\n",
      "Added sequence with accession 361\n",
      "Added sequence with accession 362\n",
      "Added sequence with accession 363\n",
      "Added sequence with accession 364\n",
      "Added sequence with accession 365\n",
      "Added sequence with accession 366\n",
      "Added sequence with accession 367\n",
      "Added sequence with accession 368\n",
      "Added sequence with accession 369\n",
      "Added sequence with accession 370\n",
      "Added sequence with accession 371\n",
      "Added sequence with accession 372\n",
      "Added sequence with accession 373\n",
      "Added sequence with accession 374\n",
      "Added sequence with accession 375\n",
      "Added sequence with accession 376\n",
      "Added sequence with accession 377\n",
      "Added sequence with accession 378\n",
      "Added sequence with accession 379\n",
      "Added sequence with accession 380\n",
      "Added sequence with accession 381\n",
      "Added sequence with accession 382\n",
      "Added sequence with accession 383\n",
      "Added sequence with accession 384\n",
      "Added sequence with accession 385\n",
      "Added sequence with accession 386\n",
      "Added sequence with accession 387\n",
      "Added sequence with accession 388\n",
      "Added sequence with accession 389\n",
      "Added sequence with accession 390\n",
      "Added sequence with accession 391\n",
      "Added sequence with accession 392\n",
      "Added sequence with accession 393\n",
      "Added sequence with accession 394\n",
      "Added sequence with accession 395\n",
      "Added sequence with accession 396\n",
      "Added sequence with accession 397\n",
      "Added sequence with accession 398\n",
      "Added sequence with accession 399\n",
      "Added sequence with accession 400\n",
      "Added sequence with accession 401\n",
      "Added sequence with accession 402\n",
      "Added sequence with accession 403\n",
      "Added sequence with accession 404\n",
      "Added sequence with accession 405\n",
      "Added sequence with accession 406\n",
      "Added sequence with accession 407\n",
      "Added sequence with accession 408\n",
      "Added sequence with accession 409\n",
      "Added sequence with accession 410\n",
      "Added sequence with accession 411\n",
      "Added sequence with accession 412\n",
      "Added sequence with accession 413\n",
      "Added sequence with accession 414\n",
      "Added sequence with accession 415\n",
      "Added sequence with accession 416\n",
      "Added sequence with accession 417\n",
      "Added sequence with accession 418\n",
      "Added sequence with accession 419\n",
      "Added sequence with accession 420\n",
      "Added sequence with accession 421\n",
      "Added sequence with accession 422\n",
      "Added sequence with accession 423\n",
      "Added sequence with accession 424\n",
      "Added sequence with accession 425\n",
      "Added sequence with accession 426\n",
      "Added sequence with accession 427\n",
      "Added sequence with accession 428\n",
      "Added sequence with accession 429\n",
      "Added sequence with accession 430\n",
      "Added sequence with accession 431\n",
      "Added sequence with accession 432\n",
      "Added sequence with accession 433\n",
      "Added sequence with accession 434\n",
      "Added sequence with accession 435\n",
      "Added sequence with accession 436\n",
      "Added sequence with accession 437\n",
      "Added sequence with accession 438\n",
      "Added sequence with accession 439\n",
      "Added sequence with accession 440\n",
      "Added sequence with accession 441\n",
      "Added sequence with accession 442\n",
      "Added sequence with accession 443\n",
      "Added sequence with accession 444\n",
      "Added sequence with accession 445\n",
      "Added sequence with accession 446\n",
      "Added sequence with accession 447\n",
      "Added sequence with accession 448\n",
      "Added sequence with accession 449\n",
      "Added sequence with accession 450\n",
      "Added sequence with accession 451\n",
      "Added sequence with accession 452\n",
      "Added sequence with accession 453\n",
      "Added sequence with accession 454\n",
      "Added sequence with accession 455\n",
      "Added sequence with accession 456\n",
      "Added sequence with accession 457\n",
      "Added sequence with accession 458\n",
      "Added sequence with accession 459\n",
      "Added sequence with accession 460\n",
      "Added sequence with accession 461\n",
      "Added sequence with accession 462\n",
      "Added sequence with accession 463\n",
      "Added sequence with accession 464\n",
      "Added sequence with accession 465\n",
      "Added sequence with accession 466\n",
      "Added sequence with accession 467\n",
      "Added sequence with accession 468\n",
      "Added sequence with accession 469\n",
      "Added sequence with accession 470\n",
      "Added sequence with accession 471\n",
      "Added sequence with accession 472\n",
      "Added sequence with accession 473\n",
      "Added sequence with accession 474\n",
      "Added sequence with accession 475\n",
      "Added sequence with accession 476\n",
      "Added sequence with accession 477\n",
      "Added sequence with accession 478\n",
      "Added sequence with accession 479\n",
      "Added sequence with accession 480\n",
      "Added sequence with accession 481\n",
      "Added sequence with accession 482\n",
      "Added sequence with accession 483\n",
      "Added sequence with accession 484\n",
      "Added sequence with accession 485\n",
      "Added sequence with accession 486\n",
      "Added sequence with accession 487\n",
      "Added sequence with accession 488\n",
      "Added sequence with accession 489\n",
      "Added sequence with accession 490\n",
      "Added sequence with accession 491\n",
      "Added sequence with accession 492\n",
      "Added sequence with accession 493\n",
      "Added sequence with accession 494\n",
      "Added sequence with accession 495\n",
      "Added sequence with accession 496\n",
      "Added sequence with accession 497\n",
      "Added sequence with accession 498\n",
      "Added sequence with accession 499\n",
      "Added sequence with accession 500\n",
      "Added sequence with accession 501\n",
      "Added sequence with accession 502\n",
      "Added sequence with accession 503\n",
      "Added sequence with accession 504\n",
      "Added sequence with accession 505\n",
      "Added sequence with accession 506\n",
      "Added sequence with accession 507\n",
      "Added sequence with accession 508\n",
      "Added sequence with accession 509\n",
      "Added sequence with accession 510\n",
      "Added sequence with accession 511\n",
      "Added sequence with accession 512\n",
      "Added sequence with accession 513\n",
      "Added sequence with accession 514\n",
      "Added sequence with accession 515\n",
      "Added sequence with accession 516\n",
      "Added sequence with accession 517\n",
      "Added sequence with accession 518\n",
      "Added sequence with accession 519\n",
      "Added sequence with accession 520\n",
      "Added sequence with accession 521\n",
      "Added sequence with accession 522\n",
      "Added sequence with accession 523\n",
      "Added sequence with accession 524\n",
      "Added sequence with accession 525\n",
      "Added sequence with accession 526\n",
      "Added sequence with accession 527\n",
      "Added sequence with accession 528\n",
      "Added sequence with accession 529\n",
      "Added sequence with accession 530\n",
      "Added sequence with accession 531\n",
      "Added sequence with accession 532\n",
      "Added sequence with accession 533\n",
      "Added sequence with accession 534\n",
      "Added sequence with accession 535\n",
      "Added sequence with accession 536\n",
      "Added sequence with accession 537\n",
      "Added sequence with accession 538\n",
      "Added sequence with accession 539\n",
      "Added sequence with accession 540\n",
      "Added sequence with accession 541\n",
      "Added sequence with accession 542\n",
      "Added sequence with accession 543\n",
      "Added sequence with accession 544\n",
      "Added sequence with accession 545\n",
      "Added sequence with accession 546\n",
      "Added sequence with accession 547\n",
      "Added sequence with accession 548\n",
      "Added sequence with accession 549\n",
      "Added sequence with accession 550\n",
      "Added sequence with accession 551\n",
      "Added sequence with accession 552\n",
      "Added sequence with accession 553\n",
      "Added sequence with accession 554\n",
      "Added sequence with accession 555\n",
      "Added sequence with accession 556\n",
      "Added sequence with accession 557\n",
      "Added sequence with accession 558\n",
      "Added sequence with accession 559\n",
      "Added sequence with accession 560\n",
      "Added sequence with accession 561\n",
      "Added sequence with accession 562\n",
      "Added sequence with accession 563\n",
      "Added sequence with accession 564\n",
      "Added sequence with accession 565\n",
      "Added sequence with accession 566\n",
      "Added sequence with accession 567\n",
      "Added sequence with accession 568\n",
      "Added sequence with accession 569\n",
      "Added sequence with accession 570\n",
      "Added sequence with accession 571\n",
      "cp histones.csv backups/histones.csv-May2524072945\n",
      "Previous data backuped to backups/histones.csv-May2524072945\n",
      "Results saved to histones.csv\n"
     ]
    }
   ],
   "source": [
    "cs.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".conda-histdb_env",
   "language": "python",
   "name": "conda-env-.conda-histdb_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
