{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/scratch/l_singh/hdb/project_dir/histonedb/CURATED_SET\n"
     ]
    }
   ],
   "source": [
    "%cd '..'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import inspect\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import Entrez, SeqIO\n",
    "from curated_set_services import CuratedSet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((555, 16),\n",
       " Index(['accession', 'type', 'variant_group', 'variant', 'doublet', 'gi',\n",
       "        'ncbi_gene_id', 'hgnc_gene_name', 'taxonomy_id', 'organism', 'phylum',\n",
       "        'class', 'taxonomy_group', 'info', 'references', 'sequence'],\n",
       "       dtype='object'))"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load data from histones.csv\n",
    "curated_set = CuratedSet()\n",
    "cs = curated_set\n",
    "\n",
    "cs.data.shape, cs.data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "my_columns = cs.data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(cs.has_duplicates())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NP_505463.1</th>\n",
       "      <td>NP_505463.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A_(Animals)</td>\n",
       "      <td></td>\n",
       "      <td>17562014</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>6239</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>Nematoda</td>\n",
       "      <td>Chromadorea</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26989147 22650316(?)</td>\n",
       "      <td>MSGRGKGGKAKTGGKAKSRSSRAGLQFPVGRLHRILRKGNYAQRVG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EEC09557.1</th>\n",
       "      <td>EEC09557.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A_(Animals)</td>\n",
       "      <td></td>\n",
       "      <td>215500063</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>6945</td>\n",
       "      <td>Ixodes scapularis</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Arachnida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26989147 22650316(?)</td>\n",
       "      <td>MSGRGKGGKVKGKSKTRSSRAGLQFPVGRIHRLLRKGNYAERVGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_724343.1</th>\n",
       "      <td>NP_724343.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A_(Animals)</td>\n",
       "      <td></td>\n",
       "      <td>24585673</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26989147 22650316(?)</td>\n",
       "      <td>MSGRGKGGKVKGKAKSRSNRAGLQFPVGRIHRLLRKGNYAERVGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_001119899.1</th>\n",
       "      <td>XP_001119899.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A_(Animals)</td>\n",
       "      <td></td>\n",
       "      <td>110764935</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7460</td>\n",
       "      <td>Apis mellifera</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26989147 22650316(?)</td>\n",
       "      <td>MSGRGKGGKAKAKAKSRSNRAGLQFPVGRIHRLLRKGNYAERVGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EDO48405.1</th>\n",
       "      <td>EDO48405.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A_(Animals)</td>\n",
       "      <td></td>\n",
       "      <td>156227602</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>45351</td>\n",
       "      <td>Nematostella vectensis</td>\n",
       "      <td>Cnidaria</td>\n",
       "      <td>Anthozoa</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26989147 22650316(?)</td>\n",
       "      <td>MSGRGKGKAKGTKSKTRSSRAGLQFPVGRIHRHLRKGNYAERVGAG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     accession type variant_group         variant doublet  \\\n",
       "NP_505463.1        NP_505463.1  H2A          cH2A  cH2A_(Animals)           \n",
       "EEC09557.1          EEC09557.1  H2A          cH2A  cH2A_(Animals)           \n",
       "NP_724343.1        NP_724343.1  H2A          cH2A  cH2A_(Animals)           \n",
       "XP_001119899.1  XP_001119899.1  H2A          cH2A  cH2A_(Animals)           \n",
       "EDO48405.1          EDO48405.1  H2A          cH2A  cH2A_(Animals)           \n",
       "\n",
       "                       gi ncbi_gene_id hgnc_gene_name taxonomy_id  \\\n",
       "NP_505463.1      17562014                                    6239   \n",
       "EEC09557.1      215500063                                    6945   \n",
       "NP_724343.1      24585673                                    7227   \n",
       "XP_001119899.1  110764935                                    7460   \n",
       "EDO48405.1      156227602                                   45351   \n",
       "\n",
       "                               organism      phylum        class  \\\n",
       "NP_505463.1      Caenorhabditis elegans    Nematoda  Chromadorea   \n",
       "EEC09557.1            Ixodes scapularis  Arthropoda    Arachnida   \n",
       "NP_724343.1     Drosophila melanogaster  Arthropoda      Insecta   \n",
       "XP_001119899.1           Apis mellifera  Arthropoda      Insecta   \n",
       "EDO48405.1       Nematostella vectensis    Cnidaria     Anthozoa   \n",
       "\n",
       "               taxonomy_group info            references  \\\n",
       "NP_505463.1                         26989147 22650316(?)   \n",
       "EEC09557.1                          26989147 22650316(?)   \n",
       "NP_724343.1                         26989147 22650316(?)   \n",
       "XP_001119899.1                      26989147 22650316(?)   \n",
       "EDO48405.1                          26989147 22650316(?)   \n",
       "\n",
       "                                                         sequence  \n",
       "NP_505463.1     MSGRGKGGKAKTGGKAKSRSSRAGLQFPVGRLHRILRKGNYAQRVG...  \n",
       "EEC09557.1      MSGRGKGGKVKGKSKTRSSRAGLQFPVGRIHRLLRKGNYAERVGAG...  \n",
       "NP_724343.1     MSGRGKGGKVKGKAKSRSNRAGLQFPVGRIHRLLRKGNYAERVGAG...  \n",
       "XP_001119899.1  MSGRGKGGKAKAKAKSRSNRAGLQFPVGRIHRLLRKGNYAERVGAG...  \n",
       "EDO48405.1      MSGRGKGKAKGTKSKTRSSRAGLQFPVGRIHRHLRKGNYAERVGAG...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(555, 16)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Данные были взяты из статьи DOI:10.5772/intechopen.81409. С помощью Blastp были найдены accessions и записаны в curated_service/H2BZ.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>XP_001349046.1</th>\n",
       "      <td>XP_001349046.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td></td>\n",
       "      <td>124511826</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>36329</td>\n",
       "      <td>Plasmodium falciparum 3D7</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Aconoidasida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MSGKGPAQKSQAAKKTAGKTLGPRHKRKRRTESFSLYIFKVLKQVH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_002369740.1</th>\n",
       "      <td>XP_002369740.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td></td>\n",
       "      <td>237840885</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>508771</td>\n",
       "      <td>Toxoplasma gondii ME49</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Conoidasida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MSGKGPAQKSQAAKKTAGKSLGPRYRRRKRTESFALYIYKVLKQVH...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     accession type variant_group variant doublet         gi  \\\n",
       "XP_001349046.1  XP_001349046.1  H2B         H2B.Z   H2B.Z          124511826   \n",
       "XP_002369740.1  XP_002369740.1  H2B         H2B.Z   H2B.Z          237840885   \n",
       "\n",
       "               ncbi_gene_id hgnc_gene_name taxonomy_id  \\\n",
       "XP_001349046.1                                   36329   \n",
       "XP_002369740.1                                  508771   \n",
       "\n",
       "                                 organism       phylum         class  \\\n",
       "XP_001349046.1  Plasmodium falciparum 3D7  Apicomplexa  Aconoidasida   \n",
       "XP_002369740.1     Toxoplasma gondii ME49  Apicomplexa   Conoidasida   \n",
       "\n",
       "               taxonomy_group info references  \\\n",
       "XP_001349046.1                                  \n",
       "XP_002369740.1                                  \n",
       "\n",
       "                                                         sequence  \n",
       "XP_001349046.1  MSGKGPAQKSQAAKKTAGKTLGPRHKRKRRTESFSLYIFKVLKQVH...  \n",
       "XP_002369740.1  MSGKGPAQKSQAAKKTAGKSLGPRYRRRKRTESFALYIYKVLKQVH...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h2bz_sequences = SeqIO.index(\"curated_service/H2BZ.fasta\", \"fasta\")\n",
    "cs.data.loc[cs.data.index.intersection(list(h2bz_sequences.keys())), :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MSGKGPAQKSQAAKKTAGKTLGPRHKRKRRTESFSLYIFKVLKQVHPETGVTKKSMNIMNSFINDIFDRLVTEATRLIRYNKKRTLSSREIQTAVRLLLPGELSKHAVSEGTKAVTKYTTSAA',\n",
       " 'MSGKGPAQKSQAAKKTAGKSLGPRYRRRKRTESFALYIYKVLKQVHPETGVSKKSMSIMNSFINDIFDRLADEAVRLIRYNKKRTLSSREIQTAVRLLLPGELSKHAVSEGTKAVTKYTTSGA',\n",
       " 'MSGKGPAQKSQAAKKTAGKSLGPRYRRRKRTESFALYIYKVLKQVHPETGVSKKSMSIMNSFINDIFDRLADEAVRLIRYNKKRTLSSREIQTAVRLLLPGELSKHAVSEGTKAVSKYTTSGA',\n",
       " 'MPGKGPAEKRQAAKKTAGKTPAEAGKKRRRKRTESFALYIYKVLKQVHPETGISKKSMSIMNSFINDIFDRMATEATNLIRFNKKKTLSSREVQTSVRLMLPGELSKHAVSEGTKAVTKYTTAAGN',\n",
       " 'MSGKVPSSKSQAAKKTAGKSLGIRYRRKKRIESFSLYIYKVLKQVHPETGVSKKSMSIMNSFINDIFDRMALEATRLIRYNKKSTLSSREIQTAVRLLLPGELSKHAVSEGTKAVTKYTTSGA',\n",
       " 'MSGKVPSTKSQAAKKTAGKTLGVRYRRKKRIESFALYIYKVLKQVHPETGVSKKSMSIMNSFINDIFDRLALEATRLIRYNKKSTLSSREIQTAVRLLLPGELSKHAVSEGTKAVTKYTTSGV',\n",
       " 'MSGKGGKQQLAKKTAANKLPSHHLDKNKKRRRRGETFSIYIYKVLRQVQPKLGMSRKSMAIMNSFINDIFERLATEAVKLIQYNKKRTLSSREMQTSVRLLLPGELSKHAATEGAKAVEKYENRPIA',\n",
       " 'MSGKSGKSIKGPAQKQQAAKKTAGKSPADGGKRKRRKRTESFALYIYKVLKQVHPETGISKKSMSIMNSFINDVFDRLSAEAVKLVQYNKKRTLSSREVQTSVRLMLPGELSKHAVSEGTKAVTKYTSASA']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(map(lambda x: str(x.seq), h2bz_sequences.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((8, 6),\n",
       " Index(['accession', 'type', 'variant_group', 'variant', 'references',\n",
       "        'sequence'],\n",
       "       dtype='object'))"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(\n",
    "    {\n",
    "        \"accession\": list(h2bz_sequences.keys()),\n",
    "        \"type\": [\"H2B\"] * len(h2bz_sequences),\n",
    "        \"variant_group\": [\"H2B.Z\"] * len(h2bz_sequences),\n",
    "        \"variant\": [\"H2B.Z\"] * len(h2bz_sequences),\n",
    "        \"references\": [\"DOI:10.5772/intechopen.81409\"] * len(h2bz_sequences),\n",
    "        \"sequence\": list(map(lambda x: str(x.seq), h2bz_sequences.values())),\n",
    "    }\n",
    ")\n",
    "df.index = df.accession\n",
    "# df.at[cs.data.index.intersection(list(h2bz_sequences.keys())), \"references\"] = (\n",
    "#     \"DOI:10.5772/intechopen.81409 \"\n",
    "#     + cs.data.loc[cs.data.index.intersection(list(h2bz_sequences.keys())), \"references\"]\n",
    "# )\n",
    "df.shape, df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>XP_001349046.1</th>\n",
       "      <td>XP_001349046.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MSGKGPAQKSQAAKKTAGKTLGPRHKRKRRTESFSLYIFKVLKQVH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_002369740.1</th>\n",
       "      <td>XP_002369740.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MSGKGPAQKSQAAKKTAGKSLGPRYRRRKRTESFALYIYKVLKQVH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_013228334.1</th>\n",
       "      <td>XP_013228334.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MSGKGPAQKSQAAKKTAGKSLGPRYRRRKRTESFALYIYKVLKQVH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CEM32013.1</th>\n",
       "      <td>CEM32013.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MPGKGPAEKRQAAKKTAGKTPAEAGKKRRRKRTESFALYIYKVLKQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_001610608.1</th>\n",
       "      <td>XP_001610608.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MSGKVPSSKSQAAKKTAGKSLGIRYRRKKRIESFSLYIYKVLKQVH...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     accession type variant_group variant  \\\n",
       "accession                                                   \n",
       "XP_001349046.1  XP_001349046.1  H2B         H2B.Z   H2B.Z   \n",
       "XP_002369740.1  XP_002369740.1  H2B         H2B.Z   H2B.Z   \n",
       "XP_013228334.1  XP_013228334.1  H2B         H2B.Z   H2B.Z   \n",
       "CEM32013.1          CEM32013.1  H2B         H2B.Z   H2B.Z   \n",
       "XP_001610608.1  XP_001610608.1  H2B         H2B.Z   H2B.Z   \n",
       "\n",
       "                                  references  \\\n",
       "accession                                      \n",
       "XP_001349046.1  DOI:10.5772/intechopen.81409   \n",
       "XP_002369740.1  DOI:10.5772/intechopen.81409   \n",
       "XP_013228334.1  DOI:10.5772/intechopen.81409   \n",
       "CEM32013.1      DOI:10.5772/intechopen.81409   \n",
       "XP_001610608.1  DOI:10.5772/intechopen.81409   \n",
       "\n",
       "                                                         sequence  \n",
       "accession                                                          \n",
       "XP_001349046.1  MSGKGPAQKSQAAKKTAGKTLGPRHKRKRRTESFSLYIFKVLKQVH...  \n",
       "XP_002369740.1  MSGKGPAQKSQAAKKTAGKSLGPRYRRRKRTESFALYIYKVLKQVH...  \n",
       "XP_013228334.1  MSGKGPAQKSQAAKKTAGKSLGPRYRRRKRTESFALYIYKVLKQVH...  \n",
       "CEM32013.1      MPGKGPAEKRQAAKKTAGKTPAEAGKKRRRKRTESFALYIYKVLKQ...  \n",
       "XP_001610608.1  MSGKVPSSKSQAAKKTAGKSLGIRYRRKKRIESFSLYIYKVLKQVH...  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6, 6)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# cs.data = cs.data[cs.data['accession'] != cs.data.index.intersection(g_h2b_accessions).values[0]]\n",
    "# cs.data.shape\n",
    "df = df[~df['accession'].isin(cs.data.index.intersection(list(h2bz_sequences.keys())).values)]\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>XP_013228334.1</th>\n",
       "      <td>XP_013228334.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MSGKGPAQKSQAAKKTAGKSLGPRYRRRKRTESFALYIYKVLKQVH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CEM32013.1</th>\n",
       "      <td>CEM32013.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MPGKGPAEKRQAAKKTAGKTPAEAGKKRRRKRTESFALYIYKVLKQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_001610608.1</th>\n",
       "      <td>XP_001610608.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MSGKVPSSKSQAAKKTAGKSLGIRYRRKKRIESFSLYIYKVLKQVH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HISTDB_H2B_Z_0</th>\n",
       "      <td>HISTDB_H2B_Z_0</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MSGKVPSTKSQAAKKTAGKTLGVRYRRKKRIESFALYIYKVLKQVH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_011128492.1</th>\n",
       "      <td>XP_011128492.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MSGKGGKQQLAKKTAANKLPSHHLDKNKKRRRRGETFSIYIYKVLR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_628349.1</th>\n",
       "      <td>XP_628349.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MSGKSGKSIKGPAQKQQAAKKTAGKSPADGGKRKRRKRTESFALYI...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     accession type variant_group variant  \\\n",
       "accession                                                   \n",
       "XP_013228334.1  XP_013228334.1  H2B         H2B.Z   H2B.Z   \n",
       "CEM32013.1          CEM32013.1  H2B         H2B.Z   H2B.Z   \n",
       "XP_001610608.1  XP_001610608.1  H2B         H2B.Z   H2B.Z   \n",
       "HISTDB_H2B_Z_0  HISTDB_H2B_Z_0  H2B         H2B.Z   H2B.Z   \n",
       "XP_011128492.1  XP_011128492.1  H2B         H2B.Z   H2B.Z   \n",
       "XP_628349.1        XP_628349.1  H2B         H2B.Z   H2B.Z   \n",
       "\n",
       "                                  references  \\\n",
       "accession                                      \n",
       "XP_013228334.1  DOI:10.5772/intechopen.81409   \n",
       "CEM32013.1      DOI:10.5772/intechopen.81409   \n",
       "XP_001610608.1  DOI:10.5772/intechopen.81409   \n",
       "HISTDB_H2B_Z_0  DOI:10.5772/intechopen.81409   \n",
       "XP_011128492.1  DOI:10.5772/intechopen.81409   \n",
       "XP_628349.1     DOI:10.5772/intechopen.81409   \n",
       "\n",
       "                                                         sequence  \n",
       "accession                                                          \n",
       "XP_013228334.1  MSGKGPAQKSQAAKKTAGKSLGPRYRRRKRTESFALYIYKVLKQVH...  \n",
       "CEM32013.1      MPGKGPAEKRQAAKKTAGKTPAEAGKKRRRKRTESFALYIYKVLKQ...  \n",
       "XP_001610608.1  MSGKVPSSKSQAAKKTAGKSLGIRYRRKKRIESFSLYIYKVLKQVH...  \n",
       "HISTDB_H2B_Z_0  MSGKVPSTKSQAAKKTAGKTLGVRYRRKKRIESFALYIYKVLKQVH...  \n",
       "XP_011128492.1  MSGKGGKQQLAKKTAANKLPSHHLDKNKKRRRRGETFSIYIYKVLR...  \n",
       "XP_628349.1     MSGKSGKSIKGPAQKQQAAKKTAGKSPADGGKRKRRKRTESFALYI...  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((561, 16), [])"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data = pd.concat([cs.data, df]).fillna(\"\")\n",
    "cs.data.shape, list(cs.has_duplicates())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CEM32013.1</th>\n",
       "      <td>CEM32013.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MPGKGPAEKRQAAKKTAGKTPAEAGKKRRRKRTESFALYIYKVLKQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_001610608.1</th>\n",
       "      <td>XP_001610608.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MSGKVPSSKSQAAKKTAGKSLGIRYRRKKRIESFSLYIYKVLKQVH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HISTDB_H2B_Z_0</th>\n",
       "      <td>HISTDB_H2B_Z_0</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MSGKVPSTKSQAAKKTAGKTLGVRYRRKKRIESFALYIYKVLKQVH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_011128492.1</th>\n",
       "      <td>XP_011128492.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MSGKGGKQQLAKKTAANKLPSHHLDKNKKRRRRGETFSIYIYKVLR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_628349.1</th>\n",
       "      <td>XP_628349.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MSGKSGKSIKGPAQKQQAAKKTAGKSPADGGKRKRRKRTESFALYI...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     accession type variant_group variant doublet gi  \\\n",
       "CEM32013.1          CEM32013.1  H2B         H2B.Z   H2B.Z              \n",
       "XP_001610608.1  XP_001610608.1  H2B         H2B.Z   H2B.Z              \n",
       "HISTDB_H2B_Z_0  HISTDB_H2B_Z_0  H2B         H2B.Z   H2B.Z              \n",
       "XP_011128492.1  XP_011128492.1  H2B         H2B.Z   H2B.Z              \n",
       "XP_628349.1        XP_628349.1  H2B         H2B.Z   H2B.Z              \n",
       "\n",
       "               ncbi_gene_id hgnc_gene_name taxonomy_id organism phylum class  \\\n",
       "CEM32013.1                                                                     \n",
       "XP_001610608.1                                                                 \n",
       "HISTDB_H2B_Z_0                                                                 \n",
       "XP_011128492.1                                                                 \n",
       "XP_628349.1                                                                    \n",
       "\n",
       "               taxonomy_group info                    references  \\\n",
       "CEM32013.1                          DOI:10.5772/intechopen.81409   \n",
       "XP_001610608.1                      DOI:10.5772/intechopen.81409   \n",
       "HISTDB_H2B_Z_0                      DOI:10.5772/intechopen.81409   \n",
       "XP_011128492.1                      DOI:10.5772/intechopen.81409   \n",
       "XP_628349.1                         DOI:10.5772/intechopen.81409   \n",
       "\n",
       "                                                         sequence  \n",
       "CEM32013.1      MPGKGPAEKRQAAKKTAGKTPAEAGKKRRRKRTESFALYIYKVLKQ...  \n",
       "XP_001610608.1  MSGKVPSSKSQAAKKTAGKSLGIRYRRKKRIESFSLYIYKVLKQVH...  \n",
       "HISTDB_H2B_Z_0  MSGKVPSTKSQAAKKTAGKTLGVRYRRKKRIESFALYIYKVLKQVH...  \n",
       "XP_011128492.1  MSGKGGKQQLAKKTAANKLPSHHLDKNKKRRRRGETFSIYIYKVLR...  \n",
       "XP_628349.1     MSGKSGKSIKGPAQKQQAAKKTAGKSPADGGKRKRRKRTESFALYI...  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9, 16)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data[cs.data[\"variant\"] == \"H2B.Z\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/l_singh/.conda/envs/histdb_env/lib/python3.8/site-packages/Bio/Entrez/__init__.py:658: UserWarning: \n",
      "Email address is not specified.\n",
      "\n",
      "To make use of NCBI's E-utilities, NCBI requires you to specify your\n",
      "email address with each request.  As an example, if your email address\n",
      "is A.N.Other@example.com, you can specify it as follows:\n",
      "   from Bio import Entrez\n",
      "   Entrez.email = 'A.N.Other@example.com'\n",
      "In case of excessive usage of the E-utilities, NCBI will attempt to contact\n",
      "a user at the email address provided before blocking access to the\n",
      "E-utilities.\n",
      "  warnings.warn(\n",
      "/home/l_singh/.conda/envs/histdb_env/lib/python3.8/site-packages/Bio/GenBank/__init__.py:1143: BiopythonParserWarning: Dropping bond qualifier in feature location\n",
      "  warnings.warn(\n",
      "/mnt/scratch/l_singh/hdb/project_dir/histonedb/CURATED_SET/curated_set_services.py:267: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  updating_data['accession'] = new_accessions\n"
     ]
    }
   ],
   "source": [
    "cs.update_accession_version()\n",
    "cs.data = cs.data.set_index(cs.data.accession.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9, 16)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data[cs.data[\"variant\"] == \"H2B.Z\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 5802\n",
      "Fetched taxid from NCBI 1169540\n",
      "Fetched taxid from NCBI 484906\n",
      "Fetched taxid from NCBI 110365\n",
      "Fetched taxid from NCBI 353152\n",
      " changes to 5802\n",
      " changes to Eimeria tenella\n",
      " changes to Apicomplexa\n",
      " changes to Conoidasida\n",
      " changes to 1169540\n",
      " changes to Vitrella brassicaformis CCMP3155\n",
      " changes to None\n",
      " changes to None\n",
      " changes to 484906\n",
      " changes to Babesia bovis T2Bo\n",
      " changes to Apicomplexa\n",
      " changes to Aconoidasida\n",
      " changes to 110365\n",
      " changes to Gregarina niphandrodes\n",
      " changes to Apicomplexa\n",
      " changes to Conoidasida\n",
      " changes to 353152\n",
      " changes to Cryptosporidium parvum Iowa II\n",
      " changes to Apicomplexa\n",
      " changes to Conoidasida\n"
     ]
    }
   ],
   "source": [
    "curated_set.update_taxids(blank_data=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9, 16)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data[cs.data[\"variant\"] == \"H2B.Z\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>XP_013228334.1</th>\n",
       "      <td>XP_013228334.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5802</td>\n",
       "      <td>Eimeria tenella</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Conoidasida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MSGKGPAQKSQAAKKTAGKSLGPRYRRRKRTESFALYIYKVLKQVH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CEM32013.1</th>\n",
       "      <td>CEM32013.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1169540</td>\n",
       "      <td>Vitrella brassicaformis CCMP3155</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MPGKGPAEKRQAAKKTAGKTPAEAGKKRRRKRTESFALYIYKVLKQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_001610608.1</th>\n",
       "      <td>XP_001610608.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>484906</td>\n",
       "      <td>Babesia bovis T2Bo</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Aconoidasida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MSGKVPSSKSQAAKKTAGKSLGIRYRRKKRIESFSLYIYKVLKQVH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_011128492.1</th>\n",
       "      <td>XP_011128492.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>110365</td>\n",
       "      <td>Gregarina niphandrodes</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Conoidasida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MSGKGGKQQLAKKTAANKLPSHHLDKNKKRRRRGETFSIYIYKVLR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_628349.1</th>\n",
       "      <td>XP_628349.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>353152</td>\n",
       "      <td>Cryptosporidium parvum Iowa II</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Conoidasida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MSGKSGKSIKGPAQKQQAAKKTAGKSPADGGKRKRRKRTESFALYI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_678689.1</th>\n",
       "      <td>XP_678689.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td></td>\n",
       "      <td>68073549</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5823</td>\n",
       "      <td>Plasmodium berghei ANKA</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Aconoidasida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MSGKGPAQKSQAAKKTAGKTLGPRHKRKRRTESFSLYIFKVLKQVH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_001349046.1</th>\n",
       "      <td>XP_001349046.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td></td>\n",
       "      <td>124511826</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>36329</td>\n",
       "      <td>Plasmodium falciparum 3D7</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Aconoidasida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MSGKGPAQKSQAAKKTAGKTLGPRHKRKRRTESFSLYIFKVLKQVH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_002369740.1</th>\n",
       "      <td>XP_002369740.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td></td>\n",
       "      <td>237840885</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>508771</td>\n",
       "      <td>Toxoplasma gondii ME49</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Conoidasida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MSGKGPAQKSQAAKKTAGKSLGPRYRRRKRTESFALYIYKVLKQVH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HISTDB_H2B_Z_0</th>\n",
       "      <td>HISTDB_H2B_Z_0</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td>H2B.Z</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>DOI:10.5772/intechopen.81409</td>\n",
       "      <td>MSGKVPSTKSQAAKKTAGKTLGVRYRRKKRIESFALYIYKVLKQVH...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     accession type variant_group variant doublet         gi  \\\n",
       "XP_013228334.1  XP_013228334.1  H2B         H2B.Z   H2B.Z                      \n",
       "CEM32013.1          CEM32013.1  H2B         H2B.Z   H2B.Z                      \n",
       "XP_001610608.1  XP_001610608.1  H2B         H2B.Z   H2B.Z                      \n",
       "XP_011128492.1  XP_011128492.1  H2B         H2B.Z   H2B.Z                      \n",
       "XP_628349.1        XP_628349.1  H2B         H2B.Z   H2B.Z                      \n",
       "XP_678689.1        XP_678689.1  H2B         H2B.Z   H2B.Z           68073549   \n",
       "XP_001349046.1  XP_001349046.1  H2B         H2B.Z   H2B.Z          124511826   \n",
       "XP_002369740.1  XP_002369740.1  H2B         H2B.Z   H2B.Z          237840885   \n",
       "HISTDB_H2B_Z_0  HISTDB_H2B_Z_0  H2B         H2B.Z   H2B.Z                      \n",
       "\n",
       "               ncbi_gene_id hgnc_gene_name taxonomy_id  \\\n",
       "XP_013228334.1                                    5802   \n",
       "CEM32013.1                                     1169540   \n",
       "XP_001610608.1                                  484906   \n",
       "XP_011128492.1                                  110365   \n",
       "XP_628349.1                                     353152   \n",
       "XP_678689.1                                       5823   \n",
       "XP_001349046.1                                   36329   \n",
       "XP_002369740.1                                  508771   \n",
       "HISTDB_H2B_Z_0                                           \n",
       "\n",
       "                                        organism       phylum         class  \\\n",
       "XP_013228334.1                   Eimeria tenella  Apicomplexa   Conoidasida   \n",
       "CEM32013.1      Vitrella brassicaformis CCMP3155         None          None   \n",
       "XP_001610608.1                Babesia bovis T2Bo  Apicomplexa  Aconoidasida   \n",
       "XP_011128492.1            Gregarina niphandrodes  Apicomplexa   Conoidasida   \n",
       "XP_628349.1       Cryptosporidium parvum Iowa II  Apicomplexa   Conoidasida   \n",
       "XP_678689.1              Plasmodium berghei ANKA  Apicomplexa  Aconoidasida   \n",
       "XP_001349046.1         Plasmodium falciparum 3D7  Apicomplexa  Aconoidasida   \n",
       "XP_002369740.1            Toxoplasma gondii ME49  Apicomplexa   Conoidasida   \n",
       "HISTDB_H2B_Z_0                                                                \n",
       "\n",
       "               taxonomy_group info                    references  \\\n",
       "XP_013228334.1                      DOI:10.5772/intechopen.81409   \n",
       "CEM32013.1                          DOI:10.5772/intechopen.81409   \n",
       "XP_001610608.1                      DOI:10.5772/intechopen.81409   \n",
       "XP_011128492.1                      DOI:10.5772/intechopen.81409   \n",
       "XP_628349.1                         DOI:10.5772/intechopen.81409   \n",
       "XP_678689.1                                                        \n",
       "XP_001349046.1                                                     \n",
       "XP_002369740.1                                                     \n",
       "HISTDB_H2B_Z_0                      DOI:10.5772/intechopen.81409   \n",
       "\n",
       "                                                         sequence  \n",
       "XP_013228334.1  MSGKGPAQKSQAAKKTAGKSLGPRYRRRKRTESFALYIYKVLKQVH...  \n",
       "CEM32013.1      MPGKGPAEKRQAAKKTAGKTPAEAGKKRRRKRTESFALYIYKVLKQ...  \n",
       "XP_001610608.1  MSGKVPSSKSQAAKKTAGKSLGIRYRRKKRIESFSLYIYKVLKQVH...  \n",
       "XP_011128492.1  MSGKGGKQQLAKKTAANKLPSHHLDKNKKRRRRGETFSIYIYKVLR...  \n",
       "XP_628349.1     MSGKSGKSIKGPAQKQQAAKKTAGKSPADGGKRKRRKRTESFALYI...  \n",
       "XP_678689.1     MSGKGPAQKSQAAKKTAGKTLGPRHKRKRRTESFSLYIFKVLKQVH...  \n",
       "XP_001349046.1  MSGKGPAQKSQAAKKTAGKTLGPRHKRKRRTESFSLYIFKVLKQVH...  \n",
       "XP_002369740.1  MSGKGPAQKSQAAKKTAGKSLGPRYRRRKRTESFALYIYKVLKQVH...  \n",
       "HISTDB_H2B_Z_0  MSGKVPSTKSQAAKKTAGKTLGVRYRRKKRIESFALYIYKVLKQVH...  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data[cs.data[\"variant\"] == \"H2B.Z\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Added sequence with accession XP_001610608.1\n",
      "Added sequence with accession HISTDB_H2B_Z_0\n",
      "Added sequence with accession XP_011128492.1\n",
      "Added sequence with accession XP_013228334.1\n",
      "Added sequence with accession XP_628349.1\n",
      "Added sequence with accession CEM32013.1\n",
      "cp histones.csv backups/histones.csv-Mar0624162721\n",
      "Previous data backuped to backups/histones.csv-Mar0624162721\n",
      "Results saved to histones.csv\n"
     ]
    }
   ],
   "source": [
    "cs.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".conda-histdb_env",
   "language": "python",
   "name": "conda-env-.conda-histdb_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
