{% extends "base.html" %}
{% load static %}

{% block includes %}
<script src="https://d3js.org/d3.v7.min.js"></script>
<!-- <script src="../js/msa.min.js"></script> -->
<script src="../CURATED_SET/js/msa.min.js"></script>
<script type="text/javascript">
  $(function() { 
    $('#browse_navbar_button').addClass('active');
  });
</script>
<script >

  //polyfill for at
  function at(n) {
      // ToInteger() abstract op
      n = Math.trunc(n) || 0;
      // Allow negative indexing from the end
      if (n < 0) n += this.length;
      // OOB access is guaranteed to return undefined
      if (n < 0 || n >= this.length) return undefined;
      // Otherwise, this is just normal property access
      return this[n];
  }
  
  const TypedArray = Reflect.getPrototypeOf(Int8Array);
  for (const C of [Array, String, TypedArray]) {
      Object.defineProperty(C.prototype, "at",
                            { value: at,
                              writable: true,
                              enumerable: false,
                              configurable: true });
  }
  //////
  
  
  d3.json("https://intbio.org/histonedb/CURATED_SET/classification.json").then(function(data) {
      
  //console.log(data);
  
  var datah={"name": "root"};
  
  function dict2tree(tree,d)
  {
  tree["children"]=[]
  
  for (var ind in d)
  {
  tree["children"].push({"name":ind})
  if (typeof d[ind]  === "object") { 
  dict2tree(tree["children"].at(-1),d[ind])}
  }
  }
  
  dict2tree(datah,data['tree']);
  
  const margin = { top: 10, right: 120, bottom: 10, left: 40 };
  const width = d3.width || 960;
  const root = d3.hierarchy(datah);
  const dx = 10;
  const dy = width / 8;
  const tree = d3.tree().nodeSize([dx, dy]);
  const diagonal = d3
    .linkHorizontal()
    .x((d) => d.y)
    .y((d) => d.x);
  
  root.x0 = dy / 2;
  root.y0 = 0;
  root.descendants().forEach((d, i) => {
    d.id = i;
    d._children = d.children;
    // if (d.depth && d.data.name.length !== 7) d.children = null;
  });
  
  tree(root);
  
  const svg = d3
    .create("svg")
    .attr("viewBox", [-margin.left, -margin.top, width, dx])
    .style("font", "10px sans-serif")
    .style("user-select", "none");
  
  const gLink = svg
    .append("g")
    .attr("fill", "none")
    .attr("stroke", "#555")
    .attr("stroke-opacity", 0.4)
    .attr("stroke-width", 1.5);
  
  const gNode = svg
    .append("g")
    .attr("cursor", "pointer")
    .attr("pointer-events", "all");
  
  update(root);
  
  document.querySelector("#tree").appendChild(svg.node());
  
  function update(source) {
    const duration = d3.event && d3.event.altKey ? 2500 : 250;
     //const duration = 10000;
    const nodes = root.descendants().reverse();
    const links = root.links();
  
    // Compute the new tree layout.
    tree(root);
    
    let left = root;
    let right = root;
    root.eachBefore((node) => {
      if (node.x < left.x) left = node;
      if (node.x > right.x) right = node;
    });
  
    const height = right.x - left.x + margin.top + margin.bottom;
  
    const transition = svg
      .transition()
      .duration(duration)
      .attr("viewBox", [-margin.left, left.x - margin.top, width, height])
      .tween(
        "resize",
        window.ResizeObserver ? null : () => () => svg.dispatch("toggle")
      );
  
    // Update the nodes…
    const node = gNode.selectAll("g").data(nodes, (d) => d.id);
  
    // Enter any new nodes at the parent's previous position.
    const nodeEnter = node
      .enter()
      .append("g")
      .attr("transform", (d) => `translate(${source.y0},${source.x0})`)
      .attr("fill-opacity", 0)
      .attr("stroke-opacity", 0)
     // .on("click", (event, d) => {
      //  d.children = d.children ? null : d._children;
      //  update(d);
      //})
      ;
  
  
  
  function histcolor(d) {
  let r = 0xF, g=0xF, b=0xF;
  if(d.data.name.includes('H3')) { r= 0; g=0; b = 0xF ;}
  if(d.data.name.includes('H4')) { r= 0; g=0xF; b = 0 ;}
  if(d.data.name.includes('H2A')) { r= 0xF; g=0xF; b = 0 ;}
  if(d.data.name.includes('H2B')) { r= 0xF; g=0; b = 0 ;}
  if(d.data.name.includes('H1')) { r= 0xF; g=0; b = 0xF ;}
  if(d.data.name.includes('Arch')) { r= 0; g=0xF; b = 0xF ;}
  return d._children ? "#"+(r & 0x5).toString(16)+(g& 0x5).toString(16)+(b& 0x5).toString(16) :  "#"+(r& 0x9).toString(16)+(g& 0x9).toString(16)+(b& 0x9).toString(16)
  }
  
    nodeEnter
      .append("circle")
      .attr("r", 2.5)
      .attr("fill", (d) => histcolor(d) )
      .attr("stroke-width", 10)
      .on("click", (event, d) => {
      d.children = d.children ? null : d._children;
      update(d);
      })
      ;
  
    nodeEnter
      .append("a")
      .attr("xlink:href", (d)=>"javascript:upd_msa(\""+d.data.name+"\")")
      .append("text")
      .attr("dy", "0.31em")
      .attr("x", (d) => (d._children ? -6 : 6))
      .attr("text-anchor", (d) => (d._children ? "end" : "start"))
      .text((d) => d.data.name)
      .clone(true)
      .lower()
      .attr("stroke-linejoin", "round")
      .attr("stroke-width", 3)
      .attr("stroke", "white");
  
    // Transition nodes to their new position.
    node
      .merge(nodeEnter)
      .transition(transition)
      .attr("transform", (d) => `translate(${d.y},${d.x})`)
      .attr("fill-opacity", 1)
      .attr("stroke-opacity", 1);
  
    // Transition exiting nodes to the parent's new position.
    node
      .exit()
      .transition(transition)
      .remove()
      .attr("transform", (d) => `translate(${source.y},${source.x})`)
      .attr("fill-opacity", 0)
      .attr("stroke-opacity", 0);
  
    // Update the links…
    const link = gLink.selectAll("path").data(links, (d) => d.target.id);
  
    // Enter any new links at the parent's previous position.
    const linkEnter = link
      .enter()
      .append("path")
      .attr("stroke", (d) => histcolor(d.source))
      .attr("d", (d) => {
        const o = { x: source.x0, y: source.y0 };
        return diagonal({ source: o, target: o });
      });
  
    // Transition links to their new position.
    link.merge(linkEnter).transition(transition).attr("d", diagonal);
  
    // Transition exiting nodes to the parent's new position.
    link
      .exit()
      .transition(transition)
      .remove()
      .attr("d", (d) => {
        const o = { x: source.x, y: source.y };
        return diagonal({ source: o, target: o });
      });
  
    // Stash the old positions for transition.
    root.eachBefore((d) => {
      d.x0 = d.x;
      d.y0 = d.y;
    });
  }
   });
  
  var rootDiv = document.getElementById("msa");
  var menuDiv = document.getElementById("menu");
  
  
  
  
  function upd_msa(name) {
  
  // var seqs;
  
  
  // var xhr = msa.io.xhr;
  // var gffParser = msa.io.gff;
  
  // xhr("../seeds/"+name+".gff3", function(err, request, body) {
  //   var features = gffParser.parseSeqs(body);
  //   m.seqs.addFeatures(features);
  
  // xhr("../seeds/"+name+".fasta", function(err, request, body) {
  // seqs =  msa.io.fasta.parse(body);
  // m.seqs.add(seqs);
  // m.seqs.add(seq2);
  // m.render();
  // });
  
  // });
  
  
  
  
  
  //Add consensus
  
  
   var m = msa({
    el: rootDiv,
   importURL: "../seeds/"+name+".fasta",
   // seqs:seqs,
    all: true,
    vis: {
      labelId: false,
      conserv: false,
      seqlogo: false //should be false, we trigger it below manually.
    },
    zoomer: {
    labelNameLength:350,
    alignmentHeight:  "auto" 
  },
     
     conf: {
      debug: false,
      dropImport: true//,
     // manualRendering:true
    }
  //colorscheme: {scheme: "pid"}
  });
  
  
  
  //m.g.zoomer.set("alignmentHeight", 5);
  
  var fun={}
  
  fun.init = function(){
    // you have here access to the conservation or the sequence object
    this.cons = this.opt.conservation();
    this.consensus= this.opt.consensus();
    this.frequency= this.opt.frequency();
    this.consensus80="";
    this.frequency.forEach((it,i) => {
      let aa="*";
      for(let i2 in it) {
       if (it[i2] >=0.8) {
        aa=i2;}
      }
      this.consensus80+=aa;
    });
  
    this.consensus50="";
    this.frequency.forEach((it,i) => {
      let aa="*";
      for(let i2 in it) {
       if (it[i2] >=0.5) {
        aa=i2;}
      }
      this.consensus50+=aa;
    });
  }
  
  function aasimilar(aa1,aa2) {
  const groups = {pos:"KRH", neg:"ED", aliph:"VLAIM",oh:"TS", small:"SAG",aroma:"FWY"}
  
  for(let i in groups) {
    if(groups[i].includes(aa1) && groups[i].includes(aa2)) {return true;}
  }
  return false;
  }
  
  fun.run = function(letter,opts){
  
    if(letter == "-") {return "#fff";}
    if (letter == this.consensus80[opts.pos]) {return "#74b";}
    if (letter == this.consensus50[opts.pos]) {return "#08c";} 
    if (aasimilar(letter,this.consensus80[opts.pos]) || aasimilar(letter,this.consensus50[opts.pos]) ) {return "#f09";}
    else {return "#fff";} 
  };
  
  m.g.colorscheme.addDynScheme("fscheme", fun);
  m.g.colorscheme.set("scheme", "fscheme");
  //m.render();
  
  
  
  
  // the menu is independent to the MSA container
  var menuOpts = {};
  menuOpts.el = menuDiv;
  menuOpts.msa = m;
  //menuOpts.menu = "small";
  var defMenu = new msa.menu.defaultmenu(menuOpts);
  m.addView("menu", defMenu);
  
  var xhr = msa.io.xhr;
  var gffParser = msa.io.gff;
  
  //Now we will add consensus with features
  //first we need to get features and load them.
  xhr("../seeds/"+name+".gff3", function(err, request, body) {
    var features = gffParser.parseSeqs(body);
    m.seqs.addFeatures(features);
  
  //Nex in this asyn call we will wait till msa has loaded all sequences and add consensus
  waitForSeqs();
  
      });
  
  function waitForSeqs(){
      if(m.seqs.length > 0){
          //variable exists, do what you want
          var seqc = {
            seq: m.g.stats._consensus,
            id: "10c",
            ids: ["Consensus"], //This is important for tree not to hide it. Tree shoud have a root node named consensus.
            name: "Consensus",
            label:"Consensus"
          }
          m.seqs.unshift(seqc);
    //After that we add a tree
   m.importURL("../seeds/"+name+".tree", function(){
   console.log("tree loaded");
   //Trigger SeqLogo from menu, see msa.js:24482 why we need to do this hack!
   console.log(defMenu.views["40_vis"]._nodes[3].callback());
  });
    //m.render();
   
  }
      else{
          setTimeout(waitForSeqs, 100);
      }
  }
  
  // call render at the end to display the whole MSA
  //m.render();
  
  
  
  
  
  
  
  }
  
  
  </script>
{% endblock %}

{% block content %}
<div class="jumbotron" style="padding: 0px 0;">
  <div class="container">
    <h1 class="text-center">Histone variation in eukaryotes</h1>
    <p class="text-center">Classification tree of histones and their variants</p>
  </div>
</div>
<div class="container">
  <h3>Click on node to collapse/expand branches. Click on node label to view info, sequence alignment and phylogeny. </h3>
  <div id="tree"></div>
  <h1>Multiple sequence alignment will appear below</h1>
  <div id="menu"></div>
  <div id="msa"></div>
  <!-- <div class="col-sm-3">
    
  </div>
  <div class="col-sm-9">

  </div> -->
</div>

{#<div class="container-fluid img-responsive">#}
{#  <div class="row">#}
{#    <div class="col-xs-4">#}
{#      <a href="{% url 'browse:browse_variants' 'H3' %}">#}
{#        <div class="row">#}
{#          <div class="col-xs-1 col-md-offset-9"><h2>H3</h2></div>#}
{#        </div>#}
{#        <div class="row">#}
{#          <div class="col-xs-12">#}
{#            <img class="img-responsive pull-right" src="{% static 'browse/img/H3_front.png' %}" width="35%" />#}
{#          </div>#}
{#        </div>#}
{#      </a>#}
{#      <a href="{% url 'browse:browse_variants' 'H2A' %}">#}
{#        <div class="row">#}
{#          <div class="col-xs-1 col-md-offset-9"><h2>H2A</h2></div>#}
{#        </div>#}
{#        <div class="row">#}
{#          <div class="col-xs-12"><img class="img-responsive pull-right" src="{% static 'browse/img/H2A_front.png' %}" width="35%" /></div>#}
{#        </div>#}
{#      </a>#}
{#    </div>#}
{#    <div class="col-xs-4">#}
{#      <div class="vcenter center-block" style="position: relative; left: 0; top: 0;">#}
{#        <img class="img-responsive center-block" src="{% static 'browse/img/nucleosome.png' %}" style="position: relative; top: 0; left: 0;"/>#}
{#        <a href="{% url 'browse:browse_variants' 'H1' %}">#}
{#        	<div style="position: absolute; top: -40px; left: 50%;"><h3>H1</h3></div>#}
{#	        <div style="position: absolute; top: 0px; left: 40%;"><img class="img-responsive" src="{% static 'browse/img/H1.png' %}" width="50%"></div>#}
{#	    </a>#}
{#        #}
{#        The Nucleosome Core Particle (PDB: <a href="http://www.ncbi.nlm.nih.gov/Structure/mmdb/mmdbsrv.cgi?uid=1AOI">1AOI</a>) is comprised of H2A, H2B, H3, and H4 core histones types, wrapping ~147 bp of DNA. H2A-H2B form a dimer and H3-H4 form a tetramer.#}
{#      </div>#}
{#    </div>#}
{#    <div class="col-xs-4">#}
{#      <div class="row">#}
{#        <a href="{% url 'browse:browse_variants' 'H4' %}">#}
{#          <div class="row">#}
{#            <div class="col-xs-1 col-md-offset-2"><h2>H4</h2></div>#}
{#          </div>#}
{#          <div class="row">#}
{#            <div class="col-xs-12">#}
{#              <img class="img-responsive pull-left" src="{% static 'browse/img/H4_front.png' %}" width="35%" />#}
{#            </div>#}
{#          </div>#}
{#        </a>#}
{#        <a href="{% url 'browse:browse_variants' 'H2B' %}">#}
{#          <div class="row">#}
{#            <div class="col-xs-1 col-md-offset-2"><h2>H2B</h2></div>#}
{#          </div>#}
{#          <div class="row">#}
{#            <div class="col-xs-12">#}
{#              <img class="img-responsive pull-left" src="{% static 'browse/img/H2B_front.png' %}" width="35%" />#}
{#            </div>#}
{#          </div>#}
{#        </a>#}
{#      </div>#}
{#    </div>#}
{#  </div>#}
{#</div>#}
{% endblock %}
